#R code for Stem Cell Hierarchy manuscript version 3

# remotes::install_version("spatstat", version = "1.64-1")
# remotes::install_version("Seurat", version = "3.2.3")

library(dplyr)
library(Seurat)
library(patchwork)
library(reshape2)
library(gplots)
require(gtools)
library(psych)
library(RColorBrewer)
library(mclust)
library(factoextra)
library(ggrepel)
library(GGally)
library(Hmisc)
library(corrplot)
library(network)
library(sna)
library(intergraph)
library(destiny)
library(ggbiplot)
library(monocle3)
library(viridis)
library(garnett)
library(org.Mm.eg.db)
library(Matrix)
library(igraph)
library(intergraph)
library(ggVennDiagram)
library(stringr)
library(topGO)
library(WGCNA)
library("openxlsx")
library(lsmeans)
library(emmeans)
library(multcomp)
library(ggcorrplot)
library(Category)
library("ggpubr")
library(scran)
library(CellChat)
library(ggplot2)                  
library(patchwork)
library(ggalluvial)
library(igraph)
library(dplyr)
options(stringsAsFactors = FALSE)
library(limma)
library(xgboost)
library(caTools)

options(future.globals.maxSize = 20000 * 1024^2) #CC uses this to bypass vector memory limits
#plan("multiprocess", workers = 4)

`%!in%` <- Negate(`%in%`) #Define this operator

multiplot <- function(..., plotlist=NULL, file, cols=1, layout=NULL) {
  library(grid)
  
  # Make a list from the ... arguments and plotlist
  plots <- c(list(...), plotlist)
  
  numPlots = length(plots)
  
  # If layout is NULL, then use 'cols' to determine layout
  if (is.null(layout)) {
    # Make the panel
    # ncol: Number of columns of plots
    # nrow: Number of rows needed, calculated from # of cols
    layout <- matrix(seq(1, cols * ceiling(numPlots/cols)),
                     ncol = cols, nrow = ceiling(numPlots/cols))
  }
  
  if (numPlots==1) {
    print(plots[[1]])
    
  } else {
    # Set up the page
    grid.newpage()
    pushViewport(viewport(layout = grid.layout(nrow(layout), ncol(layout))))
    
    # Make each plot, in the correct location
    for (i in 1:numPlots) {
      # Get the i,j matrix positions of the regions that contain this subplot
      matchidx <- as.data.frame(which(layout == i, arr.ind = TRUE))
      
      print(plots[[i]], vp = viewport(layout.pos.row = matchidx$row,
                                      layout.pos.col = matchidx$col))
    }
  }
}

GM_state <- function(cds){
  if (length(unique(pData(cds)$State)) > 1){
    T0_counts <- table(pData(cds)$State, pData(cds)$Hours)[,"0"]
    return(as.numeric(names(T0_counts)[which
                                       (T0_counts == max(T0_counts))]))
  } else {
    return (1)
  }
}

flattenCorrMatrix <- function(cormat, pmat) {
  ut <- upper.tri(cormat)
  data.frame(
    row = rownames(cormat)[row(cormat)[ut]],
    column = rownames(cormat)[col(cormat)[ut]],
    cor  =(cormat)[ut],
    p = pmat[ut]
  )
}

format_summary = function(x) {
  output = paste(round(x[1],0)," (",round(x[2],0),", ",round(x[3],0),")",sep="")
  return(output)
}

cbind.fill <- function(...){
  nm <- list(...) 
  nm <- lapply(nm, as.matrix)
  n <- max(sapply(nm, nrow)) 
  do.call(cbind, lapply(nm, function (x) 
    rbind(x, matrix(, n-nrow(x), ncol(x))))) 
}

#summary function
summarySE <- function(data=NULL, measurevar, groupvars=NULL, na.rm=FALSE,
                      conf.interval=.95, .drop=TRUE) {
  require(plyr)
  
  # New version of length which can handle NA's: if na.rm==T, don't count them
  length2 <- function (x, na.rm=FALSE) {
    if (na.rm) sum(!is.na(x))
    else       length(x)
  }
  
  # This does the summary. For each group's data frame, return a vector with
  # N, mean, and sd
  datac <- ddply(data, groupvars, .drop=.drop,
                 .fun = function(xx, col) {
                   c(N    = length(na.omit(xx[[col]])),
                     mean = mean   (na.omit(xx[[col]])),
                     sd   = sd     (na.omit(xx[[col]]))
                   )
                 },
                 measurevar
  )
  
  # Rename the "mean" column    
  datac <- rename(datac, c("mean" = measurevar))
  
  datac$se <- datac$sd / sqrt(datac$N)  # Calculate standard error of the mean
  
  # Confidence interval multiplier for standard error
  # Calculate t-statistic for confidence interval: 
  # e.g., if conf.interval is .95, use .975 (above/below), and use df=N-1
  ciMult <- qt(conf.interval/2 + .5, datac$N-1)
  datac$ci <- datac$se * ciMult
  
  return(datac)
}

# Basic function to convert human to mouse gene names: https://www.r-bloggers.com/2016/10/converting-mouse-to-human-gene-names-with-biomart-package/
convertHumanGeneList <- function(x){
  require("biomaRt")
  human = useMart("ensembl", dataset = "hsapiens_gene_ensembl")
  mouse = useMart("ensembl", dataset = "mmusculus_gene_ensembl")
  genesV2 = getLDS(attributes = c("hgnc_symbol"), filters = "hgnc_symbol", values = x , mart = human, attributesL = c("mgi_symbol"), martL = mouse, uniqueRows=T)
  humanx <- unique(genesV2[, 2])
  # Print the first 6 genes found to the screen
  print(head(humanx))
  return(humanx)
}

convertMouseGeneList <- function(x){
  require("biomaRt")
  human = useMart("ensembl", dataset = "hsapiens_gene_ensembl")
  mouse = useMart("ensembl", dataset = "mmusculus_gene_ensembl")
  genesV2 = getLDS(attributes = c("mgi_symbol"), filters = "mgi_symbol", values = x , mart = mouse, attributesL = c("hgnc_symbol"), martL = human, uniqueRows=T)
  humanx <- unique(genesV2[, 2])
  # Print the first 6 genes found to the screen
  print(head(humanx))
  return(humanx)
}

blank_theme <- theme_minimal()+
  theme(
    axis.title.x = element_blank(),
    axis.title.y = element_blank(),
    panel.border = element_blank(),
    panel.grid=element_blank(),
    axis.ticks = element_blank(),
    plot.title=element_text(size=14, face="bold")
  )

#------------------------------------------------------------------------------------------------------#
#------------------------------------------------------------------------------------------------------#
#------------------------------------------------------------------------------------------------------#
#------------------------------------------------------------------------------------------------------#

aggregate_gene_expression#loop through 10X Genomics (CellRanger) output and QC
direcs = c("second_run_CA_Lgr6","second_run_CA_Progeny","second_run_CA_US",
           "second_run_PAP_Lgr6","second_run_PAP_Tom","second_run_PAP_US",
           "second_run_Skin_Lgr6","second_run_Skin_Tom",
           "first_run_LGR6","first_run_NO_TAM_BS","first_run_PROGENY","first_run_TAM_BS")
samp_names = c("CA Lgr6","CA Progeny","CA Unsorted",
               "Pap Lgr6","Pap Progeny","Pap Unsorted",
               "TxSkin Lgr6","TxSkin Progeny",
               "UnTxSkin Lgr6","UnTxSkin Unsorted","UnTxSkin Progeny","TxSkin Unsorted")
tx_names = c("Treated","Treated","Treated",
             "Treated","Treated","Treated",
             "Treated","Treated",
             "Untreated","Untreated","Untreated","Treated")
batch_names = c("Second run","Second run","Second run",
                "Second run","Second run","Second run",
                "Second run","Second run",
                "First run","First run","First run","First run")
mingene = c(500,250,500,
            500,500,750,
            500,500,
            500,250,350,500 ) #vector of minimum genes per sample to shwo low-quality/empty droplets
maxgene = c(6000,5500,6000,
            5500,5500,5000,
            3000,3500,
            4000,2500,3500,3750) #max genes per cell (doublet)
mito_cutoff = c(10,10,10,
                10,10,10,
                15,15,
                10,10,15,17) #max % mitochondrial genes per cell


pdf(file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/figures/QC/initial_QC_matvei_cellranger_output.pdf",height=12,width=10)

colData(big_cds)

for(i in 1:length(direcs)) {
  
  #i=1
  print(i)
  #1. load data from 10X output: 
  matrix_dir = paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/matvei/cell_ranger_output/avicenna/khorms/projects/collaborations/Allan_Bailman/cellranger_outputs/",direcs[i],"/outs/filtered_feature_bc_matrix/",sep="")
  barcode.path <- paste0(matrix_dir, "barcodes.tsv")
  features.path <- paste0(matrix_dir, "features.tsv")
  matrix.path <- paste0(matrix_dir, "matrix.mtx")
  
  mat <- readMM(file = matrix.path) #read in the expression matrix
  
  feature.names = read.delim(features.path,  #read in gene names
                             header = FALSE,
                             stringsAsFactors = FALSE)
  names(feature.names)[1:2] = c("ensemble_id","gene_short_name")
  
  barcode.names = read.delim(barcode.path, #read in barcode names
                             header = FALSE,
                             stringsAsFactors = FALSE)
  
  #create cell_metadata dataframe
  cell_metadat = data.frame(matrix(samp_names[i],ncol=1,nrow=nrow(barcode.names)))
  names(cell_metadat) = "sample"
  row.names(cell_metadat) = barcode.names[,1]
  
  #create gene_annotation_file
  gene_metadat = feature.names
  row.names(gene_metadat) = gene_metadat$ensemble_id
  
  #create expression_data matrix
  colnames(mat) = row.names(cell_metadat)
  row.names(mat) = row.names(gene_metadat)
  
  cds.balmain1 <- new_cell_data_set(expression_data = mat,
                                    cell_metadata = cell_metadat,
                                    gene_metadata = gene_metadat)
  
  #----------------------------------------------------------------------------------------------#
  #Start Seurat QC --> cds.balmain2
  #----------------------------------------------------------------------------------------------#
  #Quality Control after: https://broadinstitute.github.io/2019_scWorkshop/data-wrangling-scrnaseq.html#filtering-low-quality-cells
  
  counts_per_cell <- Matrix::colSums(mat)
  counts_per_gene <- Matrix::rowSums(mat)
  genes_per_cell <- Matrix::colSums(mat>0) # count gene only if it has non-zero reads mapped.
  cells_per_gene <- Matrix::rowSums(mat>0) # only count cells where the gene is expressed
  
  #plot counts
  par(mfrow=c(2,2))
  hist((counts_per_cell+1),xlab='counts per cell',main=paste(samp_names[i],'counts per cell'),col='wheat')
  tryCatch({  hist((genes_per_cell+1), xlab='genes per cell', main=paste(samp_names[i],'genes per cell'), col='wheat') + abline(v=c(mingene[i],maxgene[i]), col="red",lwd=3)},  error=function(e){})
  plot(counts_per_cell, xlab='counts per cell',ylab='genes per cell',genes_per_cell, log='xy', main=paste(samp_names[i],'counts vs genees per cell'),col='wheat')
  tryCatch({ plot(sort(genes_per_cell), xlab='cell', ylab='genes per cell', main=paste(samp_names[i],'genes per cell (ordered)'))  + abline(h=c(mingene[i],maxgene[i]), col="red",lwd=3) },  error=function(e){})#distribution of library complexity --> libraries with hardly any genes should probably be dropped
  
  #plots for powerpoint
  #tryCatch({  hist((genes_per_cell+1), xlab='genes per cell', main=paste(samp_names[i],'genes per cell'), col='wheat')},  error=function(e){})
  #tryCatch({ plot(sort(genes_per_cell), xlab='cell', ylab='genes per cell', main=paste(samp_names[i],'genes per cell (ordered)'))},  error=function(e){})#distribution of library complexity --> libraries with hardly any genes should probably be dropped
  #tryCatch({  hist((genes_per_cell+1), xlab='genes per cell', main=paste(samp_names[i],'genes per cell'), col='wheat') + abline(v=c(mingene[i],maxgene[i]), col="red",lwd=3)},  error=function(e){})
  #tryCatch({ plot(sort(genes_per_cell), xlab='cell', ylab='genes per cell', main=paste(samp_names[i],'genes per cell (ordered)'))  + abline(h=c(mingene[i],maxgene[i]), col="red",lwd=3) },  error=function(e){})#distribution of library complexity --> libraries with hardly any genes should probably be dropped
  
  
  par(mfrow=c(1,1))
  tryCatch({ plot(sort(genes_per_cell), xlab='cell', ylab='genes per cell', main=paste(samp_names[i],'genes per cell (ordered)')) + abline(h=c(mingene[i],maxgene[i]), col="red",lwd=3) },  error=function(e){}) #distribution of library complexity --> libraries with hardly any genes should probably be dropped
  
  #create Seurat object
  #make row names into gene symbol, not ENSEMBL
  row.names(mat) = feature.names$gene_short_name
  seurat<-CreateSeuratObject(counts = mat, min.cells = 3, min.features = 200, project = "CA_US")
  
  #QC: add mitochondrial percent
  seurat[["percent.mt"]] <- PercentageFeatureSet(seurat, pattern = "mt-")
  violplot = VlnPlot(seurat, features = c("percent.mt"), ncol = 1) + geom_violin(fill="wheat") + theme(legend.position = "none") + ggtitle(paste(samp_names[i],"% mitochondrial")) + xlab("") + geom_hline(yintercept=mito_cutoff[i], color="red",size=1.25) + theme(axis.text.x = element_text(size=0))
  
  #Getting rid of low-quality cells/empty droplets (<200 genes) ; doublets (>6k genes) ; stresst/dying cells (<10% mito)
  seurat <- subset(seurat, subset = nFeature_RNA > mingene[i] & nFeature_RNA < maxgene[i] & percent.mt < 10)
  
  #normalize
  ?NormalizeData
  #Klein lab normalization https://www.nature.com/articles/nature25741#MOESM1 The gene expression counts of each cell were then normalized using a variant of total-count normalization that avoids distortion from very highly expressed genes. Specifically, we calculated , the normalized transcript counts for gene j in cell i, from the raw counts xi,j as follows: , in which  and  is the average of Xi over all cells. To prevent very highly expressed genes (for example, haemoglobin) from correspondingly decreasing the relative expression of other genes, we excluded genes comprising >10% of the total counts of any cell when calculating  and Xi.
  seurat <- NormalizeData(seurat, normalization.method = "LogNormalize", scale.factor = 10000)
  
  #find highly variable genes ?FindVariableFeatures
  seurat <- FindVariableFeatures(seurat, selection.method = "vst", nfeatures = 2000)
  
  # Identify the 10 most highly variable genes
  top10 <- head(VariableFeatures(seurat), 10)
  
  # plot variable features with and without labels: a lot of goddam hemoglobin genes maybe I should do something about this ?VariableFeaturePlot
  plot1 <- VariableFeaturePlot(seurat) + theme(legend.position = "bottom") + ggtitle(samp_names[i])
  plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)+ theme(legend.position = "bottom") + ggtitle(samp_names[i])
  
  print( multiplot(plot2,violplot,cols=2))
  
  #Seurat object --> CDS
  exp_mat <- seurat@assays[["RNA"]]@data #pull out NORMALIZED counts from Seurat object
  cell_metadat <- seurat@meta.data #pull out cell meta-data from Seurat object
  cell_metadat$tissue = strsplit(samp_names[i]," ")[[1]][1] #put column of tissue in the cell metadata
  cell_metadat$sorting = strsplit(samp_names[i]," ")[[1]][2] #put column of tissue in the cell metadata
  cell_metadat$sample_nom = samp_names[i]
  cell_metadat$tx_name = tx_names[i]
  cell_metadat$batch_name = as.factor(batch_names[i])
  gene_annot = data.frame(seurat@assays[["RNA"]]@data@Dimnames[[1]])#pull out gene names from Seurat object
  names(gene_annot) = "gene_short_name"
  row.names(gene_annot) = gene_annot$gene_short_name #row.names of gene_metadata must be equal to row.names of expression_data
  
  #Make the CDS object (cds.balmain2) with only Seurat-processed var features
  #exp_mat.varfeat = exp_mat[match(VariableFeatures(seruat),rownames(exp_mat)),] #get only variable features
  #gene_annot.varfeat = data.frame(gene_annot[match(VariableFeatures(seruat),gene_annot$gene_short_name),])
  #names(gene_annot.varfeat) = "gene_short_name"
  #row.names(gene_annot.varfeat) = gene_annot.varfeat$gene_short_name #row.names of gene_metadata must be equal to row.names of expression_data
  
  #assign each cds object to a new name
  name <- paste("cds.", i, sep = "")
  assign(name, new_cell_data_set(exp_mat,
                                 cell_metadata = cell_metadat,
                                 gene_metadata = gene_annot)    #create a bunch of cds objects that we are going to combine
  )
  
  #summary BEFORE & AFTER QC metrics
  #extract: median IQR
  format_summary = function(x) {
    output = paste(round(x[1],0)," (",round(x[2],0),", ",round(x[3],0),")",sep="")
    return(output)
  }
  #pre-processing matrix : mat
  n_cells = dim(mat)[2]
  counts_per_cell <- format_summary(data.frame(as.vector(summary(Matrix::colSums(mat))))[c(3,2,5),])
  counts_per_gene <- format_summary(data.frame(as.vector(summary(Matrix::rowSums(mat))))[c(3,2,5),])
  genes_per_cell <- format_summary(data.frame(as.vector(summary(Matrix::colSums(mat>0))))[c(3,2,5),]) # count gene only if it has non-zero reads mapped.
  cells_per_gene <- format_summary(data.frame(as.vector(summary(Matrix::rowSums(mat>0))))[c(3,2,5),]) # only count cells where the gene is expressed
  #post-processing matrix : exp_mat
  n_cells.exp = dim(exp_mat)[2]
  counts_per_cell.exp <- format_summary(data.frame(as.vector(summary(Matrix::colSums(exp_mat))))[c(3,2,5),])
  counts_per_gene.exp <- format_summary(data.frame(as.vector(summary(Matrix::rowSums(exp_mat))))[c(3,2,5),])
  genes_per_cell.exp <- format_summary(data.frame(as.vector(summary(Matrix::colSums(exp_mat>0))))[c(3,2,5),]) # count gene only if it has non-zero reads mapped.
  cells_per_gene.exp <- format_summary(data.frame(as.vector(summary(Matrix::rowSums(exp_mat>0))))[c(3,2,5),]) # only count cells where the gene is expressed
  
  sum.name <- paste("sum.", i, sep = "")
  assign(sum.name, data.frame(samp_names[i],
                              n_cells, counts_per_cell, counts_per_gene, genes_per_cell, cells_per_gene,
                              n_cells.exp,counts_per_cell.exp, counts_per_gene.exp, genes_per_cell.exp, cells_per_gene.exp)
  )
  
} 

dev.off()

#write out QC summary
totcombsum = rbind(sum.10,sum.11,sum.9,sum.12,sum.8,sum.7,sum.6,sum.5,sum.4,sum.3,sum.2,sum.1)
write.table(totcombsum, file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/QC/QC_matvei_summary.txt",row.names=F,quote=F,sep="\t")

#=====================================================================================
#Combine cds objects and begin analysis
#=====================================================================================

#create 1 giant ass cds object from all the cds's created above
big_cds <- combine_cds(list(cds.10,cds.11,cds.9,cds.12,cds.8,cds.7,cds.6,cds.5,cds.4,cds.3,cds.2,cds.1))
rm(cds.10,cds.11,cds.9,cds.12,cds.8,cds.7,cds.6,cds.5,cds.4,cds.3,cds.2,cds.1,cds.balmain1)
gc()

dim(big_cds@assays@data@listData$counts)
counts_per_cell.big_cds <- format_summary(data.frame(as.vector(summary(Matrix::colSums(big_cds@assays@data@listData$counts))))[c(3,2,5),])
counts_per_gene.big_cds <- format_summary(data.frame(as.vector(summary(Matrix::rowSums(big_cds@assays@data@listData$counts))))[c(3,2,5),])
genes_per_cell.big_cds <- format_summary(data.frame(as.vector(summary(Matrix::colSums(big_cds@assays@data@listData$counts>0))))[c(3,2,5),]) # count gene only if it has non-zero reads mapped.
cells_per_gene.big_cds <- format_summary(data.frame(as.vector(summary(Matrix::rowSums(big_cds@assays@data@listData$counts>0))))[c(3,2,5),]) # only count cells where the gene is expressed

#3b. Pre-process with Seurat-processed object
big_cds <- preprocess_cds(big_cds, 
                          method="PCA",
                          num_dim = 100,
                          norm_method = "none", #these data have already been normalized by Seurat
                          alignment_group = NULL)


#3c. plot PC variation explained ?plot_pc_variance_explained
plot_pc_variance_explained(big_cds) + ggtitle("All-sample Scree Plot") +
  theme(text = element_text(size=18)) +
  theme(axis.text=element_text(size=18,color='black'),
        axis.title=element_text(size=20,color='black')) +
  theme(plot.title = element_text(size = 22, vjust=1)) +
  theme(axis.text=element_text(size=15,color='black'),
        axis.title=element_text(size=15,color='black')) +
  theme(panel.grid.major=element_line(colour = "darkgrey"), 
        panel.grid.minor=element_blank(), 
        panel.background=element_blank()) +
  theme(legend.key = element_rect(fill = "white")) +
  theme(strip.text.x = element_text(size = 14, colour = "black", angle = 0)) +
  theme(strip.text.y = element_text(size = 14, colour = "black", angle = 0)) +
  theme(legend.text=element_text(size=14), legend.title=element_text(size=14))

#4. A further round of (nonlinear) dimensionality reduction using UMAP ?reduce_dimension
#4b.
big_cds <- reduce_dimension(
  big_cds,
  max_components = 3,
  reduction_method = c("UMAP"),
  preprocess_method = "PCA",
  umap.metric = "cosine",
  umap.min_dist = 0.1,
  umap.n_neighbors = 15L,
  umap.fast_sgd = FALSE,
  umap.nn_method = "annoy",
  cores = 1,
  verbose = FALSE
)

#5. Cluster and partition ?cluster_cells cells into supergroups community detection (Phenograph algorithm) Community detection of cells as part of Phenograph algorithm https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4508757/

#5b. 
?cluster_cells
big_cds <- cluster_cells(big_cds,
                         reduction_method = c("UMAP"),
                         k = 10, #a bigger k will result in lower resolution 
                         cluster_method = c("leiden"),
                         num_iter = 2,
                         partition_qval = 0.05, #Numeric, the q-value cutoff to determine when to partition
                         weight = FALSE,
                         resolution = NULL,
                         random_seed = NULL,
                         verbose = F)

color_map <- c("red","darkblue","purple")

plot_cells_3d(
  big_cds,
  dims = c(1, 2, 3),
  reduction_method = c("UMAP"),
  color_cells_by = "tissue_updated",
  color_palette = color_map,
  genes = NULL,
  show_trajectory_graph = TRUE,
  trajectory_graph_color = "black",
  trajectory_graph_segment_size = 2,
  norm_method = c("log", "size_only"),
  cell_size = 1,
  alpha = 1,
  min_expr = 0.1
)

?learn_graph
big_cds = learn_graph(
  big_cds, use_partition = F)


#DEG upper spike b/w spindle carcinoma

#------------------------------------------------------------------------------------#
#cells in manifold without any colors ?plot_cells
dims12.clust = plot_cells(big_cds, group_cells_by="cluster",
                          group_label_size = 0,
                          x=1,
                          y=2,
                          alpha=0.5,
                          cell_size=0) +
  geom_point(alpha=0.2,size=0.2,color="grey") +
  ggtitle("Manifold: Dim. 1 and 2") + xlab("Dim. 1") + ylab("Dim. 2") + 
  theme(text = element_text(size=18)) + theme(axis.text=element_text(size=18,color='black'),
                                              axis.title=element_text(size=20,color='black')) + theme(plot.title = element_text(size = 22, vjust=1)) + theme(axis.text=element_text(size=15,color='black'),
                                                                                                                                                             axis.title=element_text(size=15,color='black'))

dims23.clust = plot_cells(big_cds, group_cells_by="cluster",
                          group_label_size = 0,
                          x=2,
                          y=3) +
  geom_point(alpha=1,size=0.2,color="black") +
  ggtitle("Manifold: Dim. 2 and 3") + xlab("Dim. 2") + ylab("Dim. 3") + 
  theme(text = element_text(size=18)) + theme(axis.text=element_text(size=18,color='black'),
                                              axis.title=element_text(size=20,color='black')) + theme(plot.title = element_text(size = 22, vjust=1)) + 
  theme(axis.text=element_text(size=15,color='black'), axis.title=element_text(size=15,color='black'))                                                                                                                       

multiplot(dims12.clust,dims23.clust,cols=2)

#------------------------------------------------------------------------------------#
#plot clusters and partitions along dims 1&2
dims12.clust = plot_cells(big_cds, color_cells_by="cluster", group_cells_by="cluster",
                          group_label_size = 5,
                          label_groups_by_cluster=T,
                          x=1,
                          y=2) +
  ggtitle("Cluster: Dim. 1 and 2") + xlab("Dim. 1") + ylab("Dim. 2") + 
  theme(text = element_text(size=18)) + theme(axis.text=element_text(size=18,color='black'),
                                              axis.title=element_text(size=20,color='black')) + 
  theme(plot.title = element_text(size = 22, vjust=1)) + 
  theme(axis.text=element_text(size=15,color='black')) +
  scale_color_manual(values=getPalette(length(unique(colData(big_cds)$cluster))))

#plot clusters and partitions along dims 1&2
dims12.clust_clustlabel = plot_cells(big_cds, color_cells_by="cluster", group_cells_by="cluster",
                                     group_label_size = 5,
                                     label_groups_by_cluster=T,
                                     show_trajectory_graph = F,
                                     x=1,
                                     y=2,
                                     alpha=0.8) +
  ggtitle("Cluster: Dim. 1 and 2") + xlab("Dim. 1") + ylab("Dim. 2") + 
  theme(text = element_text(size=18)) + theme(axis.text=element_text(size=18,color='black'),
                                              axis.title=element_text(size=20,color='black')) + 
  theme(plot.title = element_text(size = 22, vjust=1)) + 
  theme(axis.text=element_text(size=15,color='black')) +
  scale_color_manual(values=getPalette(length(unique(colData(big_cds)$cluster))))

dims12.clust_trajectory = plot_cells(big_cds, color_cells_by="cluster", group_cells_by="cluster",
                                     group_label_size = 0,
                                     label_groups_by_cluster=F,
                                     show_trajectory_graph = T,
                                     x=1,
                                     y=2,
                                     alpha=0.3) +
  ggtitle("Cluster: Dim. 1 and 2") + xlab("Dim. 1") + ylab("Dim. 2") + 
  theme(text = element_text(size=18)) + theme(axis.text=element_text(size=18,color='black'),
                                              axis.title=element_text(size=20,color='black')) + 
  theme(plot.title = element_text(size = 22, vjust=1)) + 
  theme(axis.text=element_text(size=15,color='black')) +
  scale_color_manual(values=getPalette(length(unique(colData(big_cds)$cluster))))

#Clusters in dimensional rotation
png("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/figures/cell_clusters/clusters_dims_123.png",width=36,height=12,units="in",res=400)
print(
  multiplot(dims12.clust, dims12.clust_clustlabel,
            dims12.clust_trajectory,
            cols=3)
)
dev.off()

#Clusters in dimensional rotation
png("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/figures/cell_clusters/clusters_dims_123.png",width=14,height=7,units="in",res=400)
print(
  multiplot(dims12.clust, dims23.clust,
            cols=2)
)
dev.off()

#UMAP with no label
dims12.clust_nolabel = plot_cells(big_cds, color_cells_by="cluster", group_cells_by="cluster",
                                  group_label_size = 0,
                                  label_groups_by_cluster=F,
                                  show_trajectory_graph = F,
                                  x=1,
                                  y=2,
                                  alpha=0.3) +
  ggtitle("Cluster: Dim. 1 and 2") + xlab("Dim. 1") + ylab("Dim. 2") + 
  theme(text = element_text(size=18)) + theme(axis.text=element_text(size=18,color='black'),
                                              axis.title=element_text(size=20,color='black')) + 
  theme(plot.title = element_text(size = 22, vjust=1)) + 
  theme(axis.text=element_text(size=15,color='black')) +
  scale_color_manual(values=getPalette(length(unique(colData(big_cds)$cluster))))

png("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/figures/cell_clusters/clusters_dims_123_noclustlabel.png",
    width=7,height=7,units="in",res=400)
print(
  dims12.clust_nolabel
)
dev.off()

#=====================================================================================
#Summarize gene expression and sparsity across the entire data set
#=====================================================================================
count_dat = as.matrix(big_cds@assays@data$counts)
str(count_dat)
rm(count_dat)

gene_summaries = do.call(rbind, lapply(1:nrow(count_dat), function(i){
  
  #i=1
  print(i)
  gene_name = row.names(count_dat)[i]
  sparsity = (length(count_dat[i,count_dat[i,]>0]) / dim(count_dat)[2])*100
  mean_expression = mean(count_dat[i,count_dat[i,]>0])
  data.frame(gene_name, mean_expression, sparsity)
  
}))
str(count_dat)

write.table(gene_summaries, 
            file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/gene_summaeries/gene_summaries.txt",
            sep="\t",quote=F,row.names=F)

get_citations(big_cds)
#------------------------------------------------------------------------------------#
#------------------------------------------------------------------------------------#
#------------------------------------------------------------------------------------#
#UN-BATCH CORRECTED
#plot cells by experimental characteristics ?plot_cells

colData(big_cds)

colData(big_cds)$tissue_sorting = paste(colData(big_cds)$sorting, colData(big_cds)$tissue, sep="_")

unique(colData(big_cds)$sorting)
unique(colData(big_cds)$tissue)


#Tissue: Skin, Pap, CA
dims12.clust = plot_cells(big_cds, color_cells_by="tissue_sorting", group_cells_by="cluster",
                          group_label_size = 0,
                          x=1,
                          y=2) +
  theme(legend.position="right")


plot_cells(big_cds, color_cells_by="tissue_sorting", group_cells_by="cluster",
           group_label_size = 0,
           x=1,
           y=2)

ggtitle("Tissue: Dim. 1 and 2") + xlab("Dim. 1") + ylab("Dim. 2") +
  theme(axis.text=element_text(size=0,color='black'),
        title=element_text(size=18),
        axis.title=element_text(size=0),
        axis.ticks = element_line(size=0),
        legend.position="bottom",
        legend.text=element_text(size=15))+ 
  theme(strip.text.x = element_text(size = 10)) +
  scale_color_manual(values = c("UnTxSkin" = "skyblue3","TxSkin" = "darkblue", "Pap" = "darkgreen","CA" = "darkred")) + theme(legend.position="bottom")

dims23.clust = plot_cells(big_cds, color_cells_by="tissue", group_cells_by="cluster",
                          group_label_size = 0,
                          x=2,
                          y=3) +
  ggtitle("Tissue: Dim. 2 and 3") + xlab("Dim. 2") + ylab("Dim. 3") + 
  theme(axis.text=element_text(size=0,color='black'),
        title=element_text(size=18),
        axis.title=element_text(size=0),
        axis.ticks = element_line(size=0),
        legend.position="bottom",
        legend.text=element_text(size=15))+ 
  theme(strip.text.x = element_text(size = 10)) +
  scale_color_manual(values = c("UnTxSkin" = "skyblue3","TxSkin" = "darkblue", "Pap" = "darkgreen","CA" = "darkred")) + theme(legend.position="bottom")

multiplot(dims12.clust,dims23.clust,cols=2)

#Sorting: Unsorted, Progeny, Lgr6

unique(cell_metadat$sorting)
facs_unc = plot_cells(big_cds, color_cells_by="sorting", group_cells_by="cluster",
                          group_label_size = 0,
                          x=1,
                          y=2) +
  ggtitle("Sorting: Dim. 1 and 2") + xlab("Dim. 1") + ylab("Dim. 2") + 
  theme(axis.text=element_text(size=0,color='black'),
        title=element_text(size=18),
        axis.title=element_text(size=0),
        axis.ticks = element_line(size=0),
        legend.position="bottom",
        legend.text=element_text(size=15))+ 
  theme(strip.text.x = element_text(size = 10)) +
  scale_color_manual(values = c("Unsorted" = "lightsteelblue2", "Progeny" = "olivedrab","Lgr6" = "burlywood4")) + theme(legend.position="bottom")

png("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/batch\ effects/figs/facs_sorting_UNbatchcorrected.png",
    width=8,height=8,units="in",res=300)
  print(facs_unc)
dev.off()


#check batch effects
unique(cell_metadat$batch_name)
batch_unc = plot_cells(big_cds, color_cells_by="batch_name", group_cells_by="cluster",
                          group_label_size = 0,
                          x=1,
                          y=2) +
  ggtitle("Batch: Dim. 1 and 2") + xlab("Dim. 1") + ylab("Dim. 2") + 
  theme(axis.text=element_text(size=0,color='black'),
        title=element_text(size=18),
        axis.title=element_text(size=0),
        axis.ticks = element_line(size=0),
        legend.position="bottom",
        legend.text=element_text(size=15))+ 
  theme(strip.text.x = element_text(size = 10)) +
  scale_color_manual(name="Batch",values = c("First run" = "wheat3", "Second run" = "tomato3")) + theme(legend.position="bottom")

png("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/batch\ effects/figs/batch_UNbatchcorrected.png",
    width=8,height=8,units="in",res=300)
print(batch_unc)
dev.off()

#=====================================================================================
#Re-align to remove batch effects: batch is covarying strongly with Unsorted cells
#=====================================================================================
big_cds = align_cds(cds = big_cds,
                    num_dim = 100,
                    preprocess_method = "PCA",
                    alignment_group = "batch_name",
                    verbose=T)

#4b. ?reduce_dimension
big_cds <- reduce_dimension(
  big_cds,
  max_components = 3,
  reduction_method = c("UMAP"),
  preprocess_method = "Aligned", #must now used ALIGNEMNT PRE-PROCESSING METHOD bc it replaces the other pre-process shit
  umap.metric = "cosine",
  umap.min_dist = 0.1,
  umap.n_neighbors = 15L,
  umap.fast_sgd = FALSE,
  umap.nn_method = "annoy",
  cores = 1,
  verbose = FALSE
)

#5. Cluster and partition ?cluster_cells cells into supergroups community detection (Phenograph algorithm) Community detection of cells as part of Phenograph algorithm https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4508757/
#5b. 
big_cds <- cluster_cells(big_cds,
                         reduction_method = c("UMAP"),
                         k = 10, #a bigger k will result in lower resolution 
                         cluster_method = c("leiden"),
                         num_iter = 2,
                         partition_qval = 0.05, #Numeric, the q-value cutoff to determine when to partition
                         weight = FALSE,
                         resolution = NULL,
                         random_seed = NULL,
                         verbose = F)

#write out the QC'd, batch-corrected, PCA- and UMAP-reduced, clustered and partitioned object
saveRDS(big_cds, file = "/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/matvei/cds_objects/integrated_corrected_cds.RDS") 

#=====================================================================================
#Read in batch correct CDS
#=====================================================================================
big_cds = readRDS("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/matvei/cds_objects/integrated_corrected_cds_v2.RDS")

#=====================================================================================#
#Main text fig: Fig. 1C
#=====================================================================================#
#Make a New Tissue specificity columns for Normal skin (UnTxSkin and TxSkin), Pap, and CA
colData(big_cds)$tissue_updated = colData(big_cds)$tissue
colData(big_cds)$tissue_updated = gsub( "\\bUnTxSkin\\b", "Normal skin", colData(big_cds)$tissue_updated,perl=TRUE)
colData(big_cds)$tissue_updated = gsub( "\\bTxSkin\\b", "Normal skin", colData(big_cds)$tissue_updated,perl=TRUE)
colData(big_cds)$tissue_updated_sorting = paste(colData(big_cds)$tissue_updated, colData(big_cds)$sorting)

#FIG.2 PLOT
#Plot everything together
all_cells_plot = plot_cells(big_cds, color_cells_by="tissue_updated", group_cells_by="cluster",
                            group_label_size = 0,
                            x=1,
                            y=2,
                            alpha=0.6) +
  ggtitle("") + xlab("") + ylab("") + 
  theme_void() +
  theme(axis.text=element_text(size=0,color='black'),
        axis.title=element_text(size=0),
        title=element_text(size=0),
        axis.ticks = element_line(size=0),
        legend.position="none",
        line = element_blank()) +
  scale_color_manual(values = c("Normal skin" = "darkblue", "Pap" = "plum4", "CA" = "darkred"))

png("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Figs/Fig2/Fig2h_allcells.png",
    width=8,height=8,units="in",res=300)
print(all_cells_plot)
dev.off()

normal_unsort = plot_cells(big_cds, color_cells_by="tissue_updated_sorting", group_cells_by="cluster",
                           group_label_size = 0,
                           x=1,
                           y=2,
                           alpha=0.2) +
  ggtitle("") + xlab("") + ylab("") + 
  theme(axis.text=element_text(size=0,color='black'),
        axis.title=element_text(size=0),
        axis.ticks = element_line(size=0),
        legend.position="none") +
  scale_color_manual(values = c("Normal skin Unsorted" = "darkblue", "Normal skin Progeny" = "lightgrey", "Normal skin Lgr6" = "lightgrey", 
                                "Pap Unsorted" = "lightgrey", "Pap Progeny" = "lightgrey", "Pap Lgr6" = "lightgrey", 
                                "CA Unsorted" = "lightgrey", "CA Progeny" = "lightgrey", "CA Lgr6"  = "lightgrey"))

normal_lgr6 = plot_cells(big_cds, color_cells_by="tissue_updated_sorting", group_cells_by="cluster",
                         group_label_size = 0,
                         x=1,
                         y=2,
                         alpha=0.2) +
  ggtitle("") + xlab("") + ylab("") + 
  theme(axis.text=element_text(size=0,color='black'),
        axis.title=element_text(size=0),
        axis.ticks = element_line(size=0),
        legend.position="none") +
  scale_color_manual(values = c("Normal skin Unsorted" = "lightgrey", "Normal skin Progeny" = "lightgrey", "Normal skin Lgr6" = "darkblue", 
                                "Pap Unsorted" = "lightgrey", "Pap Progeny" = "lightgrey", "Pap Lgr6" = "lightgrey", 
                                "CA Unsorted" = "lightgrey", "CA Progeny" = "lightgrey", "CA Lgr6"  = "lightgrey"))

normal_progeny = plot_cells(big_cds, color_cells_by="tissue_updated_sorting", group_cells_by="cluster",
                            group_label_size = 0,
                            x=1,
                            y=2,
                            alpha=0.2) +
  ggtitle("") + xlab("") + ylab("") + 
  theme(axis.text=element_text(size=0,color='black'),
        axis.title=element_text(size=0),
        axis.ticks = element_line(size=0),
        legend.position="none") +
  scale_color_manual(values = c("Normal skin Unsorted" = "lightgrey", "Normal skin Progeny" = "darkblue", "Normal skin Lgr6" = "lightgrey", 
                                "Pap Unsorted" = "lightgrey", "Pap Progeny" = "lightgrey", "Pap Lgr6" = "lightgrey", 
                                "CA Unsorted" = "lightgrey", "CA Progeny" = "lightgrey", "CA Lgr6"  = "lightgrey"))



pap_unsort = plot_cells(big_cds, color_cells_by="tissue_updated_sorting", group_cells_by="cluster",
                        group_label_size = 0,
                        x=1,
                        y=2,
                        alpha=0.2) +
  ggtitle("") + xlab("") + ylab("") + 
  theme(axis.text=element_text(size=0,color='black'),
        axis.title=element_text(size=0),
        axis.ticks = element_line(size=0),
        legend.position="none") +
  scale_color_manual(values = c("Normal skin Unsorted" = "lightgrey", "Normal skin Progeny" = "lightgrey", "Normal skin Lgr6" = "lightgrey", 
                                "Pap Unsorted" = "purple", "Pap Progeny" = "lightgrey", "Pap Lgr6" = "lightgrey", 
                                "CA Unsorted" = "lightgrey", "CA Progeny" = "lightgrey", "CA Lgr6"  = "lightgrey"))

pap_lgr6 = plot_cells(big_cds, color_cells_by="tissue_updated_sorting", group_cells_by="cluster",
                      group_label_size = 0,
                      x=1,
                      y=2,
                      alpha=0.2) +
  ggtitle("") + xlab("") + ylab("") + 
  theme(axis.text=element_text(size=0,color='black'),
        axis.title=element_text(size=0),
        axis.ticks = element_line(size=0),
        legend.position="none") +
  scale_color_manual(values = c("Normal skin Unsorted" = "lightgrey", "Normal skin Progeny" = "lightgrey", "Normal skin Lgr6" = "lightgrey", 
                                "Pap Unsorted" = "lightgrey", "Pap Progeny" = "lightgrey", "Pap Lgr6" = "purple", 
                                "CA Unsorted" = "lightgrey", "CA Progeny" = "lightgrey", "CA Lgr6"  = "lightgrey"))

pap_progeny = plot_cells(big_cds, color_cells_by="tissue_updated_sorting", group_cells_by="cluster",
                         group_label_size = 0,
                         x=1,
                         y=2,
                         alpha=0.2) +
  ggtitle("") + xlab("") + ylab("") + 
  theme(axis.text=element_text(size=0,color='black'),
        axis.title=element_text(size=0),
        axis.ticks = element_line(size=0),
        legend.position="none") +
  scale_color_manual(values = c("Normal skin Unsorted" = "lightgrey", "Normal skin Progeny" = "lightgrey", "Normal skin Lgr6" = "lightgrey", 
                                
                                "Pap Unsorted" = "lightgrey", "Pap Progeny" = "purple", "Pap Lgr6" = "lightgrey", 
                                "CA Unsorted" = "lightgrey", "CA Progeny" = "lightgrey", "CA Lgr6"  = "lightgrey"))

ca_unsort = plot_cells(big_cds, color_cells_by="tissue_updated_sorting", group_cells_by="cluster",
                       group_label_size = 0,
                       x=1,
                       y=2,
                       alpha=0.2) +
  ggtitle("") + xlab("") + ylab("") + 
  theme(axis.text=element_text(size=0,color='black'),
        axis.title=element_text(size=0),
        axis.ticks = element_line(size=0),
        legend.position="none") +
  scale_color_manual(values = c("Normal skin Unsorted" = "lightgrey", "Normal skin Progeny" = "lightgrey", "Normal skin Lgr6" = "lightgrey", 
                                
                                "Pap Unsorted" = "lightgrey", "Pap Progeny" = "lightgrey", "Pap Lgr6" = "lightgrey", 
                                "CA Unsorted" = "darkred", "CA Progeny" = "lightgrey", "CA Lgr6"  = "lightgrey"))

ca_lgr6 = plot_cells(big_cds, color_cells_by="tissue_updated_sorting", group_cells_by="cluster",
                     group_label_size = 0,
                     x=1,
                     y=2,
                     alpha=0.2) +
  ggtitle("") + xlab("") + ylab("") + 
  theme(axis.text=element_text(size=0,color='black'),
        axis.title=element_text(size=0),
        axis.ticks = element_line(size=0),
        legend.position="none") +
  scale_color_manual(values = c("Normal skin Unsorted" = "lightgrey", "Normal skin Progeny" = "lightgrey", "Normal skin Lgr6" = "lightgrey", 
                                
                                "Pap Unsorted" = "lightgrey", "Pap Progeny" = "lightgrey", "Pap Lgr6" = "lightgrey", 
                                "CA Unsorted" = "lightgrey", "CA Progeny" = "lightgrey", "CA Lgr6"  = "darkred"))

ca_progeny = plot_cells(big_cds, color_cells_by="tissue_updated_sorting", group_cells_by="cluster",
                        group_label_size = 0,
                        x=1,
                        y=2,
                        alpha=0.2) +
  ggtitle("") + xlab("") + ylab("") + 
  theme(axis.text=element_text(size=0,color='black'),
        axis.title=element_text(size=0),
        axis.ticks = element_line(size=0),
        legend.position="none") +
  scale_color_manual(values = c("Normal skin Unsorted" = "lightgrey", "Normal skin Progeny" = "lightgrey", "Normal skin Lgr6" = "lightgrey", 
                                
                                "Pap Unsorted" = "lightgrey", "Pap Progeny" = "lightgrey", "Pap Lgr6" = "lightgrey", 
                                "CA Unsorted" = "lightgrey", "CA Progeny" = "darkred", "CA Lgr6"  = "lightgrey"))

#Fig1c
png("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/stem_cell_mansucript/Figs/Fig1/Fig1c.png",width=21,height=21,units="in",res=800)
print(
  multiplot(normal_unsort, normal_lgr6, normal_progeny,
            pap_unsort, pap_lgr6, pap_progeny,
            ca_unsort, ca_lgr6, ca_progeny,
            cols=3)
)
dev.off()

#=====================================================================================#
#Extended Data Fig.1a : Unsupervised cell paritions (high-level regional cell clusters)
#=====================================================================================#
#set color palette
colourCount = length(unique(colData(big_cds)$partition))
getPalette = colorRampPalette(brewer.pal(9, "Set1"))

png("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Extended\ Data/Extended\ Data\ Fig\ 1/ExtDat_Fig1a.png",width=7,height=7,units="in",res=200)

print(plot_cells(big_cds, color_cells_by="partition", group_cells_by="partition",
                 group_label_size = 7,
                 x=1,
                 y=2,
                 alpha=0.2) +
        ggtitle("") + xlab("") + ylab("") + 
        theme(axis.text=element_text(size=0,color='black'),
              axis.title=element_text(size=0),
              axis.ticks = element_line(size=0),
              legend.position="none") +
        scale_color_manual(values=getPalette(colourCount))
)
dev.off()

#=====================================================================================#
#Extended Data Fig.1b : Proportion of Pap/cluster within each partition
#=====================================================================================#
cluster_tissue_df = as.data.frame(subset(colData(big_cds), select=c("partition","tissue_updated"))) #extract partition and tissue
cluster_tissue_df$tissue_updated = gsub("CA","SCC",cluster_tissue_df$tissue_updated)
cluster_tissue_df$tissue_updated = gsub("Pap","Papilloma",cluster_tissue_df$tissue_updated)

png("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Extended\ Data/Extended\ Data\ Fig\ 1/ExtDat_Fig1b.png",
    width=12,height=6,units="in",res=250)
print(
  ggplot(cluster_tissue_df, aes(x=partition, fill=tissue_updated))+geom_bar(position="fill")+
    ggtitle("") + xlab("Cluster") + ylab("Cell fraction") +
    scale_fill_manual(values = c("Normal skin" = "darkblue", "Papilloma" = "mediumpurple2","SCC" = "red"), name="Tissue") + 
    theme(legend.position="bottom") + theme_classic() +
    theme(axis.text=element_text(size=20,color='black'),
          axis.title=element_text(size=20,color='black'),
          legend.title=element_text(size=20), 
          legend.text=element_text(size=20)) +
    scale_y_continuous(expand = c(0,0), limits=c(0,1)) #get rid of space at bottom and top
)
dev.off()

#=====================================================================================#
#Top markers: Partition (low-resolution cell cluster)
#=====================================================================================#
#Find top markers differentiating partitions
marker_test_res <- top_markers(big_cds, group_cells_by="partition",
                               genes_to_test_per_group = 100,
                               cores=8)
write.table(marker_test_res, file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/top_markers/big_cds_partition.txt", quote=F,row.names = F,sep="\t")

#=====================================================================================#
#Top markers: Cluster (high-resolution cell cluster)
#=====================================================================================#
#Find top markers differentiating CLUSTERS
marker_test_res <- top_markers(big_cds, group_cells_by="cluster",
                               genes_to_test_per_group = 100,
                               cores=4)
marker_test_res = marker_test_res[order(marker_test_res$marker_test_q_value),] #re-order low to high q-value
marker_test_res = marker_test_res[order(marker_test_res$cell_group),] #re-order based on cluster number
write.table(marker_test_res, file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/top_markers/big_cds_cluster.txt", quote=F,row.names = F,sep="\t")


#Test for GO enrichment in top 100 markers for each unsupervised cell cluster
clusts = unique(marker_test_res$cell_group)

OUT <- createWorkbook() #create excel workbook for a sample https://stackoverflow.com/questions/27713310/easy-way-to-export-multiple-data-frame-to-multiple-excel-worksheets
null_df = do.call(rbind, lapply(1:length(unique(marker_test_res$cell_group)), function(i){
  
  tryCatch({ #suppress error
    #i=1
    
    #Loop through topmarkers for each cluster and test for GO enrichment
    geneUniverse <- unique(row.names(big_cds))
    genesOfInterest <- subset(marker_test_res, cell_group==i)$gene_id #get genes within a module
    geneList <- factor(as.integer(geneUniverse %in% genesOfInterest))
    names(geneList) <- geneUniverse
    GOdata <- new("topGOdata",
                  ontology = "BP",
                  allGenes=geneList,
                  #geneSel = selection,
                  annot = annFUN.org, 
                  mapping = "org.Mm.eg.db",
                  ID = "symbol")
    
    
    resultKS <- runTest(GOdata, algorithm = "weight01", statistic = "fisher") #Fisher exact test
    tab <- GenTable(GOdata, raw.p.value = resultKS, topNodes = length(resultKS@score), numChar = 120)
    tab$module_id = as.character(los_modules_ids[i]) #add module id column
    tab$module_color = subset(module_df, module_id==los_modules_ids[i])$colors[1] #add module color coulmn
    tab = subset(tab, raw.p.value <=0.01) #keep only sig results 
    
    #Get the genes that are contributing to a term
    go_candidates = genesOfInterest
    selcGenes <- genesInTerm(GOdata, whichGO=tab$GO.ID)
    genes_in_go = do.call(rbind, lapply(1:length(selcGenes), function(j){
      #i=1
      data.frame(names(selcGenes)[j],  paste( c(na.omit(selcGenes[[j]][match(go_candidates,  selcGenes[[j]])])) , collapse=", ") )
    }))
    names(genes_in_go) = c("GO_ID","sig_genes_in_term")
    tab = cbind(tab, genes_in_go)
    
    #Write out to excel workbook
    #Add sheet with marker results
    addWorksheet(OUT, paste("Clust ",i," markers",sep="") ) # Add some sheets to the workbook
    writeData(OUT, sheet = paste("Clust ",i," markers",sep=""), x = subset(marker_test_res, cell_group==i)) #Write data to a shee
    
    #Add sheet with marker set's GO results
    addWorksheet(OUT, paste("Clust ",i," GO",sep="") ) # Add some sheets to the workbook
    writeData(OUT, sheet = paste("Clust ",i," GO",sep=""), x = tab) #Write data to a shee
    
  },  error=function(e){})
}))

# Export the Excel file
saveWorkbook(OUT, 
             "/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/top_markers/big_cds_cluster_GO.xlsx",
             overwrite=T)


#=====================================================================================#
#Cell type expression
#=====================================================================================#
#read in updated markers file with Cho's expressed immune cell markers
markers = read.delim(file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/cell_type_clustering/integrated_cell_markers_v3.txt",header=F)
cell_types = markers[grep(">",markers[,1]),] #get only cell types
all_gene_names = rownames(big_cds)
match("Vim",all_gene_names)

str(markers)

#--------------------------------------------------------------------------------------#
#Integrated markers
pdf(file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/cell_type/cell_type_markers_integrated_expression.pdf",height=15,width=15)

for(i in 1:length(cell_types)) {
  
  tryCatch({ #suppress error
    #i=67
    expressed = markers[match(cell_types[i],markers[,1])+1,]
    losgenes = str_split(expressed, "expressed: ")[[1]][2]
    losgenes = as.vector(str_split(losgenes, ", ")[[1]])
    
    #make sure marker genes match - I hecked the non-matching ones, and they do not appear under any alias in the CDS object -- ie these marker genes are not expressed in this cds object!
    print(paste("i = ",i,cell_types[i],sep=" "))
    print(match(losgenes,all_gene_names))
    
    #aggregate gene values to get a single expression value
    marker_df = data.frame(losgenes, str_split( cell_types[i],">")[[1]][2])
    names(marker_df) = c("gene","cell_type")
    
    #Monocle eigene expression
    marker_df$module <- factor(marker_df$cell_type)
    
    agexp_plot = plot_cells(big_cds,
                            genes=marker_df,
                            label_cell_groups=FALSE,
                            label_leaves=F,
                            label_branch_points=F,
                            graph_label_size=1.5,
                            show_trajectory_graph=F) +
      xlab("Dim. 1") + ylab("Dim. 2") + 
      theme(legend.position="bottom") +
      scale_color_viridis_c(option = "inferno",name="integrated expression") +
      ggtitle(paste(marker_df$cell_type[1], "integrated expression",sep=" "))+ #continuous
      theme(text = element_text(size=18)) + theme(axis.text=element_text(size=18,color='black'),
                                                  axis.title=element_text(size=20,color='black')) + theme(plot.title = element_text(size = 22, vjust=1)) + theme(axis.text=element_text(size=15,color='black'),
                                                                                                                                                                 axis.title=element_text(size=15,color='black'))
    
    print(agexp_plot)
  },  error=function(e){})
  
} #end marker plot list
dev.off()

#--------------------------------------------------------------------------------------#
#Individual markers
pdf(file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/cell_type/cell_type_markers_individual_marker_expression.pdf",height=25,width=25)

for(i in 1:12) {
  
  tryCatch({ #suppress error
    #i=68
    expressed = markers[match(cell_types[i],markers[,1])+1,]
    losgenes = str_split(expressed, "expressed: ")[[1]][2]
    losgenes = as.vector(str_split(losgenes, ", ")[[1]])
    
    #make sure marker genes match - I hecked the non-matching ones, and they do not appear under any alias in the CDS object -- ie these marker genes are not expressed in this cds object!
    print(paste("i = ",i,cell_types[i],sep=" "))
    print(match(losgenes,all_gene_names))
    
    #aggregate gene values to get a single expression value
    marker_df = data.frame(losgenes, str_split( cell_types[i],">")[[1]][2])
    names(marker_df) = c("gene","cell_type")
    
    #Monocle eigene expression
    marker_df$module <- factor(marker_df$cell_type)
    
    agexp_plot = plot_cells(big_cds,
                            genes=marker_df$gene,
                            label_cell_groups=FALSE,
                            label_leaves=F,
                            label_branch_points=F,
                            graph_label_size=1.5,
                            show_trajectory_graph=F) +
      xlab("Dim. 1") + ylab("Dim. 2") + 
      theme(legend.position="bottom") +
      scale_color_viridis_c(option = "inferno",name="") +
      ggtitle(paste(marker_df$cell_type[1], "",sep=" "))+ #continuous
      theme(axis.text=element_text(size=0,color='black'),
            axis.title=element_text(size=0,color='black')) + 
      theme(plot.title = element_text(size = 30, vjust=1)) +
      theme(strip.text = element_text(size = 18))
    
    print(agexp_plot)
  },  error=function(e){})
  
} #end marker plot list
dev.off()

#---------------------------------------------------------------------------------
#1. Print out cell cluster to file

colourCount = length(unique(colData(big_cds)$cluster))
getPalette = colorRampPalette(brewer.pal(9, "Set1"))

png("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/cell_type/cell_cluster_high_resolution.png",width=15,height=15,units="in",res=400)
print(plot_cells(big_cds, color_cells_by="cluster", group_cells_by="cluster",
                 group_label_size = 5,
                 x=1,
                 y=2,
                 alpha=0.2) +
        ggtitle("") + xlab("") + ylab("") + 
        theme(axis.text=element_text(size=0,color='black'),
              axis.title=element_text(size=0),
              axis.ticks = element_line(size=0),
              legend.position="none") +
        scale_color_manual(values=getPalette(colourCount))
)
dev.off()

#---------------------------------------------------------------------------------

#rename various cells and plot out
colData(big_cds)$cluster = big_cds@clusters@listData$UMAP$clusters
colData(big_cds)$partition = big_cds@clusters@listData$UMAP$partitions
unique(colData(big_cds)$cluster)

colData(big_cds)$cell_type = as.character(colData(big_cds)$cluster) #re-add the partition name to change it
colData(big_cds)$cell_type = gsub( "\\b45\\b", "PrE", colData(big_cds)$cell_type,perl=TRUE)
colData(big_cds)$cell_type = gsub( "\\b11\\b", "PrE", colData(big_cds)$cell_type,perl=TRUE)
colData(big_cds)$cell_type = gsub( "\\b22\\b", "IfE", colData(big_cds)$cell_type,perl=TRUE)
colData(big_cds)$cell_type = gsub( "\\b48\\b", "NK", colData(big_cds)$cell_type,perl=TRUE)
colData(big_cds)$cell_type = gsub( "\\b39\\b", "IfE", colData(big_cds)$cell_type,perl=TRUE)
colData(big_cds)$cell_type = gsub( "\\b4\\b", "IfE", colData(big_cds)$cell_type,perl=TRUE)
colData(big_cds)$cell_type = gsub( "\\b20\\b", "IfE", colData(big_cds)$cell_type,perl=TRUE)
colData(big_cds)$cell_type = gsub( "\\b2\\b", "IfE", colData(big_cds)$cell_type,perl=TRUE)
colData(big_cds)$cell_type = gsub( "\\b1\\b", "IfE", colData(big_cds)$cell_type,perl=TRUE)
colData(big_cds)$cell_type = gsub( "\\b32\\b", "IfE", colData(big_cds)$cell_type,perl=TRUE)
colData(big_cds)$cell_type = gsub( "\\b29\\b", "IfE", colData(big_cds)$cell_type,perl=TRUE)
colData(big_cds)$cell_type = gsub( "\\b26\\b", "Bas", colData(big_cds)$cell_type,perl=TRUE)
colData(big_cds)$cell_type = gsub( "\\b38\\b", "Bas", colData(big_cds)$cell_type,perl=TRUE)
colData(big_cds)$cell_type = gsub( "\\b16\\b", "Inf", colData(big_cds)$cell_type,perl=TRUE)
colData(big_cds)$cell_type = gsub( "\\b6\\b", "Inf", colData(big_cds)$cell_type,perl=TRUE)

colData(big_cds)$cell_type = gsub( "\\b8\\b", "UHF", colData(big_cds)$cell_type,perl=TRUE)
colData(big_cds)$cell_type = gsub( "\\b17\\b", "UHF", colData(big_cds)$cell_type,perl=TRUE)

colData(big_cds)$cell_type = gsub( "\\b41\\b", "SG", colData(big_cds)$cell_type,perl=TRUE)
colData(big_cds)$cell_type = gsub( "\\b25\\b", "SG", colData(big_cds)$cell_type,perl=TRUE)

colData(big_cds)$cell_type = gsub( "\\b3\\b", "OuB", colData(big_cds)$cell_type,perl=TRUE)
colData(big_cds)$cell_type = gsub( "\\b40\\b", "OuB", colData(big_cds)$cell_type,perl=TRUE)
colData(big_cds)$cell_type = gsub( "\\b13\\b", "Isth", colData(big_cds)$cell_type,perl=TRUE)
colData(big_cds)$cell_type = gsub( "\\b27\\b", "Isth", colData(big_cds)$cell_type,perl=TRUE)
colData(big_cds)$cell_type = gsub( "\\b49\\b", "InB", colData(big_cds)$cell_type,perl=TRUE)
colData(big_cds)$cell_type = gsub( "\\b30\\b", "unF", colData(big_cds)$cell_type,perl=TRUE)

colData(big_cds)$cell_type = gsub( "\\b18\\b", "nEp", colData(big_cds)$cell_type,perl=TRUE)
colData(big_cds)$cell_type = gsub( "\\b19\\b", "nEp", colData(big_cds)$cell_type,perl=TRUE)
colData(big_cds)$cell_type = gsub( "\\b33\\b", "nEp", colData(big_cds)$cell_type,perl=TRUE)
colData(big_cds)$cell_type = gsub( "\\b5\\b", "nEp", colData(big_cds)$cell_type,perl=TRUE)
colData(big_cds)$cell_type = gsub( "\\b12\\b", "nEp", colData(big_cds)$cell_type,perl=TRUE)
colData(big_cds)$cell_type = gsub( "\\b14\\b", "nEp", colData(big_cds)$cell_type,perl=TRUE)
colData(big_cds)$cell_type = gsub( "\\b9\\b", "nEp", colData(big_cds)$cell_type,perl=TRUE)

colData(big_cds)$cell_type = gsub( "\\b44\\b", "nEp", colData(big_cds)$cell_type,perl=TRUE)
colData(big_cds)$cell_type = gsub( "\\b15\\b", "nEp", colData(big_cds)$cell_type,perl=TRUE)
colData(big_cds)$cell_type = gsub( "\\b28\\b", "nEp", colData(big_cds)$cell_type,perl=TRUE)
colData(big_cds)$cell_type = gsub( "\\b7\\b", "nEp", colData(big_cds)$cell_type,perl=TRUE)
colData(big_cds)$cell_type = gsub( "\\b10\\b", "nEp", colData(big_cds)$cell_type,perl=TRUE)

colData(big_cds)$cell_type = gsub( "\\b34\\b", "End", colData(big_cds)$cell_type,perl=TRUE)

colData(big_cds)$cell_type = gsub( "\\b24\\b", "T", colData(big_cds)$cell_type,perl=TRUE)
colData(big_cds)$cell_type = gsub( "\\b42\\b", "T", colData(big_cds)$cell_type,perl=TRUE)
colData(big_cds)$cell_type = gsub( "\\b43\\b", "T", colData(big_cds)$cell_type,perl=TRUE)

colData(big_cds)$cell_type = gsub( "\\b31\\b", "Mac", colData(big_cds)$cell_type,perl=TRUE)
colData(big_cds)$cell_type = gsub( "\\b21\\b", "LC", colData(big_cds)$cell_type,perl=TRUE)
colData(big_cds)$cell_type = gsub( "\\b23\\b", "Neut", colData(big_cds)$cell_type,perl=TRUE)

colData(big_cds)$cell_type = gsub( "\\b35\\b", "Er", colData(big_cds)$cell_type,perl=TRUE)
colData(big_cds)$cell_type = gsub( "\\b36\\b", "Mel", colData(big_cds)$cell_type,perl=TRUE)
colData(big_cds)$cell_type = gsub( "\\b37\\b", "MyFi", colData(big_cds)$cell_type,perl=TRUE)

colData(big_cds)$cell_type = gsub( "\\b46\\b", "", colData(big_cds)$cell_type,perl=TRUE)
colData(big_cds)$cell_type = gsub( "\\b47\\b", "", colData(big_cds)$cell_type,perl=TRUE)

unique(colData(big_cds)$cell_type)

PrE = Proliferating Epithelia
IfE = Interfollicular Epithelia
NK = NK cells of SALT
UHF = Upper Hair Follicle
SG = Sebaceous Gland
OuB = Outer bulge
InB = Inner bulge
End = Endothelial
T = T cells
Mac = Macrophage
LC = Langerhans Cell
Neut = Neutrophil
Inf = Infundibulum
Isth = Isthmus
Bas = Basal cells
UnF = Undifferentiated Follicular
nEp = Neoplastic epithelia (sort of in-between follicular and interfollicular)
Er = Erythrocyte
Mel = Melanocyte
MyFi = Myofibroblast

#=====================================================================================#
#Calculate how many undefined cells are there
subdat = subset(colData(big_cds), (cluster == 46 | cluster == 47))
nrow(subdat) / nrow(colData(big_cds))
nrow(colData(big_cds)) - nrow(subdat)#number of cells successfully classified
1- (nrow(subdat) / nrow(colData(big_cds)))

#=====================================================================================#
#Now combine cell types with upper spike, lower spike, squamous, spindle, lower bridge, middle bridge, upper bridge

colData(big_cds)$integrated_type_bridge = colData(big_cds)$cell_type #add column of of apparent no matches to ALL the cells; later you will replace with matches
e <- readRDS(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/matvei/cds_objects/squamous.RDS",sep=""))
f <- readRDS(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/matvei/cds_objects/spindle.RDS",sep=""))
squamous_matches = match(colnames(e),colnames(big_cds)) #match extracted cells with all cell names
spindle_matches = match(colnames(f),colnames(big_cds)) #match extracted cells with all cell names

colData(big_cds)$integrated_type_bridge[squamous_matches] <- "squamous" #now replace matching cells with new name
colData(big_cds)$integrated_type_bridge[spindle_matches] <- "spindle" #now replace matching cells with new name

#Cluster 19
cds_subset.v3 = big_cds[,colData(big_cds)$cluster %in% c("19")] #filter for cluster 19 because most of the bridge cells are only in cluster 19
matching_cells = match(colnames(cds_subset.v3),colnames(big_cds)) #match extracted cells with all cell names
colData(big_cds)$integrated_type_bridge[matching_cells] <- "upper spike" #now replace matching cells with new name

#Cluster 33
cds_subset.v3 = big_cds[,colData(big_cds)$cluster %in% c("33")] #filter for cluster 19 because most of the bridge cells are only in cluster 19
matching_cells = match(colnames(cds_subset.v3),colnames(big_cds)) #match extracted cells with all cell names
colData(big_cds)$integrated_type_bridge[matching_cells] <- "lower spike" #now replace matching cells with new name

#Bridge cells CHOSEN BELOW
upper_bridge = read.table(file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/figures/pap_ca/pap_ca_trajectory_segment/upper_bridge_cell_names_v1.txt",header=T)
middle_bridge = read.table(file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/figures/pap_ca/pap_ca_trajectory_segment/middle_bridge_cell_names_v1.txt",header=T)
lower_bridge = read.table(file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/figures/pap_ca/pap_ca_trajectory_segment/lower_bridge_cell_names_v1.txt",header=T)

matching_cells = match(upper_bridge[,1],colnames(big_cds)) #match extracted cells with all cell names
colData(big_cds)$integrated_type_bridge[matching_cells] <- "Upper Bridge" #now replace matching cells with new name

matching_cells = match(middle_bridge[,1],colnames(big_cds)) #match extracted cells with all cell names
colData(big_cds)$integrated_type_bridge[matching_cells] <- "Middle Bridge" #now replace matching cells with new name

matching_cells = match(lower_bridge[,1],colnames(big_cds)) #match extracted cells with all cell names
colData(big_cds)$integrated_type_bridge[matching_cells] <- "Lower Bridge" #now replace matching cells with new name

#set color palette this is NICE
colourCount = length(unique(colData(big_cds)$integrated_type_bridge))
getPalette = colorRampPalette(brewer.pal(9, "Set1"))

png("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Figs/Fig1/Fig1H_with_labels.png",width=7,height=7,units="in",res=400)
print(plot_cells(big_cds, color_cells_by="integrated_type_bridge", group_cells_by="cluster",
                 group_label_size = 3,
                 x=1,
                 y=2,
                 alpha=0.2) +
        ggtitle("") + xlab("") + ylab("") + 
        theme(axis.text=element_text(size=0,color='black'),
              axis.title=element_text(size=0),
              title=element_text(size=0),
              axis.ticks = element_line(size=0),
              legend.position="none") +
        scale_color_manual(values=getPalette(colourCount))
)
dev.off()

png("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Figs/Fig1/Fig1H_without_labels.png",width=7,height=7,units="in",res=400)
print(plot_cells(big_cds, color_cells_by="integrated_type_bridge", group_cells_by="cluster",
                 group_label_size = 0,
                 x=1,
                 y=2,
                 alpha=0.2) +
        ggtitle("") + xlab("") + ylab("") + 
        theme(axis.text=element_text(size=0,color='black'),
              axis.title=element_text(size=0),
              title=element_text(size=0),
              axis.ticks = element_line(size=0),
              legend.position="none") +
        scale_color_manual(values=getPalette(colourCount))
)
dev.off()



#=====================================================================================#
#Because bridge cells over-write some upper/lower spike cells with the bridge type, reconsitute the complete SPIKE CELL ANNOTATIONS
colData(big_cds)$integrated_celltype_spike = colData(big_cds)$cell_type 

#Cluster 19
cds_subset.v3 = big_cds[,colData(big_cds)$cluster %in% c("19")] #filter for cluster 19 because most of the bridge cells are only in cluster 19
matching_cells = match(colnames(cds_subset.v3),colnames(big_cds)) #match extracted cells with all cell names
colData(big_cds)$integrated_celltype_spike[matching_cells] <- "upper spike" #now replace matching cells with new name

#Cluster 33
cds_subset.v3 = big_cds[,colData(big_cds)$cluster %in% c("33")] #filter for cluster 19 because most of the bridge cells are only in cluster 19
matching_cells = match(colnames(cds_subset.v3),colnames(big_cds)) #match extracted cells with all cell names
colData(big_cds)$integrated_celltype_spike[matching_cells] <- "lower spike" #now replace matching cells with new name

#Add squamous and spindle phases to the carcinoma
e <- readRDS(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/matvei/cds_objects/squamous.RDS",sep=""))
f <- readRDS(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/matvei/cds_objects/spindle.RDS",sep=""))
squamous_matches = match(colnames(e),colnames(big_cds)) #match extracted cells with all cell names
spindle_matches = match(colnames(f),colnames(big_cds)) #match extracted cells with all cell names

colData(big_cds)$integrated_celltype_spike[squamous_matches] <- "squamous" #now replace matching cells with new name
colData(big_cds)$integrated_celltype_spike[spindle_matches] <- "spindle" #now replace matching cells with new name

#=====================================================================================#
#Revision Fig : Proportion of FACS sorting Lgr6 Progeny within PAPILLOMA CLUSTERS
#NORMAILZE BY THE NUMBER OF CELLS ACTUALLY PASSING QC IN EACH SAMPLE
#=====================================================================================#
cluster_tissue_df = as.data.frame(subset(colData(big_cds), select=c("cluster","sorting"))) #extract partition and tissue
cluster_tissue_df_sub = cluster_tissue_df[cluster_tissue_df$cluster %in% c("14","9","12","5","33","19","18"),] #These are all the unsupervised cluster in the papillom parenchyma

#At baseline here are the percentages of sorting cells passing QC in th pap pareenchyma
#Bc remember there are DIFFERENT numbers of Lgr6, Progeny, and Unsorted cells being considered despite 10k being submitted for sequencing
baseline_percents = round((table(cluster_tissue_df_sub$sorting) / nrow(cluster_tissue_df_sub) )*100, 2)

#Test whether a cluster's proportion is signfiicantly different from all the others

cluster_tissue_df = as.data.frame(subset(colData(big_cds), select=c("cluster","sorting","tissue"))) #extract partition and tissue
cluster_tissue_df_sub = cluster_tissue_df[cluster_tissue_df$cluster %in% c("14","9","12","5","33","19","18"),] #These are all the unsupervised cluster in the papillom parenchyma

table(cluster_tissue_df_sub$cluster,cluster_tissue_df_sub$tissue)

#Cluster-specific percents
pap_clusts = c("14","9","12","5","33","19","18")

clust_enrichment_scores = do.call(rbind, lapply(1:length(pap_clusts), function(i){
  #i=1
  print(i)
  
  #Get proportion in specific cluster
  pap_clust = cluster_tissue_df_sub[cluster_tissue_df_sub$cluster %in% pap_clusts[i],] #Subset to a specific cluster
  cluster_percents = (table(pap_clust$sorting) / nrow(pap_clust) )*100 #Calculate percentages of each sorting factor level within that cluster
  
  #Get proportions in all other clusters
  other_clusts = cluster_tissue_df_sub[cluster_tissue_df_sub$cluster %!in% pap_clusts[i],] #Subset to all other clusters
  allclust_percents = (table(other_clusts$sorting) / nrow(other_clusts) )*100 #Calculate percentages of each sorting factor level within that cluster
  
  #Now calculate RELATIVE ENRICHMENT for a cluster (relative to the total number of cells passing QC in the papilloma among ALL clusters)
  enrichment_df = unlist(round( cluster_percents / allclust_percents , 3 ))
  return_df = data.frame(pap_clusts[i],enrichment_df[1],enrichment_df[2],enrichment_df[3])
  names(return_df) = c("cluster","Lgr6","Progeny","Unsorted")
  
  #2-Proportions Z-test
  papclust_lgr6_n = nrow(pap_clust[pap_clust$sorting %in% c("Lgr6"),]) #Get number of Lgr6 in the papilloma cluster
  papclust_progeny_n = nrow(pap_clust[pap_clust$sorting %in% c("Progeny"),]) #Get number of Progeny in the papilloma cluster
  other_clusts_lgr6_n = nrow(other_clusts[other_clusts$sorting %in% c("Lgr6"),])
  other_clusts_progeny_n = nrow(other_clusts[other_clusts$sorting %in% c("Progeny"),])
  
  res = prop.test(c(papclust_lgr6_n,papclust_progeny_n), 
                  c(other_clusts_lgr6_n,other_clusts_progeny_n), 
                  p = NULL, alternative = "two.sided",
                  correct = TRUE)
  
  #Combine enrichment score with 2 Proportion Z-test
  return_df = data.frame(return_df, res$statistic, res$p.value)
  
}))
clust_enrichment_scores$p_adj = p.adjust(clust_enrichment_scores$res.p.value, method="BH") #adjust p-value
clust_enrichment_scores_towriteout = clust_enrichment_scores
#--------------------------------#
#Now plot out enrichment scores
#--------------------------------#
clust_enrichment_scores = clust_enrichment_scores[,c(1:3)] #Just get Lgr6 and Progeny
dfm = melt(clust_enrichment_scores, id=c("cluster")) #Melt so can be plotted
names(dfm)[2] = "sorting"

level_order <- factor(dfm$cluster, level = c("14","9","12","5","19","18","33"))
png("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Nature/Round\ 2/Figs/Revised\ Fig/FACS_propotion_papclusters.png",
    width=6,height=6,units="in",res=250)
print(
  ggplot(dfm, aes(x=level_order, y=value, fill=sorting))+
    geom_bar(stat="identity", position="dodge")+
    ggtitle("") + xlab("Cell Cluster") + ylab("Relative Enrichment") +
    scale_fill_manual(values = c("Lgr6" = "burlywood4","Progeny" = "olivedrab"), name="FACS") + 
    theme(legend.position="bottom") + theme_classic() +
    theme(axis.text=element_text(size=20,color='black'),
          axis.title=element_text(size=20,color='black'),
          legend.title=element_text(size=20), 
          legend.text=element_text(size=20)) +
    geom_hline(yintercept = 1, color="black",linetype = "dashed") # Add horizontal line at base mean
)
dev.off()

#Write out results
names(clust_enrichment_scores_towriteout) = c("cluster","Lgr6_RE","Progeny_RE","Unsorted_RE",
                                              "Chi_sq","p_raw","p_adj_BH")
write.table(clust_enrichment_scores_towriteout,
            file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Nature/Round\ 2/Figs/Revised\ Fig/FACS_propotion_papclusters_quantresults.txt",
            quote=F, row.names=F, sep="\t")


#=====================================================================================#
#Extract Lrig1 and Lgr6 data
count_mat = big_cds@assays@data$counts
coutn_mat_sub = t(as.data.frame(count_mat[row.names(count_mat) %in% c("Lgr6","Lrig1"),]))
coutn_mat_sub = data.frame(coutn_mat_sub,data.frame(colData(big_cds))) #add metadata to the counts

integrated_types = unique(colData(big_cds)$integrated_type_bridge)

counts_positive = do.call(rbind, lapply(1:length(integrated_types), function(i){
  #i=1
  lesub = subset(coutn_mat_sub, integrated_type_bridge==integrated_types[i])
  N_total_cells = nrow(lesub)
  N_Lrig1 = nrow(subset(lesub, Lrig1>0)) #num Lrig+ cells
  N_Lgr6 = nrow(subset(lesub, Lgr6>0)) #num Lrig+ cells
  N_dblpos = nrow(subset(lesub, (Lgr6>0 & Lrig1>0))) #num Lrig+ cells
  
  Percent_Lgr6_pos = round((N_Lgr6/N_total_cells)*100,2)
  Percent_Lrig1_pos = round((N_Lrig1/N_total_cells)*100,2)
  Percent_Lgr6_dblpos = round((N_dblpos/N_Lgr6)*100,2)
  Percent_Lrig1_dblpos = round((N_dblpos/N_Lrig1)*100,2)
  
  data.frame(integrated_types[i],N_Lgr6,N_Lrig1,N_dblpos,Percent_Lgr6_pos,Percent_Lrig1_pos,Percent_Lgr6_dblpos,Percent_Lrig1_dblpos)
}))
names(counts_positive)[1] = "cell group"

write.table(counts_positive, file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/cell_type/lgr6_lrig1_pos.txt",
            quote=F,row.names=F,sep="\t")

#=====================================================================================#
#Double positive within sorting stage

sortings = unique(coutn_mat_sub$sorting)
counts_stage = do.call(rbind, lapply(1:length(sortings), function(j){
  #j=3
  
  lesub_stage = subset(coutn_mat_sub, sorting==sortings[j])
  
  counts_positive = do.call(rbind, lapply(1:length(integrated_types), function(i){
    #i=1
    lesub = subset(lesub_stage, integrated_type_bridge==integrated_types[i])
    N_total_cells = nrow(lesub)
    N_Lrig1 = nrow(subset(lesub, Lrig1>0)) #num Lrig+ cells
    N_Lgr6 = nrow(subset(lesub, Lgr6>0)) #num Lgr6+ cells
    N_dblpos = nrow(subset(lesub, (Lgr6>0 & Lrig1>0))) #num Lrig+ cells
    
    Percent_Lgr6_pos = round((N_Lgr6/N_total_cells)*100,2)
    Percent_Lrig1_pos = round((N_Lrig1/N_total_cells)*100,2)
    Percent_Lgr6_dblpos = round((N_dblpos/N_Lgr6)*100,2)
    Percent_Lrig1_dblpos = round((N_dblpos/N_Lrig1)*100,2)
    
    data.frame(sortings[j],integrated_types[i],N_Lgr6,N_Lrig1,N_dblpos,Percent_Lgr6_pos,Percent_Lrig1_pos,Percent_Lgr6_dblpos,Percent_Lrig1_dblpos)
  }))
  names(counts_positive)[1:2] = c("sorting","cell group")
  counts_positive
}))

write.table(counts_stage, file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/cell_type/lgr6_lrig1_pos_SORTING.txt",
            quote=F,row.names=F,sep="\t")

#=====================================================================================#
#Subset cells save as CDS objets
#=====================================================================================#
#choose main set of skin cells
cds.skin = choose_cells(big_cds)
#write UnTxSkin main set to file
cds.UnTxSkin = cds.skin[,colData(cds.skin)$tissue %in% c("UnTxSkin")]
saveRDS(cds.UnTxSkin, file = "/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/matvei/cds_objects/UnTxSkin.RDS") 
#write TxSkin main set to file
cds.TxSkin = cds.skin[,colData(cds.skin)$tissue %in% c("TxSkin")]
saveRDS(cds.TxSkin, file = "/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/matvei/cds_objects/TxSkin.RDS") 

#choose main set of Pap cells
cds.sub = choose_cells(big_cds)
cds.sub = cds.sub[,colData(cds.sub)$tissue %in% c("Pap")]
#write to file
saveRDS(cds.sub, file = "/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/matvei/cds_objects/Pap.RDS") 

#choose main set of CA cells
cds.sub = choose_cells(big_cds)
cds.sub = cds.sub[,colData(cds.sub)$tissue %in% c("CA")]
#write to file
saveRDS(cds.sub, file = "/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/matvei/cds_objects/CA.RDS") 

#choose TransitionalPap cells subset to Pap
cds.sub = choose_cells(big_cds)
cds.sub = cds.sub[,colData(cds.sub)$tissue %in% c("Pap")]
#write to file
saveRDS(cds.sub, file = "/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/matvei/cds_objects/TransitionalPap.RDS") 

#choose TransitionalCA cells from Papilloma bridge + main cluster
cds.sub = choose_cells(big_cds)
cds.sub = cds.sub[,colData(cds.sub)$tissue %in% c("CA")]
#write to file
saveRDS(cds.sub, file = "/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/matvei/cds_objects/TransitionalCA.RDS") 

#choose T cells
cds.sub = choose_cells(big_cds)
saveRDS(cds.sub, file = "/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/matvei/cds_objects/T-cells.RDS") 

#choose innate immune cells
cds.sub = choose_cells(big_cds)
saveRDS(cds.sub, file = "/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/matvei/cds_objects/InnateImmunecells.RDS") 

#choose Macrophages
cds.sub = choose_cells(ii_cds) #ii_cds is created below in explore innate immune cells section
saveRDS(cds.sub, file = "/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/matvei/cds_objects/Macrophage.RDS") 

#choose Langerhans
cds.sub = choose_cells(ii_cds) #ii_cds is created below in explore innate immune cells section
saveRDS(cds.sub, file = "/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/matvei/cds_objects/Langerhans.RDS") 

#choose Macrophages
cds.sub = choose_cells(ii_cds) #ii_cds is created below in explore innate immune cells section
saveRDS(cds.sub, file = "/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/matvei/cds_objects/Neutrophil.RDS") 

#choose Spindle phase tumor
cds.sub = choose_cells(ca_only) #ii_cds is created below in explore innate immune cells section
saveRDS(cds.sub, file = "/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/matvei/cds_objects/Spindle.RDS") 

#choose Squamous phase tumor
cds.sub = choose_cells(ca_only) #ii_cds is created below in explore innate immune cells section
saveRDS(cds.sub, file = "/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/matvei/cds_objects/Squamous.RDS") 

#=====================================================================================#
#Create new combined parenchyma cds = cds_comb
#=====================================================================================#
#cds_comb represents the tissue parenchyma: paritions 1-4

cds_comb = big_cds[,colData(big_cds)$partition %in% c(1,2,3,4)]

#=====================================================================================#
#T cell analysis
#=====================================================================================#
#Explore these T cells in more depth --> run analyzer on them!
t_cds <- readRDS("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/matvei/cds_objects/T-cells.RDS")

#3b. Pre-process with Seurat-processed object
t_cds <- preprocess_cds(t_cds, 
                        method="PCA",
                        num_dim = 100,
                        norm_method = "none", #these data have already been normalized by Seurat
                        alignment_group = NULL)


#Re-align to remove batch effects: batch is covarying strongly with Unsorted ?align_cds
t_cds = align_cds(cds = t_cds,
                  num_dim = 100,
                  preprocess_method = "PCA",
                  alignment_group = "batch_name",
                  verbose=T)

#4b. ?reduce_dimension
t_cds <- reduce_dimension(
  t_cds,
  max_components = 2,
  reduction_method = c("UMAP"),
  preprocess_method = "Aligned", #must now used ALIGNEMNT PRE-PROCESSING METHOD bc it replaces the other pre-process shit
  umap.metric = "cosine",
  umap.min_dist = 0.1,
  umap.n_neighbors = 15L,
  umap.fast_sgd = FALSE,
  umap.nn_method = "annoy",
  cores = 1,
  verbose = FALSE
)

#5. Cluster and partition ?cluster_cells cells into supergroups community detection (Phenograph algorithm) Community detection of cells as part of Phenograph algorithm https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4508757/
#5b. 
t_cds <- cluster_cells(t_cds,
                       reduction_method = c("UMAP"),
                       k = 10, #a bigger k will result in lower resolution 
                       cluster_method = c("leiden"),
                       num_iter = 2,
                       partition_qval = 0.05, #Numeric, the q-value cutoff to determine when to partition
                       weight = FALSE,
                       resolution = NULL,
                       random_seed = NULL,
                       verbose = F)

#T-cell tissue
#Make a New Tissue specificity columns for Normal skin (UnTxSkin and TxSkin), Pap, and CA
colData(t_cds)$tissue_updated = colData(t_cds)$tissue
colData(t_cds)$tissue_updated = gsub( "\\bUnTxSkin\\b", "Normal skin", colData(t_cds)$tissue_updated,perl=TRUE)
colData(t_cds)$tissue_updated = gsub( "\\bTxSkin\\b", "Normal skin", colData(t_cds)$tissue_updated,perl=TRUE)

table(colData(t_cds)$tissue)

t_cell_tissue_plot = plot_cells(t_cds, color_cells_by="tissue_updated", group_cells_by="cluster",
                                group_label_size = 0,
                                x=1,
                                y=2,
                                cell_size=1.5) +
  ggtitle("") + xlab("") + ylab("") + 
  theme(axis.text=element_text(size=0,color='black'),
        axis.title=element_text(size=0),
        axis.ticks = element_line(size=0),
        legend.position="none") +
  scale_color_manual(values = c("Normal skin" = "darkblue", "Pap" = "mediumpurple2","CA" = "darkred"))


#Rasa2 expression in t cells
#replace correctly
t_cds <- readRDS("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/matvei/cds_objects/T-cells.RDS")
colData(t_cds)$tissue = gsub("UnTxSkin","1.Normal skin",colData(t_cds)$tissue)
colData(t_cds)$tissue = gsub("TxSkin","2.Inflamed skin",colData(t_cds)$tissue)
colData(t_cds)$tissue = gsub("Pap","3.Papilloma",colData(t_cds)$tissue)
colData(t_cds)$tissue = gsub("CA","4.Carcinoma",colData(t_cds)$tissue)

#combine gene value
agg_gene_clindat = data.frame(t(t_cds["Rasa2",]@assays@data$counts),colData(t_cds))
names(agg_gene_clindat)[1] = c("eigengene_expression")

ggplot(agg_gene_clindat, aes(x=tissue,y=eigengene_expression, fill=tissue)) +
  #stat_compare_means(comparisons = my_comparisons) +
  geom_violin(aes(fill=tissue), width=1) +
  geom_jitter(width = 0.15) +  #geom_boxplot(outlier.size=0, width=0.2, color="black") +
  #geom_bar(stat = "identity", fill="darkorange") +
  #facet_wrap(~parenchyma, scales="free", nrow=1) + 
  #geom_jitter(width = 0.1) +
  ggtitle("Rasa2 T-cell expression") + 
  xlab("") + ylab("log expression") + 
  theme(axis.text.x=element_text(size=25,color='black'),
        axis.text.y=element_text(size=25,color='black'),
        axis.title=element_text(size=25),
        axis.ticks.x = element_line(size=0),
        strip.text.x = element_text(size = 0),
        legend.position="none",
        title = element_text(size = 25, face="bold")) +
  theme(
    # Hide panel borders and remove grid lines
    panel.border = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.background = element_blank(),
    element_line(colour = 'black', size = 0.3),
    # Change axis line
    axis.line = element_line(colour = "black"),
    strip.background = element_rect(colour="white",
                                    fill="white")
  ) +
  scale_fill_manual(values = c("1.Normal skin" = "lightblue",
                               "2.Inflamed skin" = "blue",
                               "3.Papilloma" = "thistle3", 
                               "4.Carcinoma" = "red")) + theme(legend.position="none")

#=====================================================================================#
#Extended Data Fig.1e : T-cell manifold
#=====================================================================================#
png("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Extended\ Data/Extended\ Data\ Fig\ 1/ExtDat_Fig1e.png",
    width=7,height=7,units="in",res=200)
print( t_cell_tissue_plot
)
dev.off()


#-----------------------------------------------------------
#Find markers between the 2 main groups of T-cells
cds.sub.neoplasm = choose_cells(t_cds) #choose neoplastic partition of T-cells
cds.sub.non.neoplasm = choose_cells(t_cds) #choose non-neoplastic partition of T-cells

dim(cds.sub.neoplasm)[2] + dim(cds.sub.non.neoplasm)[2] == dim(t_cds)[2] #make sure i've captured all the cells

#add group based on cell id
colData(t_cds)$t_cell_group = ifelse(colnames(t_cds) %in% colnames(cds.sub.neoplasm), "Neoplasm group", "Normal skin group")

#plot neoplasm vs non-neplasm group
t_cell_neoplasm_group_plot = plot_cells(t_cds, color_cells_by="t_cell_group", group_cells_by="cluster",
                                        group_label_size = 5,
                                        x=1,
                                        y=2,
                                        cell_size=1.5) +
  ggtitle("") + xlab("") + ylab("") + 
  theme(axis.text=element_text(size=0,color='black'),
        axis.title=element_text(size=0),
        axis.ticks = element_line(size=0))

png("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/t_cell_analysis/Fig_t_cell_neoplasm_group.png",width=7,height=7,units="in",res=400)
print( t_cell_neoplasm_group_plot
)
dev.off()

#marker_test_res on activated vs non-activated T-cells
marker_test_res <- top_markers(t_cds, group_cells_by="t_cell_group",
                               genes_to_test_per_group = 200,
                               cores=8)
marker_test_res = marker_test_res[order(marker_test_res$marker_test_q_value),] #re-order low to high q-value
write.table(marker_test_res, file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/t_cell_analysis/t_cell_top_markers.txt", quote=F,row.names = F,sep="\t")

#marker_test_res on tissue type
marker_test_res <- top_markers(t_cds, group_cells_by="tissue_updated",
                               genes_to_test_per_group = 200,
                               cores=8)
marker_test_res = marker_test_res[order(marker_test_res$marker_test_q_value),] #re-order low to high q-value
write.table(marker_test_res, file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/t_cell_analysis/t_cell_by_tissue_top_markers.txt", quote=F,row.names = F,sep="\t")


#-----------------------------------------------------------------------------------------------------------------#
#-----------------------------------------------------------------------------------------------------------------#

#look at immune cell classifiers

pdf(file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/figures/cell_clusters/T-cell_markers_v2.pdf",height=8,width=12)

for(i in 15:22) { #15-18 are the T cell types!
  tryCatch({ #suppress error
    #i=15
    expressed = markers[match(cell_types[i],markers[,1])+1,]
    losgenes = str_split(expressed, "expressed: ")[[1]][2]
    losgenes = as.vector(str_split(losgenes, ", ")[[1]])
    
    #make sure marker genes match - I hecked the non-matching ones, and they do not appear under any alias in the CDS object -- ie these marker genes are not expressed in this cds object!
    print(paste("i = ",i,cell_types[i],sep=" "))
    print(match(losgenes,all_gene_names))
    
    marker_plot = plot_cells(t_cds, 
                             genes=losgenes, 
                             group_cells_by="cluster",
                             group_label_size = 0,
                             cell_size=1.5,
                             x=1,
                             y=2,
                             label_cell_groups=FALSE,
                             label_leaves=F,
                             label_branch_points=F,
                             show_trajectory_graph=F) +
      ggtitle(str_split( cell_types[i],">")[[1]][2]) + xlab("Dim. 1") + ylab("Dim. 2") +
      theme(legend.position="bottom") +
      theme(strip.text.x = element_text(size = 16))+
      scale_color_viridis_c(option = "inferno",name="expression") + #continuous
      theme(axis.text=element_text(size=18,color='black'),
            axis.title=element_text(size=20,color='black')) + theme(plot.title = element_text(size = 22, vjust=1)) + theme(axis.text=element_text(size=15,color='black'),
                                                                                                                           axis.title=element_text(size=15,color='black'))
    print(marker_plot)
    
    #aggregate gene values to get a single expression value
    marker_df = data.frame(losgenes, str_split( cell_types[i],">")[[1]][2])
    names(marker_df) = c("gene","cell_type")
    agg_mat = aggregate_gene_expression(
      t_cds,
      gene_group_df = marker_df,
      cell_group_df = NULL,
      norm_method = c("log"),
      pseudocount = 1,
      scale_agg_values = TRUE,
      max_agg_value = 5,
      min_agg_value = -5,
      exclude.na = TRUE
    )
    
    #Monocle eigene expression
    marker_df$module <- factor(marker_df$cell_type)
    
    agexp_plot = plot_cells(t_cds,
                            genes=marker_df,
                            cell_size=1.5,
                            label_cell_groups=FALSE,
                            label_leaves=F,
                            label_branch_points=F,
                            graph_label_size=1.5,
                            show_trajectory_graph=F) +
      xlab("Dim. 1") + ylab("Dim. 2") + 
      theme(legend.position="bottom") +
      theme(strip.text.x = element_text(size = 16))+
      scale_color_viridis_c(option = "inferno",name="aggregated expression") +
      ggtitle(paste(marker_df$cell_type[1], "aggregated expression",sep=" ")) + #continuous
      theme(axis.text=element_text(size=18,color='black'),
            axis.title=element_text(size=20,color='black')) + theme(plot.title = element_text(size = 22, vjust=1)) + theme(axis.text=element_text(size=15,color='black'),
                                                                                                                           axis.title=element_text(size=15,color='black'))
    
    print(agexp_plot)
  },  error=function(e){})
  
} #end marker plot list
dev.off()

#track T-cell load over time!
tissues = c("Normal skin","Pap","CA")
colData(cds_comb)$tissue = gsub( "\\bUnTxSkin\\b", "Normal skin", colData(cds_comb)$tissue,perl=TRUE)
colData(cds_comb)$tissue = gsub( "\\bTxSkin\\b", "Normal skin", colData(cds_comb)$tissue,perl=TRUE)

tload = do.call(rbind, lapply(1:length(tissues), function(i){
  #i=2
  t.sub = t_cds[,colData(t_cds)$tissue_updated %in% c(tissues[i])] #subset to tissue-specific amount of T-cells
  a = length(colData(cds_comb)$tissue[colData(cds_comb)$tissue %in% tissues[i]]) #get number of cells in tissue-specific parenchyma
  
  th = data.frame(t(as.data.frame(as.matrix(t.sub[match("Cd4",as.vector(rowData(t.sub)[,1])),]@assays@data@listData$counts)))) #isolate counts for genes
  th_num = nrow(subset(th, th$Cd4>0)) #find the number of cells expressing this over
  
  tc = data.frame(t(as.data.frame(as.matrix(t.sub[match(c("Cd8a","Cd8b1"),as.vector(rowData(t.sub)[,1])),]@assays@data@listData$counts)))) #isolate counts for genes
  tc_num = nrow(subset(tc, (tc$Cd8a>0 | tc$Cd8b1>0))) #find the number of cells expressing this over
  
  treg = data.frame(t(as.data.frame(as.matrix(t.sub[match(c("Foxp3"),as.vector(rowData(t.sub)[,1])),]@assays@data@listData$counts)))) #isolate counts for genes
  treg_num = nrow(subset(treg, (treg$Foxp3>0))) #find the number of cells expressing this over
  
  tprop = round((dim(t.sub)[2]/a)*100,3) # proportions
  th_prop = round((th_num/a)*100,3) # proportions
  tc_prop = round((tc_num/a)*100,3) # proportions
  treg_prop = round((treg_num/a)*100,3) # proportions
  
  data.frame(tissues[i],dim(t.sub)[2],th_num,tc_num,treg_num,tprop,th_prop,tc_prop,treg_prop)
  
}))

names(tload) = c("tissue","T-cell no.","Th no.","Tc no.","Treg no.","T-cell fraction","Th fraction","Tc fraction","Treg fraction")
toload.dfm = melt(tload, id=c("tissue"))

tload_plot = ggplot(data=toload.dfm, aes(x= factor(tissue, level = c('Normal skin', 'Pap', 'CA')) ,y=value,fill=tissue,color=tissue)) +
  geom_bar(stat="identity") +
  facet_wrap(~variable, ncol=4, scales="free") +
  xlab("tissue") + ylab("%                            N") + 
  theme(legend.position="bottom") +
  theme(strip.text.x = element_text(size = 16))+
  ggtitle("T cell load") + #continuous
  scale_color_manual(values = c("Normal skin" = "darkblue", "Pap" = "mediumpurple2","CA" = "darkred")) +
  scale_fill_manual(values = c("Normal skin" = "darkblue", "Pap" = "mediumpurple2","CA" = "darkred")) + theme(legend.position="bottom") +
  theme(axis.text=element_text(size=18,color='black'),
        axis.title=element_text(size=20,color='black')) + theme(plot.title = element_text(size = 22, vjust=1)) + theme(axis.text=element_text(size=15,color='black'),
                                                                                                                       axis.title=element_text(size=15,color='black')) +
  theme(text = element_text(size=18)) +
  theme(axis.text.y=element_text(size=17,color='black'),
        axis.text.x=element_text(size=9,color='black'),
        axis.title.x=element_text(size=20,color='black'),
        axis.title.y=element_text(size=22,color='black')) +
  theme(plot.title = element_text(size = 20, vjust=1)) +
  theme(axis.text=element_text(size=15,color='black'),
        axis.title=element_text(size=15,color='black')) +
  theme(panel.grid.major=element_line(colour = "darkgrey"), 
        panel.grid.minor=element_blank(), 
        panel.background=element_blank()) +
  theme(legend.key = element_rect(fill = "white")) +
  theme(strip.text.x = element_text(size = 14, colour = "black", angle = 0)) +
  theme(strip.text.y = element_text(size = 14, colour = "black", angle = 0)) +
  theme(legend.text=element_text(size=14), legend.title=element_text(size=14)) +
  theme(strip.background=element_rect(fill="white"))
print(tload_plot)

#=====================================================================================#
#Innate Immune Cell Analysis
#=====================================================================================#
ii_cds <- readRDS("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/matvei/cds_objects/InnateImmunecells.RDS")

#3b. Pre-process with Seurat-processed object
ii_cds <- preprocess_cds(ii_cds, 
                         method="PCA",
                         num_dim = 100,
                         norm_method = "none", #these data have already been normalized by Seurat
                         alignment_group = NULL)


#Re-align to remove batch effects: batch is covarying strongly with Unsorted ?align_cds
ii_cds = align_cds(cds = ii_cds,
                   num_dim = 100,
                   preprocess_method = "PCA",
                   alignment_group = "batch_name",
                   verbose=T)

#4b. ?reduce_dimension
ii_cds <- reduce_dimension(
  ii_cds,
  max_components = 2,
  reduction_method = c("UMAP"),
  preprocess_method = "Aligned", #must now used ALIGNEMNT PRE-PROCESSING METHOD bc it replaces the other pre-process shit
  umap.metric = "cosine",
  umap.min_dist = 0.1,
  umap.n_neighbors = 15L,
  umap.fast_sgd = FALSE,
  umap.nn_method = "annoy",
  cores = 1,
  verbose = FALSE
)

#5. Cluster and partition ?cluster_cells cells into supergroups community detection (Phenograph algorithm) Community detection of cells as part of Phenograph algorithm https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4508757/
#5b. 
ii_cds <- cluster_cells(ii_cds,
                        reduction_method = c("UMAP"),
                        k = 10, #a bigger k will result in lower resolution 
                        cluster_method = c("leiden"),
                        num_iter = 2,
                        partition_qval = 0.05, #Numeric, the q-value cutoff to determine when to partition
                        weight = FALSE,
                        resolution = NULL,
                        random_seed = NULL,
                        verbose = F)

#Innate immune tissue
#Make a New Tissue specificity columns for Normal skin (UnTxSkin and TxSkin), Pap, and CA
colData(ii_cds)$tissue_updated = colData(ii_cds)$tissue
colData(ii_cds)$tissue_updated = gsub( "\\bUnTxSkin\\b", "Normal skin", colData(ii_cds)$tissue_updated,perl=TRUE)
colData(ii_cds)$tissue_updated = gsub( "\\bTxSkin\\b", "Normal skin", colData(ii_cds)$tissue_updated,perl=TRUE)

#look at immune cell classifiers

pdf(file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/figures/cell_clusters/Innate_immune_cell_markers_v1.pdf",height=8,width=12)

for(i in 25:32) { #25-32 are some innate immune types
  tryCatch({ #suppress error
    #i=25
    expressed = markers[match(cell_types[i],markers[,1])+1,]
    losgenes = str_split(expressed, "expressed: ")[[1]][2]
    losgenes = as.vector(str_split(losgenes, ", ")[[1]])
    
    #make sure marker genes match - I hecked the non-matching ones, and they do not appear under any alias in the CDS object -- ie these marker genes are not expressed in this cds object!
    print(paste("i = ",i,cell_types[i],sep=" "))
    print(match(losgenes,all_gene_names))
    
    marker_plot = plot_cells(ii_cds, 
                             genes=losgenes, 
                             group_cells_by="cluster",
                             group_label_size = 0,
                             cell_size=1,
                             x=1,
                             y=2,
                             label_cell_groups=FALSE,
                             label_leaves=F,
                             label_branch_points=F,
                             show_trajectory_graph=F) +
      ggtitle(str_split( cell_types[i],">")[[1]][2]) + xlab("Dim. 1") + ylab("Dim. 2") +
      theme(legend.position="bottom") +
      theme(strip.text.x = element_text(size = 16))+
      scale_color_viridis_c(option = "inferno",name="expression") + #continuous
      theme(axis.text=element_text(size=18,color='black'),
            axis.title=element_text(size=20,color='black')) + theme(plot.title = element_text(size = 22, vjust=1)) + theme(axis.text=element_text(size=15,color='black'),
                                                                                                                           axis.title=element_text(size=15,color='black'))
    print(marker_plot)
    
    #aggregate gene values to get a single expression value
    marker_df = data.frame(losgenes, str_split( cell_types[i],">")[[1]][2])
    names(marker_df) = c("gene","cell_type")
    agg_mat = aggregate_gene_expression(
      ii_cds,
      gene_group_df = marker_df,
      cell_group_df = NULL,
      norm_method = c("log"),
      pseudocount = 1,
      scale_agg_values = TRUE,
      max_agg_value = 5,
      min_agg_value = -5,
      exclude.na = TRUE
    )
    
    #Monocle eigene expression
    marker_df$module <- factor(marker_df$cell_type)
    
    agexp_plot = plot_cells(ii_cds,
                            genes=marker_df,
                            cell_size=1,
                            label_cell_groups=FALSE,
                            label_leaves=F,
                            label_branch_points=F,
                            graph_label_size=1.5,
                            show_trajectory_graph=F) +
      xlab("Dim. 1") + ylab("Dim. 2") + 
      theme(legend.position="bottom") +
      theme(strip.text.x = element_text(size = 16))+
      scale_color_viridis_c(option = "inferno",name="aggregated expression") +
      ggtitle(paste(marker_df$cell_type[1], "aggregated expression",sep=" ")) + #continuous
      theme(axis.text=element_text(size=18,color='black'),
            axis.title=element_text(size=20,color='black')) + theme(plot.title = element_text(size = 22, vjust=1)) + theme(axis.text=element_text(size=15,color='black'),
                                                                                                                           axis.title=element_text(size=15,color='black'))
    
    print(agexp_plot)
  },  error=function(e){})
  
} #end marker plot list
dev.off()

#track innate immune cell load over time
mac = readRDS(file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/matvei/cds_objects/Macrophage.RDS")
langerhans = readRDS(file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/matvei/cds_objects/Langerhans.RDS")
neutro = readRDS(file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/matvei/cds_objects/Neutrophil.RDS")
mac@colData$innate_immune_cell = "Macrophage"
langerhans@colData$innate_immune_cell = "Langerhans"
neutro@colData$innate_immune_cell = "Neutrophil"
ii_comb = combine_cds(list(mac, langerhans, neutro)) #combine separate files into 1

#3b. Pre-process with Seurat-processed object
ii_comb <- preprocess_cds(ii_comb, 
                          method="PCA",
                          num_dim = 100,
                          norm_method = "none", #these data have already been normalized by Seurat
                          alignment_group = NULL)


#Re-align to remove batch effects: batch is covarying strongly with Unsorted ?align_cds
ii_comb = align_cds(cds = ii_comb,
                    num_dim = 100,
                    preprocess_method = "PCA",
                    alignment_group = "batch_name",
                    verbose=T)

#4b. ?reduce_dimension
ii_comb <- reduce_dimension(
  ii_comb,
  max_components = 2,
  reduction_method = c("UMAP"),
  preprocess_method = "Aligned", #must now used ALIGNEMNT PRE-PROCESSING METHOD bc it replaces the other pre-process shit
  umap.metric = "cosine",
  umap.min_dist = 0.3,
  umap.n_neighbors = 15L,
  umap.fast_sgd = FALSE,
  umap.nn_method = "annoy",
  cores = 1,
  verbose = FALSE
)


tissues = c("Normal skin","Pap","CA")

colData(ii_comb)$tissue_updated = colData(ii_comb)$tissue
colData(ii_comb)$tissue_updated = gsub( "\\bUnTxSkin\\b", "Normal skin", colData(ii_comb)$tissue_updated,perl=TRUE)
colData(ii_comb)$tissue_updated = gsub( "\\bTxSkin\\b", "Normal skin", colData(ii_comb)$tissue_updated,perl=TRUE)


plot_cells(ii_comb, color_cells_by="innate_immune_cell", group_cells_by="cluster",
           group_label_size = 0,
           x=1,
           y=2,
           cell_size=1.2,
           alpha=0.5) +
  ggtitle("") + xlab("") + ylab("") + 
  theme(axis.text=element_text(size=0,color='black'),
        axis.title=element_text(size=0),
        axis.ticks = element_line(size=0),
        legend.position="right")

#-------------------------------------------------#
#Main Text Fig
#-------------------------------------------------#
#Fig. 1J
png("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/stem_cell_mansucript/Figs/Fig1/Fig1J.png",width=7,height=7,units="in",res=400)

print( plot_cells(ii_comb, color_cells_by="tissue_updated", group_cells_by="cluster",
                  group_label_size = 0,
                  x=1,
                  y=2,
                  cell_size=1.2,
                  alpha=0.5) +
         ggtitle("") + xlab("") + ylab("") + 
         theme(axis.text=element_text(size=0,color='black'),
               axis.title=element_text(size=0),
               axis.ticks = element_line(size=0),
               legend.position="none") +
         scale_color_manual(values = c("Normal skin" = "darkblue", "Pap" = "mediumpurple2","CA" = "darkred"))
)
dev.off()
#-------------------------------------------------#

iiload = do.call(rbind, lapply(1:length(tissues), function(i){
  #i=1
  ii.sub = ii_comb[,colData(ii_comb)$tissue_updated %in% c(tissues[i])] #subset to tissue-specific amount of T-cells
  a = length(colData(cds_comb)$tissue[colData(cds_comb)$tissue %in% tissues[i]]) #get number of cells in tissue-specific parenchyma
  
  macs = ii.sub[,grep("Macrophage", ii.sub@colData$innate_immune_cell)]
  mac_num = dim(macs)[2]
  
  langs = ii.sub[,grep("Langerhans", ii.sub@colData$innate_immune_cell)]
  lang_num = dim(langs)[2]
  
  neutros = ii.sub[,grep("Neutrophil", ii.sub@colData$innate_immune_cell)]
  neut_num = dim(neutros)[2]
  
  iiprop = round((dim(ii.sub)[2]/a)*100,3) # proportions
  mac_prop = round((mac_num/a)*100,3) # proportions
  lang_prop = round((lang_num/a)*100,3) # proportions
  neut_prop = round((neut_num/a)*100,3) # proportions
  
  data.frame(tissues[i],dim(ii.sub)[2],mac_num,lang_num,neut_num,iiprop,mac_prop,lang_prop,neut_prop)
  
}))

names(iiload) = c("tissue","InnateImmune no.","Mac no.","Langerhans no.","Neutrophil no.","InnateImmune percent","Mac percent","Langerhans percent","Neutrophil percent")
iiload.dfm = melt(iiload, id=c("tissue"))

iiload_plot = ggplot(data=iiload.dfm, aes(x=tissue,y=value,fill=tissue,color=tissue)) +
  geom_bar(stat="identity") +
  facet_wrap(~variable, ncol=4, scales="free") +
  xlab("tissue") + ylab("%                            N") + 
  theme(legend.position="bottom") +
  theme(strip.text.x = element_text(size = 16))+
  ggtitle("T cell load") + #continuous
  scale_color_manual(values = c("UnTxSkin" = "skyblue3","TxSkin" = "darkblue", "Pap" = "darkgreen","CA" = "darkred")) + theme(legend.position="bottom") +
  scale_fill_manual(values = c("UnTxSkin" = "skyblue3","TxSkin" = "darkblue", "Pap" = "darkgreen","CA" = "darkred")) + theme(legend.position="bottom") +
  theme(axis.text=element_text(size=18,color='black'),
        axis.title=element_text(size=20,color='black')) + theme(plot.title = element_text(size = 22, vjust=1)) + theme(axis.text=element_text(size=15,color='black'),
                                                                                                                       axis.title=element_text(size=15,color='black')) +
  theme(text = element_text(size=18)) +
  theme(axis.text.y=element_text(size=17,color='black'),
        axis.text.x=element_text(size=9,color='black'),
        axis.title.x=element_text(size=20,color='black'),
        axis.title.y=element_text(size=22,color='black')) +
  theme(plot.title = element_text(size = 20, vjust=1)) +
  theme(axis.text=element_text(size=15,color='black'),
        axis.title=element_text(size=15,color='black')) +
  theme(panel.grid.major=element_line(colour = "darkgrey"), 
        panel.grid.minor=element_blank(), 
        panel.background=element_blank()) +
  theme(legend.key = element_rect(fill = "white")) +
  theme(strip.text.x = element_text(size = 14, colour = "black", angle = 0)) +
  theme(strip.text.y = element_text(size = 14, colour = "black", angle = 0)) +
  theme(legend.text=element_text(size=14), legend.title=element_text(size=14)) +
  theme(strip.background=element_rect(fill="white"))
print(iiload_plot)

#combine T cell and innate immune cell loan
iiload.dfm$immune_type = "Innate"
toload.dfm$immune_type = "T-cell"
cellcomb = rbind(iiload.dfm,toload.dfm)
cellcomb = data.frame(cellcomb, t(data.frame(str_split(cellcomb$variable, " "))))
names(cellcomb)[c(5,6)] = c("cell","type")
cellpercent = subset(cellcomb, type=="percent" | type=="fraction")

#=====================================================================================#
#Main text fig: Fig. 1K
#=====================================================================================#

cellpercent$tissue = gsub("Normal skin","NSk",cellpercent$tissue)
cellpercent$tissue = gsub("CA","SCC",cellpercent$tissue)

png("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Figs/Fig1/Fig1K.png",width=8.5,height=7,units="in",res=400)
print(
  ggplot(data=cellpercent, aes(x=factor(tissue, level = c('NSk', 'Pap', 'SCC')),y=value,fill=tissue,color=tissue)) +
    geom_bar(stat="identity",alpha=0.6) +
    facet_wrap(~cell, ncol=4, scales="free_y") +
    xlab("") + ylab("%") + 
    theme(legend.position="bottom") +
    ggtitle("") + #continuous
    scale_color_manual(values = c("NSk" = "darkblue", "Pap" = "mediumpurple2","SCC" = "darkred")) +
    scale_fill_manual(values = c("NSk" = "darkblue", "Pap" = "mediumpurple2","SCC" = "darkred")) +
    theme(panel.grid.major=element_line(colour = "darkgrey"), 
          panel.grid.minor=element_blank(), 
          panel.background=element_blank()) +
    theme(legend.key = element_rect(fill = "white")) +
    theme(axis.text = element_text(size = 16)) +
    theme(strip.text.x = element_text(size = 16, colour = "black", angle = 0)) +
    theme(strip.text.y = element_text(size = 16, colour = "black", angle = 0)) +
    theme(legend.text=element_text(size=16), legend.title=element_text(size=14)) +
    theme(strip.background=element_rect(fill="white")) +
    theme(legend.position="none") +
    theme(plot.title = element_text(size=0))
)
dev.off()

#=====================================================================================#
#Main text fig: Fig. 1P : only the main Innate-Immune nad T-cells
#=====================================================================================#

cellpercent.sub = cellpercent[cellpercent$cell %in% c("InnateImmune","T-cell"),]
png("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Figs/Fig1/Fig1K.png",width=2,height=7,units="in",res=400)
print(
  ggplot(data=cellpercent.sub, aes(x=factor(tissue, level = c('Normal skin', 'Pap', 'CA')),y=value,fill=tissue,color=tissue)) +
    geom_bar(stat="identity") +
    facet_wrap(~cell, ncol=1, scales="free_y") +
    xlab("") + ylab("") + 
    theme(legend.position="bottom") +
    ggtitle("") + #continuous
    scale_color_manual(values = c("Normal skin" = "darkblue", "Pap" = "mediumpurple2","CA" = "darkred")) +
    scale_fill_manual(values = c("Normal skin" = "darkblue", "Pap" = "mediumpurple2","CA" = "darkred")) +
    theme(panel.grid.major=element_line(colour = "darkgrey"), 
          panel.grid.minor=element_blank(), 
          panel.background=element_blank()) +
    theme(legend.key = element_rect(fill = "white")) +
    theme(strip.text.x = element_text(size = 0, colour = "black", angle = 0)) +
    theme(axis.text.y = element_text(size = 14, colour = "black", angle = 0)) +
    theme(legend.text=element_text(size=14), legend.title=element_text(size=14)) +
    theme(strip.background=element_rect(fill="white")) +
    theme(legend.position="none") +
    theme(plot.title = element_text(size=0))
)
dev.off()


#=====================================================================================#
#Macrophage Analysis
#=====================================================================================#

mac = readRDS(file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/matvei/cds_objects/Macrophage.RDS")
langerhans = readRDS(file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/matvei/cds_objects/Langerhans.RDS")
mac@colData$innate_immune_cell = "Macrophage"
langerhans@colData$innate_immune_cell = "Langerhans"
mac_comb = combine_cds(list(mac, langerhans)) #combine separate files into 1

#3b. Pre-process with Seurat-processed object
mac_comb <- preprocess_cds(mac_comb, 
                           method="PCA",
                           num_dim = 100,
                           norm_method = "none", #these data have already been normalized by Seurat
                           alignment_group = NULL)


#Re-align to remove batch effects: batch is covarying strongly with Unsorted ?align_cds
mac_comb = align_cds(cds = mac_comb,
                     num_dim = 100,
                     preprocess_method = "PCA",
                     alignment_group = "batch_name",
                     verbose=T)

#4b. ?reduce_dimension
mac_comb <- reduce_dimension(
  mac_comb,
  max_components = 2,
  reduction_method = c("UMAP"),
  preprocess_method = "Aligned", #must now used ALIGNEMNT PRE-PROCESSING METHOD bc it replaces the other pre-process shit
  umap.metric = "cosine",
  umap.min_dist = 0.3,
  umap.n_neighbors = 15L,
  umap.fast_sgd = FALSE,
  umap.nn_method = "annoy",
  cores = 1,
  verbose = FALSE
)


tissues = c("Normal skin","Pap","CA")

colData(mac_comb)$tissue_updated = colData(mac_comb)$tissue
colData(mac_comb)$tissue_updated = gsub( "\\bUnTxSkin\\b", "Normal skin", colData(mac_comb)$tissue_updated,perl=TRUE)
colData(mac_comb)$tissue_updated = gsub( "\\bTxSkin\\b", "Normal skin", colData(mac_comb)$tissue_updated,perl=TRUE)

mac_comb_fig = plot_cells(mac_comb, color_cells_by="tissue_updated", group_cells_by="cluster",
                          group_label_size = 0,
                          x=1,
                          y=2,
                          cell_size=1.4,
                          alpha=0.5) +
  ggtitle("") + xlab("") + ylab("") + 
  theme(axis.text=element_text(size=0,color='black'),
        axis.title=element_text(size=0),
        axis.ticks = element_line(size=0),
        legend.position="none") +
  scale_color_manual(values = c("Normal skin" = "darkblue", "Pap" = "mediumpurple2","CA" = "darkred"))

#=====================================================================================#
#Extended Data Fig.1d : Macrophage and Langerhans Cell manifold
#=====================================================================================#
png("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Extended\ Data/Extended\ Data\ Fig\ 1/ExtDat_Fig1d.png",
    width=7,height=7,units="in",res=200)
print( mac_comb_fig
)
dev.off()

#---------------------------------------------------------------------------------------------------------------------------------#
#---------------------------------------------------------------------------------------------------------------------------------#
#---------------------------------------------------------------------------------------------------------------------------------#

#M1/M2 markers
#read in eigengenes
mac_markers = read.xlsx(
  xlsxFile="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/macrophage_analysis/integrated_mac_markers.xlsx",
  colNames=T
)


for (i in 1:ncol(mac_markers)) {
  #i=1
  print(i)
  #only if more than 1 constituent gene
  marker_df = data.frame(na.omit(mac_markers[,i]),names(mac_markers)[i])
  names(marker_df) = c("gene","cell_type")
  
  #UPDATED NCBI MATCHING!! --> this will product lots of non-matching names that i must resolve with NCBI aliases
  allnetnames = do.call(rbind, lapply(1:nrow(marker_df), function(l){
    #l=1
    netnames = data.frame(marker_df$gene[l],rowData(big_cds)$gene_short_name[match(marker_df$gene[l],rowData(big_cds)$gene_short_name)])
    names(netnames) = c("balmain_network_gene_name","cds_gene_name")
    netnames
  }))
  cds_gene_names = rowData(big_cds)$gene_short_name #get vector of all gene names in cds object
  #get consistent names between cds and network gene names
  match_noms = do.call(rbind, lapply(1:nrow(allnetnames), function(k){
    #k=1 ; 22
    tryCatch({ #suppress error
      if(is.na(allnetnames$cds_gene_name[k])==F) {cds_alias_match = allnetnames$cds_gene_name[k]} else
      { 
        #search through Gene aliases for a match
        ncbi_symbol = ncbi_names[grep(allnetnames$balmain_network_gene_name[k],ncbi_names$Aliases),] #search through aliases for a match
        if (nrow(ncbi_symbol)>0) {
          ncbi_matches_vector = c(as.character(ncbi_symbol$Symbol), as.character(data.frame(strsplit(as.character(ncbi_symbol$Aliases),", "))[,1]) ) #vector of all aliases that marker_gene matches
          cds_alias_match = na.omit(cds_gene_names[match(ncbi_matches_vector,cds_gene_names)]) #match this symbol to the cds gene names
        } else {
          #search through Gene Symbols
          ncbi_symbol.1 = ncbi_names[grep(allnetnames$balmain_network_gene_name[k],ncbi_names$Symbol),] 
          ncbi_matches_vector = c(as.character(ncbi_symbol.1$Symbol), as.character(data.frame(strsplit(as.character(ncbi_symbol.1$Aliases),", "))[,1]) ) #vector of all aliases that marker_gene matches
          cds_alias_match = na.omit(cds_gene_names[match(ncbi_matches_vector,cds_gene_names)]) #match this symbol to the cds gene names
        }
      }
      data.frame(allnetnames[k,],cds_alias_match)
    },  error=function(e){})
  }))
  
  #remake the marker df dataframe with the matched gene names
  marker_df = data.frame(na.omit(unique(match_noms$cds_alias_match)),names(mac_markers)[i])
  names(marker_df) = c("gene","cell_type")
  
  #Monocle eigene expression
  marker_df$module <- factor(marker_df$cell_type)
  
  #all cells eigengene
  agexp_plot = plot_cells(mac_comb,
                          genes=marker_df,
                          label_cell_groups=FALSE,
                          label_leaves=F,
                          label_branch_points=F,
                          graph_label_size=1.5,
                          show_trajectory_graph=F,
                          cell_size = 1) +
    xlab("Dim. 1") + ylab("Dim. 2") + 
    theme(legend.position="bottom") +
    scale_color_viridis_c(option = "inferno",name="integrated expression") +
    ggtitle(paste(names(mac_markers)[i],"integrated markers",sep=" "))+ #continuous
    theme(axis.text=element_text(size=8,color='black'),
          axis.title=element_text(size=10,color='black')) + theme(plot.title = element_text(size = 14, vjust=1)) +
    theme(strip.text.x = element_text(size = 0)) #remove facet label
  
  png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/macrophage_analysis/integrated_marker__plots/macrophage_",i,"_integrated_marker_manifold.png",sep=""),width=7,height=7,units="in",res=200)
  print(agexp_plot)
  dev.off()
  
}

#plot mac genes of interest
plot_cells(mac_comb,
           genes=c("Trem1","Trem2"),
           label_cell_groups=FALSE,
           label_leaves=F,
           label_branch_points=F,
           graph_label_size=1.5,
           show_trajectory_graph=F,
           cell_size = 1) +
  xlab("Dim. 1") + ylab("Dim. 2") + 
  theme(legend.position="bottom") +
  scale_color_viridis_c(option = "inferno",name="log expression") +
  ggtitle("")+ #continuous
  theme(axis.text=element_text(size=8,color='black'),
        axis.title=element_text(size=10,color='black')) + theme(plot.title = element_text(size = 14, vjust=1)) +
  theme(strip.text.x = element_text(size = 10)) #remove facet label



#------------------------------------------------------------------------------------#
#marker_test_res on tissue-specific 
colData(mac_comb)$immune_cell_tissue = paste(colData(mac_comb)$innate_immune_cell,colData(mac_comb)$tissue_updated,sep=" ")
marker_test_res <- top_markers(mac_comb, group_cells_by="immune_cell_tissue",
                               genes_to_test_per_group = 200,
                               cores=8)
marker_test_res = marker_test_res[order(marker_test_res$marker_test_q_value),] #re-order low to high q-value
write.table(marker_test_res, file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/macrophage_analysis/macrophage_LC_top_markers.txt", quote=F,row.names = F,sep="\t")


#---------------------------------------------------------------------------------------------------------------------------------#
#---------------------------------------------------------------------------------------------------------------------------------#
#---------------------------------------------------------------------------------------------------------------------------------#
#---------------------------------------------------------------------------------------------------------------------------------#
#------------------------------------------------------------------------------------#
#------------------------------------------------------------------------------------#
#----------------------------------------------------------------------------------------------------#
#----------------------------------------------------------------------------------------------------#
#----------------------------------------------------------------------------------------------------#

####################################################################################################################
####################################################################################################################
#cell complexity: number of genes
colData(big_cds)$nFeature_RNA
cell_complexity_manifold = plot_cells(big_cds,
                                      color_cells_by= "nFeature_RNA",
                                      label_cell_groups=FALSE,
                                      label_leaves=F,
                                      label_branch_points=F,
                                      graph_label_size=1.5,
                                      show_trajectory_graph=F) +
  xlab("Dim. 1") + ylab("Dim. 2") + 
  theme(legend.position="bottom") +
  scale_color_viridis_c(option = "inferno",name="no. genes") +
  ggtitle("Cell complexity")+ #continuous
  theme(axis.text=element_text(size=8,color='black'),
        axis.title=element_text(size=10,color='black')) + theme(plot.title = element_text(size = 14, vjust=1)) +
  theme(strip.text.x = element_text(size = 0)) #remove facet label



####################################################################################################################


####################################################################################################################
####################################################################################################################
##################################################################################################################################
##################################################################################################################################
##################################################################################################################################
##################################################################################################################################
#extract Jacks lung clsuter 5 stem cell markers
ind_sheet = read.xlsx(
  xlsxFile=c("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/apriori_eigengenes/Jacks_lung_stem\ cells_Cancer\ CEll_Table\ S4.xlsx"),
  sheet = "cluster5_msigdbStem",
)

genevec = paste(ind_sheet$List.of.interesecting.genes, collapse=" ")
genedf = data.frame(strsplit(genevec, " "))
names(genedf) = "cluster5_msigdbStem"
unique_genedf = data.frame(na.omit(unique(genedf$cluster5_msigdbStem)))

write.table(unique_genedf, file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/apriori_eigengenes/lung_stem_cluster5_msigdbStem.txt",sep="\t",quote=F,row.names=F)


##################################################################################################################################
##################################################################################################################################
#Extract stem cell markers (SCGS Stem Cell Gene Signature) from Palmer et al 2012: https://link.springer.com/article/10.1186/gb-2012-13-8-r71#author-information
library(readxl)
eigengene_excel = "/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/stem_markers/Palmer\ et\ al\ 2012\ Genome\ Biol/Human\ SCGS.xlsx"
sheets = excel_sheets(eigengene_excel) #et a vector of sheets names


mouse_scgs = do.call(rbind, lapply(1:length(sheets), function(i){
  
  #i=1
  print(paste(i,sheets[i],sep=" "))
  
  datasheet = data.frame(read_excel(eigengene_excel, sheet = sheets[i]))
  mouse_ortholog = data.frame(sheets[i], convertHumanGeneList(datasheet$gene_name))
  names(mouse_ortholog) = c("gene_type","gene")
  mouse_ortholog
  
}))

write.table(mouse_scgs, 
            file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/stem_markers/Palmer\ et\ al\ 2012\ Genome\ Biol/mouse_orthologous_scgs.txt",
            sep="\t",quote=F,row.names=F)

##################################################################################################################################
#Commonly mutated genes found to be important cancer markers: https://www.science.org/doi/full/10.1126/science.abf3067?casa_token=gQyLaVK3yz8AAAAA%3A3iy9RoN0Vqsmo8q3rTjL82TKDuIKn9is6YWys9xqyk0qBlzYQ_TwLTeLqCQPJaBqmAjVqdf7YK4jNUc
library(readxl)
eigengene_excel = "/Users/marktaylor/Desktop/Papers/cancer\ gene\ networks/Science\ 2021\ Interpretation\ of\ cancer\ mutations\ using\ a\ multiscale\ map\ of\ protein\ systems/science.abf3067_tables_s1_to_s8/science.abf3067_table_s8.xlsx"
sheets = excel_sheets(eigengene_excel) #et a vector of sheets names

datasheet = data.frame(read_excel(eigengene_excel, sheet = sheets[1])) #extract the gene list
#Convert to mouse orthologs
names(datasheet)[1] = "gene"
mouse_ortholog = data.frame(convertHumanGeneList(datasheet$gene))

names(mouse_ortholog)[1] = "gene"

#get the network data in this paper

ppi_networks = read.delim(file="/Users/marktaylor/Desktop/Papers/cancer\ gene\ networks/Science\ 2021\ Interpretation\ of\ cancer\ mutations\ using\ a\ multiscale\ map\ of\ protein\ systems/IAS_score.tsv"
                          ,header=T)
head(ppi_networks)

lgr6_sub = subset(ppi_networks, (Protein.1=="LGR6" | Protein.2=="LGR6"))



##################################################################################################################################
##################################################################################################################################
##################################################################################################################################
##################################################################################################################################
##################################################################################################################################
####################################################################################################################################################################################################################################################################
##################################################################################################################################
##################################################################################################################################
##################################################################################################################################
##################################################################################################################################
##################################################################################################################################
#Re-make manifold for particular movements

#read in parenchyma tissue CDS's and combine
parenchymas = c("UnTxSkin","TxSkin","TransitionalPap", "Pap", "CA")
colpal_transitionnetwork.1 = c("lightblue","blue","cyan","green", "red")

a <- readRDS(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/matvei/cds_objects/",parenchymas[1],".RDS",sep=""))
b <- readRDS(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/matvei/cds_objects/",parenchymas[2],".RDS",sep=""))
c <- readRDS(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/matvei/cds_objects/",parenchymas[3],".RDS",sep=""))
d <- readRDS(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/matvei/cds_objects/",parenchymas[4],".RDS",sep=""))
e <- readRDS(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/matvei/cds_objects/",parenchymas[5],".RDS",sep=""))

#add comparison names
a@colData$compare_var = parenchymas[1]
b@colData$compare_var = parenchymas[2]
c@colData$compare_var = parenchymas[3]
d@colData$compare_var = parenchymas[4]
e@colData$compare_var = parenchymas[5]

#Overall tissue
a@colData$tissue_type = "Normal skin"
b@colData$tissue_type = "Normal skin"
c@colData$tissue_type = "Papilloma"
d@colData$tissue_type = "Papilloma"
e@colData$tissue_type = "Carcinoma"

#---------------------------------------------------------------------------------------------------#
#Carcinoma only
ca_only = e

#3b. Pre-process object
ca_only <- preprocess_cds(ca_only, 
                          method="PCA",
                          num_dim = 50,
                          norm_method = "none", #these data have already been normalized by Seurat
                          alignment_group = NULL)

#4b. ?reduce_dimension
ca_only <- reduce_dimension(
  ca_only,
  max_components = 2,
  reduction_method = c("UMAP"),
  preprocess_method = "PCA", #must now used ALIGNEMNT PRE-PROCESSING METHOD bc it replaces the other pre-process shit
  umap.metric = "cosine",
  umap.min_dist = 0.1,
  umap.n_neighbors = 15L,
  umap.fast_sgd = FALSE,
  umap.nn_method = "annoy",
  cores = 1,
  verbose = FALSE
)

#5. Cluster and partition ?cluster_cells cells into supergroups community detection (Phenograph algorithm) Community detection of cells as part of Phenograph algorithm https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4508757/
#5b. 
ca_only <- cluster_cells(ca_only,
                         reduction_method = c("UMAP"),
                         k = 10, #a bigger k will result in lower resolution 
                         cluster_method = c("leiden"),
                         num_iter = 2,
                         partition_qval = 0.05, #Numeric, the q-value cutoff to determine when to partition
                         weight = FALSE,
                         resolution = NULL,
                         random_seed = NULL,
                         verbose = F)

#=====================================================================================#
#Extended Data Fig.1c : Vimentin expression of SCC parenchyma
#=====================================================================================#

png("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Extended\ Data/Extended\ Data\ Fig\ 1/ExtDat_Fig1c.png",
    width=6,height=5,units="in",res=200)

print(plot_cells(ca_only,
                 genes="Vim", #seed gene is the first row of the markers file
                 label_cell_groups=FALSE,
                 label_leaves=F,
                 label_branch_points=F,
                 graph_label_size=1.5,
                 show_trajectory_graph=F,
                 cell_size = 1.3) +
        xlab("") + ylab("") + 
        theme(axis.ticks = element_blank()) +
        theme(legend.position="right") +
        scale_color_viridis_c(option = "inferno",name="Vimentin\nexpression") +
        ggtitle("")+ #continuous
        theme(axis.text=element_text(size=0,color='black'),
              axis.title=element_text(size=0,color='black'),
              legend.title=element_text(size=15), 
              legend.text=element_text(size=15)) + theme(plot.title = element_text(size = 0, vjust=0)) +
        theme(strip.text.x = element_text(size = 0)) #remove facet label
)

dev.off()

##################################################################################################################################
##################################################################################################################################
##################################################################################################################################
##################################################################################################################################
##################################################################################################################################
##################################################################################################################################

#Generate eigengenes systematically from seed gene

#1. read in the data
aim2dat = read.table(file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/bulk_network_data/Aim2-4/AIM2_AIM4_BS_CARC_HFC_NOTED_ENTREZ_03_16_15_EXPRESSION.txt",header=T)
aim2dat[1:10,1:10]
dim(aim2dat)
aim2phenodat = read.table(file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/bulk_network_data/Aim2-4/AIM2_AIM4_BS_CARC_HFC_NOTED_ENTREZ_03_16_15_PHENOTYPE_DATA.txt",header=T)
aim2phenodat[1:10,1:10]
dim(aim2phenodat)
aim2probedat = read.table(file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/bulk_network_data/Aim2-4/AIM2_AIM4_BS_CARC_HFC_NOTED_ENTREZ_03_16_15_PROBE_DATA.txt",header=T)
aim2probedat[1:10,]


#2a. subset to specific samples and get sample IDs
# normal_wt_ids = subset(aim2phenodat, TYPE=="NORMAL" & GENOTYPE=="WT") my original settings cf Allan's settings
normal_wt_ids = subset(aim2phenodat, TISSUE=="BACK" & GENOTYPE_INFERRED=="WT") #use Allan's filter settings
normal_wt_ids = normal_wt_ids$IDENTIFIER
normal_wt_ids = gsub("-",".",normal_wt_ids)  #gsub hyphens to decimals  bc R doesn't allow hyphens in column names and changes hyphens to decimals in the aim2dat dataframe
length(normal_wt_ids)

#2b. subset to specific samples and get sample IDs
tumor_wt_ids = subset(aim2phenodat, TISSUE=="CARC" & GENOTYPE_INFERRED=="WT")
tumor_wt_ids = tumor_wt_ids$IDENTIFIER
tumor_wt_ids = gsub("-",".",tumor_wt_ids)  #gsub hyphens to decimals  bc R doesn't allow hyphens in column names and changes hyphens to decimals in the aim2dat dataframe
length(tumor_wt_ids)

#3a. subset expression dat from sample IDs
normal_wt_dat = subset(aim2dat, select=normal_wt_ids)
#normal_wt_dat.v2 = aim2dat[,match(normal_wt_ids, colnames(aim2dat))] #alternate way of selecting correct columns

#3b. subset expression dat from sample IDs
tumor_wt_dat = subset(aim2dat, select=tumor_wt_ids)
#tumor_wt_dat.v2 = aim2dat[,match(tumor_wt_ids, colnames(aim2dat))] #alternate way of selecting correct columns

#4a. Normal skin WT: calculate correlation matrix
deg_matx = as.matrix(normal_wt_dat) #convert to normal matrix
deg_matx = t(deg_matx) #transpose so genes are being correlated
cormat = rcorr(deg_matx, type = c("spearman")) #calc correlation matrix
rhodf = data.frame(cormat$r) #extract rho values and coerce to dataframe
p_df = data.frame(cormat$P) #extract rho values and coerce to dataframe

#4b. Tumor WT: calculate correlation matrix
tumor_matx = as.matrix(tumor_wt_dat) #convert to normal matrix
tumor_matx = t(tumor_matx) #transpose so genes are being correlated
tumor_cormat = rcorr(tumor_matx, type = c("spearman")) #calc correlation matrix
tumor_rhodf = data.frame(tumor_cormat$r) #extract rho values and coerce to dataframe
tumor_p_df = data.frame(tumor_cormat$P) #extract rho values and coerce to dataframe


tumor_rhodf[1:10,1:10]

#-----------------------------------------------------------------------------------------------------------------#
#-----------------------------------------------------------------------------------------------------------------#
#Loop through all genes and calculate standardized eigengenes chagnes at bulk leel

unique(aim2probedat$SYMBOL)
head(aim2probedat)
aim2probedat$unique_symbol = make.unique(aim2probedat$SYMBOL, "_")
unique_genes = aim2probedat$unique_symbol

bulk_eigengenome = do.call(cbind, lapply(1:length(unique_genes), function(i){
  #i=1
  tryCatch({ #suppress error
    
    print(i)
    match_index = match(unique_genes[i], aim2probedat$unique_symbol) #match seed gene to its place in the matrix
    
    #Normal skin
    symbol_match = data.frame(aim2probedat$unique_symbol, rhodf[,match_index]) #put together gene names, expression data, and p-vals
    symbol_match = symbol_match[order(-symbol_match$rhodf...match_index.),] #order from high to low gene values
    names(symbol_match) = c("gene","rho")
    nsk_match = symbol_match[1:eigengene_size,]
    
    nsk_dat = t(normal_wt_dat[match_index,])[,1] #single gene
    nsk_mean = mean(nsk_dat)
    nsk_var = var(nsk_dat)
    
    nsk_eigengene = rowSums(normal_wt_dat[match(nsk_match$gene,aim2probedat$unique_symbol),]) #eigengene
    nsk_eigen_mean = mean(nsk_eigengene)
    nsk_eigen_var = var(nsk_eigengene)
    
    #Tumor
    symbol_match = data.frame(aim2probedat$unique_symbol, tumor_rhodf[,match_index]) #put together gene names, expression data, and p-vals
    symbol_match = symbol_match[order(-symbol_match$tumor_rhodf...match_index.),] #order from high to low gene values
    names(symbol_match) = c("gene","rho")
    symbol_match$seed_gene = seed_gene[i]
    tumor_match = symbol_match[1:eigengene_size,]
    
    ca_dat = t(tumor_wt_dat[match_index,])[,1]  #single gene
    ca_mean = mean(ca_dat)
    ca_var = var(ca_dat)
    
    ca_eigengene = rowSums(tumor_wt_dat[match(tumor_match$gene,aim2probedat$unique_symbol),]) #eigengene
    ca_eigen_mean = mean(ca_eigengene)
    ca_eigen_var = var(ca_eigengene)
    
    #Calculate FC
    fc_singlegene = ca_mean/nsk_mean
    fc_eigengenee = ca_eigen_mean/nsk_eigen_mean
    
    #Make single-gene dataframe to graph
    nsk_df_singlegene = data.frame(t(normal_wt_dat[match_index,])[,1],"NSk")
    names(nsk_df_singlegene) = c("expression","tissue")
    ca_df_singlegene = data.frame(t(tumor_wt_dat[match_index,])[,1],"Car")
    names(ca_df_singlegene) = c("expression","tissue")
    
    single_gene_df = rbind(nsk_df_singlegene,ca_df_singlegene)
    single_gene_df$tissue <- factor(single_gene_df$tissue, levels=c("NSk", "Car"))
    
    my_comparisons <- list( c("NSk", "Car"))
    
    single_gene_plot = ggplot(single_gene_df, aes(x=tissue,y=expression, fill=tissue)) +
      stat_compare_means(comparisons = my_comparisons, method="wilcox.test")  +
      geom_violin(width=1, alpha=0.4) +
      geom_boxplot(outlier.size=0, width=0.3, color="black", size=0.7, alpha=0.4) +
      theme_classic() +
      geom_jitter(width=0.25, size=0.8, alpha=0.8) +
      #geom_bar(stat = "identity", fill="darkorange") +
      #facet_wrap(~parenchyma, scales="free", nrow=1) + 
      #geom_jitter(width = 0.1, size=0.3, alpha=0.4) +
      ggtitle(paste(unique_genes[i]," single gene",sep="")) + 
      xlab("") + ylab("") + 
      theme(axis.text.x=element_text(size=15,color='black'),
            axis.text.y=element_text(size=15,color='black'),
            title=element_text(size=20),
            axis.ticks.x = element_line(size=0),
            strip.text.x = element_text(size = 0),
            legend.position="none") +
      scale_fill_manual(name="",values = c("Car" = "darkred",
                                           "NSk" = "darkblue"))
    
    #Make single-gene dataframe to graph
    nsk_df_eigengeene = data.frame(nsk_eigengene,"NSk")
    names(nsk_df_eigengeene) = c("expression","tissue")
    ca_df_eigengeene = data.frame(ca_eigengene,"Car")
    names(ca_df_eigengeene) = c("expression","tissue")
    
    eigengene_df = rbind(nsk_df_eigengeene,ca_df_eigengeene)
    eigengene_df$tissue <- factor(eigengene_df$tissue, levels=c("NSk", "Car"))
    
    eigen_gene_plot = ggplot(eigengene_df, aes(x=tissue,y=expression, fill=tissue)) +
      stat_compare_means(comparisons = my_comparisons, method="wilcox.test")  +
      geom_violin(width=1, alpha=0.4) +
      geom_boxplot(outlier.size=0, width=0.3, color="black", size=0.7, alpha=0.4) +
      theme_classic() +
      geom_jitter(width=0.25, size=0.8, alpha=0.8) +
      #geom_bar(stat = "identity", fill="darkorange") +
      #facet_wrap(~parenchyma, scales="free", nrow=1) + 
      #geom_jitter(width = 0.1, size=0.3, alpha=0.4) +
      ggtitle(paste(unique_genes[i]," Metagene",sep="")) + 
      xlab("") + ylab("") + 
      theme(axis.text.x=element_text(size=15,color='black'),
            axis.text.y=element_text(size=15,color='black'),
            title=element_text(size=20),
            axis.ticks.x = element_line(size=0),
            strip.text.x = element_text(size = 0),
            legend.position="none") +
      scale_fill_manual(name="",values = c("Car" = "darkred",
                                           "NSk" = "darkblue"))
    
    #Print out figs
    png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/Eigengenome/bulk_eigengenome_aim2/figs/",
              unique_genes[i],".png",sep=""),
        width=12,height=6,units="in",res=150)
    print(multiplot(single_gene_plot,eigen_gene_plot,cols=2))
    dev.off()
    
    #Combine all the genes togethre
    to_return = data.frame(unique_genes[i],nsk_mean,nsk_var,nsk_eigen_mean,nsk_eigen_var,fc_singlegene,
                           ca_mean,ca_var,ca_eigen_mean,ca_eigen_var,fc_eigengenee)
    
    names(to_return)[1] = "gene"
    
    write.table(to_return,
                file=paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/Eigengenome/bulk_eigengenome_aim2/textfiles/",
                           unique_genes[i],".txt",sep=""),
                sep="\t",row.names=F,quote=F
    )
    
    return(to_return)
  },  error=function(e){})
}))

#concatenate individual files into one dataframe and write out
file_list <- list.files(path="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/Eigengenome/bulk_eigengenome_aim2/textfiles/", 
                        pattern=".txt",full.names=T) #get list of all files
dataset <- ldply(file_list, read.table, header=TRUE, sep="\t") #concatenate
write.table(dataset, file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/Eigengenome/bulk_eigengenome_aim2/combined.txt", quote=F,row.names = F,sep="\t")


#-----------------------------------------------------------------------------------------------------------------#
#-----------------------------------------------------------------------------------------------------------------#
#-----------------------------------------------------------------------------------------------------------------#
#-----------------------------------------------------------------------------------------------------------------#
#Kras network

thresholds = seq(0.6,0.8,by=0.01)
network_summaries = do.call(rbind, lapply(1:length(thresholds), function(j){
  tryCatch({ #suppress error
    #j=1
    print(j)
    
    
    #--------------------------------------------------------------------------------------------
    #Normal skin
    #Kras network
    match_index = match("Kras", aim2probedat$SYMBOL) #match seed gene to its place in the matrix
    symbol_match = data.frame(aim2probedat$IDENTIFIER, rhodf[,match_index], rhodf[,match_index]) #put together gene names, expression data, and p-vals
    names(symbol_match) = c("gene","rho","p")
    symbol_match = symbol_match[order(-symbol_match$rho),] #order from high to low gene value
    symbol_match_kras = subset(symbol_match, rho > thresholds[j])
    
    combined_genes = unique(symbol_match_kras$gene)
    
    #now subset to the correct parts of the correlation matrix
    combined_matrix = rhodf[match(combined_genes, aim2probedat$IDENTIFIER),match(combined_genes, aim2probedat$IDENTIFIER)]
    combined_matrix[combined_matrix < thresholds[j]] <- 0 #replace lements of the matrix less than rho threshold with zero
    row.names(combined_matrix) =  make.names( aim2probedat[match(combined_genes, aim2probedat$IDENTIFIER),]$SYMBOL , unique=T)
    
    #my data
    combined_matrix[1:6,1:6]
    row.names(combined_matrix)
    grep("Metap2",row.names(combined_matrix))
    
    net = network(combined_matrix, directed = FALSE)
    ?network
    net = network(net, directed = FALSE)
    network.vertex.names(net) = row.names(combined_matrix) #add gene names to nodes
    #now add a vector of yes/no gene names for the 3 genes we're interested in: Lgr6, Lrig1, Lgr5
    
    dummy_vec = rep("non-seed",nrow(combined_matrix)) #make a dummy vector
    dummy_vec[match(c("Kras"), row.names(combined_matrix))] <- "seed"
    net %v% "seed_status" = dummy_vec
    
    set.seed(1) #this puts the nodes in the same position
    netgraph_back = ggnet2(net, mode = "kamadakawai",color="seed_status",
                           label = T,label.color = "black", label.size=3,
                           size="degree",max_size = 25,
                           size.min = 0,#remove any isolated nodes
                           edge.color="lightgrey",
                           alpha=0.6) +
      theme(legend.position="none") +
      theme(panel.background = element_rect(fill = "white")) +
      theme(legend.position="none") +
      ggtitle(paste("Dorsal Network, threshold=",thresholds[j],sep="")) +
      theme(plot.title = element_text(size = 30, vjust=1, face="bold")) +
      scale_colour_manual(values = c("seed"="#E1AF00","non-seed"="#3B9AB2")) +
      coord_equal()
    
    #-------------------------------------------------------------------------------------------------------------------------
    #Tail
    
    #Lgr5 network
    match_index = match("Lgr5", cellrepdat$IDENTIFIER) #match seed gene to its place in the matrix
    symbol_match = data.frame(cellrepdat$IDENTIFIER, tail_rhodf[,match_index], tail_p_df[,match_index]) #put together gene names, expression data, and p-vals
    names(symbol_match) = c("gene","rho","p")
    symbol_match = symbol_match[order(-symbol_match$rho),] #order from high to low gene value
    symbol_match_lgr5 = subset(symbol_match, rho > thresholds[j])
    
    #Lgr6 network
    match_index = match("Lgr6", cellrepdat$IDENTIFIER) #match seed gene to its place in the matrix
    symbol_match = data.frame(cellrepdat$IDENTIFIER, tail_rhodf[,match_index], tail_p_df[,match_index]) #put together gene names, expression data, and p-vals
    names(symbol_match) = c("gene","rho","p")
    symbol_match = symbol_match[order(-symbol_match$rho),] #order from high to low gene value
    symbol_match_lgr6 = subset(symbol_match, rho > thresholds[j])
    
    #Lrig1 network
    match_index = match("Lrig1", cellrepdat$IDENTIFIER) #match seed gene to its place in the matrix
    symbol_match = data.frame(cellrepdat$IDENTIFIER, tail_rhodf[,match_index], tail_p_df[,match_index]) #put together gene names, expression data, and p-vals
    names(symbol_match) = c("gene","rho","p")
    symbol_match = symbol_match[order(-symbol_match$rho),] #order from high to low gene value
    symbol_match_lrig1 = subset(symbol_match, rho > thresholds[j])
    
    #now get all the genes across these three networks
    combined_genes = unique(c(symbol_match_lgr5$gene,symbol_match_lgr6$gene,symbol_match_lrig1$gene)) #combined unique genes called in each of these networks
    
    #get the size of unique 
    combined_networks_n = length(combined_genes)
    
    #get size of networks
    tail_summary = data.frame("tail skin", thresholds[j],nrow(symbol_match_lgr5), nrow(symbol_match_lgr6), nrow(symbol_match_lrig1),combined_networks_n)
    
    #now subset to the correct parts of the correlation matrix
    combined_matrix = tail_rhodf[match(combined_genes, cellrepdat$IDENTIFIER),match(combined_genes, cellrepdat$IDENTIFIER)]
    combined_matrix[combined_matrix <= thresholds[j]] <- 0 #replace lements of the matrix less than rho threshold with zero
    row.names(combined_matrix) = combined_genes #replace this baddies
    
    #my data
    net = network(combined_matrix, directed = FALSE)
    net = network(net, directed = FALSE)
    network.vertex.names(net) = row.names(combined_matrix) #add gene names to nodes
    #now add a vector of yes/no gene names for the 3 genes we're interested in: Lgr6, Lrig1, Lgr5
    
    dummy_vec = rep("non-seed",nrow(combined_matrix)) #make a dummy vector
    dummy_vec[match(c("Lgr5","Lgr6","Lrig1"), row.names(combined_matrix))] <- "seed"
    net %v% "seed_status" = dummy_vec
    
    set.seed(1) #this puts the nodes in the same position
    netgraph_tail = ggnet2(net, mode = "kamadakawai",color="seed_status",
                           label = T,label.color = "black", label.size=3,
                           size="degree",max_size = 25,
                           size.min = 0,#remove any isolated nodes
                           edge.color="lightgrey",
                           alpha=0.6) +
      theme(legend.position="none") +
      theme(panel.background = element_rect(fill = "white")) +
      theme(legend.position="none") +
      ggtitle(paste("Tail Network, threshold=",thresholds[j],sep="")) +
      theme(plot.title = element_text(size = 30, vjust=1, face="bold")) +
      scale_colour_manual(values = c("seed"="#EBCC2A","non-seed"="#78B7C5")) +
      coord_equal()
    
    #---------------------------------------------------------------------------------------#
    #Print the network out to a large object
    png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Eve_ms/threshold_sensitivity/hairball_figs/dorsal_threshold_",thresholds[j],".png",sep=""),
        width=50,height=25,units="in",res=150)
    print(multiplot(netgraph_back,netgraph_tail,cols=2))
    dev.off()
    
    #---------------------------------------------------------------------------------------#
    names(tail_summary) = c("site","rho_threshold","Lgr5_N","Lgr6_N","Lrig1__N","total_network_N")
    names(back_summary) = c("site","rho_threshold","Lgr5_N","Lgr6_N","Lrig1__N","total_network_N")
    return(rbind(back_summary,tail_summary))
  },  error=function(e){})
  
}))

write.table(network_summaries,
            file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Eve_ms/threshold_sensitivity/hairball_figs/1_network_summaries.txt",
            row.names=F,quote=F,sep="\t")



#-----------------------------------------------------------------------------------------------------------------#
#-----------------------------------------------------------------------------------------------------------------#
#-----------------------------------------------------------------------------------------------------------------#
#-----------------------------------------------------------------------------------------------------------------#

#compare variability in single-cell data with bulk data

genes_to_examine = c("Lgr5",
                     "Bmi1","Sox4","Lrig1",
                     "Lgr6","Sox9",
                     "Sox2","Prom1","Foxa1","Krt15","Krt19","Pitx1","Psca","Prom2","Klf5",
                     "Cd44")

tumor_wt_dat[1:10,1:10]
dim(tumor_wt_dat)
dim(normal_wt_dat)
row.names(tumor_wt_dat) = make.unique(aim2probedat$SYMBOL) #add gene symbols as row-names
row.names(normal_wt_dat) = make.unique(aim2probedat$SYMBOL) #add gene symbols as row-names

unique(colData(big_cds)$tissue_updated)
nsk_cds = big_cds[,colData(big_cds)$tissue_updated %in% "Normal skin"] #filter for cluster 19 because most of the bridge cells are only in cluster 19
car_cds = big_cds[,colData(big_cds)$tissue_updated %in% "CA"] #filter for cluster 19 because most of the bridge cells are only in cluster 19


var_metrics = do.call(rbind, lapply(1:length(genes_to_examine), function(i){
  #i=1
  
  print(i)
  #Bulk variability metrics
  bulk_nsk_var = var(  t(normal_wt_dat[match(genes_to_examine[i],row.names(normal_wt_dat)),] )[,1] )
  bulk_nsk_sd = sd(  t(normal_wt_dat[match(genes_to_examine[i],row.names(normal_wt_dat)),] )[,1] )
  #nsk_cv = (nsk_sd / (mean(  t(normal_wt_dat[match(genes_to_examine[i],row.names(normal_wt_dat)),] )[,1] )) ) * 100
  
  bulk_car_var = var(  t(tumor_wt_dat[match(genes_to_examine[i],row.names(tumor_wt_dat)),] )[,1] )
  bulk_car_sd = sd(  t(tumor_wt_dat[match(genes_to_examine[i],row.names(tumor_wt_dat)),] )[,1] )
  #car_cv = (car_sd / (mean(  t(tumor_wt_dat[match(genes_to_examine[i],row.names(tumor_wt_dat)),] )[,1] )) ) * 100
  
  #Single cell variability metrics
  sc_nsk_var = var(  nsk_cds@assays@data$counts[match(genes_to_examine[i],row.names(big_cds)),] )
  sc_nsk_sd = sd(  nsk_cds@assays@data$counts[match(genes_to_examine[i],row.names(big_cds)),] )
  #sc_nsk_cv = ( sc_nsk_sd / (mean(  nsk_cds@assays@data$counts[match(genes_to_examine[i],row.names(big_cds)),] )) )*100
  
  sc_car_var = var(  car_cds@assays@data$counts[match(genes_to_examine[i],row.names(big_cds)),] )
  sc_car_sd = sd(  car_cds@assays@data$counts[match(genes_to_examine[i],row.names(big_cds)),] )
  #sc_car_cv = ( sc_car_sd / (mean(  car_cds@assays@data$counts[match(genes_to_examine[i],row.names(big_cds)),] )) )*100
  
  data.frame(genes_to_examine[i],
             bulk_nsk_var, bulk_nsk_sd,
             bulk_car_var, bulk_car_sd,
             sc_nsk_var, sc_nsk_sd,
             sc_car_var, sc_car_sd)
  
}))

names(var_metrics)[1] = "gene"

dfm = melt(var_metrics, id=c("gene"))
var_split = data.frame(t(data.frame(strsplit(as.character(dfm$variable), "_"))))
names(var_split) = c("expression_type","tissue","variability_metric")
dfm_comb = cbind(dfm,var_split)
dfm_comb$expression_type = gsub("bulk","Bulk",dfm_comb$expression_type)
dfm_comb$expression_type = gsub("sc","Single-cell",dfm_comb$expression_type)
dfm_comb$tissue = gsub("nsk","Normal skin",dfm_comb$tissue)
dfm_comb$tissue = gsub("car","Carcinoma",dfm_comb$tissue)
dfm_comb$variability_metric = gsub("var","Variance",dfm_comb$variability_metric)
dfm_comb$variability_metric = gsub("sd","Standard deviation",dfm_comb$variability_metric)

ggplot(dfm_comb, aes(x=gene,y=value, fill=expression_type)) +
  geom_bar(stat = "identity", position="dodge") +
  facet_grid(tissue~variability_metric, scales="free") + 
  ggtitle("Gene expression variability is greater in bulk than single-cell data") + 
  xlab("gene") + ylab("") + 
  theme(axis.text.x=element_text(size=10,color='black', angle=90),
        axis.text.y=element_text(size=12,color='black'),
        axis.title=element_text(size=15),
        axis.ticks.x = element_line(size=0),
        strip.text.x = element_text(size =15),
        strip.text.y = element_text(size =20),
        legend.position="bottom",
        title = element_text(size = 17.5, face="bold")) +
  scale_fill_manual(values = c("Bulk" = "burlywood2",
                               "Single-cell" = "darkseagreen4")) +
  theme(
    # Hide panel borders and remove grid lines
    #panel.border = element_blank(),
    #panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.background = element_blank(),
    element_line(colour = 'black', size = 0.3),
    # Change axis line
    axis.line = element_line(colour = "black"),
    strip.background = element_rect(colour="white",
                                    fill="white")) +
  scale_fill_manual(values = c("Bulk" = "burlywood2",
                               "Single-cell" = "darkseagreen4"))



#-----------------------------------------------------------------------------------------------------------------#
#-----------------------------------------------------------------------------------------------------------------#
#-----------------------------------------------------------------------------------------------------------------#
#Repeat this on our multi-sample data
nsk_cds = big_cds[,colData(big_cds)$tissue_updated %in% "Normal skin"] #filter for cluster 19 because most of the bridge cells are only in cluster 19
car_cds = big_cds[,colData(big_cds)$tissue_updated %in% "CA"] #filter for cluster 19 because most of the bridge cells are only in cluster 19


var_metrics = do.call(rbind, lapply(1:length(genes_to_examine), function(i){
  #i=1
  
  print(i)
  #Bulk variability metrics
  bulk_nsk_var = var(  t(normal_wt_dat[match(genes_to_examine[i],row.names(normal_wt_dat)),] )[,1] )
  bulk_nsk_sd = sd(  t(normal_wt_dat[match(genes_to_examine[i],row.names(normal_wt_dat)),] )[,1] )
  #nsk_cv = (nsk_sd / (mean(  t(normal_wt_dat[match(genes_to_examine[i],row.names(normal_wt_dat)),] )[,1] )) ) * 100
  
  bulk_car_var = var(  t(tumor_wt_dat[match(genes_to_examine[i],row.names(tumor_wt_dat)),] )[,1] )
  bulk_car_sd = sd(  t(tumor_wt_dat[match(genes_to_examine[i],row.names(tumor_wt_dat)),] )[,1] )
  #car_cv = (car_sd / (mean(  t(tumor_wt_dat[match(genes_to_examine[i],row.names(tumor_wt_dat)),] )[,1] )) ) * 100
  
  #Single cell variability metrics
  sc_nsk_var = var(  nsk_cds@assays@data$counts[match(genes_to_examine[i],row.names(big_cds)),] )
  sc_nsk_sd = sd(  nsk_cds@assays@data$counts[match(genes_to_examine[i],row.names(big_cds)),] )
  #sc_nsk_cv = ( sc_nsk_sd / (mean(  nsk_cds@assays@data$counts[match(genes_to_examine[i],row.names(big_cds)),] )) )*100
  
  sc_car_var = var(  car_cds@assays@data$counts[match(genes_to_examine[i],row.names(big_cds)),] )
  sc_car_sd = sd(  car_cds@assays@data$counts[match(genes_to_examine[i],row.names(big_cds)),] )
  #sc_car_cv = ( sc_car_sd / (mean(  car_cds@assays@data$counts[match(genes_to_examine[i],row.names(big_cds)),] )) )*100
  
  data.frame(genes_to_examine[i],
             bulk_nsk_var, bulk_nsk_sd,
             bulk_car_var, bulk_car_sd,
             sc_nsk_var, sc_nsk_sd,
             sc_car_var, sc_car_sd)
  
}))

names(var_metrics)[1] = "gene"

dfm = melt(var_metrics, id=c("gene"))
var_split = data.frame(t(data.frame(strsplit(as.character(dfm$variable), "_"))))
names(var_split) = c("expression_type","tissue","variability_metric")
dfm_comb = cbind(dfm,var_split)
dfm_comb$expression_type = gsub("bulk","Bulk",dfm_comb$expression_type)
dfm_comb$expression_type = gsub("sc","Single-cell",dfm_comb$expression_type)
dfm_comb$tissue = gsub("nsk","Normal skin",dfm_comb$tissue)
dfm_comb$tissue = gsub("car","Carcinoma",dfm_comb$tissue)
dfm_comb$variability_metric = gsub("var","Variance",dfm_comb$variability_metric)
dfm_comb$variability_metric = gsub("sd","Standard deviation",dfm_comb$variability_metric)

ggplot(dfm_comb, aes(x=gene,y=value, fill=expression_type)) +
  geom_bar(stat = "identity", position="dodge") +
  facet_grid(tissue~variability_metric, scales="free") + 
  ggtitle("Gene expression variability is greater in bulk than single-cell data") + 
  xlab("gene") + ylab("") + 
  theme(axis.text.x=element_text(size=10,color='black', angle=90),
        axis.text.y=element_text(size=12,color='black'),
        axis.title=element_text(size=15),
        axis.ticks.x = element_line(size=0),
        strip.text.x = element_text(size =15),
        strip.text.y = element_text(size =20),
        legend.position="bottom",
        title = element_text(size = 17.5, face="bold")) +
  scale_fill_manual(values = c("Bulk" = "burlywood2",
                               "Single-cell" = "darkseagreen4")) +
  theme(
    # Hide panel borders and remove grid lines
    #panel.border = element_blank(),
    #panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.background = element_blank(),
    element_line(colour = 'black', size = 0.3),
    # Change axis line
    axis.line = element_line(colour = "black"),
    strip.background = element_rect(colour="white",
                                    fill="white")) +
  scale_fill_manual(values = c("Bulk" = "burlywood2",
                               "Single-cell" = "darkseagreen4"))


#-----------------------------------------------------------------------------------------------------------------#
#-----------------------------------------------------------------------------------------------------------------#
#-----------------------------------------------------------------------------------------------------------------#
#-----------------------------------------------------------------------------------------------------------------#
#-----------------------------------------------------------------------------------------------------------------#



#---------------------------------------#
#Write out tumor rhodf to disk
write.table(tumor_rhodf,
            file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/cormat/aim2_tumor_rho.txt",
            row.names=F,col.names=F,sep="\t",quote=F)

#---------------------------------------#
#Table S1 Supplemental Table Eigengene composition
seed_gene = c("Lgr5",
              "Bmi1","Sox4","Lrig1",
              "Lgr6","Sox9",
              "Sox2","Prom1","Foxa1","Krt15","Krt19","Pitx1","Psca","Prom2","Klf5",
              "Cd44")
eigengene_size = 100

allgenes_rhos = do.call(cbind, lapply(1:length(seed_gene), function(i){
  #i=1
  print(i)
  if( seed_gene[i]=="Prom1" ) {match_index = 12747} else {
    match_index = match(seed_gene[i], aim2probedat$SYMBOL) #match seed gene to its place in the matrix
  }
  
  #Normall skin
  symbol_match = data.frame(aim2probedat$SYMBOL, rhodf[,match_index]) #put together gene names, expression data, and p-vals
  symbol_match = symbol_match[order(-symbol_match$rhodf...match_index.),] #order from high to low gene values
  names(symbol_match) = c("gene","rho")
  symbol_match$seed_gene = seed_gene[i]
  symbol_match$report_col = paste(symbol_match$gene, " (",round(symbol_match$rho,3), ")",sep="")
  nsk_match = symbol_match[1:eigengene_size,]
  
  1+1
  
  #Tumor
  symbol_match = data.frame(aim2probedat$SYMBOL, tumor_rhodf[,match_index]) #put together gene names, expression data, and p-vals
  symbol_match = symbol_match[order(-symbol_match$tumor_rhodf...match_index.),] #order from high to low gene values
  names(symbol_match) = c("gene","rho")
  symbol_match$seed_gene = seed_gene[i]
  symbol_match$report_col = paste(symbol_match$gene, " (",round(symbol_match$rho,3), ")",sep="")
  tumor_match = symbol_match[1:eigengene_size,]
  
  to_return = data.frame(cbind(nsk_match$report_col,tumor_match$report_col))
  names(to_return) = c(paste(seed_gene[i]," NSk",sep=""),
                       paste(seed_gene[i]," Car",sep="")
  )
  return(to_return)
}))


write.table(allgenes_rhos,
            file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Supplemental\ Information/eigengene_constituents.txt",
            row.names=F,col.names=T,sep="\t",quote=F)


#---------------------------------------#
#Read in tumor data and get it ready to export and test for PPIs
tumor_rhodf = read.table(file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/cormat/aim2_tumor_rho.txt",
                         header=F)

#Now go through seed genes and get ordered correlations for each
seed_gene = c("Lgr5",
              "Bmi1","Sox4","Lrig1",
              "Lgr6","Sox9",
              "Sox2","Prom1","Foxa1","Krt15","Krt19","Pitx1","Psca","Prom2","Klf5",
              "Cd44")
eigengene_size = 100

tumor_rhos = do.call(rbind, lapply(1:length(seed_gene), function(i){
  #i=5
  print(i)
  if( seed_gene[i]=="Prom1" ) {match_index = 12747} else {
    match_index = match(seed_gene[i], aim2probedat$SYMBOL) #match seed gene to its place in the matrix
  }
  symbol_match = data.frame(aim2probedat$SYMBOL, tumor_rhodf[,match_index]) #put together gene names, expression data, and p-vals
  symbol_match = symbol_match[order(-symbol_match$tumor_rhodf...match_index.),] #order from high to low gene values
  names(symbol_match) = c("gene","rho")
  symbol_match$seed_gene = seed_gene[i]
  symbol_match
}))

write.table(tumor_rhos,
            file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/cormat/aim2_tumor_rho_stem_seed_genes.txt",
            row.names=F, quote=F, sep="\t")

tumor_rhodf = read.table(file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/cormat/aim2_tumor_rho_stem_seed_genes.txt",
                         header=T)

#---------------------------------------#
#---------------------------------------#
#---------------------------------------#
#Write out Lgr5/6 tumor/normal skin networks
match_index = match("Lgr6", aim2probedat$SYMBOL) #match seed gene to its place in the matrix

#Normal skin
symbol_match = data.frame(aim2probedat$SYMBOL, rhodf[,match_index], p_df[,match_index]) #put together gene names, expression data, and p-vals
symbol_match = symbol_match[order(-symbol_match$rhodf...match_index.),] #order from high to low gene values
names(symbol_match) = c("gene","rho","p")
symbol_match$rho = round(symbol_match$rho,3)

#Tumor
tumor_symbol_match = data.frame(aim2probedat$SYMBOL, tumor_rhodf[,match_index], tumor_p_df[,match_index]) #put together gene names, expression data, and p-vals
tumor_symbol_match = tumor_symbol_match[order(-tumor_symbol_match$tumor_rhodf...match_index.),] #order from high to low gene values
names(tumor_symbol_match) = c("gene","rho","p")
tumor_symbol_match$rho = round(tumor_symbol_match$rho,3)

lgr6_combined = cbind(symbol_match[,c(1,2)],tumor_symbol_match[,c(1,2)])
names(lgr6_combined)[c(1,3)] = c("Normal skin eigengene","SCC eigengene")

write.table(lgr6_combined,
            file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Extended\ Data/Extended\ Data\ Fig\ 2/Lgr6_eigengene_components.txt",
            sep="\t",quote=F,row.names=F)

#-------------------------------------------------------------------------------------------#
#-------------------------------------------------------------------------------------------#
#-------------------------------------------------------------------------------------------#
#-------------------------------------------------------------------------------------------#
#-------------------------------------------------------------------------------------------#
#-------------------------------------------------------------------------------------------#

#-------------------------------------------------------------------------------------------#
#EXTENDED DATA FIG. TABLE 2 network overlaps
#-------------------------------------------------------------------------------------------#
#Write function to find overlaps in NSk vs carcinoma networks

seed_gene = c(
  "Sox2","Prom1","Foxa1","Krt15","Krt19","Pitx1","Psca","Prom2","Klf5",
  "Cd44")

network_overlaps = do.call(rbind, lapply(1:length(seed_gene), function(i){
  
  #i=1
  print(i)
  
  #Get the correct match_index
  if( seed_gene[i]=="Prom1" ) {match_index = 12747} else {
    match_index = match(seed_gene[i], aim2probedat$SYMBOL) #match seed gene to its place in the matrix
  }
  
  #Normal skin
  symbol_match = data.frame(aim2probedat$SYMBOL, rhodf[,match_index], p_df[,match_index]) #put together gene names, expression data, and p-vals
  symbol_match = symbol_match[order(-symbol_match$rhodf...match_index.),] #order from high to low gene values
  names(symbol_match) = c("gene","rho","p")
  symbol_match$rho = round(symbol_match$rho,3)
  
  #Tumor
  tumor_symbol_match = data.frame(aim2probedat$SYMBOL, tumor_rhodf[,match_index], tumor_p_df[,match_index]) #put together gene names, expression data, and p-vals
  tumor_symbol_match = tumor_symbol_match[order(-tumor_symbol_match$tumor_rhodf...match_index.),] #order from high to low gene values
  names(tumor_symbol_match) = c("gene","rho","p")
  tumor_symbol_match$rho = round(tumor_symbol_match$rho,3)
  
  networks_combined = cbind(seed_gene[i],symbol_match[1:eigengene_size,c(1,2)],tumor_symbol_match[1:eigengene_size,c(1,2)])
  names(networks_combined)[c(1,2,4)] = c("seed_gene","nsk_eigengene","scc_eigengene")
  networks_combined
  
}))

#Summarize how often genes occur in normak skin networks
freq_table = data.frame(table(network_overlaps$nsk_eigengene))
freq_table = freq_table[order(-freq_table$Freq),]

mean(freq_table[freq_table$Freq>1,]$Freq) #mean number of shared nodes: ie how many times the same genes occurs in different networks
length(freq_table[freq_table$Freq>1,]$Freq) #total number of shared nodes: ie how many times the same genes occurs in different networks
(length(freq_table[freq_table$Freq>1,]$Freq) / nrow(freq_table))*100 #percent shared nodes of all unique gene nodes in all networks

#Summarize how often genes occur in scc networks
freq_table = data.frame(table(network_overlaps$scc_eigengene))
freq_table = freq_table[order(-freq_table$Freq),]

mean(freq_table[freq_table$Freq>1,]$Freq) #mean number of shared nodes: ie how many times the same genes occurs in different networks
length(freq_table[freq_table$Freq>1,]$Freq) #total number of shared nodes: ie how many times the same genes occurs in different networks
(length(freq_table[freq_table$Freq>1,]$Freq) / nrow(freq_table))*100 #percent shared nodes of all unique gene nodes in all networks


#-------------------------------------------------------------------------------------------#
#-------------------------------------------------------------------------------------------#
#-------------------------------------------------------------------------------------------#
#-------------------------------------------------------------------------------------------#
#read in the seed genes to create my own eigengenes
stem_hierachy_genes = read.xlsx(
  xlsxFile="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/mark_eigengenes/stem_gene_hierarchy_genes.xlsx",
  colNames=T
)

seed_gene = unique(stem_hierachy_genes$genes)

#Set analysis parameters
eigengene_size = 100 #how many genes to include in eigengene
bonferroni_correction_factor = length(seed_gene)*2*eigengene_size #number of genes being tested and *2 becomes making a network for noromal and carcinoma
le_text = paste("Network collapse: No significant edges")

#make sure that seed genes match to gene symtols in the probe data
data.frame(seed_gene,match(seed_gene,aim2probedat$SYMBOL))
match(alias2SymbolTable(seed_gene, species = "Mm"),aim2probedat$SYMBOL)

#--------------------------------------------------------------#
#Function to quantify stemness of a network matrix
stem_markers = read.xlsx(
  xlsxFile="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/stem_markers/Integrated_stem_cell_markers_frotiers_gene.xlsx",
  colNames=T
)
stem_markers_unique = unique(stem_markers$Gene.Symbol)
length(stem_markers_unique)

define_stemness = function(input_mat, bulk_expression_dat) {
  
  no_stem_genes = length(na.omit(match(stem_markers_unique,rownames(input_mat))))
  propotion_stem_genes = no_stem_genes / nrow(input_mat)
  
  overlapping_stem_genes = na.omit(rownames(input_mat)[match(stem_markers_unique,rownames(input_mat))]) #find gene names of overlapping stem genes
  
  #get overlapping probes for stem and all genes in input_mat
  overlapping_stem_match_index = na.omit(match(overlapping_stem_genes, aim2probedat$SYMBOL))
  allgenes_match_index = na.omit(match(rownames(input_mat),aim2probedat$SYMBOL))
  
  #extract the data
  stem_ovlerap_dat = bulk_expression_dat[overlapping_stem_match_index,]
  allmatgenes_dat = bulk_expression_dat[allgenes_match_index,]
  
  #proporiton of stem expression
  stem_expression = sum(stem_ovlerap_dat)
  all_expression = sum(allmatgenes_dat)
  stem_expression_propotion = stem_expression / all_expression
  
  #collate results
  stemness_results = data.frame(no_stem_genes,propotion_stem_genes,stem_expression,all_expression,stem_expression_propotion)
  return(stemness_results)
}

#end function definition
#--------------------------------------------------------------#
#--------------------------------------------------------------#
#--------------------------------------------------------------#

manifolds_network_df = do.call(rbind, lapply(1:length(seed_gene), function(i){
  #manifolds_network_df = do.call(rbind, lapply(1:2, function(i){
  
  tryCatch({ #suppress error
    #i=1
    print(i)
    
    #Find some aliases
    match_index = match(seed_gene[i], aim2probedat$SYMBOL) #match seed gene to its place in the matrix
    
    #---------------------------------------#
    #Normal skin
    symbol_match = data.frame(aim2probedat$SYMBOL, rhodf[,match_index], p_df[,match_index]) #put together gene names, expression data, and p-vals
    symbol_match = symbol_match[order(-symbol_match$rhodf...match_index.),] #order from high to low gene values
    names(symbol_match) = c("gene","rho","p")
    
    #loop to examine network characteristics: number of genes, median correlation, signifiance of least correlated gene
    med_sig_df = do.call(rbind, lapply(2:eigengene_size, function(j){
      #j=2
      sub_symbolmatch = symbol_match[1:j,]
      if(symbol_match$p[j] < 0.05) {cor_significance = "significant"} else {cor_significance = "not significant"} #ask whether the last gene is significant or not
      data.frame(j, median(sub_symbolmatch$rho), cor_significance)
    }))
    names(med_sig_df) = c("N","rho","sig")
    
    #make marker df for grahing eigengene
    marker_df = data.frame(unique(alias2SymbolTable(symbol_match$gene[1:(eigengene_size+1)], species = "Mm")),paste(seed_gene[i]," Normal skin",sep=""),paste(seed_gene[i]," Normal skin",sep=""))
    names(marker_df) = c("gene","cell_type","module")
    
    normal_skin_network_charct = ggplot(med_sig_df, aes(x=N, y=rho, color=sig)) +
      geom_point(alpha=0.1) +
      ggtitle(paste(seed_gene[i]," Normal skin networks",sep="")) +
      xlab("Network size (N constituent genes)") +
      ylab("Median rho") +
      scale_color_manual(values = c("significant" = "darkblue", "not significant" = "red"), name="Significance") +
      theme(axis.text=element_text(size=10,color='black'),
            axis.title=element_text(size=12,color='black')) +
      theme(plot.title = element_text(size = 15, vjust=1)) +
      theme(panel.grid.major=element_line(colour = "darkgrey"), 
            panel.grid.minor=element_blank(), 
            panel.background=element_blank()) +
      theme(legend.key = element_rect(fill = "white")) +
      theme(legend.text=element_text(size=12), legend.title=element_text(size=12)) +
      theme(legend.position="bottom")
    
    #plot individual genes and their rhos
    symbol_match_indexed = symbol_match[2:eigengene_size,]
    symbol_match_indexed$gene_index = seq(2,eigengene_size,by=1)
    normal_skin_network_charct.ind_gene = ggplot(symbol_match_indexed, aes(x=gene_index, y=rho)) +
      geom_point(alpha=1, color="goldenrod4") +
      ggtitle(paste(seed_gene[i]," Normal skin ordinated ind genes",sep="")) +
      xlab("Gene index") +
      ylab("Rho") +
      theme(axis.text=element_text(size=10,color='black'),
            axis.title=element_text(size=12,color='black')) +
      theme(plot.title = element_text(size = 15, vjust=1)) +
      theme(panel.grid.major=element_line(colour = "darkgrey"), 
            panel.grid.minor=element_blank(), 
            panel.background=element_blank()) +
      theme(legend.key = element_rect(fill = "white")) +
      theme(legend.text=element_text(size=12), legend.title=element_text(size=12)) +
      theme(legend.position="bottom")
    
    #now subset to network (eigengene) that you will create the eigengen with
    #try creating eigengene from top 1% of significant network correlations
    # tigit_sub = subset(symbol_match, p<0.05)
    # top1_percent = (nrow(subset(symbol_match, p<0.05))*0.01)
    # normal_skin_genes_for_eigengene = symbol_match[1:top1_percent,]
    normal_skin_genes_for_eigengene = symbol_match[1:eigengene_size,]
    write.table(normal_skin_genes_for_eigengene, file=paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/mark_eigengenes/eigengene_constituent_genes_v2/",seed_gene[i],"_normal_skin_eigengene.txt",sep=""))
    
    #------------------------------#
    #4a. Normal skin WT: calculate correlation matrix
    #------------------------------#
    #all cells: Seed gene
    agexp_plot_seedgene.12 = plot_cells(big_cds,
                                        genes=marker_df$gene[1], #seed gene is the first row of the markers file
                                        label_cell_groups=FALSE,
                                        label_leaves=F,
                                        label_branch_points=F,
                                        graph_label_size=1.5,
                                        show_trajectory_graph=F) +
      xlab("Dim. 1") + ylab("Dim. 2") + 
      theme(legend.position="bottom") +
      scale_color_viridis_c(option = "inferno",name="expression") +
      ggtitle(paste(seed_gene[i],"seed gene",sep=" "))+ #continuous
      theme(axis.text=element_text(size=8,color='black'),
            axis.title=element_text(size=10,color='black')) + theme(plot.title = element_text(size = 14, vjust=1)) +
      theme(strip.text.x = element_text(size = 0)) #remove facet label
    
    median_rho = round(summary(normal_skin_genes_for_eigengene$rho)[3],3)
    #all cells eigengene
    marker_df = na.omit(marker_df)
    agexp_plot.12 = plot_cells(big_cds,
                               genes=marker_df,
                               label_cell_groups=FALSE,
                               label_leaves=F,
                               label_branch_points=F,
                               graph_label_size=1.5,
                               show_trajectory_graph=F) +
      xlab("Dim. 1") + ylab("Dim. 2") + 
      theme(legend.position="bottom") +
      scale_color_viridis_c(option = "inferno",name="integrated expression") +
      ggtitle(paste(seed_gene[i],"normal skin eigengene, med rho =",as.vector(median_rho),sep=" "))+ #continuous
      theme(axis.text=element_text(size=8,color='black'),
            axis.title=element_text(size=10,color='black')) + theme(plot.title = element_text(size = 14, vjust=1)) +
      theme(strip.text.x = element_text(size = 0)) #remove facet label
    
    #------------------------------------------------------------------#
    #Start network analysis: Normal skin
    #------------------------------------------------------------------#
    match_index = match(symbol_match$gene[1:eigengene_size], aim2probedat$SYMBOL)
    r_cormat = cormat$r[match_index,match_index]
    p_cormat =  cormat$P[match_index,match_index]
    
    #replace non-sig correlation matrix values with 0
    sigmat = do.call(rbind, lapply(1:nrow(r_cormat), function(rr) {
      #rr=1
      loscols = do.call(cbind, lapply(1:ncol(r_cormat), function(cc) {
        #cc=3
        if((is.na(p_cormat[rr,cc]) ==T)) {cval = 0} else if((p_cormat[rr,cc] < (0.05/bonferroni_correction_factor))) {cval = (r_cormat[rr,cc])} else {cval = 0}
        data.frame(cval)
      }))
      loscols
    }))
    row.names(sigmat) = make.names(aim2probedat$SYMBOL[match_index], unique=TRUE) #take care of any duplicate probe maps
    
    #find where rows/cols are all 0 which represent genes with no significant correlation
    rowsumz = data.frame(rowSums(sigmat))
    rowsumz$index = c(1:nrow(rowsumz))
    index_to_drop = subset(rowsumz, rowSums.sigmat.==0)$index
    if(length(index_to_drop)==0) {sigmat = sigmat} else {sigmat = sigmat[-index_to_drop,-index_to_drop]}
    
    #my data
    net = network(sigmat, directed = FALSE)
    net = network(net, directed = FALSE)
    network.vertex.names(net) = row.names(sigmat) #add gene names to nodes
    net %v% "direction" = c("Seed",rep("Correlated",(nrow(sigmat)-1)))
    
    #highlight any important stem genes
    stem_match_index = match(stem_markers_unique,row.names(sigmat)) #match stem gene to its position in sigmat matrix
    color_vector = rep("wheat",nrow(sigmat)) #make color palette for all genes
    color_vector = replace(color_vector, na.omit(stem_match_index), "darkred") #now replace colors for stem genes
    color_vector = replace(color_vector, 1, "seagreen4") #now replace color for seed gene
    
    set.seed(1) #this puts the nodes in the same position
    netgraph = ggnet2(net, mode = "kamadakawai",color=color_vector,
                      label = T,label.color = "black", label.size=4,
                      size="degree",max_size = 21,
                      alpha=0.6,
                      edge.size=0.1) +
      theme(legend.position="none") +
      theme(panel.background = element_rect(fill = "white")) +
      theme(legend.position="none") +
      ggtitle(paste(seed_gene[i],"Normal skin Network",sep=" ")) +
      theme(plot.title = element_text(size = 18, vjust=1))
    
    ig2 <- asIgraph(net) #make network object into igraph object
    firstdegneighb = neighborhood.size(ig2, 1) #find number of first degree neighbors
    seconddegneighb = neighborhood.size(ig2, 2) #find number of second degreeneighbors
    
    neighbors = data.frame(netgraph$data[,1],firstdegneighb,seconddegneighb)
    neighbors$tissue_type = "Normal skin"
    
    #ask how much stemness is in this network based onthe function defined above
    normal_stemness = define_stemness(sigmat,normal_wt_dat)
    names(normal_stemness) = paste("normalskin_",names(normal_stemness),sep="") #add a normal ski thing to these column names
    
    #######################################################################################################################
    #######################################################################################################################
    #---------------------------------------#
    #tumor
    match_index = match(seed_gene[i], aim2probedat$SYMBOL) #match seed gene to its place in the matrix
    symbol_match = data.frame(aim2probedat$SYMBOL, tumor_rhodf[,match_index], tumor_p_df[,match_index]) #put together gene names, expression data, and p-vals
    symbol_match = symbol_match[order(-symbol_match$tumor_rhodf...match_index.),] #order from high to low gene values
    names(symbol_match) = c("gene","rho","p")
    
    #loop to examine network characteristics: number of genes, median correlation, signifiance of least correlated gene
    med_sig_df = do.call(rbind, lapply(2:eigengene_size, function(j){
      #j=100
      sub_symbolmatch = symbol_match[1:j,]
      if(symbol_match$p[j] < 0.05) {cor_significance = "significant"} else {cor_significance = "not significant"} #ask whether the last gene is significant or not
      data.frame(j, median(sub_symbolmatch$rho), cor_significance)
    }))
    names(med_sig_df) = c("N","rho","sig")
    
    #make marker df for grahing eigengene
    marker_df = data.frame(unique(alias2SymbolTable(symbol_match$gene[1:(eigengene_size+1)], species = "Mm")),paste(seed_gene[i],"Carcinoma",sep=""),paste(seed_gene[i],"Carcinoma",sep=""))
    names(marker_df) = c("gene","cell_type","module")
    
    tumor_network_charct = ggplot(med_sig_df, aes(x=N, y=rho, color=sig)) +
      geom_point(alpha=0.1) +
      ggtitle(paste(seed_gene[i]," tumor networks",sep="")) +
      xlab("Network size (N constituent genes)") +
      ylab("Median rho") +
      scale_color_manual(values = c("significant" = "darkblue", "not significant" = "red"), name="Significance") +
      theme(axis.text=element_text(size=10,color='black'),
            axis.title=element_text(size=12,color='black')) +
      theme(plot.title = element_text(size = 15, vjust=1)) +
      theme(panel.grid.major=element_line(colour = "darkgrey"), 
            panel.grid.minor=element_blank(), 
            panel.background=element_blank()) +
      theme(legend.key = element_rect(fill = "white")) +
      theme(legend.text=element_text(size=12), legend.title=element_text(size=12)) +
      theme(legend.position="bottom")
    
    #plot individual genes and their rhos
    symbol_match_indexed = symbol_match[2:eigengene_size,]
    symbol_match_indexed$gene_index = seq(2,eigengene_size,by=1)
    tumor_network_charct.ind_gene = ggplot(symbol_match_indexed, aes(x=gene_index, y=rho)) +
      geom_point(alpha=0.1, color="goldenrod4") +
      ggtitle(paste(seed_gene[i]," Tumor ordinated ind genes",sep="")) +
      xlab("Gene index") +
      ylab("Rho") +
      theme(axis.text=element_text(size=10,color='black'),
            axis.title=element_text(size=12,color='black')) +
      theme(plot.title = element_text(size = 15, vjust=1)) +
      theme(panel.grid.major=element_line(colour = "darkgrey"), 
            panel.grid.minor=element_blank(), 
            panel.background=element_blank()) +
      theme(legend.key = element_rect(fill = "white")) +
      theme(legend.text=element_text(size=12), legend.title=element_text(size=12)) +
      theme(legend.position="bottom")
    
    #now subset to network (eigengene) that you will create the eigengen with
    #try creating eigengene from top 1% of significant network correlations
    #top1_percent = (nrow(subset(symbol_match, p<0.05))*0.01)
    #tumor_genes_for_eigengene = symbol_match[1:top1_percent,]
    tumor_genes_for_eigengene = symbol_match[1:eigengene_size,]
    write.table(tumor_genes_for_eigengene, file=paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/mark_eigengenes/eigengene_constituent_genes_v2/",seed_gene[i],"_tumor_eigengene.txt",sep=""))
    
    median_rho_tumor = round(summary(tumor_genes_for_eigengene$rho)[3],3)
    #all cells eigengene
    marker_df = na.omit(marker_df)
    agexp_plot_tumor.12 = plot_cells(big_cds,
                                     genes=marker_df,
                                     label_cell_groups=FALSE,
                                     label_leaves=F,
                                     label_branch_points=F,
                                     graph_label_size=1.5,
                                     show_trajectory_graph=F) +
      xlab("Dim. 1") + ylab("Dim. 2") + 
      theme(legend.position="bottom") +
      scale_color_viridis_c(option = "inferno",name="integrated expression") +
      ggtitle(paste(seed_gene[i],"tumor eigengene, med rho =",median_rho_tumor,sep=" "))+ #continuous
      theme(axis.text=element_text(size=8,color='black'),
            axis.title=element_text(size=10,color='black')) + theme(plot.title = element_text(size = 14, vjust=1)) +
      theme(strip.text.x = element_text(size = 0)) #remove facet label
    
    
    #------------------------------------------------------------------#
    #Start network analysis: Carcinoma
    #------------------------------------------------------------------#
    match_index = match(symbol_match$gene[1:eigengene_size], aim2probedat$SYMBOL)
    r_tumor_cormat = tumor_cormat$r[match_index,match_index]
    p_tumor_cormat =  tumor_cormat$P[match_index,match_index]
    
    #replace non-sig correlation matrix values with 0
    sigmat = do.call(rbind, lapply(1:nrow(r_tumor_cormat), function(rr) {
      #rr=1
      loscols = do.call(cbind, lapply(1:ncol(r_tumor_cormat), function(cc) {
        #cc=3
        if((is.na(p_tumor_cormat[rr,cc]) ==T)) {cval = 0} else if((p_tumor_cormat[rr,cc] < (0.05/bonferroni_correction_factor))) {cval = (r_tumor_cormat[rr,cc])} else {cval = 0}
        data.frame(cval)
      }))
      loscols
    }))
    row.names(sigmat) = make.names(aim2probedat$SYMBOL[match_index], unique=TRUE) #take care of any duplicate probe maps
    
    #find where rows/cols are all 0 which represent genes with no significant correlation
    rowsumz = data.frame(rowSums(sigmat))
    rowsumz$index = c(1:nrow(rowsumz))
    index_to_drop = subset(rowsumz, rowSums.sigmat.==0)$index
    if(length(index_to_drop)==0) {sigmat = sigmat} else {sigmat = sigmat[-index_to_drop,-index_to_drop]}
    
    #Set conditional statement for if there is a network collapse
    if(nrow(sigmat)==0) { 
      tumor_netgraph = ggplot() + 
        annotate("text", x = 4, y = 25, size=8, label = le_text) + 
        theme_void()
      neighbors_tumor = data.frame("no_genes",0,0,0)
      names(neighbors_tumor) = c("gene","firstdegneighb","seconddegneighb","tissue_type")
    } else{
      
      #my data
      net = network(sigmat, directed = FALSE)
      net = network(net, directed = FALSE)
      network.vertex.names(net) = row.names(sigmat) #add gene names to nodes
      net %v% "direction" = c("Seed",rep("Correlated",(nrow(sigmat)-1)))
      
      #highlight any important stem genes
      stem_match_index = match(stem_markers_unique,row.names(sigmat)) #match stem gene to its position in sigmat matrix
      color_vector = rep("wheat",nrow(sigmat)) #make color palette for all genes
      color_vector = replace(color_vector, na.omit(stem_match_index), "darkred") #now replace colors for stem genes
      color_vector = replace(color_vector, 1, "seagreen4") #now replace color for seed gene
      
      set.seed(1) #this puts the nodes in the same position
      tumor_netgraph = ggnet2(net, mode = "kamadakawai",color=color_vector,
                              label = T,label.color = "black", label.size=4,
                              size="degree",max_size = 21,
                              alpha=0.6,
                              edge.size=0.1) +
        theme(legend.position="none") +
        theme(panel.background = element_rect(fill = "white")) +
        theme(legend.position="none") +
        ggtitle(paste(seed_gene[i],"Tumor Network",sep=" ")) +
        theme(plot.title = element_text(size = 18, vjust=1))
      
      ig2 <- asIgraph(net) #make network object into igraph object
      firstdegneighb = neighborhood.size(ig2, 1) #find number of first degree neighbors
      seconddegneighb = neighborhood.size(ig2, 2) #find number of second degreeneighbors
      
      #neighbors
      neighbors_tumor = data.frame(tumor_netgraph$data[,1],firstdegneighb,seconddegneighb)
      neighbors_tumor$tissue_type = "Carcinoma"
      
      #ask how much stemness is in this network based onthe function defined above
      tumor_stemness = define_stemness(sigmat,tumor_wt_dat)
      names(tumor_stemness) = paste("carcinoma_",names(tumor_stemness),sep="") #add a normal ski thing to these column names
      
    } #end if conditional statement
    
    
    #------------------------------------------------------------------#
    #Quantify network change
    #------------------------------------------------------------------#
    names(neighbors)[1] = "gene"
    names(neighbors_tumor)[1] = "gene"
    combined_neighbor_data = rbind(neighbors,neighbors_tumor)
    
    #gene uniqueness
    gene_counts = data.frame(table(combined_neighbor_data$gene))
    num_normal_genes = nrow(neighbors)
    num_tumor_genes = nrow(neighbors_tumor)
    num_shared_genes = nrow(subset(gene_counts, Freq==2))
    num_genes_changed = num_normal_genes - num_shared_genes
    percent_genes_changed = round((num_genes_changed / num_normal_genes)*100,2)
    
    #first deg neighbor
    normal_med_first_deg_neighbor = median(neighbors$firstdegneighb)
    tumor_med_first_deg_neighbor = median(neighbors_tumor$firstdegneighb)
    percent_change_first_deg_neighbor = round(((tumor_med_first_deg_neighbor-normal_med_first_deg_neighbor)/normal_med_first_deg_neighbor)*100,2)
    
    #second deg neighbor
    normal_med_second_deg_neighbor = median(neighbors$seconddegneighb)
    tumor_med_second_deg_neighbor = median(neighbors_tumor$seconddegneighb)
    percent_change_second_deg_neighbor = round(((tumor_med_second_deg_neighbor-normal_med_second_deg_neighbor)/normal_med_second_deg_neighbor)*100,2)
    
    #network coherence = correlation coefficients
    percent_change_med_rho = round(((median_rho_tumor-median_rho)/median_rho)*100,2)
    
    #combined percent change object
    integrated_network_change_metric = abs(percent_change_first_deg_neighbor) + abs(percent_change_second_deg_neighbor) + abs(percent_change_med_rho)
    
    #Stemness calculations
    stem_genes_change = tumor_stemness$carcinoma_no_stem_genes - normal_stemness$normalskin_no_stem_genes
    stem_gene_expression_proportion_change = tumor_stemness$carcinoma_stem_expression_propotion - normal_stemness$normalskin_stem_expression_propotion
    
    network_change_metrics = data.frame(
      seed_gene[i],
      num_normal_genes,num_tumor_genes,num_shared_genes,num_genes_changed,percent_genes_changed,
      normal_med_first_deg_neighbor, tumor_med_first_deg_neighbor, percent_change_first_deg_neighbor,
      normal_med_second_deg_neighbor, tumor_med_second_deg_neighbor, percent_change_second_deg_neighbor,
      as.vector(median_rho), as.vector(median_rho_tumor), as.vector(percent_change_med_rho),
      as.vector(integrated_network_change_metric),normal_stemness,tumor_stemness,stem_genes_change,stem_gene_expression_proportion_change)
    names(network_change_metrics)[c(1,13:16)] = c("seed_gene","normal_med_rho","tumor_med_rho","percent_change_med_rho","integrated_network_rewiring_metric")
    
    write.table(network_change_metrics, file=paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/mark_eigengenes/network_metric_results/",seed_gene[i],"network_change_metric.txt",sep=""),
                sep="\t",quote=F,row.names=F)
    
    #------------------------------------------------------------------#
    #Graphing
    #------------------------------------------------------------------#
    png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/mark_eigengenes/manifold_networkplots_v2/",seed_gene[i],"_manifolds.png",sep=""),width=21,height=7,units="in",res=200)
    print(multiplot(agexp_plot_seedgene.12,
                    agexp_plot.12, 
                    agexp_plot_tumor.12, cols=3))
    dev.off()
    
    png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/mark_eigengenes/manifold_networkplots_v2/",seed_gene[i],"_network_size_sensitivity.png",sep=""),width=10,height=10,units="in",res=100)
    print(multiplot(normal_skin_network_charct, normal_skin_network_charct.ind_gene,
                    tumor_network_charct, tumor_network_charct.ind_gene,cols=2))
    dev.off()
    
    png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/mark_eigengenes/network_figs_v2/",seed_gene[i],"_networks.png",sep=""),width=16,height=8,units="in",res=100)
    print(multiplot(netgraph, tumor_netgraph,cols=2))
    dev.off()
    
    #------------------------------------------------------------------#
    #End loop by returning network rewiring metric
    #------------------------------------------------------------------#
    return(network_change_metrics)
    
  },  error=function(e){})
  
}))   #end loop through seed genes        

#Print out integrated network rewiring results
write.table(manifolds_network_df, file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/mark_eigengenes/network_rewiring_analysis.txt", row.names=F, quote=F, sep="\t")

#End network rewiring analysis

###################################################################################################################################################
###################################################################################################################################################
###################################################################################################################################################
###################################################################################################################################################
###################################################################################################################################################
###################################################################################################################################################
# NETWORK FIG FOR MANUSCRIPT

#--------------------------------------------------------------------------------------------
#Carcinoma eigengenes: Sox9, Lgr6, Lrig1
#Sox9 network
match_index = match("Sox9", aim2probedat$SYMBOL) #match seed gene to its place in the matrix
symbol_match = data.frame(aim2probedat$SYMBOL, tumor_rhodf[,match_index], tumor_p_df[,match_index]) #put together gene names, expression data, and p-vals
names(symbol_match) = c("gene","rho","p")
symbol_match = symbol_match[order(-symbol_match$rho),] #order from high to low gene value
symbol_match_sox9 = symbol_match[1:100,]

#Lgr6 network
match_index = match("Lgr6", aim2probedat$SYMBOL) #match seed gene to its place in the matrix
symbol_match = data.frame(aim2probedat$SYMBOL, tumor_rhodf[,match_index], tumor_p_df[,match_index]) #put together gene names, expression data, and p-vals
names(symbol_match) = c("gene","rho","p")
symbol_match = symbol_match[order(-symbol_match$rho),] #order from high to low gene value
symbol_match_lgr6 = symbol_match[1:100,]

#Lrig1 network
match_index = match("Lrig1", aim2probedat$SYMBOL) #match seed gene to its place in the matrix
symbol_match = data.frame(aim2probedat$SYMBOL, tumor_rhodf[,match_index], tumor_p_df[,match_index]) #put together gene names, expression data, and p-vals
names(symbol_match) = c("gene","rho","p")
symbol_match = symbol_match[order(-symbol_match$rho),] #order from high to low gene value
symbol_match_lrig1 = symbol_match[1:100,]

#now get all the genes across these three networks
combined_genes = unique(c(symbol_match_sox9$gene,symbol_match_lgr6$gene,symbol_match_lrig1$gene)) #combined unique genes called in each of these networks
combined_genes = combined_genes[-match(c("Nnt"),combined_genes)] #drop Nnt because it is by itself

#Find any overlapping genes between these networks
combined_genes_df = rbind(symbol_match_sox9,symbol_match_lgr6,symbol_match_lrig1)
combined_count = data.frame(table(combined_genes_df$gene))
overlapping_genes = subset(combined_count, Freq>1)$Var1

rho_threshold = min(combined_genes_df$rho)
#color genes by unique or non-unique to the different networks

#now subset to the correct parts of the correlation matrix
combined_matrix = tumor_rhodf[match(combined_genes, aim2probedat$SYMBOL),match(combined_genes, aim2probedat$SYMBOL)]
combined_matrix[combined_matrix < rho_threshold] <- 0 #replace lements of the matrix less than rho threshold, which is minimum rho 
row.names(combined_matrix) = combined_genes #replace this baddies

#my data
net = network(combined_matrix, directed = FALSE)
net = network(net, directed = FALSE)
network.vertex.names(net) = row.names(combined_matrix) #add gene names to nodes
#now add a vector of yes/no gene names for the 3 genes we're interested in: Lgr6, Lrig1, Lgr5

dummy_vec = rep("Sox9",nrow(combined_matrix)) #make a dummy vector
dummy_vec[match(c(symbol_match_lgr6$gene), row.names(combined_matrix))] <- "Lgr6"
dummy_vec[match(c(symbol_match_lrig1$gene), row.names(combined_matrix))] <- "Lrig1"
dummy_vec[match(overlapping_genes, row.names(combined_matrix))] <- "Overlapping"
dummy_vec[match(c("Sox9","Lgr6","Lrig1"), row.names(combined_matrix))] <- "Seed"


net %v% "seed_status" = dummy_vec

set.seed(4) #this puts the nodes in the same position
network_fig_lgr = ggnet2(net, mode = "kamadakawai",color="seed_status",
                         label = T,label.color = "black", label.size=3,
                         size="degree",max_size = 15,
                         size.min = 0,#remove any isolated nodes
                         edge.color="black",
                         alpha=0.6,
                         edge.size = 0.03) +
  theme(panel.background = element_rect(fill = "white")) +
  theme(legend.position="none") +
  ggtitle(paste("Dorsal Network, threshold=",rho_threshold,sep="")) +
  theme(plot.title = element_text(size = 0, vjust=0, face="bold")) +
  scale_colour_manual(values = c("Overlapping"="darkorange3",
                                 "Sox9"="burlywood3","Lgr6"="darkkhaki","Lrig1"="darkolivegreen4",
                                 "Seed"="coral4")) +
  coord_equal()
network_fig_lgr

png("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Figs/Fig2/network_fig.png",
    width=8,height=8,units="in",res=300)
print(network_fig_lgr)
dev.off()

set.seed(4) #this puts the nodes in the same position
network_fig_lgr_w_legend = ggnet2(net, mode = "kamadakawai",color="seed_status",
                                  label = T,label.color = "black", label.size=3,
                                  size="degree",max_size = 15,
                                  size.min = 0,#remove any isolated nodes
                                  edge.color="black",
                                  alpha=0.6,
                                  edge.size = 0.03) +
  theme(panel.background = element_rect(fill = "white")) +
  theme(legend.position="right") +
  ggtitle(paste("Dorsal Network, threshold=",rho_threshold,sep="")) +
  theme(plot.title = element_text(size = 0, vjust=0, face="bold")) +
  scale_colour_manual(values = c("Overlapping"="darkorange3",
                                 "Sox9"="burlywood3","Lgr6"="darkkhaki","Lrig1"="darkolivegreen4",
                                 "Seed"="coral4")) +
  coord_equal()
png("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Figs/Fig2/network_fig_w_legend.png",
    width=15,height=15,units="in",res=300)
print(network_fig_lgr_w_legend)
dev.off()

###################################################################################################################################################
###################################################################################################################################################
###################################################################################################################################################

#NETWORK ANALYSIS: BULK EXPRESSION TRANCRIPTOME-WIDE ANALYSIS

names(rhodf) = aim2probedat$SYMBOL

#Run this for every single goddamn probe: re-name all the probes to make the unique if there are more than one gene symbols
make.unique(aim2probedat$SYMBOL, sep="_")

#Ask which genes has multiple probes and how many
probe_counts = data.frame(table(aim2probedat$SYMBOL))
probe_counts = subset(probe_counts, Freq>1)
nrow(probe_counts) #the number of genes measured by >1 probe

#Number of eigengenes calculable in the data = 13492
length(unique(aim2probedat$SYMBOL)) #Number of unique gene symbols
15850 - 13492
18273 - 2358 - 1273
15850 - 2358

dim(aim2probedat)

#BEGING THE NETWORK ANALYSIS BITCH
seed_gene = unique(aim2probedat$SYMBOL) #Get all gene symbols in the aim2 data
thresholds = seq(0.3,0.9,by=0.01) #Set a range of thresholds
j=21 #set a threshold of 0.5

manifolds_network_df = do.call(rbind, lapply(1:length(seed_gene), function(i){
  #manifolds_network_df = do.call(rbind, lapply(1:2, function(i){
  
  tryCatch({ #suppress error
    #i=1
    print(i)
    
    #Find some aliases
    match_index = match(seed_gene[i], aim2probedat$SYMBOL) #match seed gene to its place in the matrix
    
    #---------------------------------------#
    #Normal skin
    symbol_match = data.frame(aim2probedat$SYMBOL, rhodf[,match_index]) #put together gene names, expression data, and p-vals
    names(symbol_match) = c("gene","rho")
    symbol_match = symbol_match[order(-symbol_match$rho),] #order from high to low gene value
    symbol_match = subset(symbol_match, rho > thresholds[j])
    
    #get a special lil matrix to calculate second degree neighbors
    match_index = match(symbol_match$gene, aim2probedat$SYMBOL)
    sigmat = as.matrix(rhodf[,match_index])
    row.names(sigmat) = aim2probedat$SYMBOL
    
    if(dim(sigmat)[2]==1) {seconddegneighb = 0} else {
      sigmat = sigmat[,-1] #drop the first column because it consists of the seed gene
      edges = do.call(c, lapply(1:ncol(sigmat), function(p) {
        #p=1
        names(sigmat[,p][sigmat[,p]>thresholds[j]]) #get gene name of significant nodes that each first deg neighbor has --> thesea re the second degree neighbors
      }))
      seconddegneighb = length(unique(edges))
    }
    
    firstdegneighb = (nrow(symbol_match)-1)
    if(firstdegneighb==0) {coherence=0} else {
      coherence = mean(symbol_match$rho) }
    
    #---------------------------------------#
    #Tumor
    match_index = match(seed_gene[i], aim2probedat$SYMBOL) #match seed gene to its place in the matrix
    symbol_match_tumor = data.frame(aim2probedat$SYMBOL, tumor_rhodf[,match_index]) #put together gene names, expression data, and p-vals
    names(symbol_match_tumor) = c("gene","rho")
    symbol_match_tumor = symbol_match_tumor[order(-symbol_match_tumor$rho),] #order from high to low gene value
    symbol_match_tumor = subset(symbol_match_tumor, rho > thresholds[j])
    
    #get a special lil matrix to calculate second degree neighbors
    match_index = match(symbol_match_tumor$gene, aim2probedat$SYMBOL)
    sigmat = as.matrix(tumor_rhodf[,match_index])
    row.names(sigmat) = aim2probedat$SYMBOL
    
    if(dim(sigmat)[2]==1) {seconddegneighb_tumor = 0} else {
      sigmat = sigmat[,-1] #drop the first column because it consists of the seed gene
      
      edges = do.call(c, lapply(1:ncol(sigmat), function(p) {
        #p=1
        names(sigmat[,p][sigmat[,p]>thresholds[j]]) #get gene name of significant nodes that each first deg neighbor has --> thesea re the second degree neighbors
      }))
      seconddegneighb_tumor = length(unique(edges))
    }
    
    firstdegneighb_tumor = (nrow(symbol_match_tumor)-1)
    if(firstdegneighb_tumor==0) {coherence_tumor=0} else {
      coherence_tumor = mean(symbol_match_tumor$rho) }
    
    #------------------------------------------------------------------#
    #Quantify network change
    #------------------------------------------------------------------#
    
    #gene uniqueness
    num_normal_genes = nrow(symbol_match[-1,])
    num_tumor_genes = nrow(symbol_match_tumor[-1,])
    num_shared_genes = length(intersect(symbol_match$gene[-1],symbol_match_tumor$gene[-1])) #THIS HAS TOO MANY GODDAMN MATCHES BECAUSE THERE ARE MULTIPLE PROBES FOR THE SAME GENE SYMBOL
    num_genes_changed = num_tumor_genes - num_normal_genes - num_shared_genes
    percent_genes_changed = round(((num_genes_changed+1) / (num_normal_genes+1))*100,2)
    
    #first deg neighbor
    percent_change_first_deg_neighbor = round((((firstdegneighb_tumor+1)-(firstdegneighb+1))/(firstdegneighb+1))*100,2)
    
    #second deg neighbor
    percent_change_second_deg_neighbor = round((((seconddegneighb_tumor+1)-(seconddegneighb+1))/(seconddegneighb+1))*100,2)
    
    #network coherence = correlation coefficients
    if(coherence==0) {percent_change_coherence = round(coherence_tumor*100,2)} else{
      percent_change_coherence = round(((coherence_tumor-coherence)/coherence)*100,2) }
    
    #combined percent change object
    integrated_network_change_metric = abs(percent_change_first_deg_neighbor) + abs(percent_change_second_deg_neighbor) + abs(percent_change_coherence) + abs(percent_genes_changed)
    
    network_change_metrics = data.frame(
      seed_gene[i],
      num_normal_genes,num_tumor_genes,
      seconddegneighb,seconddegneighb_tumor,
      coherence, coherence_tumor,
      num_shared_genes,
      num_genes_changed,
      percent_genes_changed,
      percent_change_first_deg_neighbor,
      percent_change_second_deg_neighbor,
      percent_change_coherence,
      integrated_network_change_metric)
    
    names(network_change_metrics) = c("seed_gene",
                                      "firstdegneighb_nsk","firstdegneighb_tumor",
                                      "seconddegneighb_nsk","seconddegneighb_tumor",
                                      "coherence_nsk","coherence_tumor",
                                      "num_shared_genes",
                                      "num_genes_changed",
                                      "percent_genes_changed",
                                      "percent_change_first_deg_neighbor",
                                      "percent_change_second_deg_neighbor",
                                      "percent_change_coherence",
                                      "integrated_network_change_metric"
    )
    
    write.table(network_change_metrics, file=paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/mark_eigengenes/network_metric_results/",seed_gene[i],"_network_analysis.txt",sep=""),
                sep="\t",quote=F,row.names=F)
    
    #------------------------------------------------------------------#
    #End loop by returning network rewiring metric
    #------------------------------------------------------------------#
    return(network_change_metrics)
    
  },  error=function(e){})
  
}))   #end loop through seed genes        

#Print out integrated network rewiring results
write.table(manifolds_network_df, file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/mark_eigengenes/network_rewiring_analysis_rho_threshold_0.5.txt", row.names=F, quote=F, sep="\t")

#End network rewiring analysis

##############################################################################################################################
##################################################################################################################################
##################################################################################################################################
##################################################################################################################################
##################################################################################################################################
#Positive and negative correlations: up and down eigengenes

#read in the seed genes to create my own eigengenes
stem_hierachy_genes = read.xlsx(
  xlsxFile="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/mark_eigengenes/stem_gene_hierarchy_genes.xlsx",
  colNames=T
)

seed_gene = unique(stem_hierachy_genes$genes)

#------------------------------------------------------------------------------------------------#
#Get cell cycle seed genes from Seurat: https://satijalab.org/seurat/archive/v3.2/cell_cycle_vignette.html
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes

#Get mouse cell cycle genes
m.s.genes <- convertHumanGeneList(cc.genes.updated.2019$s.genes)
sphase_genes = data.frame("S phase",m.s.genes)
names(sphase_genes) = c("phase","gene")

m.g2m.genes <- convertHumanGeneList(cc.genes.updated.2019$g2m.genes)
g2mphase_genes = data.frame("G2/M phase",m.g2m.genes)
names(g2mphase_genes) = c("phase","gene")

cell_cycle_markers = rbind(sphase_genes,g2mphase_genes)

#------------------------------------------------------------------------------------------------#
#Emma Lundberg post-translationally controlled cell cycle genes: https://www.nature.com/articles/s41586-021-03232-9
emma_genes_1 = data.frame("silencing decreases proliferation", convertHumanGeneList(c("DUSP18","KLHL38","CD2BP2","SOX12")) )
names(emma_genes_1) = c("gene_type","gene")
emma_genes_2 = data.frame("silencing increases proliferation", convertHumanGeneList(c("JPH3")) )
names(emma_genes_2) = c("gene_type","gene")
emma_genes_3 = data.frame("uncoupled transcript-protein (Fig. 3)", convertHumanGeneList(c("DUSP19","DUSP18","N6AMT1","PHLDB1","TTC21B","UGDH","PAPSS1","FLI1","PC")) )
names(emma_genes_3) = c("gene_type","gene")

combined_emma= rbind(emma_genes_1,emma_genes_2,emma_genes_3)

#------------------------------------------------------------------------------------------------#
#------------------------------------------------------------------------------------------------#
#------------------------------------------------------------------------------------------------#
#------------------------------------------------------------------------------------------------#
#------------------------------------------------------------------------------------------------#
#------------------------------------------------------------------------------------------------#
#------------------------------------------------------------------------------------------------#
#------------------------------------------------------------------------------------------------#
#correlations

#CHANGE THE LINE BELOW FOR NEW SEED GENES BELOW: put a dataframe in: column 1 is gene type and column 2 is gene name
seed_gene_df = combined_emma
names(seed_gene_df) = c("gene_type","gene")

#CHANGE THE LINE BELOW FOR NEW SEED GENES BELOW: create and add the output dir here
output_dir = "/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/mark_eigengenes/cell_cycle_emma_lundberg/"

combined_neg_and_pos_genes = do.call(rbind, lapply(1:nrow(seed_gene_df), function(i){
  tryCatch({ #suppress error
    
    #i=1
    print(i)
    
    #Find some aliases
    match_index = match(seed_gene_df$gene[i], aim2probedat$SYMBOL) #match seed gene to its place in the matrix
    
    seed_gene_fig = plot_cells(big_cds,
                               genes=seed_gene_df$gene[i],
                               label_cell_groups=FALSE,
                               label_leaves=F,
                               label_branch_points=F,
                               graph_label_size=1.5,
                               show_trajectory_graph=F) +
      ggtitle(paste(seed_gene_df$gene[i],"seed gene")) + xlab("") + ylab("") + 
      theme(axis.text=element_text(size=0,color='black'),
            title=element_text(size=18),
            axis.title=element_text(size=0),
            axis.ticks = element_line(size=0),
            legend.position="none")+ 
      theme(strip.text.x = element_text(size = 0)) + #remove facet label
      scale_color_viridis_c(option = "inferno",name="integrated expression")
    
    
    plot_cells(big_cds,
               genes=c("Hba-a1","Hbb-bs","Hbb-bt"),
               label_cell_groups=FALSE,
               label_leaves=F,
               label_branch_points=F,
               graph_label_size=1.5,
               show_trajectory_graph=F) +
      #ggtitle(paste(seed_gene_df$gene[i],"seed gene")) + xlab("") + ylab("") + 
      theme(axis.text=element_text(size=0,color='black'),
            title=element_text(size=18),
            axis.title=element_text(size=0),
            axis.ticks = element_line(size=0),
            legend.position="none")+ 
      theme(strip.text.x = element_text(size = 10)) + #remove facet label
      scale_color_viridis_c(option = "inferno",name="integrated expression")
    
    colData(chemo_cds)
    ?plot_cells
    plot_cells(chemo_cds,
               color_cells_by = "nFeature_RNA",
               label_cell_groups=FALSE,
               label_leaves=F,
               label_branch_points=F,
               graph_label_size=1.5,
               show_trajectory_graph=F) +
      #ggtitle(paste(seed_gene_df$gene[i],"seed gene")) + xlab("") + ylab("") + 
      theme(axis.text=element_text(size=0,color='black'),
            title=element_text(size=18),
            axis.title=element_text(size=0),
            axis.ticks = element_line(size=0),
            legend.position="bottom")+ 
      theme(strip.text.x = element_text(size = 10)) + #remove facet label
      scale_color_viridis_c(option = "inferno",name="integrated expression")
    
    match("Vim",row.names(chemo_cds))
    plot_cells(chemo_cds,
               genes = "Vim",
               label_cell_groups=FALSE,
               label_leaves=F,
               label_branch_points=F,
               graph_label_size=1.5,
               show_trajectory_graph=F) +
      #ggtitle(paste(seed_gene_df$gene[i],"seed gene")) + xlab("") + ylab("") + 
      theme(axis.text=element_text(size=0,color='black'),
            title=element_text(size=18),
            axis.title=element_text(size=0),
            axis.ticks = element_line(size=0),
            legend.position="bottom")+ 
      theme(strip.text.x = element_text(size = 10)) + #remove facet label
      scale_color_viridis_c(option = "inferno",name="integrated expression")
    
    #------------------------------------------------------------------------------------------------#
    #------------------------------------------------------------------------------------------------#
    #Normal skin
    symbol_match = data.frame(aim2probedat$SYMBOL, rhodf[,match_index], p_df[,match_index]) #put together gene names, expression data, and p-vals
    symbol_match = symbol_match[order(-symbol_match$rhodf...match_index.),] #order from high to low gene values
    names(symbol_match) = c("gene","rho","p")
    
    #------------------------------#
    #POSTIIVE network: now subset to network (eigengene) that you will create the eigengen with
    marker_df = data.frame(unique(alias2SymbolTable(symbol_match$gene[1:(eigengene_size+1)], species = "Mm")),paste(seed_gene_df$gene[i]," Normal skin",sep=""),paste(seed_gene_df$gene[i]," Normal skin",sep=""))
    names(marker_df) = c("gene","cell_type","module")
    
    #all cells eigengene
    marker_df = na.omit(marker_df)
    nsk_pos_genes = marker_df$gene
    nsk_pos = plot_cells(big_cds,
                         genes=marker_df,
                         label_cell_groups=FALSE,
                         label_leaves=F,
                         label_branch_points=F,
                         graph_label_size=1.5,
                         show_trajectory_graph=F) +
      ggtitle(paste(seed_gene_df$gene[i],"NSk Positive")) + xlab("") + ylab("") + 
      theme(axis.text=element_text(size=0,color='black'),
            title=element_text(size=18),
            axis.title=element_text(size=0),
            axis.ticks = element_line(size=0),
            legend.position="none")+ 
      theme(strip.text.x = element_text(size = 0)) + #remove facet label
      scale_color_viridis_c(option = "inferno",name="integrated expression")
    
    #------------------------------#
    #NEGATIVE network: now subset to network (eigengene) that you will create the eigengen with
    marker_df = data.frame(unique(alias2SymbolTable(symbol_match$gene[(nrow(symbol_match)-(eigengene_size+1)):nrow(symbol_match)], species = "Mm")),paste(seed_gene_df$gene[i]," Normal skin",sep=""),paste(seed_gene_df$gene[i]," Normal skin",sep=""))
    names(marker_df) = c("gene","cell_type","module")
    
    #all cells eigengene
    marker_df = na.omit(marker_df)
    nsk_neg_genes = marker_df$gene
    nsk_neg = plot_cells(big_cds,
                         genes=marker_df,
                         label_cell_groups=FALSE,
                         label_leaves=F,
                         label_branch_points=F,
                         graph_label_size=1.5,
                         show_trajectory_graph=F) +
      ggtitle(paste(seed_gene_df$gene[i],"NSk Negative")) + xlab("") + ylab("") + 
      theme(axis.text=element_text(size=0,color='black'),
            title=element_text(size=18),
            axis.title=element_text(size=0),
            axis.ticks = element_line(size=0),
            legend.position="none")+ 
      theme(strip.text.x = element_text(size = 0)) + #remove facet label
      scale_color_viridis_c(option = "inferno",name="integrated expression")
    
    #------------------------------------------------------------------------------------------------#
    #------------------------------------------------------------------------------------------------#
    #Tumor
    symbol_match = data.frame(aim2probedat$SYMBOL, tumor_rhodf[,match_index], tumor_p_df[,match_index]) #put together gene names, expression data, and p-vals
    symbol_match = symbol_match[order(-symbol_match$tumor_rhodf...match_index.),] #order from high to low gene values
    names(symbol_match) = c("gene","rho","p")
    
    #------------------------------#
    #POSTIIVE network: now subset to network (eigengene) that you will create the eigengen with
    marker_df = data.frame(unique(alias2SymbolTable(symbol_match$gene[1:(eigengene_size+1)], species = "Mm")),paste(seed_gene_df$gene[i]," Normal skin",sep=""),paste(seed_gene_df$gene[i]," Normal skin",sep=""))
    names(marker_df) = c("gene","cell_type","module")
    
    #all cells eigengene
    marker_df = na.omit(marker_df)
    tumor_pos_genes = marker_df$gene
    tumor_pos = plot_cells(big_cds,
                           genes=marker_df,
                           label_cell_groups=FALSE,
                           label_leaves=F,
                           label_branch_points=F,
                           graph_label_size=1.5,
                           show_trajectory_graph=F) +
      ggtitle(paste(seed_gene_df$gene[i],"CA Positive")) + xlab("") + ylab("") + 
      theme(axis.text=element_text(size=0,color='black'),
            title=element_text(size=18),
            axis.title=element_text(size=0),
            axis.ticks = element_line(size=0),
            legend.position="none")+ 
      theme(strip.text.x = element_text(size = 0)) + #remove facet label
      scale_color_viridis_c(option = "inferno",name="integrated expression")
    
    #------------------------------#
    #NEGATIVE network: now subset to network (eigengene) that you will create the eigengen with
    marker_df = data.frame(unique(alias2SymbolTable(symbol_match$gene[(nrow(symbol_match)-(eigengene_size+1)):nrow(symbol_match)], species = "Mm")),paste(seed_gene_df$gene[i]," Normal skin",sep=""),paste(seed_gene_df$gene[i]," Normal skin",sep=""))
    names(marker_df) = c("gene","cell_type","module")
    
    #all cells eigengene
    marker_df = na.omit(marker_df)
    tumor_neg_genes = marker_df$gene
    tumor_neg = plot_cells(big_cds,
                           genes=marker_df,
                           label_cell_groups=FALSE,
                           label_leaves=F,
                           label_branch_points=F,
                           graph_label_size=1.5,
                           show_trajectory_graph=F) +
      ggtitle(paste(seed_gene_df$gene[i],"CA Negative")) + xlab("") + ylab("") + 
      theme(axis.text=element_text(size=0,color='black'),
            title=element_text(size=18),
            axis.title=element_text(size=0),
            axis.ticks = element_line(size=0),
            legend.position="none")+ 
      theme(strip.text.x = element_text(size = 0)) + #remove facet label
      scale_color_viridis_c(option = "inferno",name="integrated expression")
    
    #------------------------------------------------------------------------------------------------#
    #------------------------------------------------------------------------------------------------#
    #Combine and print out eigengene figs
    
    png(paste(output_dir,seed_gene_df$gene[i],"_eigengene.png",sep=""),width=12,height=12,units="in",res=300)
    print(multiplot(nsk_pos,nsk_neg,
                    tumor_pos,tumor_neg,
                    cols=2))
    dev.off()
    
    #Print out seed gene
    png(paste(output_dir,seed_gene_df$gene[i],"_seedgene.png",sep=""),width=6,height=6,units="in",res=300)
    print(seed_gene_fig)
    dev.off()
    
    #------------------------------------------------------------------------------------------------#
    #print out the genes in each network being plotted as eigengenes
    combined_network_genes = data.frame(seed_gene_df$gene[i],cbind.fill(nsk_pos_genes, nsk_neg_genes,
                                                                        tumor_pos_genes, tumor_neg_genes))
    names(combined_network_genes) = c("seed_gene","nsk_pos_genes","nsk_neg_genes","tumor_pos_genes","tumor_neg_genes")
    return(combined_network_genes)
    
  },  error=function(e){})
}))

write.table(combined_neg_and_pos_genes,
            file=paste(output_dir,"2_eigengene_constituent_genes.txt",sep=""),
            row.names=F,quote=F,sep="\t")

write.table(seed_gene_df,
            file=paste(output_dir,"1_eigengene_constituent_genes.txt",sep=""),
            row.names=F,quote=F,sep="\t")



##################################################################################################################################
##################################################################################################################################
##################################################################################################################################
#Positive and Negative eigengenes for a Carmen Report : Rasa4
library(readxl)

eigengene_excel = "/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/apriori_eigengenes/gapdhs-carwt-probe\ 2.xls"
sheets = excel_sheets(eigengene_excel) #et a vector of sheets names


for(i in 1:length(sheets)) {
  
  #i=1
  print(paste(i,sheets[i],sep=" "))
  
  datasheet = data.frame(read_excel(eigengene_excel, sheet = sheets[i]))
  datasheet = datasheet[-c(1:21),] #get rid of Carmen report shit at the top
  names(datasheet) = datasheet[1,]
  datasheet = datasheet[-1,]
  datasheet$rho_a = as.numeric(datasheet$rho_a) #order from high to low
  
  #-----------------------------------------------------#
  #Positive 
  datasheet = datasheet[order(-datasheet$rho_a),] #order from high to low
  
  #remake the marker df dataframe with the matched gene names
  marker_df = data.frame(unique(alias2SymbolTable(unique(datasheet$name_2[1:100]), species = "Mm")),"random","random")
  names(marker_df) = c("gene","cell_type","module")
  marker_df = na.omit(marker_df) #remove non-matching genes
  
  #construct binary expression dataframe
  subcounts = t(as.matrix(big_cds@assays@data$counts[na.omit(match(marker_df$gene,row.names(big_cds@assays@data$counts))),]))  #exctract gene counts and transpose as a matrix
  subcounts = replace(subcounts, subcounts != 0, 1) #replace non-zero elements in the matrix with 1
  summed_binary_exprssion = data.frame(big_cds@int_colData@listData$reducedDims@listData$UMAP, rowSums(subcounts) ) #combine UMAP coordinates with summed binary expression across genes, per cell
  names(summed_binary_exprssion) = c("UMAP1","UMAP2","UMAP3","summed_binary_expression_value")
  
  #Now graph summed binary expression
  summed_binary_expression_plot_pos = ggplot(data=summed_binary_exprssion, aes(x=UMAP1,y=UMAP2,color=summed_binary_expression_value)) +
    geom_point(alpha=0.4, size=0.2) +
    theme(axis.ticks = element_blank()) +
    ggtitle(paste("POSTIVIE Summed binary expression: ", sheets[i],sep=" ")) +
    scale_color_viridis_c(option = "inferno",name="summed binary expression") +
    theme(legend.position="bottom") +
    theme(axis.text=element_text(size=0,color='black'),
          axis.title=element_text(size=0,color='black')) + 
    theme(plot.title = element_text(size = 15, vjust=0)) +
    theme(strip.text.x = element_text(size = 0)) + #remove facet label
    theme(panel.grid.major=element_blank(), 
          panel.grid.minor=element_blank(), 
          panel.background=element_blank()) +
    theme(axis.line = element_line(size = 0.3, colour = "black"))
  
  classic_eigengene_expression_plot_pos = 
    plot_cells(big_cds,
               genes=marker_df,
               label_cell_groups=FALSE,
               label_leaves=F,
               label_branch_points=F,
               graph_label_size=1.5,
               show_trajectory_graph=F) +
    theme(axis.text=element_text(size=0,color='black'),
          axis.title=element_text(size=0,color='black')) + 
    theme(axis.ticks = element_blank()) +
    theme(plot.title = element_text(size = 15, vjust=0)) +
    theme(legend.position="bottom") +
    scale_color_viridis_c(option = "inferno",name="integrated expression") +
    ggtitle(paste("POSITIVE Classic eigengene: ", sheets[i],sep=" "))+ #continuous
    theme(strip.text.x = element_text(size = 0)) #remove facet label
  
  #-----------------------------------------------------#
  #NEGATIVE 
  datasheet = datasheet[order(datasheet$rho_a),] #order from low to high
  
  #remake the marker df dataframe with the matched gene names
  marker_df = data.frame(unique(alias2SymbolTable(unique(datasheet$name_2[1:100]), species = "Mm")),"random","random")
  names(marker_df) = c("gene","cell_type","module")
  marker_df = na.omit(marker_df) #remove non-matching genes
  
  #construct binary expression dataframe
  subcounts = t(as.matrix(big_cds@assays@data$counts[na.omit(match(marker_df$gene,row.names(big_cds@assays@data$counts))),]))  #exctract gene counts and transpose as a matrix
  subcounts = replace(subcounts, subcounts != 0, 1) #replace non-zero elements in the matrix with 1
  summed_binary_exprssion = data.frame(big_cds@int_colData@listData$reducedDims@listData$UMAP, rowSums(subcounts) ) #combine UMAP coordinates with summed binary expression across genes, per cell
  names(summed_binary_exprssion) = c("UMAP1","UMAP2","UMAP3","summed_binary_expression_value")
  
  #Now graph summed binary expression
  summed_binary_expression_plot_neg = ggplot(data=summed_binary_exprssion, aes(x=UMAP1,y=UMAP2,color=summed_binary_expression_value)) +
    geom_point(alpha=0.4, size=0.2) +
    theme(axis.ticks = element_blank()) +
    ggtitle(paste("NEGATIVE Summed binary expression: ", sheets[i],sep=" ")) +
    scale_color_viridis_c(option = "inferno",name="summed binary expression") +
    theme(legend.position="bottom") +
    theme(axis.text=element_text(size=0,color='black'),
          axis.title=element_text(size=0,color='black')) + 
    theme(plot.title = element_text(size = 15, vjust=0)) +
    theme(strip.text.x = element_text(size = 0)) + #remove facet label
    theme(panel.grid.major=element_blank(), 
          panel.grid.minor=element_blank(), 
          panel.background=element_blank()) +
    theme(axis.line = element_line(size = 0.3, colour = "black"))
  
  classic_eigengene_expression_plot_neg = 
    plot_cells(big_cds,
               genes=marker_df,
               label_cell_groups=FALSE,
               label_leaves=F,
               label_branch_points=F,
               graph_label_size=1.5,
               show_trajectory_graph=F) +
    theme(axis.text=element_text(size=0,color='black'),
          axis.title=element_text(size=0,color='black')) + 
    theme(axis.ticks = element_blank()) +
    theme(plot.title = element_text(size = 15, vjust=0)) +
    theme(legend.position="bottom") +
    scale_color_viridis_c(option = "inferno",name="integrated expression") +
    ggtitle(paste("NEGATIVE Classic eigengene: ", sheets[i],sep=" "))+ #continuous
    theme(strip.text.x = element_text(size = 0)) #remove facet label
  
  
  
  #-----------#
  #print out binary and classic plots
  png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/balmain_eigene/eigengene_PNGs_v",eigengene_verison,"/eigengene_",sheets[i],".png",sep=""),
      width=14,height=15,units="in",res=400)
  
  print(multiplot(classic_eigengene_expression_plot_pos,
                  summed_binary_expression_plot_pos,
                  classic_eigengene_expression_plot_neg,
                  summed_binary_expression_plot_neg,
                  cols=2))
  
  dev.off()
  
}

#-----------------------------------------------------#
#-----------------------------------------------------#
#-----------------------------------------------------#
#-----------------------------------------------------#
#Positive and negative eigengenes from seed gene #Return flag
#I need to try this for all the paper's 17 stem genes and see if we see switching from one spike to another in the positive and negative networks

seed_gene = c("Lgr5",
              "Bmi1","Sox4","Lrig1",
              "Lgr6","Sox9",
              "Sox2","Prom1","Foxa1","Krt15","Krt19","Pitx1","Psca","Prom2","Klf5",
              "Cd44","Dkkl1")
eigengene_size = 100

for(i in 1:length(seed_gene)) {
  
  #i=17
  print(i)
  
  #---------------------------------------#
  #CARCINOMA EIGENGENEE
  
  if( seed_gene[i]=="Prom1" ) {match_index = 12747} else {
    match_index = match(seed_gene[i], aim2probedat$SYMBOL) #match seed gene to its place in the matrix
  }
  
  symbol_match = data.frame(aim2probedat$SYMBOL, tumor_rhodf[,match_index], tumor_p_df[,match_index]) #put together gene names, expression data, and p-vals
  symbol_match = symbol_match[order(-symbol_match$tumor_rhodf...match_index.),] #order from high to low gene values
  names(symbol_match) = c("gene","rho","p")
  
  #---------------------------------------#
  #POSITIVE SCC
  #make marker df for grahing eigengene
  marker_df = data.frame(unique(alias2SymbolTable(symbol_match$gene[1:(eigengene_size)], species = "Mm")),paste(seed_gene[i]," Normal skin",sep=""),paste(seed_gene[i]," Normal skin",sep=""))
  names(marker_df) = c("gene","cell_type","module")
  marker_df = na.omit(marker_df)
  
  classic_eigengene_expression_plot_pos_scc = 
    plot_cells(big_cds,
               genes=marker_df,
               label_cell_groups=FALSE,
               label_leaves=F,
               label_branch_points=F,
               graph_label_size=1.5,
               show_trajectory_graph=F,
               cell_size=1.2) +
    xlab("") + ylab("") + 
    theme(axis.ticks = element_blank()) +
    scale_color_viridis_c(option = "inferno",name="") +
    theme(axis.text=element_text(size=0,color='black'),
          axis.title=element_text(size=0,color='black')) + theme(plot.title = element_text(size = 0, vjust=0)) +
    theme(strip.text.x = element_text(size = 0))  +
    annotate("text", label = paste(seed_gene[i]," Car\n+metagene",sep=""), 
             x = -3, y = -4.5, vjust = 0,size=10) +
    theme(legend.position=c(5, 5),
          legend.key.size = unit(1, 'cm'),
          legend.text = element_text(size=20)) +
    xlim(-4.1,0.9) +
    ylim(-7.5, -3.8)
  
  #-----------------------------------------------------#
  #NEGATIVE SCC
  marker_df = data.frame(unique(alias2SymbolTable(symbol_match$gene[(nrow(symbol_match)-eigengene_size):nrow(symbol_match)], species = "Mm")),paste(seed_gene[i]," Normal skin",sep=""),paste(seed_gene[i]," Normal skin",sep=""))
  names(marker_df) = c("gene","cell_type","module")
  marker_df = na.omit(marker_df)
  
  classic_eigengene_expression_plot_neg_scc = 
    plot_cells(big_cds,
               genes=marker_df,
               label_cell_groups=FALSE,
               label_leaves=F,
               label_branch_points=F,
               graph_label_size=1.5,
               show_trajectory_graph=F,
               cell_size=1.2)+
    xlab("") + ylab("") + 
    theme(axis.ticks = element_blank()) +
    scale_color_viridis_c(option = "inferno",name="") +
    theme(axis.text=element_text(size=0,color='black'),
          axis.title=element_text(size=0,color='black')) + theme(plot.title = element_text(size = 0, vjust=0)) +
    theme(strip.text.x = element_text(size = 0))  +
    annotate("text", label = paste(seed_gene[i]," Car\n-metagene",sep=""), 
             x = -3, y = -4.5, vjust = 0,size=10) +
    theme(legend.position=c(5, 5),
          legend.key.size = unit(1, 'cm'),
          legend.text = element_text(size=20)) +
    xlim(-4.1,0.9) +
    ylim(-7.5, -3.8)
  
  
  #-----------#
  #print out binary and classic plots
  png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Figs/Fig4/pos_neg_manifolds/",seed_gene[i],".png",sep=""),
      width=10,height=5,units="in",res=400)
  
  print(multiplot(
    
    classic_eigengene_expression_plot_pos_scc,
    classic_eigengene_expression_plot_neg_scc,
    cols=2))
  
  dev.off()
  
}#end positive and negative eigengene loop


################################################################################################################
################################################################################################################

#Seed genes only

seed_gene = c("Lgr5",
              "Bmi1","Sox4","Lrig1",
              "Lgr6","Sox9",
              "Sox2","Prom1","Foxa1","Krt15","Krt19","Pitx1","Psca","Prom2","Klf5",
              "Cd44")
length(seed_gene)

seed_gene_only = 
  plot_cells(big_cds,
             genes=seed_gene,
             label_cell_groups=FALSE,
             label_leaves=F,
             label_branch_points=F,
             graph_label_size=1.5,
             show_trajectory_graph=F)+
  xlab("") + ylab("") + 
  theme(axis.ticks = element_blank()) +
  scale_color_viridis_c(option = "inferno",name="") +
  theme(axis.text=element_text(size=0,color='black'),
        axis.title=element_text(size=0,color='black')) + theme(plot.title = element_text(size = 0, vjust=0)) +
  theme(strip.text.x = element_text(size = 19))  +
  theme(legend.position="none",
        legend.key.size = unit(1, 'cm'),
        legend.text = element_text(size=20))

png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Nature/Round\ 2/Figs/Revised\ Fig/seed_genes.png",sep=""),
    width=16,height=16,units="in",res=400)
print(seed_gene_only)
dev.off()


#With legend
seed_gene_only_legend = 
  plot_cells(big_cds,
             genes=seed_gene,
             label_cell_groups=FALSE,
             label_leaves=F,
             label_branch_points=F,
             graph_label_size=1.5,
             show_trajectory_graph=F)+
  xlab("") + ylab("") + 
  theme(axis.ticks = element_blank()) +
  scale_color_viridis_c(option = "inferno",name="") +
  theme(axis.text=element_text(size=0,color='black'),
        axis.title=element_text(size=0,color='black')) + theme(plot.title = element_text(size = 0, vjust=0)) +
  theme(strip.text.x = element_text(size = 19))  +
  theme(legend.position="bottom")

png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Nature/Round\ 2/Figs/Revised\ Fig/seed_genes_with_legend.png",sep=""),
    width=16,height=16,units="in",res=600)
print(seed_gene_only_legend)
dev.off()


################################################################################################################
################################################################################################################

row.names(big_cds)[grep("Dkk",row.names(big_cds))]

#Now plot several interesting genes
ind_genes = plot_cells(big_cds,
                       genes=c("Dkk1"),
                       label_cell_groups=FALSE,
                       label_leaves=F,
                       label_branch_points=F,
                       graph_label_size=1.5,
                       show_trajectory_graph=F) +
  theme(axis.text=element_text(size=0,color='black'),
        axis.title=element_text(size=0,color='black')) + 
  theme(axis.ticks = element_blank()) +
  theme(plot.title = element_text(size = 15, vjust=0)) +
  theme(legend.position="bottom") +
  scale_color_viridis_c(option = "inferno",name="log expression") +
  ggtitle(paste(""))+ #continuous
  theme(strip.text.x = element_text(size = 15)) #Increase facet label text size

png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/balmain_eigene/eigengene_PNGs_v",eigengene_verison,"/1_ind_genes.png",sep=""),
    width=21,height=7,units="in",res=400)

print(ind_genes)

dev.off()

##################################################################################################################################
##################################################################################################################################
##################################################################################################################################
#Fig. 1 manifolds

seed_gene = c("Lgr5","Lgr6")

eigengene_size = 100
manifolds_network_df = do.call(rbind, lapply(1:length(seed_gene), function(i){
  #manifolds_network_df = do.call(rbind, lapply(1:2, function(i){
  
  tryCatch({ #suppress error
    #i=2
    print(i)
    
    match_index = match(seed_gene[i], aim2probedat$SYMBOL) #match seed gene to its place in the matrix
    
    #---------------------------------------#
    #Normal skin
    symbol_match = data.frame(aim2probedat$SYMBOL, rhodf[,match_index], p_df[,match_index]) #put together gene names, expression data, and p-vals
    symbol_match = symbol_match[order(-symbol_match$rhodf...match_index.),] #order from high to low gene values
    names(symbol_match) = c("gene","rho","p")
    
    #make marker df for grahing eigengene
    marker_df = data.frame(unique(alias2SymbolTable(symbol_match$gene[1:(eigengene_size+1)], species = "Mm")),paste(seed_gene[i]," Normal skin",sep=""),paste(seed_gene[i]," Normal skin",sep=""))
    names(marker_df) = c("gene","cell_type","module")
    marker_df = na.omit(marker_df)
    
    #------------------------------#
    #4a. Normal skin WT: calculate correlation matrix
    #------------------------------#
    #all cells: Seed gene
    agexp_plot_seedgene.12 = plot_cells(big_cds,
                                        genes=marker_df$gene[1], #seed gene is the first row of the markers file
                                        label_cell_groups=FALSE,
                                        label_leaves=F,
                                        label_branch_points=F,
                                        graph_label_size=1.5,
                                        show_trajectory_graph=F) +
      xlab("") + ylab("") + 
      theme(axis.ticks = element_blank()) +
      scale_color_viridis_c(option = "inferno",name="expression") +
      ggtitle(paste(seed_gene[i],"seed gene",sep=" "))+ #continuous
      theme(axis.text=element_text(size=0,color='black'),
            axis.title=element_text(size=0,color='black')) + theme(plot.title = element_text(size = 0, vjust=0)) +
      theme(strip.text.x = element_text(size = 0)) +
      theme(legend.position=c(0.8, 0.2),
            legend.key.size = unit(1, 'cm'),
            legend.text = element_text(size=12.5)) #remove facet label
    
    #all cells eigengene
    agexp_plot.12 = plot_cells(big_cds,
                               genes=marker_df,
                               label_cell_groups=FALSE,
                               label_leaves=F,
                               label_branch_points=F,
                               graph_label_size=1.5,
                               show_trajectory_graph=F) +
      xlab("") + ylab("") + 
      theme(axis.ticks = element_blank()) +
      scale_color_viridis_c(option = "inferno",name="integrated expression") +
      theme(axis.text=element_text(size=0,color='black'),
            axis.title=element_text(size=0,color='black')) + theme(plot.title = element_text(size = 0, vjust=0)) +
      theme(strip.text.x = element_text(size = 0))  +
      theme(legend.position=c(0.8, 0.2),
            legend.key.size = unit(1, 'cm'),
            legend.text = element_text(size=12.5)) #remove facet label
    
    #---------------------------------------#
    #tumor
    match_index = match(seed_gene[i], aim2probedat$SYMBOL) #match seed gene to its place in the matrix
    symbol_match = data.frame(aim2probedat$SYMBOL, tumor_rhodf[,match_index], tumor_p_df[,match_index]) #put together gene names, expression data, and p-vals
    symbol_match = symbol_match[order(-symbol_match$tumor_rhodf...match_index.),] #order from high to low gene values
    names(symbol_match) = c("gene","rho","p")
    
    #make marker df for grahing eigengene
    marker_df = data.frame(unique(alias2SymbolTable(symbol_match$gene[1:(eigengene_size+1)], species = "Mm")),paste(seed_gene[i],"Carcinoma",sep=""),paste(seed_gene[i],"Carcinoma",sep=""))
    names(marker_df) = c("gene","cell_type","module")
    marker_df = na.omit(marker_df)
    
    #all cells eigengene
    agexp_plot_tumor.12 = plot_cells(big_cds,
                                     genes=marker_df,
                                     label_cell_groups=FALSE,
                                     label_leaves=F,
                                     label_branch_points=F,
                                     graph_label_size=1.5,
                                     show_trajectory_graph=F) +
      xlab("") + ylab("") + 
      theme(axis.ticks = element_blank()) +
      theme(legend.position="none") +
      scale_color_viridis_c(option = "inferno",name="integrated expression") +
      theme(axis.text=element_text(size=0,color='black'),
            axis.title=element_text(size=0,color='black')) + theme(plot.title = element_text(size = 0, vjust=0)) +
      theme(strip.text.x = element_text(size = 0))  +
      theme(legend.position=c(0.8, 0.2),
            legend.key.size = unit(1, 'cm'),
            legend.text = element_text(size=12.5))#remove facet label
    
    #------------------------------------------------------------------#
    #Graphing
    #------------------------------------------------------------------#
    png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Figs/Fig1/eigengene_manifolds/",seed_gene[i],"_manifolds.png",sep=""),width=21,height=7,units="in",res=200)
    print(multiplot(agexp_plot_seedgene.12,
                    agexp_plot.12, 
                    agexp_plot_tumor.12, cols=3))
    dev.off()
    
    
  },  error=function(e){})
  
}))   #end loop through seed genes        

################################################################################################################
#Supplementary methods showing eigengene stability across different sizes of eigengenes

seed_gene = c("Lgr5",
              "Bmi1","Sox4","Lrig1",
              "Lgr6","Sox9",
              "Sox2","Prom1","Foxa1","Krt15","Krt19","Pitx1","Psca","Prom2","Klf5",
              "Cd44",
              "Ppp2ca",
              "Wnt4")

#Make vectors of eigengene sizes where there are 21 possible sizes and the FIRST position in the vector is a placehodler
eigengene_size = c(0,10,20,30,40,50,
                   75,100,125,150,175,
                   200,300,500,750,1000,
                   2000,5000,7500,10000,15000)

dim(tumor_rhodf)

manifolds_network_df = do.call(rbind, lapply(1:length(seed_gene), function(i){
  #manifolds_network_df = do.call(rbind, lapply(17:18, function(i){
  
  tryCatch({ #suppress error
    #i=17
    print(i)
    
    #Get the correct match_index in the bulk expression data
    if( seed_gene[i]=="Prom1" ) {match_index = 12747} else {
      match_index = match(seed_gene[i], aim2probedat$SYMBOL) #match seed gene to its place in the matrix
    }
    
    #---------------------------------------#
    #tumor
    match_index = match(seed_gene[i], aim2probedat$SYMBOL) #match seed gene to its place in the matrix
    symbol_match = data.frame(aim2probedat$SYMBOL, tumor_rhodf[,match_index], tumor_p_df[,match_index]) #put together gene names, expression data, and p-vals
    symbol_match = symbol_match[order(-symbol_match$tumor_rhodf...match_index.),] #order from high to low gene values
    names(symbol_match) = c("gene","rho","p")
    
    #---------------------------------------#
    #Now start looping through eigengene definitions
    
    for(j in 2:length(eigengene_size)) { #start indexing at 2 bc the first position is 0
      
      #j=2
      print(paste("i=",i," j=",j,sep=""))
      
      #make marker df for grahing eigengene
      marker_df = data.frame(unique(alias2SymbolTable(symbol_match$gene[1:(eigengene_size[j])], species = "Mm")),paste(seed_gene[i],"Carcinoma",sep=""),paste(seed_gene[i],"Carcinoma",sep=""))
      names(marker_df) = c("gene","cell_type","module")
      marker_df = na.omit(marker_df)
      
      #Plot eigengene size
      agexp_plot_tumor.12 = plot_cells(big_cds,
                                       genes=marker_df,
                                       label_cell_groups=FALSE,
                                       label_leaves=F,
                                       label_branch_points=F,
                                       graph_label_size=1.5,
                                       show_trajectory_graph=F,
                                       cell_size=0.2) +
        xlab("") + ylab("") +
        ggtitle(eigengene_size[j]) +
        theme(axis.ticks = element_blank()) +
        theme(legend.position="none") +
        scale_color_viridis_c(option = "inferno",name="integrated expression") +
        theme(axis.text=element_text(size=0,color='black'),
              axis.title=element_text(size=0,color='black')) + theme(plot.title = element_text(size = 15, vjust=0, face="bold")) +
        theme(strip.text.x = element_text(size = 0))  +
        theme(legend.position="none")#remove facet label
      
      #assign the manifold plot to a looped variable name "pj" where j is the index of eigengene size
      assign(paste("p", j, sep=""), agexp_plot_tumor.12)
    }
    
    #------------------------------------------------------------------#
    #Graphing
    #------------------------------------------------------------------#
    png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Supplemental\ Information/eigengene_stability_figs/"
              ,seed_gene[i],"_eigengene_sizes.png",sep=""),width=16,height=20,units="in",res=400)
    print(multiplot(p2,p3,p4,p5,p6,p7,p8,p9,p10,p11,
                    p12,p13,p14,p15,p16,p17,p18,p19,p20,p21,
                    cols=4))
    dev.off()
    
    
  },  error=function(e){})
  
}))   #end loop through 

#----------------------------------------------------------------------------------------------------#

#Ferroptosis: Gpx4 stuff

row.names(big_cds)[grep("Ml",row.names(big_cds))]

#Molecular network driving intrinsic Gpx4 dependency
c("Hif1a","Plin2","Hilpda","Acsl4")

row.names(big_cds)[grep("G0",row.names(big_cds))]

match(c("Hif1a","Plin2","Hilpda","Acsl4",), marker_df$gene)

#make marker df for grahing eigengene
marker_df = data.frame(unique(alias2SymbolTable(c("Hif1a","Plin2","Hilpda","Acsl4","Gpx4","G0s2"), species = "Mm")),paste(seed_gene[i],"Carcinoma",sep=""),paste(seed_gene[i],"Carcinoma",sep=""))
names(marker_df) = c("gene","cell_type","module")
marker_df = na.omit(marker_df)

#all cells eigengene
agexp_plot_tumor.12 = plot_cells(big_cds,
                                 genes=marker_df,
                                 label_cell_groups=FALSE,
                                 label_leaves=F,
                                 label_branch_points=F,
                                 graph_label_size=1.5,
                                 show_trajectory_graph=F) +
  xlab("") + ylab("") + 
  theme(axis.ticks = element_blank()) +
  theme(legend.position="none") +
  scale_color_viridis_c(option = "inferno",name="integrated expression") +
  theme(axis.text=element_text(size=0,color='black'),
        axis.title=element_text(size=0,color='black')) + theme(plot.title = element_text(size = 0, vjust=0)) +
  theme(strip.text.x = element_text(size = 0))  +
  theme(legend.position=c(0.8, 0.2),
        legend.key.size = unit(1, 'cm'),
        legend.text = element_text(size=12.5))#remove facet label

#aggregate gene values to get a single expression value
agg_mat = aggregate_gene_expression(
  big_cds,
  gene_group_df = marker_df,
  cell_group_df = NULL,
  norm_method = c("log"),
  pseudocount = 1,
  scale_agg_values = F,
  max_agg_value = NA,
  min_agg_value = NA,
  exclude.na = TRUE
)

#combine aggregated values with CNV calls
agg_gene_clindat = data.frame(c(agg_mat),colData(big_cds))
names(agg_gene_clindat)[1] = c("eigengene_expression")

# compute lower and upper whiskers
ylim1 = boxplot.stats(agg_gene_clindat$eigengene_expression)$stats[c(1, 5)]

head(agg_gene_clindat)

#Now graph comparisons
my_comparisons <- list( c("19", "33"))

agg_gene_clindat.1 = agg_gene_clindat[agg_gene_clindat$cluster %in% c(19,33),]

ggplot(agg_gene_clindat.1, aes(x=cluster,y=eigengene_expression, fill=cluster)) +
  stat_compare_means(comparisons = my_comparisons) +
  geom_violin(aes(fill=cluster), width=1, alpha=0.4) +
  geom_boxplot(outlier.size=0, width=0.3, color="white", size=1.4, alpha=0.4) +
  #geom_bar(stat = "identity", fill="darkorange") +
  #facet_wrap(~parenchyma, scales="free", nrow=1) + 
  geom_jitter(width = 0.1, size=0.3, alpha=0.4) +
  ggtitle("") + 
  xlab("papilloma parenchyma") + ylab("integrated expression") + 
  theme(axis.text.x=element_text(size=25,color='black'),
        axis.text.y=element_text(size=25,color='black'),
        axis.title=element_text(size=25),
        axis.ticks.x = element_line(size=0),
        strip.text.x = element_text(size = 0),
        legend.position="none",
        title = element_text(size = 0, face="bold")) +
  theme(
    # Hide panel borders and remove grid lines
    panel.border = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.background = element_blank(),
    element_line(colour = 'black', size = 0.3),
    # Change axis line
    axis.line = element_line(colour = "black"),
    strip.background = element_rect(colour="white",
                                    fill="white")
  ) +
  scale_fill_manual(values = c("19" = "darkgreen",
                               "33" = "darkblue"))

##############################################################################################################################
##################################################################################################################################
##################################################################################################################################
##################################################################################################################################
#=====================================================================================#
#Extended Data Fig.2a : Test the Lgr5 tumor signal after serially removing clusters 13 and 14
#=====================================================================================#

table(colData(big_cds)$partition)

cluster13 = big_cds[,colData(big_cds)$partition %in% c(13)] #remove cell partition 14
table(colData(cluster13)$tissue_updated)

#---------------------------------------#
#1. Define the Lfr5 SCC eigengene 
match_index = match("Lgr5", aim2probedat$SYMBOL) #match seed gene to its place in the matrix
symbol_match = data.frame(aim2probedat$SYMBOL, tumor_rhodf[,match_index], tumor_p_df[,match_index]) #put together gene names, expression data, and p-vals
symbol_match = symbol_match[order(-symbol_match$tumor_rhodf...match_index.),] #order from high to low gene values
names(symbol_match) = c("gene","rho","p")

#make marker df for grahing eigengene
marker_df = data.frame(unique(alias2SymbolTable(symbol_match$gene[1:(eigengene_size+1)], species = "Mm")),paste(seed_gene[i],"Carcinoma",sep=""),paste(seed_gene[i],"Carcinoma",sep=""))
names(marker_df) = c("gene","cell_type","module")

#---------------------------------------#
#2. Drop cell cluster 14
big_cds_nopartition14 = big_cds[,colData(big_cds)$partition %!in% c(14)] #remove cell partition 14
table(colData(big_cds_nopartition14)$partition)

agexp_plot_tumor.no14 = plot_cells(big_cds_nopartition14,
                                   genes=marker_df,
                                   label_cell_groups=FALSE,
                                   label_leaves=F,
                                   label_branch_points=F,
                                   graph_label_size=1.5,
                                   show_trajectory_graph=F) +
  ggtitle("Cluster 14 removed") +
  xlab("") + ylab("") + 
  theme(axis.ticks = element_blank()) +
  theme(legend.position="bottom") +
  scale_color_viridis_c(option = "inferno",name="integrated expression") +
  theme(axis.text=element_text(size=0,color='black'),
        axis.title=element_text(size=0,color='black')) + theme(plot.title = element_text(size = 15, vjust=0)) +
  theme(strip.text.x = element_text(size = 0)) #remove facet label

#---------------------------------------#
#3. Drop cell cluster 13
big_cds_nopartition13 = big_cds_nopartition14[,colData(big_cds_nopartition14)$partition %!in% c(13)] #remove cell partition 13
table(colData(big_cds_nopartition13)$partition)

agexp_plot_tumor.no13 = plot_cells(big_cds_nopartition13,
                                   genes=marker_df,
                                   label_cell_groups=FALSE,
                                   label_leaves=F,
                                   label_branch_points=F,
                                   graph_label_size=1.5,
                                   show_trajectory_graph=F) +
  ggtitle("Cluster 13 removed") +
  xlab("") + ylab("") + 
  theme(axis.ticks = element_blank()) +
  theme(legend.position="bottom") +
  scale_color_viridis_c(option = "inferno",name="integrated expression") +
  theme(axis.text=element_text(size=0,color='black'),
        axis.title=element_text(size=0,color='black')) + theme(plot.title = element_text(size = 15, vjust=0)) +
  theme(strip.text.x = element_text(size = 0)) #remove facet label


#---------------------------------------#
#4. Lgr5 all cells
agexp_plot_tumor.allcells = plot_cells(big_cds,
                                       genes=marker_df,
                                       label_cell_groups=FALSE,
                                       label_leaves=F,
                                       label_branch_points=F,
                                       graph_label_size=1.5,
                                       show_trajectory_graph=F) +
  ggtitle("All cells") +
  xlab("") + ylab("") + 
  theme(axis.ticks = element_blank()) +
  theme(legend.position="bottom") +
  scale_color_viridis_c(option = "inferno",name="integrated expression") +
  theme(axis.text=element_text(size=0,color='black'),
        axis.title=element_text(size=0,color='black')) + theme(plot.title = element_text(size = 15, vjust=0)) +
  theme(strip.text.x = element_text(size = 0)) #remove facet label


#------------------------------------------------------------------#
#5. Plot all Lgr5 eigengenes together
png("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Extended\ Data/Extended\ Data\ Fig\ 2/Lgr5_manifold.png",
    width=12,height=5,units="in",res=200)
print(multiplot(agexp_plot_tumor.allcells,
                agexp_plot_tumor.no14, 
                agexp_plot_tumor.no13, cols=3))
dev.off()

#------------------------------------------------------------------#
#------------------------------------------------------------------#
#------------------------------------------------------------------#
#------------------------------------------------------------------#
#Write a function to graph the trend for any eigengene going from normal skin -> papilloma -> SCC

make_eigengene_violinplot = function(seed_gene, eigengene_type, cds_object) {
  
  # seed_gene = "Lgr6"
  # eigengene_type = "tumor"
  # cds_object = big_cds
  
  #---------------------------------------#
  #1. Define the Lfr5 SCC eigengene 
  match_index = match(seed_gene, aim2probedat$SYMBOL) #match seed gene to its place in the matrix
  
  if(eigengene_type == "tumor") { symbol_match = data.frame(aim2probedat$SYMBOL, tumor_rhodf[,match_index], tumor_p_df[,match_index])} else
  {symbol_match = data.frame(aim2probedat$SYMBOL, rhodf[,match_index], p_df[,match_index])}  #put together gene names, expression data, and p-vals
  names(symbol_match) = c("gene","rho","p")
  symbol_match = symbol_match[order(-symbol_match$rho),] #order from high to low gene values
  
  #make marker df for grahing eigengene
  marker_df = data.frame(unique(alias2SymbolTable(symbol_match$gene[1:(eigengene_size)], species = "Mm")),paste(seed_gene[i],"Carcinoma",sep=""),paste(seed_gene[i],"Carcinoma",sep=""))
  names(marker_df) = c("gene","cell_type","module")
  
  #aggregate gene values to get a single expression value
  agg_mat = aggregate_gene_expression(
    cds_object,
    gene_group_df = marker_df,
    cell_group_df = NULL,
    norm_method = c("log"),
    pseudocount = 1,
    scale_agg_values = F,
    max_agg_value = NA,
    min_agg_value = NA,
    exclude.na = TRUE
  )
  
  #replace correctly
  colData(cds_object)$tissue_updated = gsub("Normal skin","NSk",colData(cds_object)$tissue_updated)
  colData(cds_object)$tissue_updated = gsub("CA","SCC",colData(cds_object)$tissue_updated)
  
  #combine aggregated values with CNV calls
  agg_gene_clindat = data.frame(c(agg_mat),colData(cds_object))
  names(agg_gene_clindat)[1] = c("eigengene_expression")
  
  # compute lower and upper whiskers
  ylim1 = boxplot.stats(agg_gene_clindat$eigengene_expression)$stats[c(1, 5)]
  
  #Now graph comparisons
  my_comparisons <- list( c("NSk", "Pap"), c("Pap", "SCC"), c("NSk", "SCC"))
  
  eigengene_violin_plot = ggplot(agg_gene_clindat, aes(x=tissue_updated,y=eigengene_expression, fill=tissue_updated)) +
    #stat_compare_means(comparisons = my_comparisons) +
    geom_violin(aes(fill=tissue_updated), width=1) +
    geom_boxplot(outlier.size=0, width=0.2, color="black") +
    #geom_bar(stat = "identity", fill="darkorange") +
    #facet_wrap(~parenchyma, scales="free", nrow=1) + 
    #geom_jitter(width = 0.1) +
    ggtitle("") + 
    xlab("papilloma parenchyma") + ylab("eigengene expression") + 
    theme(axis.text.x=element_text(size=25,color='black'),
          axis.text.y=element_text(size=25,color='black'),
          axis.title=element_text(size=0),
          axis.ticks.x = element_line(size=0),
          strip.text.x = element_text(size = 0),
          legend.position="none",
          title = element_text(size = 0, face="bold")) +
    theme(
      # Hide panel borders and remove grid lines
      panel.border = element_blank(),
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      panel.background = element_blank(),
      element_line(colour = 'black', size = 0.3),
      # Change axis line
      axis.line = element_line(colour = "black"),
      strip.background = element_rect(colour="white",
                                      fill="white")
    ) +
    scale_fill_manual(values = c("NSk" = "lightblue",
                                 "Pap" = "thistle3", 
                                 "SCC" = "red")) + theme(legend.position="none") +
    coord_cartesian(ylim = ylim1*1.5) #Don't plot outliers
  
  return(eigengene_violin_plot)
  
}


#=====================================================================================#
#Extended Data Fig.2b : Test the Lgr5 tumor signal after serially removing clusters 13 and 14
#=====================================================================================#
png("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Extended\ Data/Extended\ Data\ Fig\ 2/Lgr5_eigengene_violinplot_allcells.png",
    width=5,height=5,units="in",res=200)
print( make_eigengene_violinplot("Lgr5","tumor",big_cds) )
dev.off()

png("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Extended\ Data/Extended\ Data\ Fig\ 2/Lgr5_eigengene_violinplot_cluster14removed.png",
    width=5,height=5,units="in",res=200)
print( make_eigengene_violinplot("Lgr5","tumor",big_cds_nopartition14) )
dev.off()

png("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Extended\ Data/Extended\ Data\ Fig\ 2/Lgr5_eigengene_violinplot_cluster13removed.png",
    width=5,height=5,units="in",res=200)
print( make_eigengene_violinplot("Lgr5","tumor",big_cds_nopartition13) )
dev.off()

#=====================================================================================#
#Main text Fig.1f : Violin plot Lgr5 and Lgr6 eigengenes
#=====================================================================================#
png("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Figs/Fig1/Fig1f_Lgr5.png",
    width=5,height=5,units="in",res=200)
print( make_eigengene_violinplot("Lgr5","tumor",big_cds) )
dev.off()

png("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Figs/Fig1/Fig1f_Lgr6.png",
    width=5,height=5,units="in",res=200)
print( make_eigengene_violinplot("Lgr6","tumor",big_cds) )
dev.off()


##############################################################################################################################
##################################################################################################################################
##################################################################################################################################
##################################################################################################################################
##################################################################################################################################

#number of cells per tissue
table(colData(big_cds)$tissue_updated)
table(colData(big_cds)$tissue_updated) / dim(big_cds)[2]

#number of cells per tissue*FACS sorting
table(colData(big_cds)$tissue_updated_sorting)
ca_total = table(colData(big_cds)$tissue_updated)[1]
normal_skin_total = table(colData(big_cds)$tissue_updated)[2]
pap_total = table(colData(big_cds)$tissue_updated)[3]
tissue_tots = c(ca_total,ca_total,ca_total,
                normal_skin_total,normal_skin_total,normal_skin_total,
                pap_total,pap_total,pap_total)
table(colData(big_cds)$tissue_updated_sorting) / tissue_tots

#number of cells within partitions
table(colData(big_cds)$partition) / dim(big_cds)[2]
part_counts = data.frame(table(colData(big_cds)$partition))
part_counts$Var1 = as.numeric(as.character(part_counts$Var1))
non_four = subset(part_counts, Var1 > 4)
sum(non_four$Freq) / dim(big_cds)[2]

#number of cells within pap_ca
table(colData(cds_comb)$parenchyma_tissue) / dim(big_cds)[2]

#number of cells within pap_ca
table(colData(t_cds)$tissue) / dim(big_cds)[2]
t_cds

##############################################################################################################################
##################################################################################################################################
##############################################################################################################################
##################################################################################################################################
#Create eigengenes from top and bottom DEGs for the 3 transitions: 1) NSk->Pap 2) Pap -> Control Carcinomas 3) Control Carcinomas -> Cisplatin-treated carcinomas

file_transitions = c("Normal skin_Pap","Pap_Control","Control_Cisplatin")

deg_eigengenes = do.call(cbind, lapply(1:length(file_transitions), function(i){
  #i=1
  results_file = read.delim(file=paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/DEG/integrated_original_chemo/",
                                       file_transitions[i],"_deg.txt",sep="") ,
                            header=T)
  results_file = subset(results_file, variable_tested=="tissue")
  
  #get top and bottom and FC results
  numeric_fc = results_file[!is.infinite(results_file$fc),]
  numeric_fc = numeric_fc[order(-numeric_fc$fc),] #order numeric file from high to low FC
  infinite_fc = results_file[is.infinite(results_file$fc),]
  
  #get gene names for top and bottom FC results
  top_numeric_genees = head(numeric_fc, 100)$gene #top 100 genes
  bottom_numeric_genees = tail(numeric_fc, 100)$gene #bottom 100 genes
  top_inf_genes = head(infinite_fc, 100)$gene #top 100 genes
  
  #combine into 3-column dataframe and return
  largest_degs = data.frame(top_numeric_genees,bottom_numeric_genees,top_inf_genes)
  names(largest_degs) = c(
    paste(file_transitions[i],"_top_numeric_deg",sep=""),
    paste(file_transitions[i],"_bottom_numeric_deg",sep=""),
    paste(file_transitions[i],"_top_infinite_deg",sep="")
  )
  
  return(largest_degs)
  
}))

write.table(deg_eigengenes,
            file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/apriori_mark_eigengenes/integrated-original-chemo-DEG-eigengenes.txt",
            quote=F, row.names=F, sep="\t")

##################################################################################################################################
##################################################################################################################################

lung_human_genes = read.xlsx(
  xlsxFile="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/apriori_eigengenes/KRT19_eigengenes_Human_Lung_NORMAL_Tumor.xlsx",
  colNames=T
)

mouse_ortholog1 = data.frame(convertHumanGeneList(lung_human_genes$Krt19_human.lung_normal))
mouse_ortholog2 = data.frame(convertHumanGeneList(na.omit(lung_human_genes$Krt19_human.lung_tumor)))

mouse_ortholog2 = do.call(rbind, lapply(1:length(lung_human_genes$Krt19_human.lung_tumor), function(i){
  tryCatch({ #suppress error
    #i=1
    print(i)
    mouse_ortholog = convertHumanGeneList(lung_human_genes$Krt19_human.lung_tumor[i])
    mouse_ortholog
  },  error=function(e){})
}))

##################################################################################################################################
##################################################################################################################################
#Make EMT eigeengenes from marker list: EMT gene list from the Nanostring nCounter® PanCancer Progression Panel

#Read in a sheet convert to mouse orthologs and find functional definitions
sheets = c(4,5)
library(erer)

do.call(rbind, lapply(1:length(sheets), function(i){
  #i=5
  
  #Read in excel file+sheeet
  sheet = read.xlsx(
    xlsxFile = "/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/apriori_eigengenes/PanCancer_Progression_Panel_Gene_List.xlsx",
    sheet=i
  )
  
  #Now go through eeach column of interest and get list of genes that fall into that category
  category_list = do.call(list, lapply(3:(ncol(sheet)), function(j){
    #j=3
    names(sheet)
    ortholog_marker = list(convertHumanGeneList(na.omit(data.frame(sheet[,c(1,j)]))$Genes)) #This requires the internet fuccc
    names(ortholog_marker) = names(sheet)[j]
    ortholog_marker
  }))
  
  #write out list which is annoying
  write.list(category_list,
             file=paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/apriori_eigengenes/pangenome_sheet_",sheets[i],".txt",sep=""),
             row.names = F,quote=F)
  
}))

##################################################################################################################################
##################################################################################################################################
#Eigengenes from Haensel et al

#Cell marker types

#----------------------------------#
#Unwounded (UW) skin
#----------------------------------#

sheet = read.xlsx(
  xlsxFile = "/Users/marktaylor/Desktop/Papers/mouse_skin_markers/Defining\ Epidermal\ Basal\ Cell\ States\ during\ Skin\ Homeostasis\ and\ Wound\ Healing\ Using\ Single-Cell\ Transcriptomics/SI/1-s2.0-S2211124720302655-mmc2.xlsx",
  sheet=1,
  startRow = 2
)
head(sheet)
clusters = unique(sheet$Cluster)

library("dplyr")

uw_markers = do.call(bind_cols, lapply(1:length(clusters), function(i){
  #i=1
  print(i)
  top_markers = data.frame(subset(sheet, Cluster==clusters[i])$Gene[1:100])
  names(top_markers) = paste("UW ",clusters[i],sep="")
  top_markers
}))

#----------------------------------#
#Wounded (WO) skin
#----------------------------------#

sheet = read.xlsx(
  xlsxFile = "/Users/marktaylor/Desktop/Papers/mouse_skin_markers/Defining\ Epidermal\ Basal\ Cell\ States\ during\ Skin\ Homeostasis\ and\ Wound\ Healing\ Using\ Single-Cell\ Transcriptomics/SI/1-s2.0-S2211124720302655-mmc2.xlsx",
  sheet=2,
  startRow = 2
)
head(sheet)
clusters = unique(sheet$Cluster)

library("dplyr")

wo_markers = do.call(bind_cols, lapply(1:length(clusters), function(i){
  #i=1
  print(i)
  top_markers = data.frame(subset(sheet, Cluster==clusters[i])$Gene[1:100])
  names(top_markers) = paste("WO ",clusters[i],sep="")
  top_markers
}))


#Combine together and print out
write.table(cbind(uw_markers,wo_markers),
            file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/apriori_eigengenes/Haensel\ et\ al\ Basal/TableS1_totalsample_markers.txt",
            sep="\t",row.names=F,quote=F)


#----------------------------------#
#Basal Cell DEGs b/w UW and WO skin
#----------------------------------#
sheet = read.xlsx(
  xlsxFile = "/Users/marktaylor/Desktop/Papers/mouse_skin_markers/Defining\ Epidermal\ Basal\ Cell\ States\ during\ Skin\ Homeostasis\ and\ Wound\ Healing\ Using\ Single-Cell\ Transcriptomics/SI/1-s2.0-S2211124720302655-mmc5.xlsx",
  sheet=1,
  startRow = 2
)
head(sheet)
clusters = unique(sheet$cluster)

#Bind columns with uneven lengths by: 1) transoisee to rows 2) use dyplr's bind_rows function 3) transpose back
markers = do.call(bind_rows, lapply(1:length(clusters), function(i){
  #i=1
  print(i)
  top_markers = data.frame(subset(sheet, cluster==clusters[i])$gene)
  names(top_markers) = paste("Basal cell DEG ",clusters[i],sep="")
  as.data.frame(t(top_markers))
}))

markers = t(markers)
write.table(markers,
            file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/apriori_eigengenes/Haensel\ et\ al\ Basal/TableS4_basal_cell_DEGs_UW_vs_WO.txt",
            sep="\t",row.names=F,quote=F)

#----------------------------------#
#UW skin Basal cell cluster markers
#----------------------------------#
sheet = read.xlsx(
  xlsxFile = "/Users/marktaylor/Desktop/Papers/mouse_skin_markers/Defining\ Epidermal\ Basal\ Cell\ States\ during\ Skin\ Homeostasis\ and\ Wound\ Healing\ Using\ Single-Cell\ Transcriptomics/SI/1-s2.0-S2211124720302655-mmc6.xlsx",
  sheet=1,
  startRow = 2
)
head(sheet)
clusters = unique(sheet$cluster)

#Bind columns with uneven lengths by: 1) transoisee to rows 2) use dyplr's bind_rows function 3) transpose back
markers = do.call(bind_rows, lapply(1:length(clusters), function(i){
  #i=1
  print(i)
  top_markers = data.frame(subset(sheet, cluster==clusters[i])$gene)
  names(top_markers) = paste("Basal cell UW ",clusters[i],sep="")
  as.data.frame(t(top_markers))
}))

markers = t(markers)
write.table(markers,
            file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/apriori_eigengenes/Haensel\ et\ al\ Basal/TableS5_UW_basal_cell_cluster_markers.txt",
            sep="\t",row.names=F,quote=F)

#----------------------------------#
#WO skin Basal cell cluster markers
#----------------------------------#
sheet = read.xlsx(
  xlsxFile = "/Users/marktaylor/Desktop/Papers/mouse_skin_markers/Defining\ Epidermal\ Basal\ Cell\ States\ during\ Skin\ Homeostasis\ and\ Wound\ Healing\ Using\ Single-Cell\ Transcriptomics/SI/1-s2.0-S2211124720302655-mmc7.xlsx",
  sheet=1,
  startRow = 2
)
head(sheet)
clusters = unique(sheet$cluster)

#Bind columns with uneven lengths by: 1) transoisee to rows 2) use dyplr's bind_rows function 3) transpose back
markers = do.call(bind_rows, lapply(1:length(clusters), function(i){
  #i=1
  print(i)
  top_markers = data.frame(subset(sheet, cluster==clusters[i])$gene)
  names(top_markers) = paste("Basal cell WO ",clusters[i],sep="")
  as.data.frame(t(top_markers))
}))

markers = t(markers)
write.table(markers,
            file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/apriori_eigengenes/Haensel\ et\ al\ Basal/TableS6_WO_basal_cell_cluster_markers.txt",
            sep="\t",row.names=F,quote=F)

##################################################################################################################################
##################################################################################################################################
#TCGA Survival genes and genomic features data from: 
#Table S1C - Gene Expression
# Z-score > 1.96 -> ADVERSE features upregulation of tumor features associted with SHORTER SURVIVAL
# Z-score < 1.96 -> FAVORABLE feature = upregulation of tumor feature associated with LONGER SURVIVAL

#Read in a sheet convert to mouse orthologs and find functional definitions
survival_genes = read.xlsx(
  xlsxFile = "/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/apriori_eigengenes/Smith\ et\ al\ TCGA\ survival\ genome\ features/table_s1.xlsx",
  sheet=4
)
head(survival_genes)

#Add mouse orthologs bc biomaRt servers are offline what the actual fuck
#read in database of homologs
human_homologs = read.delim(file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/human_scRNAseq/human_mouse_homologs/HOM_MouseHumanSequence.txt", header=T)

#match mouse genes with Human genes
mouse_ortholog_survival_genes = do.call(rbind, lapply(1:nrow(survival_genes), function(z){
  #z=1
  print(z)
  tryCatch({ #suppress error
    homologmatch = human_homologs[match(survival_genes$Gene[z], human_homologs$Symbol),][,1] #get the HomoloGene.ID for a human gene matching mouse gene symbol
    mouse_ortholog = subset(human_homologs, HomoloGene.ID==homologmatch & Common.Organism.Name=="mouse, laboratory")$Symbol[1] #get the human natcg
    data.frame(mouse_ortholog, survival_genes[z,])
  },  error=function(e){})
}))

length(na.omit(mouse_ortholog_survival_genes$mouse_ortholog))
length()

head(mouse_ortholog_survival_genes)

#Now create top and bottom metagenes from survival data
#Start at column 3 bc that's where the survival data starts
survival_top_and_bottom = do.call(cbind, lapply(27:ncol(mouse_ortholog_survival_genes), function(i){
  #i=27
  
  print(i)
  re_ord = na.omit(mouse_ortholog_survival_genes[order(mouse_ortholog_survival_genes[,i]),c(1,i)]) #re-order from high-to-low Z-score from univariate Cox hazard model: favorable --> adverse
  
  #Get top adverse and favorable features
  favorable = re_ord[1:100,1] #lowest Z-score are FAVORABEL
  adverse = tail(re_ord,100)[,1] #highest Z-scores are ADVERSE
  
  #Favorable
  favorable_plot = plot_cells(big_cds,
                              genes=data.frame(unique(favorable),"eigengene","eigengene"), #seed gene is the first row of the markers file
                              label_cell_groups=FALSE,
                              label_leaves=F,
                              label_branch_points=F,
                              graph_label_size=1.5,
                              show_trajectory_graph=F) +
    xlab("") + ylab("") + 
    theme(legend.position="bottom") +
    scale_color_viridis_c(option = "inferno",name="log expression") +
    ggtitle(paste(names(mouse_ortholog_survival_genes)[i],"top 100 FAVORABLE genes",sep=" "))+ #continuous
    theme(plot.title = element_text(size = 22, vjust=1)) +
    theme(axis.text=element_text(size=0,color='black'),
          axis.title=element_text(size=0),
          axis.ticks = element_line(size=0)) +
    theme(strip.text.x = element_text(size = 0)) #remove facet label
  
  #Adverse
  adverse_plot = plot_cells(big_cds,
                            genes=data.frame(unique(adverse),"eigengene","eigengene"), #seed gene is the first row of the markers file
                            label_cell_groups=FALSE,
                            label_leaves=F,
                            label_branch_points=F,
                            graph_label_size=1.5,
                            show_trajectory_graph=F) +
    xlab("") + ylab("") + 
    theme(legend.position="bottom") +
    scale_color_viridis_c(option = "inferno",name="log expression") +
    ggtitle(paste(names(mouse_ortholog_survival_genes)[i],"top 100 ADVERSE genes",sep=" "))+ #continuous
    theme(plot.title = element_text(size = 22, vjust=1)) +
    theme(axis.text=element_text(size=0,color='black'),
          axis.title=element_text(size=0),
          axis.ticks = element_line(size=0)) +
    theme(strip.text.x = element_text(size = 0)) #remove facet label
  
  #-----------#
  #print out eigengene plots
  png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/TCGA/scRNAseq_TCGA/Smith_et_al_top_survival/UMAP/",(i-2),"_",names(mouse_ortholog_survival_genes)[i],"_top_survival_genes.png",sep=""),
      width=14,height=7,units="in",res=400)
  
  print(multiplot(
    favorable_plot,adverse_plot,
    cols=2))
  
  dev.off()
  
  #Combine results and return to loop
  res_df = data.frame(favorable, adverse)
  names(res_df) = c(paste(names(mouse_ortholog_survival_genes)[i],"favorable"),
                    paste(names(mouse_ortholog_survival_genes)[i],"adverse"))
  return(res_df)
  
}))

#Get the actual genes in each one
survival_top_and_bottom = do.call(data.frame, lapply(3:ncol(mouse_ortholog_survival_genes), function(i){
  #i=3
  
  print(i)
  re_ord = na.omit(mouse_ortholog_survival_genes[order(mouse_ortholog_survival_genes[,i]),c(1,i)]) #re-order from high-to-low Z-score from univariate Cox hazard model: favorable --> adverse
  
  #Get top adverse and favorable features
  favorable = re_ord[1:100,1] #lowest Z-score are FAVORABEL
  adverse = tail(re_ord,100)[,1] #highest Z-scores are ADVERSE
  
  #Combine results and return to loop
  res_df = data.frame(favorable, adverse)
  names(res_df) = c(paste(names(mouse_ortholog_survival_genes)[i],"favorable"),
                    paste(names(mouse_ortholog_survival_genes)[i],"adverse"))
  return(res_df)
  
}))
write.table(survival_top_and_bottom,
            file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/TCGA/scRNAseq_TCGA/Smith_et_al_top_survival/UMAP/0_top100_survival_genes.txt",
            row.names = F, quote=F, sep="\t"
)



##################################################################################################################################
##################################################################################################################################
##################################################################################################################################

#Eigengene generated by ALLAN figs
balmain_genes = read.xlsx(
  xlsxFile="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/apriori_eigengenes/Mark_Eigengenes.xlsx",
  colNames=T
)


##################################################################################################################################
##################################################################################################################################
#make directory of version of graphs

eigengene_verison = 36 #increment this to show different versions
dir.create(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/balmain_eigene/eigengene_PNGs_v",eigengene_verison,sep=""))

##################################################################################################################################
##################################################################################################################################
##################################################################################################################################
##################################################################################################################################

eigengene_agg = do.call(cbind, lapply(506:ncol(balmain_genes), function(i){
  tryCatch({ #suppress error
    
    #i=506
    print(i)
    #only if more than 1 constituent gene
    marker_df = data.frame(na.omit(balmain_genes[,i]),names(balmain_genes)[i])
    names(marker_df) = c("gene","cell_type")
    
    #Remove IFN and OLF genes 
    marker_df = marker_df[-grep("Ifn",marker_df$gene),]
    marker_df = marker_df[-grep("Olf",marker_df$gene),]
    
    marker_df_top = marker_df[1:100,] #top
    marker_df_bttm = tail(marker_df, 100) #bottom 
    
    #remake the marker df dataframe with the matched gene names
    marker_df_top = data.frame(unique(alias2SymbolTable(unique(marker_df_top$gene), species = "Mm")),names(balmain_genes)[i],names(balmain_genes)[i])
    names(marker_df_top) = c("gene","cell_type","module")
    marker_df_top = na.omit(marker_df_top) #remove non-matching genes
    
    #all cells eigengene
    agexp_plot_top = plot_cells(big_cds,
                            genes=marker_df_top, #seed gene is the first row of the markers file
                            label_cell_groups=FALSE,
                            label_leaves=F,
                            label_branch_points=F,
                            graph_label_size=1.5,
                            show_trajectory_graph=F) +
      xlab("") + ylab("") + 
      theme(legend.position="bottom") +
      scale_color_viridis_c(option = "inferno",name="log expression") +
      ggtitle(paste("All cells:", names(balmain_genes)[i],"top",sep=" "))+ #continuous
      theme(plot.title = element_text(size = 22, vjust=1)) +
      theme(axis.text=element_text(size=0,color='black'),
            axis.title=element_text(size=0),
            axis.ticks = element_line(size=0)) +
      theme(strip.text.x = element_text(size = 0)) #remove facet label
    
    #bottom
    #remake the marker df dataframe with the matched gene names
    marker_df_bttm = data.frame(unique(alias2SymbolTable(unique(marker_df_bttm$gene), species = "Mm")),names(balmain_genes)[i],names(balmain_genes)[i])
    names(marker_df_bttm) = c("gene","cell_type","module")
    marker_df_bttm = na.omit(marker_df_bttm) #remove non-matching genes
    
    #all cells eigengene
    agexp_plot_bttm = plot_cells(big_cds,
                                genes=marker_df_bttm, #seed gene is the first row of the markers file
                                label_cell_groups=FALSE,
                                label_leaves=F,
                                label_branch_points=F,
                                graph_label_size=1.5,
                                show_trajectory_graph=F) +
      xlab("") + ylab("") + 
      theme(legend.position="bottom") +
      scale_color_viridis_c(option = "inferno",name="log expression") +
      ggtitle(paste("All cells:", names(balmain_genes)[i],"bottom",sep=" "))+ #continuous
      theme(plot.title = element_text(size = 22, vjust=1)) +
      theme(axis.text=element_text(size=0,color='black'),
            axis.title=element_text(size=0),
            axis.ticks = element_line(size=0)) +
      theme(strip.text.x = element_text(size = 0)) #remove facet label
    
    
    #-----------#
    #print out eigengene plots
    png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/balmain_eigene/eigengene_PNGs_v",eigengene_verison,"/",names(balmain_genes)[i],".png",sep=""),
        width=14,height=7,units="in",res=400)
    
    print(multiplot(
      agexp_plot_top, agexp_plot_bttm,
      cols=2))
    
    dev.off()
    
  },  error=function(e){})
}))


##################################################################################################################################
#Mark eigengenes papilloma only

eigengene_agg = do.call(cbind, lapply(506:ncol(balmain_genes), function(i){
  tryCatch({ #suppress error
    
    #i=506
    print(i)
    #only if more than 1 constituent gene
    marker_df = data.frame(na.omit(balmain_genes[,i]),names(balmain_genes)[i])
    names(marker_df) = c("gene","cell_type")
    
    #Remove IFN and OLF genes 
    marker_df = marker_df[-grep("Ifn",marker_df$gene),]
    marker_df = marker_df[-grep("Olf",marker_df$gene),]
    
    marker_df_top = marker_df[1:100,] #top
    marker_df_bttm = tail(marker_df, 100) #bottom 
    
    #remake the marker df dataframe with the matched gene names
    marker_df_top = data.frame(unique(alias2SymbolTable(unique(marker_df_top$gene), species = "Mm")),names(balmain_genes)[i],names(balmain_genes)[i])
    names(marker_df_top) = c("gene","cell_type","module")
    marker_df_top = na.omit(marker_df_top) #remove non-matching genes
    
    #all cells eigengene
    agexp_plot_top = plot_cells(big_cds,
                                genes=marker_df_top, #seed gene is the first row of the markers file
                                label_cell_groups=FALSE,
                                label_leaves=F,
                                label_branch_points=F,
                                graph_label_size=1.5,
                                show_trajectory_graph=F,
                                cell_size = 1.2) +
      xlab("") + ylab("") + 
      theme(axis.ticks = element_blank()) +
      scale_color_viridis_c(option = "inferno",name="") +
      theme(axis.text=element_text(size=0,color='black'),
            axis.title=element_text(size=0,color='black')) + theme(plot.title = element_text(size = 0, vjust=0)) +
      theme(strip.text.x = element_text(size = 0))  +
      annotate("text", label = paste(names(balmain_genes)[i],"top",sep=""), 
               x = -5, y = 7, vjust = 0,size=10) +
      theme(legend.position=c(5, 5),
            legend.key.size = unit(1, 'cm'),
            legend.text = element_text(size=20)) +
      xlim(-4.1,0.9) +
      ylim(-7.5, -3.8)
    
    #bottom
    #remake the marker df dataframe with the matched gene names
    marker_df_bttm = data.frame(unique(alias2SymbolTable(unique(marker_df_bttm$gene), species = "Mm")),names(balmain_genes)[i],names(balmain_genes)[i])
    names(marker_df_bttm) = c("gene","cell_type","module")
    marker_df_bttm = na.omit(marker_df_bttm) #remove non-matching genes
    
    #all cells eigengene
    agexp_plot_bttm = plot_cells(big_cds,
                                 genes=marker_df_bttm, #seed gene is the first row of the markers file
                                 label_cell_groups=FALSE,
                                 label_leaves=F,
                                 label_branch_points=F,
                                 graph_label_size=1.5,
                                 show_trajectory_graph=F,
                                 cell_size = 1.2) +
      xlab("") + ylab("") + 
      theme(axis.ticks = element_blank()) +
      scale_color_viridis_c(option = "inferno",name="") +
      theme(axis.text=element_text(size=0,color='black'),
            axis.title=element_text(size=0,color='black')) + theme(plot.title = element_text(size = 0, vjust=0)) +
      theme(strip.text.x = element_text(size = 0))  +
      annotate("text", label = paste(names(balmain_genes)[i]," bottom",sep=""), 
               x = -5, y = 7, vjust = 0,size=10) +
      theme(legend.position=c(5, 5),
            legend.key.size = unit(1, 'cm'),
            legend.text = element_text(size=20)) +
      xlim(-4.1,0.9) +
      ylim(-7.5, -3.8)
    
    
    #-----------#
    #print out eigengene plots
    png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/balmain_eigene/eigengene_PNGs_v",eigengene_verison,"/",names(balmain_genes)[i],"_pap_only.png",sep=""),
        width=10,height=5,units="in",res=400)
    
    print(multiplot(
      agexp_plot_top, agexp_plot_bttm,
      cols=2))
    
    dev.off()
    
  },  error=function(e){})
}))

##################################################################################################################################
#Asymmetric cell division genes: Table 1 from Lyttle et al 2018 https://www.nature.com/articles/s41568-018-0056-x
##################################################################################################################################

c("Llgl1","Numb","Msi1","Msi2","Pafah1b1","Trim3","Trp53")

png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/figures/individual_gene_UMAPs/asymmetric_cell_division_genes.png",sep=""),
    width=20,height=20,units="in",res=350)

print(
  plot_cells(big_cds,
             genes=c("Llgl1","Numb","Msi1","Msi2","Pafah1b1","Trim3","Trp53"), #seed gene is the first row of the markers file
             label_cell_groups=FALSE,
             label_leaves=F,
             label_branch_points=F,
             graph_label_size=1.5,
             show_trajectory_graph=F) +
    xlab("") + ylab("") + 
    theme(legend.position="bottom") +
    scale_color_viridis_c(option = "inferno",name="log expression") +
    theme(plot.title = element_text(size = 0, vjust=1)) +
    theme(axis.text=element_text(size=0,color='black'),
          axis.title=element_text(size=0),
          axis.ticks = element_line(size=0)) +
    theme(strip.text.x = element_text(size = 15))
)
dev.off()

##################################################################################################################################
##################################################################################################################################
#GO EIGENGENES BIOLOGICAL PROCESS: BP

# Download GO annotation codes
#http://www.informatics.jax.org/downloads/reports/go_terms.mgi
#Download all GO terms and genes
#http://www.informatics.jax.org/downloads/reports/index.html#ftp

#Read in codes
go_codes = read.xlsx(
  xlsxFile="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/GO/GO\ codes.xlsx",
  colNames=T
)
go_codes = subset(go_codes, GO.process=="Biological Process") #usbset to GO code

#Read in GO data
go_data = read.delim(file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/GO/GO\ database.txt",
                     header=T)

nrow(go_codes)

match("GO:0042110",go_codes$GO.code)

for(i in 3:nrow(go_codes)){
  
  tryCatch({ #suppress error
    
    #i=5340
    print(i)
    #remake the marker df dataframe with the matched gene names
    marker_df = data.frame(unique(subset(go_data, GO.Term.ID==go_codes$GO.code[i])$Mouse.Marker.Symbol),"eigengene","eigengene")
    names(marker_df) = c("gene","cell_type","module")
    marker_df = na.omit(marker_df) #remove non-matching genes
    
    #all cells eigengene
    agexp_plot = plot_cells(big_cds,
                            genes=marker_df, #seed gene is the first row of the markers file
                            label_cell_groups=FALSE,
                            label_leaves=F,
                            label_branch_points=F,
                            graph_label_size=1.5,
                            show_trajectory_graph=F) +
      xlab("") + ylab("") + 
      scale_color_viridis_c(option = "inferno",name="log expression") +
      ggtitle(paste(i," " ,go_codes$GO.code[i], go_codes$GO.key[i],"\nN=",nrow(marker_df),sep="") )+ #continuous
      theme(plot.title = element_text(size = 13, vjust=0)) +
      theme(axis.text=element_text(size=0,color='black'),
            axis.title=element_text(size=0),
            axis.ticks = element_line(size=0)) +
      theme(strip.text.x = element_text(size = 0)) + #remove facet label
      theme(legend.position=c(0.8, 0.2),
            legend.key.size = unit(1, 'cm'),
            legend.text = element_text(size=13))
    #-----------#
    #print out eigengene plots
    png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/GO/MGI_BP_eigengenes/umaps/",i,"_",go_codes$GO.key[i],".png",sep=""),
        width=7,height=7,units="in",res=350)
    print(agexp_plot)
    dev.off()
    
  },  error=function(e){})
  
}

#--------------------------------------------------------------------------------------#
#Calculate the average MGI GO eigengene expression for each unsupervised cell culture
#--------------------------------------------------------------------------------------#

all_mgi_eigengenes = do.call(rbind, lapply(1:nrow(go_codes), function(i){
  
  tryCatch({ #suppress error
    
    #i=2
    print(i)
    #remake the marker df dataframe with the matched gene names
    marker_df = data.frame(unique(subset(go_data, GO.Term.ID==go_codes$GO.code[i])$Mouse.Marker.Symbol),"eigengene","eigengene")
    names(marker_df) = c("gene","cell_type","module")
    marker_df = na.omit(marker_df) #remove non-matching genes
    
    if(nrow(marker_df)>1){
      agg_mat = aggregate_gene_expression(
        big_cds,
        gene_group_df = marker_df,
        cell_group_df = NULL,
        norm_method = c("log"),
        pseudocount = 1,
        scale_agg_values = F,
        max_agg_value = NA,
        min_agg_value = NA,
        exclude.na = TRUE
      )
      
      agg_gene_clindat = data.frame(t(agg_mat),colData(big_cds)$cluster )
      names(agg_gene_clindat) = c("eigengene_expression","cluster")
      
      dfc = summarySE(agg_gene_clindat, measurevar='eigengene_expression', groupvars=c('cluster'))
      dfc.reord = dfc[order(-dfc$eigengene_expression),] #re-order from high to low eigengene expression
      
      topcluster1 = dfc.reord$cluster[1]
      topcluster2 = dfc.reord$cluster[2]
      topcluster3 = dfc.reord$cluster[3]
      
      mgi_eigengene_expression = data.frame(i, go_codes$GO.code[i], go_codes$GO.key[i],
                                            nrow(marker_df),
                                            topcluster1,topcluster2,topcluster3,
                                            t(dfc$eigengene_expression),
                                            paste(marker_df$gene, collapse=", ")
      )
      
      names(mgi_eigengene_expression) = c("i","MGI_GO_code","MGI_GO_key",
                                          "n_gene",
                                          "topcluster1","topcluster2","topcluster3",
                                          paste("cluster",seq(1,49),sep=""),
                                          "genes"
      )
      
      write.table(
        mgi_eigengene_expression,
        file=paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/GO/MGI_BP_eigengenes/unsupervised_cluster_textfiles/",i,"_",go_codes$GO.key[i],".txt",sep=""),
        sep="\t", quote=F, row.names=F
      )
      
      return(mgi_eigengene_expression)
      
    } #end if statement asking for marker files with at least 2 genes (2 rows)
  },  error=function(e){})
}))    

write.table(
  all_mgi_eigengenes,
  file=paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/GO/MGI_BP_eigengenes/unsupervised_cluster_mgo_eigengenes.txt",sep=""),
  sep="\t", quote=F, row.names=F
)

#--------------------------------------------------------------------------------------#
#Make Excel sheet for top GO eigengene expression for each unsupervised cell cluster
#--------------------------------------------------------------------------------------#

all_mgi_eigengenes = read.delim(file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/GO/MGI_BP_eigengenes/unsupervised_cluster_mgo_eigengenes.txt",
                                header=T)

topclusts = seq(1:49) 
OUT <- createWorkbook() #create excel workbook for a sample https://stackoverflow.com/questions/27713310/easy-way-to-export-multiple-data-frame-to-multiple-excel-worksheets

for(i in 1:length(topclusts)) {
  tryCatch({ #suppress error
    #i=33
    print(i)
    subclust = all_mgi_eigengenes[all_mgi_eigengenes$topcluster1 %in% topclusts[i],] #subset to GOs where a specific cluster has highest expression
    
    #Calculate specificity metric: percent difference in avg expression of cluster vs all others = (specific_cluster - all others) / all others
    
    #Get average across all cell clusters (cols) EXCEPT for cluster of interest
    col_of_interest = subclust[,match(paste("cluster",i,sep=""),names(subclust))]
    dropped_col = subclust[,-match(paste("cluster",i,sep=""),names(subclust))] #Drop column of interest
    other_mean = (col_of_interest-rowMeans(dropped_col[,8:46])) / rowMeans(dropped_col[,8:46]) #only compare to clusters 1-39 with at least 150 cells
    # other_mean = (col_of_interest-rowMeans(dropped_col[,8:46])) #only compare to clusters 1-39 with at least 150 cells
    
    subclust$specificity = other_mean #Add specificity metric
    
    re_ord = subclust[order(-subclust[,match(paste("cluster",i,sep=""),names(subclust))]),] #order from high to low gene expression in thee clustere
    #re_ord = subclust[order(-subclust$specificity),] #order from high to low gene expression in thee clustere
    
    # Write out to excel workbook
    sheet_name = topclusts[i]
    addWorksheet(OUT, sheet_name) # Add a sheet to the workbook OUT
    writeData(OUT, sheet = sheet_name, x = re_ord) #Write data to a sheet
  },  error=function(e){})
  
}

# Export the Excel file
saveWorkbook(OUT, 
             "/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/GO/MGI_BP_eigengenes/unsupervised_cluster_mgo_eigengenes_PROCESSSED.xlsx",
             overwrite=T)

table(all_mgi_eigengenes$topcluster1)

#--------------------------------------------------------------------------#
#High definition plot of the 2 spikes bebe
print(paste("j=",j,sep=""))

#Add dummy column
colData(big_cds)$isolated_cell_population = "Non-match"

#Match Upper spike
j=19
matching_cell_ids = row.names(colData(big_cds)[ colData(big_cds)[,match_index] %in% factor_levels[j] , ]) #get cell id's that match a factor level
sub_cell_matches = match(matching_cell_ids,colnames(big_cds)) #match extracted cells with all cell names
colData(big_cds)$isolated_cell_population[sub_cell_matches] <- "Upper spike" #now replace matching cells with new name

#Match Lower spike
j=33
matching_cell_ids = row.names(colData(big_cds)[ colData(big_cds)[,match_index] %in% factor_levels[j] , ]) #get cell id's that match a factor level
sub_cell_matches = match(matching_cell_ids,colnames(big_cds)) #match extracted cells with all cell names
colData(big_cds)$isolated_cell_population[sub_cell_matches] <- "Lower spike" #now replace matching cells with new name

#Extract low-D embeddings and make a graph object bitch
graph_df = data.frame( big_cds@int_colData@listData$reducedDims$UMAP , colData(big_cds)$isolated_cell_population)
names(graph_df) = c("UMAP1","UMAP2","UMAP3","cell_pop")

#Now plot out
png("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Nature/Round\ 2/Figs/Revised\ Fig/Spike\ exemplars/Spikes_betch.png",
    width=7,height=7,units="in",res=600)

print(
  ggplot(data=graph_df, aes(x=UMAP1,y=UMAP2)) +
    theme(legend.position = "none",
          panel.grid = element_blank(),
          axis.title = element_blank(),
          axis.text = element_blank(),
          axis.ticks = element_blank(),
          panel.background = element_blank()) +
    geom_point(size=1.45, aes(alpha=cell_pop,color=cell_pop)) +
    ggtitle( paste( metadatcols[i] , factor_levels[j], sep=" ") ) + xlab("") + ylab("") + 
    theme(axis.text=element_text(size=0,color='black'),
          axis.title=element_text(size=0),
          title=element_text(size=0),
          axis.ticks = element_line(size=0),
          legend.position="none") +
    scale_color_manual(values = c("Non-match" = "lightgrey", 
                                  "Lower spike" = "darkred",
                                  "Upper spike" = "darkgreen")) +
    scale_alpha_manual(values = c("Non-match" = 0.5, 
                                  "Lower spike" = 0.8,
                                  "Upper spike" = 0.8)) +
    xlim(-4.1,0.9) +
    ylim(-7.5, -3.8)
)
dev.off()


#--------------------------------------------------------------------------#
#Now plot each cluster independently/individually

match_index = 14 #this is the column number in the metadata whose individual factor levels you wanna plot out
i = 14 #this is the column number in the metadata whose individual factor levels you wanna plot out
metadatcols = names(colData(big_cds))
factor_levels = na.omit(unique(colData(big_cds)[,match_index]))

factor_levels = as.factor(1:49) #Put the cell clusters in order



for(j in 1:length(factor_levels)){
  tryCatch({ #suppress error
    #j=1
    print(paste("j=",j,sep=""))
    
    matching_cell_ids = row.names(colData(big_cds)[ colData(big_cds)[,match_index] %in% factor_levels[j] , ]) #get cell id's that match a factor level
    
    #Add dummy column
    colData(big_cds)$isolated_cell_population = "Non-match"
    
    #Now match isolated cell IDs with all cell IDs and replace non-match with match in dummy column
    sub_cell_matches = match(matching_cell_ids,colnames(big_cds)) #match extracted cells with all cell names
    colData(big_cds)$isolated_cell_population[sub_cell_matches] <- "Match" #now replace matching cells with new name
    
    #Exract low-D embeddings and make a graph object bitch
    graph_df = data.frame( big_cds@int_colData@listData$reducedDims$UMAP , colData(big_cds)$isolated_cell_population)
    names(graph_df) = c("UMAP1","UMAP2","UMAP3","cell_pop")
    
    #Now plot out
    png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/figures/cell_clusters/clusters_individual/",
              metadatcols[i],"_",factor_levels[j],".png",sep=""),
        width=7,height=7,units="in",res=600)
    
    print(
      ggplot(data=graph_df, aes(x=UMAP1,y=UMAP2)) +
        theme_classic() +
        geom_point(size=0.01, aes(alpha=cell_pop,color=cell_pop)) +
        ggtitle( paste( metadatcols[i] , factor_levels[j], sep=" ") ) + xlab("") + ylab("") + 
        theme(axis.text=element_text(size=0,color='black'),
              axis.title=element_text(size=0),
              title=element_text(size=15),
              axis.ticks = element_line(size=0),
              legend.position="none") +
        scale_color_manual(values = c("Non-match" = "lightgrey", 
                                      "Match" = "darkred")) +
        scale_alpha_manual(values = c("Non-match" = 0.1, 
                                      "Match" = 0.8))
    )
    dev.off()
    
  },  error=function(e){})
  
} #End factor level loop



##################################################################################################################################
##################################################################################################################################
##################################################################################################################################
##################################################################################################################################
#Eigengene pipelines in each Lgr6 stage
#---------------------------------------------------------------------------------------------------#
#Papilloma to Carcinoma: get Pap or CA but all NEOEPITHELIAL CELLS!! this is a good parenchyma definition
all_parenchyma = big_cds[,(colData(big_cds)$tissue_updated %in% c("Pap","CA") & colData(big_cds)$cell_type %in% "nEp") |
                           (colData(big_cds)$cell_type=="PrE" | colData(big_cds)$cell_type=="IfE" | colData(big_cds)$cell_type=="Bas" | 
                              colData(big_cds)$cell_type=="Inf" | colData(big_cds)$cell_type=="UHF" | colData(big_cds)$cell_type=="SG" | 
                              colData(big_cds)$cell_type=="OuB" | colData(big_cds)$cell_type=="Isth" | colData(big_cds)$cell_type=="InB" | 
                              colData(big_cds)$cell_type=="unF")]

min_distances = seq(0.05,1,by=0.05) #This is showing a lot of interesting bridgeness with min distanes bewteen 0.4 and 0.6
#min_distances = seq(0.4,0.6,by=0.01) #This is showing a lot of interesting bridgeness with min distanes bewteen 0.4 and 0.6

sortings = c("Unsorted","Lgr6","Progeny")
for (j in 1:length(sortings)) {
  
  #j=2
  parenchyma_sub = all_parenchyma[,colData(all_parenchyma)$sorting %in% sortings[j]] #subset to a particular sorting parenchyma
  parenchyma_sub <- preprocess_cds(parenchyma_sub, 
                                   method="PCA",
                                   num_dim = 50,
                                   norm_method = "none", #these data have already been normalized by Seurat
                                   alignment_group = NULL)
  
  pdf(file=paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/parenchyma_sorting/all_parenchyma_",sortings[j],"manifold_min_distances_v1.pdf",sep=""),height=10,width=10)
  for (i in 1:length(min_distances)) {
    #i=11
    print(paste("j=",j," i=",i,sep=""))
    #loop to plot in Pap and CA in different ways
    #4b. ?reduce_dimension 
    parenchyma_sub.test <- reduce_dimension(
      parenchyma_sub,
      max_components = 2,
      reduction_method = c("UMAP"),
      preprocess_method = "PCA", #must now used ALIGNEMNT PRE-PROCESSING METHOD bc it replaces the other pre-process shit
      umap.metric = "cosine",
      umap.min_dist = min_distances[i],
      umap.n_neighbors = 15L,
      umap.fast_sgd = FALSE,
      umap.nn_method = "annoy",
      cores = 1,
      verbose = FALSE
    )
    
    #5. Cluster and partition ?cluster_cells cells into supergroups community detection (Phenograph algorithm) Community detection of cells as part of Phenograph algorithm https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4508757/
    #5b.  ?cluster_cells
    parenchyma_sub.test <- cluster_cells(parenchyma_sub.test,
                                         reduction_method = c("UMAP"),
                                         k = 10, #a bigger k will result in lower resolution 
                                         cluster_method = c("leiden"),
                                         num_iter = 2,
                                         partition_qval = 0.05, #Numeric, the q-value cutoff to determine when to partition
                                         weight = FALSE,
                                         resolution = NULL,
                                         random_seed = NULL,
                                         verbose = F)
    print( 
      plot_cells(parenchyma_sub.test, color_cells_by="tissue_updated", group_cells_by="cluster",
                 group_label_size = 0,
                 show_trajectory_graph=F,
                 x=1,
                 y=2,
                 alpha=0.8,
                 cell_size=1) +
        ggtitle(paste(sortings[j]," Minimal distance = ",min_distances[i],sep="")) + xlab("Dim. 1") + ylab("Dim. 2") +
        theme(plot.title = element_text(size = 22, vjust=1)) +
        theme(axis.text=element_text(size=0,color='black'),
              axis.title=element_text(size=0),
              axis.ticks = element_line(size=0)) +
        scale_color_manual(values = c("Normal skin" = "darkblue", "Pap" = "mediumpurple2","CA" = "darkred"))
    )
  }
  dev.off()
  
} #End sortings loop
#--------------------------------------------------------------------------------------#
#--------------------------------------------------------------------------------------#
#Now plot all eigengenes in each of these

sortings = c("Unsorted","Lgr6","Progeny")
optimal_min_distances = c(0.55) #vector of optimal min distances based on visually looking at the above

sortings = c("Unsorted","Lgr6","Progeny")
for (j in 1:length(sortings)) {
  
  #j=1
  
  output_dir = paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/parenchyma_sorting/",sortings[j],"/",sep="")
  dir.create(output_dir)
  
  parenchyma_sub = all_parenchyma[,colData(all_parenchyma)$sorting %in% sortings[j]] #subset to a particular sorting parenchyma
  parenchyma_sub <- preprocess_cds(parenchyma_sub, 
                                   method="PCA",
                                   num_dim = 50,
                                   norm_method = "none", #these data have already been normalized by Seurat
                                   alignment_group = NULL)
  
  #loop to plot in Pap and CA in different ways
  #4b. ?reduce_dimension 
  parenchyma_sub.test <- reduce_dimension(
    parenchyma_sub,
    max_components = 2,
    reduction_method = c("UMAP"),
    preprocess_method = "PCA", #must now used ALIGNEMNT PRE-PROCESSING METHOD bc it replaces the other pre-process shit
    umap.metric = "cosine",
    umap.min_dist = 0.53,
    umap.n_neighbors = 15L,
    umap.fast_sgd = FALSE,
    umap.nn_method = "annoy",
    cores = 1,
    verbose = FALSE
  )
  
  #5. Cluster and partition ?cluster_cells cells into supergroups community detection (Phenograph algorithm) Community detection of cells as part of Phenograph algorithm https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4508757/
  #5b.  ?cluster_cells
  parenchyma_sub.test <- cluster_cells(parenchyma_sub.test,
                                       reduction_method = c("UMAP"),
                                       k = 10, #a bigger k will result in lower resolution 
                                       cluster_method = c("leiden"),
                                       num_iter = 2,
                                       partition_qval = 0.05, #Numeric, the q-value cutoff to determine when to partition
                                       weight = FALSE,
                                       resolution = NULL,
                                       random_seed = NULL,
                                       verbose = F)
  
  #plot out these lovely results
  all_parenchyma_manifold = plot_cells(parenchyma_sub.test, color_cells_by="tissue_updated", group_cells_by="cluster",
                                       group_label_size = 0,
                                       show_trajectory_graph=F,
                                       x=1,
                                       y=2,
                                       alpha=0.8,
                                       cell_size=1) +
    ggtitle(paste(sortings[j]," parenchyma tissue",sep="")) + xlab("Dim. 1") + ylab("Dim. 2") +
    theme(plot.title = element_text(size = 22, vjust=1)) +
    theme(axis.text=element_text(size=0,color='black'),
          axis.title=element_text(size=0),
          axis.ticks = element_line(size=0)) +
    scale_color_manual(values = c("Normal skin" = "darkblue", "Pap" = "mediumpurple2","CA" = "darkred")) +
    theme(legend.position="bottom")
  
  png(paste(output_dir,"1_tissue_manifold.png",sep=""),
      width=10,height=10,units="in",res=100)
  print(all_parenchyma_manifold)
  dev.off()
  
  plot_cells(big_cds,
             genes=c("Lgr6","Sox9"), #seed gene is the first row of the markers file
             label_cell_groups=FALSE,
             label_leaves=F,
             label_branch_points=F,
             graph_label_size=1.5,
             show_trajectory_graph=F,
             cell_size=1) +
    theme(legend.position="bottom") +
    scale_color_viridis_c(option = "inferno",name="log expression") +
    ggtitle(paste(names(balmain_genes)[i],"Seed Gene:", marker_df$gene[1],sep=" "))+ #continuous
    theme(plot.title = element_text(size = 22, vjust=1)) +
    theme(axis.text=element_text(size=0,color='black'),
          axis.title=element_text(size=0),
          axis.ticks = element_line(size=0)) +
    theme(strip.text.x = element_text(size = 15)) #remove facet label
  
  eigengene_agg = do.call(cbind, lapply(1:ncol(balmain_genes), function(i){
    #subset eigengene_agg = do.call(cbind, lapply(172:175, function(i){
    tryCatch({ #suppress error
      
      #i=1
      print(i)
      #only if more than 1 constituent gene
      marker_df = data.frame(na.omit(balmain_genes[,i]),names(balmain_genes)[i])
      names(marker_df) = c("gene","cell_type")
      
      #remake the marker df dataframe with the matched gene names
      marker_df = data.frame(alias2SymbolTable(unique(marker_df$gene), species = "Mm"),names(balmain_genes)[i],names(balmain_genes)[i])
      names(marker_df) = c("gene","cell_type","module")
      marker_df = na.omit(marker_df) #remove non-matching genes
      
      #all cells: Seed gene
      agexp_plot_seedgene = plot_cells(parenchyma_sub.test,
                                       genes=marker_df$gene[1], #seed gene is the first row of the markers file
                                       label_cell_groups=FALSE,
                                       label_leaves=F,
                                       label_branch_points=F,
                                       graph_label_size=1.5,
                                       show_trajectory_graph=F,
                                       cell_size=1) +
        theme(legend.position="bottom") +
        scale_color_viridis_c(option = "inferno",name="log expression") +
        ggtitle(paste(names(balmain_genes)[i],"Seed Gene:", marker_df$gene[1],sep=" "))+ #continuous
        theme(plot.title = element_text(size = 22, vjust=1)) +
        theme(axis.text=element_text(size=0,color='black'),
              axis.title=element_text(size=0),
              axis.ticks = element_line(size=0)) +
        theme(strip.text.x = element_text(size = 0)) #remove facet label
      
      #all cells eigengene
      agexp_plot = plot_cells(parenchyma_sub.test,
                              genes=marker_df,
                              label_cell_groups=FALSE,
                              label_leaves=F,
                              label_branch_points=F,
                              graph_label_size=1.5,
                              show_trajectory_graph=F,
                              cell_size=1) +
        theme(legend.position="bottom") +
        scale_color_viridis_c(option = "inferno",name="integrated expression") +
        ggtitle(paste("All cells:", names(balmain_genes)[i],"Eigengene",sep=" "))+ #continuous
        theme(plot.title = element_text(size = 22, vjust=1)) +
        theme(axis.text=element_text(size=0,color='black'),
              axis.title=element_text(size=0),
              axis.ticks = element_line(size=0)) +
        theme(strip.text.x = element_text(size = 0)) #remove facet label
      
      #-----------#
      #print out eigengene plots
      png(paste(output_dir,names(balmain_genes)[i],".png",sep=""),
          width=18,height=9,units="in",res=100)
      print(multiplot(agexp_plot_seedgene, agexp_plot,
                      cols=2))
      dev.off()
      
    },  error=function(e){})
    
  }))
  
}

##################################################################################################################################
##################################################################################################################################
##################################################################################################################################
##################################################################################################################################
##################################################################################################################################
##################################################################################################################################

#heatmap of all the eigengenes

virpal = viridis_pal(alpha = 1, begin = 0, end = 1, direction = 1,
                     option = "C") #set the colorblind palette
library(BBmisc)
eigengenvals.norm = as.matrix(normalize(t(eigengene_agg), method = "range", range = c(-5, 5), margin = 1L, on.constant = "quiet"))
#subsample to 1k cells
colstosample = sample(1:ncol(eigengenvals.norm),2000)
eigengenvals.norm = eigengenvals.norm[,colstosample]
heatmap.2(eigengenvals.norm, scale = "none", col = virpal, labCol="",
          trace = "none", density.info = "none",
          key.xlab="standardized integrated expression",main = "Eigene expression",
          cexRow = 0.8,
          margins=c(2,15))
detach("package:BBmisc", unload = TRUE)

#-----------------------------------------------------------------#
#order cells by ascending eigenene a plot together
eigengene_agg = as.data.frame(eigengene_agg)
eigengene_agg = eigengene_agg[order(eigengene_agg$Lgr6_carcinoma),]
eigengene_agg$index = seq(1:nrow(eigengene_agg))
metadat_reordered = colData(big_cds)[match(row.names(eigengene_agg),row.names(colData(big_cds))),] #extract correctly ordered metadata
eigengene_agg$tissue = metadat_reordered$tissue
eigen.dfm = melt(eigengene_agg, id=c("index","tissue"))

#loop thru each of the eigengenes and plot against increasing Lgr6_carcinoma
for(j in 1:ncol(balmain_genes)) {
  print(paste("j=",j,sep=""))
  #j=2
  subeigen.dfm = subset(eigen.dfm, (variable=="Lgr6_carcinoma" | variable==names(balmain_genes)[j])) #subset to an eigengene and Lgr6
  subeigen.dfm.lgr6 = subset(eigen.dfm, (variable=="Lgr6_carcinoma")) #subset to an eigengene and Lgr6
  subeigen.dfm.lgr6$value = subeigen.dfm.lgr6$value+(0-min(subeigen.dfm.lgr6$value)) #adjust everything up to 0
  
  subeigen.dfm.eigengene = subset(eigen.dfm, (variable==names(balmain_genes)[j])) #subset to an eigengene and Lgr6
  subeigen.dfm.eigengene$value = subeigen.dfm.eigengene$value+(0-min(subeigen.dfm.eigengene$value)) #adjust everything up to 0
  
  #now loop through each of the 
  subeigen.dfm.comb = rbind(subeigen.dfm.lgr6,subeigen.dfm.eigengene)
  
  ascending_eigen = ggplot(subeigen.dfm.comb, aes(x=index/1000, y=value,color=variable,shape=variable)) +
    geom_point(alpha=0.3, size=0.2) +
    facet_wrap(~tissue, ncol=2, scales="free") +
    theme(legend.key = element_rect(fill = "white")) +
    theme(legend.text=element_text(size=14), legend.title=element_text(size=14)) +
    theme(strip.background=element_rect(fill="white")) +
    ggtitle(paste("Eigengene:",names(balmain_genes)[j],"co-trajectory",sep=" ")) +
    xlab("individual 1000 cell ordinated by increasing Lgr6_carcinoma") +
    ylab("integrated expression") +
    theme(text = element_text(size=18)) +
    theme(axis.text=element_text(size=18,color='black'),
          axis.title=element_text(size=20,color='black')) +
    theme(plot.title = element_text(size = 22, vjust=1)) +
    theme(axis.text=element_text(size=15,color='black'),
          axis.title=element_text(size=15,color='black')) +
    theme(panel.grid.major=element_line(colour = "darkgrey"), 
          panel.grid.minor=element_blank(), 
          panel.background=element_blank()) +
    theme(legend.key = element_rect(fill = "white")) +
    theme(strip.text.x = element_text(size = 15, colour = "black", angle = 0)) +
    theme(strip.text.y = element_text(size = 15, colour = "black", angle = 0)) +
    theme(legend.text=element_text(size=10), legend.title=element_text(size=15)) +
    scale_color_manual(values = c("darkgreen","red")) + 
    theme(legend.position="bottom")
  
  #print(ascending_eigen)
  
}

dev.off()


##################################################################################################################################
##################################################################################################################################
################################################################################################################################## 

#Calculate eigencell of eigengenes: group by gene and by parenchyma cell

#loop thru eigengenes and calculate eigencell values for tissue type (compare_var) parenchymas    
eigencell_agg = do.call(rbind, lapply(1:ncol(balmain_genes), function(i){
  #i=1
  print(i)
  #only if more than 1 constituent gene
  marker_df = data.frame(na.omit(balmain_genes[,i]),names(balmain_genes)[i])
  names(marker_df) = c("gene","cell_type")
  
  #UPDATED NCBI MATCHING!! --> this will product lots of non-matching names that i must resolve with NCBI aliases
  allnetnames = do.call(rbind, lapply(1:nrow(marker_df), function(l){
    #l=1
    netnames = data.frame(marker_df$gene[l],rowData(big_cds)$gene_short_name[match(marker_df$gene[l],rowData(big_cds)$gene_short_name)])
    names(netnames) = c("balmain_network_gene_name","cds_gene_name")
    netnames
  }))
  cds_gene_names = rowData(big_cds)$gene_short_name #get vector of all gene names in cds object
  #get consistent names between cds and network gene names
  match_noms = do.call(rbind, lapply(1:nrow(allnetnames), function(k){
    #k=1 ; 22
    tryCatch({ #suppress error
      if(is.na(allnetnames$cds_gene_name[k])==F) {cds_alias_match = allnetnames$cds_gene_name[k]} else
      { 
        #search through Gene aliases for a match
        ncbi_symbol = ncbi_names[grep(allnetnames$balmain_network_gene_name[k],ncbi_names$Aliases),] #search through aliases for a match
        if (nrow(ncbi_symbol)>0) {
          ncbi_matches_vector = c(as.character(ncbi_symbol$Symbol), as.character(data.frame(strsplit(as.character(ncbi_symbol$Aliases),", "))[,1]) ) #vector of all aliases that marker_gene matches
          cds_alias_match = na.omit(cds_gene_names[match(ncbi_matches_vector,cds_gene_names)]) #match this symbol to the cds gene names
        } else {
          #search through Gene Symbols
          ncbi_symbol.1 = ncbi_names[grep(allnetnames$balmain_network_gene_name[k],ncbi_names$Symbol),] 
          ncbi_matches_vector = c(as.character(ncbi_symbol.1$Symbol), as.character(data.frame(strsplit(as.character(ncbi_symbol.1$Aliases),", "))[,1]) ) #vector of all aliases that marker_gene matches
          cds_alias_match = na.omit(cds_gene_names[match(ncbi_matches_vector,cds_gene_names)]) #match this symbol to the cds gene names
        }
      }
      data.frame(allnetnames[k,],cds_alias_match)
    },  error=function(e){})
  }))
  
  #remake the marker df dataframe with the matched gene names
  marker_df = data.frame(na.omit(unique(match_noms$cds_alias_match)),names(balmain_genes)[i])
  names(marker_df) = c("gene","cell_type")
  
  #cell IDs and treatment
  eigen_cell_def = data.frame(row.names(colData(cds_comb)),colData(cds_comb)$compare_var)
  names(eigen_cell_def) = c("cell_id","cell_group")
  
  #aggregate across cells within all individual genes
  agg_mat = aggregate_gene_expression(
    cds_comb,
    gene_group_df = NULL,
    cell_group_df = eigen_cell_def,
    norm_method = c("log"),
    pseudocount = 1,
    scale_agg_values = TRUE,
    max_agg_value = 10,
    min_agg_value = -5,
    exclude.na = TRUE
  )
  
  #extract genes from eigencell
  agg_matsub = agg_mat[row.names(agg_mat) %in% marker_df$gene,] #get cds names that match the balmain sub network names
  
  #sum across genes to get eigengene
  #if(nrow(agg_matsub)>25) {agg_matsub = agg_matsub[1:25,]} else {agg_matsub = agg_matsub} #balance number of eigengenes by setting max as top 25
  sumdf = data.frame(t(data.frame(colSums(agg_matsub)))) #sum all eigengenes within a cell class
  sumdf = sumdf/nrow(agg_matsub) #standardize by eigengene size: divide by number of included genes
  eigensumdf = data.frame(sumdf$UnTxSkin, sumdf$TxSkin, sumdf$TransitionalPap, sumdf$Pap, sumdf$CA) #re-order
  row.names(eigensumdf) = names(balmain_genes)[i]
  names(eigensumdf) = parenchymas
  eigensumdf
  
}))

eigencell_agg

#now heatmap of those eigencell eigengenes:  columns are eigencells, rows are eigengenes
virpal = viridis_pal(alpha = 1, begin = 0, end = 1, direction = 1,
                     option = "C") #set the colorblind palette
eigencell_agg.mat = as.matrix(eigencell_agg)

pdf(file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/balmain_eigene/eigencell_eigengene_v7.pdf",height=15,width=15)
heatmap.2(eigencell_agg.mat, scale = "none", col = virpal, dendrogram="row",
          trace = "none", density.info = "none",key.xlab="eigencell eigengene expression",
          main = "Parenchyma Eigencell Eigengene Expression",
          key=T,
          cexCol = 1,
          cexRow = 1,
          Rowv=T,
          Colv=F,
          keysize=1,
          colsep=c(1:27),
          rowsep=c(1:28),
          sepcolor="lightgrey",
          sepwidth=c(0.02,0.02),
          margins=c(8,15))
dev.off()

match("Fat1",row.names(big_cds))
plot_cells(big_cds,
           genes="Fat1", #seed gene is the first row of the markers file
           label_cell_groups=FALSE,
           label_leaves=F,
           label_branch_points=F,
           graph_label_size=1.5,
           show_trajectory_graph=F) +
  xlab("Dim. 1") + ylab("Dim. 2") + 
  theme(legend.position="bottom") +
  scale_color_viridis_c(option = "inferno",name="integrated expression") +
  ggtitle("Fat1")+ #continuous
  theme(axis.text=element_text(size=8,color='black'),
        axis.title=element_text(size=10,color='black')) + theme(plot.title = element_text(size = 14, vjust=1)) +
  theme(strip.text.x = element_text(size = 0)) #remove facet label


####################################################################################################################
####################################################################################################################
####################################################################################################################
#Eigengene DEG analysis along 5 tissues simultaneously with cds_comb with only tissue parenchyma

eigengene_cld = do.call(rbind, lapply(1:ncol(balmain_genes), function(i){
  #i=1
  print(i)
  #only if more than 1 constituent gene
  marker_df = data.frame(na.omit(balmain_genes[,i]),names(balmain_genes)[i])
  names(marker_df) = c("gene","cell_type")
  
  #UPDATED NCBI MATCHING!! --> this will product lots of non-matching names that i must resolve with NCBI aliases
  allnetnames = do.call(rbind, lapply(1:nrow(marker_df), function(l){
    #l=1
    netnames = data.frame(marker_df$gene[l],rowData(big_cds)$gene_short_name[match(marker_df$gene[l],rowData(big_cds)$gene_short_name)])
    names(netnames) = c("balmain_network_gene_name","cds_gene_name")
    netnames
  }))
  cds_gene_names = rowData(big_cds)$gene_short_name #get vector of all gene names in cds object
  #get consistent names between cds and network gene names
  match_noms = do.call(rbind, lapply(1:nrow(allnetnames), function(k){
    #k=1 ; 22
    tryCatch({ #suppress error
      if(is.na(allnetnames$cds_gene_name[k])==F) {cds_alias_match = allnetnames$cds_gene_name[k]} else
      { 
        #search through Gene aliases for a match
        ncbi_symbol = ncbi_names[grep(allnetnames$balmain_network_gene_name[k],ncbi_names$Aliases),] #search through aliases for a match
        if (nrow(ncbi_symbol)>0) {
          ncbi_matches_vector = c(as.character(ncbi_symbol$Symbol), as.character(data.frame(strsplit(as.character(ncbi_symbol$Aliases),", "))[,1]) ) #vector of all aliases that marker_gene matches
          cds_alias_match = na.omit(cds_gene_names[match(ncbi_matches_vector,cds_gene_names)]) #match this symbol to the cds gene names
        } else {
          #search through Gene Symbols
          ncbi_symbol.1 = ncbi_names[grep(allnetnames$balmain_network_gene_name[k],ncbi_names$Symbol),] 
          ncbi_matches_vector = c(as.character(ncbi_symbol.1$Symbol), as.character(data.frame(strsplit(as.character(ncbi_symbol.1$Aliases),", "))[,1]) ) #vector of all aliases that marker_gene matches
          cds_alias_match = na.omit(cds_gene_names[match(ncbi_matches_vector,cds_gene_names)]) #match this symbol to the cds gene names
        }
      }
      data.frame(allnetnames[k,],cds_alias_match)
    },  error=function(e){})
  }))
  
  #remake the marker df dataframe with the matched gene names
  marker_df = data.frame(na.omit(unique(match_noms$cds_alias_match)),names(balmain_genes)[i])
  names(marker_df) = c("gene","cell_type")
  
  #aggregate eigengene expression
  agg_mat = aggregate_gene_expression(
    cds_comb,
    gene_group_df = marker_df,
    cell_group_df = NULL,
    norm_method = c("log"),
    pseudocount = 1,
    scale_agg_values = TRUE,
    max_agg_value = 10,
    min_agg_value = -5,
    exclude.na = TRUE
  )
  
  agg_matdf = data.frame(t(data.frame(agg_mat))) #dataframe where each cell has eigengene value
  agg_matdf$tissue = as.vector(colData(cds_comb)$compare_var) #add tissue 
  names(agg_matdf) = c("eigengene_expression","tissue")
  
  #scale everything up to 0 for negative binormail test
  rescale_value = 0-min(agg_matdf$eigengene_expression)
  agg_matdf$eigengene_expression = agg_matdf$eigengene_expression+rescale_value
  
  #fit negative binomial GLM to eigengene testing for amount-tissue differences
  glmnb_res <- glm.nb(eigengene_expression~tissue, data = agg_matdf)
  
  #CLD on negative binomial GLM
  lsmeanres = data.frame(cld(emmeans(glmnb_res,specs = c("tissue"))))
  names(lsmeanres) = c("tissue","mean","SE","df","lower.CL","upper.CL","CLD")
  
  #lsmeanres$Kruskal_Wallis_p = kruskal.test(eigengene_expression~tissue, data=agg_matdf)$p.value #add K-W p-value to output
  
  lsmeanres$eigengene = names(balmain_genes)[i] #add column with the eigengene 
  lsmeanres
  
})) #end do.call loop

write.table(eigengene_cld, file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/balmain_eigene/eigengene_expression_glm.txt",quote=F,sep="\t",row.names=F)

####################################################################################################################
####################################################################################################################
####################################################################################################################

pdf(file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/balmain_eigene/eigengene_correlations_v1.pdf",height=19,width=19)

all_corrplot = ggcorr(eigengene_everycell, 
                      method = c("complete.obs","spearman"),
                      geom = "circle",
                      low="slateblue4",high="darkorange3",
                      max_size = 11,
                      hjust = 1,
                      angle = 0,
                      size = 6,layout.exp=10,
                      legend.position = "right",
                      legend.size = 15,
) + ggtitle("Parenchymaous Eigengene Expression\nPairwise Spearman Correlations") +
  theme(plot.title = element_text(size = 25, face = "bold"))

print(all_corrplot)

all_corrplot_pearson = ggcorr(eigengene_everycell, 
                              method = c("complete.obs","pearson"),
                              geom = "circle",
                              low="slateblue4",high="darkorange3",
                              max_size = 11,
                              hjust = 1,
                              angle = 0,
                              size = 6,layout.exp=10,
                              legend.position = "right",
                              legend.size = 15,
) + ggtitle("Parenchymaous Eigengene Expression\nPairwise Pearson Correlations") +
  theme(plot.title = element_text(size = 25, face = "bold"))

print(all_corrplot_pearson)

#-----------------------------------------------------------------#
#Plot indiivdual eigengenes vs Lgr6

#order cells by ascending eigenene a plot together
f = c("Lgr6_carcinoma_49gene",
      "stem.cell.markers.1",
      "Prom1a","Prom1b","Sox9","Lgr5","Psca")

#subset to metrics to be used in pairsplot
lesubset= subset(eigengene_everycell, select=f)
eigengene_agg = lesubset[order(lesubset$Lgr6_carcinoma_49gene),]
eigengene_agg$index = seq(1:nrow(eigengene_agg))

cellord = gsub("[.]","-",row.names(eigengene_agg), ) #replace the decimal with dash so that they match
metadat_reordered = colData(cds_comb)[match(cellord, row.names(colData(cds_comb))),] #extract correctly ordered metadata
eigengene_agg$tissue = metadat_reordered$compare_var
eigen.dfm = melt(eigengene_agg, id=c("index","tissue"))

#plot each of the eigengenes and plot against increasing Lgr6_carcinoma

eigen.dfm$value = eigen.dfm$value+(0-min(eigen.dfm$value)) #adjust everything up to 0

ascending_eigen = ggplot(eigen.dfm, aes(x=index/1000, y=value,color=variable,group=variable)) +
  #geom_point(alpha=0.05, size=0.2) +
  stat_smooth(se = F) +
  theme(legend.key = element_rect(fill = "white")) +
  theme(legend.text=element_text(size=14), legend.title=element_text(size=14)) +
  theme(strip.background=element_rect(fill="white")) +
  ggtitle(paste("Eigengene co-trajectories",sep=" ")) +
  xlab("Ordinated Lgr6 Eigengene (1000 cells)") +
  ylab("integrated expression") +
  theme(text = element_text(size=18)) +
  theme(axis.text=element_text(size=18,color='black'),
        axis.title=element_text(size=20,color='black')) +
  theme(plot.title = element_text(size = 22, vjust=1)) +
  theme(axis.text=element_text(size=15,color='black'),
        axis.title=element_text(size=20,color='black')) +
  theme(panel.grid.major=element_line(colour = "darkgrey"), 
        panel.grid.minor=element_blank(), 
        panel.background=element_blank()) +
  theme(legend.key = element_rect(fill = "white")) +
  theme(strip.text.x = element_text(size = 15, colour = "black", angle = 0)) +
  theme(strip.text.y = element_text(size = 15, colour = "black", angle = 0)) +
  theme(legend.text=element_text(size=15), legend.title=element_text(size=15)) +
  scale_color_manual(values = cbbPalette, name="Eigengene") + 
  theme(legend.position="bottom")

print(ascending_eigen)

head(eigen.dfm)
#tissue distribution
tissue_dfm = subset(eigen.dfm, variable=="Lgr6_carcinoma_49gene" ) #get only 1 variable so only 1 set of points
tissue_dfm$dummyvar = 0
tissue_eigen = ggplot(tissue_dfm, aes(x=index/1000, y=dummyvar,color=tissue)) +
  geom_point(position = position_jitter(height=100), alpha=0.4, size=0.5) +
  #stat_smooth(se = F) +
  theme(legend.key = element_rect(fill = "white")) +
  theme(legend.text=element_text(size=14), legend.title=element_text(size=14)) +
  theme(strip.background=element_rect(fill="white")) +
  ggtitle(paste("Tissue distribution",sep=" ")) +
  xlab("Ordinated Lgr6 Eigengene (1000 cells)") +
  ylab("") +
  theme(text = element_text(size=18)) +
  theme(axis.text.y=element_text(size=0,color='black'),
        axis.text.x=element_text(size=18,color='black'),
        axis.title=element_text(size=20,color='black')) +
  theme(plot.title = element_text(size = 22, vjust=1)) +
  theme(axis.text=element_text(size=15,color='black'),
        axis.title=element_text(size=20,color='black')) +
  theme(panel.grid.major=element_line(colour = "darkgrey"), 
        panel.grid.minor=element_blank(), 
        panel.background=element_blank()) +
  theme(legend.key = element_rect(fill = "white")) +
  theme(strip.text.x = element_text(size = 15, colour = "black", angle = 0)) +
  theme(strip.text.y = element_text(size = 15, colour = "black", angle = 0)) +
  theme(legend.text=element_text(size=15), legend.title=element_text(size=15)) +
  scale_color_manual(name="Tissue",values = c("UnTxSkin" = "skyblue3","TxSkin" = "darkblue", "TransitionalPap" = "green","Pap" = "darkgreen","CA" = "darkred")) + theme(legend.position="bottom") +
  theme(legend.position="bottom")

print(tissue_eigen)


dev.off()

##################################################################################################################################
##################################################################################################################################
##################################################################################################################################  
##################################################################################################################################
##################################################################################################################################
##################################################################################################################################
#---------------------------------------------------------------------------------------------------#
#---------------------------------------------------------------------------------------------------#
#---------------------------------------------------------------------------------------------------#
#---------------------------------------------------------------------------------------------------#
#Define bridge cells
#START BRIDGE ANALYSIS
#---------------------------------------------------------------------------------------------------#
#Papilloma to Carcinoma: get Pap or CA but all NEOEPITHELIAL CELLS!! this is a good parenchyma definition
pap_ca = big_cds[,colData(big_cds)$tissue_updated %in% c("Pap","CA") & colData(big_cds)$cell_type %in% "nEp"]

min_distances = seq(0.05,1,by=0.05) #This is showing a lot of interesting bridgeness with min distanes bewteen 0.4 and 0.6
min_distances = seq(0.4,0.6,by=0.01) #This is showing a lot of interesting bridgeness with min distanes bewteen 0.4 and 0.6

pap_ca <- preprocess_cds(pap_ca, 
                         method="PCA",
                         num_dim = 50,
                         norm_method = "none", #these data have already been normalized by Seurat
                         alignment_group = NULL)


#now do it between 0.4 and 0.6

pdf(file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/figures/manifold_optimization/pap_ca_manifold_min_distances_v3.2.pdf",height=10,width=10)
for (i in 1:length(min_distances)) {
  #i=11
  print(i)
  #loop to plot in Pap and CA in different ways
  #4b. ?reduce_dimension 
  pap_ca.test <- reduce_dimension(
    pap_ca,
    max_components = 2,
    reduction_method = c("UMAP"),
    preprocess_method = "PCA", #must now used ALIGNEMNT PRE-PROCESSING METHOD bc it replaces the other pre-process shit
    umap.metric = "cosine",
    umap.min_dist = min_distances[i],
    umap.n_neighbors = 15L,
    umap.fast_sgd = FALSE,
    umap.nn_method = "annoy",
    cores = 1,
    verbose = FALSE
  )
  
  #5. Cluster and partition ?cluster_cells cells into supergroups community detection (Phenograph algorithm) Community detection of cells as part of Phenograph algorithm https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4508757/
  #5b.  ?cluster_cells
  pap_ca.test <- cluster_cells(pap_ca.test,
                               reduction_method = c("UMAP"),
                               k = 10, #a bigger k will result in lower resolution 
                               cluster_method = c("leiden"),
                               num_iter = 2,
                               partition_qval = 0.05, #Numeric, the q-value cutoff to determine when to partition
                               weight = FALSE,
                               resolution = NULL,
                               random_seed = NULL,
                               verbose = F)
  print( 
    plot_cells(pap_ca.test, color_cells_by="tissue", group_cells_by="cluster",
               group_label_size = 0,
               show_trajectory_graph=F,
               x=1,
               y=2,
               alpha=0.8,
               cell_size=1) +
      ggtitle(paste("Minimal distance = ",min_distances[i],sep="")) + xlab("Dim. 1") + ylab("Dim. 2") +
      theme(text = element_text(size=18)) + theme(axis.text=element_text(size=18,color='black'),
                                                  axis.title=element_text(size=20,color='black')) + theme(plot.title = element_text(size = 22, vjust=1)) + theme(axis.text=element_text(size=15,color='black'),
                                                                                                                                                                 axis.title=element_text(size=15,color='black'))+
      scale_color_manual(values = c("UnTxSkin" = "skyblue3","TxSkin" = "darkblue", "Pap" = "darkgreen","CA" = "darkred")) + theme(legend.position="bottom")
  )
  
}
dev.off()

#--------------------------------------------------------------------------------------#
#--------------------------------------------------------------------------------------#
#--------------------------------------------------------------------------------------#
#--------------------------------------------------------------------------------------#
#Optimal min_ditance = 0.44
pap_ca = big_cds[,colData(big_cds)$tissue_updated %in% c("Pap","CA") & colData(big_cds)$cell_type %in% "nEp"]

pap_ca <- preprocess_cds(pap_ca, 
                         method="PCA",
                         num_dim = 50,
                         norm_method = "none", #these data have already been normalized by Seurat
                         alignment_group = NULL)

#4b. ?reduce_dimension 
pap_ca <- reduce_dimension(
  pap_ca,
  max_components = 2,
  reduction_method = c("UMAP"),
  preprocess_method = "PCA", #must now used ALIGNEMNT PRE-PROCESSING METHOD bc it replaces the other pre-process shit
  umap.metric = "cosine",
  umap.min_dist = 0.44, #this is based on optimizing the manifold to visualize bridge cells (See pap_ca_manifold_min_distances.pdf)
  umap.n_neighbors = 15L,
  umap.fast_sgd = FALSE,
  umap.nn_method = "annoy",
  cores = 1,
  verbose = FALSE
)

#5. Cluster and partition ?cluster_cells cells into supergroups community detection (Phenograph algorithm) Community detection of cells as part of Phenograph algorithm https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4508757/
#5b.  ?cluster_cells
pap_ca <- cluster_cells(pap_ca,
                        reduction_method = c("UMAP"),
                        k = 10, #a bigger k will result in lower resolution 
                        cluster_method = c("leiden"),
                        num_iter = 2,
                        partition_qval = 0.05, #Numeric, the q-value cutoff to determine when to partition
                        weight = FALSE,
                        resolution = NULL,
                        random_seed = NULL,
                        verbose = F)

png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/figures/pap_ca/pap_ca_integrated_markers.png",sep=""),
    width=8,height=8,units="in",res=300)

pap_cafig = plot_cells(pap_ca, color_cells_by="integrated_type_bridge", group_cells_by="cluster",
                       group_label_size = 0,
                       show_trajectory_graph=F,
                       x=1,
                       y=2,
                       alpha=0.8,
                       cell_size=1) +
  ggtitle("") + xlab("") + ylab("") +
  theme(plot.title = element_text(size=20),
        axis.text=element_text(size=0,color='black'),
        axis.title=element_text(size=0),
        axis.ticks = element_line(size=0)) +
  scale_color_manual(values = c("nEp" = "mediumpurple","lower spike" = "olivedrab4","upper spike" = "yellow4","squamous"="red","spindle"="darkred",
                                "nEp"="lightgrey",
                                "Upper Bridge"="darkorange4","Lower Bridge"="goldenrod3","Middle Bridge"="navajowhite3")) + 
  theme(legend.position="bottom")

print(pap_cafig)
dev.off()

#-----------------------------------------------------------#
#6. Learn trajectory for the optimal min-dstance graph
pap_ca <- learn_graph(
  pap_ca,
  use_partition = F, #lear graph across this all
  close_loop = TRUE,
  learn_graph_control = NULL,
  verbose = FALSE
)

#Fig. 3
png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Figs/Fig3/pap_ca_trajectory.png",sep=""),
    width=8,height=8,units="in",res=300)
pap_cafig = plot_cells(pap_ca, color_cells_by="integrated_type_bridge", group_cells_by="cluster",
                       group_label_size = 0,
                       show_trajectory_graph=T,
                       x=1,
                       y=2,
                       alpha=0.8,
                       cell_size=1.2,
                       label_branch_points = T,
                       label_roots = F,
                       label_leaves = T,
                       graph_label_size=1) +
  ggtitle("") + xlab("") + ylab("") +
  theme(plot.title = element_text(size=0),
        axis.text=element_text(size=0,color='black'),
        axis.title=element_text(size=0),
        axis.ticks = element_line(size=0)) +
  scale_color_manual(values = c("nEp" = "mediumpurple","lower spike" = "olivedrab4","upper spike" = "yellow4","squamous"="red","spindle"="darkred",
                                "nEp"="lightgrey",
                                "Upper Bridge"="darkorange4","Lower Bridge"="goldenrod3","Middle Bridge"="navajowhite3")) + theme(legend.position="none")
print(pap_cafig)
dev.off()

##################################################################################################################################
##################################################################################################################################
#Pap_ca eigengenes for the mansucript

#Get all the Tgf genes and map out
seed_gene = c( aim2probedat$SYMBOL[grep("Tgf",aim2probedat$SYMBOL)], "Cd47","Tmem173","Mb21d1",
               "Cd274","Ctla4","Pdcd1", "Cd80")
match("Pdcd1",row.names(big_cds))
row.names(big_cds)[grep("Cg",row.names(big_cds))]
aim2probedat$SYMBOL[grep("Mb21d1",aim2probedat$SYMBOL)]

eigengene_size=99
eigengene_agg = do.call(cbind, lapply(1:length(seed_gene), function(i){
  #subset eigengene_agg = do.call(cbind, lapply(106:118, function(i){
  tryCatch({ #suppress error
    
    #i=17
    print(i)
    
    #---------------------------------------#
    #tumor
    match_index = match(seed_gene[i], aim2probedat$SYMBOL) #match seed gene to its place in the matrix
    symbol_match = data.frame(aim2probedat$SYMBOL, tumor_rhodf[,match_index], tumor_p_df[,match_index]) #put together gene names, expression data, and p-vals
    symbol_match = symbol_match[order(-symbol_match$tumor_rhodf...match_index.),] #order from high to low gene values
    names(symbol_match) = c("gene","rho","p")
    
    #make marker df for grahing eigengene
    marker_df = data.frame(unique(alias2SymbolTable(symbol_match$gene[1:(eigengene_size+1)], species = "Mm")),paste(seed_gene[i],"Carcinoma",sep=""),paste(seed_gene[i],"Carcinoma",sep=""))
    names(marker_df) = c("gene","cell_type","module")
    marker_df = na.omit(marker_df)
    
    #Pap-ca
    agexp_plot = plot_cells(pap_ca,
                            genes=marker_df,
                            label_cell_groups=FALSE,
                            label_leaves=F,
                            label_branch_points=F,
                            graph_label_size=1.5,
                            show_trajectory_graph=F,
                            cell_size=1.2) +
      xlab("Dim. 1") + ylab("Dim. 2") + 
      theme(legend.position="bottom") +
      scale_color_viridis_c(option = "inferno",name="") +
      ggtitle(paste(seed_gene[i],"SCC Eigengene",sep=" "))+ #continuous
      theme(axis.text=element_text(size=0,color='black'),
            axis.title=element_text(size=0,color='black')) + theme(plot.title = element_text(size = 0, vjust=0)) +
      theme(strip.text.x = element_text(size = 0)) +
      xlim(-1,6) +
      ylim(-2.5,6) +
      theme(axis.ticks = element_blank()) +
      ggtitle(seed_gene[i]) +
      theme(plot.title = element_text(size=15),
            legend.text=element_text(size=12))
    
    #-----------#
    #print out eigengene plots
    png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Figs/Fig3/manifolds_pap_ca/",seed_gene[i],".png",sep=""),
        width=2.25,height=6,units="in",res=200)
    print(agexp_plot)
    dev.off()
    
  },  error=function(e){})
  
}))


##################################################################################################################################
##################################################################################################################################

#Fig. 4 manifolds
match("Atm",row.names(big_cds))
match("Mb21d1",row.names(big_cds))

seed_gene = c("Cdkn2a","Cdkn2b","Trp53","Slc4a11","Tigit","Tacstd2","Ly6d","Lypd3","Tmem173",
              "Atm","Atr","Foxm1","E2f1",
              "Banf1","Mb21d1")

#Loop through all eigengenes
eigengene_agg = do.call(cbind, lapply(1:length(seed_gene), function(i){
  tryCatch({ #suppress error
    
    #---------------------------------------#
    #i=15
    print(i)
    
    #Get the correct match_index
    if( seed_gene[i]=="Prom1" ) {match_index = 12747} else {
      match_index = match(seed_gene[i], aim2probedat$SYMBOL) #match seed gene to its place in the matrix
    }
    
    #---------------------------------------#
    #Normal skin
    symbol_match = data.frame(aim2probedat$SYMBOL, rhodf[,match_index], p_df[,match_index]) #put together gene names, expression data, and p-vals
    symbol_match = symbol_match[order(-symbol_match$rhodf...match_index.),] #order from high to low gene values
    names(symbol_match) = c("gene","rho","p")
    
    #make marker df for grahing eigengene
    marker_df = data.frame(unique(alias2SymbolTable(symbol_match$gene[1:(eigengene_size+1)], species = "Mm")),paste(seed_gene[i]," Normal skin",sep=""),paste(seed_gene[i]," Normal skin",sep=""))
    names(marker_df) = c("gene","cell_type","module")
    marker_df = na.omit(marker_df)
    
    #------------------------------#
    #4a. Seed gene
    #------------------------------#
    #all cells eigengene
    seed_gene_manifold = plot_cells(big_cds,
                                    genes=seed_gene[i],
                                    label_cell_groups=FALSE,
                                    label_leaves=F,
                                    label_branch_points=F,
                                    graph_label_size=1.5,
                                    show_trajectory_graph=F) +
      xlab("") + ylab("")  +
      theme(axis.ticks = element_blank()) +
      scale_color_viridis_c(option = "inferno",name="") +
      theme(axis.text=element_text(size=0,color='black'),
            axis.title=element_text(size=0,color='black')) + theme(plot.title = element_text(size = 0, vjust=0)) +
      theme(strip.text.x = element_text(size = 0)) +
      annotate("text", label = paste(seed_gene[i],"\nseed gene",sep=""), 
               x = -6, y = 7, vjust = 0,size=10) +
      theme(legend.position=c(0.8, 0.2),
            legend.key.size = unit(1, 'cm'),
            legend.text = element_text(size=20))
    
    
    #------------------------------#
    #4a. Normal skin WT: calculate correlation matrix
    #------------------------------#
    #all cells eigengene
    agexp_plot.12 = plot_cells(big_cds,
                               genes=marker_df,
                               label_cell_groups=FALSE,
                               label_leaves=F,
                               label_branch_points=F,
                               graph_label_size=1.5,
                               show_trajectory_graph=F) +
      xlab("") + ylab("") + 
      theme(axis.ticks = element_blank()) +
      scale_color_viridis_c(option = "inferno",name="") +
      theme(axis.text=element_text(size=0,color='black'),
            axis.title=element_text(size=0,color='black')) + theme(plot.title = element_text(size = 0, vjust=0)) +
      theme(strip.text.x = element_text(size = 0))  +
      annotate("text", label = paste(seed_gene[i],"\nNSk eigengene",sep=""), 
               x = -5, y = 7, vjust = 0,size=10) +
      theme(legend.position=c(0.8, 0.2),
            legend.key.size = unit(1, 'cm'),
            legend.text = element_text(size=20))
    
    #---------------------------------------#
    #tumor
    #Get the correct match_index
    if( seed_gene[i]=="Prom1" ) {match_index = 12747} else {
      match_index = match(seed_gene[i], aim2probedat$SYMBOL) #match seed gene to its place in the matrix
    }
    symbol_match = data.frame(aim2probedat$SYMBOL, tumor_rhodf[,match_index], tumor_p_df[,match_index]) #put together gene names, expression data, and p-vals
    symbol_match = symbol_match[order(-symbol_match$tumor_rhodf...match_index.),] #order from high to low gene values
    names(symbol_match) = c("gene","rho","p")
    
    
    #make marker df for grahing eigengene
    marker_df = data.frame(unique(alias2SymbolTable(symbol_match$gene[1:(eigengene_size+1)], species = "Mm")),paste(seed_gene[i],"Carcinoma",sep=""),paste(seed_gene[i],"Carcinoma",sep=""))
    names(marker_df) = c("gene","cell_type","module")
    marker_df = na.omit(marker_df)
    
    #all cells eigengene
    agexp_plot_tumor.12 = plot_cells(big_cds,
                                     genes=marker_df,
                                     label_cell_groups=FALSE,
                                     label_leaves=F,
                                     label_branch_points=F,
                                     graph_label_size=1.5,
                                     show_trajectory_graph=F) +
      xlab("") + ylab("") + 
      theme(axis.ticks = element_blank()) +
      scale_color_viridis_c(option = "inferno",name="") +
      theme(axis.text=element_text(size=0,color='black'),
            axis.title=element_text(size=0,color='black')) + theme(plot.title = element_text(size = 0, vjust=0)) +
      theme(strip.text.x = element_text(size = 0))  +
      annotate("text", label = paste(seed_gene[i],"\nCar eigengene",sep=""), 
               x = -5, y = 7, vjust = 0,size=10) +
      theme(legend.position=c(0.8, 0.2),
            legend.key.size = unit(1, 'cm'),
            legend.text = element_text(size=20))
    
    #-----------#
    #print out seed gene and NSk eigengene plots for extended data
    png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Extended\ Data/manifolds/seed_nsk/",seed_gene[i],".png",sep=""),
        width=14,height=7,units="in",res=300)
    
    print(multiplot(seed_gene_manifold, agexp_plot.12,
                    cols=2))
    dev.off()
    
    #-----------#
    #print out tumor eigengene plots
    png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Figs/Fig4/fig4_eigengene_manifolds/",seed_gene[i],".png",sep=""),
        width=7,height=7,units="in",res=300)
    
    print(multiplot(agexp_plot_tumor.12,
                    cols=1))
    dev.off()
    
  },  error=function(e){})
  
}))

##################################################################################################################################
#The Marjonovic et al cluster 5 lung high plasticity gene markers

cluster5_lung = c("Slc4a11","Il23a","Rrad","Tnf","Abtb2","Lif","Tnfrsf11b","Tigit","Dusp10","Ntm","Trim15","Tm4sf1","Chac1","Creb5","5031415H12Rik","Cxcl1","Mcpt8","Trp53inp1","Abcg1","Mapk11","Klhl2","Il24","Zfand2a","Tspan5","Ddit3","Ascc3","Sgms2","Grpel2","Gdf15","Dennd2c","Ly6a","Slc9a4","Asns","Ets1","Ugcg","Gcnt2","Rnf19b","Pvt1","3110043O21Rik","Cln8","Hmces","Maff","Tnfaip3","Prkx","Tgfa","Pxdc1","Cdk17","Slc1a4","Arid5a","Cldn4","Iars","Ptgs2","Relb","Mxd1","Ppp1r13l","Btg1","Slc38a2","Sema7a","Smcr8","Cwh43","Aaed1","B930095G15Rik","Pvr","Fbxo30","Zdhhc5","Chrnb1","Rhbdf2","Fam50a","Cxcl16","Itga2","Hs3st1","Trib1","Tgif1","B3gnt2","Rdh10","Rhob","Gm28048","Chmp4c","Chic2","Klf6","Bcl10","2610035D17Rik","Gm10157","Cars","Bax","Agap1","Llph-ps2","Otud4","Prkca","AI413582","Procr","9930111J21Rik1","Fam83g","Ocln","1810013L24Rik","Otud7b","Actn1","Gtpbp4","Nars","Pik3c2a","Vps37b","Cdkn1a","Med13l","Mdm2","Trp53","Klc1","Cxadr","Sgpl1","Ptrh1","Arid3b","Sars","Btbd10","2900026A02Rik","Galnt3","Coq10b","Fgd6","Mafk","Ddit4","Mier1","Tsc22d2","Stam","Fdx1l","Hgs","Sbf2","Pdlim5","Fam96b","Itga3","Mpzl3","Wnt7b","Mob3a","Atp2b1","Slk","Map2k4","Ier3","Ets2","Erdr1","Slc7a1","Llph","Aff4","Kdm6b","Cdcp1","Ddx21","Utp14a","Irf6","Krt8-ps","Smndc1","Itgav","Tnfrsf23","Efna5","Ptbp3","Usp6nl","Kctd13","Malt1","Stk17b","Ptpn23","Rap2b","Lad1","Abt1","Nfkb2","Sertad2","Sh3d19","Tnfrsf10b","Nras","Wipf2","Ubr2","Cyth2","Ankrd12","Mapk6","Slc6a9","Cldn7","Trip10","Mnt","Tead1","Kif5b","Atf5","Cep120","Ccnl1","Esyt2","Rhbdf1","Gmeb1","Ppp4r3b","Prrg4","Ago2","Utrn","Atxn1","Gm7809","Poldip3","Gm42899","Src","Ahctf1","Areg","Tuft1","Zfp57","Pdpk1","Grk2","Atp6v0b","Cd2ap","Leprotl1","Flnb","Pip5k1a","Osbpl3","Mak16","Smim15","Fam69a","4931406P16Rik","BC016579","Spag9","Sh2d1b1","Ints12","Crtc2","Ercc1","Togaram1","Gm37401","Sqstm1","Cdc42ep1","Tmem184b","Rusc2","Cpsf7","Slc25a19","Psmd8","Rnf19a","St13","Ddx54","Rab1a","Syf2","Rbms1","Ube2b","Manba","Brd4","Appl1","Mafg","Kdm5c","Ammecr1l","Smox","Gm9892","Cdv3","Myc","Fbxw11","Serinc1","Mrpl38","Adm2","Samd5","Arhgap26","Fhl2","Brd2","Cdc42bpb","Fam13b","Apba3","Clu","Pdcd2","Mpp5","Tnip1","Ccdc22","Stx11","Rc3h1","Ggps1","Rabgef1","Heatr5a","Arpc5","Gm14253","Eif2s2","Tjp2","Tars","Eif1","Wasf2","Atf4","Rictor","St14","Krt23","Atp6ap2","Sfn","Krt18","Smg1","Rela","Fzd5","Pogk","Slc3a2","Ube2d3","Fosl2","2810403A07Rik","Tsg101-ps","AC160980.1","Cct2","Krt80","Ccng2","Kbtbd2","Comtd1","Tpd52l2","Txndc9","Gm13050","Trmt10c","Rab6a","Arl8a","March7","Actr2","Akirin1","Smg7","Mindy3","Eea1","Neat1","Zfp7","Pdcl3","Tnfrsf26","Lrrc8a","Gng5","Yars","Rhoa","Ctnna1","Vps35","Actg-ps1","Rasa2","Etf1","Eif5","Tnip3","Tnfrsf12a","Junb","Slc25a25","Rnf44","Krt8","Pwp2","Wnt7a","Aars","Dnajb6","Afdn","Hspa8","Vmp1","Klf7","Ezr","Tsg101","Hspa9","Atg101","Eif3g","Trib3","Eif3a","Chd1","Tmf1","Rab7","Nrbp1","Wdr36","Ghitm","Perp","Furin","Mat2a","2410006H16Rik","Tpm1","Cycs","Ubald2","Klf11","Wipi2","Eif6","Rabggtb","Fnbp1l","Sbno1","F11r","Prmt2","Pls3","Dusp16","Nfe2l1","Polk","Cltc","Npepps","Eef2","Hnrnpk","Ncoa2","Trappc4","Cdc42","Vps26a","Lasp1","Rack1","Dusp3","Plcd3","Pak1ip1","Rlim","Eif4g2","Naa15","Adgrl2","Cast","Tecpr1","Ubxn6","Eps15","Picalm","H3f3b","Dph3","Tcf20","Map3k20","Hsp90ab1","Tbrg1","Zfand5","Mea1","Stub1","Actb","Crb3","Anxa3","Lgals3","Epcam","Cct4","Gm6210","Ndrg1","Gm19680","Anxa7","Snrnp48","Hagh","Fam107b","Bcr","Frg1","Plscr1")

#make marker df for grahing eigengene
marker_df = data.frame(unique(alias2SymbolTable(cluster5_lung, species = "Mm")),"eigengene","eigengene")
names(marker_df) = c("gene","cell_type","module")
marker_df = na.omit(marker_df)

#all cells eigengene
cluster5 = plot_cells(big_cds,
                      genes=marker_df,
                      label_cell_groups=FALSE,
                      label_leaves=F,
                      label_branch_points=F,
                      graph_label_size=1.5,
                      show_trajectory_graph=F) +
  xlab("") + ylab("") + 
  theme(axis.ticks = element_blank()) +
  scale_color_viridis_c(option = "inferno",name="") +
  theme(axis.text=element_text(size=0,color='black'),
        axis.title=element_text(size=0,color='black')) + theme(plot.title = element_text(size = 0, vjust=0)) +
  theme(strip.text.x = element_text(size = 0))  +
  annotate("text", label = "Lung SCC\nhigh-plasticity markers", 
           x = -5, y = 7, vjust = 0,size=8) +
  theme(legend.position=c(0.8, 0.2),
        legend.key.size = unit(1, 'cm'),
        legend.text = element_text(size=20))

png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Figs/Fig4/fig4_eigengene_manifolds/","cluster5_lung_jaks",".png",sep=""),
    width=7,height=7,units="in",res=300)
print(cluster5)
dev.off()

#-------------------------------------------------------#
#-------------------------------------------------------#
#-------------------------------------------------------#
#Extended data fig 7
#Subtractive analysis for an eigengene
#-------------------------------------------------------#
seed_gene="Trp53"
i=1
#Get the correct match_index
match_index = match(seed_gene[i], aim2probedat$SYMBOL) #match seed gene to its place in the matrix

#---------------------------------------#
#Normal skin eigengene
symbol_match.nsk = data.frame(aim2probedat$SYMBOL, rhodf[,match_index], p_df[,match_index]) #put together gene names, expression data, and p-vals
names(symbol_match.nsk) = c("gene","rho","p")
symbol_match.nsk = symbol_match.nsk[order(-symbol_match.nsk$rho),] #order from high to low gene values

#---------------------------------------#
#Tumor eigengene
symbol_match.scc = data.frame(aim2probedat$SYMBOL, tumor_rhodf[,match_index], tumor_p_df[,match_index]) #put together gene names, expression data, and p-vals
names(symbol_match.scc) = c("gene","rho","p")
symbol_match.scc = symbol_match.scc[order(-symbol_match.scc$rho),] #order from high to low gene values

#---------------------------------------#
#Subtractive analysis nsk vs scc
nsk_subtractive = setdiff(symbol_match.nsk$gene[1:eigengene_size],symbol_match.scc$gene[1:eigengene_size])
scc_subtractive = setdiff(symbol_match.scc$gene[1:eigengene_size],symbol_match.nsk$gene[1:eigengene_size])

#------------------------------#
#4a. Seed gene
#------------------------------#
#all cells eigengene
seed_gene_manifold = plot_cells(big_cds,
                                genes=seed_gene[i],
                                label_cell_groups=FALSE,
                                label_leaves=F,
                                label_branch_points=F,
                                graph_label_size=1.5,
                                show_trajectory_graph=F) +
  xlab("") + ylab("")  +
  theme(axis.ticks = element_blank()) +
  scale_color_viridis_c(option = "inferno",name="") +
  theme(axis.text=element_text(size=0,color='black'),
        axis.title=element_text(size=0,color='black')) + theme(plot.title = element_text(size = 0, vjust=0)) +
  theme(strip.text.x = element_text(size = 0)) +
  annotate("text", label = paste(seed_gene[i],"\nseed gene",sep=""), 
           x = -6, y = 7, vjust = 0,size=10) +
  theme(legend.position=c(0.8, 0.2),
        legend.key.size = unit(1, 'cm'),
        legend.text = element_text(size=20))


#------------------------------#
#4b. Normal skin WT: subtractive eigengene
#------------------------------#
marker_df = data.frame(unique(alias2SymbolTable(nsk_subtractive, species = "Mm")),paste(seed_gene[i]," Normal skin",sep=""),paste(seed_gene[i]," Normal skin",sep=""))
names(marker_df) = c("gene","cell_type","module")
marker_df = na.omit(marker_df)

#all cells eigengene
agexp_plot.12 = plot_cells(big_cds,
                           genes=marker_df,
                           label_cell_groups=FALSE,
                           label_leaves=F,
                           label_branch_points=F,
                           graph_label_size=1.5,
                           show_trajectory_graph=F) +
  xlab("") + ylab("") + 
  theme(axis.ticks = element_blank()) +
  scale_color_viridis_c(option = "inferno",name="") +
  theme(axis.text=element_text(size=0,color='black'),
        axis.title=element_text(size=0,color='black')) + theme(plot.title = element_text(size = 0, vjust=0)) +
  theme(strip.text.x = element_text(size = 0))  +
  annotate("text", label = paste(seed_gene[i],"\nNSk subtractive\neigengene",sep=""), 
           x = -5, y = 7, vjust = 0,size=10) +
  theme(legend.position=c(0.8, 0.2),
        legend.key.size = unit(1, 'cm'),
        legend.text = element_text(size=20))

#------------------------------#
#4c. SCC: subtractive eigengene
#------------------------------#
#tumor
#make marker df for grahing eigengene
marker_df = data.frame(unique(alias2SymbolTable(scc_subtractive, species = "Mm")),paste(seed_gene[i],"Carcinoma",sep=""),paste(seed_gene[i],"Carcinoma",sep=""))
names(marker_df) = c("gene","cell_type","module")
marker_df = na.omit(marker_df)

#all cells eigengene
agexp_plot_tumor.12 = plot_cells(big_cds,
                                 genes=marker_df,
                                 label_cell_groups=FALSE,
                                 label_leaves=F,
                                 label_branch_points=F,
                                 graph_label_size=1.5,
                                 show_trajectory_graph=F) +
  xlab("") + ylab("") + 
  theme(axis.ticks = element_blank()) +
  scale_color_viridis_c(option = "inferno",name="") +
  theme(axis.text=element_text(size=0,color='black'),
        axis.title=element_text(size=0,color='black')) + theme(plot.title = element_text(size = 0, vjust=0)) +
  theme(strip.text.x = element_text(size = 0))  +
  annotate("text", label = paste(seed_gene[i],"\nSCC subtractive\neigengene",sep=""), 
           x = -5, y = 7, vjust = 0,size=10) +
  theme(legend.position=c(0.8, 0.2),
        legend.key.size = unit(1, 'cm'),
        legend.text = element_text(size=20))


#-----------#
#print out eigengene plots
png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Extended\ Data/Extended\ Data\ Fig\ 7/",seed_gene[i],".png",sep=""),
    width=21,height=7,units="in",res=300)

print(multiplot(seed_gene_manifold, agexp_plot.12, agexp_plot_tumor.12,
                cols=3))
dev.off()

##################################################################################################################################
##################################################################################################################################
##################################################################################################################################
##################################################################################################################################
#Fig. 4 - Loess curves at the single cell level for LI, spindle path, and tumor suppressors
#1. LI exemplar eigengene: Klf5
#2. Spindle path exeplar: Bmi1
#3. Tumor suppressors: Cdkn2a, Cdkn2b, Trp53

seed_gene = c("Klf5","Bmi1","Cdkn2a","Cdkn2b","Trp53","Tigit","Slc4a11")

clust19_33_cds = big_cds[, colData(big_cds)$cluster %in% c(19,33)] #Extract upper and lower spikes

agg_eigengene_vals = do.call(cbind, lapply(1:length(seed_gene), function(i){
  #i=1
  print(i)
  #---------------------------------------#
  #tumor
  match_index = match(seed_gene[i], aim2probedat$SYMBOL) #match seed gene to its place in the matrix
  symbol_match = data.frame(aim2probedat$SYMBOL, tumor_rhodf[,match_index], tumor_p_df[,match_index]) #put together gene names, expression data, and p-vals
  symbol_match = symbol_match[order(-symbol_match$tumor_rhodf...match_index.),] #order from high to low gene values
  names(symbol_match) = c("gene","rho","p")
  
  #make marker df for grahing eigengene
  marker_df = data.frame(unique(alias2SymbolTable(symbol_match$gene[1:(eigengene_size)], species = "Mm")),paste(seed_gene[i],"Carcinoma",sep=""),paste(seed_gene[i],"Carcinoma",sep=""))
  names(marker_df) = c("gene","cell_type","module")
  marker_df = na.omit(marker_df)
  
  #CALCULATE EIGENGENE VALUE OF ASSOCIATED DISEASED GENES AND GRAPH OUT
  #negative eigengene
  agg_mat = aggregate_gene_expression(
    clust19_33_cds,
    gene_group_df = marker_df,
    cell_group_df = NULL,
    norm_method = c("log"),
    pseudocount = 1,
    scale_agg_values = F,
    #max_agg_value = NA,
    min_agg_value = 0,
    exclude.na = TRUE
  )
  agg_mat = t(data.frame(agg_mat))
  agg_mat = data.frame(agg_mat-mean(agg_mat)) #bring average expression to zero
  names(agg_mat) = seed_gene[i]
  agg_mat
  
}))

assoc_genes = agg_eigengene_vals
#2. Add eigengene data frame, and order cells from low-to-high ZFP36L2
assoc_genes = assoc_genes[order(assoc_genes$Klf5),] #low to high Klf5 SCC eigengene expression

#3. Add ordinal column of cell order number
assoc_genes$cell_order = seq(1:nrow(assoc_genes))

#4. Melt dataframe by cell_order
dfm = melt(assoc_genes, id=c("cell_order"))

unique(dfm$variable)
seed_gene = c("Klf5","Bmi1","Cdkn2a","Cdkn2b","Trp53","Tigit","Slc4a11")
dfm$variable = gsub("Klf5","LI exemplar (Klf5)",dfm$variable)
dfm$variable = gsub("Cdkn2a","Tumor suppressor (Cdkn2a)",dfm$variable)
dfm$variable = gsub("Cdkn2b","Tumor suppressor (Cdkn2b)",dfm$variable)
dfm$variable = gsub("Bmi1","Spindle path exemplar (Bmi1)",dfm$variable)
dfm$variable = gsub("Trp53","Tumor suppressor (Trp53)",dfm$variable)
dfm$variable = gsub("Tigit","Drug resistance (Tigit)",dfm$variable)
dfm$variable = gsub("Slc4a11","Drug resistance (Slc4a11)",dfm$variable)


#5. Plot out
tumor_suppressor_stem = ggplot(dfm, aes(x=cell_order, y=value, color=as.factor(variable))) +
  stat_smooth(aes(group=variable),method="loess",linetype="solid", alpha=1, size=1.1, se=F) +
  #stat_smooth(geom='line', linetype ="dashed", alpha=1.3, size=0.7, se=FALSE) +
  #facet_wrap( ~ module, ncol=4, scales="free")+
  theme(strip.background = element_rect(colour="white", fill="white", 
                                        size=1, linetype="solid")) +
  ylab("scaled SCC eigengene expression") +
  xlab("") +
  ggtitle("") +
  theme(panel.grid.major=element_line(colour = "darkgrey"), 
        panel.grid.minor=element_blank(), 
        panel.background=element_blank()) +
  theme(legend.key = element_rect(fill = "white")) +
  theme(legend.text=element_text(size=15), legend.title=element_text(size=20)) +
  theme(axis.text=element_text(size=15,color='black'),
        axis.title=element_text(size=15),
        title=element_text(size=0),
        axis.ticks = element_line(size=0)) +
  theme(legend.position="right") +
  scale_color_manual(name="",values = c("LI exemplar (Klf5)" = "goldenrod2",
                                        "Tumor suppressor (Cdkn2a)" = "burlywood3",
                                        "Tumor suppressor (Cdkn2b)" = "peachpuff2",
                                        "Spindle path exemplar (Bmi1)" = "darkgreen",
                                        "Tumor suppressor (Trp53)" = "darkolivegreen3",
                                        "Drug resistance (Tigit)" = "darkorange",
                                        "Drug resistance (Slc4a11)" = "chocolate1"))

#6. Print out
png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Figs/Fig4/spike_stem_tumorsuppressor_withlegend.png",sep=""),
    width=6,height=4,units="in",res=200)
print(tumor_suppressor_stem)
dev.off()

png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Figs/Fig4/spike_stem_tumorsuppressor_nolegend.png",sep=""),
    width=6,height=4,units="in",res=200)
print(tumor_suppressor_stem + theme(legend.position="none"))
dev.off()


##################################################################################################################################
##################################################################################################################################
#---------------------------------------------------------------------------------------------------#
#---------------------------------------------------------------------------------------------------#
#Define papilloma cells ALONE
#Analyze upper spike and lower spike trajectories!
#---------------------------------------------------------------------------------------------------#
#Papilloma to Carcinoma: get Pap or CA but all NEOEPITHELIAL CELLS!! this is a good parenchyma definition
pap_cds = big_cds[,colData(big_cds)$tissue_updated %in% c("Pap") & colData(big_cds)$cell_type %in% "nEp"]

min_distances = seq(0.05,1,by=0.05) #This is showing a lot of interesting bridgeness with min distanes bewteen 0.4 and 0.6
#min_distances = seq(0.4,0.6,by=0.01) #This is showing a lot of interesting bridgeness with min distanes bewteen 0.4 and 0.6

pap_cds <- preprocess_cds(pap_cds, 
                          method="PCA",
                          num_dim = 50,
                          norm_method = "none", #these data have already been normalized by Seurat
                          alignment_group = NULL)


#now do it between 0.4 and 0.6

pdf(file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/figures/manifold_optimization/pap_alone_v1.pdf",height=10,width=10)
for (i in 1:length(min_distances)) {
  #i=1
  print(i)
  #loop to plot in Pap and CA in different ways
  #4b. ?reduce_dimension 
  pap_cds.test <- reduce_dimension(
    pap_cds,
    max_components = 2,
    reduction_method = c("UMAP"),
    preprocess_method = "PCA", #must now used ALIGNEMNT PRE-PROCESSING METHOD bc it replaces the other pre-process shit
    umap.metric = "cosine",
    umap.min_dist = min_distances[i],
    umap.n_neighbors = 15L,
    umap.fast_sgd = FALSE,
    umap.nn_method = "annoy",
    cores = 1,
    verbose = FALSE
  )
  
  #5. Cluster and partition ?cluster_cells cells into supergroups community detection (Phenograph algorithm) Community detection of cells as part of Phenograph algorithm https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4508757/
  #5b.  ?cluster_cells
  pap_cds.test <- cluster_cells(pap_cds.test,
                                reduction_method = c("UMAP"),
                                k = 10, #a bigger k will result in lower resolution 
                                cluster_method = c("leiden"),
                                num_iter = 2,
                                partition_qval = 0.05, #Numeric, the q-value cutoff to determine when to partition
                                weight = FALSE,
                                resolution = NULL,
                                random_seed = NULL,
                                verbose = F)
  print( 
    plot_cells(pap_cds.test, color_cells_by="integrated_type_bridge", group_cells_by="cluster",
               group_label_size = 0,
               show_trajectory_graph=F,
               x=1,
               y=2,
               alpha=0.8,
               cell_size=1) +
      ggtitle(paste("Minimal distance = ",min_distances[i],sep="")) + xlab("Dim. 1") + ylab("Dim. 2") +
      theme(text = element_text(size=18)) + 
      theme(axis.text=element_text(size=0,color='black'),
            axis.title=element_text(size=0,color='black'),
            axis.ticks = element_line(size=0)) + 
      theme(plot.title = element_text(size = 22, vjust=1)) +
      scale_color_manual(values = c("nEp" = "mediumpurple","lower spike" = "olivedrab4","upper spike" = "yellow4","squamous"="red","spindle"="darkred",
                                    "Upper Bridge"="darkorange4","Lower Bridge"="goldenrod3","Middle Bridge"="navajowhite3")) + 
      theme(legend.position="bottom")
  )
  
}
dev.off()


#--------------------------------------------------------------------------------------#
#Optimal min_ditance for the pap alone is = 0.1
pap_cds = big_cds[,colData(big_cds)$tissue_updated %in% c("Pap") & colData(big_cds)$cell_type %in% "nEp"]

pap_cds <- preprocess_cds(pap_cds, 
                          method="PCA",
                          num_dim = 50,
                          norm_method = "none", #these data have already been normalized by Seurat
                          alignment_group = NULL)

#4b. ?reduce_dimension 
pap_cds <- reduce_dimension(
  pap_cds,
  max_components = 2,
  reduction_method = c("UMAP"),
  preprocess_method = "PCA", #must now used ALIGNEMNT PRE-PROCESSING METHOD bc it replaces the other pre-process shit
  umap.metric = "cosine",
  umap.min_dist = 0.1, #this is based on optimizing the manifold to visualize bridge cells (See pap_ca_manifold_min_distances.pdf)
  umap.n_neighbors = 15L,
  umap.fast_sgd = FALSE,
  umap.nn_method = "annoy",
  cores = 1,
  verbose = FALSE
)

#5. Cluster and partition ?cluster_cells cells into supergroups community detection (Phenograph algorithm) Community detection of cells as part of Phenograph algorithm https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4508757/
#5b.  ?cluster_cells
pap_cds <- cluster_cells(pap_cds,
                         reduction_method = c("UMAP"),
                         k = 10, #a bigger k will result in lower resolution 
                         cluster_method = c("leiden"),
                         num_iter = 2,
                         partition_qval = 0.05, #Numeric, the q-value cutoff to determine when to partition
                         weight = FALSE,
                         resolution = NULL,
                         random_seed = NULL,
                         verbose = F)

#6. Learn trajectory for the optimal min-dstance graph
pap_cds <- learn_graph(
  pap_cds,
  use_partition = F, #lear graph across this all
  close_loop = TRUE,
  learn_graph_control = NULL,
  verbose = FALSE
)

#Pap alone with cells explained
png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/pap_alone_analysis/pap_alone_legend.png",sep=""),
    width=7,height=7,units="in",res=300)
pap_fig = plot_cells(pap_cds, color_cells_by="integrated_type_bridge", group_cells_by="cluster",
                     group_label_size = 0,
                     show_trajectory_graph=F,
                     x=1,
                     y=2,
                     alpha=0.7,
                     cell_size=1.2) +
  ggtitle("") + xlab("") + ylab("") +
  theme(plot.title = element_text(size=0),
        axis.text=element_text(size=0,color='black'),
        axis.title=element_text(size=0),
        axis.ticks = element_line(size=0)) +
  scale_color_manual(values = c("nEp" = "mediumpurple","lower spike" = "olivedrab4","upper spike" = "yellow4","squamous"="red","spindle"="darkred",
                                "nEp"="lightgrey",
                                "Upper Bridge"="darkorange4","Lower Bridge"="goldenrod3","Middle Bridge"="navajowhite3")) + theme(legend.position="bottom")
print(pap_fig)
dev.off()

#Pap alone trajectory
png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/pap_alone_analysis/pap_alone_trajectory.png",sep=""),
    width=7,height=7,units="in",res=300)
pap_fig = plot_cells(pap_cds, color_cells_by="integrated_type_bridge", group_cells_by="cluster",
                     group_label_size = 0,
                     show_trajectory_graph=T,
                     x=1,
                     y=2,
                     alpha=0.2,
                     cell_size=1.2,
                     label_branch_points = T,
                     label_roots = T,
                     label_leaves = T,
                     graph_label_size=3) +
  ggtitle("") + xlab("") + ylab("") +
  theme(plot.title = element_text(size=0),
        axis.text=element_text(size=0,color='black'),
        axis.title=element_text(size=0),
        axis.ticks = element_line(size=0)) +
  scale_color_manual(values = c("nEp" = "mediumpurple","lower spike" = "olivedrab4","upper spike" = "yellow4","squamous"="red","spindle"="darkred",
                                "nEp"="lightgrey",
                                "Upper Bridge"="darkorange4","Lower Bridge"="goldenrod3","Middle Bridge"="navajowhite3")) + theme(legend.position="none")
print(pap_fig)
dev.off()

#Pap alone unsupervised clusters
png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/pap_alone_analysis/pap_alone_unsupervised_clusters.png",sep=""),
    width=7,height=7,units="in",res=300)
pap_fig = plot_cells(pap_cds, color_cells_by="cluster", group_cells_by="cluster",
                     group_label_size = 7,
                     show_trajectory_graph=F,
                     x=1,
                     y=2,
                     alpha=0.2,
                     cell_size=1.2,
                     label_branch_points = T,
                     label_roots = T,
                     label_leaves = T,
                     graph_label_size=3) +
  ggtitle("") + xlab("") + ylab("") +
  theme(plot.title = element_text(size=0),
        axis.text=element_text(size=0,color='black'),
        axis.title=element_text(size=0),
        axis.ticks = element_line(size=0)) +
  theme(legend.position="none")
print(pap_fig)
dev.off()


#---------------------------------------------------------------------------------------#
#plot a nice upper-spike marker eigengeene
match("BANF1_CAR_WT",names(balmain_genes))
i=197
print(paste(i,names(balmain_genes)[i],sep=" "))

print(i)
#only if more than 1 constituent gene
marker_df = data.frame(na.omit(balmain_genes[,i]),names(balmain_genes)[i])
names(marker_df) = c("gene","cell_type")

#remake the marker df dataframe with the matched gene names
marker_df = data.frame(unique(alias2SymbolTable(unique(marker_df$gene), species = "Mm")),names(balmain_genes)[i],names(balmain_genes)[i])
names(marker_df) = c("gene","cell_type","module")
marker_df = na.omit(marker_df) #remove non-matching genes


#Pap alone trajectory
png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/pap_alone_analysis/pap_upperspike_eigengene.png",sep=""),
    width=14,height=7,units="in",res=300)

allcells_fig = plot_cells(big_cds,
                          genes=marker_df,
                          label_cell_groups=FALSE,
                          label_leaves=F,
                          label_branch_points=F,
                          graph_label_size=1.5,
                          show_trajectory_graph=F) +
  theme(axis.text=element_text(size=0,color='black'),
        axis.title=element_text(size=0,color='black')) + 
  theme(axis.ticks = element_blank()) +
  theme(plot.title = element_text(size = 15, vjust=0)) +
  theme(legend.position="bottom") +
  scale_color_viridis_c(option = "inferno",name="integrated expression") +
  ggtitle("All cells BANF1_CAR_WT")+ #continuous
  theme(strip.text.x = element_text(size = 0)) #remove facet label

pap_fig = plot_cells(pap_cds,
                     genes=marker_df,
                     label_cell_groups=FALSE,
                     label_leaves=F,
                     label_branch_points=F,
                     graph_label_size=1.5,
                     show_trajectory_graph=F,
                     cell_size = 1) +
  theme(axis.text=element_text(size=0,color='black'),
        axis.title=element_text(size=0,color='black')) + 
  theme(axis.ticks = element_blank()) +
  theme(plot.title = element_text(size = 15, vjust=0)) +
  theme(legend.position="bottom") +
  scale_color_viridis_c(option = "inferno",name="integrated expression") +
  ggtitle("Papilloma BANF1_CAR_WT")+ #continuous
  theme(strip.text.x = element_text(size = 0)) #remove facet label

print(multiplot(allcells_fig,pap_fig,cols=2))
dev.off()

#Wnt
row.names(pap_cds)[grep("Wnt",row.names(pap_cds))]
wnt_ligands = plot_cells(pap_cds,
                         genes=row.names(pap_cds)[grep("Wnt",row.names(pap_cds))],
                         label_cell_groups=FALSE,
                         label_leaves=F,
                         label_branch_points=F,
                         graph_label_size=1.5,
                         show_trajectory_graph=F,
                         cell_size = 1) +
  theme(axis.text=element_text(size=0,color='black'),
        axis.title=element_text(size=0,color='black')) + 
  theme(axis.ticks = element_blank()) +
  theme(plot.title = element_text(size = 20, vjust=0)) +
  theme(legend.position="bottom") +
  scale_color_viridis_c(option = "inferno",name="integrated expression") +
  ggtitle("Wnt ligands")+ #continuous
  theme(strip.text.x = element_text(size = 15)) #remove facet label

#Wnt receptors
fzd_wntreceptors = plot_cells(pap_cds,
                              genes=row.names(pap_cds)[grep("Fzd",row.names(pap_cds))],
                              label_cell_groups=FALSE,
                              label_leaves=F,
                              label_branch_points=F,
                              graph_label_size=1.5,
                              show_trajectory_graph=F,
                              cell_size = 1) +
  theme(axis.text=element_text(size=0,color='black'),
        axis.title=element_text(size=0,color='black')) + 
  theme(axis.ticks = element_blank()) +
  theme(plot.title = element_text(size = 15, vjust=0)) +
  theme(legend.position="bottom") +
  scale_color_viridis_c(option = "inferno",name="integrated expression") +
  ggtitle("Wnt receptors: Frizzled proteins")+ #continuous
  theme(strip.text.x = element_text(size = 15)) #remove facet label

lrp_wntreceptors = plot_cells(pap_cds,
                              genes=row.names(pap_cds)[grep("Lrp",row.names(pap_cds))],
                              label_cell_groups=FALSE,
                              label_leaves=F,
                              label_branch_points=F,
                              graph_label_size=1.5,
                              show_trajectory_graph=F,
                              cell_size = 1) +
  theme(axis.text=element_text(size=0,color='black'),
        axis.title=element_text(size=0,color='black')) + 
  theme(axis.ticks = element_blank()) +
  theme(plot.title = element_text(size = 20, vjust=0)) +
  theme(legend.position="bottom") +
  scale_color_viridis_c(option = "inferno",name="integrated expression") +
  ggtitle("Wnt receptors: Lrp")+ #continuous
  theme(strip.text.x = element_text(size = 15)) #remove facet label

grep("Nsun2",row.names(big_cds))


png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/pap_alone_analysis/DEG_results/wnt_ligands.png",sep=""),
    width=14,height=14,units="in",res=300)
print(wnt_ligands)
dev.off()

png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/pap_alone_analysis/DEG_results/fzd_wntreceptors.png",sep=""),
    width=14,height=14,units="in",res=300)
print(fzd_wntreceptors)
dev.off()

png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/pap_alone_analysis/DEG_results/lrp_wntreceptors.png",sep=""),
    width=14,height=14,units="in",res=300)
print(lrp_wntreceptors)
dev.off()


#--------------------------------------------------------------------------------------#
#--------------------------------------------------------------------------------------#
#-----------------------------------------------------------#
#Pap alone trajectory analysis
cds_sub <- choose_cells(pap_cds)
subset_pr_test_res <- graph_test(cds_sub, neighbor_graph="principal_graph", cores=4) #perform trajectory analysis
re_order_gene_changes = data.frame(subset_pr_test_res[order(subset_pr_test_res$q_value),]) #re-order single gene results from high to low
graph_test_df_comb.reord = re_order_gene_changes[,c("gene_short_name","morans_I","morans_test_statistic","p_value","q_value")] #re-order columns
write.table(graph_test_df_comb.reord, file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/pap_alone_analysis/trajecotry\ results/pap_alone_unsupervised_trajecotry.txt",sep="\t",row.names=F,quote=F)
write.table(data.frame(colnames(cds_sub)), file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/figures/pap_ca/pap_ca_trajectory_segment/upper_bridge_cell_names_v1.txt",sep="\t",row.names=F,quote=F)

#--------------------------------------------------------------------------------------#
#-----------------------------------------------------------#

names(colData(big_cds))
table(colData(big_cds)$partition)
round((table(colData(big_cds)$partition) / nrow(colData(big_cds)) )*100,2)
#--------------------------------------------------------------------------------------#
#--------------------------------------------------------------------------------------#
#--------------------------------------------------------------------------------------#
#--------------------------------------------------------------------------------------#
#--------------------------------------------------------------------------------------#
#-----------------------------------------------------------#
#Function to test for DEGs between 2 clusters of cells

#Here are the things you can test bitch
unique(colData(big_cds)$integrated_celltype_spike)


test_group_1 = c("upper spike","upper spike","upper spike","upper spike","lower spike",
                 "squamous","PrE","UHF","UHF")
test_group_2 = c("spindle","squamous","lower spike","nEp","nEp",
                 "spindle","upper spike","upper spike","lower spike")

for (i in 2:length(test_group_2)) {
  
  tryCatch({ #suppress error
    #i=1
    pap_cds = big_cds[,colData(big_cds)$integrated_celltype_spike %in% c(test_group_1[i],test_group_2[i])] #subset cds object to variable genes
    
    
    minimum_cells = round(dim(pap_cds)[2]*0.05)
    #now create Seurat object from the patient-condition specific object
    seurat_obj = CreateSeuratObject(
      pap_cds@assays@data$counts,
      project = "CreateSeuratObject",
      assay = "RNA",
      names.field = 1,
      names.delim = "_",
      meta.data = NULL,
      min.cells = minimum_cells)
    
    #FindVariableFeatures
    seurat_obj <- FindVariableFeatures(seurat_obj, selection.method = "vst", nfeatures = 2000)
    variable_genes <- VariableFeatures(seurat_obj) #get variable gene names
    
    to_test_cds = pap_cds[rowData(pap_cds)$gene_short_name %in% variable_genes,] #subset cds object to variable genes
    
    meandiffs = do.call(rbind, lapply(1:length(variable_genes), function(e){
      # meandiffs = do.call(rbind, lapply(1:10, function(e){
      
      tryCatch({ #suppress error
        #e=1
        
        print(paste("i=",i," e=",e,sep=""))
        
        #test group 2
        ca_data = to_test_cds[rowData(to_test_cds)$gene_short_name %in% variable_genes[e],colData(to_test_cds)$integrated_celltype_spike %in% test_group_2[i]]@assays@data$counts
        ca_mean = mean(ca_data)
        ca_cov = (sd(ca_data) / ca_mean )*100
        ca_percent_expressed = (length(ca_data[ca_data>0]) /  length(ca_data))*100
        
        #test group 1
        normaskin_data = to_test_cds[rowData(to_test_cds)$gene_short_name %in% variable_genes[e],colData(to_test_cds)$integrated_celltype_spike %in% test_group_1[i]]@assays@data$counts
        normakin_mean = mean(normaskin_data)
        normaskin_cov = (sd(normaskin_data) / normakin_mean )*100
        normaskin_percent_expressed = (length(normaskin_data[normaskin_data>0]) /  length(normaskin_data))*100
        
        #differences
        mean_abs_diff = ca_mean - normakin_mean
        mean_per_diff = ((ca_mean - normakin_mean)/normakin_mean)*100
        fc = ca_mean/normakin_mean
        
        #test expression diffs: make dataframe with tissue and p16 status
        ca_dat = data.frame(as.numeric(ca_data), test_group_2[i])
        normaskin_dat = data.frame(as.numeric(normaskin_data), test_group_1[i])
        names(ca_dat) = c("expression","tissue")
        names(normaskin_dat) = c("expression","tissue")
        agg_matdf = rbind(normaskin_dat,ca_dat)
        names(agg_matdf) = c("expression","tissue")
        
        #scale everything up to 0 for negative binormail test
        rescale_value = 0-min(agg_matdf$expression)
        agg_matdf$expression = agg_matdf$expression+rescale_value
        
        #fit negative binomial GLM to eigengene testing for amount-tissue differences
        glmnb_res <- glm.nb(expression~tissue, data = agg_matdf)
        
        #return df not testing for p16 status
        standard_result_df = data.frame(variable_genes[e],
                                        ca_mean, ca_cov, ca_percent_expressed, 
                                        normakin_mean, normaskin_cov, normaskin_percent_expressed,
                                        mean_abs_diff, mean_per_diff, fc,
                                        t(summary(glmnb_res)$coefficients[2,])
        )
        standard_result_df$variable_tested = "tissue"
        names(standard_result_df)[1:7] = c("gene",
                                           paste(test_group_2[i],"_mean",sep=""),  paste(test_group_2[i],"_cov",sep=""),  paste(test_group_2[i],"_percent_expressed",sep=""),
                                           paste(test_group_1[i],"_mean",sep=""),  paste(test_group_1[i],"_cov",sep=""),  paste(test_group_1[i],"_percent_expressed",sep="")
        )
        
        return(standard_result_df) #if there is p16 variation return both results
        
      },  error=function(e){})
    }))
    
    #re-order from high to low percent change
    meandiffs = meandiffs[order(-meandiffs$mean_abs_diff),]
    
    #the infinity values mean that normal_skin is at 0; re-order within the infitnite group based on absolute mean diffchanges
    infdiffs = meandiffs[is.infinite(meandiffs$mean_per_diff),] #extract inifity values
    infdiffs = infdiffs[order(-infdiffs$mean_abs_diff),] #re-order absolute differences high to low
    
    #remove infinite values and then add the re-ordered inf_diffs
    no_inf_diffs <- meandiffs[!is.infinite(meandiffs$mean_per_diff),]
    meandiffs = rbind(infdiffs,no_inf_diffs)
    
    names(meandiffs)[1:7] = c("gene",
                              paste(test_group_2[i],"_mean",sep=""),  paste(test_group_2[i],"_cov",sep=""),  paste(test_group_2[i],"_percent_expressed",sep=""),
                              paste(test_group_1[i],"_mean",sep=""),  paste(test_group_1[i],"_cov",sep=""),  paste(test_group_1[i],"_percent_expressed",sep="")
    )
    
    #write out results
    write.table(meandiffs, 
                file=paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/DEG/selected_cell_clusters/",
                           test_group_1[i],"_",test_group_2[i],
                           "_degs.txt",sep=""),
                quote=F,sep="\t",row.names=F)
    
    #--------------------------------------------#
    #GO analysis on top DEGs
    
    #create topGO object: top 100 POSITIVE DEGs
    geneUniverse <- row.names(big_cds)
    genesOfInterest = meandiffs$gene[1:100]
    geneList <- factor(as.integer(geneUniverse %in% genesOfInterest))
    names(geneList) <- geneUniverse
    GOdata <- new("topGOdata",
                  ontology = "BP",
                  allGenes=geneList,
                  #geneSel = selection,
                  annot = annFUN.org, 
                  mapping = "org.Mm.eg.db",
                  ID = "symbol")
    
    resultKS <- runTest(GOdata, algorithm = "weight01", statistic = "fisher") #Fisher exact test
    tab <- GenTable(GOdata, raw.p.value = resultKS, topNodes = length(resultKS@score), numChar = 120)
    tab$comparison = paste(test_group_2[i]," vs ", test_group_1[i],sep="")
    tab = subset(tab, raw.p.value <=0.01) #keep only sig results 
    
    write.table(tab, 
                file=paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/DEG/selected_cell_clusters/",
                           test_group_1[i],"_",test_group_2[i],
                           "_top100deg_GOterms.txt",sep=""),
                quote=F,sep="\t",row.names=F)
    
  },  error=function(e){})
  
} #End outer loop teesting for main DEGs between groups


#--------------------------------------------------------------------------------------#
#--------------------------------------------------------------------------------------#
#--------------------------------------------------------------------------------------#
#--------------------------------------------------------------------------------------#

#PAPILLOMA-CARCINOMA MANIFOLD WITH EIGENGENES
#pap_ca.test = loop through the eigengenes
eigengene_agg = do.call(cbind, lapply(446:ncol(balmain_genes), function(i){
  #subset eigengene_agg = do.call(cbind, lapply(106:118, function(i){
  tryCatch({ #suppress error
    
    #i=1
    print(i)
    #only if more than 1 constituent gene
    marker_df = data.frame(na.omit(balmain_genes[,i]),names(balmain_genes)[i])
    names(marker_df) = c("gene","cell_type")
    
    #remake the marker df dataframe with the matched gene names
    marker_df = data.frame(alias2SymbolTable(unique(marker_df$gene), species = "Mm"),names(balmain_genes)[i],names(balmain_genes)[i])
    names(marker_df) = c("gene","cell_type","module")
    marker_df = na.omit(marker_df) #remove non-matching genes
    
    #All cells seed genee
    all_cells_seed = plot_cells(big_cds,
                                genes=marker_df$gene[1],
                                label_cell_groups=FALSE,
                                label_leaves=F,
                                label_branch_points=F,
                                graph_label_size=1.5,
                                show_trajectory_graph=F) +
      xlab("") + ylab("") + 
      ggtitle(paste("All cells: ",names(balmain_genes)[i]," Seed",sep="")) +
      theme(axis.ticks = element_blank()) +
      scale_color_viridis_c(option = "inferno",name="") +
      theme(axis.text=element_text(size=0,color='black'),
            axis.title=element_text(size=0,color='black')) + theme(plot.title = element_text(size = 15, vjust=0)) +
      theme(strip.text.x = element_text(size = 0)) +
      theme(legend.position="none")
    
    #All cells eigengene
    all_cells_eigengene = plot_cells(big_cds,
                                     genes=marker_df,
                                     label_cell_groups=FALSE,
                                     label_leaves=F,
                                     label_branch_points=F,
                                     graph_label_size=1.5,
                                     show_trajectory_graph=F) +
      xlab("") + ylab("") + 
      ggtitle(paste("All cells: ",names(balmain_genes)[i]," Eigengene",sep="")) +
      theme(axis.ticks = element_blank()) +
      scale_color_viridis_c(option = "inferno",name="") +
      theme(axis.text=element_text(size=0,color='black'),
            axis.title=element_text(size=0,color='black')) + theme(plot.title = element_text(size = 15, vjust=0)) +
      theme(strip.text.x = element_text(size = 0)) +
      theme(legend.position="none")
    
    #Pap_ca: Seed gene
    pap_ca_seed = plot_cells(pap_ca,
                             genes=marker_df$gene[1], #seed gene is the first row of the markers file
                             label_cell_groups=FALSE,
                             label_leaves=F,
                             label_branch_points=F,
                             graph_label_size=1.2,
                             show_trajectory_graph=F,
                             cell_size=1,
                             alpha=0.8) +
      xlab("") + ylab("") + 
      ggtitle(paste("Neoplasia parenchyma: ",names(balmain_genes)[i]," Seed",sep="")) +
      theme(axis.ticks = element_blank()) +
      scale_color_viridis_c(option = "inferno",name="") +
      theme(axis.text=element_text(size=0,color='black'),
            axis.title=element_text(size=0,color='black')) + theme(plot.title = element_text(size = 15, vjust=0)) +
      theme(strip.text.x = element_text(size = 0)) +
      theme(legend.position="none")
    
    #Pap_ca eigengene
    pap_ca_eigengene = plot_cells(pap_ca,
                                  genes=marker_df,
                                  label_cell_groups=FALSE,
                                  label_leaves=F,
                                  label_branch_points=F,
                                  graph_label_size=1.5,
                                  show_trajectory_graph=F,
                                  cell_size=1,
                                  alpha=0.8) +
      xlab("") + ylab("") + 
      ggtitle(paste("Neoplalsia parenchyma ",names(balmain_genes)[i]," Eigengene",sep="")) +
      theme(axis.ticks = element_blank()) +
      scale_color_viridis_c(option = "inferno",name="") +
      theme(axis.text=element_text(size=0,color='black'),
            axis.title=element_text(size=0,color='black')) + theme(plot.title = element_text(size = 15, vjust=0)) +
      theme(strip.text.x = element_text(size = 0)) +
      theme(legend.position="none")
    
    #-----------#
    #print out eigengene plots
    png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/figures/pap_ca/pap_ca_eigengene/",names(balmain_genes)[i],".png",sep=""),
        width=14,height=14,units="in",res=200)
    
    print(multiplot(all_cells_seed, all_cells_eigengene, 
                    pap_ca_seed, pap_ca_eigengene,
                    cols=2))
    dev.off()
    
  },  error=function(e){})
  
}))


#-----------------------------------------------------------#
#-----------------------------------------------------------#
#-----------------------------------------------------------#
#Upper bridge
cds_sub <- choose_cells(pap_ca)
subset_pr_test_res <- graph_test(cds_sub, neighbor_graph="principal_graph", cores=4) #perform trajectory analysis
re_order_gene_changes = data.frame(subset_pr_test_res[order(subset_pr_test_res$q_value),]) #re-order single gene results from high to low
graph_test_df_comb.reord = re_order_gene_changes[,c("gene_short_name","morans_I","morans_test_statistic","p_value","q_value")] #re-order columns
write.table(graph_test_df_comb.reord, file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/figures/pap_ca/pap_ca_trajectory_segment/upper_bridge_unsupervised_trajecotry.txt",sep="\t",row.names=F,quote=F)
write.table(data.frame(colnames(cds_sub)), file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/figures/pap_ca/pap_ca_trajectory_segment/upper_bridge_cell_names_v1.txt",sep="\t",row.names=F,quote=F)

#-----------------------------------------------------------#
#-----------------------------------------------------------#
#-----------------------------------------------------------#
#-----------------------------------------------------------#
#Middle bridge
cds_sub <- choose_cells(pap_ca)
subset_pr_test_res <- graph_test(cds_sub, neighbor_graph="principal_graph", cores=4) #perform trajectory analysis
pr_deg_ids <- row.names(subset(subset_pr_test_res, q_value < 0.05)) #extract only sig genes by Moran I test
gene_module_df <- data.frame(find_gene_modules(cds_sub[pr_deg_ids,], resolution=0.001)) #collapse genes to unsupervised modules
re_order_gene_changes = data.frame(subset_pr_test_res[order(subset_pr_test_res$q_value),]) #re-order single gene results from high to low
graph_test_df_comb = data.frame(re_order_gene_changes, gene_module_df[match(re_order_gene_changes$gene_short_name, gene_module_df$id),])    #combine with module
graph_test_df_comb.reord = graph_test_df_comb[,c(5,8,9,10,11,4,3,2,6)] #re-order columns
write.table(graph_test_df_comb.reord, file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/figures/pap_ca/pap_ca_trajectory_segment/middle_bridge_unsupervised_trajecotry.txt",sep="\t",row.names=F,quote=F)
write.table(data.frame(colnames(cds_sub)), file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/figures/pap_ca/pap_ca_trajectory_segment/middle_bridge_cell_names_v1.txt",sep="\t",row.names=F,quote=F)

#-----------------------------------------------------------#
#-----------------------------------------------------------#
#-----------------------------------------------------------#
#-----------------------------------------------------------#
#Lower bridge
cds_sub <- choose_cells(pap_ca)
subset_pr_test_res <- graph_test(cds_sub, neighbor_graph="principal_graph", cores=4) #perform trajectory analysis
pr_deg_ids <- row.names(subset(subset_pr_test_res, q_value < 0.05)) #extract only sig genes by Moran I test
gene_module_df <- data.frame(find_gene_modules(cds_sub[pr_deg_ids,], resolution=0.001)) #collapse genes to unsupervised modules
re_order_gene_changes = data.frame(subset_pr_test_res[order(subset_pr_test_res$q_value),]) #re-order single gene results from high to low
graph_test_df_comb = data.frame(re_order_gene_changes, gene_module_df[match(re_order_gene_changes$gene_short_name, gene_module_df$id),])    #combine with module
graph_test_df_comb.reord = graph_test_df_comb[,c(5,8,9,10,11,4,3,2,6)] #re-order columns
write.table(graph_test_df_comb.reord, file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/figures/pap_ca/pap_ca_trajectory_segment/lower_bridge_unsupervised_trajecotry.txt",sep="\t",row.names=F,quote=F)
write.table(data.frame(colnames(cds_sub)), file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/figures/pap_ca/pap_ca_trajectory_segment/lower_bridge_cell_names_v1.txt",sep="\t",row.names=F,quote=F)

#-----------------------------------------------------------#
#-----------------------------------------------------------#
#-----------------------------------------------------------#
#-----------------------------------------------------------#
#-----------------------------------------------------------#

#-----------------------------------------------------------#
#do GO analysis on bridge

top_marker_files = list.files("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/figures/pap_ca/pap_ca_trajectory_segment",full.names = T,pattern="unsupervised_trajecotry")
top_marker_files_short_name = list.files("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/figures/pap_ca/pap_ca_trajectory_segment",full.names = F,pattern="unsupervised_trajecotry")
num_markers = 100 #decide how many marker I wanan consider

combined_go_res = do.call(rbind, lapply(1:length(top_marker_files), function(i){
  
  #i=1
  print(i)
  sub_marker_file = read.delim(file=top_marker_files[i],header=T)
  
  #create topGO object
  geneUniverse <- row.names(big_cds)
  genesOfInterest = sub_marker_file$gene_short_name[1:num_markers]
  geneList <- factor(as.integer(geneUniverse %in% genesOfInterest))
  names(geneList) <- geneUniverse
  GOdata <- new("topGOdata",
                ontology = "BP",
                allGenes=geneList,
                #geneSel = selection,
                annot = annFUN.org, 
                mapping = "org.Mm.eg.db",
                ID = "symbol")
  
  resultKS <- runTest(GOdata, algorithm = "weight01", statistic = "fisher") #Fisher exact test
  tab <- GenTable(GOdata, raw.p.value = resultKS, topNodes = length(resultKS@score), numChar = 120)
  tab$marker_file = top_marker_files_short_name[i]
  tab = subset(tab, raw.p.value <=0.01) #keep only sig results 
  
  pdf(file=paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/figures/pap_ca/pap_ca_trajectory_segment/GO_figs/",top_marker_files_short_name[i],"_GO.pdf",sep=""),width=12,height=8)
  print(showSigOfNodes(GOdata, score(resultKS), firstSigNodes = 10, useInfo = "def"))
  print(showSigOfNodes(GOdata, score(resultKS), firstSigNodes = 5, useInfo = "def"))
  dev.off()
  
  write.table(tab, file=paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/figures/pap_ca/pap_ca_trajectory_segment/GO_enrichment/",top_marker_files_short_name[i],"_GO_enrichment.txt",sep=""),
              quote=F, row.names=F,sep="\t")
  
  return(tab)
  
})) #end topGO loop plz

write.table(combined_go_res, file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/figures/pap_ca/pap_ca_trajectory_segment/combined_bridge_GO_analysis.txt",
            quote=F, row.names=F,sep="\t")


#---------------------------------------------------------------------------------------------------------------------------------------------------# 
#---------------------------------------------------------------------------------------------------------------------------------------------------# 
#plot combined GO analysis

#get GO results files
#combine topGO results from DEG and plot
go_comb = read.delim(file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/figures/pap_ca/pap_ca_trajectory_segment/combined_bridge_GO_analysis.txt", 
                     header=T)

#Relative enrichment: Significant / Expected
go_comb$enrichment_metric = go_comb$Significant / go_comb$Expected
go_comb$test = go_comb$marker_file
tests = unique(go_comb$marker_file)
go_comb$test = gsub("lower_bridge_unsupervised_trajecotry.txt","Lower bridge",go_comb$test)
go_comb$test = gsub("middle_bridge_unsupervised_trajecotry.txt","Middle bridge",go_comb$test)
go_comb$test = gsub("upper_bridge_unsupervised_trajecotry.txt","Upper bridge",go_comb$test)


tests = c("Upper bridge","Middle bridge","Lower bridge")
#get top 10 GO terms for each DEG comparison
highest_go = do.call(rbind, lapply(1:length(tests), function(p) {
  #p=1
  degind = subset(go_comb, test==tests[p])
  degind[c(1:10),]
}))
nrow(highest_go)

#get highest GO terms for each comparison
go_comp_red = do.call(rbind, lapply(1:nrow(highest_go), function(i) {
  subset(go_comb, Term==highest_go$Term[i])
}))
nrow(go_comp_red)

#get only comparison I care about
go_comp_red.sub = go_comp_red
dcasted = dcast(data = go_comp_red.sub,formula = Term~test,fun.aggregate = sum,value.var = "enrichment_metric")
dcasted = dcasted[match(highest_go$Term,dcasted$Term),]
row.names(dcasted) = dcasted$Term
dcasted = dcasted[,-1] #drop row col
dcasted = scale(as.matrix(dcasted))
#dcasted = -log10(as.matrix(dcasted))

greens_pal = brewer.pal(12,"Greens")

redblue(100)
greenWhiteRed(100)

#Now plot out
png("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Figs/Fig3/pap_ca_bridge_GO.png",
    width=10,height=10,units="in",res=200)

heatmap.2(dcasted,col=greenWhiteRed(100),
          trace = "none", density.info = "none",key.xlab="-log10(p)",
          main = "Top GO terms",
          key=T,
          cexCol = 1,
          Rowv=F,
          Colv=F,
          keysize=1,
          cexRow = 1.3,
          margins=c(12,40))
dev.off()

#----------------------------------------------------------------------------------------------------------#
#----------------------------------------------------------------------------------------------------------#
#----------------------------------------------------------------------------------------------------------#
#-----------------------------------------------------------#
#-------------------------------------------------------------------------------------------#
#-------------------------------------------------------------------------------------------#
#-------------------------------------------------------------------------------------------#
#-------------------------------------------------------------------------------------------#
#DEG for all my data: normal skin -> Pap -> untreated carcinoma -> cisplatin treated carcinoma

colData(tumor_big_cds)$united_control_carcinoma = colData(tumor_big_cds)$tissue_updated #add new comparison variable
colData(tumor_big_cds)$united_control_carcinoma = gsub("CA","Control",colData(tumor_big_cds)$united_control_carcinoma) #convert original carcinoma name to Control
colData(tumor_big_cds)$p16_status = ifelse(colData(tumor_big_cds)$sample_name %in% c("EK-3205_01","EK-3205_02","EK-3205_04"),
                                           "p16_negative","p16_positive") #add p16 status
table(colData(tumor_big_cds)$p16_status) #get numbers of p16 status
table(colData(tumor_big_cds)$united_control_carcinoma)

#make new comparison columns
first_compare = c("Normal skin","Pap","Control")
colpal_transitionnetwork.1 = c("lightblue","blue","cyan","green", "green","purple")

second_compare = c("Pap","Control","Cisplatin")
colpal_transitionnetwork.2 = c("blue","cyan","green","red","purple","red")

#-------------------------------------------------------------------------------------------#
#Start DEG loop 

for(v in 1:length(first_compare)) {
  tryCatch({ #suppress error
    #v=2
    
    #Subset
    celltype_cds = tumor_big_cds[,colData(tumor_big_cds)$united_control_carcinoma %in% c(first_compare[v],second_compare[v])]
    
    #now create Seurat object from the patient-condition specific object
    seurat_obj = CreateSeuratObject(
      celltype_cds@assays@data$counts,
      project = "CreateSeuratObject",
      assay = "RNA",
      names.field = 1,
      names.delim = "_",
      meta.data = NULL)
    
    #FindVariableFeatures
    seurat_obj <- FindVariableFeatures(seurat_obj, selection.method = "vst", nfeatures = 5000)
    variable_genes <- VariableFeatures(seurat_obj) #get variable gene names
    
    to_test_cds = celltype_cds[rowData(celltype_cds)$gene_short_name %in% variable_genes,] #subset cds object to variable genes
    
    meandiffs = do.call(rbind, lapply(1:length(variable_genes), function(e){
      # meandiffs = do.call(rbind, lapply(1:10, function(e){
      
      tryCatch({ #suppress error
        #e=1
        
        print(paste("v=",v," e=",e,sep=""))
        ca_data = to_test_cds[rowData(to_test_cds)$gene_short_name %in% variable_genes[e],colData(to_test_cds)$united_control_carcinoma %in% second_compare[v]]@assays@data$counts
        ca_mean = mean(ca_data)
        ca_cov = (sd(ca_data) / ca_mean )*100
        ca_percent_expressed = (length(ca_data[ca_data>0]) /  length(ca_data))*100
        
        normaskin_data = to_test_cds[rowData(to_test_cds)$gene_short_name %in% variable_genes[e],colData(to_test_cds)$united_control_carcinoma %in% first_compare[v]]@assays@data$counts
        normakin_mean = mean(normaskin_data)
        normaskin_cov = (sd(normaskin_data) / normakin_mean )*100
        normaskin_percent_expressed = (length(normaskin_data[normaskin_data>0]) /  length(normaskin_data))*100
        
        mean_abs_diff = ca_mean - normakin_mean
        mean_per_diff = ((ca_mean - normakin_mean)/normakin_mean)*100
        fc = ca_mean/normakin_mean
        
        #extract metadata
        secondcompar_metadat = colData(to_test_cds)[colData(to_test_cds)$united_control_carcinoma %in% second_compare[v],]
        firstcompar_metadat = colData(to_test_cds)[colData(to_test_cds)$united_control_carcinoma %in% first_compare[v],]
        
        #test expression diffs: make dataframe with tissue and p16 status
        ca_dat = data.frame(as.numeric(ca_data), second_compare[v], secondcompar_metadat$p16_status)
        normaskin_dat = data.frame(as.numeric(normaskin_data), first_compare[v], firstcompar_metadat$p16_status)
        names(ca_dat) = c("expression","tissue","p16")
        names(normaskin_dat) = c("expression","tissue","p16")
        agg_matdf = rbind(normaskin_dat,ca_dat)
        names(agg_matdf) = c("expression","tissue","p16")
        
        #scale everything up to 0 for negative binormail test
        rescale_value = 0-min(agg_matdf$expression)
        agg_matdf$expression = agg_matdf$expression+rescale_value
        
        #fit negative binomial GLM to eigengene testing for amount-tissue differences
        glmnb_res <- glm.nb(expression~tissue, data = agg_matdf)
        
        #return df not testing for p16 status
        standard_result_df = data.frame(variable_genes[e],
                                        ca_mean, ca_cov, ca_percent_expressed, 
                                        normakin_mean, normaskin_cov, normaskin_percent_expressed,
                                        mean_abs_diff, mean_per_diff, fc,
                                        t(summary(glmnb_res)$coefficients[2,])
        )
        standard_result_df$variable_tested = "tissue"
        names(standard_result_df)[1] = "gene"
        
        #test for p16 status - fit negative binomial GLM and test for p16 interaction
        if(first_compare[v]=="Control" | second_compare[v]=="Control" | second_compare[v]=="Cisplatin") {
          glmnb_res_p16 <- glm.nb(formula = expression ~ tissue + p16 + tissue:p16, data = agg_matdf)
          
          glmp16_result_df = data.frame(variable_genes[e],
                                        ca_mean, ca_cov, ca_percent_expressed, 
                                        normakin_mean, normaskin_cov, normaskin_percent_expressed,
                                        mean_abs_diff, mean_per_diff, fc,
                                        summary(glmnb_res_p16)$coefficients
          )
          glmp16_result_df$variable_tested = row.names(summary(glmnb_res_p16)$coefficients)
          names(glmp16_result_df)[1] = "gene"
          
          return(rbind(standard_result_df,glmp16_result_df)) #if there is p16 variation return both results
          
        } else return(standard_result_df) #if there is not p16 variation, just return the standard result
        
      },  error=function(e){})
    }))
    
    #re-order from high to low percent change
    meandiffs = meandiffs[order(-meandiffs$mean_per_diff),]
    
    #the infinity values mean that normal_skin is at 0; re-order within the infitnite group based on absolute mean diffchanges
    infdiffs = meandiffs[is.infinite(meandiffs$mean_per_diff),] #extract inifity values
    infdiffs = infdiffs[order(-infdiffs$mean_abs_diff),] #re-order absolute differences high to low
    
    #remove infinite values and then add the re-ordered inf_diffs
    no_inf_diffs <- meandiffs[!is.infinite(meandiffs$mean_per_diff),]
    meandiffs = rbind(infdiffs,no_inf_diffs)
    
    names(meandiffs)[2:7] = c("second_stage_mean", "second_stage_cov", "second_stage_percent_expressed", 
                              "first_stage_mean", "first_stage_cov", "first_stage_percent_expressed")
    
    meandiffs_final = data.frame(paste(first_compare[v],second_compare[v],sep=" -> "), meandiffs)
    names(meandiffs_final)[1] = "test"
    
    #write out results
    write.table(meandiffs_final, file=paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/DEG/integrated_original_chemo/",
                                            first_compare[v],"_",second_compare[v],"_deg.txt",sep=""),quote=F,sep="\t",row.names=F)
    
  },  error=function(e){})
  
} #end inner loop

#End DEG loop
#-------------------------------------------------------------------------------------------#
#----------------------------------------------------------------------------------------------------#
#----------------------------------------------------------------------------------------------------#
#----------------------------------------------------------------------------------------------------#
#Post-process DEG results
#1. combine DEG results

deg_comb = do.call(rbind, lapply(1:length(first_compare), function(v) {
  #v=1
  degind = read.delim(file=paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/DEG/integrated_original_chemo/",
                                 first_compare[v],"_",second_compare[v],"_deg.txt",sep=""), header=T)
  degind
}))

#2. Filter DEGs: keep infinites (cannot be tested) and tests with less than 
multi_test_correction = 3
deg_comb = subset(deg_comb, (fc==Inf | Pr...z.. < (0.01/multi_test_correction)))
deg_comb = subset(deg_comb, variable_tested != "(Intercept)") #get rid of intercept

#3. Loop through the different tests
unique(deg_comb$variable_tested)
deg_comb$variable_tested = gsub("tissuePap","tissue multivariate (Pap)",deg_comb$variable_tested)
deg_comb$variable_tested = gsub("tissueControl","tissue multivariate (Cntrl)",deg_comb$variable_tested)
deg_comb$variable_tested = gsub("p16p16_positive","p16 multivariate (+)",deg_comb$variable_tested)
deg_comb$variable_tested = gsub("tissue multivariate (Cntrl):p16 multivariate (+)","tissue-p16 interaction",deg_comb$variable_tested)
deg_comb$score = abs(deg_comb$Estimate)*(1-deg_comb$Pr...z..) #calculate score = |effect|*(1-q)
unique(deg_comb$variable_tested)
head(deg_comb)

#4. Start Loop to go through GLM models
variables_tested = unique(deg_comb$variable_tested)
for(i in 1:length(unique(variables_tested))) {
  tryCatch({ #suppress error
    #i=1
    print(i)
    #4aa. create results diretory
    results_dir = paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/DEG/integrated_original_chemo/combined_results/",variables_tested[i],"/",sep="")
    dir.create(results_dir)
    
    #4b. Subset to the right test
    deg_comb_sub = subset(deg_comb, variable_tested==variables_tested[i])
    
    #4c. Overlaps
    x = list("NSk-Pap" = subset(deg_comb_sub, test=="Normal skin -> Pap")$gene,
             "Pap-Cntrl" = subset(deg_comb_sub, test=="Pap -> Control")$gene,
             "Cntrl-Cispltn" = subset(deg_comb_sub, test=="Control -> Cisplatin")$gene
    )
    
    png(paste(results_dir,"venn_deg_overlaps.png",sep=""),width=6,height=6,units="in",res=150)
    print(
      ggVennDiagram(x) +
        ggtitle(paste("GLM:",variables_tested[i],sep=" "))
    )
    dev.off()
    
    #4d. make Manhattan plots of scores
    ncbi_names = read.delim(file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Balmains_networks/mouse_genome_gene_result_reduced.txt",header=T)
    
    deg_comb.1 = data.frame(deg_comb_sub, ncbi_names[match(deg_comb_sub$gene,ncbi_names$Symbol),] )
    deg_comb.1 = na.omit(deg_comb.1)
    deg_comb.1 = subset(deg_comb.1, chromosome != "MT")
    deg_comb.1$chr_ord = factor(deg_comb.1$chromosome, levels=c(1:19,"X","Y")) #order chrs correctly
    deg_comb.1 = deg_comb.1[order(-deg_comb.1$score),] #order by score high to low
    
    #4di. Overall Manhattan plot
    chrs = unique(deg_comb.1$chromosome)
    
    #get the top 2 score genes for each  in each chromosome for each comparison
    tests = unique(deg_comb.1$test)
    topscoregenes = do.call(rbind, lapply(1:length(tests), function(k){
      #k=1
      test_subset = subset(deg_comb.1, test==tests[k])
      test_top = do.call(rbind, lapply(1:length(chrs), function(j){
        #j=1
        chrsub = subset(test_subset, chromosome==chrs[j])
        chrsub[c(1:2),] #get top 2
      }))
      test_top
    }))
    
    #Overall result
    overall_manhattan_plot = ggplot(data=deg_comb.1, aes(x = start_position_on_the_genomic_accession/1000000, y = score, color=test)) + 
      geom_point(size=0.5, alpha=0.6) +
      facet_grid(test ~ chr_ord, scales="free") +
      geom_text_repel(data=topscoregenes, aes(x=start_position_on_the_genomic_accession/1000000, y=score,color=test,label = gene),alpha=1, size=3) + #label only top trajectory genes
      xlab('position (Mb)') +
      ylab('Score = |effect|*(1-q)') +
      ggtitle("DEG Score plot") +
      theme(text = element_text(size=17)) +
      theme(axis.text=element_text(size=14,color='black'),
            axis.title=element_text(size=14,color='black')) +
      theme(strip.background = element_rect(fill = 'white', color='black', size=1)) +
      theme(plot.title = element_text(size = 24)) +
      theme(panel.background=element_blank(),
            panel.grid.major=element_blank(),
            panel.grid.minor=element_blank()) +
      #scale_x_continuous(breaks=number_ticks(10)) +
      #scale_y_continuous(breaks=number_ticks(10)) +
      theme(axis.line = element_line(colour = 'black')) +
      theme(legend.position = 'none') +
      theme(axis.line = element_line(colour = 'black')) +
      scale_color_manual(name="tissue\ncomparison",values = c("Normal skin -> Pap" = "darkblue","Pap -> Control" = "purple", "Control -> Cisplatin" = "darkred")) +
      theme(panel.background = element_rect(colour = "black")) +
      theme(panel.spacing = unit(0, "lines")) +
      theme_classic() +
      theme(panel.grid.major = element_blank(),
            panel.grid.minor = element_blank(),
            strip.background = element_blank()) +
      theme(legend.position="bottom") +
      theme(axis.text.x=element_text(size=3,angle=90,color='black'))
    
    png(paste(results_dir,"combined_deg_genome_scan.png",sep=""),width=16,height=8,units="in",res=200)
    print(
      overall_manhattan_plot
    )
    dev.off()
    
    #4dii. print out top 20 genes in each chr for each test
    pdf(file=paste(results_dir,"chromosome_wise_DEG_genome_scan.pdf",sep=""),height=8,width=8)
    for(j in 1:length(chrs)) {
      #j=1
      
      outersub = subset(deg_comb.1, chromosome==chrs[j])
      innersub = do.call(rbind, lapply(1:length(tests), function(o){
        #i=4
        comparsub = subset(outersub, test==tests[o])
        comparsub = comparsub[order(-comparsub$score),]
        comparsub[c(1:20),] #get top 20
      }))
      
      top_deg_plot = ggplot(data=outersub, aes(x = start_position_on_the_genomic_accession/1000000, y = score, color=test)) + 
        geom_point(size=0.5, alpha=0.6) +
        facet_wrap(test ~ ., scales="free",ncol=1) +
        geom_text_repel(data=innersub, aes(x=start_position_on_the_genomic_accession/1000000, y=score,color=test,label = gene),alpha=1, size=3) + #label only top trajectory genes
        xlab('position (Mb)') +
        ylab('Score = |effect|*(1-q)') +
        ggtitle(paste("Chr",chrs[j],"DEG Score plot",sep=" ")) +
        theme(text = element_text(size=17)) +
        theme(axis.text=element_text(size=14,color='black'),
              axis.title=element_text(size=14,color='black')) +
        theme(strip.background = element_rect(fill = 'white', color='black', size=1)) +
        theme(panel.background=element_blank(),
              panel.grid.major=element_blank(),
              panel.grid.minor=element_blank()) +
        #scale_x_continuous(breaks=number_ticks(10)) +
        #scale_y_continuous(breaks=number_ticks(10)) +
        theme(axis.line = element_line(colour = 'black')) +
        theme(legend.position = 'none') +
        theme(axis.line = element_line(colour = 'black')) +
        scale_color_manual(name="tissue\ncomparison",values = c("Normal skin -> Pap" = "darkblue","Pap -> Control" = "purple", "Control -> Cisplatin" = "darkred")) +
        theme(panel.background = element_rect(colour = "black")) +
        theme(panel.spacing = unit(0, "lines")) +
        theme_classic() +
        theme(panel.grid.major = element_blank(),
              panel.grid.minor = element_blank(),
              strip.background = element_blank()) +
        theme(legend.position="bottom") +
        theme(axis.text.x=element_text(size=10,angle=0,color='black')) +
        theme(plot.title = element_text(size = 20))
      print(top_deg_plot)
    }
    dev.off()
    
    
    #Now do GO analysis on top DEGs
    num_markers = 100 #decide how many marker I wanan consider
    
    combined_go_res = do.call(rbind, lapply(1:length(tests), function(l){
      tryCatch({ #suppress error
        #l=1
        sub_marker_file  = subset(deg_comb.1, test==tests[l])
        
        #create topGO object
        geneUniverse <- row.names(tumor_big_cds)
        genesOfInterest = sub_marker_file$gene[1:num_markers]
        geneList <- factor(as.integer(geneUniverse %in% genesOfInterest))
        names(geneList) <- geneUniverse
        GOdata <- new("topGOdata",
                      ontology = "BP",
                      allGenes=geneList,
                      #geneSel = selection,
                      annot = annFUN.org, 
                      mapping = "org.Mm.eg.db",
                      ID = "symbol")
        
        resultKS <- runTest(GOdata, algorithm = "weight01", statistic = "fisher") #Fisher exact test
        tab <- GenTable(GOdata, raw.p.value = resultKS, topNodes = length(resultKS@score), numChar = 120)
        tab$test = tests[l]
        tab = subset(tab, raw.p.value <=0.01) #keep only sig results 
        
        pdf(file=paste(results_dir,tests[l],"_top_100_DEG_GO_enrichment.pdf",sep=""),
            width=12,height=8)
        print(showSigOfNodes(GOdata, score(resultKS), firstSigNodes = 10, useInfo = "def"))
        print(showSigOfNodes(GOdata, score(resultKS), firstSigNodes = 5, useInfo = "def"))
        dev.off()
        
        write.table(tab, file=paste(results_dir,tests[l],"_top_100_DEG_GO_enrichment.txt",sep=""),
                    quote=F, row.names=F,sep="\t")
        
        return(tab)
      },  error=function(e){})
    })) #end topGO loop plz
    
  },  error=function(e){})
} #End Venn, Manhattan, GO term loop



#---------------------------------------------------------------------------------------------------------------------------------------------------# 
#---------------------------------------------------------------------------------------------------------------------------------------------------# 
#plot combined GO analysis

#get GO results files
?list.files
go_res_files = list.files(path="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/DEG/integrated_original_chemo/combined_results/tissue",
                          pattern="enrichment.txt",
                          full.names = T)

#combine topGO results from DEG and plot
go_comb = do.call(rbind, lapply(1:length(go_res_files), function(p) {
  #p=1
  degind = read.delim(file=go_res_files[p], header=T)
  degind
}))

#get top 10 GO terms for each DEG comparison
highest_go = do.call(rbind, lapply(1:length(go_res_files), function(p) {
  #p=1
  degind = read.delim(file=go_res_files[p], header=T)
  degind[c(1:20),]
}))

#get highest GO terms for each comparison
go_comp_red = do.call(rbind, lapply(1:nrow(highest_go), function(i) {
  subset(go_comb, Term==highest_go$Term[i])
}))

#get only comparison I care about
go_comp_red.sub = go_comp_red

#unmelt top GO terms
dcasted = dcast(data = go_comp_red.sub,formula = Term~test,fun.aggregate = sum,value.var = "raw.p.value")
row.names(dcasted) = dcasted$Term
dcasted = dcasted[,-1] #drop row col
dcasted = -log10(as.matrix(dcasted))

low_colors = brewer.pal(9, "Blues")[9:4]
high_colors = brewer.pal(9, "YlOrRd")[4:9]
zero_color = "white"
col = c(low_colors,zero_color,high_colors)

#cluster row NOT column
par(mar=c(7,4,4,2)+0.1)
pdf(file='/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/DEG/integrated_original_chemo/combined_results/tissue/combined_topGO.pdf', width=14, height=12)
heatmap.2(dcasted,col=bluered(100),dendrogram="row",
          trace = "none", density.info = "none",key.xlab="-log10(p)",
          main = "Top GO terms",
          key=T,
          cexCol = 1,
          Rowv=T,
          Colv=F,
          keysize=1,
          cexRow = 1,
          margins=c(12,40))
dev.off()


#-----------------------------------------------------------#
#-----------------------------------------------------------#
#-----------------------------------------------------------#
#-----------------------------------------------------------#
#-----------------------------------------------------------#
#Get individual terms from models and write out files for each comparison
#Get DEGs whose change depends ONLY on tissue
deg_comb = do.call(rbind, lapply(1:length(first_compare), function(v) {
  #v=1
  degind = read.delim(file=paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/DEG/integrated_original_chemo/",
                                 first_compare[v],"_",second_compare[v],"_deg.txt",sep=""), header=T)
  degind = subset(degind, (fc==Inf | Pr...z.. < (0.01/multi_test_correction))) #subset to nice p-vals
  degind = subset(degind, variable_tested=="tissue") #subset to nice p-vals
  
  degind$score = abs(degind$Estimate)*(1-degind$Pr...z..) #calculate score = |effect|*(1-q)
  re_ordered = degind[order(-degind$score),]
  write.table(re_ordered, 
              file=paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/DEG/integrated_original_chemo/individual_term_results/",first_compare[v],"_",second_compare[v],"_tissue_only_deg.txt",
                         sep=""),
              row.names=F,sep="\t",quote=F)
}))


#Get DEGs who depend on p16 status
#v=3
degind = read.delim(file=paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/DEG/integrated_original_chemo/",
                               first_compare[v],"_",second_compare[v],"_deg.txt",sep=""), header=T)
degind = subset(degind, (fc==Inf | Pr...z.. < (0.01/multi_test_correction))) #subset to nice p-vals
degind = subset(degind, variable_tested=="tissueControl:p16p16_positive") #subset to interaction terms only
degind$variable_tested = "tissue-p16 interaction"
degind$score = abs(degind$Estimate)*(1-degind$Pr...z..) #calculate score = |effect|*(1-q)
re_ordered = degind[order(-degind$score),]
write.table(re_ordered, 
            file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/DEG/integrated_original_chemo/individual_term_results/control_cisplatin_tissue_p16_interaction_deg.txt",
            row.names=F,sep="\t",quote=F)



#-----------------------------------------------------------#
#-----------------------------------------------------------#
#-----------------------------------------------------------#
#-----------------------------------------------------------#
#-----------------------------------------------------------#
#-----------------------------------------------------------#

#-----------------------------------------------------------#
#Plot modules
#Now test specific lineages 
module_freq_df = data.frame(table(gene_module_df$module))
names(module_freq_df) = c("module","N")
module_freq_df = module_freq_df[order(module_freq_df$module),]

agg_mat <- aggregate_gene_expression(cds_subset, gene_module_df)
module_dendro <- hclust(dist(agg_mat))
gene_module_df$module <- factor(gene_module_df$module, 
                                levels = row.names(agg_mat)[module_dendro$order])

#plot eigengene expression of all unsupervised modules of significant Moran I genes
plot_cells(cds_subset,
           genes=gene_module_df, #seed gene is the first row of the markers file
           label_cell_groups=FALSE,
           label_leaves=F,
           label_branch_points=F,
           graph_label_size=1.5,
           show_trajectory_graph=F,
           cell_size=1.2) +
  xlab("Dim. 1") + ylab("Dim. 2") + 
  theme(legend.position="bottom") +
  scale_color_viridis_c(option = "inferno",name="integrated expression") +
  ggtitle("Gene modules")+ #continuous
  theme(axis.text=element_text(size=8,color='black'),
        axis.title=element_text(size=10,color='black')) + theme(plot.title = element_text(size = 14, vjust=1))

#so it looks like module 2 genes are the most highly expressed let's see what they are


#--------------------------------------------------#
#heatmap of groups of co-regulated genes
gene_module_df <- find_gene_modules(cds_subset[pr_deg_ids,], resolution=0.001) #collapse genes to unsupervised modules
cell_group_df <- tibble::tibble(cell=row.names(colData(cds_subset)), 
                                cell_group=partitions(pap_ca.test)[colnames(cds_subset)])
agg_mat <- aggregate_gene_expression(cds_subset, gene_module_df, cell_group_df)
row.names(agg_mat) <- stringr::str_c("Module ", row.names(agg_mat))
colnames(agg_mat) <- stringr::str_c("Partition ", colnames(agg_mat))

?pheatmap
pheatmap::pheatmap(agg_mat, cluster_rows=TRUE, cluster_cols=TRUE,
                   scale="column", clustering_method="ward.D2",
                   fontsize=12)

#plot partition ?plot_cells
plot_cells(cds_subset,
           color_cells_by = "partition",
           group_cells_by = "partition",
           label_cell_groups=T,
           group_label_size = 9,
           label_leaves=T,
           label_branch_points=T,
           graph_label_size=1.5,
           show_trajectory_graph=F,
           cell_size=1.2) +
  xlab("Dim. 1") + ylab("Dim. 2") + 
  theme(legend.position="bottom") +
  ggtitle("Partitions")+ #continuous
  theme(axis.text=element_text(size=8,color='black'),
        axis.title=element_text(size=10,color='black')) + theme(plot.title = element_text(size = 14, vjust=1))

#----------------------------------------------------------------------------------------------------------------------------------------------#
#----------------------------------------------------------------------------------------------------------------------------------------------#
#----------------------------------------------------------------------------------------------------------------------------------------------#
#----------------------------------------------------------------------------------------------------------------------------------------------#
#Spike analysis
#-----------------------------------------------------------#
#Cluster 19: all cells 
cds_subset.v3 = big_cds[,colData(big_cds)$cluster %in% c("19")] #filter for cluster 19 because most of the bridge cells are only in cluster 19
write.table(data.frame(colnames(cds_subset.v3)),file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/cell_subsets_by_cell_id/cluster_19_cells.txt", sep="\t",row.names=F, quote=F) #write out the cell names of the cells I'm calling the upper spike

#add this to the big_cds object
matching_cells = match(colnames(cds_subset.v3),colnames(big_cds)) #match extracted cells with all cell names
colData(big_cds)$cluster_19 = "non-cluster-19" #add column of of apparent no matches to ALL the cells; later you will replace with matches
colData(big_cds)$cluster_19[matching_cells] <- "cluster-19" #now replace matching cells with new name

#Now find markers: one region vs all others
marker_test_res <- top_markers(big_cds, group_cells_by="cluster_19",
                               genes_to_test_per_group = 250,
                               cores=8)
marker_test_res = subset(marker_test_res, cell_group == "cluster-19") #get only the region
marker_test_res = marker_test_res[order(marker_test_res$marker_test_q_value),] #re-order low to high q-value
write.table(marker_test_res, file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/top_markers/cluster_19_vs_all_other_top_markers.txt", quote=F,row.names = F,sep="\t")

#-----------------------------------------------------------#
#cluster 19 papilloma nEP only
colData(cds_comb)
pap_nep = big_cds[,colData(big_cds)$tissue_updated %in% c("Pap") & colData(big_cds)$cell_type %in% c("nEp")] #filter for cluster 19 because most of the bridge cells are only in cluster 19
clust19 = big_cds[,colData(big_cds)$cluster %in% c("19")] #filter for cluster 19 because most of the bridge cells are only in cluster 19
matching_cells = match(colnames(clust19),colnames(pap_nep)) #match extracted cells with all cell names
colData(pap_nep)$cluster_19 = "non-cluster-19" #add column of of apparent no matches to ALL the cells; later you will replace with matches
colData(pap_nep)$cluster_19[matching_cells] <- "cluster-19" #now replace matching cells with new name
#Now find markers: one region vs all others

marker_test_res <- top_markers(pap_nep, group_cells_by="cluster_19",
                               genes_to_test_per_group = 250,
                               cores=8)
marker_test_res = subset(marker_test_res, cell_group == "cluster-19") #get only the region
marker_test_res = marker_test_res[order(marker_test_res$marker_test_q_value),] #re-order low to high q-value
write.table(marker_test_res, file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/top_markers/pap_nEp_cluster_19_vs_all_other_top_markers.txt", quote=F,row.names = F,sep="\t")

#Then enter genes into this manual GO thing: http://geneontology.org/

#-----------------------------------------------------------#
#-----------------------------------------------------------#
#Cluster 33: all cells 
cds_subset.v3 = big_cds[,colData(big_cds)$cluster %in% c("33")] #filter for cluster 19 because most of the bridge cells are only in cluster 19
write.table(data.frame(colnames(cds_subset.v3)),file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/cell_subsets_by_cell_id/cluster_33_cells.txt", sep="\t",row.names=F, quote=F) #write out the cell names of the cells I'm calling the upper spike

#add this to the big_cds object
matching_cells = match(colnames(cds_subset.v3),colnames(big_cds)) #match extracted cells with all cell names
colData(big_cds)$cluster_33 = "non-cluster-33" #add column of of apparent no matches to ALL the cells; later you will replace with matches
colData(big_cds)$cluster_33[matching_cells] <- "cluster-33" #now replace matching cells with new name

#Now find markers: one region vs all others
marker_test_res <- top_markers(big_cds, group_cells_by="cluster_33",
                               genes_to_test_per_group = 250,
                               cores=8)
marker_test_res = subset(marker_test_res, cell_group == "cluster-33") #get only the region
marker_test_res = marker_test_res[order(marker_test_res$marker_test_q_value),] #re-order low to high q-value
write.table(marker_test_res, file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/top_markers/cluster_33_vs_all_other_top_markers.txt", quote=F,row.names = F,sep="\t")

#-----------------------------------------------------------#
#çlsuter 33 papilloma nEP only
colData(cds_comb)
pap_nep = big_cds[,colData(big_cds)$tissue_updated %in% c("Pap") & colData(big_cds)$cell_type %in% c("nEp")] #filter for cluster 19 because most of the bridge cells are only in cluster 19
clust33 = big_cds[,colData(big_cds)$cluster %in% c("33")] #filter for cluster 19 because most of the bridge cells are only in cluster 19
matching_cells = match(colnames(clust33),colnames(pap_nep)) #match extracted cells with all cell names
colData(pap_nep)$cluster_33 = "non-cluster-33" #add column of of apparent no matches to ALL the cells; later you will replace with matches
colData(pap_nep)$cluster_33[matching_cells] <- "cluster-33" #now replace matching cells with new name
#Now find markers: one region vs all others

marker_test_res <- top_markers(pap_nep, group_cells_by="cluster_33",
                               genes_to_test_per_group = 250,
                               cores=8)
marker_test_res = subset(marker_test_res, cell_group == "cluster-33") #get only the region
marker_test_res = marker_test_res[order(marker_test_res$marker_test_q_value),] #re-order low to high q-value
write.table(marker_test_res, file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/top_markers/pap_nEp_cluster_33_vs_all_other_top_markers.txt", quote=F,row.names = F,sep="\t")

#-----------------------------------------------------------#
#Squamous and Spindle cells vs all nEp papilloma cells (including the spikes)
e <- readRDS(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/matvei/cds_objects/",parenchymas_updated[5],".RDS",sep=""))
f <- readRDS(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/matvei/cds_objects/",parenchymas_updated[6],".RDS",sep=""))

#add comparison names
e@colData$parenchyma_tissue = "Squamous"
f@colData$parenchyma_tissue = "Spindle"

#get CA and Pap nEp
pap_ca_nep = big_cds[,colData(big_cds)$tissue_updated %in% c("Pap","CA") & colData(big_cds)$cell_type %in% c("nEp")] #filter for cluster 19 because most of the bridge cells are only in cluster 19
unique(colData(pap_ca_nep)$tissue_updated)
unique(colData(pap_ca_nep)$cell_type)

squamous_matches = match(colnames(e),colnames(pap_ca_nep)) #match extracted cells with all cell names
spindle_matches = match(colnames(f),colnames(pap_ca_nep)) #match extracted cells with all cell names

colData(pap_ca_nep)$cell_group = "pap_nEp" #add column of of apparent no matches to ALL the cells; later you will replace with matches
colData(pap_ca_nep)$cell_group[squamous_matches] <- "squamous" #now replace matching cells with new name
colData(pap_ca_nep)$cell_group[spindle_matches] <- "spindle" #now replace matching cells with new name

#Now find markers: one region vs all others

marker_test_res <- top_markers(pap_ca_nep, group_cells_by="cell_group",
                               genes_to_test_per_group = 250,
                               cores=8)
marker_test_res = marker_test_res[order(marker_test_res$marker_test_q_value),] #re-order low to high q-value
write.table(marker_test_res, file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/top_markers/pap_ca_nEp_squamous_spindle.txt", quote=F,row.names = F,sep="\t")


#-----------------------------------------------------------#
#Get top markers of bridge cells
bridge_names = c("Lower Bridge","Middle Bridge","Upper Bridge")

for(i in 1:length(bridge_names)) {
  
  #i=1
  #Get bridge cell names
  bridge_cell_names = row.names(colData(pap_ca)[ colData(pap_ca)$integrated_type_bridge %in% bridge_names[i], ]) #match metadata rows, which are for cells
  
  #Pap and carcinoma parenchyma
  matching_cells = match(bridge_cell_names,colnames(pap_ca)) #match extracted cells with all cell names
  colData(pap_ca)$bridge = paste("non-",bridge_names[i],sep="") #add column of of apparent no matches to ALL the cells; later you will replace with matches
  colData(pap_ca)$bridge[matching_cells] <- bridge_names[i] #now replace matching cells with new name
  marker_test_res <- top_markers(pap_ca, group_cells_by="bridge",
                                 genes_to_test_per_group = 100,
                                 cores=8)
  marker_test_res = marker_test_res[order(marker_test_res$marker_test_q_value),] #re-order low to high q-value
  marker_test_res = marker_test_res[marker_test_res$cell_group %in% bridge_names[i],]  #Keep only bridge-specific results
  write.table(marker_test_res, 
              file=paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/top_markers/",
                         "bridge_pap_ca_parenchyma_",bridge_names[i],".txt",sep=""), 
              quote=F,row.names = F,sep="\t")
  
  #ALL cells: change pap_ca --> big_cds
  matching_cells = match(bridge_cell_names,colnames(big_cds)) #match extracted cells with all cell names
  colData(big_cds)$bridge = paste("non-",bridge_names[i],sep="") #add column of of apparent no matches to ALL the cells; later you will replace with matches
  colData(big_cds)$bridge[matching_cells] <- bridge_names[i] #now replace matching cells with new name
  marker_test_res <- top_markers(big_cds, group_cells_by="bridge",
                                 genes_to_test_per_group = 100,
                                 cores=8)
  marker_test_res = marker_test_res[order(marker_test_res$marker_test_q_value),] #re-order low to high q-value
  marker_test_res = marker_test_res[marker_test_res$cell_group %in% bridge_names[i],]  #Keep only bridge-specific results
  write.table(marker_test_res, 
              file=paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/top_markers/",
                         "bridge_allcells_",bridge_names[i],".txt",sep=""), 
              quote=F,row.names = F,sep="\t")
  
}


#-----------------------------------------------------------#
#Fig. 4
#plot spikes together
#-----------------------------------------------------------#
colData(big_cds)$combined_spike = "non-spike" #add column of of apparent no matches to ALL the cells; later you will replace with matches

cds_subset.v3 = big_cds[,colData(big_cds)$cluster %in% c("19")] #filter for cluster 19 because most of the bridge cells are only in cluster 19
matching_cells = match(colnames(cds_subset.v3),colnames(big_cds)) #match extracted cells with all cell names
colData(big_cds)$combined_spike[matching_cells] <- "upper spike" #now replace matching cells with new name

cds_subset.v3 = big_cds[,colData(big_cds)$cluster %in% c("33")] #filter for cluster 19 because most of the bridge cells are only in cluster 19
matching_cells = match(colnames(cds_subset.v3),colnames(big_cds)) #match extracted cells with all cell names
colData(big_cds)$combined_spike[matching_cells] <- "lower spike" #now replace matching cells with new name

combined_spike_fig = plot_cells(big_cds, color_cells_by="combined_spike", group_cells_by="cluster",
                                group_label_size = 0,
                                x=1,
                                y=2,
                                alpha=0.2,
                                cell_size=1) +
  ggtitle("Combined spike") + xlab("") + ylab("") + 
  theme(plot.title = element_text(size=20),
        axis.text=element_text(size=0,color='black'),
        axis.title=element_text(size=0),
        axis.ticks = element_line(size=0),
        legend.position="none") +
  scale_color_manual(values = c("non-spike" = "lightgrey", "lower spike" = "olivedrab4","upper spike" = "yellow4"))


#Fig1c
png("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/stem_cell_mansucript/Figs/Fig4/Fig4a.png",width=21,height=21,units="in",res=800)
print(combined_spike_fig
)
dev.off()


#----------------------------------------------------------------------------------------------------------------------------------#
#----------------------------------------------------------------------------------------------------------------------------------#
#----------------------------------------------------------------------------------------------------------------------------------#
#----------------------------------------------------------------------------------------------------------------------------------#
#Get all cluster-specific gene markers, and prepare an input file for STRING to be run on my other computer

#First copy the top 100 genes from trajectory analysis to the top_markers directory so it can be included in GO enrichment

file_paths = list.files("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/top_markers/",
                        full.names = T)
file_names = list.files("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/top_markers/",
                        full.names = F)

#Combine all files into a single big ass file
markers_combined = do.call(rbind, lapply(1:length(file_paths), function(i){
  #i=1
  
  marker_list = read.table(file=file_paths[i],header=T)
  resdf = data.frame(file_names[i],marker_list)
  names(resdf)[1] = "marker_file_name"
  return(resdf)
}))

write.table(markers_combined,
            file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/top_markers/combined/comb_markers.txt",
            row.names=F, quote=F, sep="\t")

#----------------------------------------------------------------------------------------------------------------------------------#
#----------------------------------------------------------------------------------------------------------------------------------#
#Process STRING results into a lovely Excel file

string_res = read.delim(file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/STRING/STRING_GO/STRING_enrichment_singlecell_marker_sets.txt",
                        header=T)
file_names = unique(string_res$marker_file_name)

file_names = file_names[4:12] #Pick only a few to focus on

OUT <- createWorkbook() #create excel workbook for a sample https://stackoverflow.com/questions/27713310/easy-way-to-export-multiple-data-frame-to-multiple-excel-worksheets

for(i in 1:length(file_names)) {
  #i=1
  marker_sub = subset(string_res, marker_file_name==file_names[i])
  cell_groups = unique(marker_sub$cell_group)
  
  for(j in 1:length(cell_groups)){
    #j=1
    #tryCatch({ #suppress error
    
    print(paste("i=",i," j=",j,sep=""))
    marker_sub_sub = subset(marker_sub, cell_group==cell_groups[j])
    
    # Write out to excel workbook
    sheet_name = substr(paste(file_names[i], cell_groups[j],sep="_"),1,31)
    addWorksheet(OUT, sheet_name) # Add a sheet to the workbook OUT
    writeData(OUT, sheet = sheet_name, x = marker_sub_sub) #Write data to a sheet
    #},  error=function(e){})
    
  }
}

# Export the Excel file
saveWorkbook(OUT, 
             "/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/STRING/STRING_GO/STRING_enrichment_combined.xlsx",
             overwrite=T)


#-----------------------------------------------------------#
#do GO on the makrers for lower spike, upper spike, papilloma neoplasm, squamous, non-squamous, and BRIDGE CELLS

top_marker_files = list.files("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/top_markers",full.names = T)
top_marker_files_short_name = list.files("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/top_markers",full.names = F)
num_markers = 100 #decide how many marker I wanan consider

combined_go_res = do.call(rbind, lapply(7:9, function(i){
  
  #i=7
  marker_file = read.delim(file=top_marker_files[i],header=T)
  cell_groups = unique(marker_file$cell_group)
  
  #loop through cell groups and run topGO on the top marker genes
  sig_go_results_cell_groups = do.call(rbind, lapply(1:length(cell_groups), function(j){
    
    #j=1
    print(j)
    sub_marker_file = subset(marker_file, cell_group==cell_groups[j])
    
    #create topGO object
    geneUniverse <- row.names(big_cds)
    genesOfInterest = sub_marker_file$gene_id[1:num_markers]
    geneList <- factor(as.integer(geneUniverse %in% genesOfInterest))
    names(geneList) <- geneUniverse
    GOdata <- new("topGOdata",
                  ontology = "BP",
                  allGenes=geneList,
                  #geneSel = selection,
                  annot = annFUN.org, 
                  mapping = "org.Mm.eg.db",
                  ID = "symbol")
    
    resultKS <- runTest(GOdata, algorithm = "weight01", statistic = "fisher") #Fisher exact test
    tab <- GenTable(GOdata, raw.p.value = resultKS, topNodes = length(resultKS@score), numChar = 120)
    tab$cell_group = cell_groups[j]
    tab$marker_file = top_marker_files_short_name[i]
    tab = subset(tab, raw.p.value <=0.01) #keep only sig results 
    
    pdf(file=paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/GO/spike_and_carcinoma_results/figs/",cell_groups[j],"_GO.pdf",sep=""),width=12,height=8)
    print(showSigOfNodes(GOdata, score(resultKS), firstSigNodes = 10, useInfo = "def"))
    print(showSigOfNodes(GOdata, score(resultKS), firstSigNodes = 5, useInfo = "def"))
    dev.off()
    
    return(tab)
    
  })) #end topGO loop plz
  
  return(sig_go_results_cell_groups)
  
}))

write.table(combined_go_res, file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/GO/spike_and_carcinoma_results/combined_sig_go_spikes_spindle_squamous.txt",
            quote=F, row.names=F,sep="\t")

#----------------------------------------------------------------------------------------------------------#
#---------------------------------------------------------------------------------------------------------------------------------------------------# 
#---------------------------------------------------------------------------------------------------------------------------------------------------# 
#Bridge analysis on STRING GO enrichment

#get GO results files
go_res_files = read.xlsx(
  xlsxFile="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/STRING/STRING_GO/ENRICHMENT_HEATMAP.xlsx",
  colNames=T
)
cell_groups = unique(go_res_files$cell_group)

#combine topGO results from DEG and plot
go_comb = do.call(rbind, lapply(1:length(go_res_files), function(p) {
  #p=1
  degind = subset(go_res_files, cell_group==cell_groups[p] & include_in_fig=="Y") #extract a specific cell population's GO terms
  degind_others = subset(go_res_files, cell_group!=cell_groups[p]) #now find if any GO terms are shared elsewhere among the other cell populations
  other_matches = degind_others[match(degind$term,degind_others$term),]
  na.omit(rbind(degind,other_matches))
}))

#unmelt top GO terms
# dcasted = dcast(data = go_comb,formula = description~cell_group,fun.aggregate = sum,value.var = "fdr")
dcasted = dcast(data = go_comb,formula = factor(description,levels=unique(description))~cell_group,fun.aggregate = sum,value.var = "fdr")
names(dcasted)[1] = "description"
row.names(dcasted) = dcasted$description
dcasted = dcasted[,-1] #drop row col
dcasted = as.matrix(dcasted)
dcasted[dcasted ==0 ] <- 1
dcasted = -log10(as.matrix(dcasted))


RColorBrewer::display.brewer.all()
low_colors = brewer.pal(9, "Blues")[9:4]
high_colors = brewer.pal(9, "YlOrRd")[4:9]
zero_color = "white"
col = c(low_colors,zero_color,high_colors)

#cluster row NOT column
par(mar=c(7,4,4,2)+0.1)

brewer.pal(15, "Purples")


pdf(file='/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Figs/Fig3/STRING_GO.pdf', 
    width=10, height=7)
heatmap.2(dcasted,col=brewer.pal(15, "Greens"),dendrogram="row",
          trace = "none", density.info = "none",key.xlab="-log10(FDR)",
          main = "Top GO terms",
          key=T,
          cexCol = 1,
          Rowv=F,
          Colv=F,
          keysize=1,
          cexRow = 1,
          #lhei = 3,
          margins=c(10,25))
dev.off()

pdf(file='/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Figs/Fig3/STRING_GO_nokey.pdf', 
    width=6, height=10)
heatmap.2(dcasted,col=brewer.pal(15, "Greens"),dendrogram="row",
          trace = "none", density.info = "none",key.xlab="-log10(FDR)",
          main = "",
          key=F,
          cexCol = 1,
          Rowv=F,
          Colv=F,
          keysize=1,
          cexRow = 1,
          #lhei = 3,
          margins=c(10,25))
dev.off()


#-----------------------------------------------------------#
#-----------------------------------------------------------#
#-----------------------------------------------------------#
#-----------------------------------------------------------#
#-----------------------------------------------------------#
#Get individual terms from models and write out files for each comparison
#Get DEGs whose change depends ONLY on tissue
deg_comb = do.call(rbind, lapply(1:length(first_compare), function(v) {
  #v=1
  degind = read.delim(file=paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/DEG/integrated_original_chemo/",
                                 first_compare[v],"_",second_compare[v],"_deg.txt",sep=""), header=T)
  degind = subset(degind, (fc==Inf | Pr...z.. < (0.01/multi_test_correction))) #subset to nice p-vals
  degind = subset(degind, variable_tested=="tissue") #subset to nice p-vals
  
  degind$score = abs(degind$Estimate)*(1-degind$Pr...z..) #calculate score = |effect|*(1-q)
  re_ordered = degind[order(-degind$score),]
  write.table(re_ordered, 
              file=paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/DEG/integrated_original_chemo/individual_term_results/",first_compare[v],"_",second_compare[v],"_tissue_only_deg.txt",
                         sep=""),
              row.names=F,sep="\t",quote=F)
}))


#Get DEGs who depend on p16 status
#v=3
degind = read.delim(file=paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/DEG/integrated_original_chemo/",
                               first_compare[v],"_",second_compare[v],"_deg.txt",sep=""), header=T)
degind = subset(degind, (fc==Inf | Pr...z.. < (0.01/multi_test_correction))) #subset to nice p-vals
degind = subset(degind, variable_tested=="tissueControl:p16p16_positive") #subset to interaction terms only
degind$variable_tested = "tissue-p16 interaction"
degind$score = abs(degind$Estimate)*(1-degind$Pr...z..) #calculate score = |effect|*(1-q)
re_ordered = degind[order(-degind$score),]
write.table(re_ordered, 
            file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/DEG/integrated_original_chemo/individual_term_results/control_cisplatin_tissue_p16_interaction_deg.txt",
            row.names=F,sep="\t",quote=F)



##################################################################################################################################
##################################################################################################################################
# Spike interactome analysis
#Start CellChat analysis
library(CellChat)
library(ggplot2)                  
library(patchwork)
library(ggalluvial)
library(igraph)
library(dplyr)
options(stringsAsFactors = FALSE)

#------------------------------------------------------------------------------------------------#
#make master output dir
outdir_1 = "/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/CellChat/results/"

#------------------------------------------------------------------------------------------------#
#make factors within the metadata to loop through (called tissue_sources) -> factor cells, will infer communication only within cells

#------------------------------------------------------------------------------------------------#
#Start CellChat to loop through papilloma cell types combined with the spikes
#-----------------------------------------------------------#
colData(big_cds)$combined_spike_cell_type = colData(big_cds)$cell_type #add column of of apparent no matches to ALL the cells; later you will replace with matches

cds_subset.v3 = big_cds[,colData(big_cds)$cluster %in% c("19")] #filter for cluster 19 because most of the bridge cells are only in cluster 19
matching_cells = match(colnames(cds_subset.v3),colnames(big_cds)) #match extracted cells with all cell names
colData(big_cds)$combined_spike_cell_type[matching_cells] <- "upper spike" #now replace matching cells with new name

cds_subset.v3 = big_cds[,colData(big_cds)$cluster %in% c("33")] #filter for cluster 19 because most of the bridge cells are only in cluster 19
matching_cells = match(colnames(cds_subset.v3),colnames(big_cds)) #match extracted cells with all cell names
colData(big_cds)$combined_spike_cell_type[matching_cells] <- "lower spike" #now replace matching cells with new name

# get scRNAseq data from CDS object
data.input = big_cds@assays@data$counts # normalized data matrix
meta = colData(big_cds) # a dataframe with rownames containing cell mata data

#subset to Papilloma cells only
cell.use = rownames(meta)[meta$tissue_updated == "Pap" & meta$cell_type!=""] # extract the cell names from disease data & Exclude cells that I cannot annotate as a particular cell type
print(paste("Number of cells = ",length(cell.use),sep=""))

# Prepare input data for CellChat analysis
data.input = data.input[, cell.use]
meta = data.frame(meta[cell.use, ])
unique(meta$tissue_updated) # check the cell labels
unique(meta$combined_spike_cell_type)

# Create CellChat Object
cellchat <- createCellChat(object = data.input, meta = meta, group.by = "combined_spike_cell_type")

# Set which interactome database to mine
CellChatDB <- CellChatDB.mouse # use CellChatDB.human if running on mouse data
CellChatDB.use <- CellChatDB # simply use the default CellChatDB
# set the used database in the object
cellchat@DB <- CellChatDB.use  
showDatabaseCategory(CellChatDB)

cellchat <- subsetData(cellchat) # subset the expression data of signaling genes for saving computation cost
#future::plan("multiprocess", workers = 4) # do parallel

#Preprocessing the expression data for cell-cell communication analysis
cellchat <- identifyOverExpressedGenes(cellchat)
cellchat <- identifyOverExpressedInteractions(cellchat)
cellchat <- projectData(cellchat, PPI.mouse)

#Compute the communication probability and infer cellular communication network
cellchat <- computeCommunProb(cellchat, raw.use = TRUE)

# Filter out the cell-cell communication if there are only few number of cells in certain cell groups
cellchat <- filterCommunication(cellchat, min.cells = 10)

# Create output directory
outdir = paste(outdir_1,"integrated_spike_papilloma",sep="")
dir.create(outdir)

#write out interaction results
df.net <- subsetCommunication(cellchat)
write.table(df.net, file=paste(outdir,"/interaction_results_cell_type.txt",sep=""), quote=F,row.names=F,sep="\t")

#Infer the cell-cell communication at a signaling pathway level
cellchat <- computeCommunProbPathway(cellchat)

#Calculate the aggregated cell-cell communication network
cellchat <- aggregateNet(cellchat)

#Visualize these baddies
groupSize <- as.numeric(table(cellchat@idents))

png(paste(outdir,"/aggregated_cell_cell_communication_network_NumInt_cell_type.png",sep=""),width=7,height=7,units="in",res=200)
print( netVisual_circle(cellchat@net$count, vertex.weight = groupSize, weight.scale = T, label.edge= F, title.name = paste("Number of interactions")) )
dev.off()

png(paste(outdir,"/aggregated_cell_cell_communication_network_IntWeight_cell_type.png",sep=""),width=7,height=7,units="in",res=200)
print( netVisual_circle(cellchat@net$weight, vertex.weight = groupSize, weight.scale = T, label.edge= F, title.name = paste("Interaction weights/strengths")) )
dev.off()

# Print out individual graphs
mat <- cellchat@net$weight
par(mfrow = c(3,4), xpd=TRUE)
pdf(file=paste(outdir,"/individual_cell_type_interactions.pdf",sep=""),height=7,width=7)
for (i in 1:nrow(mat)) {
  #i=1
  mat2 <- matrix(0, nrow = nrow(mat), ncol = ncol(mat), dimnames = dimnames(mat))
  mat2[i, ] <- mat[i, ]
  print(netVisual_circle(mat2, vertex.weight = groupSize, weight.scale = T, edge.weight.max = max(mat), title.name = rownames(mat)[i]) )
}
dev.off()

# show all the significant interactions (L-R pairs) from some cell groups (defined by 'sources.use') to other cell groups (defined by 'targets.use')
# netVisual_bubble(cellchat, sources.use = 4, targets.use = c(5:11), remove.isolate = FALSE)
png(paste(outdir,"/bubble_plot_communication_prob.png",sep=""),width=15,height=30,units="in",res=200)
netVisual_bubble(cellchat) + theme(axis.text.x = element_text(size=12))
dev.off()

#plotGeneExpression(cellchat, signaling = "CXCL")

# Compute the network centrality scores
cellchat <- netAnalysis_computeCentrality(cellchat, slot.name = "netP") # the slot 'netP' means the inferred intercellular communication network of signaling pathways

#Visualize the dominant senders (sources) and receivers (targets) in a 2D space
# Signaling role analysis on the aggregated cell-cell communication network from all signaling pathways
png(paste(outdir,"/interaction_summary.png",sep=""),width=6,height=6,units="in",res=200)
print( gg1 <- netAnalysis_signalingRole_scatter(cellchat) + ggtitle("Dominant senders and sources") +
         theme(axis.text= element_text(size=14),
               axis.title = element_text(size=14),
               plot.title= element_text(size=16))
       
) 
dev.off()

#Identify signals contributing most to outgoing or incoming signaling of certain cell groups  
# Signaling role analysis on the aggregated cell-cell communication network from all signaling pathways
png(paste(outdir,"/outgoing_signals.png",sep=""),width=6,height=10,units="in",res=200)
tryCatch({ #suppress error
  print( netAnalysis_signalingRole_heatmap(cellchat, pattern = "outgoing",
                                           height = 18) )
},  error=function(e){})
dev.off()

png(paste(outdir,"/incoming_signals.png",sep=""),width=10,height=10,units="in",res=200)
tryCatch({ #suppress error
  print( netAnalysis_signalingRole_heatmap(cellchat, pattern = "incoming",
                                           height = 18) )
},  error=function(e){})
dev.off()


#Identify and visualize outgoing communication pattern of secreting cells

#Incoming communication patterns
#selectK(cellchat, pattern = "outgoing")
nPatterns = 5
cellchat <- identifyCommunicationPatterns(cellchat, pattern = "incoming", k = nPatterns)
png(paste(outdir,"/incoming_patterns.png",sep=""),width=10,height=10,units="in",res=200)
netAnalysis_river(cellchat, pattern = "incoming")
dev.off()

png(paste(outdir,"/incoming_contribution_patterns.png",sep=""),width=20,height=5,units="in",res=200)
netAnalysis_dot(cellchat, pattern = "incoming") +
  theme(axis.text= element_text(size=14),
        axis.title = element_text(size=14),
        plot.title= element_text(size=16))
dev.off()

#Outgoing communication patterns
#selectK(cellchat, pattern = "outgoing")
nPatterns = 5
cellchat <- identifyCommunicationPatterns(cellchat, pattern = "outgoing", k = nPatterns)
png(paste(outdir,"/outgoing_patterns.png",sep=""),width=10,height=10,units="in",res=200)
netAnalysis_river(cellchat, pattern = "outgoing")
dev.off()

png(paste(outdir,"/outgoing_contribution_patterns.png",sep=""),width=20,height=5,units="in",res=200)
netAnalysis_dot(cellchat, pattern = "outgoing") +
  theme(axis.text= element_text(size=14),
        axis.title = element_text(size=14),
        plot.title= element_text(size=16))
dev.off()

} #end tissue loop


#-----------------------------------------------------------------------------------#
#-----------------------------------------------------------------------------------#
#-----------------------------------------------------------------------------------#
#Consensus eigengene: tumor meta-eigengene
#-----------------------------------------------------------------------------------#
#-----------------------------------------------------------------------------------#
#-----------------------------------------------------------------------------------#

seed_gene = c("Lgr5",
              "Bmi1","Sox4","Lrig1",
              "Lgr6","Sox9",
              "Sox2","Prom1","Foxa1","Krt15","Krt19","Pitx1","Psca","Prom2","Klf5",
              "Cd44")
seed_gene_class = c("Lgr5",
                    "US","US","US",
                    "Ubiquitous","Ubiquitous",
                    "LS","LS","LS","LS","LS","LS","LS","LS","LS","LS")

eigengene_size=100
#Loop through all eigengenes
eigengene_agg = do.call(rbind, lapply(1:length(seed_gene), function(i){
  
  #i=6
  print(i)
  
  #Get the correct match_index
  if( seed_gene[i]=="Prom1" ) {match_index = 12747} else {
    match_index = match(seed_gene[i], aim2probedat$SYMBOL) #match seed gene to its place in the matrix
  }
  
  #---------------------------------------#
  #Normal skin
  match_index = match(seed_gene[i], aim2probedat$SYMBOL) #match seed gene to its place in the matrix
  symbol_match = data.frame(aim2probedat$SYMBOL, rhodf[,match_index], p_df[,match_index]) #put together gene names, expression data, and p-vals
  symbol_match = symbol_match[order(-symbol_match$rhodf...match_index.),] #order from high to low gene values
  names(symbol_match) = c("gene","rho","p")
  network_type="Normal skin"
  
  return_df_nsk = data.frame(network_type,seed_gene_class[i],seed_gene[i],symbol_match$gene,symbol_match$rho)
  names(return_df_nsk) = c("network_type","seed_class","seed_gene","cor_gene","rho")
  
  #---------------------------------------#
  #Tumor
  match_index = match(seed_gene[i], aim2probedat$SYMBOL) #match seed gene to its place in the matrix
  symbol_match = data.frame(aim2probedat$SYMBOL, tumor_rhodf[,match_index], tumor_p_df[,match_index]) #put together gene names, expression data, and p-vals
  symbol_match = symbol_match[order(-symbol_match$tumor_rhodf...match_index.),] #order from high to low gene values
  names(symbol_match) = c("gene","rho","p")
  network_type="Tumor"
  
  return_df_tumor = data.frame(network_type, seed_gene_class[i],seed_gene[i],symbol_match$gene,symbol_match$rho)
  names(return_df_tumor) = c("network_type","seed_class","seed_gene","cor_gene","rho")
  
  rbind( return_df_nsk[1:eigengene_size,] , return_df_tumor[1:eigengene_size,])
  
}))


#Now get summaries of network changes
seed_gene_types = unique(eigengene_agg$seed_class)

shared_networks = do.call(rbind, lapply(2:length(seed_gene_types), function(i){
  #i=3
  nsk_sub = subset(eigengene_agg, seed_class==seed_gene_types[i] & network_type=="Normal skin")
  nsk_freq = data.frame(table(nsk_sub$cor_gene)) #get freq of correlated genes within a seed gene class eg upper spike
  nsk_freq = nsk_freq[order(-nsk_freq$Freq),] #order from high to low (shared to non-shared)
  nsk_freq_morethan2 = nsk_freq[nsk_freq$Freq>1,] #get frequently occurring nodes
  nsk_percent_shared = round((nrow(nsk_freq_morethan2) / length(unique(nsk_sub$cor_gene)))*100,2) #perecentage of shared nodes
  
  tumor_sub = subset(eigengene_agg, seed_class==seed_gene_types[i] & network_type=="Tumor")
  tumor_freq = data.frame(table(tumor_sub$cor_gene)) #get freq of correlated genes within a seed gene class eg upper spike
  tumor_freq = tumor_freq[order(-tumor_freq$Freq),] #order from high to low (shared to non-shared)
  tumor_freq_morethan2 = tumor_freq[tumor_freq$Freq>1,] #get frequently occurring nodes
  tumor_percent_shared = round((nrow(tumor_freq_morethan2) / length(unique(tumor_sub$cor_gene)))*100,2) #perecentage of shared nodes
  
  data.frame(seed_gene_types[i], nsk_percent_shared, tumor_percent_shared)
  
}))

names(shared_networks) = c("cluster","Nsk","SCC")
shared_networks$cluster = gsub("LS",3,shared_networks$cluster)
shared_networks$cluster = gsub("Ubiquitous",2,shared_networks$cluster)
shared_networks$cluster = gsub("US",1,shared_networks$cluster)

dfm = melt(shared_networks, id=c("cluster"))

network_node_fig = ggplot(data=dfm, aes(x=cluster, y=value, fill=variable)) +
  geom_bar(alpha=1, aes(x=cluster, y=value, fill=variable), stat="identity",position=position_dodge(),color="black") +
  theme_classic() +
  theme(axis.text.x=element_text(size=25,color='black'),
        axis.text.y=element_text(size=25,color='black')) +
  #xlab("Network seed gene cluster") +
  #ylab("Shared network nodes (%)") +
  xlab("") +
  ylab("") +
  #scale_x_discrete(limits = c("Normal skin","Pap","CA"),
  #labels=c("NSk","Pap","SCC")) +
  theme(legend.position="right") +
  scale_fill_manual(name="",values = c("Nsk" = "lightblue",
                                       "SCC" = "red"))

png("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Figs/Fig1/Node_sharing_with_legend.png",
    width=5,height=5,units="in",res=200)
print( network_node_fig)
dev.off()

png("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Figs/Fig1/Node_sharing_without_legend.png",
    width=5,height=5,units="in",res=200)
print( network_node_fig + theme(legend.position="none"))
dev.off()


#Calculate percent change in proportion of shared nodes between normal skin and tumor
shared_networks$percent_change = round(((shared_networks$SCC-shared_networks$Nsk)/shared_networks$Nsk)*100,2)


#-----------------------------------------------------------------------------------#
#-----------------------------------------------------------------------------------#
#Make Lower spike meta-eigengene, Consensus Eigengene
ls_genes = subset(eigengene_agg, network_type=="Tumor" & seed_class=="LS")
ls_freq = data.frame(table(ls_genes$cor_gene))
ls_freq = ls_freq[order(-ls_freq$Freq),]
nrow(ls_freq)
ls_gene_go = subset(ls_freq, Freq>1) #meta-eigengene is of genes that occur at least twice among the US eigengenes ie all shared nodes
ls_gene_go = ls_gene_go$Var1

go_candidates = ls_gene_go
length(go_candidates)

#create topGO object: top 100 POSITIVE DEGs
geneUniverse <- row.names(big_cds)
genesOfInterest = go_candidates
geneList <- factor(as.integer(geneUniverse %in% genesOfInterest))
names(geneList) <- geneUniverse
GOdata <- new("topGOdata",
              ontology = "BP",
              allGenes=geneList,
              #geneSel = selection,
              annot = annFUN.org, 
              mapping = "org.Mm.eg.db",
              ID = "symbol")

resultKS <- runTest(GOdata, algorithm = "weight01", statistic = "fisher") #Fisher exact test
tab <- GenTable(GOdata, raw.p.value = resultKS, topNodes = length(resultKS@score), numChar = 120)
tab = subset(tab, raw.p.value <=0.05) #keep only sig results 

#Get the genes that are contributing to a term
selcGenes <- genesInTerm(GOdata, whichGO=tab$GO.ID)
genes_in_go = do.call(rbind, lapply(1:length(selcGenes), function(i){
  #i=1
  data.frame(names(selcGenes)[i],  paste( c(na.omit(selcGenes[[i]][match(go_candidates,  selcGenes[[i]])])) , collapse=", ") )
}))
names(genes_in_go) = c("GO_ID","sig_genes_in_term")
tab = cbind(tab, genes_in_go)

write.table(tab, 
            file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/GO/LS_metaeigengene/LS-metaeigengene.txt",
            quote=F,sep="\t",row.names=F)

#-----------------------------------------------------------------------------------#
#-----------------------------------------------------------------------------------#
#-----------------------------------------------------------------------------------#
#Get consensus EIGENGENES from eigengenome results
#-----------------------------------------------------------------------------------#
#-----------------------------------------------------------------------------------#
#-----------------------------------------------------------------------------------#

#Cluster 33 = Lower Spike
#Cluster 19 = Upper Spike

eigengenome_res = read.table(file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/Eigengenome/eigengenome_compressed/eigegenome_expression_patterns_cell_clusters_combined.txt", header=T)
head(eigengenome_res)

#------------------------------------#
#Function to get consensus eigengene
get_consensus_eigengene = function(cluster,eigengene_type,eigengene_size) {
  
  #cluster=33
  #eigengene_type="Tumor"
  #eigengene_size=100
  
  eigengenome_res.sub = subset(eigengenome_res, (topcluster1==cluster & tissue_network==eigengene_type ) )
  
  #get eigengene constituents for each eigengene 
  all_eigengenes = do.call(rbind, lapply(1:nrow(eigengenome_res.sub), function(j){
    #j=1
    print(j)
    match_index = match(eigengenome_res.sub$seed_gene[j], aim2probedat$SYMBOL) #match seed gene to its place in the matrix
    
    if(eigengene_type=="Tumor") {
      symbol_match = data.frame(aim2probedat$SYMBOL, tumor_rhodf[,match_index], tumor_p_df[,match_index])} else
      {symbol_match = data.frame(aim2probedat$SYMBOL, rhodf[,match_index], p_df[,match_index])}   #put together gene names, expression data, and p-vals
    
    names(symbol_match) = c("gene","rho","p")
    symbol_match = symbol_match[order(-symbol_match$rho),] #order from high to low gene values
    constituents = data.frame(eigengenome_res.sub$seed_gene[j],symbol_match$gene[1:eigengene_size])
    names(constituents) = c("seed_gene","constituent_genes")
    constituents
  }))
  
  summary_df = data.frame(table(all_eigengenes$constituent_genes)) #count how often a gene is included
  summary_df = summary_df[order(-summary_df$Freq),] #order from most to least freq
  return(summary_df$Var1[1:eigengene_size])
}

#Call this function for different clusters
conensus_eigengene_33 = get_consensus_eigengene(33, "Tumor", 100)
conensus_eigengene_19 = get_consensus_eigengene(19, "Tumor", 100)

consensus_spikes = data.frame(conensus_eigengene_33,conensus_eigengene_19)
write.table(consensus_spikes,
            file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/Consensus\ eigengenes/spikes/spike_consensus_eigengenes.txt",
            quote=F,row.names=F,sep="\t")

#---------------------------------------------#
#Lower spike
i=1
marker_df = data.frame(unique(alias2SymbolTable(consensus_spikes[,i], species = "Mm")),"eigengene","eigengene")
names(marker_df) = c("gene","cell_type","module")
marker_df = na.omit(marker_df)

lower_spike_fig = plot_cells(big_cds,
                             genes=marker_df, #seed gene is the first row of the markers file
                             label_cell_groups=FALSE,
                             label_leaves=F,
                             label_branch_points=F,
                             graph_label_size=1.5,
                             show_trajectory_graph=F,
                             alpha=0.4) +
  xlab("Dim. 1") + ylab("Dim. 2") + 
  theme(legend.position="none") +
  scale_color_viridis_c(option = "inferno",name="expression") +
  ggtitle(paste("Lower spike consensus eigengene",sep=" "))+ #continuous
  theme(axis.text=element_text(size=0,color='black'),
        axis.title=element_text(size=0,color='black')) + 
  theme(plot.title = element_text(size = 15, vjust=0)) +
  theme(strip.text.x = element_text(size = 0)) + #remove facet label
  theme(panel.grid.major=element_blank(), 
        panel.grid.minor=element_blank(), 
        panel.background=element_blank()) +
  theme(axis.line = element_line(size = 0.3, colour = "black")) +
  theme(axis.ticks = element_blank())

#---------------------------------------------#
#Upper spike
i=2
marker_df = data.frame(unique(alias2SymbolTable(consensus_spikes[,i], species = "Mm")),"eigengene","eigengene")
names(marker_df) = c("gene","cell_type","module")
marker_df = na.omit(marker_df)

upper_spike_fig = plot_cells(big_cds,
                             genes=marker_df, #seed gene is the first row of the markers file
                             label_cell_groups=FALSE,
                             label_leaves=F,
                             label_branch_points=F,
                             graph_label_size=1.5,
                             show_trajectory_graph=F,
                             alpha=0.4) +
  xlab("Dim. 1") + ylab("Dim. 2") + 
  theme(legend.position="none") +
  scale_color_viridis_c(option = "inferno",name="expression") +
  ggtitle(paste("Upper spike consensus eigengene",sep=" "))+ #continuous
  theme(axis.text=element_text(size=0,color='black'),
        axis.title=element_text(size=0,color='black')) + 
  theme(plot.title = element_text(size = 15, vjust=0)) +
  theme(strip.text.x = element_text(size = 0)) + #remove facet label
  theme(panel.grid.major=element_blank(), 
        panel.grid.minor=element_blank(), 
        panel.background=element_blank()) +
  theme(axis.line = element_line(size = 0.3, colour = "black")) +
  theme(axis.ticks = element_blank())

#---------------------------------------------#
png("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/Consensus\ eigengenes/spikes/consensus_spike_eigengenes.png",
    width=16,height=8,units="in",res=300)
print(
  multiplot(lower_spike_fig,upper_spike_fig,cols=2)
)
dev.off()

#=====================================================================================
prostate_cds = readRDS("/Users/marktaylor/Desktop/Projects/Balmain\ lab/human_scRNAseq/cutaneous_SCC_scRNAseq/human_prostate_normal_tumor_cds.RDS")
optimal_min_dist = 0.75

prostate_cds <- preprocess_cds(prostate_cds, 
                               method="PCA",
                               num_dim = 100,
                               norm_method = "none", #Have these data definitely not been normalized by Seurat?
                               alignment_group = NULL)

#4. A further round of (nonlinear) dimensionality reduction using UMAP ?reduce_dimension
#4b.
prostate_cds <- reduce_dimension(
  prostate_cds,
  max_components = 2,
  reduction_method = c("UMAP"),
  preprocess_method = "PCA",
  umap.metric = "cosine",
  umap.min_dist = optimal_min_dist,
  umap.n_neighbors = 15L,
  umap.fast_sgd = FALSE,
  umap.nn_method = "annoy",
  cores = 1,
  verbose = FALSE
)

#5. Cluster and partition ?cluster_cells cells into supergroups community detection (Phenograph algorithm) Community detection of cells as part of Phenograph algorithm https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4508757/
#5b. 
prostate_cds <- cluster_cells(prostate_cds,
                              reduction_method = c("UMAP"),
                              k = 10, #a bigger k will result in lower resolution 
                              cluster_method = c("leiden"),
                              num_iter = 2,
                              partition_qval = 0.05, #Numeric, the q-value cutoff to determine when to partition
                              weight = FALSE,
                              resolution = NULL,
                              random_seed = NULL,
                              verbose = F)
colData(prostate_cds)

dims12.tissue = plot_cells(prostate_cds, color_cells_by="tissue", group_cells_by="cluster",
                           group_label_size = 0,
                           x=1,
                           y=2,
                           cell_size=0.4,
                           alpha=0.8) +
  ggtitle("") + theme(axis.ticks = element_blank()) +
  scale_color_viridis_c(option = "inferno",name="") +
  theme(axis.text=element_text(size=0,color='black'),
        axis.title=element_text(size=0,color='black')) + theme(plot.title = element_text(size = 0, vjust=0)) +
  theme(strip.text.x = element_text(size = 0)) +
  scale_color_manual(name="tissue",values = c("normal" = "skyblue3","tumor" = "darkred")) + theme(legend.position="bottom")

png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Figs/Fig4/human_manifold/","disease_stage",".png",sep=""),
    width=7,height=7,units="in",res=300)
print(dims12.tissue)
dev.off()


match("Tacstd2" ,row.names(big_cds))
#Loop through all eigengenes
eigengene_agg = do.call(cbind, lapply(c(440,313), function(i){
  #subset eigengene_agg = do.call(cbind, lapply(172:175, function(i){
  tryCatch({ #suppress error
    
    #i=313
    print(i)
    #only if more than 1 constituent gene
    marker_df = data.frame(na.omit(balmain_genes[,i]),names(balmain_genes)[i])
    names(marker_df) = c("gene","cell_type")
    
    #remake the marker df dataframe with the matched gene names
    marker_df = data.frame(unique(alias2SymbolTable(unique(marker_df$gene), species = "Mm")),names(balmain_genes)[i],names(balmain_genes)[i])
    names(marker_df) = c("gene","cell_type","module")
    marker_df = na.omit(marker_df) #remove non-matching genes
    
    #all cells eigengene
    agexp_plot_tumor.12 = plot_cells(big_cds,
                                     genes=marker_df,
                                     label_cell_groups=FALSE,
                                     label_leaves=F,
                                     label_branch_points=F,
                                     graph_label_size=1.5,
                                     show_trajectory_graph=F) +
      xlab("") + ylab("") + 
      ggtitle(paste(names(balmain_genes)[i]," eigengene",sep="")) +
      theme(axis.ticks = element_blank()) +
      scale_color_viridis_c(option = "inferno",name="") +
      theme(axis.text=element_text(size=0,color='black'),
            axis.title=element_text(size=0,color='black')) + theme(plot.title = element_text(size = 0, vjust=0)) +
      theme(strip.text.x = element_text(size = 0)) +
      theme(legend.position="none")
    
    #---------------------------------------#
    
    #make marker df for grahing eigengene
    marker_df_human = data.frame(convertMouseGeneList(marker_df$gene),"eigengene","eigengene")
    names(marker_df_human) = c("gene","cell_type","module")
    marker_df_human = na.omit(marker_df_human)
    
    #all cells eigengene
    agexp_plot_tumor.12_cscc = plot_cells(prostate_cds,
                                          genes=marker_df_human,
                                          label_cell_groups=FALSE,
                                          label_leaves=F,
                                          label_branch_points=F,
                                          graph_label_size=1.5,
                                          show_trajectory_graph=F) +
      xlab("") + ylab("") + 
      ggtitle(paste(names(balmain_genes)[i]," eigengene",sep="")) +
      theme(axis.ticks = element_blank()) +
      scale_color_viridis_c(option = "inferno",name="") +
      theme(axis.text=element_text(size=0,color='black'),
            axis.title=element_text(size=0,color='black')) + theme(plot.title = element_text(size = 0, vjust=0)) +
      theme(strip.text.x = element_text(size = 0)) +
      theme(legend.position="none")
    
    
    #-----------#
    #print out seed gene and NSk eigengene plots for extended data
    png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Figs/Fig4/human_manifold/",names(balmain_genes)[i],".png",sep=""),
        width=14,height=7,units="in",res=300)
    
    print(multiplot( agexp_plot_tumor.12,
                     agexp_plot_tumor.12_cscc, 
                     cols=2))
    dev.off()
    
  },  error=function(e){})
  
}))




#=====================================================================================
#Human cutaneous SCC
cscc_cds = readRDS("/Users/marktaylor/Desktop/Projects/Balmain\ lab/human_scRNAseq/cutaneous_SCC_scRNAseq/human_cSCC_cds.RDS")

cscc_cds <- preprocess_cds(cscc_cds, 
                           method="PCA",
                           num_dim = 100,
                           norm_method = "none", #Have these data definitely not been normalized by Seurat?
                           alignment_group = NULL)

cscc_cds = align_cds(cds = cscc_cds,
                     num_dim = 100,
                     preprocess_method = "PCA",
                     alignment_group = "patient",
                     verbose=T)

#4b. ?reduce_dimension
cscc_cds <- reduce_dimension(
  cscc_cds,
  max_components = 2,
  reduction_method = c("UMAP"),
  preprocess_method = "Aligned", #must now used ALIGNEMNT PRE-PROCESSING METHOD bc it replaces the other pre-process shit
  umap.metric = "cosine",
  umap.min_dist = 0.1,
  umap.n_neighbors = 15L,
  umap.fast_sgd = FALSE,
  umap.nn_method = "annoy",
  cores = 1,
  verbose = FALSE
)

#5. Cluster and partition ?cluster_cells cells into supergroups community detection (Phenograph algorithm) Community detection of cells as part of Phenograph algorithm https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4508757/
cscc_cds <- cluster_cells(cscc_cds,
                          reduction_method = c("UMAP"),
                          k = 10, #a bigger k will result in lower resolution 
                          cluster_method = c("leiden"),
                          num_iter = 2,
                          partition_qval = 0.05, #Numeric, the q-value cutoff to determine when to partition
                          weight = FALSE,
                          resolution = NULL,
                          random_seed = NULL,
                          verbose = F)

dims12.tissue = plot_cells(cscc_cds, color_cells_by="tum.norm", group_cells_by="cluster",
                           group_label_size = 0,
                           x=1,
                           y=2,
                           cell_size=0.4,
                           alpha=0.8) +
  ggtitle("Inter-patient Heterogeneity Controlled") + theme(axis.ticks = element_blank()) +
  scale_color_viridis_c(option = "inferno",name="") +
  theme(axis.text=element_text(size=0,color='black'),
        axis.title=element_text(size=0,color='black')) + theme(plot.title = element_text(size = 15, vjust=0)) +
  theme(strip.text.x = element_text(size = 0)) +
  scale_color_manual(name="tissue",values = c("Normal" = "skyblue3","Tumor" = "darkred")) + theme(legend.position="bottom")

png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/balmain_eigene/eigengene_PNGs_mouse-based-human_v6/","1_tumor_status",".png",sep=""),
    width=7,height=7,units="in",res=300)
print(dims12.tissue)
dev.off()


#-----------------------------------------------------------------------------------#
#-----------------------------------------------------------------------------------#
#cSCC eigengenes of interest based on Ji et al markers for their TSK population
human_seed_gene = c("MMP10","LAMC2","NT5E","SLITKR6",
                    "ELK3","KLF7","FOSL1","ETS1","KLF6","SOX4",
                    "HDAC1","NR3C1",
                    "INHBA","PTHLH","MAGEA4","FEZ1","NT5E","IL24","LAMC2",
                    "KCNMA","CD276")

seed_gene = convertHumanGeneList(human_seed_gene) #Convert human markers to mouse

seed_gene = c(seed_gene, #now add the seed genes from the paper to this list
              "Lgr5",
              "Bmi1","Sox4","Lrig1",
              "Lgr6","Sox9",
              "Sox2","Prom1","Foxa1","Krt15","Krt19","Pitx1","Psca","Prom2","Klf5",
              "Cd44",
              "Mmp10")

match("Mmp10",row.names(big_cds))

#Loop through all eigengenes
eigengene_agg = do.call(cbind, lapply(1:length(seed_gene), function(i){
  tryCatch({ #suppress error
    
    #i=17
    print(i)
    
    #---------------------------------------#
    #Mouse
    #---------------------------------------#
    seed_gene_manifold = plot_cells(big_cds,
                                    genes=seed_gene[i],
                                    label_cell_groups=FALSE,
                                    label_leaves=F,
                                    label_branch_points=F,
                                    graph_label_size=1.5,
                                    show_trajectory_graph=F) +
      xlab("") + ylab("")  +
      ggtitle(paste(seed_gene[i]," seed gene",sep="")) +
      theme(axis.ticks = element_blank()) +
      scale_color_viridis_c(option = "inferno",name="") +
      theme(axis.text=element_text(size=0,color='black'),
            axis.title=element_text(size=0,color='black')) + theme(plot.title = element_text(size = 15, vjust=0)) +
      theme(strip.text.x = element_text(size = 0)) +
      theme(legend.position="bottom")
    
    #---------------------------------------#
    #tumor
    match_index = match(seed_gene[i], aim2probedat$SYMBOL) #match seed gene to its place in the matrix
    symbol_match = data.frame(aim2probedat$SYMBOL, tumor_rhodf[,match_index], tumor_p_df[,match_index]) #put together gene names, expression data, and p-vals
    symbol_match = symbol_match[order(-symbol_match$tumor_rhodf...match_index.),] #order from high to low gene values
    names(symbol_match) = c("gene","rho","p")
    
    #make marker df for grahing eigengene
    marker_df = data.frame(unique(alias2SymbolTable(symbol_match$gene[1:(eigengene_size+1)], species = "Mm")),paste(seed_gene[i],"Carcinoma",sep=""),paste(seed_gene[i],"Carcinoma",sep=""))
    names(marker_df) = c("gene","cell_type","module")
    marker_df = na.omit(marker_df)
    
    #all cells eigengene
    agexp_plot_tumor.12 = plot_cells(big_cds,
                                     genes=marker_df,
                                     label_cell_groups=FALSE,
                                     label_leaves=F,
                                     label_branch_points=F,
                                     graph_label_size=1.5,
                                     show_trajectory_graph=F) +
      xlab("") + ylab("") + 
      ggtitle(paste(seed_gene[i]," eigengene",sep="")) +
      theme(axis.ticks = element_blank()) +
      scale_color_viridis_c(option = "inferno",name="") +
      theme(axis.text=element_text(size=0,color='black'),
            axis.title=element_text(size=0,color='black')) + theme(plot.title = element_text(size = 15, vjust=0)) +
      theme(strip.text.x = element_text(size = 0)) +
      theme(legend.position="bottom")
    
    #---------------------------------------#
    #Human cSCC
    #---------------------------------------#
    
    seed_gene_manifold_cscc = plot_cells(cscc_cds,
                                         genes="MMP10",
                                         label_cell_groups=FALSE,
                                         label_leaves=F,
                                         label_branch_points=F,
                                         graph_label_size=1.5,
                                         show_trajectory_graph=F) +
      xlab("") + ylab("")  +
      ggtitle(paste(seed_gene[i]," seed gene",sep="")) +
      theme(axis.ticks = element_blank()) +
      scale_color_viridis_c(option = "inferno",name="") +
      theme(axis.text=element_text(size=0,color='black'),
            axis.title=element_text(size=0,color='black')) + theme(plot.title = element_text(size = 15, vjust=0)) +
      theme(strip.text.x = element_text(size = 0)) +
      theme(legend.position="bottom")
    
    #---------------------------------------#
    
    #make marker df for grahing eigengene
    marker_df_human = data.frame(convertMouseGeneList(marker_df$gene),"eigengene","eigengene")
    names(marker_df_human) = c("gene","cell_type","module")
    marker_df_human = na.omit(marker_df_human)
    
    #all cells eigengene
    agexp_plot_tumor.12_cscc = plot_cells(cscc_cds,
                                          genes=marker_df_human,
                                          label_cell_groups=FALSE,
                                          label_leaves=F,
                                          label_branch_points=F,
                                          graph_label_size=1.5,
                                          show_trajectory_graph=F) +
      xlab("") + ylab("") + 
      ggtitle(paste(seed_gene[i]," eigengene",sep="")) +
      theme(axis.ticks = element_blank()) +
      scale_color_viridis_c(option = "inferno",name="") +
      theme(axis.text=element_text(size=0,color='black'),
            axis.title=element_text(size=0,color='black')) + theme(plot.title = element_text(size = 15, vjust=0)) +
      theme(strip.text.x = element_text(size = 0)) +
      theme(legend.position="bottom")
    
    
    #-----------#
    #print out seed gene and NSk eigengene plots for extended data
    png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/balmain_eigene/eigengene_PNGs_mouse-based-human_v6/",seed_gene[i],".png",sep=""),
        width=14,height=14,units="in",res=300)
    
    print(multiplot(seed_gene_manifold, agexp_plot_tumor.12,
                    seed_gene_manifold_cscc, agexp_plot_tumor.12_cscc, 
                    cols=2))
    dev.off()
    
  },  error=function(e){})
  
}))

#-------------------------------------------------------------------------------#
#loop through a priori balmain genes
#TSK signature genes from table S2 of Ji et al: https://www.cell.com/cell/fulltext/S0092-8674(20)30672-3?_returnURL=https%3A%2F%2Flinkinghub.elsevier.com%2Fretrieve%2Fpii%2FS0092867420306723%3Fshowall%3Dtrue#supplementaryMaterial
tsk_human_genes = c("MMP10","PTHLH","FEZ1","IL24","KCNMA1","INHBA","MAGEA4","NT5E","LAMC2","SLITRK6","MMP1","SERPINE1","SERPINE2","IGFBP6","TNC","ITGA5","ECM1","SPINK6","TMSB10","KRT18","FST","ODC1","LAMB3","ITGA6","ACTB","ITGB1","SORL1","PFN1","EMP3","S100A10","EREG","LAMA3","PRKCDBP","OCIAD2","CTSV","ANXA3","S100A2","MET","CD99","TMSB4X","TIMP3","TPM4","NEFM","SCG5","SH3BGRL3","PLAU","PKM","CD63","LGALS1","BMP1","TNFRSF12A","ATP1B1","CFL1","MBOAT2","F3","TNFRSF21","CLIC1","CAP1","PDPN","SERINC2","ANXA1","BSG","DSG2","RHOC","KRT17","PDLIM7","PHLDA1","GLO1","LTBP1","TAGLN2","CD151","CAPN2","COL5A2","IGFBP2","P4HA2","COL17A1","PLEK2","AREG","ARPC1B","GLIPR1","S100A6","PLOD3","YWHAZ","CAV1","ADIRF","TGFBI","RBP1","FSTL3","C16orf74","UPP1","TNNT1","ANXA5","SDC1","PRDX5","MYL12A","GJB6","SLC7A8","RAB32","FXYD5","FSCN1")
tsk_mouse_orthologs = convertHumanGeneList(tsk_human_genes) #motherfucking MMP10 is not being found by an orthology finder godamit
data.frame(tsk_mouse_orthologs)

#Barrett's esophagus undifferentiated eesophagus Table S9&S10: https://science.sciencemag.org/content/373/6556/760

be_undiff_s9 = c("CD44","EPHB2","SNORC","HES6","CDCA7","ASCL2","RHOBTB3","MYB","PPP1R1B","HIST3H2A","TRABD2A","THBS2","CPVL","ARSE","WNK4","ZNF511","GABRP","TCF12","TACC1","RCN1","LHFPL3-AS2","CNTNAP2","ST6GAL1","FABP6","ADRA2A","SLC38A11","AC097478.1","GPRIN3","PROX1","MAP3K20-AS1","CHRNA5","EVA1A","NMNAT3","PRKX","RPL5P12","F12","CRYZL2P","HOXB4","GABPB1-AS1","C1QTNF1","TMEM173","NEUROG3","TYMS","CADPS","RPL5P4","HOXA7","CXCR4","SOD3","KLHL23","TMEM150C","CNDP1","AL390729.1","AC124067.4","SOX4","SCD5","EPHX4","GAS6-AS1","KCTD12","ZNF703","INSM1","TTYH1","OLFM4","MAP3K20","FGFR4","WNK2","GABRA4","C6orf58","TYMSOS","NOTCH1","PDZRN3","E2F3","NOA1","RHOV","RPS2","SKAP1","MLEC","SCG5","MAP2K6","TPH1","MEX3A","EEF1B2P6","RANBP17","BICD1","TEX45","LINC02418","BTNL9","FAIM2","NNT-AS1","ACTA2","ZIC2","HOXA10-AS","AC239800.2","AGMAT","RPL22L1","SRPK3","GLIS3","PLA2R1","KCNF1","HOXA3","RAC2","ZMAT3","GRB14","AQP1","BMP4","AL354707.1","BCL2","PAICS","MEST","PNLIPRP2","CYFIP2","PLEKHB1","FOXJ1","AXIN2","SCML4","IMPDH2","PLEKHS1","GJA1","PHB2","ZNF37BP","CTTNBP2","C1QBP","HOXA9","CLDN15","TMEM161A","UTP14A","DPYSL2","CD8B","NOS2","SNHG3","PRR7","LINC01659","MCM4","SMC2","FAS","ALDH1B1","SLC39A8","RPL14P1","LINC00926","GTF2F2","CEP57","NT5DC3","HOXB8","CELF2","PCM1","CEMIP","SBF2-AS1","ZNF462","NIPSNAP1","KIAA1257","QPRT","ACSM3","CTPS1","VSNL1","ETV4","CCDC196","LMCD1","YARS2","C19orf48","SMOC2","ARL9","CCL25","COL18A1","DDC","TSFM","CYP2W1","HSPA4L","PURPL","USP13","STK33","NRARP","EEF1B2","ARHGEF40","SYT13","ASB9","TGFB3","ZBED1","HMGA1","SLCO4A1-AS1","LAMB1","MT-TC","GINS2","POU5F2","AC007342.9","ARID5B","CNNM1","RPL7AP66","AL353622.1","CCDC14","AC074183.2","ITPR2","SLC43A3","EDAR","SAMD5","AEN","LINC00920","SH3PXD2B")
data.frame(convertHumanGeneList(be_undiff_s9))
be_endocrine_vs_other_endocrine_s10 = c("NEUROG3","PAX4","TPPP3","RNF186","KRT40","SOX4","MEX3A","MDK","AL450311.2","APOL4","RGS16","AC097478.1","PRUNE2","GCH1","ETV1","FOXJ1","LRRIQ1","ASIC5","TRMT9B","AC004540.2","LINC01014","BAMBI","CADPS","RNF183","MMP10","LDLRAD1","CDKAL1","GADD45G","TUBA4B","AC023300.2","ARMC3","SINHCAF","INSL5","KCNK17","MAP2K6","TUBB2B","ARL9","HRK","DNAH5","CHST9","GCLC","NHLH2","FAM216B","FAM241B","PEMT","PAH","GSTA3","AC013264.1","MAP3K19","CDKN1C","ZMYND10","C6","AC013470.2","ELFN2","LINC01571","SYNGR4","WDR38","ANKRD66","ZC3H12C","PPFIA2","IHH","HOXA9","CFAP52","CCDC190","C6orf52","CFAP126","CFAP157","DRC1","SNTN","ELAVL4","LINC01158","TMEM200A","IGFBP1","OR52N1","MID1","AC010980.1","GPRC6A","KCNJ2","C9orf135","SPAG17","OR51E2","C11orf97","CFAP45","LRTM1","C5orf66","THEG","TP73","WDR63","FAM81B","CFAP77","C1orf127","ROPN1L","CFAP53","TMED6","HOXA11","RCOR2","LINC01831","SLC5A12","SCN9A","AC009414.2","C11orf70","FAM183A","OR8D4","SYT13","DAND5","NME5","PON1","LINC00907","TEX45","HOXC9","CATSPERD","DYDC2","PPP1R42","AC022568.1","ZBBX","BRINP3","PRODH2","SBK1","ZSCAN16","B3GALT5-AS1","LRRC74B","AP000757.1","TMEM236","PRAC2","STOML3","ERICH3-AS1","PAX3","ADAM22","TMSB15B","ZDHHC1","CFAP43","RXFP4","RSPH4A","PKHD1","AL162742.1","LINC02345","FEV","GSDMA","EFCAB10","AP000357.2","CAPS","AC011497.2","PLXNA1","TEKT1","CFAP100","DCDC1","DAW1","DCDC2B","AL512638.2","AC002076.1","AC022726.2","AC005077.4","AL031316.1","RGR","GABRB1","SLC23A1","SDK2")
data.frame(convertHumanGeneList(be_endocrine_vs_other_endocrine_s10))

#Human colon cancer Lgr5 network from Allan
human_colon_cancer_lgr5 = c("CTPS2","ZNRF3","PROX1","ZC3H8","MGC27345","ZNF3","DKC1","HNRNPA1L2","RNF43","TRAF5","NKD1","RBMX","MIR17HG","BCL11A","NAA38","UTP14A","ZNF507","HUWE1","SOX4","ZNF280C","PHF16","TGFBRAP1","---","GNL3L","TAS2R3","PTCD3","XKRX","DIAPH2","SNORA56","MBTD1","TXLNG","---","PHF14","ZNF273","IMMP2L","ZNF138","---","OTUD6B","SNORD61","RBMX2","THUMPD2","UBN2","THOC2","NUDT4","AMACR","MTERFD1","RAD54B","---","POLR1A","BRD7","SNORA15","SNORA15","ZNF182","ASCL2","ZFC3H1","WDR35","SPATA13","---","REV1","TAS2R5","UTP23","SNORA15","POLA1","IGFL4","ATRX","TMTC4","SLC25A14","CPSF6","C13orf23","DDX55","TAS2R4","PCMTD2","ANKRD46","SLC6A6","TIA1","---","ADNP","ANAPC1","RBBP6","CXorf23","---","RPIA","MACC1","SMARCC1","SUPT4H1","PUS7","ZNF146","CBLL1","FAM123B","GTPBP10","PLAGL2","ANAPC1","SNORA70C","ZBTB33","CDK13","SLC5A6","PLCB4","NKRF","ANAPC1","TAB3","STAG2","UPF3B","COIL","GTF3A","NSMAF","SCML1","SNORD42A","DDX52","CHD7","ANAPC1","CDCA7","PRKDC","NKAP","TRIM24","CASK","XPO4","ZNF107","SNORA69","PCID2","ZDHHC9","COX19","ABCB7","MEMO1","---","C2orf34","RBM28","SFRS6","WDR75","ZBTB39","FSD1L","HDAC2","CSPP1","RAD50","MYC","RBM12B","BRCC3","DDX10","ZNF547","TSPAN6","PNPT1","CHD6","PAN3","ZNF713","AGK","WDR43","DCAF4","RRN3P1","SNORA70","NONO","---","NOL11","GTF2IRD1","USP7","RNF219","DNAJC2","LOC151009","NUPL2","CENPJ","TPP2","NARS2","RRS1","ZNF623","SNORD53","AXIN2","RBM41","ZNF673","MRE11A","USPL1","NCOA5","MTBP","RBM26","LOC100287552","LAS1L","FMR1","ACTR3C","DNAJC15","PSPH","PUM2","FAM122B","NOL8","C13orf18","HNRNPA1","---","C19orf2","PMFBP1","ANGEL1","INE1","TMEM209","HOXB9","NR6A1","TMLHE","XIAP","FHDC1","PMS2L1","---","PHKA2","MGC72080","PCDHB18","MOCS3","ROCK2","RRN3","SNORD4A","---","ZSCAN2","BMS1","TOP1MT","ATR","ANAPC5","MPHOSPH10","PAPD5","RP9P","RBM33","---","SRPK1","C8orf33","C7orf64","PCID2","MIR196A1","ZKSCAN1","PFDN4","KRIT1","ZNF250","OTUD5","SNORA70","ENGASE","PABPC1","RRN3P1","CFTR","SSBP1","SNORD102","LUC7L","NOL10","RPL23","MGC72080","ATRX","C6orf134","MEMO1","BAZ1B","RPA4","MTMR1","RNASEH2B","SNORD52","CUL4B","RPL39","PCDHB15","NAALADL2","C2orf15","SNTB1","SMOC2","THADA","C6orf134","C6orf134","SUGT1L1","---","MKI67IP","POLR1D","FAM192A","PTK2","CCT6A","---","RPL39","PHF8","APLF","GPR75","PAXIP1","MTIF3","PDK3","C12orf45","SLC2A12","NAE1","ALDH1B1","NARG2","SYDE2","MGC72080","OFD1","DPEP1","TRIM13","DTNB","ZNF706","DMTF1","KANK1","MGC72080","ZNF195","GTF3C3","PTCH1","NUFIP1","MGC72080","C14orf101","POLR1B","RP9","MGC72080","RFXAP","EXOSC5","MORC2","UCHL3","CDK12","PRMT3","CDKAL1","PTRH2","NEBL","DIDO1","MLXIPL","XRN2")
data.frame(convertHumanGeneList(human_colon_cancer_lgr5))

#Loop through all eigengenes
eigengene_agg = do.call(cbind, lapply(439:ncol(balmain_genes), function(i){
  #subset eigengene_agg = do.call(cbind, lapply(172:175, function(i){
  tryCatch({ #suppress error
    
    #i=445
    print(i)
    #only if more than 1 constituent gene
    marker_df = data.frame(na.omit(balmain_genes[,i]),names(balmain_genes)[i])
    names(marker_df) = c("gene","cell_type")
    
    #remake the marker df dataframe with the matched gene names
    marker_df = data.frame(unique(alias2SymbolTable(unique(marker_df$gene), species = "Mm")),names(balmain_genes)[i],names(balmain_genes)[i])
    names(marker_df) = c("gene","cell_type","module")
    marker_df = na.omit(marker_df) #remove non-matching genes
    
    #all cells eigengene
    agexp_plot_tumor.12 = plot_cells(big_cds,
                                     genes=marker_df,
                                     label_cell_groups=FALSE,
                                     label_leaves=F,
                                     label_branch_points=F,
                                     graph_label_size=1.5,
                                     show_trajectory_graph=F) +
      xlab("") + ylab("") + 
      ggtitle(paste(names(balmain_genes)[i]," eigengene",sep="")) +
      theme(axis.ticks = element_blank()) +
      scale_color_viridis_c(option = "inferno",name="") +
      theme(axis.text=element_text(size=0,color='black'),
            axis.title=element_text(size=0,color='black')) + theme(plot.title = element_text(size = 15, vjust=0)) +
      theme(strip.text.x = element_text(size = 0)) +
      theme(legend.position="bottom")
    
    #---------------------------------------#
    
    #make marker df for grahing eigengene
    marker_df_human = data.frame(convertMouseGeneList(marker_df$gene),"eigengene","eigengene")
    #marker_df_human = data.frame(tsk_human_genes,"eigengene","eigengene") #For TSK just use the human eigengenes bc the orhtology finder is so bad
    names(marker_df_human) = c("gene","cell_type","module")
    marker_df_human = na.omit(marker_df_human)
    
    #all cells eigengene
    agexp_plot_tumor.12_cscc = plot_cells(cscc_cds,
                                          genes=marker_df_human,
                                          label_cell_groups=FALSE,
                                          label_leaves=F,
                                          label_branch_points=F,
                                          graph_label_size=1.5,
                                          show_trajectory_graph=F) +
      xlab("") + ylab("") + 
      ggtitle(paste(names(balmain_genes)[i]," eigengene",sep="")) +
      theme(axis.ticks = element_blank()) +
      scale_color_viridis_c(option = "inferno",name="") +
      theme(axis.text=element_text(size=0,color='black'),
            axis.title=element_text(size=0,color='black')) + theme(plot.title = element_text(size = 15, vjust=0)) +
      theme(strip.text.x = element_text(size = 0)) +
      theme(legend.position="bottom")
    
    #-----------#
    #print out seed gene and NSk eigengene plots for extended data
    png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/balmain_eigene/eigengene_PNGs_mouse-based-human_v6/",names(balmain_genes)[i],".png",sep=""),
        width=14,height=7,units="in",res=300)
    
    print(multiplot( agexp_plot_tumor.12,
                     agexp_plot_tumor.12_cscc, 
                     cols=2))
    dev.off()
    
  },  error=function(e){})
  
}))

#-----------------------------------------------------------------------------------#
#-----------------------------------------------------------------------------------#
#-----------------------------------------------------------------------------------#
#Extract gene marker lists from SI data Barrett's esophagus data: https://science.sciencemag.org/content/373/6556/760

files_si = list.files(path="/Users/marktaylor/Desktop/Projects/Balmain\ lab/human_scRNAseq/data/human_eigengenes/esophageal/abd1449_Tables_S1_to_S12/",
                      full.names=T)

#Loop through excel files
collated_barret_data = do.call(list, lapply(6:16, function(i) {
  
  tryCatch({ #suppress error
    
    #i=9
    
    marker_list = do.call(list, lapply(2:40, function(j){
      
      tryCatch({ #suppress error
        
        #j=2 #Start indexing at 2 because the first sheet is usually the legend sheet
        print(j)
        marker_genes = read.xlsx(
          xlsxFile=files_si[i],
          colNames=T,
          sheet=j
        )
        gene_markers = data.frame(marker_genes$Symbol) #extract gene symbols
        names(gene_markers) = getSheetNames(files_si[i])[j] #get sheet name
        
        #Mouse orthologous mapping
        #remake the marker df dataframe with the matched gene names
        marker_df = data.frame(unique(alias2SymbolTable(unique(convertHumanGeneList(gene_markers[,1])), species = "Mm")),"eigengene","eigengene")
        names(marker_df) = c("gene","cell_type","module")
        marker_df = na.omit(marker_df) #remove non-matching genes
        
        #all cells eigengene
        agexp_plot_tumor.12 = plot_cells(big_cds,
                                         genes=marker_df,
                                         label_cell_groups=FALSE,
                                         label_leaves=F,
                                         label_branch_points=F,
                                         graph_label_size=1.5,
                                         show_trajectory_graph=F) +
          xlab("") + ylab("") + 
          ggtitle(paste(getSheetNames(files_si[i])[j],"all-genes eigengene",sep="")) +
          theme(axis.ticks = element_blank()) +
          scale_color_viridis_c(option = "inferno",name="") +
          theme(axis.text=element_text(size=0,color='black'),
                axis.title=element_text(size=0,color='black')) + theme(plot.title = element_text(size = 15, vjust=0)) +
          theme(strip.text.x = element_text(size = 0)) +
          theme(legend.position="bottom")
        
        png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/human_scRNAseq/esophageal_SCC/SI_gene_signature_results/",getSheetNames(files_si[i])[j],
                  "_allgenes.png",sep=""),
            width=7,height=7,units="in",res=250)
        print(agexp_plot_tumor.12)
        dev.off()
        
        #If the marker set is larger than 100, try it with just the top 100
        if(nrow(marker_df)>100) {
          
          marker_df = data.frame(unique(alias2SymbolTable(unique(convertHumanGeneList(gene_markers[1:100,1])), species = "Mm")),"eigengene","eigengene")
          names(marker_df) = c("gene","cell_type","module")
          marker_df = na.omit(marker_df) #remove non-matching genes
          
          #all cells eigengene
          agexp_plot_tumor.12.top = plot_cells(big_cds,
                                               genes=marker_df,
                                               label_cell_groups=FALSE,
                                               label_leaves=F,
                                               label_branch_points=F,
                                               graph_label_size=1.5,
                                               show_trajectory_graph=F) +
            xlab("") + ylab("") + 
            ggtitle(paste(getSheetNames(files_si[i])[j],"top 100 eigengene",sep="")) +
            theme(axis.ticks = element_blank()) +
            scale_color_viridis_c(option = "inferno",name="") +
            theme(axis.text=element_text(size=0,color='black'),
                  axis.title=element_text(size=0,color='black')) + theme(plot.title = element_text(size = 15, vjust=0)) +
            theme(strip.text.x = element_text(size = 0)) +
            theme(legend.position="bottom")
          
          png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/human_scRNAseq/esophageal_SCC/SI_gene_signature_results/",getSheetNames(files_si[i])[j],
                    "_top100genes.png",sep=""),
              width=7,height=7,units="in",res=250)
          print(agexp_plot_tumor.12.top)
          dev.off()
          
          return(gene_markers)
          
        } #End if statement getting top 100 signature genes if longer than 100
        
      },  error=function(e){})
      
    })) #End loop through individual excel file
    
    #Get rid of null list elements
    marker_list <- marker_list[!sapply(marker_list,is.null)]
    marker_list
    
  },  error=function(e){})
  
})) #End going through file 


plot_cells(big_cds,
           genes="Ppp2ca",
           label_cell_groups=FALSE,
           label_leaves=F,
           label_branch_points=F,
           graph_label_size=1.5,
           show_trajectory_graph=F) +
  xlab("") + ylab("") + 
  #ggtitle(paste(getSheetNames(files_si[i])[j],"top 100 eigengene",sep="")) +
  theme(axis.ticks = element_blank()) +
  scale_color_viridis_c(option = "inferno",name="") +
  theme(axis.text=element_text(size=0,color='black'),
        axis.title=element_text(size=0,color='black')) + theme(plot.title = element_text(size = 15, vjust=0)) +
  theme(strip.text.x = element_text(size = 0)) +
  theme(legend.position="bottom")

#-----------------------------------------------------------------------------------#
#-----------------------------------------------------------------------------------#
#-----------------------------------------------------------------------------------#
#Fig. 2  Spike-specific eigengene exemplars

# Upper spike exemplar
match_index = match("Trp53", aim2probedat$SYMBOL) #match seed gene to its place in the matrix
symbol_match = data.frame(aim2probedat$SYMBOL, tumor_rhodf[,match_index], tumor_p_df[,match_index]) #put together gene names, expression data, and p-vals
symbol_match = symbol_match[order(-symbol_match$tumor_rhodf...match_index.),] #order from high to low gene values
names(symbol_match) = c("gene","rho","p")
marker_df = data.frame(unique(alias2SymbolTable(symbol_match$gene[1:(eigengene_size+1)], species = "Mm")),"eigengene","eigengene")
names(marker_df) = c("gene","cell_type","module")
tumor_genes_for_eigengene = symbol_match[1:eigengene_size,]

#all cells eigengene
marker_df = na.omit(marker_df)
upper_spike_exemplar = plot_cells(big_cds,
                                  genes=marker_df,
                                  label_cell_groups=FALSE,
                                  label_leaves=F,
                                  label_branch_points=F,
                                  graph_label_size=1.5,
                                  show_trajectory_graph=F,
                                  cell_size = 1.3,
                                  alpha=0.7) +
  xlim(-4,1) + ylim(-7.9,-3.9) +
  xlab("Dim. 1") + ylab("Dim. 2") + 
  theme(legend.position="none") +
  scale_color_viridis_c(option = "inferno",name="integrated expression") +
  ggtitle("Trp53 tumor eigengene")+ #continuous
  theme(axis.text=element_text(size=0,color='black'),
        axis.title=element_text(size=0,color='black')) + 
  theme(plot.title = element_text(size = 15, vjust=0)) +
  theme(strip.text.x = element_text(size = 0)) + #remove facet label
  theme(panel.grid.major=element_blank(), 
        panel.grid.minor=element_blank(), 
        panel.background=element_blank()) +
  theme(axis.line = element_line(size = 0.3, colour = "black")) +
  theme(axis.ticks = element_blank())

# Lower spike exemplar
match_index = match("Wnt4", aim2probedat$SYMBOL) #match seed gene to its place in the matrix
symbol_match = data.frame(aim2probedat$SYMBOL, tumor_rhodf[,match_index], tumor_p_df[,match_index]) #put together gene names, expression data, and p-vals
symbol_match = symbol_match[order(-symbol_match$tumor_rhodf...match_index.),] #order from high to low gene values
names(symbol_match) = c("gene","rho","p")
marker_df = data.frame(unique(alias2SymbolTable(symbol_match$gene[1:(eigengene_size+1)], species = "Mm")),"eigengene","eigengene")
names(marker_df) = c("gene","cell_type","module")
tumor_genes_for_eigengene = symbol_match[1:eigengene_size,]

#all cells eigengene
marker_df = na.omit(marker_df)
lower_spike_exemplar = plot_cells(big_cds,
                                  genes=marker_df,
                                  label_cell_groups=FALSE,
                                  label_leaves=F,
                                  label_branch_points=F,
                                  graph_label_size=1.5,
                                  show_trajectory_graph=F,
                                  cell_size = 1.3,
                                  alpha=0.7) +
  xlim(-4,1) + ylim(-7.9,-3.9) +
  xlab("Dim. 1") + ylab("Dim. 2") + 
  theme(legend.position="none") +
  scale_color_viridis_c(option = "inferno",name="integrated expression") +
  ggtitle("Wnt4 tumor eigengene")+ #continuous
  theme(axis.text=element_text(size=0,color='black'),
        axis.title=element_text(size=0,color='black')) + 
  theme(plot.title = element_text(size = 15, vjust=0)) +
  theme(strip.text.x = element_text(size = 0)) + #remove facet label
  theme(panel.grid.major=element_blank(), 
        panel.grid.minor=element_blank(), 
        panel.background=element_blank()) +
  theme(axis.line = element_line(size = 0.3, colour = "black")) +
  theme(axis.ticks = element_blank())


#---------------------------------------------#
png("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Figs/Fig2/spike\ exemplars/spike_exemplars.png",
    width=12,height=8,units="in",res=300)
print(
  multiplot(upper_spike_exemplar, lower_spike_exemplar,
            cols=2)
)
dev.off()


##################################################################################################################################
##################################################################################################################################
##################################################################################################################################

#Lgr6 subsets

lgr6_subsets = read.xlsx(
  xlsxFile="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/apriori_eigengenes/Mark_Lgr6-sub-eigengenes.xlsx",
  colNames=T
)

pdf(file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/balmain_eigene/Lgr6-sub-eigengenes_v1.pdf",height=16,width=16)

#loop through the eigengenes
eigengene_agg = do.call(cbind, lapply(1:ncol(lgr6_subsets), function(i){
  #subset eigengene_agg = do.call(cbind, lapply(49:57, function(i){
  tryCatch({ #suppress error
    
    #i=2
    print(i)
    #only if more than 1 constituent gene
    marker_df = data.frame(na.omit(lgr6_subsets[,i]),names(lgr6_subsets)[i])
    names(marker_df) = c("gene","cell_type")
    
    #match Balmain network names to cds gene names --> this will product lots of non-matching names that i must resolve with NCBI aliases
    allnetnames = do.call(rbind, lapply(1:nrow(marker_df), function(l){
      #l=1
      netnames = data.frame(marker_df$gene[l],rowData(big_cds)$gene_short_name[match(marker_df$gene[l],rowData(big_cds)$gene_short_name)])
      names(netnames) = c("balmain_network_gene_name","cds_gene_name")
      netnames
    }))
    cds_gene_names = rowData(big_cds)$gene_short_name #get vector of all gene names in cds object
    #get consistent names between cds and network gene names
    match_noms = do.call(rbind, lapply(1:nrow(allnetnames), function(k){
      #k=5 ; 22
      tryCatch({ #suppress error
        if(is.na(allnetnames$cds_gene_name[k])==F) {cds_alias_match = allnetnames$cds_gene_name[k]} else
        { ncbi_symbol = ncbi_names[grep(allnetnames$balmain_network_gene_name[k],ncbi_names$Aliases),]$Symbol #get the NCBI gene symbol for a balmain netowkr name
        cds_alias_match = cds_gene_names[match(ncbi_symbol,cds_gene_names)] #match this symbol to the cds gene names
        }
        data.frame(allnetnames[k,],cds_alias_match)
      },  error=function(e){})
    }))
    
    #remake the marker df dataframe with the matched gene names
    marker_df = data.frame(na.omit(unique(match_noms$cds_alias_match)),names(lgr6_subsets)[i])
    names(marker_df) = c("gene","cell_type")
    
    agg_mat = aggregate_gene_expression(
      big_cds,
      gene_group_df = marker_df,
      cell_group_df = NULL,
      norm_method = c("log"),
      pseudocount = 1,
      scale_agg_values = TRUE,
      max_agg_value = 10,
      min_agg_value = -5,
      exclude.na = TRUE
    )
    
    agg_mat = t(as.data.frame(agg_mat)) #coerce aggregated expression values into data.frame
    names(agg_mat)[1] = names(lgr6_subsets)[i]
    #agg_mat$eigene_name = names(lgr6_subsets)[i]
    
    #Monocle eigene expression
    marker_df$module <- factor(marker_df$cell_type)
    
    #all cells eigengene
    agexp_plot = plot_cells(big_cds,
                            genes=marker_df,
                            label_cell_groups=FALSE,
                            label_leaves=F,
                            label_branch_points=F,
                            graph_label_size=1.5,
                            show_trajectory_graph=F) +
      xlab("Dim. 1") + ylab("Dim. 2") + 
      theme(legend.position="bottom") +
      scale_color_viridis_c(option = "inferno",name="integrated expression") +
      ggtitle(paste("All cells:", names(lgr6_subsets)[i],":",paste(marker_df$gene,collapse="/"),sep=" "))+ #continuous
      theme(axis.text=element_text(size=8,color='black'),
            axis.title=element_text(size=10,color='black')) + theme(plot.title = element_text(size = 14, vjust=1)) +
      theme(strip.text.x = element_text(size = 0)) #remove facet label
    
    #Unsorted cells eigengene
    unsorted_agexp = plot_cells(unsorted_cds,
                                genes=marker_df,
                                label_cell_groups=FALSE,
                                label_leaves=F,
                                label_branch_points=F,
                                graph_label_size=1.5,
                                show_trajectory_graph=F) +
      xlab("Dim. 1") + ylab("Dim. 2") + 
      theme(legend.position="bottom") +
      scale_color_viridis_c(option = "inferno",name="integrated expression") +
      ggtitle(paste("Unsorted cells:", names(lgr6_subsets)[i],":",paste(marker_df$gene,collapse="/"),sep=" "))+ #continuous
      theme(axis.text=element_text(size=8,color='black'),
            axis.title=element_text(size=10,color='black')) + theme(plot.title = element_text(size = 14, vjust=1)) +
      theme(strip.text.x = element_text(size = 0)) #remove facet label
    
    #Lgr6+ cells
    lgr6_agexp = plot_cells(lgr6_cds,
                            genes=marker_df,
                            label_cell_groups=FALSE,
                            label_leaves=F,
                            label_branch_points=F,
                            graph_label_size=1.5,
                            show_trajectory_graph=F) +
      xlab("Dim. 1") + ylab("Dim. 2") + 
      theme(legend.position="bottom") +
      scale_color_viridis_c(option = "inferno",name="integrated expression") +
      ggtitle(paste("Lgr6+ FACS cells:", names(lgr6_subsets)[i],":",paste(marker_df$gene,collapse="/"),sep=" "))+ #continuous
      theme(axis.text=element_text(size=8,color='black'),
            axis.title=element_text(size=10,color='black')) + theme(plot.title = element_text(size = 14, vjust=1)) +
      theme(strip.text.x = element_text(size = 0)) #remove facet label
    
    #Progeny cells
    progeny_agexp = plot_cells(progeny_cds,
                               genes=marker_df,
                               label_cell_groups=FALSE,
                               label_leaves=F,
                               label_branch_points=F,
                               graph_label_size=1.5,
                               show_trajectory_graph=F) +
      xlab("Dim. 1") + ylab("Dim. 2") + 
      theme(legend.position="bottom") +
      scale_color_viridis_c(option = "inferno",name="integrated expression") +
      ggtitle(paste("Progeny FACS cells:", names(lgr6_subsets)[i],":",paste(marker_df$gene,collapse="/"),sep=" "))+ #continuous
      theme(axis.text=element_text(size=8,color='black'),
            axis.title=element_text(size=10,color='black')) + theme(plot.title = element_text(size = 14, vjust=1)) +
      theme(strip.text.x = element_text(size = 0)) #remove facet label
    
    print(multiplot(agexp_plot, unsorted_agexp, lgr6_agexp, progeny_agexp,cols=2))
    
    return(agg_mat)
  },  error=function(e){})
  
}))

#now heatmap of those eigengenes
virpal = viridis_pal(alpha = 1, begin = 0, end = 1, direction = 1,
                     option = "C") #set the colorblind palette
library(BBmisc)
eigengenvals.norm = as.matrix(normalize(t(eigengene_agg), method = "range", range = c(-5, 5), margin = 1L, on.constant = "quiet"))
#subsample to 1k cells
colstosample = sample(1:ncol(eigengenvals.norm),2000)
eigengenvals.norm = eigengenvals.norm[,colstosample]
heatmap.2(eigengenvals.norm, scale = "none", col = virpal, labCol="",
          trace = "none", density.info = "none",
          key.xlab="standardized integrated expression",main = "Eigene expression",
          cexRow = 0.8,
          margins=c(2,15))
detach("package:BBmisc", unload = TRUE)

dev.off()

##################################################################################################################################
##################################################################################################################################
################################################################################################################################## 

#plot interesting cells in the papilloma

pap_ca_tissue_no_trajectory = plot_cells(pap_ca, group_cells_by="cluster",
                                         gene="C330027C09Rik", 
                                         group_label_size = 0,
                                         show_trajectory_graph=F,
                                         x=1,
                                         y=2) +
  scale_color_viridis_c(option = "inferno",name="integrated expression") +
  ggtitle()+ #continuous
  theme(axis.text=element_text(size=8,color='black'),
        axis.title=element_text(size=10,color='black')) + theme(plot.title = element_text(size = 14, vjust=1)) +
  theme(strip.text.x = element_text(size = 0)) #remove facet label


####################################################################################################################
####################################################################################################################
####################################################################################################################
#Optimize by ranking and dropping
#rank: 1) calculate each constituent genes percent change normal skin-carcinoma 2) rank all constituent genes from highest percent change to lowest
#dorp: drop the least % change

for(i in c(6:8,11:48)) { #look at non-repetitive 
  for(i in c(27:48)) { #look at non-repetitive 
    
    #i=5
    print(i)
    #only if more than 1 constituent gene
    marker_df = data.frame(na.omit(balmain_genes[,i]),names(balmain_genes)[i])
    names(marker_df) = c("gene","cell_type")
    
    #UPDATED NCBI MATCHING!! --> this will product lots of non-matching names that i must resolve with NCBI aliases
    allnetnames = do.call(rbind, lapply(1:nrow(marker_df), function(l){
      #l=1
      netnames = data.frame(marker_df$gene[l],rowData(big_cds)$gene_short_name[match(marker_df$gene[l],rowData(big_cds)$gene_short_name)])
      names(netnames) = c("balmain_network_gene_name","cds_gene_name")
      netnames
    }))
    cds_gene_names = rowData(big_cds)$gene_short_name #get vector of all gene names in cds object
    #get consistent names between cds and network gene names
    match_noms = do.call(rbind, lapply(1:nrow(allnetnames), function(k){
      #k=1 ; 22
      tryCatch({ #suppress error
        if(is.na(allnetnames$cds_gene_name[k])==F) {cds_alias_match = allnetnames$cds_gene_name[k]} else
        { 
          #search through Gene aliases for a match
          ncbi_symbol = ncbi_names[grep(allnetnames$balmain_network_gene_name[k],ncbi_names$Aliases),] #search through aliases for a match
          if (nrow(ncbi_symbol)>0) {
            ncbi_matches_vector = c(as.character(ncbi_symbol$Symbol), as.character(data.frame(strsplit(as.character(ncbi_symbol$Aliases),", "))[,1]) ) #vector of all aliases that marker_gene matches
            cds_alias_match = na.omit(cds_gene_names[match(ncbi_matches_vector,cds_gene_names)]) #match this symbol to the cds gene names
          } else {
            #search through Gene Symbols
            ncbi_symbol.1 = ncbi_names[grep(allnetnames$balmain_network_gene_name[k],ncbi_names$Symbol),] 
            ncbi_matches_vector = c(as.character(ncbi_symbol.1$Symbol), as.character(data.frame(strsplit(as.character(ncbi_symbol.1$Aliases),", "))[,1]) ) #vector of all aliases that marker_gene matches
            cds_alias_match = na.omit(cds_gene_names[match(ncbi_matches_vector,cds_gene_names)]) #match this symbol to the cds gene names
          }
        }
        data.frame(allnetnames[k,],cds_alias_match)
      },  error=function(e){})
    }))
    
    #remake the marker df dataframe with the matched gene names
    marker_df = data.frame(na.omit(unique(match_noms$cds_alias_match)),names(balmain_genes)[i])
    names(marker_df) = c("gene","cell_type")
    
    #get mean differences between carcinoma and normal skin for each gene
    meandiffs = do.call(rbind, lapply(1:nrow(marker_df), function(e){
      #e=1
      ca_mean = mean(cds_comb[rowData(cds_comb)$gene_short_name %in% marker_df$gene[e],colData(cds_comb)$compare_var %in% "CA"]@assays@data$counts)
      normakin_mean = mean(cds_comb[rowData(cds_comb)$gene_short_name %in% marker_df$gene[e],colData(cds_comb)$compare_var %in% "UnTxSkin"]@assays@data$counts)
      mean_abs_diff = ca_mean - normakin_mean
      mean_per_diff = ((ca_mean - normakin_mean)/normakin_mean)*100
      data.frame(marker_df[e,],mean_abs_diff,mean_per_diff)
    }))
    
    #re-order from high to low percent change
    meandiffs = meandiffs[order(-meandiffs$mean_per_diff),]
    
    #the infinity values mean that normal_skin is at 0; re-order within the infitnite group based on absolute mean diffchanges
    infdiffs = meandiffs[is.infinite(meandiffs$mean_per_diff),] #extract inifity values
    infdiffs = infdiffs[order(-infdiffs$mean_abs_diff),] #re-order absolute differences high to low
    
    #remove infinite values and then add the re-ordered inf_diffs
    no_inf_diffs <- meandiffs[!is.infinite(meandiffs$mean_per_diff),]
    meandiffs = rbind(infdiffs,no_inf_diffs)
    
    #now go thru and drop the genes with the least big differences between high and low
    optimize_results = do.call(rbind, lapply(0:(nrow(meandiffs)-2), function(p){
      #p=44
      print(p)
      markder_df_sub = meandiffs[1:(nrow(meandiffs)-p),1:2] #increment and drop last rows
      
      #aggregate eigengene expression from subset of genes above
      agg_mat = aggregate_gene_expression(
        cds_comb,
        gene_group_df = markder_df_sub,
        cell_group_df = NULL,
        norm_method = c("log"),
        pseudocount = 1,
        scale_agg_values = TRUE,
        max_agg_value = 10,
        min_agg_value = -5,
        exclude.na = TRUE
      )
      
      agg_matdf = data.frame(t(data.frame(agg_mat))) #dataframe where each cell has eigengene value
      agg_matdf$tissue = as.vector(colData(cds_comb)$compare_var) #add tissue 
      names(agg_matdf) = c("eigengene_expression","tissue")
      
      #scale everything up to 0 for negative binormail test
      rescale_value = 0-min(agg_matdf$eigengene_expression)
      agg_matdf$eigengene_expression = agg_matdf$eigengene_expression+rescale_value
      
      #fit negative binomial GLM to eigengene testing for amount-tissue differences
      glmnb_res <- glm.nb(eigengene_expression~tissue, data = agg_matdf)
      
      #CLD on negative binomial GLM
      lsmeanres = data.frame(cld(emmeans(glmnb_res,specs = c("tissue"))))
      names(lsmeanres) = c("tissue","mean","SE","df","lower.CL","upper.CL","CLD")
      
      #lsmeanres$Kruskal_Wallis_p = kruskal.test(eigengene_expression~tissue, data=agg_matdf)$p.value #add K-W p-value to output
      
      lsmeanres$eigengene = names(balmain_genes)[i] #add column with the eigengene 
      lsmeanres$constituent_genes = paste(markder_df_sub$gene,collapse=", ")
      
      #scale everything up to 1 for correct percent change calculation
      rescale_value.2 = 1-min(lsmeanres$mean)
      lsmeanres$mean = lsmeanres$mean+rescale_value.2
      
      #abs change UnTxSkin --> Carcinoma
      abs_change = subset(lsmeanres, tissue=="CA")$mean - subset(lsmeanres, tissue=="UnTxSkin")$mean
      
      #% change UnTxSkin --> Carcinoma
      percent_change = ((subset(lsmeanres, tissue=="CA")$mean - subset(lsmeanres, tissue=="UnTxSkin")$mean) / abs(subset(lsmeanres, tissue=="UnTxSkin")$mean))*100
      
      data.frame(lsmeanres$eigengene[1], lsmeanres$constituent_genes[1], abs_change, percent_change) #result
      
    }))
    
    #graph optimizer performance
    names(optimize_results)[1:2] = c("eigengene","constituent_genes")
    optimize_results$index = as.vector(1:nrow(optimize_results))
    
    ascending_eigen = ggplot(optimize_results, aes(x=index, y=percent_change,color="darkred",shape="triangle")) +
      geom_point(alpha=1, size=1.5) +
      theme(legend.key = element_rect(fill = "white")) +
      theme(legend.text=element_text(size=14), legend.title=element_text(size=14)) +
      theme(strip.background=element_rect(fill="white")) +
      ggtitle(paste("Eigengene:",names(balmain_genes)[i],"Optimizer Performance",sep=" ")) +
      ylab("Optimizer: percent change Normal Skin to Carcinoma") +
      xlab("Constituent-gene combination index") +
      theme(text = element_text(size=18)) +
      theme(axis.text=element_text(size=18,color='black'),
            axis.title=element_text(size=20,color='black')) +
      theme(plot.title = element_text(size = 22, vjust=1)) +
      theme(axis.text=element_text(size=15,color='black'),
            axis.title=element_text(size=15,color='black')) +
      theme(panel.grid.major=element_line(colour = "darkgrey"), 
            panel.grid.minor=element_blank(), 
            panel.background=element_blank()) +
      theme(legend.key = element_rect(fill = "white")) +
      theme(strip.text.x = element_text(size = 15, colour = "black", angle = 0)) +
      theme(strip.text.y = element_text(size = 15, colour = "black", angle = 0)) +
      theme(legend.text=element_text(size=10), legend.title=element_text(size=15)) +
      scale_color_manual(values = c("darkgreen","red")) + 
      theme(legend.position="none") +
      theme(axis.text=element_text(size=8,color='black'),
            axis.title=element_text(size=10,color='black')) + theme(plot.title = element_text(size = 14, vjust=1))
    
    #optimal genes
    #re-order from high to low percentant change and then print out
    optimize_results.1 = optimize_results[order(-optimize_results$percent_change),]
    optimal_genes = data.frame(strsplit(as.character((optimize_results.1$constituent_genes[1])),", "))[,1] #get the eigengene set with the biggest percent change!
    marker_df_optimal = data.frame(optimal_genes,names(balmain_genes)[i],names(balmain_genes)[i])
    names(marker_df_optimal) = c("gene","cell_type","module")
    
    agexp = plot_cells(big_cds,
                       genes=marker_df_optimal,
                       label_cell_groups=FALSE,
                       label_leaves=F,
                       label_branch_points=F,
                       graph_label_size=1.5,
                       show_trajectory_graph=F) +
      xlab("Dim. 1") + ylab("Dim. 2") + 
      theme(legend.position="bottom") +
      scale_color_viridis_c(option = "inferno",name="integrated expression") +
      ggtitle(paste("Optimal ",names(balmain_genes)[i], " Eigengene: ",optimize_results.1$constituent_genes[1],sep=""))+ #continuous
      theme(axis.text=element_text(size=8,color='black'),
            axis.title=element_text(size=10,color='black')) + theme(plot.title = element_text(size = 14, vjust=1))
    
    
    pdf(file=paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/balmain_eigene/optimization/three_gene/",names(balmain_genes)[i],"_v2.pdf",sep=""),
        height=10,width=20)
    print(multiplot(ascending_eigen,agexp,cols=2))
    dev.off()
    
    #write out results
    write.table(optimize_results.1, file=paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/balmain_eigene/optimization/three_gene/",names(balmain_genes)[i],"_optimizer_results_v2.txt",sep=""),quote=F,sep="\t",row.names=F)
    
  } #end loop
  
  #concatenate individual files into one dataframe and write out
  file_list <- list.files(path="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/balmain_eigene/optimization/multigene/", 
                          pattern="_v2.txt",full.names=T) #get list of all files
  dataset <- ldply(file_list, read.table, header=TRUE, sep="\t") #concatenate
  write.table(dataset, file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/balmain_eigene/optimization/multigene/combined_optimization_runs_eigengenes.txt", quote=F,row.names = F,sep="\t")
  
  ####################################################################################################################
  ####################################################################################################################
  ####################################################################################################################
  #Optimize across ALL PARENCHYMA CELLS ALL GENES: calculate scores across every single genes!
  
  #find 2000 most variable genes
  #Create Seurat object from cds
  seurat_obj = CreateSeuratObject(
    cds_comb@assays@data$counts,
    project = "CreateSeuratObject",
    assay = "RNA",
    names.field = 1,
    names.delim = "_",
    meta.data = NULL)
  
  #FindVariableFeatures
  seurat_obj <- FindVariableFeatures(seurat_obj, selection.method = "vst", nfeatures = 2000)
  #variable_genes <- VariableFeatures(seurat_obj) #get variable gene names
  
  #TRY IT ON LITERALLY ALL GENES #I'M RE-DOING 
  variable_genes = rowData(cds_comb)$gene_short_name
  
  sub_cds = cds_comb[rowData(cds_comb)$gene_short_name %in% variable_genes,] #subset cds object to variable genes
  rm(seurat_obj)
  
  #get mean differences between carcinoma and normal skin for 2000 most variable genes
  meandiffs = do.call(rbind, lapply(1:length(variable_genes), function(e){
    #e=6756
    tryCatch({ #suppress error
      
      print(e)
      ca_data = sub_cds[rowData(sub_cds)$gene_short_name %in% variable_genes[e],colData(sub_cds)$compare_var %in% "CA"]@assays@data$counts
      ca_mean = mean(ca_data)
      ca_cov = (sd(ca_data) / ca_mean )*100
      ca_percent_expressed = (length(ca_data[ca_data>0]) /  length(ca_data))*100
      
      normaskin_data = sub_cds[rowData(sub_cds)$gene_short_name %in% variable_genes[e],colData(sub_cds)$compare_var %in% "UnTxSkin"]@assays@data$counts
      normakin_mean = mean(normaskin_data)
      normaskin_cov = (sd(normaskin_data) / normakin_mean )*100
      normaskin_percent_expressed = (length(normaskin_data[normaskin_data>0]) /  length(normaskin_data))*100
      
      mean_abs_diff = ca_mean - normakin_mean
      mean_per_diff = ((ca_mean - normakin_mean)/normakin_mean)*100
      fc = log2(ca_mean/normakin_mean)
      
      #test expression diffs
      ca_dat = data.frame(as.numeric(ca_data), "Carcinoma")
      normaskin_dat = data.frame(as.numeric(normaskin_data), "Normal Skin")
      names(ca_dat) = c("expression","tissue")
      names(normaskin_dat) = c("expression","tissue")
      agg_matdf = rbind(normaskin_dat,ca_dat)
      names(agg_matdf) = c("expression","tissue")
      
      #scale everything up to 0 for negative binormail test
      rescale_value = 0-min(agg_matdf$expression)
      agg_matdf$expression = agg_matdf$expression+rescale_value
      
      #fit negative binomial GLM to eigengene testing for amount-tissue differences
      glmnb_res <- glm.nb(expression~tissue, data = agg_matdf)
      
      data.frame(variable_genes[e],
                 ca_mean, ca_cov, ca_percent_expressed, 
                 normakin_mean, normaskin_cov, normaskin_percent_expressed,
                 mean_abs_diff, mean_per_diff, fc,
                 t(summary(glmnb_res)$coefficients[2,])
      )
    },  error=function(e){})
    
  }))
  
  #re-order from high to low percent change
  meandiffs = meandiffs[order(-meandiffs$mean_per_diff),]
  
  #the infinity values mean that normal_skin is at 0; re-order within the infitnite group based on absolute mean diffchanges
  infdiffs = meandiffs[is.infinite(meandiffs$mean_per_diff),] #extract inifity values
  infdiffs = infdiffs[order(-infdiffs$mean_abs_diff),] #re-order absolute differences high to low
  
  #remove infinite values and then add the re-ordered inf_diffs
  no_inf_diffs <- meandiffs[!is.infinite(meandiffs$mean_per_diff),]
  meandiffs = rbind(infdiffs,no_inf_diffs)
  
  #write out results
  write.table(meandiffs, file=paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/DEG/carcinoma_vs_normalskin/carcinoma_vs_normalskin_allgenes.txt",sep=""),quote=F,sep="\t",row.names=F)
  
  
  ####################################################################################################################
  ####################################################################################################################
  ####################################################################################################################
  #Combinametrics optimizations: find optimal 3-gene eigengene from all possible combinations of 10 (10 most variable if more than 10, founding by Seurat)
  
  for(i in 8:ncol(balmain_genes)) {
    #i=1
    print(i)
    #only if more than 1 constituent gene
    marker_df = data.frame(na.omit(balmain_genes[,i]),names(balmain_genes)[i])
    names(marker_df) = c("gene","cell_type")
    
    #UPDATED NCBI MATCHING!! --> this will product lots of non-matching names that i must resolve with NCBI aliases
    allnetnames = do.call(rbind, lapply(1:nrow(marker_df), function(l){
      #l=1
      netnames = data.frame(marker_df$gene[l],rowData(big_cds)$gene_short_name[match(marker_df$gene[l],rowData(big_cds)$gene_short_name)])
      names(netnames) = c("balmain_network_gene_name","cds_gene_name")
      netnames
    }))
    cds_gene_names = rowData(big_cds)$gene_short_name #get vector of all gene names in cds object
    #get consistent names between cds and network gene names
    match_noms = do.call(rbind, lapply(1:nrow(allnetnames), function(k){
      #k=1 ; 22
      tryCatch({ #suppress error
        if(is.na(allnetnames$cds_gene_name[k])==F) {cds_alias_match = allnetnames$cds_gene_name[k]} else
        { 
          #search through Gene aliases for a match
          ncbi_symbol = ncbi_names[grep(allnetnames$balmain_network_gene_name[k],ncbi_names$Aliases),] #search through aliases for a match
          if (nrow(ncbi_symbol)>0) {
            ncbi_matches_vector = c(as.character(ncbi_symbol$Symbol), as.character(data.frame(strsplit(as.character(ncbi_symbol$Aliases),", "))[,1]) ) #vector of all aliases that marker_gene matches
            cds_alias_match = na.omit(cds_gene_names[match(ncbi_matches_vector,cds_gene_names)]) #match this symbol to the cds gene names
          } else {
            #search through Gene Symbols
            ncbi_symbol.1 = ncbi_names[grep(allnetnames$balmain_network_gene_name[k],ncbi_names$Symbol),] 
            ncbi_matches_vector = c(as.character(ncbi_symbol.1$Symbol), as.character(data.frame(strsplit(as.character(ncbi_symbol.1$Aliases),", "))[,1]) ) #vector of all aliases that marker_gene matches
            cds_alias_match = na.omit(cds_gene_names[match(ncbi_matches_vector,cds_gene_names)]) #match this symbol to the cds gene names
          }
        }
        data.frame(allnetnames[k,],cds_alias_match)
      },  error=function(e){})
    }))
    
    #remake the marker df dataframe with the matched gene names
    marker_df = data.frame(na.omit(unique(match_noms$cds_alias_match)),names(balmain_genes)[i])
    names(marker_df) = c("gene","cell_type")
    
    #if more than 10 genes in eigengene use Findvariable genes
    if(nrow(marker_df)>10) {
      
      #get list of 10 genes with the highest variation in the data!
      celltype_cds = cds_comb[rowData(cds_comb)$gene_short_name %in% marker_df$gene,]
      
      #Create Seurat object from celltype_cds
      seurat_obj = CreateSeuratObject(
        celltype_cds@assays@data$counts,
        project = "CreateSeuratObject",
        assay = "RNA",
        names.field = 1,
        names.delim = "_",
        meta.data = NULL)
      
      #FindVariableFeatures get the 15 most variable genes
      seurat_obj <- FindVariableFeatures(seurat_obj, selection.method = "vst", nfeatures = 10)
      variable_genes <- VariableFeatures(seurat_obj) #get variable gene names
      
    } else {variable_genes =   marker_df$gene} #if there are fewer than 15 genes than the variable_genes list just becomes those genes
    
    #optimize differences between 
    combinations_to_try = combinations(n=length(variable_genes), r=3, v=as.character(variable_genes), set=TRUE, repeats.allowed=FALSE) #now test the 15 most variable genes for optimized differences among 3 genes
    
    optimize_results = do.call(rbind, lapply(1:nrow(combinations_to_try), function(p){
      #p=1
      print(p)
      markder_df_sub = data.frame(combinations_to_try[p,],"eigengene3")
      names(markder_df_sub) = c("gene","cell_type")
      
      #aggregate eigengene expression from subset of genes above
      agg_mat = aggregate_gene_expression(
        cds_comb,
        gene_group_df = markder_df_sub,
        cell_group_df = NULL,
        norm_method = c("log"),
        pseudocount = 1,
        scale_agg_values = TRUE,
        max_agg_value = 10,
        min_agg_value = -5,
        exclude.na = TRUE
      )
      
      agg_matdf = data.frame(t(data.frame(agg_mat))) #dataframe where each cell has eigengene value
      agg_matdf$tissue = as.vector(colData(cds_comb)$compare_var) #add tissue 
      names(agg_matdf) = c("eigengene_expression","tissue")
      
      #scale everything up to 0 for negative binormail test
      rescale_value = 0-min(agg_matdf$eigengene_expression)
      agg_matdf$eigengene_expression = agg_matdf$eigengene_expression+rescale_value
      
      #fit negative binomial GLM to eigengene testing for amount-tissue differences
      glmnb_res <- glm.nb(eigengene_expression~tissue, data = agg_matdf)
      
      #CLD on negative binomial GLM
      lsmeanres = data.frame(cld(emmeans(glmnb_res,specs = c("tissue"))))
      names(lsmeanres) = c("tissue","mean","SE","df","lower.CL","upper.CL","CLD")
      
      #lsmeanres$Kruskal_Wallis_p = kruskal.test(eigengene_expression~tissue, data=agg_matdf)$p.value #add K-W p-value to output
      
      lsmeanres$eigengene = names(balmain_genes)[i] #add column with the eigengene 
      lsmeanres$constituent_genes = paste(markder_df_sub$gene,collapse=", ")
      
      #scale everything up to 1 for correct percent change calculation
      rescale_value.2 = 1-min(lsmeanres$mean)
      lsmeanres$mean = lsmeanres$mean+rescale_value.2
      
      #abs change UnTxSkin --> Carcinoma
      abs_change = subset(lsmeanres, tissue=="CA")$mean - subset(lsmeanres, tissue=="UnTxSkin")$mean
      
      #% change UnTxSkin --> Carcinoma
      percent_change = ((subset(lsmeanres, tissue=="CA")$mean - subset(lsmeanres, tissue=="UnTxSkin")$mean) / abs(subset(lsmeanres, tissue=="UnTxSkin")$mean))*100
      
      data.frame(lsmeanres$eigengene[1], lsmeanres$constituent_genes[1], abs_change, percent_change) #result
      
    }))
    
    #order from high to low
    names(optimize_results)[1:2] = c("eigengene","constituent_genes")
    optimize_results.1 = optimize_results[order(optimize_results$percent_change),]
    optimize_results.1$index = as.vector(1:nrow(optimize_results.1))
    
    ascending_eigen = ggplot(optimize_results.1, aes(x=index, y=percent_change,color="darkred",shape="triangle")) +
      geom_point(alpha=0.3, size=1.5) +
      theme(legend.key = element_rect(fill = "white")) +
      theme(legend.text=element_text(size=14), legend.title=element_text(size=14)) +
      theme(strip.background=element_rect(fill="white")) +
      ggtitle(paste("Eigengene:",names(balmain_genes)[i],"Optimizer Performance",sep=" ")) +
      ylab("Optimizer: percent change Normal Skin to Carcinoma") +
      xlab("Trigenic combination index") +
      theme(text = element_text(size=18)) +
      theme(axis.text=element_text(size=18,color='black'),
            axis.title=element_text(size=20,color='black')) +
      theme(plot.title = element_text(size = 22, vjust=1)) +
      theme(axis.text=element_text(size=15,color='black'),
            axis.title=element_text(size=15,color='black')) +
      theme(panel.grid.major=element_line(colour = "darkgrey"), 
            panel.grid.minor=element_blank(), 
            panel.background=element_blank()) +
      theme(legend.key = element_rect(fill = "white")) +
      theme(strip.text.x = element_text(size = 15, colour = "black", angle = 0)) +
      theme(strip.text.y = element_text(size = 15, colour = "black", angle = 0)) +
      theme(legend.text=element_text(size=10), legend.title=element_text(size=15)) +
      scale_color_manual(values = c("darkgreen","red")) + 
      theme(legend.position="none") +
      theme(axis.text=element_text(size=8,color='black'),
            axis.title=element_text(size=10,color='black')) + theme(plot.title = element_text(size = 14, vjust=1))
    
    #optimal genes
    optimal_genes = data.frame(strsplit(as.character((optimize_results.1$constituent_genes[nrow(optimize_results.1)])),", "))[,1] #get the 3 genes with the biggest change
    marker_df_optimal = data.frame(optimal_genes,names(balmain_genes)[i],names(balmain_genes)[i])
    names(marker_df_optimal) = c("gene","cell_type","module")
    
    agexp = plot_cells(big_cds,
                       genes=marker_df_optimal,
                       label_cell_groups=FALSE,
                       label_leaves=F,
                       label_branch_points=F,
                       graph_label_size=1.5,
                       show_trajectory_graph=F) +
      xlab("Dim. 1") + ylab("Dim. 2") + 
      theme(legend.position="bottom") +
      scale_color_viridis_c(option = "inferno",name="integrated expression") +
      ggtitle(paste("Optimal ",names(balmain_genes)[i], " Eigengene: ",optimize_results.1$constituent_genes[nrow(optimize_results.1)],sep=""))+ #continuous
      theme(axis.text=element_text(size=8,color='black'),
            axis.title=element_text(size=10,color='black')) + theme(plot.title = element_text(size = 14, vjust=1))
    
    
    pdf(file=paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/balmain_eigene/optimization/three_gene/",names(balmain_genes)[i],".pdf",sep=""),
        height=10,width=20)
    print(multiplot(ascending_eigen,agexp,cols=2))
    dev.off()
    
    #re-order from high to low percentant change and then print out
    optimize_results.1 = optimize_results[order(-optimize_results$percent_change),]
    write.table(optimize_results.1, file=paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/balmain_eigene/optimization/three_gene/",names(balmain_genes)[i],"_optimizer_results.txt",sep=""),quote=F,sep="\t",row.names=F)
    
  } #end loop
  
  ##################################################################################################################################
  ##################################################################################################################################
  ##################################################################################################################################
  ########################################################################################################################
  ##################################################################################################################################
  ##################################################################################################################################
  #Plot individual genes within an eigengene
  
  i=334
  names(balmain_genes)[i]
  
  dir.create(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/balmain_eigene/individualgenes_PNGs_",names(balmain_genes)[i],sep=""))
  
  print(i)
  #only if more than 1 constituent gene
  marker_df = data.frame(na.omit(balmain_genes[,i]),names(balmain_genes)[i])
  names(marker_df) = c("gene","cell_type")
  
  #remake the marker df dataframe with the matched gene names
  marker_df = data.frame(unique(alias2SymbolTable(unique(marker_df$gene), species = "Mm")),names(balmain_genes)[i],names(balmain_genes)[i])
  names(marker_df) = c("gene","cell_type","module")
  marker_df = na.omit(marker_df) #remove non-matching genes
  
  #Loop through individual genes within an eigengene
  for(j in 1:nrow(marker_df)) {
    
    #j=12  #I had to manually increment and I don't know why
    print(j)
    
    png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/balmain_eigene/individualgenes_PNGs_",names(balmain_genes)[i],"/", marker_df$gene[j],".png",sep=""),width=6,height=6,units="in",res=400)
    
    print(plot_cells(big_cds,
                     cell_size=0.5,
                     genes=marker_df$gene[j],
                     label_cell_groups=FALSE,
                     label_leaves=F,
                     label_branch_points=F,
                     graph_label_size=1.5,
                     show_trajectory_graph=F,
                     alpha=0.8) +
            xlab("Dim. 1") + ylab("Dim. 2") + 
            theme(legend.position="bottom") +
            scale_color_viridis_c(option = "inferno",name="expression") +
            ggtitle(marker_df$gene[j])+ #continuous
            theme(axis.text=element_text(size=8,color='black'),
                  axis.title=element_text(size=10,color='black')) + theme(plot.title = element_text(size = 14, vjust=1)) +
            theme(strip.text.x = element_text(size = 0)) #remove facet label
    )
    dev.off()
    
  } 
  
  ##################################################################################################################################
  ##################################################################################################################################
  ##################################################################################################################################
  
  #Replace quantitative expression values with binary 0/1 for yes or no expression
  
  eigengene_agg = do.call(cbind, lapply(381:ncol(balmain_genes), function(i){
    tryCatch({ #suppress error
      
      #i=380
      print(paste(i,names(balmain_genes)[i],sep=" "))
      
      print(i)
      #only if more than 1 constituent gene
      marker_df = data.frame(na.omit(balmain_genes[,i]),names(balmain_genes)[i])
      names(marker_df) = c("gene","cell_type")
      
      #remake the marker df dataframe with the matched gene names
      marker_df = data.frame(unique(alias2SymbolTable(unique(marker_df$gene), species = "Mm")),names(balmain_genes)[i],names(balmain_genes)[i])
      names(marker_df) = c("gene","cell_type","module")
      marker_df = na.omit(marker_df) #remove non-matching genes
      
      #-----------------------------------------------------#
      #Full manifold
      #construct binary expression dataframe
      subcounts = t(as.matrix(big_cds@assays@data$counts[na.omit(match(marker_df$gene,row.names(big_cds@assays@data$counts))),]))  #exctract gene counts and transpose as a matrix
      subcounts = replace(subcounts, subcounts != 0, 1) #replace non-zero elements in the matrix with 1
      summed_binary_exprssion = data.frame(big_cds@int_colData@listData$reducedDims@listData$UMAP, rowSums(subcounts) ) #combine UMAP coordinates with summed binary expression across genes, per cell
      names(summed_binary_exprssion) = c("UMAP1","UMAP2","UMAP3","summed_binary_expression_value")
      
      #now make a column of whether a cell has full expression: ie does binary expression add up to number of genes
      full_expression_sum = length(marker_df$gene)
      summed_binary_exprssion$full_expression = ifelse(summed_binary_exprssion$summed_binary_expression==full_expression_sum,"Complete","Incomplete")
      names(summed_binary_exprssion)[ncol(summed_binary_exprssion)] = "CompleteExpression"
      
      #Now graph summed binary expression
      summed_binary_expression_plot = ggplot(data=summed_binary_exprssion, aes(x=UMAP1,y=UMAP2,color=summed_binary_expression_value)) +
        geom_point(alpha=0.4, size=0.2) +
        theme(axis.ticks = element_blank()) +
        ggtitle(paste("Summed binary expression: ", names(balmain_genes)[i],sep=" ")) +
        scale_color_viridis_c(option = "inferno",name="summed binary expression") +
        theme(legend.position="bottom") +
        theme(axis.text=element_text(size=0,color='black'),
              axis.title=element_text(size=0,color='black')) + 
        theme(plot.title = element_text(size = 15, vjust=0)) +
        theme(strip.text.x = element_text(size = 0)) + #remove facet label
        theme(panel.grid.major=element_blank(), 
              panel.grid.minor=element_blank(), 
              panel.background=element_blank()) +
        theme(axis.line = element_line(size = 0.3, colour = "black"))
      
      
      classic_eigengene_expression_plot = 
        plot_cells(big_cds,
                   genes=marker_df,
                   label_cell_groups=FALSE,
                   label_leaves=F,
                   label_branch_points=F,
                   graph_label_size=1.5,
                   show_trajectory_graph=F) +
        theme(axis.text=element_text(size=0,color='black'),
              axis.title=element_text(size=0,color='black')) + 
        theme(axis.ticks = element_blank()) +
        theme(plot.title = element_text(size = 15, vjust=0)) +
        theme(legend.position="bottom") +
        scale_color_viridis_c(option = "inferno",name="integrated expression") +
        ggtitle(paste("Classic eigengene: ", names(balmain_genes)[i],sep=" "))+ #continuous
        theme(strip.text.x = element_text(size = 0)) #remove facet label
      
      #-----------------------------------------------------#
      #Pap_ca manifold
      #construct binary expression dataframe
      subcounts = t(as.matrix(pap_ca@assays@data$counts[na.omit(match(marker_df$gene,row.names(pap_ca@assays@data$counts))),]))  #exctract gene counts and transpose as a matrix
      subcounts = replace(subcounts, subcounts != 0, 1) #replace non-zero elements in the matrix with 1
      summed_binary_exprssion = data.frame(pap_ca@int_colData@listData$reducedDims@listData$UMAP, rowSums(subcounts) ) #combine UMAP coordinates with summed binary expression across genes, per cell
      names(summed_binary_exprssion) = c("UMAP1","UMAP2","summed_binary_expression_value")
      
      #Now graph summed binary expression
      summed_binary_expression_plot_pap_ca = ggplot(data=summed_binary_exprssion, aes(x=UMAP1,y=UMAP2,color=summed_binary_expression_value)) +
        geom_point(alpha=0.8, size=1) +
        theme(axis.ticks = element_blank()) +
        ggtitle(paste("Summed binary expression: ", names(balmain_genes)[i],sep=" ")) +
        scale_color_viridis_c(option = "inferno",name="summed binary expression") +
        theme(legend.position="bottom") +
        theme(axis.text=element_text(size=0,color='black'),
              axis.title=element_text(size=0,color='black')) + 
        theme(plot.title = element_text(size = 15, vjust=0)) +
        theme(strip.text.x = element_text(size = 0)) + #remove facet label
        theme(panel.grid.major=element_blank(), 
              panel.grid.minor=element_blank(), 
              panel.background=element_blank()) +
        theme(axis.line = element_line(size = 0.3, colour = "black"))
      
      classic_eigengene_expression_plot_pap_ca = 
        plot_cells(pap_ca,
                   genes=marker_df,
                   label_cell_groups=FALSE,
                   label_leaves=F,
                   label_branch_points=F,
                   graph_label_size=1.5,
                   show_trajectory_graph=F,
                   cell_size=1) +
        theme(axis.text=element_text(size=0,color='black'),
              axis.title=element_text(size=0,color='black')) + 
        theme(axis.ticks = element_blank()) +
        theme(plot.title = element_text(size = 15, vjust=0)) +
        theme(legend.position="bottom") +
        scale_color_viridis_c(option = "inferno",name="integrated expression") +
        ggtitle(paste("Classic eigengene: ", names(balmain_genes)[i],sep=" "))+ #continuous
        theme(strip.text.x = element_text(size = 0)) #remove facet label
      
      
      #-----------#
      #print out binary and classic plots
      png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/balmain_eigene/eigengene_PNGs_v",eigengene_verison,"/eigengene_",names(balmain_genes)[i],".png",sep=""),
          width=14,height=15,units="in",res=400)
      
      print(multiplot(classic_eigengene_expression_plot,
                      summed_binary_expression_plot,
                      classic_eigengene_expression_plot_pap_ca,
                      summed_binary_expression_plot_pap_ca,
                      cols=2))
      
      dev.off()
      
    },  error=function(e){})
    
  }))
  
  ##################################################################################################################################
  ##################################################################################################################################
  ##################################################################################################################################
  ##################################################################################################################################
  #HERE IT IS: EIGENGENE SENSITIVITY ANALYSIS: SENSITIVITY OF EIGENGENE PATTERNS TO PARTICULAR CONSTITUENT GENEs
  
  #Bin cells -> calculate mean eigengene expression for each bin -> drop a gene -> recalculate -> see how dramatically it has changed
  
  #High exprssion zones are more important so possibly weight the change the high expression zones by mean expression?
  
  #AMPLITUDE = mean eigengene expression change
  #SPECIFICITY = ability of eigengenes to discriminate among celltype or cluster
  
  eigengene_agg = do.call(cbind, lapply(c(1:232,239:ncol(balmain_genes)), function(i){
    eigengene_agg = do.call(cbind, lapply(c(90:ncol(balmain_genes)), function(i){
      
      tryCatch({ #suppress error
        
        #i=172
        
        #only if more than 1 constituent gene
        marker_df = data.frame(na.omit(balmain_genes[,i]),names(balmain_genes)[i])
        names(marker_df) = c("gene","cell_type")
        
        #remake the marker df dataframe with the matched gene names
        marker_df = data.frame(unique(alias2SymbolTable(unique(marker_df$gene), species = "Mm")),names(balmain_genes)[i],names(balmain_genes)[i])
        names(marker_df) = c("gene","cell_type","module")
        marker_df = na.omit(marker_df) #remove non-matching genes
        
        #aggregate gene values to get a single expression value
        agg_mat = aggregate_gene_expression(
          big_cds,
          gene_group_df = marker_df,
          cell_group_df = NULL,
          norm_method = c("log"),
          pseudocount = 1,
          scale_agg_values = F,
          max_agg_value = NA,
          min_agg_value = NA,
          exclude.na = TRUE
        )
        
        #Cluster-specific BEFORE GENE DROPPED
        agg_gene_before = data.frame(c(agg_mat),colData(big_cds))
        names(agg_gene_before)[1] = c("eigengene_expression")
        dfc_before = summarySE(agg_gene_before, measurevar='eigengene_expression', groupvars=c("cluster","cell_type"))
        cluster_variance_before = var(dfc_before$eigengene_expression) #calculate variance among clusters
        
        #Celltype-specific BEFORE GENE DROPPED
        dfc_before_celltype = summarySE(agg_gene_before, measurevar='eigengene_expression', groupvars=c("cell_type"))
        celltype_variance_before = var(dfc_before_celltype$eigengene_expression) #calculate variance among clusters
        
        
        #Start loop to test effect of dropping a gene from the eigengene
        genes_dropped_effect = do.call(rbind, lapply(1:nrow(marker_df), function(j) {
          # genes_dropped_effect = do.call(rbind, lapply(1:5, function(j) {
          
          tryCatch({ #suppress error
            #j=1
            print(paste("i =",i,names(balmain_genes)[i],"j =",j,sep=" "))
            
            marker_df_sub = marker_df[-j,]
            nrow(marker_df_sub)
            
            agg_mat_sub = aggregate_gene_expression(
              big_cds,
              gene_group_df = marker_df_sub,
              cell_group_df = NULL,
              norm_method = c("log"),
              pseudocount = 1,
              scale_agg_values = F,
              max_agg_value = NA,
              min_agg_value = NA,
              exclude.na = TRUE
            )
            
            #Calculate eigengene expression
            agg_gene_after = data.frame(c(agg_mat_sub),colData(big_cds))
            names(agg_gene_after)[1] = c("eigengene_expression")
            dfc_after = summarySE(agg_gene_after, measurevar='eigengene_expression', groupvars=c("cluster","cell_type"))
            cluster_variance_after = var(dfc_after$eigengene_expression) #calculate variance among clusters
            
            #Celltype-specific BEFORE GENE DROPPED
            dfc_after_celltype = summarySE(agg_gene_after, measurevar='eigengene_expression', groupvars=c("cell_type"))
            celltype_variance_after = var(dfc_after_celltype$eigengene_expression) #calculate variance among clusters
            
            #calculate variance change
            cluster_variance_change = cluster_variance_after-cluster_variance_before
            celltype_variance_change = celltype_variance_after-celltype_variance_before
            
            #get eigengene expression change
            agg_gene_clindat_change = data.frame(c(agg_mat_sub-agg_mat),colData(big_cds)) #calculate cluster-specific change
            names(agg_gene_clindat_change)[1] = c("eigengene_expression_change")
            dfc = summarySE(agg_gene_clindat_change, measurevar='eigengene_expression_change', groupvars=c("cluster","cell_type"))
            
            gene_change_summary = data.frame(marker_df$gene[j],dfc[,c(1:4)],dfc$eigengene_expression_change*dfc_before$eigengene_expression,cluster_variance_change,celltype_variance_change)
            names(gene_change_summary) = c("gene_dropped","cell_cluster","cell_type","N","eigengene_expression_change",
                                           "eigengene_expression_change_weighted_expression_amount",
                                           "cluster_variance_change","celltype_variance_change")
            return(gene_change_summary)
          },  error=function(e){})
        }))
        
        #-------------------------------------------------------------------------------------------#
        #Overall MEAN GENE EXPRESSION (AMPLITUDE) change
        
        #unweighted change
        dfc_overall = summarySE(genes_dropped_effect, measurevar='eigengene_expression_change', groupvars=c("gene_dropped"))
        dfc_overall$absolute_value = abs(dfc_overall$eigengene_expression_change)
        dfc_overall = dfc_overall[order(-dfc_overall$absolute_value),] #high to low mean change
        
        #weighted change
        dfc_overall_weighted = summarySE(genes_dropped_effect, measurevar='eigengene_expression_change_weighted_expression_amount', groupvars=c("gene_dropped"))
        dfc_overall_weighted$absolute_value = abs(dfc_overall_weighted$eigengene_expression_change_weighted_expression_amount)
        dfc_overall_weighted = dfc_overall_weighted[order(-dfc_overall_weighted$absolute_value),] #high to low mean change
        
        #-------------------------------------------------------------------------------------------#
        #Cluster-specific MEAN GENE EXPRESSION (AMPLITUDE) 
        
        #unweighted change
        dfc_cluster_type = summarySE(genes_dropped_effect, measurevar='eigengene_expression_change', groupvars=c("gene_dropped","cell_cluster"))
        dfc_cluster_type$absolute_value = abs(dfc_cluster_type$eigengene_expression_change)
        dfc_cluster_type = dfc_cluster_type[order(-dfc_cluster_type$absolute_value),] #high to low mean change
        
        #weighted change
        dfc_cluster_type_weighted = summarySE(genes_dropped_effect, measurevar='eigengene_expression_change_weighted_expression_amount', groupvars=c("gene_dropped","cell_cluster"))
        dfc_cluster_type_weighted$absolute_value = abs(dfc_cluster_type_weighted$eigengene_expression_change_weighted_expression_amount)
        dfc_cluster_type_weighted = dfc_cluster_type_weighted[order(-dfc_cluster_type_weighted$absolute_value),] #high to low mean change
        
        #weighted change
        upperspike_unweighted = subset(dfc_cluster_type, cell_cluster==19)
        lowerspike_unweighted = subset(dfc_cluster_type, cell_cluster==33)
        upperspike_weighted = subset(dfc_cluster_type_weighted, cell_cluster==19)
        lowerspike_weighted = subset(dfc_cluster_type_weighted, cell_cluster==33)
        
        #-------------------------------------------------------------------------------------------#
        #Cell type-specific MEAN GENE EXPRESSION (AMPLITUDE) 
        dfc_cluster_type = summarySE(genes_dropped_effect, measurevar='eigengene_expression_change', groupvars=c("gene_dropped","cell_type"))
        dfc_cluster_type$absolute_value = abs(dfc_cluster_type$eigengene_expression_change)
        dfc_cluster_type = dfc_cluster_type[order(-dfc_cluster_type$absolute_value),] #high to low mean change
        
        #weighted change
        dfc_cluster_type_weighted = summarySE(genes_dropped_effect, measurevar='eigengene_expression_change_weighted_expression_amount', groupvars=c("gene_dropped","cell_type"))
        dfc_cluster_type_weighted$absolute_value = abs(dfc_cluster_type_weighted$eigengene_expression_change_weighted_expression_amount)
        dfc_cluster_type_weighted = dfc_cluster_type_weighted[order(-dfc_cluster_type_weighted$absolute_value),] #high to low mean change
        
        #weighted change
        nep_unweighted = subset(dfc_cluster_type, cell_type=="nEp")
        nep_weighted = subset(dfc_cluster_type_weighted, cell_type=="nEp")
        
        #-------------------------------------------------------------------------------------------#
        #get variance change from high to low loss of specificity
        ordered_cluster_var_change = genes_dropped_effect[ order(genes_dropped_effect$cluster_variance_change), ] #a loss of specificity would be NEGATIVE bc after_variance - before_variance; so we need to order from low to high
        cluster_change_genes = unique(ordered_cluster_var_change$gene_dropped)
        
        ordered_celltype_var_change = genes_dropped_effect[ order(genes_dropped_effect$celltype_variance_change), ] #a loss of specificity would be NEGATIVE bc after_variance - before_variance; so we need to order from low to high
        celltype_change_genes = unique(ordered_celltype_var_change$gene_dropped)
        
        
        #-------------------------------------------------------------------------------------------#
        #Combine all the results
        
        combined_result_genes = data.frame(cluster_change_genes,
                                           celltype_change_genes,
                                           dfc_overall_weighted$gene_dropped,
                                           upperspike_weighted$gene_dropped,
                                           lowerspike_weighted$gene_dropped,
                                           nep_weighted$gene_dropped)
        
        names(combined_result_genes) = c("cluster_specificity",
                                         "celltype_specificity",
                                         "overall_amplitude",
                                         "upperspike_amplitude",
                                         "lowerspike_amplitude",
                                         "nep_amplitude")
        
        write.table(combined_result_genes,
                    file=paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/eigengene_sensitivity_analysis/",names(balmain_genes)[i],"_sensitivity.txt",sep=""),
                    row.names = F, quote=F, sep="\t")
      },  error=function(e){})
      
    }))
    
    #-------------------------------------------------------------------------------------------#
    #-------------------------------------------------------------------------------------------#
    #-------------------------------------------------------------------------------------------#
    
    #Now graph the effect of dropping genes
    los_files = list.files(path="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/eigengene_sensitivity_analysis",
                           full.names = T)
    los_files_short = list.files("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/eigengene_sensitivity_analysis",
                                 full.names = F)
    
    #start loop through eigengenes
    for(i in 45:48) {
      tryCatch({ #suppress error
        
        #i=45
        
        #create directory where figs will be printed
        output_dir = paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/eigengene_sensitivity_analysis/",
                           strsplit(los_files_short[i],".txt")[[1]][1],
                           sep="")
        dir.create(output_dir)
        
        #Read in file  
        marker_file = read.delim(file=los_files[i], header=T)
        
        #Cluster specificity: full eigengene
        marker_df = data.frame(marker_file$cluster_specificity,"dummary_var","dummary_var")
        names(marker_df) = c("gene","cell_type","module")
        
        full_eigengene_plot = plot_cells(big_cds,
                                         genes=marker_df,
                                         label_cell_groups=FALSE,
                                         label_leaves=F,
                                         label_branch_points=F,
                                         graph_label_size=1.5,
                                         show_trajectory_graph=F,
                                         cell_size=0.4,
                                         alpha=0.4) +
          theme(legend.position="bottom") +
          scale_color_viridis_c(option = "inferno",name="integrated expression") +
          ggtitle("Full eigengene")+ #continuous
          theme(plot.title = element_text(size = 22, vjust=1)) +
          theme(axis.text=element_text(size=0,color='black'),
                axis.title=element_text(size=0),
                axis.ticks = element_line(size=0)) +
          theme(strip.text.x = element_text(size = 0)) #remove facet label
        
        #Print out full eigengene plot
        png(paste(output_dir,"/","0_gene_reduction_full_eigengene.png",sep=""),
            width=7.5,height=7.5,units="in",res=250)
        print(full_eigengene_plot)
        dev.off()
        
        
        #Start loop dropping single and cimulative genes based on sensitviity analysis doen above
        
        for(j in 1:nrow(marker_file)) {
          tryCatch({ #suppress error
            
            print(paste("i=",i," j=",j,sep=""))
            #j=8
            #----------------------------------------------------------------------------------#
            #SINGLE-GENE REDUCTION
            marker_df_sub = marker_file[-j,] #drop a single gene
            
            #Cluster specificity
            marker_df = data.frame(marker_df_sub$cluster_specificity,"dummary_var","dummary_var")
            names(marker_df) = c("gene","cell_type","module")
            
            cluster_specificity_one_dropped = plot_cells(big_cds,
                                                         genes=marker_df,
                                                         label_cell_groups=FALSE,
                                                         label_leaves=F,
                                                         label_branch_points=F,
                                                         graph_label_size=1.5,
                                                         show_trajectory_graph=F,
                                                         cell_size=0.4,
                                                         alpha=0.4) +
              theme(legend.position="bottom") +
              scale_color_viridis_c(option = "inferno",name="integrated expression") +
              ggtitle(paste("Specificity Single-gene reduction ", j,"th ",marker_file$cluster_specificity[j],sep=""))+ #continuous
              theme(plot.title = element_text(size = 22, vjust=1)) +
              theme(axis.text=element_text(size=0,color='black'),
                    axis.title=element_text(size=0),
                    axis.ticks = element_line(size=0)) +
              theme(strip.text.x = element_text(size = 0)) #remove facet label
            
            marker_df = data.frame(marker_df_sub$overall_amplitude,"dummary_var","dummary_var")
            names(marker_df) = c("gene","cell_type","module")
            
            mean_one_dropped = plot_cells(big_cds,
                                          genes=marker_df,
                                          label_cell_groups=FALSE,
                                          label_leaves=F,
                                          label_branch_points=F,
                                          graph_label_size=1.5,
                                          show_trajectory_graph=F,
                                          cell_size=0.4,
                                          alpha=0.4) +
              theme(legend.position="bottom") +
              scale_color_viridis_c(option = "inferno",name="integrated expression") +
              ggtitle(paste("Amplitude Single-gene reduction ", j,"th ",marker_file$overall_amplitude[j],sep=""))+ #continuous
              theme(plot.title = element_text(size = 22, vjust=1)) +
              theme(axis.text=element_text(size=0,color='black'),
                    axis.title=element_text(size=0),
                    axis.ticks = element_line(size=0)) +
              theme(strip.text.x = element_text(size = 0)) #remove facet label
            
            #----------------------------------------------------------------------------------#
            #MULTIPLE-GENE REDUCTION
            marker_df_sub = marker_file[-c(1:j),] #drop a single gene
            
            #Cluster specificity
            marker_df = data.frame(marker_df_sub$cluster_specificity,"dummary_var","dummary_var")
            names(marker_df) = c("gene","cell_type","module")
            
            cluster_specificity_cumulative_dropped = plot_cells(big_cds,
                                                                genes=marker_df,
                                                                label_cell_groups=FALSE,
                                                                label_leaves=F,
                                                                label_branch_points=F,
                                                                graph_label_size=1.5,
                                                                show_trajectory_graph=F,
                                                                cell_size=0.4,
                                                                alpha=0.4) +
              theme(legend.position="bottom") +
              scale_color_viridis_c(option = "inferno",name="integrated expression") +
              ggtitle(paste("Specificity cumulative reduction ", j,"th ",marker_file$cluster_specificity[j],sep=""))+ #continuous
              theme(plot.title = element_text(size = 22, vjust=1)) +
              theme(axis.text=element_text(size=0,color='black'),
                    axis.title=element_text(size=0),
                    axis.ticks = element_line(size=0)) +
              theme(strip.text.x = element_text(size = 0)) #remove facet label
            
            marker_df = data.frame(marker_df_sub$overall_amplitude,"dummary_var","dummary_var")
            names(marker_df) = c("gene","cell_type","module")
            
            mean_cumulative_dropped = plot_cells(big_cds,
                                                 genes=marker_df,
                                                 label_cell_groups=FALSE,
                                                 label_leaves=F,
                                                 label_branch_points=F,
                                                 graph_label_size=1.5,
                                                 show_trajectory_graph=F,
                                                 cell_size=0.4,
                                                 alpha=0.4) +
              theme(legend.position="bottom") +
              scale_color_viridis_c(option = "inferno",name="integrated expression") +
              ggtitle(paste("Amplitude cumulative reduction ", j,"th ",marker_file$overall_amplitude[j],sep=""))+ #continuous
              theme(plot.title = element_text(size = 22, vjust=1)) +
              theme(axis.text=element_text(size=0,color='black'),
                    axis.title=element_text(size=0),
                    axis.ticks = element_line(size=0)) +
              theme(strip.text.x = element_text(size = 0)) #remove facet label
            
            png(paste(output_dir,"/",j,"_gene_reduction.png",sep=""),
                width=15,height=15,units="in",res=250)
            
            print(multiplot(cluster_specificity_one_dropped, 
                            mean_one_dropped,
                            cluster_specificity_cumulative_dropped,
                            mean_cumulative_dropped,
                            cols=2))
            dev.off()
            
          },  error=function(e){})
          
        } #End manifold loop
        
      },  error=function(e){})
      
    } #End loop through eigengenes
    
    
    ##################################################################################################################################
    ##################################################################################################################################
    ##################################################################################################################################
    ##################################################################################################################################
    #What genes are actually mapping in these data
    
    gene_summaries = read.table(file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/gene_summaeries/gene_summaries.txt",header=T)
    
    eigengene_agg = do.call(cbind, lapply(349:ncol(balmain_genes), function(i){
      tryCatch({ #suppress error
        
        #i=172
        print(paste(i,names(balmain_genes)[i],sep=" "))
        
        print(i)
        #only if more than 1 constituent gene
        marker_df = data.frame(na.omit(balmain_genes[,i]),names(balmain_genes)[i])
        names(marker_df) = c("gene","cell_type")
        
        #remake the marker df dataframe with the matched gene names
        marker_df = data.frame(unique(alias2SymbolTable(unique(marker_df$gene), species = "Mm")),names(balmain_genes)[i],names(balmain_genes)[i])
        names(marker_df) = c("gene","cell_type","module")
        marker_df = na.omit(marker_df) #remove non-matching genes
        
        data.frame( na.omit(gene_summaries[match(marker_df$gene,gene_summaries$gene_name),]) )
        
        ##################################################################################################################################
        ##################################################################################################################################
        ##################################################################################################################################
        #Test eigengene expression differences among cell types for the top DEG eigengenes
        
        eigengene_agg_test_differences = do.call(rbind, lapply(334:ncol(balmain_genes), function(i){
          
          #subset eigengene_agg = do.call(cbind, lapply(172:175, function(i){
          
          tryCatch({ #suppress error
            
            #i=335
            print(i)
            #only if more than 1 constituent gene
            marker_df = data.frame(na.omit(balmain_genes[,i]),names(balmain_genes)[i])
            names(marker_df) = c("gene","cell_type")
            
            #remake the marker df dataframe with the matched gene names
            marker_df = data.frame(unique(alias2SymbolTable(unique(marker_df$gene), species = "Mm")),names(balmain_genes)[i],names(balmain_genes)[i])
            names(marker_df) = c("gene","cell_type","module")
            marker_df = na.omit(marker_df) #remove non-matching genes
            
            #aggregate gene values to get a single expression value
            agg_mat = aggregate_gene_expression(
              big_cds,
              gene_group_df = marker_df,
              cell_group_df = NULL,
              norm_method = c("log"),
              pseudocount = 1,
              scale_agg_values = F,
              max_agg_value = NA,
              min_agg_value = NA,
              exclude.na = TRUE
            )
            
            #combine aggregated values with CNV calls
            agg_gene_clindat = data.frame(c(agg_mat),colData(big_cds))
            names(agg_gene_clindat)[1] = c("eigengene_expression")
            
            #Subset to papilloma parenchyma only: nEp, lower spike, upper spike
            agg_gene_clindat_sub = subset(agg_gene_clindat, (tissue_updated=="Pap" & (integrated_celltype_spike=="nEp" | integrated_celltype_spike=="upper spike" | integrated_celltype_spike=="lower spike")  ))
            
            #Now graph comparisons
            my_comparisons <- list( c("lower spike", "nEp"), c("nEp", "upper spike"), c("lower spike", "upper spike"))
            
            comparison_res_fig = ggplot(agg_gene_clindat_sub, aes(x=integrated_celltype_spike,y=eigengene_expression, fill=integrated_celltype_spike)) +
              stat_compare_means(comparisons = my_comparisons) +
              geom_violin(aes(fill=integrated_celltype_spike), width=1) +
              geom_boxplot(outlier.size=0, width=0.2, color="white") +
              #geom_bar(stat = "identity", fill="darkorange") +
              #facet_wrap(~parenchyma, scales="free", nrow=1) + 
              #geom_jitter(width = 0.1) +
              ggtitle(names(balmain_genes)[i]) + 
              xlab("papilloma parenchyma") + ylab("eigengene expression") + 
              theme(axis.text.x=element_text(size=15,color='black'),
                    axis.text.y=element_text(size=10,color='black'),
                    axis.title=element_text(size=15),
                    axis.ticks.x = element_line(size=0),
                    strip.text.x = element_text(size = 15),
                    legend.position="none",
                    title = element_text(size = 15, face="bold")) +
              theme(
                # Hide panel borders and remove grid lines
                panel.border = element_blank(),
                panel.grid.major = element_blank(),
                panel.grid.minor = element_blank(),
                panel.background = element_blank(),
                element_line(colour = 'black', size = 0.3),
                # Change axis line
                axis.line = element_line(colour = "black"),
                strip.background = element_rect(colour="white",
                                                fill="white")
              ) +
              scale_fill_manual(values = c("lower spike" = "red",
                                           "nEp" = "purple", 
                                           "upper spike" = "orange")) + theme(legend.position="none")
            
            #-----------#
            #print out binary and classic plots
            png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/DEG/eigengenes/DEG_eigengenes/",names(balmain_genes)[i],".png",sep=""),
                width=8,height=8,units="in",res=200)
            
            print(comparison_res_fig)
            
            dev.off()
            
          },  error=function(e){})
          
        }))
        
        
        
        ##################################################################################################################################
        ##################################################################################################################################
        ##################################################################################################################################
        ##################################################################################################################################
        ##################################################################################################################################
        ##################################################################################################################################
        #Test sparsity threshold:  how many genes are still in the data when I increase the threshold:
        rm(our_cds)
        #Fraction of cells
        thresholds = seq(0,1,by=0.01)
        
        #before running this, make sure to unload the package SeuratObject and reload Seurat
        gene_distributions = do.call(rbind, lapply(1:length(thresholds), function(i){
          tryCatch({ #suppress error
            
            #i=11
            #Min cells a gene must be expressed in
            no_cells = dim(big_cds)[2] #get number of cells in this subset
            min.cells_sub = round(no_cells*thresholds[i]) #get min number
            
            print(i)
            #now create Seurat object from the patient-condition specific object
            seurat_obj = CreateSeuratObject(
              big_cds@assays@data$counts,
              project = "CreateSeuratObject",
              assay = "RNA",
              names.field = 1,
              names.delim = "_",
              meta.data = NULL,
              min.cells = min.cells_sub)
            
            passing_genes = dim(seurat_obj)[1] #number of genes passing the min.cells filter
            
            gene_number = data.frame(thresholds[i], passing_genes)
            gene_number
            
          },  error=function(e){})
        }))
        
        names(gene_distributions)[1] = c("min_cell_percent_threshold")
        gene_distributions = data.frame(round(gene_distributions$min_cell_percent_threshold*dim(big_cds)[2]),gene_distributions)
        names(gene_distributions)[1] = c("min_cell_number_threshold")
        
        #Plot thresholds
        head(gene_distributions)
        ggplot(gene_distributions, aes(x=min_cell_percent_threshold, y=passing_genes), color="darkorange") +
          geom_point() +
          geom_line() +
          ggtitle("No. Mapping Genes across Sparsity Thresholds") +
          xlab("Fraction of cells gene must be expresed in") +
          ylab("No. genes") +
          theme(axis.text=element_text(size=10,color='black'),
                axis.title=element_text(size=12,color='black')) +
          theme(plot.title = element_text(size = 15, vjust=1)) +
          theme(panel.grid.major=element_line(colour = "darkgrey"), 
                panel.grid.minor=element_blank(), 
                panel.background=element_blank()) +
          theme(legend.key = element_rect(fill = "white")) +
          theme(legend.text=element_text(size=12), legend.title=element_text(size=12)) +
          theme(legend.position="bottom")
        
        
        #---------------------------------------------------------------------------------------------------------------------------------------------#
        #---------------------------------------------------------------------------------------------------------------------------------------------#
        #---------------------------------------------------------------------------------------------------------------------------------------------#
        #---------------------------------------------------------------------------------------------------------------------------------------------#
        #---------------------------------------------------------------------------------------------------------------------------------------------#
        #---------------------------------------------------------------------------------------------------------------------------------------------#
        #---------------------------------------------------------------------------------------------------------------------------------------------#
        #WELCOME TO THE EIGENGENOME: HERE ALL GENE INTERACTIONS SHALL BE UNDERSTOOD
        #Expression of all eigengenes presentin these data
        
        #Find genes present in at least 1% of cells
        
        #Min cells a gene must be expressed in
        no_cells = dim(big_cds)[2] #get number of cells in this subset
        min.cells_sub = round(no_cells*0.0001) #get min number
        
        #now create Seurat object from the patient-condition specific object
        seurat_obj = CreateSeuratObject(
          big_cds@assays@data$counts,
          project = "CreateSeuratObject",
          assay = "RNA",
          names.field = 1,
          names.delim = "_",
          meta.data = NULL,
          min.cells = 1)
        
        seed_genes = row.names(seurat_obj)
        rm(seurat_obj)
        
        #find overlapping genes in microarray and scRNAseq data
        
        #microarray genes
        bulk_genes = alias2SymbolTable(unique(aim2probedat$SYMBOL),species = "Mm")
        
        #find overlaps between bulk microarray genes and single-cell genes
        length(na.omit(match(seed_genes,bulk_genes)))
        seed_genes_to_use = na.omit(bulk_genes[match(seed_genes,bulk_genes)])
        seed_genes_to_use = data.frame(seed_genes_to_use)
        names(seed_genes_to_use) = "gene_name"
        write.table(data.frame(seed_genes_to_use), 
                    file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/Eigengenome/seed_genes.txt",
                    quote=F,sep="\t",row.names=F)
        seed_genes_to_use
        
        eigengene_size=100
        
        eigengenome_expression_patterns = do.call(rbind, lapply(1:length(seed_genes_to_use), function(i){
          #eigengenome_expression_patterns = do.call(rbind, lapply(1:5, function(i){
          
          #subset eigengene_agg = do.call(cbind, lapply(172:175, function(i){
          
          tryCatch({ #suppress error
            
            #i=1
            print(paste(i,seed_genes_to_use[i],sep=" "))
            
            #Find some aliases
            match_index = match(seed_genes_to_use[i], aim2probedat$SYMBOL) #match seed gene to its place in the matrix
            
            #---------------------------------------#
            #Normal skin
            symbol_match = data.frame(aim2probedat$SYMBOL, rhodf[,match_index], p_df[,match_index]) #put together gene names, expression data, and p-vals
            symbol_match = symbol_match[order(-symbol_match$rhodf...match_index.),] #order from high to low gene values
            names(symbol_match) = c("gene","rho","p")
            
            #make marker df for grahing eigengene
            marker_df = data.frame(unique(alias2SymbolTable(symbol_match$gene[1:(eigengene_size+1)], species = "Mm")),paste(seed_genes_to_use[i]," Normal skin",sep=""),paste(seed_genes_to_use[i]," Normal skin",sep=""))
            names(marker_df) = c("gene","cell_type","module")
            
            #now subset to network (eigengene) that you will create the eigengen with
            #try creating eigengene from top 1% of significant network correlations
            normal_skin_genes_for_eigengene = symbol_match[1:eigengene_size,]
            write.table(normal_skin_genes_for_eigengene, file=paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/Eigengenome/eigengene_constituents/",seed_genes_to_use[i],"_normal_skin_eigengene.txt",sep=""), quote=F, row.names=F, sep="\t")
            
            #------------------------------#
            #4a. Normal skin WT
            #------------------------------#
            #all cells: Seed gene
            agexp_plot_seedgene.12 = plot_cells(big_cds,
                                                genes=marker_df$gene[1], #seed gene is the first row of the markers file
                                                label_cell_groups=FALSE,
                                                label_leaves=F,
                                                label_branch_points=F,
                                                graph_label_size=1.5,
                                                show_trajectory_graph=F,
                                                alpha=0.4) +
              xlab("Dim. 1") + ylab("Dim. 2") + 
              theme(legend.position="bottom") +
              scale_color_viridis_c(option = "inferno",name="expression") +
              ggtitle(paste(seed_genes_to_use[i],"seed gene",sep=" "))+ #continuous
              theme(axis.text=element_text(size=0,color='black'),
                    axis.title=element_text(size=0,color='black')) + 
              theme(plot.title = element_text(size = 15, vjust=0)) +
              theme(strip.text.x = element_text(size = 0)) + #remove facet label
              theme(panel.grid.major=element_blank(), 
                    panel.grid.minor=element_blank(), 
                    panel.background=element_blank()) +
              theme(axis.line = element_line(size = 0.3, colour = "black")) +
              theme(axis.ticks = element_blank())
            
            
            median_rho = round(summary(normal_skin_genes_for_eigengene$rho)[3],3)
            #all cells eigengene
            marker_df = na.omit(marker_df)
            agexp_plot.12 = plot_cells(big_cds,
                                       genes=marker_df,
                                       label_cell_groups=FALSE,
                                       label_leaves=F,
                                       label_branch_points=F,
                                       graph_label_size=1.5,
                                       show_trajectory_graph=F,
                                       alpha=0.4) +
              xlab("Dim. 1") + ylab("Dim. 2") + 
              theme(legend.position="bottom") +
              scale_color_viridis_c(option = "inferno",name="integrated expression") +
              ggtitle(paste(seed_genes_to_use[i],"NSk eigengene, med rho =",as.vector(median_rho),sep=" "))+ #continuous
              theme(axis.text=element_text(size=0,color='black'),
                    axis.title=element_text(size=0,color='black')) + 
              theme(plot.title = element_text(size = 15, vjust=0)) +
              theme(strip.text.x = element_text(size = 0)) + #remove facet label
              theme(panel.grid.major=element_blank(), 
                    panel.grid.minor=element_blank(), 
                    panel.background=element_blank()) +
              theme(axis.line = element_line(size = 0.3, colour = "black")) +
              theme(axis.ticks = element_blank())
            
            #construct binary expression dataframe
            subcounts = t(as.matrix(big_cds@assays@data$counts[na.omit(match(marker_df$gene,row.names(big_cds@assays@data$counts))),]))  #exctract gene counts and transpose as a matrix
            subcounts = replace(subcounts, subcounts != 0, 1) #replace non-zero elements in the matrix with 1
            summed_binary_exprssion = data.frame(big_cds@int_colData@listData$reducedDims@listData$UMAP, rowSums(subcounts) ) #combine UMAP coordinates with summed binary expression across genes, per cell
            names(summed_binary_exprssion) = c("UMAP1","UMAP2","UMAP3","summed_binary_expression_value")
            
            #Now graph summed binary expression
            summed_binary_expression_plot = ggplot(data=summed_binary_exprssion, aes(x=UMAP1,y=UMAP2,color=summed_binary_expression_value)) +
              geom_point(alpha=0.4, size=0.2) +
              theme(axis.ticks = element_blank()) +
              ggtitle(paste("Summed binary expression NSk",sep=" ")) +
              scale_color_viridis_c(option = "inferno",name="summed binary expression") +
              theme(legend.position="bottom") +
              theme(axis.text=element_text(size=0,color='black'),
                    axis.title=element_text(size=0,color='black')) + 
              theme(plot.title = element_text(size = 15, vjust=0)) +
              theme(strip.text.x = element_text(size = 0)) + #remove facet label
              theme(panel.grid.major=element_blank(), 
                    panel.grid.minor=element_blank(), 
                    panel.background=element_blank()) +
              theme(axis.line = element_line(size = 0.3, colour = "black"))
            
            #aggregate gene values to get a single expression value
            agg_mat = aggregate_gene_expression(
              big_cds,
              gene_group_df = marker_df,
              cell_group_df = NULL,
              norm_method = c("log"),
              pseudocount = 1,
              scale_agg_values = F,
              max_agg_value = NA,
              min_agg_value = NA,
              exclude.na = TRUE
            )
            
            #combine aggregated values with CNV calls
            agg_gene_clindat = data.frame(c(agg_mat),colData(big_cds))
            names(agg_gene_clindat)[1] = c("eigengene_expression")
            
            dfc = summarySE(agg_gene_clindat, measurevar='eigengene_expression', groupvars=c('cluster'))
            dfc.reord = dfc[order(-dfc$eigengene_expression),] #re-order from high to low eigengene expression
            
            topcluster1 = dfc.reord$cluster[1]
            topcluster2 = dfc.reord$cluster[2]
            topcluster3 = dfc.reord$cluster[3]
            
            nsk_eigengene_expression = data.frame(seed_genes_to_use[i],
                                                  "NSk",
                                                  topcluster1,topcluster2,topcluster3,
                                                  t(dfc$eigengene_expression)
            )
            names(nsk_eigengene_expression)[1:2] = c("seed_gene","tissue_network")
            
            #######################################################################################################################
            #######################################################################################################################
            #---------------------------------------#
            #tumor
            match_index = match(seed_genes_to_use[i], aim2probedat$SYMBOL) #match seed gene to its place in the matrix
            symbol_match = data.frame(aim2probedat$SYMBOL, tumor_rhodf[,match_index], tumor_p_df[,match_index]) #put together gene names, expression data, and p-vals
            symbol_match = symbol_match[order(-symbol_match$tumor_rhodf...match_index.),] #order from high to low gene values
            names(symbol_match) = c("gene","rho","p")
            
            #make marker df for grahing eigengene
            marker_df = data.frame(unique(alias2SymbolTable(symbol_match$gene[1:(eigengene_size+1)], species = "Mm")),paste(seed_genes_to_use[i],"Carcinoma",sep=""),paste(seed_genes_to_use[i],"Carcinoma",sep=""))
            names(marker_df) = c("gene","cell_type","module")
            
            #now subset to network (eigengene) that you will create the eigengen with
            tumor_genes_for_eigengene = symbol_match[1:eigengene_size,]
            write.table(tumor_genes_for_eigengene, file=paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/Eigengenome/eigengene_constituents/",seed_genes_to_use[i],"_tumor_eigengene.txt",sep=""), , quote=F, row.names=F, sep="\t")
            
            median_rho_tumor = round(summary(tumor_genes_for_eigengene$rho)[3],3)
            #all cells eigengene
            marker_df = na.omit(marker_df)
            agexp_plot_tumor.12 = plot_cells(big_cds,
                                             genes=marker_df,
                                             label_cell_groups=FALSE,
                                             label_leaves=F,
                                             label_branch_points=F,
                                             graph_label_size=1.5,
                                             show_trajectory_graph=F,
                                             alpha=0.4) +
              xlab("Dim. 1") + ylab("Dim. 2") + 
              theme(legend.position="bottom") +
              scale_color_viridis_c(option = "inferno",name="integrated expression") +
              ggtitle(paste(seed_genes_to_use[i],"tumor eigengene, med rho =",median_rho_tumor,sep=" "))+ #continuous
              theme(axis.text=element_text(size=0,color='black'),
                    axis.title=element_text(size=0,color='black')) + 
              theme(plot.title = element_text(size = 15, vjust=0)) +
              theme(strip.text.x = element_text(size = 0)) + #remove facet label
              theme(panel.grid.major=element_blank(), 
                    panel.grid.minor=element_blank(), 
                    panel.background=element_blank()) +
              theme(axis.line = element_line(size = 0.3, colour = "black")) +
              theme(axis.ticks = element_blank())
            
            #construct binary expression dataframe
            subcounts = t(as.matrix(big_cds@assays@data$counts[na.omit(match(marker_df$gene,row.names(big_cds@assays@data$counts))),]))  #exctract gene counts and transpose as a matrix
            subcounts = replace(subcounts, subcounts != 0, 1) #replace non-zero elements in the matrix with 1
            summed_binary_exprssion = data.frame(big_cds@int_colData@listData$reducedDims@listData$UMAP, rowSums(subcounts) ) #combine UMAP coordinates with summed binary expression across genes, per cell
            names(summed_binary_exprssion) = c("UMAP1","UMAP2","UMAP3","summed_binary_expression_value")
            
            #Now graph summed binary expression
            summed_binary_expression_plot_tumor = ggplot(data=summed_binary_exprssion, aes(x=UMAP1,y=UMAP2,color=summed_binary_expression_value)) +
              geom_point(alpha=0.4, size=0.2) +
              theme(axis.ticks = element_blank()) +
              ggtitle(paste("Summed binary expression tumor",sep="")) +
              scale_color_viridis_c(option = "inferno",name="summed binary expression") +
              theme(legend.position="bottom") +
              theme(axis.text=element_text(size=0,color='black'),
                    axis.title=element_text(size=0,color='black')) + 
              theme(plot.title = element_text(size = 15, vjust=0)) +
              theme(strip.text.x = element_text(size = 0)) + #remove facet label
              theme(panel.grid.major=element_blank(), 
                    panel.grid.minor=element_blank(), 
                    panel.background=element_blank()) +
              theme(axis.line = element_line(size = 0.3, colour = "black"))
            
            #aggregate gene values to get a single expression value
            agg_mat = aggregate_gene_expression(
              big_cds,
              gene_group_df = marker_df,
              cell_group_df = NULL,
              norm_method = c("log"),
              pseudocount = 1,
              scale_agg_values = F,
              max_agg_value = NA,
              min_agg_value = NA,
              exclude.na = TRUE
            )
            
            #combine aggregated values with CNV calls
            agg_gene_clindat = data.frame(c(agg_mat),colData(big_cds))
            names(agg_gene_clindat)[1] = c("eigengene_expression")
            
            dfc = summarySE(agg_gene_clindat, measurevar='eigengene_expression', groupvars=c('cluster'))
            dfc.reord = dfc[order(-dfc$eigengene_expression),] #re-order from high to low eigengene expression
            
            topcluster1 = dfc.reord$cluster[1]
            topcluster2 = dfc.reord$cluster[2]
            topcluster3 = dfc.reord$cluster[3]
            
            tumor_eigengene_expression = data.frame(seed_genes_to_use[i],
                                                    "carcinoma",
                                                    topcluster1,topcluster2,topcluster3,
                                                    t(dfc$eigengene_expression)
            )
            names(tumor_eigengene_expression)[1:2] = c("seed_gene","tissue_network")
            
            #------------------------------------------------------------------------------------#
            #Start graphing
            
            png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/Eigengenome/manifolds/",seed_genes_to_use[i],".png",sep=""),
                width=12,height=9,units="in",res=200)
            
            print(multiplot(agexp_plot_seedgene.12,
                            agexp_plot.12, summed_binary_expression_plot,
                            agexp_plot_tumor.12, summed_binary_expression_plot_tumor
                            ,cols=3))
            
            dev.off()
            
            #------------------------------------------------------------------------------------#
            #write_out_eigengene_patterns
            combined_patterns = rbind(nsk_eigengene_expression,
                                      tumor_eigengene_expression)
            names(combined_patterns)[6:54] = paste("cluster_",seq(1:49),sep="")
            
            write.table(combined_patterns, file=paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/Eigengenome/eigengene_patterns_cell_clusters/",
                                                      seed_genes_to_use[i],".txt",sep=""),
                        sep="\t", row.names=F, quote=F)
            
            #------------------------------------------------------------------------------------#
            #Return the lovely eigengene expression patterns
            return(
              combined_patterns
            )
            
          },  error=function(e){})
          
        }))   
        
        #Write out integrated results
        names(eigengenome_expression_patterns)[6:54] = paste("cluster_",seq(1:49),sep="")
        write.table(eigengenome_expression_patterns, 
                    file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/Eigengenome/eigenome_expression_patterns.txt",
                    quote=F, row.names=F, sep="\t")
        
        
        #---------------------------------------------------------------------------------------------------------------------------------------------#
        #---------------------------------------------------------------------------------------------------------------------------------------------#
        #---------------------------------------------------------------------------------------------------------------------------------------------#
        #---------------------------------------------------------------------------------------------------------------------------------------------#
        
        ##################################################################################################################################
        ##################################################################################################################################        
        ##################################################################################################################################
        ##################################################################################################################################
        ##################################################################################################################################
        #Fig. 2 : Seed gene + Eigengene manifolds
        
        #selected eigengenes
        fig2_genes = read.xlsx(
          xlsxFile="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Figs/Fig2/stem_cell_markers_for_allan.xlsx",
          colNames=T
        )
        
        seed_gene = fig2_genes$Fig_2a
        match(fig2_genes$Fig_2c,row.names(big_cds)) #oh fuck yes all these genes exist in these data!
        #get the correct prom1 probe
        aim2probedat[grep("Prom1", aim2probedat$SYMBOL),]#oh fuck yes all these genes exist in these data!
        eigengene_size=99
        
        #Loop through all eigengenes
        eigengene_agg = do.call(cbind, lapply(1:length(seed_gene), function(i){
          tryCatch({ #suppress error
            
            #######################################################################################################################
            #######################################################################################################################
            #---------------------------------------#
            #i=1
            print(i)
            
            #Get the correct match_index
            if( seed_gene[i]=="Prom1" ) {match_index = 12747} else {
              match_index = match(seed_gene[i], aim2probedat$SYMBOL) #match seed gene to its place in the matrix
            }
            
            #---------------------------------------#
            #Normal skin
            symbol_match = data.frame(aim2probedat$SYMBOL, rhodf[,match_index], p_df[,match_index]) #put together gene names, expression data, and p-vals
            symbol_match = symbol_match[order(-symbol_match$rhodf...match_index.),] #order from high to low gene values
            names(symbol_match) = c("gene","rho","p")
            
            #make marker df for grahing eigengene
            marker_df = data.frame(unique(alias2SymbolTable(symbol_match$gene[1:(eigengene_size+1)], species = "Mm")),paste(seed_gene[i]," Normal skin",sep=""),paste(seed_gene[i]," Normal skin",sep=""))
            names(marker_df) = c("gene","cell_type","module")
            marker_df = na.omit(marker_df)
            
            #------------------------------#
            #4a. Seed gene
            #------------------------------#
            #all cells eigengene
            seed_gene_manifold = plot_cells(big_cds,
                                            genes=seed_gene[i],
                                            label_cell_groups=FALSE,
                                            label_leaves=F,
                                            label_branch_points=F,
                                            graph_label_size=1.5,
                                            show_trajectory_graph=F) +
              xlab("") + ylab("")  +
              theme(axis.ticks = element_blank()) +
              scale_color_viridis_c(option = "inferno",name="") +
              theme(axis.text=element_text(size=0,color='black'),
                    axis.title=element_text(size=0,color='black')) + theme(plot.title = element_text(size = 0, vjust=0)) +
              theme(strip.text.x = element_text(size = 0)) +
              annotate("text", label = paste(seed_gene[i],"\nseed gene",sep=""), 
                       x = -6, y = 7, vjust = 0,size=10) +
              theme(legend.position=c(0.8, 0.2),
                    legend.key.size = unit(1, 'cm'),
                    legend.text = element_text(size=20))
            
            
            #------------------------------#
            #4a. Normal skin WT: calculate correlation matrix
            #------------------------------#
            #all cells eigengene
            agexp_plot.12 = plot_cells(big_cds,
                                       genes=marker_df,
                                       label_cell_groups=FALSE,
                                       label_leaves=F,
                                       label_branch_points=F,
                                       graph_label_size=1.5,
                                       show_trajectory_graph=F) +
              xlab("") + ylab("") + 
              theme(axis.ticks = element_blank()) +
              scale_color_viridis_c(option = "inferno",name="") +
              theme(axis.text=element_text(size=0,color='black'),
                    axis.title=element_text(size=0,color='black')) + theme(plot.title = element_text(size = 0, vjust=0)) +
              theme(strip.text.x = element_text(size = 0))  +
              annotate("text", label = paste(seed_gene[i],"\nNSk eigengene",sep=""), 
                       x = -5, y = 7, vjust = 0,size=10) +
              theme(legend.position=c(0.8, 0.2),
                    legend.key.size = unit(1, 'cm'),
                    legend.text = element_text(size=20))
            
            #---------------------------------------#
            #tumor
            if( seed_gene[i]=="Prom1" ) {match_index = 12747} else {
              match_index = match(seed_gene[i], aim2probedat$SYMBOL) #match seed gene to its place in the matrix
            }
            symbol_match = data.frame(aim2probedat$SYMBOL, tumor_rhodf[,match_index], tumor_p_df[,match_index]) #put together gene names, expression data, and p-vals
            symbol_match = symbol_match[order(-symbol_match$tumor_rhodf...match_index.),] #order from high to low gene values
            names(symbol_match) = c("gene","rho","p")
            
            #make marker df for grahing eigengene
            marker_df = data.frame(unique(alias2SymbolTable(symbol_match$gene[1:(eigengene_size+1)], species = "Mm")),paste(seed_gene[i],"Carcinoma",sep=""),paste(seed_gene[i],"Carcinoma",sep=""))
            names(marker_df) = c("gene","cell_type","module")
            marker_df = na.omit(marker_df)
            
            #all cells eigengene
            agexp_plot_tumor.12 = plot_cells(big_cds,
                                             genes=marker_df,
                                             label_cell_groups=FALSE,
                                             label_leaves=F,
                                             label_branch_points=F,
                                             graph_label_size=1.5,
                                             show_trajectory_graph=F) +
              xlab("") + ylab("") + 
              theme(axis.ticks = element_blank()) +
              scale_color_viridis_c(option = "inferno",name="") +
              theme(axis.text=element_text(size=0,color='black'),
                    axis.title=element_text(size=0,color='black')) + theme(plot.title = element_text(size = 0, vjust=0)) +
              theme(strip.text.x = element_text(size = 0))  +
              annotate("text", label = paste(seed_gene[i],"\nCar eigengene",sep=""), 
                       x = -5, y = 7, vjust = 0,size=10) +
              theme(legend.position=c(0.8, 0.2),
                    legend.key.size = unit(1, 'cm'),
                    legend.text = element_text(size=20))
            
            
            #-----------#
            #print out eigengene plots as a nice long row
            png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Figs/Fig2/Fig2_manifolds/",seed_gene[i],".png",sep=""),
                width=14,height=7,units="in",res=300)
            
            print(multiplot(seed_gene_manifold, agexp_plot_tumor.12,
                            cols=2))
            dev.off()
            
            #-----------#
            #print out as column
            png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Figs/Fig2/Fig2_manifolds/",seed_gene[i],"_column.png",sep=""),
                width=7,height=14,units="in",res=300)
            
            print(multiplot(seed_gene_manifold, agexp_plot_tumor.12,
                            cols=1))
            dev.off()
            
          },  error=function(e){})
          
        }))
        
        
        
        
        ##################################################################################################################################
        ##################################################################################################################################
        #Fig. 2B : Eigencell eigengene heatmap
        
        #Because bridge cells over-write some upper/lower spike cells, reconsitute these cells
        colData(big_cds)$integrated_celltype_spike = colData(big_cds)$cell_type 
        
        #Cluster 19
        cds_subset.v3 = big_cds[,colData(big_cds)$cluster %in% c("19")] #filter for cluster 19 because most of the bridge cells are only in cluster 19
        matching_cells = match(colnames(cds_subset.v3),colnames(big_cds)) #match extracted cells with all cell names
        colData(big_cds)$integrated_celltype_spike[matching_cells] <- "upper spike" #now replace matching cells with new name
        
        #Cluster 33
        cds_subset.v3 = big_cds[,colData(big_cds)$cluster %in% c("33")] #filter for cluster 19 because most of the bridge cells are only in cluster 19
        matching_cells = match(colnames(cds_subset.v3),colnames(big_cds)) #match extracted cells with all cell names
        colData(big_cds)$integrated_celltype_spike[matching_cells] <- "lower spike" #now replace matching cells with new name
        
        #Add squamous and spindle phases to the carcinoma
        e <- readRDS(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/matvei/cds_objects/squamous.RDS",sep=""))
        f <- readRDS(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/matvei/cds_objects/spindle.RDS",sep=""))
        squamous_matches = match(colnames(e),colnames(big_cds)) #match extracted cells with all cell names
        spindle_matches = match(colnames(f),colnames(big_cds)) #match extracted cells with all cell names
        
        colData(big_cds)$integrated_celltype_spike[squamous_matches] <- "squamous" #now replace matching cells with new name
        colData(big_cds)$integrated_celltype_spike[spindle_matches] <- "spindle" #now replace matching cells with new name
        
        
        #selected eigengenes
        cds_comb = big_cds[,colData(big_cds)$partition %in% c(1,2,3,4)]
        unique(colData(cds_comb)$integrated_celltype_spike)
        unique(colData(cds_comb)$heatmap_vars)
        
        colData(cds_comb)$heatmap_vars = colData(cds_comb)$integrated_celltype_spike
        colData(cds_comb)$heatmap_vars = gsub("IfE","NSk-IF",colData(cds_comb)$heatmap_vars)
        colData(cds_comb)$heatmap_vars = gsub("Inf","NSk-F",colData(cds_comb)$heatmap_vars)
        colData(cds_comb)$heatmap_vars = gsub("OuB","NSk-F",colData(cds_comb)$heatmap_vars)
        colData(cds_comb)$heatmap_vars = gsub("UHF","NSk-F",colData(cds_comb)$heatmap_vars)
        colData(cds_comb)$heatmap_vars = gsub("Isth","NSk-F",colData(cds_comb)$heatmap_vars)
        colData(cds_comb)$heatmap_vars = gsub("unF","NSk-F",colData(cds_comb)$heatmap_vars)
        colData(cds_comb)$heatmap_vars = gsub("PrE","NSk-F",colData(cds_comb)$heatmap_vars)
        colData(cds_comb)$heatmap_vars = gsub("nEp","Pap-nEp",colData(cds_comb)$heatmap_vars)
        colData(cds_comb)$heatmap_vars = gsub("upper spike","Pap-Sp",colData(cds_comb)$heatmap_vars)
        colData(cds_comb)$heatmap_vars = gsub("lower spike","Pap-LI",colData(cds_comb)$heatmap_vars)
        colData(cds_comb)$heatmap_vars = gsub("squamous","SCC-Sq",colData(cds_comb)$heatmap_vars)
        colData(cds_comb)$heatmap_vars = gsub("spindle","SCC-Sp",colData(cds_comb)$heatmap_vars)
        
        unique(colData(cds_comb)$heatmap_vars)
        unique(colData(cds_comb)$integrated_celltype_spike)
        
        #---------------------------------------------------------------------------------#
        #CARCINOMA eigencell heatmap
        #loop Thru eigengenes listed in the eigencell_genes file
        eigengene_size=99
        eigencell_agg = do.call(rbind, lapply(1:length(seed_gene), function(i){
          tryCatch({ #suppress error
            
            #i=1
            print(i)
            
            #tumor
            if( seed_gene[i]=="Prom1" ) {match_index = 12747} else {
              match_index = match(seed_gene[i], aim2probedat$SYMBOL) #match seed gene to its place in the matrix
            }
            
            symbol_match = data.frame(aim2probedat$SYMBOL, tumor_rhodf[,match_index], tumor_p_df[,match_index]) #put together gene names, expression data, and p-vals
            symbol_match = symbol_match[order(-symbol_match$tumor_rhodf...match_index.),] #order from high to low gene values
            names(symbol_match) = c("gene","rho","p")
            
            #make marker df for grahing eigengene
            marker_df = data.frame(unique(alias2SymbolTable(symbol_match$gene[1:(eigengene_size+1)], species = "Mm")),paste(seed_gene[i],"Carcinoma",sep=""),paste(seed_gene[i],"Carcinoma",sep=""))
            names(marker_df) = c("gene","cell_type","module")
            
            #Create data frame of cells and parenchyma tissue to aggregate over
            #eigen_cell_def = data.frame(row.names(colData(cds_comb)),colData(cds_comb)$parenchyma_tissue) #unbalancec cells
            eigen_cell_def = data.frame(row.names(colData(cds_comb)),colData(cds_comb)$heatmap_vars)
            names(eigen_cell_def) = c("cell_id","cell_group")
            
            #aggregate across cells within all individual genes
            agg_mat = aggregate_gene_expression(
              cds_comb, #unbalanced all cells
              gene_group_df = NULL,
              cell_group_df = eigen_cell_def,
              norm_method = c("log"),
              pseudocount = 1,
              scale_agg_values = TRUE,
              #max_agg_value = 10,
              #min_agg_value = -5,
              exclude.na = TRUE
            )
            
            #extract genes from eigencell
            agg_matsub = agg_mat[row.names(agg_mat) %in% marker_df$gene,] #get cds names that match the balmain sub network names
            
            #sum across genes to get eigengene
            #if(nrow(agg_matsub)>25) {agg_matsub = agg_matsub[1:25,]} else {agg_matsub = agg_matsub} #balance number of eigengenes by setting max as top 25
            sumdf = data.frame(t(data.frame(colSums(agg_matsub)))) #sum all eigengenes within a cell class
            sumdf = sumdf/nrow(agg_matsub) #standardize by eigengene size: divide by number of included genes
            eigensumdf = data.frame(sumdf$NSk.IF, sumdf$NSk.F, sumdf$Pap.nEp, sumdf$Pap.LI,sumdf$Pap.Sp,sumdf$SCC.Sq,sumdf$SCC.Sp) #re-order
            
            row.names(eigensumdf) = paste(seed_gene[i])
            names(eigensumdf) = c("NSk-IF", "NSk-F","Pap-nEp","Pap-LI","Pap-Sp","SCC-Sq","SCC-Sp")
            eigensumdf
            
          },  error=function(e){})
        }))
        
        #now heatmap of those eigencell eigengenes:  columns are eigencells, rows are eigengenes
        virpal = viridis_pal(alpha = 1, begin = 0, end = 1, direction = 1,
                             option = "C") #set the colorblind palette
        eigencell_agg.mat = as.matrix(eigencell_agg)
        
        #Re-order matrix to look like what I want it to betch
        row_order = c("Lgr5",
                      "Bmi1","Sox4","Lrig1",
                      "Lgr6","Sox9",
                      "Sox2","Prom1","Foxa1","Krt15","Krt19","Pitx1","Psca","Prom2","Klf5",
                      "Cd44")
        eigencell_agg.mat = eigencell_agg.mat[match(row_order,row.names(eigencell_agg.mat)),]
        
        length(row_order) == nrow(eigencell_agg.mat)
        
        nr = nrow(eigencell_agg.mat)
        pdf(file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Figs/Fig2/heatmaps/Fig2B_eigencell_heatmap_carcinoma.pdf",height=6,width=6)
        heatmap.2(eigencell_agg.mat, scale = "none", col = virpal, dendrogram="row",
                  trace = "none", density.info = "none",key.xlab="eigencell eigengene expression",
                  main = "",
                  key=F,
                  cexCol = 1,
                  cexRow = 0.2 + 1/log10(nr),
                  Rowv=F,
                  Colv=F,
                  keysize=1,
                  colsep=c(1:27),
                  rowsep=c(1:28),
                  sepcolor="lightgrey",
                  sepwidth=c(0.02,0.02),
                  adjRow=c(0.3,0.3),
                  srtCol=0,
                  margins=c(7,8),
                  adjCol = c(0.45,0.3),
                  lwid=c(1,1), 
                  lhei=c(0.1,0.6)
        )
        dev.off()
        
        #Scale each row individually
        eigencell_agg.mat_rowscaled = do.call(rbind, lapply(1:nrow(eigencell_agg.mat), function(i){
          #i=1
          return_row = t(data.frame(scale(eigencell_agg.mat[i,])))
          row.names(return_row) = row.names(eigencell_agg.mat)[i]
          return_row
        }))
        
        pdf(file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Figs/Fig2/heatmaps/Fig2B_eigencell_heatmap_carcinoma_rowscaled.pdf",height=6,width=6)
        heatmap.2(eigencell_agg.mat_rowscaled, scale = "none", col = virpal, dendrogram="row",
                  trace = "none", density.info = "none",key.xlab="eigencell eigengene expression",
                  main = "",
                  key=F,
                  cexCol = 1,
                  cexRow = 0.5 + 1/log10(nr),
                  Rowv=F,
                  Colv=F,
                  keysize=1,
                  colsep=c(1:27),
                  rowsep=c(1:28),
                  sepcolor="lightgrey",
                  sepwidth=c(0.02,0.02),
                  adjRow=c(0.3,0.3),
                  srtCol=0,
                  margins=c(7,8),
                  adjCol = c(0.45,0.3),
                  lwid=c(1,1), 
                  lhei=c(0.1,0.6)
        )
        dev.off()
        
        pdf(file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Figs/Fig2/heatmaps/Fig2B_eigencell_heatmap_carcinoma_rowscaled_withscale.pdf",height=8,width=6)
        heatmap.2(eigencell_agg.mat_rowscaled,  scale = "none",col = virpal, dendrogram="row",
                  trace = "none", density.info = "none",key.xlab="eigencell eigengene expression",
                  main = "",
                  key=T,
                  cexCol = 1,
                  cexRow = 0.7 + 1/log10(nr),
                  Rowv=F,
                  Colv=F,
                  keysize=1,
                  colsep=c(1:27),
                  rowsep=c(1:28),
                  sepcolor="lightgrey",
                  sepwidth=c(0.02,0.02),
                  adjRow=c(0.3,0.3),
                  srtCol=0,
                  margins=c(7,8),
                  adjCol = c(0.45,0.3),
                  lwid=c(0.55,1), 
                  lhei=c(0.115,0.6)
        )
        dev.off()
        
        #---------------------------------------------------------------------------------#
        #---------------------------------------------------------------------------------#
        #---------------------------------------------------------------------------------#
        #---------------------------------------------------------------------------------#
        #---------------------------------------------------------------------------------#
        #Fig.1 correlation plot
        
        match_index1 = match(c("Lgr5",
                               "Bmi1","Sox4","Lrig1",
                               "Lgr6","Sox9",
                               "Sox2"), aim2probedat$SYMBOL) #match seed gene to its place in the matrix
        Prom1=12747 #because this is a different fucking probe
        match_index2 = match(c("Foxa1","Krt15","Krt19","Pitx1","Psca","Prom2","Klf5",
                               "Cd44"), aim2probedat$SYMBOL) #match seed gene to its place in the matrix
        match_index = c(match_index1,12747,match_index2)
        
        #---------------------------------------#
        #Get normal skin and tumor data
        nskdat = normal_wt_dat[match_index,]
        row.names(nskdat) = row_order
        nskdat=t(nskdat)
        
        tumordat = tumor_wt_dat[match_index,]
        row.names(tumordat) = row_order
        tumordat=t(tumordat)
        
        #---------------------------------------#
        #Calculate mean rho values in correlation matrices and compare them
        mean(as.vector(rcorr(nskdat, type = c("spearman"))$r)) #NSk correlation
        mean(as.vector(rcorr(tumordat, type = c("spearman"))$r)) #Tumor correlation
        t.test(as.vector(rcorr(nskdat, type = c("spearman"))$r) , as.vector(rcorr(tumordat, type = c("spearman"))$r) ,alternative = c("less"))
        
        #---------------------------------------#
        #Tumor seed gene correlations in bulk data
        png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Figs/Fig1/Fig1x_tumor_correlations.png",sep=""),
            width=10,height=10,units="in",res=200)
        #plot tumor data
        print(ggcorr(tumordat, 
                     method = c("complete.obs","spearman"),
                     label_round=2,
                     label=T,
                     label_size=4.5,
                     geom = "tile",
                     low="darkgreen",high="darkorange",
                     min_size= 11,
                     max_size = 11,
                     hjust = 1,
                     angle = 0,
                     size = 6,layout.exp=5,
                     legend.position = "right",
                     legend.size = 15,
        ) + ggtitle("") +
          theme(plot.title = element_text(size = 25, face = "bold"))
        )
        dev.off()
        
        #---------------------------------------#
        #NSk seed gene correlations in bulk data
        png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Figs/Fig1/Fig1y_nsk_correlations.png",sep=""),
            width=10,height=10,units="in",res=200)
        #plot tumor data
        print(ggcorr(nskdat, 
                     method = c("complete.obs","spearman"),
                     label_round=2,
                     label=T,
                     label_size=4.5,
                     geom = "tile",
                     low="darkgreen",high="darkorange",
                     min_size= 11,
                     max_size = 11,
                     hjust = 1,
                     angle = 0,
                     size = 6,layout.exp=5,
                     legend.position = "right",
                     legend.size = 15,
        ) + ggtitle("") +
          theme(plot.title = element_text(size = 25, face = "bold"))
        )
        dev.off()
        
        
        #---------------------------------------------------------------------------------#
        #---------------------------------------------------------------------------------#
        #---------------------------------------------------------------------------------#
        #---------------------------------------------------------------------------------#
        #Fig. 5 : Carcinoma clulster 3 gene that INCREASE across the lineage trajectory
        #Sorting versus stem eigengene
        
        #selected eigengenes
        seed_gene = c(
          "Sox2","Prom1","Foxa1","Krt15","Krt19","Pitx1","Psca","Prom2","Klf5",
          "Cd44")
        eigengene_size=100
        
        #Loop through all eigengenes
        eigengene_agg = do.call(rbind, lapply(1:length(seed_gene), function(i){
          
          #i=5
          print(i)
          
          #Get the correct match_index
          if( seed_gene[i]=="Prom1" ) {match_index = 12747} else {
            match_index = match(seed_gene[i], aim2probedat$SYMBOL) #match seed gene to its place in the matrix
          }
          
          #---------------------------------------#
          #tumor
          match_index = match(seed_gene[i], aim2probedat$SYMBOL) #match seed gene to its place in the matrix
          symbol_match = data.frame(aim2probedat$SYMBOL, tumor_rhodf[,match_index], tumor_p_df[,match_index]) #put together gene names, expression data, and p-vals
          symbol_match = symbol_match[order(-symbol_match$tumor_rhodf...match_index.),] #order from high to low gene values
          names(symbol_match) = c("gene","rho","p")
          
          #make marker df for grahing eigengene
          marker_df = data.frame(unique(alias2SymbolTable(symbol_match$gene[1:(eigengene_size+1)], species = "Mm")),paste(seed_gene[i],"Carcinoma",sep=""),paste(seed_gene[i],"Carcinoma",sep=""))
          names(marker_df) = c("gene","cell_type","module")
          marker_df = na.omit(marker_df)
          
          #CALCULATE EIGENGENE VALUE OF ASSOCIATED DISEASED GENES AND GRAPH OUT
          #negative eigengene
          agg_mat = aggregate_gene_expression(
            cds_comb,
            gene_group_df = marker_df,
            cell_group_df = NULL,
            norm_method = c("log"),
            pseudocount = 1,
            scale_agg_values = F,
            #max_agg_value = NA,
            min_agg_value = 0,
            exclude.na = TRUE
          )
          
          agg_mat = t(data.frame(agg_mat))
          names(agg_mat)[1] = "eigengene_value"
          combined_eigen = data.frame(colData(cds_comb)$tissue_updated,colData(cds_comb)$sorting,agg_mat) #Add metadata
          names(combined_eigen) = c("tissue","sorting","eigengene_value")
          dfc = summarySE(combined_eigen, measurevar='eigengene_value', groupvars=c("tissue","sorting"))
          
          #Reorder
          dfc = dfc[c(6,4,5,9,7,8,3,1,2),]
          dfc$order = seq(1:nrow(dfc))
          dfc$gene_type = "eigengene"
          dfc$seed_gene = seed_gene[i]
          
          #test whether Lgr6 -> Progeny within a tissue is significant and positive or negative
          pap_test.p = t.test( subset(combined_eigen, tissue=="Pap" & sorting=="Lgr6")$eigengene_value,
                               subset(combined_eigen, tissue=="Pap" & sorting=="Progeny")$eigengene_value )$p.value # independent 2-group Mann-Whitney U Test
          if(pap_test.p < 0.05 & dfc$eigengene_value[6] < dfc$eigengene_value[5]) {test.pap = "significant decrease"} else if 
          (pap_test.p < 0.05 & dfc$eigengene_value[6] > dfc$eigengene_value[5]) {test.pap = "significant increase"} else {test.pap = "not significant"} 
          
          car_test.p = t.test( subset(combined_eigen, tissue=="CA" & sorting=="Lgr6")$eigengene_value,
                               subset(combined_eigen, tissue=="CA" & sorting=="Progeny")$eigengene_value )$p.value # independent 2-group Mann-Whitney U Test
          if(car_test.p < 0.05 & dfc$eigengene_value[9] < dfc$eigengene_value[8]) {test.car = "significant decrease"} else if 
          (car_test.p < 0.05 & dfc$eigengene_value[9] > dfc$eigengene_value[8]) {test.car = "significant increase"} else {test.car = "not significant"} 
          
          dfc$test_result = c(NA,NA,NA,test.pap,test.pap,test.pap,test.car,test.car,test.car)
          
          #--------------------------------------------#
          #Do the same thing for individual seed gene
          agg_mat = data.frame(cds_comb@assays@data$counts[seed_gene[i],])
          names(agg_mat)[1] = "eigengene_value"
          combined_eigen = data.frame(colData(cds_comb)$tissue_updated,colData(cds_comb)$sorting,agg_mat) #Add metadata
          names(combined_eigen) = c("tissue","sorting","eigengene_value")
          dfc_seed = summarySE(combined_eigen, measurevar='eigengene_value', groupvars=c("tissue","sorting"))
          
          #Reorder
          dfc_seed = dfc_seed[c(6,4,5,9,7,8,3,1,2),]
          dfc_seed$order = seq(1:nrow(dfc_seed))
          dfc_seed$gene_type = "seed gene"
          dfc_seed$seed_gene = seed_gene[i]
          dfc_seed
          
          #test whether Lgr6 -> Progeny within a tissue is significant and positive or negative
          pap_test.p = t.test( subset(combined_eigen, tissue=="Pap" & sorting=="Lgr6")$eigengene_value,
                               subset(combined_eigen, tissue=="Pap" & sorting=="Progeny")$eigengene_value )$p.value # independent 2-group Mann-Whitney U Test
          if(pap_test.p < 0.05 & dfc_seed$eigengene_value[6] < dfc_seed$eigengene_value[5]) {test.pap = "significant decrease"} else if 
          (pap_test.p < 0.05 & dfc_seed$eigengene_value[6] > dfc_seed$eigengene_value[5]) {test.pap = "significant increase"} else {test.pap = "not significant"} 
          
          car_test.p = t.test( subset(combined_eigen, tissue=="CA" & sorting=="Lgr6")$eigengene_value,
                               subset(combined_eigen, tissue=="CA" & sorting=="Progeny")$eigengene_value )$p.value # independent 2-group Mann-Whitney U Test
          if(car_test.p < 0.05 & dfc_seed$eigengene_value[9] < dfc_seed$eigengene_value[8]) {test.car = "significant decrease"} else if 
          (car_test.p < 0.05 & dfc_seed$eigengene_value[9] > dfc_seed$eigengene_value[8]) {test.car = "significant increase"} else {test.car = "not significant"} 
          
          dfc_seed$test_result = c(NA,NA,NA,test.pap,test.pap,test.pap,test.car,test.car,test.car)
          
          
          rbind(dfc,dfc_seed)
          
        }))
        
        #LINE PLOT Carcinoma eigengene
        eigengene_aggset = subset(eigengene_agg, tissue=="CA" & gene_type=="eigengene" & tissue!="Normal skin" & sorting!="Unsorted" &
                                    seed_gene!="Krt15" & seed_gene!="Prom1" & seed_gene!="Prom2")
        
        ca_eigengene = ggplot(data=eigengene_aggset, aes(x = sorting, y = eigengene_value, color=test_result)) + 
          geom_point(size=1.5, alpha=1) +
          geom_line(size=0.5, aes(group=seed_gene)) +
          geom_errorbar(data=eigengene_aggset, aes(sorting, ymin=eigengene_value-abs(se), ymax=eigengene_value+abs(se)), 
                        alpha=0.5, width=0.1, size=0.8) +
          #facet_wrap(~seed_gene, scales = "free") +
          facet_wrap(~ factor(seed_gene, levels=c(
            "Sox2","Foxa1","Krt19","Pitx1","Psca","Klf5",
            "Cd44")), scales = "free", nrow=1) +
          
          xlab('') +
          ylab('') +
          ggtitle("") +
          theme(axis.text=element_text(size=14,color='black'),
                axis.title=element_text(size=0,color='black')) +
          theme(strip.background = element_rect(fill = 'white', color='black', size=1)) +
          theme(plot.title = element_text(size = 0)) +
          theme(panel.background=element_blank(),
                panel.grid.major=element_blank(),
                panel.grid.minor=element_blank())+
          theme(axis.line = element_line(colour = 'black')) +
          theme(legend.position = 'none') +
          theme(axis.line = element_line(colour = 'black')) +
          theme(panel.background = element_rect(colour = "black")) +
          theme(panel.spacing = unit(0, "lines")) +
          theme_classic() +
          theme(legend.position="none") +
          theme(axis.text.x=element_text(size=0,angle=0,color='black')) +
          #theme(axis.text.y=element_text(size=25,angle=0,color='black')) +
          scale_x_discrete(breaks=c("Lgr6","Progeny"),
                           labels=c("L", "P")) +
          theme(strip.background = element_blank(),
                axis.ticks.x = element_blank(),
                strip.text.x = element_text(size = 16)) +
          scale_color_manual(name="",values = c("significant decrease" = "blue",
                                                "significant increase" = "red",
                                                "not significant" = "grey"))
        
        png("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Figs/Fig5/ca_eigengene.png",
            width=11,height=3,units="in",res=200)
        print(ca_eigengene)
        dev.off()
        
        #LINE PLOT Carcinoma seed gene
        eigengene_aggset = subset(eigengene_agg, tissue=="CA" & gene_type=="seed gene" & tissue!="Normal skin" & sorting!="Unsorted"&
                                    seed_gene!="Krt15" & seed_gene!="Prom1" & seed_gene!="Prom2")
        ca_seedgene = ggplot(data=eigengene_aggset, aes(x = sorting, y = eigengene_value, color=test_result)) + 
          geom_point(size=1.5, alpha=1) +
          geom_line(size=0.5, aes(group=seed_gene)) +
          geom_errorbar(data=eigengene_aggset, aes(sorting, ymin=eigengene_value-abs(se), ymax=eigengene_value+abs(se)), 
                        alpha=0.5, width=0.1, size=0.8) +
          #facet_wrap(~seed_gene, scales = "free") +
          facet_wrap(~ factor(seed_gene, levels=c(
            "Sox2","Foxa1","Krt19","Pitx1","Psca","Klf5",
            "Cd44")), scales = "free", nrow=1) +
          
          xlab('') +
          ylab('') +
          ggtitle("") +
          theme(axis.text=element_text(size=14,color='black'),
                axis.title=element_text(size=0,color='black')) +
          theme(strip.background = element_rect(fill = 'white', color='black', size=1)) +
          theme(plot.title = element_text(size = 0)) +
          theme(panel.background=element_blank(),
                panel.grid.major=element_blank(),
                panel.grid.minor=element_blank())+
          theme(axis.line = element_line(colour = 'black')) +
          theme(legend.position = 'none') +
          theme(axis.line = element_line(colour = 'black')) +
          theme(panel.background = element_rect(colour = "black")) +
          theme(panel.spacing = unit(0, "lines")) +
          theme_classic() +
          theme(legend.position="none") +
          theme(axis.text.x=element_text(size=0,angle=0,color='black')) +
          #theme(axis.text.y=element_text(size=25,angle=0,color='black')) +
          scale_x_discrete(breaks=c("Lgr6","Progeny"),
                           labels=c("L", "P")) +
          theme(strip.background = element_blank(),
                axis.ticks.x = element_blank(),
                strip.text.x = element_text(size = 16)) +
          scale_color_manual(name="",values = c("significant decrease" = "blue",
                                                "significant increase" = "red",
                                                "not significant" = "grey"))
        
        png("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Figs/Fig5/ca_seedgene.png",
            width=11,height=3,units="in",res=200)
        print(ca_seedgene)
        dev.off()
        
        
        #-----------------------------------------------------------------------------------------------------------------------#
        #-----------------------------------------------------------------------------------------------------------------------#
        #-----------------------------------------------------------------------------------------------------------------------#
        #-----------------------------------------------------------------------------------------------------------------------#
        #FIGURE 5 extended data: selected eigengenes
        seed_gene = c(
          "Bmi1","Sox4","Lrig1")
        eigengene_size=100
        
        #Loop through all eigengenes
        eigengene_agg = do.call(rbind, lapply(1:length(seed_gene), function(i){
          
          #i=5
          print(i)
          
          #Get the correct match_index
          if( seed_gene[i]=="Prom1" ) {match_index = 12747} else {
            match_index = match(seed_gene[i], aim2probedat$SYMBOL) #match seed gene to its place in the matrix
          }
          
          #---------------------------------------#
          #tumor
          match_index = match(seed_gene[i], aim2probedat$SYMBOL) #match seed gene to its place in the matrix
          symbol_match = data.frame(aim2probedat$SYMBOL, tumor_rhodf[,match_index], tumor_p_df[,match_index]) #put together gene names, expression data, and p-vals
          symbol_match = symbol_match[order(-symbol_match$tumor_rhodf...match_index.),] #order from high to low gene values
          names(symbol_match) = c("gene","rho","p")
          
          #make marker df for grahing eigengene
          marker_df = data.frame(unique(alias2SymbolTable(symbol_match$gene[1:(eigengene_size+1)], species = "Mm")),paste(seed_gene[i],"Carcinoma",sep=""),paste(seed_gene[i],"Carcinoma",sep=""))
          names(marker_df) = c("gene","cell_type","module")
          marker_df = na.omit(marker_df)
          
          #CALCULATE EIGENGENE VALUE OF ASSOCIATED DISEASED GENES AND GRAPH OUT
          #negative eigengene
          agg_mat = aggregate_gene_expression(
            cds_comb,
            gene_group_df = marker_df,
            cell_group_df = NULL,
            norm_method = c("log"),
            pseudocount = 1,
            scale_agg_values = F,
            #max_agg_value = NA,
            min_agg_value = 0,
            exclude.na = TRUE
          )
          
          agg_mat = t(data.frame(agg_mat))
          names(agg_mat)[1] = "eigengene_value"
          combined_eigen = data.frame(colData(cds_comb)$tissue_updated,colData(cds_comb)$sorting,agg_mat) #Add metadata
          names(combined_eigen) = c("tissue","sorting","eigengene_value")
          dfc = summarySE(combined_eigen, measurevar='eigengene_value', groupvars=c("tissue","sorting"))
          
          #Reorder
          dfc = dfc[c(6,4,5,9,7,8,3,1,2),]
          dfc$order = seq(1:nrow(dfc))
          dfc$gene_type = "eigengene"
          dfc$seed_gene = seed_gene[i]
          
          #test whether Lgr6 -> Progeny within a tissue is significant and positive or negative
          pap_test.p = t.test( subset(combined_eigen, tissue=="Pap" & sorting=="Lgr6")$eigengene_value,
                               subset(combined_eigen, tissue=="Pap" & sorting=="Progeny")$eigengene_value )$p.value # independent 2-group Mann-Whitney U Test
          if(pap_test.p < 0.05 & dfc$eigengene_value[6] < dfc$eigengene_value[5]) {test.pap = "significant decrease"} else if 
          (pap_test.p < 0.05 & dfc$eigengene_value[6] > dfc$eigengene_value[5]) {test.pap = "significant increase"} else {test.pap = "not significant"} 
          
          car_test.p = t.test( subset(combined_eigen, tissue=="CA" & sorting=="Lgr6")$eigengene_value,
                               subset(combined_eigen, tissue=="CA" & sorting=="Progeny")$eigengene_value )$p.value # independent 2-group Mann-Whitney U Test
          if(car_test.p < 0.05 & dfc$eigengene_value[9] < dfc$eigengene_value[8]) {test.car = "significant decrease"} else if 
          (car_test.p < 0.05 & dfc$eigengene_value[9] > dfc$eigengene_value[8]) {test.car = "significant increase"} else {test.car = "not significant"} 
          
          dfc$test_result = c(NA,NA,NA,test.pap,test.pap,test.pap,test.car,test.car,test.car)
          
          #--------------------------------------------#
          #Do the same thing for individual seed gene
          agg_mat = data.frame(cds_comb@assays@data$counts[seed_gene[i],])
          names(agg_mat)[1] = "eigengene_value"
          combined_eigen = data.frame(colData(cds_comb)$tissue_updated,colData(cds_comb)$sorting,agg_mat) #Add metadata
          names(combined_eigen) = c("tissue","sorting","eigengene_value")
          dfc_seed = summarySE(combined_eigen, measurevar='eigengene_value', groupvars=c("tissue","sorting"))
          
          #Reorder
          dfc_seed = dfc_seed[c(6,4,5,9,7,8,3,1,2),]
          dfc_seed$order = seq(1:nrow(dfc_seed))
          dfc_seed$gene_type = "seed gene"
          dfc_seed$seed_gene = seed_gene[i]
          dfc_seed
          
          #test whether Lgr6 -> Progeny within a tissue is significant and positive or negative
          pap_test.p = t.test( subset(combined_eigen, tissue=="Pap" & sorting=="Lgr6")$eigengene_value,
                               subset(combined_eigen, tissue=="Pap" & sorting=="Progeny")$eigengene_value )$p.value # independent 2-group Mann-Whitney U Test
          if(pap_test.p < 0.05 & dfc_seed$eigengene_value[6] < dfc_seed$eigengene_value[5]) {test.pap = "significant decrease"} else if 
          (pap_test.p < 0.05 & dfc_seed$eigengene_value[6] > dfc_seed$eigengene_value[5]) {test.pap = "significant increase"} else {test.pap = "not significant"} 
          
          car_test.p = t.test( subset(combined_eigen, tissue=="CA" & sorting=="Lgr6")$eigengene_value,
                               subset(combined_eigen, tissue=="CA" & sorting=="Progeny")$eigengene_value )$p.value # independent 2-group Mann-Whitney U Test
          if(car_test.p < 0.05 & dfc_seed$eigengene_value[9] < dfc_seed$eigengene_value[8]) {test.car = "significant decrease"} else if 
          (car_test.p < 0.05 & dfc_seed$eigengene_value[9] > dfc_seed$eigengene_value[8]) {test.car = "significant increase"} else {test.car = "not significant"} 
          
          dfc_seed$test_result = c(NA,NA,NA,test.pap,test.pap,test.pap,test.car,test.car,test.car)
          
          
          rbind(dfc,dfc_seed)
          
        }))
        
        #LINE PLOT Carcinoma eigengene
        eigengene_aggset = subset(eigengene_agg, tissue=="CA" & gene_type=="eigengene" & tissue!="Normal skin" & sorting!="Unsorted" &
                                    seed_gene!="Krt15" & seed_gene!="Prom1" & seed_gene!="Prom2")
        
        ca_eigengene = ggplot(data=eigengene_aggset, aes(x = sorting, y = eigengene_value, color=test_result)) + 
          geom_point(size=1.5, alpha=1) +
          geom_line(size=0.5, aes(group=seed_gene)) +
          geom_errorbar(data=eigengene_aggset, aes(sorting, ymin=eigengene_value-abs(se), ymax=eigengene_value+abs(se)), 
                        alpha=0.5, width=0.1, size=0.8) +
          #facet_wrap(~seed_gene, scales = "free") +
          facet_wrap(~ factor(seed_gene, levels=c(
            "Bmi1","Sox4","Lrig1")), scales = "free", nrow=1) +
          
          xlab('') +
          ylab('') +
          ggtitle("") +
          theme(axis.text=element_text(size=14,color='black'),
                axis.title=element_text(size=0,color='black')) +
          theme(strip.background = element_rect(fill = 'white', color='black', size=1)) +
          theme(plot.title = element_text(size = 0)) +
          theme(panel.background=element_blank(),
                panel.grid.major=element_blank(),
                panel.grid.minor=element_blank())+
          theme(axis.line = element_line(colour = 'black')) +
          theme(legend.position = 'none') +
          theme(axis.line = element_line(colour = 'black')) +
          theme(panel.background = element_rect(colour = "black")) +
          theme(panel.spacing = unit(0, "lines")) +
          theme_classic() +
          theme(legend.position="none") +
          theme(axis.text.x=element_text(size=0,angle=0,color='black')) +
          #theme(axis.text.y=element_text(size=25,angle=0,color='black')) +
          scale_x_discrete(breaks=c("Lgr6","Progeny"),
                           labels=c("L", "P")) +
          theme(strip.background = element_blank(),
                axis.ticks.x = element_blank(),
                strip.text.x = element_text(size = 16)) +
          scale_color_manual(name="",values = c("significant decrease" = "blue",
                                                "significant increase" = "red",
                                                "not significant" = "grey"))
        
        png("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Figs/Fig5/ca_eigengene_cluster1.png",
            width=4,height=3,units="in",res=200)
        print(ca_eigengene)
        dev.off()
        
        #LINE PLOT Carcinoma seed gene
        eigengene_aggset = subset(eigengene_agg, tissue=="CA" & gene_type=="seed gene" & tissue!="Normal skin" & sorting!="Unsorted"&
                                    seed_gene!="Krt15" & seed_gene!="Prom1" & seed_gene!="Prom2")
        ca_seedgene = ggplot(data=eigengene_aggset, aes(x = sorting, y = eigengene_value, color=test_result)) + 
          geom_point(size=1.5, alpha=1) +
          geom_line(size=0.5, aes(group=seed_gene)) +
          geom_errorbar(data=eigengene_aggset, aes(sorting, ymin=eigengene_value-abs(se), ymax=eigengene_value+abs(se)), 
                        alpha=0.5, width=0.1, size=0.8) +
          #facet_wrap(~seed_gene, scales = "free") +
          facet_wrap(~ factor(seed_gene, levels=c(
            "Bmi1","Sox4","Lrig1")), scales = "free", nrow=1) +
          xlab('') +
          ylab('') +
          ggtitle("") +
          theme(axis.text=element_text(size=14,color='black'),
                axis.title=element_text(size=0,color='black')) +
          theme(strip.background = element_rect(fill = 'white', color='black', size=1)) +
          theme(plot.title = element_text(size = 0)) +
          theme(panel.background=element_blank(),
                panel.grid.major=element_blank(),
                panel.grid.minor=element_blank())+
          theme(axis.line = element_line(colour = 'black')) +
          theme(legend.position = 'none') +
          theme(axis.line = element_line(colour = 'black')) +
          theme(panel.background = element_rect(colour = "black")) +
          theme(panel.spacing = unit(0, "lines")) +
          theme_classic() +
          theme(legend.position="none") +
          theme(axis.text.x=element_text(size=14,angle=0,color='black')) +
          #theme(axis.text.y=element_text(size=25,angle=0,color='black')) +
          scale_x_discrete(breaks=c("Lgr6","Progeny"),
                           labels=c("L", "P")) +
          theme(strip.background = element_blank(),
                axis.ticks.x = element_blank(),
                strip.text.x = element_text(size = 16)) +
          scale_color_manual(name="",values = c("significant decrease" = "blue",
                                                "significant increase" = "red",
                                                "not significant" = "grey"))
        
        png("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Figs/Fig5/ca_seedgene_cluster1.png",
            width=4,height=3,units="in",res=200)
        print(ca_seedgene)
        dev.off()
        
        
        #-----------------------------------------------------------------------------------------------------------------------#
        #-----------------------------------------------------------------------------------------------------------------------#
        #-----------------------------------------------------------------------------------------------------------------------#
        #-----------------------------------------------------------------------------------------------------------------------#
        
        
        
        #LINE PLOT Papilloma eigengene
        eigengene_aggset = subset(eigengene_agg, tissue=="Pap" & gene_type=="eigengene" & tissue!="Normal skin" & sorting!="Unsorted")
        pap_eigengene = ggplot(data=eigengene_aggset, aes(x = sorting, y = eigengene_value, color=test_result)) + 
          geom_point(size=1.5, alpha=1) +
          geom_line(size=0.5, aes(group=seed_gene)) +
          geom_errorbar(data=eigengene_aggset, aes(sorting, ymin=eigengene_value-abs(se), ymax=eigengene_value+abs(se)), 
                        alpha=0.5, width=0.1, size=0.8) +
          #facet_wrap(~seed_gene, scales = "free") +
          facet_wrap(~ factor(seed_gene, levels=c("Lgr5",
                                                  "Bmi1","Sox4","Lrig1",
                                                  "Lgr6","Sox9",
                                                  "Sox2","Prom1","Foxa1","Krt15","Krt19","Pitx1","Psca","Prom2","Klf5",
                                                  "Cd44")), scales = "free", nrow=1) +
          
          xlab('') +
          ylab('') +
          ggtitle("") +
          theme(axis.text=element_text(size=14,color='black'),
                axis.title=element_text(size=0,color='black')) +
          theme(strip.background = element_rect(fill = 'white', color='black', size=1)) +
          theme(plot.title = element_text(size = 0)) +
          theme(panel.background=element_blank(),
                panel.grid.major=element_blank(),
                panel.grid.minor=element_blank())+
          theme(axis.line = element_line(colour = 'black')) +
          theme(legend.position = 'none') +
          theme(axis.line = element_line(colour = 'black')) +
          theme(panel.background = element_rect(colour = "black")) +
          theme(panel.spacing = unit(0, "lines")) +
          theme_classic() +
          theme(legend.position="none") +
          theme(axis.text.x=element_text(size=0,angle=0,color='black')) +
          #theme(axis.text.y=element_text(size=25,angle=0,color='black')) +
          scale_x_discrete(breaks=c("Lgr6","Progeny"),
                           labels=c("L", "P")) +
          theme(strip.background = element_blank(),
                axis.ticks.x = element_blank(),
                strip.text.x = element_text(size = 16)) +
          scale_color_manual(name="",values = c("significant decrease" = "blue",
                                                "significant increase" = "red",
                                                "not significant" = "grey"))
        
        png("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Figs/Fig5/pap_eigengene.png",
            width=21,height=3,units="in",res=200)
        print(pap_eigengene)
        dev.off()
        
        #LINE PLOT Papilloma seed ene
        eigengene_aggset = subset(eigengene_agg, tissue=="Pap" & gene_type=="seed gene" & tissue!="Normal skin" & sorting!="Unsorted")
        pap_seedgene = ggplot(data=eigengene_aggset, aes(x = sorting, y = eigengene_value, color=test_result)) + 
          geom_point(size=1.5, alpha=1) +
          geom_line(size=0.5, aes(group=seed_gene)) +
          geom_errorbar(data=eigengene_aggset, aes(sorting, ymin=eigengene_value-abs(se), ymax=eigengene_value+abs(se)), 
                        alpha=0.5, width=0.1, size=0.8) +
          #facet_wrap(~seed_gene, scales = "free") +
          facet_wrap(~ factor(seed_gene, levels=c("Lgr5",
                                                  "Bmi1","Sox4","Lrig1",
                                                  "Lgr6","Sox9",
                                                  "Sox2","Prom1","Foxa1","Krt15","Krt19","Pitx1","Psca","Prom2","Klf5",
                                                  "Cd44")), scales = "free", nrow=1) +
          
          xlab('') +
          ylab('') +
          ggtitle("") +
          theme(axis.text=element_text(size=14,color='black'),
                axis.title=element_text(size=0,color='black')) +
          theme(strip.background = element_rect(fill = 'white', color='black', size=1)) +
          theme(plot.title = element_text(size = 0)) +
          theme(panel.background=element_blank(),
                panel.grid.major=element_blank(),
                panel.grid.minor=element_blank())+
          theme(axis.line = element_line(colour = 'black')) +
          theme(legend.position = 'none') +
          theme(axis.line = element_line(colour = 'black')) +
          theme(panel.background = element_rect(colour = "black")) +
          theme(panel.spacing = unit(0, "lines")) +
          theme_classic() +
          theme(legend.position="none") +
          theme(axis.text.x=element_text(size=0,angle=0,color='black')) +
          #theme(axis.text.y=element_text(size=25,angle=0,color='black')) +
          scale_x_discrete(breaks=c("Lgr6","Progeny"),
                           labels=c("L", "P")) +
          theme(strip.background = element_blank(),
                axis.ticks.x = element_blank(),
                strip.text.x = element_text(size = 16)) +
          scale_color_manual(name="",values = c("significant decrease" = "blue",
                                                "significant increase" = "red",
                                                "not significant" = "grey"))
        
        png("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Figs/Fig5/pap_seedgene.png",
            width=21,height=3,units="in",res=200)
        print(pap_seedgene)
        dev.off()
        
        #----------------------------------------------------------------------------------------------------------------------------#
        #----------------------------------------------------------------------------------------------------------------------------#
        #----------------------------------------------------------------------------------------------------------------------------#
        #Fig.5 Fig5 percentage of FACS sorting fraction
        
        parenchymas = unique(colData(cds_comb)$parenchyma_tissue)
        parenchyma_proportion = data.frame(table(colData(cds_comb)$parenchyma_tissue) / dim(cds_comb)[2])
        
        tissues = unique(colData(cds_comb)$tissue_updated)
        metadata_df = colData(cds_comb)
        head(metadata_df)
        
        freq_comb = do.call(rbind, lapply(1:length(tissues), function(i){
          #i=1
          meta_sub = subset(metadata_df, tissue_updated == tissues[i])
          frequency_sum = data.frame(tissues[i], round((table(meta_sub$sorting) / nrow(meta_sub))*100,1))
          names(frequency_sum) = c("tissue","sorting","percentage")
          frequency_sum
        }))
        
        #get rid of Unsorted
        freq_comb = subset(freq_comb, sorting!="Unsorted")
        
        sorting_fraction = ggplot(data=freq_comb, aes(x=tissue, y=percentage, group=sorting)) +
          geom_point(alpha=0.5, aes(color=sorting), size=2) +
          geom_line(aes(color=sorting)) +
          theme_classic() +
          xlab("tissue") +
          ylab("cell fraction (%)") +
          scale_x_discrete(limits = c("Normal skin","Pap","CA"),
                           labels=c("NSk","Pap","SCC")) +
          theme(legend.position="none") +
          scale_color_manual(name="",values = c("Lgr6" = "darkolivegreen4",
                                                "Progeny" = "goldenrod3"))
        
        png("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Figs/Fig5/sorting_fraction.png",
            width=3,height=2,units="in",res=200)
        print(sorting_fraction)
        dev.off()
        
        sorting_fraction = ggplot(data=freq_comb, aes(x=tissue, y=percentage, group=sorting)) +
          geom_point(alpha=0.5, aes(color=sorting), size=2) +
          geom_line(aes(color=sorting)) +
          theme_classic() +
          xlab("tissue") +
          ylab("cell fraction (%)") +
          scale_x_discrete(limits = c("Normal skin","Pap","CA"),
                           labels=c("NSk","Pap","SCC")) +
          theme(legend.position="right") +
          scale_color_manual(name="",values = c("Lgr6" = "darkolivegreen4",
                                                "Progeny" = "goldenrod3"))
        
        png("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Figs/Fig5/sorting_fraction_w_legend.png",
            width=3,height=2,units="in",res=200)
        print(sorting_fraction)
        dev.off()
        
        #----------------------------------------------------------------------------------------------------------------------------#
        #----------------------------------------------------------------------------------------------------------------------------#
        #----------------------------------------------------------------------------------------------------------------------------#
        #Calculate proportion of FACS sorting in each cluster
        #Exlude tiny cluster 46 bc it has no Lgr6+ cells
        
        metadat = colData(big_cds)
        clusts = unique(metadat$cluster)
        
        sorted_clust = do.call(rbind, lapply(c(1:45,47:49), function(i) {
          #i=1
          print(i)
          clustdf = subset(metadat, cluster==i)
          res_df = data.frame(i, t(data.frame( round(( table(clustdf$sorting) / nrow(clustdf) )*100,2) )))
          names(res_df) = res_df[1,]
          names(res_df)[1] = "cell_cluster"
          res_df[-1,]
        }))
        
        
        #-------------------------------#
        #old plots below
        
        head(eigengene_agg)
        eigengene_agg$tissue = gsub("CA","SCC",eigengene_agg$tissue)
        eigengene_agg$tissue = gsub("Pap","Papilloma",eigengene_agg$tissue)
        
        eigengene_agg$eigengene_tissue = paste(eigengene_agg$eigengene, eigengene_agg$tissue)
        factor_levs = unique(eigengene_agg$eigengene_tissue)
        
        slopes = do.call(rbind, lapply(1:length(factor_levs), function(i){
          #i=1
          eigen_sub = subset(eigengene_agg, eigengene_tissue==factor_levs[i])
          lmdf = lm(eigengene_value~order,data=eigen_sub)
          slope = data.frame(eigen_sub[1,c(9,1,10)], lmdf$coefficients[2])
          names(slope)[4] = "slope"
          slope
        }))
        slopes[order(slopes$slope),]
        
        #add positivee or negativee call
        slopes = do.call(rbind, lapply(1:nrow(slopes), function(i){
          #i=1
          if(slopes[i,]$slope<0) {slope_sign = "negative"} else {slope_sign = "positive"}
          data.frame(slopes[i,],slope_sign)
          
        }))
        
        slope_plot = ggplot(slopes, aes(x=eigengene, y=tissue, color=slope_sign)) +
          geom_point(aes(size=abs(slope)), alpha=0.8) +
          theme(strip.background = element_rect(colour="white", fill="white", 
                                                size=1, linetype="solid")) +
          ylab("") +
          xlab("") +
          ggtitle("") +
          scale_size_continuous(range = c(2,15)) +
          theme(panel.grid.major=element_line(colour = "darkgrey"), 
                panel.grid.minor=element_blank(), 
                panel.background=element_blank()) +
          theme(legend.key = element_rect(fill = "white")) +
          theme(legend.text=element_text(size=15), legend.title=element_text(size=20)) +
          theme(axis.text=element_text(size=15,color='black'),
                axis.text.x=element_text(size=15,color='black',angle=90,hjust=1.2, vjust=0.7),
                axis.title=element_text(size=0),
                title=element_text(size=0),
                axis.ticks = element_line(size=0)) +
          theme(legend.position="right") +
          scale_color_manual(name="",values = c("positive" = "darkseagreen4",
                                                "negative" = "darkgoldenrod3")) +
          scale_x_discrete(limits=c("Lgr5",
                                    "Bmi1","Sox4","Lrig1",
                                    "Lgr6","Sox9","Nanos1",
                                    "Sox2","Prom1","Foxa1","Krt15","Krt19","Pitx1","Psca","Prom2","Klf5",
                                    "Cd44"))
        
        #print out eigengene plots
        png("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Figs/Fig5/slopes_withlegend.png",
            width=10,height=4,units="in",res=200)
        print(slope_plot + theme(legend.position="bottom"))
        dev.off()
        
        png("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Figs/Fig5/slopes_nolegend.png",
            width=7.5,height=3,units="in",res=200)
        print(slope_plot + theme(legend.position="none"))
        dev.off()
        
        
        #pick Cd44 and plot out slope and line of best fit
        #Facet grid plot
        
        cd44_dat = subset(eigengene_agg, eigengene_tissue=="Cd44 SCC")
        lm(eigengene_value ~ order, data = cd44_dat)
        cd44_scc_plot = ggplot(data=cd44_dat, aes(x = order, y = eigengene_value, color=tissue)) + 
          geom_point(size=5, alpha=1) +
          geom_errorbar(data=cd44_dat, aes(order, ymin=eigengene_value-abs(se), ymax=eigengene_value+abs(se)), alpha=1, width=0, size=1.5) +
          stat_smooth(se=F,method="lm",size=0.8) +
          xlab('') +
          ylab('') +
          ggtitle("") +
          theme(axis.text=element_text(size=14,color='black'),
                axis.title=element_text(size=14,color='black')) +
          theme(strip.background = element_rect(fill = 'white', color='black', size=1)) +
          theme(plot.title = element_text(size = 24)) +
          theme(panel.background=element_blank(),
                panel.grid.major=element_blank(),
                panel.grid.minor=element_blank())+
          theme(axis.line = element_line(colour = 'black')) +
          theme(legend.position = 'none') +
          theme(axis.line = element_line(colour = 'black')) +
          theme(panel.background = element_rect(colour = "black")) +
          theme(panel.spacing = unit(0, "lines")) +
          theme_classic() +
          theme(legend.position="none") +
          theme(axis.text.x=element_text(size=0,angle=0,color='black')) +
          theme(axis.text.y=element_text(size=25,angle=0,color='black')) +
          scale_x_continuous(breaks=c(7,8,9),
                             labels=c("Unsorted", "Lgr6+", "Lgr6-progeny")) +
          scale_color_manual(name="",values = c("SCC" = "darkgoldenrod3",
                                                "papilloma" = "darkseagreen4"))
        
        cd44_dat = subset(eigengene_agg, eigengene_tissue=="Cd44 Papilloma")
        lm(eigengene_value ~ order, data = cd44_dat)
        cd44_pap_plot = ggplot(data=cd44_dat, aes(x = order, y = eigengene_value, color=tissue)) + 
          geom_point(size=5, alpha=1) +
          geom_errorbar(data=cd44_dat, aes(order, ymin=eigengene_value-abs(se), ymax=eigengene_value+abs(se)), alpha=1, width=0, size=1.5) +
          stat_smooth(se=F,method="lm",size=0.8) +
          xlab('') +
          ylab('') +
          ggtitle("") +
          theme(axis.text=element_text(size=14,color='black'),
                axis.title=element_text(size=14,color='black')) +
          theme(strip.background = element_rect(fill = 'white', color='black', size=1)) +
          theme(plot.title = element_text(size = 24)) +
          theme(panel.background=element_blank(),
                panel.grid.major=element_blank(),
                panel.grid.minor=element_blank())+
          theme(axis.line = element_line(colour = 'black')) +
          theme(legend.position = 'none') +
          theme(axis.line = element_line(colour = 'black')) +
          theme(panel.background = element_rect(colour = "black")) +
          theme(panel.spacing = unit(0, "lines")) +
          theme_classic() +
          theme(legend.position="none") +
          theme(axis.text.x=element_text(size=0,angle=0,color='black')) +
          theme(axis.text.y=element_text(size=25,angle=0,color='black')) +
          scale_x_continuous(breaks=c(4,5,6),
                             labels=c("Unsorted", "Lgr6+", "Lgr6-progeny")) +
          scale_color_manual(name="",values = c("SCC" = "darkgoldenrod3",
                                                "Papilloma" = "darkseagreen4"))
        
        png("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Figs/Fig5/cd44_scc.png",
            width=4,height=4,units="in",res=200)
        print(cd44_scc_plot)
        dev.off()
        
        png("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Figs/Fig5/cd44_pap.png",
            width=4,height=4,units="in",res=200)
        print(cd44_pap_plot)
        dev.off()
        
        
        
        ####################################################################################################
        # BREAKING IT UP stepwise
        slopes = do.call(rbind, lapply(1:length(factor_levs), function(i){
          #i=1
          
          #Unsorted-> Lgr6
          eigen_sub = subset(eigengene_agg, eigengene_tissue==factor_levs[i])
          lmdf = lm(eigengene_value~order,data=eigen_sub[1:2,]) #only look at unsorted -> Lgr6 row
          slope = data.frame(eigen_sub[1,c(9,1,10)], lmdf$coefficients[2])
          names(slope)[4] = "slope"
          slope$slope_type = "Uns->Lgr6+"
          
          #Lgr6-> Progeny
          lmdf = lm(eigengene_value~order,data=eigen_sub[2:3,]) #only look at Lgr6 -> progeny row
          slope.2 = data.frame(eigen_sub[1,c(9,1,10)], lmdf$coefficients[2])
          names(slope.2)[4] = "slope"
          slope.2$slope_type = "Lgr6+->Progeny"
          
          rbind(slope,slope.2)
          
        }))
        slopes[order(slopes$slope),]
        
        #add positivee or negativee call
        slopes = do.call(rbind, lapply(1:nrow(slopes), function(i){
          #i=1
          if(slopes[i,]$slope<0) {slope_sign = "negative"} else {slope_sign = "positive"}
          data.frame(slopes[i,],slope_sign)
          
        }))
        
        slopes$tissue = gsub("Papilloma","Pap",slopes$tissue)
        slopes$tissue = gsub("Normal skin","NSk",slopes$tissue)
        
        
        slope_plot = ggplot(slopes, aes(x=eigengene, y=tissue, color=slope_sign)) +
          geom_point(aes(size=abs(slope)), alpha=0.8) +
          facet_grid(slope_type ~ .) +
          theme(strip.background = element_rect(colour="white", fill="white", 
                                                size=1, linetype="solid")) +
          ylab("") +
          xlab("") +
          ggtitle("") +
          scale_size_continuous(range = c(2,15)) +
          theme(panel.grid.major=element_line(colour = "darkgrey"), 
                panel.grid.minor=element_blank(), 
                panel.background=element_blank()) +
          theme(legend.key = element_rect(fill = "white")) +
          theme(legend.text=element_text(size=15), legend.title=element_text(size=20)) +
          theme(axis.text=element_text(size=15,color='black'),
                axis.text.x=element_text(size=15,color='black',angle=90,hjust=1.2, vjust=0.7),
                axis.title=element_text(size=0),
                title=element_text(size=0),
                axis.ticks = element_line(size=0)) +
          theme(legend.position="right") +
          scale_color_manual(name="",values = c("positive" = "darkseagreen4",
                                                "negative" = "darkgoldenrod3")) +
          scale_x_discrete(limits=c("Lgr5",
                                    "Bmi1","Sox4","Lrig1",
                                    "Lgr6","Sox9","Nanos1",
                                    "Sox2","Prom1","Foxa1","Krt15","Krt19","Pitx1","Psca","Prom2","Klf5",
                                    "Cd44"))
        
        #print out eigengene plots
        png("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Figs/Fig5/slopes_withlegend_v2.png",
            width=10,height=7,units="in",res=200)
        print(slope_plot + theme(legend.position="bottom"))
        dev.off()
        
        png("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Figs/Fig5/slopes_nolegend_v2.png",
            width=7.5,height=4,units="in",res=200)
        print(slope_plot + theme(legend.position="none"))
        dev.off()
        
        cd44_dat = subset(eigengene_agg, eigengene_tissue=="Cd44 SCC")
        #lm(eigengene_value ~ order, data = cd44_dat)
        cd44_dat$eigengene_value[2] - cd44_dat$eigengene_value[1]
        cd44_dat$eigengene_value[3] - cd44_dat$eigengene_value[2]
        
        cd44_scc_plot = ggplot(data=cd44_dat, aes(x = order, y = eigengene_value, color=tissue)) + 
          geom_point(size=5, alpha=1) +
          geom_line(size=1.5) +
          geom_errorbar(data=cd44_dat, aes(order, ymin=eigengene_value-abs(se), ymax=eigengene_value+abs(se)), alpha=1, width=0, size=1.5) +
          #stat_smooth(se=F,method="lm",size=0.8) +
          xlab('') +
          ylab('') +
          ggtitle("") +
          theme(axis.text=element_text(size=14,color='black'),
                axis.title=element_text(size=14,color='black')) +
          theme(strip.background = element_rect(fill = 'white', color='black', size=1)) +
          theme(plot.title = element_text(size = 24)) +
          theme(panel.background=element_blank(),
                panel.grid.major=element_blank(),
                panel.grid.minor=element_blank())+
          theme(axis.line = element_line(colour = 'black')) +
          theme(legend.position = 'none') +
          theme(axis.line = element_line(colour = 'black')) +
          theme(panel.background = element_rect(colour = "black")) +
          theme(panel.spacing = unit(0, "lines")) +
          theme_classic() +
          theme(legend.position="none") +
          theme(axis.text.x=element_text(size=0,angle=0,color='black')) +
          theme(axis.text.y=element_text(size=25,angle=0,color='black')) +
          scale_x_continuous(breaks=c(7,8,9),
                             labels=c("Unsorted", "Lgr6+", "Lgr6-progeny")) +
          scale_color_manual(name="",values = c("SCC" = "darkred",
                                                "Papilloma" = "purple"))
        
        
        cd44_dat = subset(eigengene_agg, eigengene_tissue=="Cd44 Papilloma")
        cd44_dat$eigengene_value[2] - cd44_dat$eigengene_value[1]
        cd44_dat$eigengene_value[3] - cd44_dat$eigengene_value[2]
        cd44_pap_plot = ggplot(data=cd44_dat, aes(x = order, y = eigengene_value, color=tissue)) + 
          geom_point(size=5, alpha=1) +
          geom_line(size=1.5) +
          geom_errorbar(data=cd44_dat, aes(order, ymin=eigengene_value-abs(se), ymax=eigengene_value+abs(se)), alpha=1, width=0, size=1.5) +
          #stat_smooth(se=F,method="lm",size=0.8) +
          xlab('') +
          ylab('') +
          ggtitle("") +
          theme(axis.text=element_text(size=14,color='black'),
                axis.title=element_text(size=14,color='black')) +
          theme(strip.background = element_rect(fill = 'white', color='black', size=1)) +
          theme(plot.title = element_text(size = 24)) +
          theme(panel.background=element_blank(),
                panel.grid.major=element_blank(),
                panel.grid.minor=element_blank())+
          theme(axis.line = element_line(colour = 'black')) +
          theme(legend.position = 'none') +
          theme(axis.line = element_line(colour = 'black')) +
          theme(panel.background = element_rect(colour = "black")) +
          theme(panel.spacing = unit(0, "lines")) +
          theme_classic() +
          theme(legend.position="none") +
          theme(axis.text.x=element_text(size=0,angle=0,color='black')) +
          theme(axis.text.y=element_text(size=25,angle=0,color='black')) +
          scale_x_continuous(breaks=c(7,8,9),
                             labels=c("Unsorted", "Lgr6+", "Lgr6-progeny")) +
          scale_color_manual(name="",values = c("SCC" = "darkred",
                                                "Papilloma" = "mediumpurple4"))
        
        png("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Figs/Fig5/cd44_scc_v2.png",
            width=5,height=4,units="in",res=200)
        print(cd44_scc_plot)
        dev.off()
        
        png("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Figs/Fig5/cd44_pap_v2.png",
            width=5,height=4,units="in",res=200)
        print(cd44_pap_plot)
        dev.off()
        
        
        #######################################################################################################################
        #######################################################################################################################
        #######################################################################################################################
        #######################################################################################################################
        #######################################################################################################################
        #######################################################################################################################
        
        #Normal skin eigencell heatmap
        #loop Thru eigengenes listed in the eigencell_genes file   
        eigengene_size=100
        eigencell_agg = do.call(rbind, lapply(1:length(seed_gene), function(i){
          tryCatch({ #suppress error
            
            #i=1
            print(i)
            
            #tumor
            if( seed_gene[i]=="Prom1" ) {match_index = 12747} else {
              match_index = match(seed_gene[i], aim2probedat$SYMBOL) #match seed gene to its place in the matrix
            }
            
            symbol_match = data.frame(aim2probedat$SYMBOL, rhodf[,match_index], p_df[,match_index]) #put together gene names, expression data, and p-vals
            symbol_match = symbol_match[order(-symbol_match$rhodf...match_index.),] #order from high to low gene values
            names(symbol_match) = c("gene","rho","p")
            
            #make marker df for grahing eigengene
            marker_df = data.frame(unique(alias2SymbolTable(symbol_match$gene[1:(eigengene_size+1)], species = "Mm")),paste(seed_gene[i],"Carcinoma",sep=""),paste(seed_gene[i],"Carcinoma",sep=""))
            names(marker_df) = c("gene","cell_type","module")
            
            #Create data frame of cells and parenchyma tissue to aggregate over
            #eigen_cell_def = data.frame(row.names(colData(cds_comb)),colData(cds_comb)$parenchyma_tissue) #unbalancec cells
            eigen_cell_def = data.frame(row.names(colData(cds_comb)),colData(cds_comb)$heatmap_vars)
            names(eigen_cell_def) = c("cell_id","cell_group")
            
            #aggregate across cells within all individual genes
            agg_mat = aggregate_gene_expression(
              cds_comb, #unbalanced all cells
              gene_group_df = NULL,
              cell_group_df = eigen_cell_def,
              norm_method = c("log"),
              pseudocount = 1,
              scale_agg_values = TRUE,
              #max_agg_value = 10,
              #min_agg_value = -5,
              exclude.na = TRUE
            )
            
            #extract genes from eigencell
            agg_matsub = agg_mat[row.names(agg_mat) %in% marker_df$gene,] #get cds names that match the balmain sub network names
            
            #sum across genes to get eigengene
            #if(nrow(agg_matsub)>25) {agg_matsub = agg_matsub[1:25,]} else {agg_matsub = agg_matsub} #balance number of eigengenes by setting max as top 25
            sumdf = data.frame(t(data.frame(colSums(agg_matsub)))) #sum all eigengenes within a cell class
            sumdf = sumdf/nrow(agg_matsub) #standardize by eigengene size: divide by number of included genes
            eigensumdf = data.frame(sumdf$NSk.IF, sumdf$NSk.F, sumdf$Pap.nEp, sumdf$Pap.LI,sumdf$Pap.Sp,sumdf$SCC.Sq,sumdf$SCC.Sp) #re-order
            
            row.names(eigensumdf) = paste(seed_gene[i])
            names(eigensumdf) = c("NSk-IF", "NSk-F","Pap-nEp","Pap-LI","Pap-Sp","SCC-Sq","SCC-Sp")
            eigensumdf
            
          },  error=function(e){})
        }))
        
        #now heatmap of those eigencell eigengenes:  columns are eigencells, rows are eigengenes
        virpal = viridis_pal(alpha = 1, begin = 0, end = 1, direction = 1,
                             option = "C") #set the colorblind palette
        eigencell_agg.mat = as.matrix(eigencell_agg)
        
        #Re-order matrix to look like what I want it to betch
        row_order = c("Lgr5","Lgr6","Sox9","Nanos1",
                      "Bmi1","Sox4","Lrig1",
                      "Cd44",
                      "Sox2","Prom1","Foxa1","Krt15","Krt19","Pitx1","Psca","Prom2","Klf5")
        eigencell_agg.mat = eigencell_agg.mat[match(row_order,row.names(eigencell_agg.mat)),]
        
        length(row_order) == nrow(eigencell_agg.mat)
        
        nr = nrow(eigencell_agg.mat)
        pdf(file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Figs/Fig2/heatmaps/Fig2B_eigencell_heatmap_normalskin.pdf",height=6,width=6)
        heatmap.2(eigencell_agg.mat, scale = "none", col = virpal, dendrogram="row",
                  trace = "none", density.info = "none",key.xlab="eigencell eigengene expression",
                  main = "",
                  key=F,
                  cexCol = 1,
                  cexRow = 0.2 + 1/log10(nr),
                  Rowv=F,
                  Colv=F,
                  keysize=1,
                  colsep=c(1:27),
                  rowsep=c(1:28),
                  sepcolor="lightgrey",
                  sepwidth=c(0.02,0.02),
                  adjRow=c(0.3,0.3),
                  srtCol=0,
                  margins=c(7,8),
                  adjCol = c(0.45,0.3),
                  lwid=c(1,1), 
                  lhei=c(0.1,0.6)
        )
        dev.off()
        
        
        #---------------------------------------------------------------------------------#
        #Seed gene only
        #loop Thru eigengenes listed in the eigencell_genes file   
        eigengene_size=100
        eigencell_agg = do.call(rbind, lapply(1:length(seed_gene), function(i){
          tryCatch({ #suppress error
            
            #i=1
            print(i)
            
            #tumor
            if( seed_gene[i]=="Prom1" ) {match_index = 12747} else {
              match_index = match(seed_gene[i], aim2probedat$SYMBOL) #match seed gene to its place in the matrix
            }
            
            symbol_match = data.frame(aim2probedat$SYMBOL, rhodf[,match_index], p_df[,match_index]) #put together gene names, expression data, and p-vals
            symbol_match = symbol_match[order(-symbol_match$rhodf...match_index.),] #order from high to low gene values
            names(symbol_match) = c("gene","rho","p")
            
            #make marker df for grahing eigengene
            marker_df = data.frame(unique(alias2SymbolTable(symbol_match$gene[1:(eigengene_size+1)], species = "Mm")),paste(seed_gene[i],"Carcinoma",sep=""),paste(seed_gene[i],"Carcinoma",sep=""))
            names(marker_df) = c("gene","cell_type","module")
            
            #Create data frame of cells and parenchyma tissue to aggregate over
            #eigen_cell_def = data.frame(row.names(colData(cds_comb)),colData(cds_comb)$parenchyma_tissue) #unbalancec cells
            eigen_cell_def = data.frame(row.names(colData(cds_comb)),colData(cds_comb)$heatmap_vars)
            names(eigen_cell_def) = c("cell_id","cell_group")
            
            #aggregate across cells within all individual genes
            agg_mat = aggregate_gene_expression(
              cds_comb, #unbalanced all cells
              gene_group_df = NULL,
              cell_group_df = eigen_cell_def,
              norm_method = c("log"),
              pseudocount = 1,
              scale_agg_values = TRUE,
              #max_agg_value = 10,
              #min_agg_value = -5,
              exclude.na = TRUE
            )
            
            #extract genes from eigencell
            agg_matsub = agg_mat[row.names(agg_mat) %in% marker_df$gene[1],] #get only seed gene!
            eigensumdf = data.frame(agg_matsub[2], agg_matsub[1],agg_matsub[4],agg_matsub[3],agg_matsub[5],agg_matsub[7],agg_matsub[6]) #re-order
            
            row.names(eigensumdf) = paste(seed_gene[i])
            names(eigensumdf) = c("NSk-IF", "NSk-F","Pap-nEp","Pap-LI","Pap-Sp","SCC-Sq","SCC-Sp")
            eigensumdf
            
          },  error=function(e){})
        }))
        
        #now heatmap of those eigencell eigengenes:  columns are eigencells, rows are eigengenes
        virpal = viridis_pal(alpha = 1, begin = 0, end = 1, direction = 1,
                             option = "C") #set the colorblind palette
        eigencell_agg.mat = as.matrix(eigencell_agg)
        
        #Re-order matrix to look like what I want it to betch
        row_order = c("Lgr5","Lgr6","Sox9","Nanos1",
                      "Bmi1","Sox4","Lrig1",
                      "Cd44",
                      "Sox2","Prom1","Foxa1","Krt15","Krt19","Pitx1","Psca","Prom2","Klf5")
        eigencell_agg.mat = eigencell_agg.mat[match(row_order,row.names(eigencell_agg.mat)),]
        
        length(row_order) == nrow(eigencell_agg.mat)
        
        nr = nrow(eigencell_agg.mat)
        pdf(file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Figs/Fig2/heatmaps/Fig2B_eigencell_heatmap_seedgene.pdf",height=6,width=6)
        heatmap.2(eigencell_agg.mat, scale = "none", col = virpal, dendrogram="row",
                  trace = "none", density.info = "none",key.xlab="eigencell eigengene expression",
                  main = "",
                  key=F,
                  cexCol = 1,
                  cexRow = 0.2 + 1/log10(nr),
                  Rowv=F,
                  Colv=F,
                  keysize=1,
                  colsep=c(1:27),
                  rowsep=c(1:28),
                  sepcolor="lightgrey",
                  sepwidth=c(0.02,0.02),
                  adjRow=c(0.3,0.3),
                  srtCol=0,
                  margins=c(7,8),
                  adjCol = c(0.45,0.3),
                  lwid=c(1,1), 
                  lhei=c(0.1,0.6)
        )
        dev.off()
        
        ##################################################################################################################################
        ##################################################################################################################################
        ##################################################################################################################################
        #Venn diagram overlaps
        
        seed_gene_venn = c("Lgr5","Lgr6","Lrig1","Sox9")
        #-----------------------------------------#
        #Normal skin
        nvsk_venn = do.call(cbind, lapply(1:length(seed_gene_venn), function(i){
          #i=1
          match_index = match(seed_gene_venn[i], aim2probedat$SYMBOL) #match seed gene to its place in the matrix
          symbol_match = data.frame(aim2probedat$SYMBOL, rhodf[,match_index]) #put together gene names, expression data, and p-vals
          symbol_match = symbol_match[order(-symbol_match$rhodf...match_index.),] #order from high to low rho values
          names(symbol_match) = c("gene","rho")
          topgenes = data.frame(symbol_match[1:100,1])  #get top 100 correlated genes
          names(topgenes) = seed_gene_venn[i]
          topgenes
        }))
        
        x = list("Lrig1" = na.omit(nvsk_venn$Lrig1),
                 "Lgr6" = na.omit(nvsk_venn$Lgr6),
                 "Lgr5" = na.omit(nvsk_venn$Lgr5),
                 "Sox9" = na.omit(nvsk_venn$Sox9)
        )
        
        nskvenn_fig = ggVennDiagram(x, color="black",label="count") +
          scale_fill_gradient(
            low = "white",
            high = "darkblue",
            guide = "colourbar",
            aesthetics = "fill") +
          theme(legend.position = "none")
        
        #-----------------------------------------#
        #Carcinoma
        nvsk_venn = do.call(cbind, lapply(1:length(seed_gene_venn), function(i){
          #i=1
          match_index = match(seed_gene_venn[i], aim2probedat$SYMBOL) #match seed gene to its place in the matrix
          symbol_match = data.frame(aim2probedat$SYMBOL, tumor_rhodf[,match_index]) #put together gene names, expression data, and p-vals
          symbol_match = symbol_match[order(-symbol_match$tumor_rhodf...match_index.),] #order from high to low rho values
          names(symbol_match) = c("gene","rho")
          topgenes = data.frame(symbol_match[1:100,1]) #get top 100 correlated genes
          names(topgenes) = seed_gene_venn[i]
          topgenes
        }))
        
        
        x_tumor = list("Lrig1" = na.omit(nvsk_venn$Lrig1),
                       "Lgr6" = na.omit(nvsk_venn$Lgr6),
                       "Lgr5" = na.omit(nvsk_venn$Lgr5),
                       "Sox9" = na.omit(nvsk_venn$Sox9)
        )
        
        tumorvenn_fig = ggVennDiagram(x_tumor, color="black", label="count") +
          scale_fill_gradient(
            low = "white",
            high = "darkred",
            guide = "colourbar",
            aesthetics = "fill") +
          theme(legend.position = "none") +
          theme(text = element_text(size=30))
        
        
        #-------------------------------------------------------------------------------------------------------------#
        #plot NSk and Tumor venn eigengenes together
        png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Extended\ Data/Extedned\ Data\ Fig\ 5/venn_eigengene_overlaps.png",sep=""),
            width=6,height=3,units="in",res=200)
        print(
          multiplot(
            nskvenn_fig,tumorvenn_fig,
            cols=2 
          )
        )
        dev.off()
        
        ##################################################################################################################################
        ##################################################################################################################################
        ####################################################################################################################################################################################################################################################################
        ##################################################################################################################################
        ##################################################################################################################################
        #Fig. 3C
        #individual gene plots
        
        #read in eigengenes
        stem_hierachy_genes = read.xlsx(
          xlsxFile="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/mark_eigengenes/stem_gene_hierarchy_genes.xlsx",
          colNames=T
        )
        
        seed_gene = unique(stem_hierachy_genes$genes)
        data.frame(seed_gene)
        #loop to plot each gene one at a time
        
        
        for(i in 1:length(seed_gene)) {
          tryCatch({ #suppress error
            
            #i=1
            print(i)
            
            #Parenchyma
            cds_ind_gene <- cds_comb[rowData(cds_comb)$gene_short_name %in% seed_gene[i],] #extract individual gene to be plotted #PARENCHYMA
            ind_gene_extract = data.frame( colData(cds_ind_gene)$parenchyma_tissue,(t(cds_ind_gene@assays@data$counts)) ) #extract the data and make a dataframe
            names(ind_gene_extract) = c("tissue","expression")
            ind_gene_extract$gene = seed_gene[i]
            ind_gene_extract.no0 = subset(ind_gene_extract, expression>0) #filter out 0's
            
            #Calculate weighting factors multiply by weighting factor = (1-percentage of cells with 0 expression)
            normal_skin_weight_factor = 1- (nrow(subset(ind_gene_extract,  tissue=="Normal skin" & expression==0)) / nrow(subset(ind_gene_extract,  tissue=="Normal skin")))
            normal_skin_df = subset(ind_gene_extract.no0,  tissue=="Normal skin") #get only cells with expression
            normal_skin_df$expression = normal_skin_df$expression * normal_skin_weight_factor
            
            pap_weight_factor = 1- (nrow(subset(ind_gene_extract,  tissue=="Pap" & expression==0)) / nrow(subset(ind_gene_extract,  tissue=="Pap")))
            pap_df = subset(ind_gene_extract.no0,  tissue=="Pap")
            pap_df$expression = pap_df$expression * pap_weight_factor 
            
            squamous_weight_factor = 1- (nrow(subset(ind_gene_extract,  tissue=="Squamous" & expression==0)) / nrow(subset(ind_gene_extract,  tissue=="Squamous")))
            squamous_df = subset(ind_gene_extract.no0,  tissue=="Squamous")
            squamous_df$expression = squamous_df$expression * squamous_weight_factor 
            
            spindle_weight_factor = 1- (nrow(subset(ind_gene_extract,  tissue=="Spindle" & expression==0)) / nrow(subset(ind_gene_extract,  tissue=="Spindle")))
            spindle_df = subset(ind_gene_extract.no0,  tissue=="Spindle")
            spindle_df$expression = spindle_df$expression * spindle_weight_factor 
            
            #combined weighted expression
            weighted_df = rbind(normal_skin_df, pap_df, squamous_df, spindle_df)
            
            parenchyma_ind_gene_plot_weighted = ggplot(weighted_df, aes(x=tissue,y=expression)) +
              facet_grid(. ~ gene) +
              #geom_violin(aes(fill=tissue, color=tissue), width=1) + #idk why this goddam motherfucking violin plot isn't working
              geom_boxplot(outlier.size=0.4, aes(fill=tissue)) +
              ggtitle("scRNAseq: Parenchyma weighted") + xlab("") + ylab("") + 
              theme(axis.text.x=element_text(size=10,color='black'),
                    axis.text.y=element_text(size=10,color='black'),
                    axis.title=element_text(size=10),
                    axis.ticks.x = element_line(size=0),
                    strip.text.x = element_text(size = 15),
                    legend.position="none") +
              scale_fill_manual(values = c("Normal skin" = "lightblue", "Pap" = "plum","Squamous" = "indianred1","Spindle" = "red"),
                                name="Tissue") +
              scale_color_manual(values = c("Normal skin" = "lightblue", "Pap" = "plum","Squamous" = "indianred1","Spindle" = "red"),
                                 name="Tissue") +
              scale_x_discrete(limits=c("Normal skin","Pap","Squamous","Spindle")) +
              theme(
                # Hide panel borders and remove grid lines
                panel.border = element_blank(),
                panel.grid.major = element_blank(),
                panel.grid.minor = element_blank(),
                panel.background = element_blank(),
                element_line(colour = 'black', size = 0.3),
                # Change axis line
                axis.line = element_line(colour = "black"),
                strip.background = element_rect(colour="white",
                                                fill="white")
              )
            
            #Parenchyma plot with Monocle's built-in violing plotting algo: it kind of sucks because it only plots cells with some epxrssion
            parenchyma_ind_gene_plot = plot_genes_violin(cds_ind_gene, group_cells_by="parenchyma_tissue", ncol=1, 
                                                         min_expr = 0,
                                                         log_scale=F,
                                                         pseudocount = 0.1) +
              #geom_boxplot(color="black", width=0.2, outlier.size=0.4) +
              ggtitle("scRNAseq: Parenchyma unweighted") + xlab("") + ylab("") + 
              theme(axis.text.x=element_text(size=10,color='black'),
                    axis.text.y=element_text(size=10,color='black'),
                    axis.title=element_text(size=10),
                    axis.ticks.x = element_line(size=0),
                    strip.text.x = element_text(size = 15),
                    legend.position="none") +
              scale_fill_manual(values = c("Normal skin" = "lightblue", "Pap" = "plum","Squamous" = "indianred1","Spindle" = "red"),
                                name="Tissue") +
              scale_x_discrete(limits=c("Normal skin","Pap","Squamous","Spindle"))
            
            #All cells
            cds_ind_gene <- big_cds[rowData(big_cds)$gene_short_name %in% seed_gene[i],] #extract individual gene to be plotted #PARENCHYMA
            ind_gene_extract = data.frame( colData(cds_ind_gene)$tissue_updated,(t(cds_ind_gene@assays@data$counts)) ) #extract the data and make a dataframe
            names(ind_gene_extract) = c("tissue","expression")
            ind_gene_extract$gene = seed_gene[i]
            ind_gene_extract.no0 = subset(ind_gene_extract, expression>0) #filter out 0's
            
            #Calculate weighting factors multiply by weighting factor = (1-percentage of cells with 0 expression)
            normal_skin_weight_factor = 1- (nrow(subset(ind_gene_extract,  tissue=="Normal skin" & expression==0)) / nrow(subset(ind_gene_extract,  tissue=="Normal skin")))
            normal_skin_df = subset(ind_gene_extract.no0,  tissue=="Normal skin") #get only cells with expression
            normal_skin_df$expression = normal_skin_df$expression * normal_skin_weight_factor
            
            pap_weight_factor = 1- (nrow(subset(ind_gene_extract,  tissue=="Pap" & expression==0)) / nrow(subset(ind_gene_extract,  tissue=="Pap")))
            pap_df = subset(ind_gene_extract.no0,  tissue=="Pap")
            pap_df$expression = pap_df$expression * pap_weight_factor 
            
            ca_weight_factor = 1- (nrow(subset(ind_gene_extract,  tissue=="CA" & expression==0)) / nrow(subset(ind_gene_extract,  tissue=="CA")))
            ca_df = subset(ind_gene_extract.no0,  tissue=="CA")
            ca_df$expression = ca_df$expression * ca_weight_factor 
            
            #combined weighted expression
            weighted_df = rbind(normal_skin_df, pap_df, ca_df)
            
            allcells_ind_gene_plot_weighted = ggplot(weighted_df, aes(x=tissue,y=expression)) +
              facet_grid(. ~ gene) +
              #geom_violin(aes(fill=tissue, color=tissue), width=1) + #idk why this goddam motherfucking violin plot isn't working
              geom_boxplot(outlier.size=0.4, aes(fill=tissue)) +
              ggtitle("scRNAseq: All cells weighted") + xlab("") + ylab("") + 
              theme(axis.text.x=element_text(size=10,color='black'),
                    axis.text.y=element_text(size=10,color='black'),
                    axis.title=element_text(size=10),
                    axis.ticks.x = element_line(size=0),
                    strip.text.x = element_text(size = 15),
                    legend.position="none") +
              scale_fill_manual(values = c("Normal skin" = "lightblue", "Pap" = "plum","CA" = "indianred1"),
                                name="Tissue") +
              scale_x_discrete(limits=c("Normal skin","Pap","CA")) +
              theme(
                # Hide panel borders and remove grid lines
                panel.border = element_blank(),
                panel.grid.major = element_blank(),
                panel.grid.minor = element_blank(),
                panel.background = element_blank(),
                element_line(colour = 'black', size = 0.3),
                # Change axis line
                axis.line = element_line(colour = "black"),
                strip.background = element_rect(colour="white",
                                                fill="white")
              )
            
            
            allcells_ind_gene_plot = plot_genes_violin(cds_ind_gene, group_cells_by="tissue_updated", ncol=1) +
              geom_boxplot(color="black", width=0.3, outlier.size=0.4) +
              ggtitle("scRNAseq: All cells unweighted") + xlab("") + ylab("") + 
              theme(axis.text.x=element_text(size=10,color='black'),
                    axis.text.y=element_text(size=10,color='black'),
                    axis.title=element_text(size=10),
                    axis.ticks.x = element_line(size=0),
                    strip.text.x = element_text(size = 15),
                    legend.position="none") +
              scale_fill_manual(values = c("Normal skin" = "lightblue", "Pap" = "plum","CA" = "indianred1"),
                                name="Tissue") +
              scale_x_discrete(limits=c("Normal skin","Pap","CA"))
            
            #bulk expression plot
            match_index = match(seed_gene[i], aim2probedat$SYMBOL) #match seed gene to its place in the matrix
            ns_bulk = data.frame("Normal skin",t(normal_wt_dat[match_index,])) #extract normal skin
            names(ns_bulk) = c("tissue","expression")
            ca_bulk = data.frame("CA",t(tumor_wt_dat[match_index,])) #extract tumor expression
            names(ca_bulk) = c("tissue","expression")
            bulk_comb = rbind(ns_bulk,ca_bulk) #combine normal skin and tumor
            bulk_comb$gene = seed_gene[i] #add column of gene name in order to facet grid
            
            bulkexpression_ind_gene_plot = ggplot(bulk_comb, aes(x=tissue,y=log2(expression))) +
              facet_grid(. ~ gene) +
              #geom_violin(aes(fill=tissue), width=1) +
              geom_boxplot(outlier.size=0.4, aes(fill=tissue)) +
              ggtitle("Bulk tissue microarray") + xlab("") + ylab("") + 
              theme(axis.text.x=element_text(size=10,color='black'),
                    axis.text.y=element_text(size=10,color='black'),
                    axis.title=element_text(size=10),
                    axis.ticks.x = element_line(size=0),
                    strip.text.x = element_text(size = 15),
                    legend.position="none") +
              scale_fill_manual(values = c("Normal skin" = "lightblue","CA" = "indianred1"),
                                name="Tissue") +
              scale_x_discrete(limits=c("Normal skin","CA")) +
              theme(
                # Hide panel borders and remove grid lines
                panel.border = element_blank(),
                panel.grid.major = element_blank(),
                panel.grid.minor = element_blank(),
                panel.background = element_blank(),
                element_line(colour = 'black', size = 0.3),
                # Change axis line
                axis.line = element_line(colour = "black"),
                strip.background = element_rect(colour="white",
                                                fill="white")
              )
            
            #Normal skin
            symbol_match = data.frame(aim2probedat$SYMBOL, rhodf[,match_index], p_df[,match_index]) #put together gene names, expression data, and p-vals
            symbol_match = symbol_match[order(-symbol_match$rhodf...match_index.),] #order from high to low gene values
            names(symbol_match) = c("gene","rho","p")
            
            #now make the eigengene and post that shit up in here
            #need to make eigengene for 
            
            
            
            png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/stem_cell_mansucript/Figs/Fig3/Fig3C_ind_gene_plots/",seed_gene[i],"_parenchyma_expression_violinplot.png",sep=""),
                width=11,height=4,units="in",res=300)
            
            print(
              multiplot(bulkexpression_ind_gene_plot, allcells_ind_gene_plot_weighted,parenchyma_ind_gene_plot_weighted,cols=3)
            )
            dev.off()  
            
          },  error=function(e){})
        }
        
        ##################################################################################################################################
        ##################################################################################################################################
        
        #Fig. 3C - UPDATED INTEGRATED PLOTS WITH METASTASES!
        #individual gene plots
        
        #0. read in eigengenes from Allan
        balmain_genes = read.xlsx(
          xlsxFile="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/apriori_eigengenes/Mark_Eigengenes.xlsx",
          colNames=T
        )
        seed_gene = t(balmain_genes[1,]) #get dataframe of all seed genes
        names(seed_gene)[1] = "seed_gene_id"
        
        #GPX4 and Slc4a11
        seed_gene = data.frame(c("Slc7a11","Gpx4"))
        names(seed_gene)[1] = "seed_gene_id"
        seed_gene = c("Slc7a11","Gpx4")
        
        #1a. read in AIM2 bulk data
        aim2dat = read.table(file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/bulk_network_data/Aim2-4/AIM2_AIM4_BS_CARC_HFC_NOTED_ENTREZ_03_16_15_EXPRESSION.txt",header=T)
        aim2phenodat = read.table(file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/bulk_network_data/Aim2-4/AIM2_AIM4_BS_CARC_HFC_NOTED_ENTREZ_03_16_15_PHENOTYPE_DATA.txt",header=T)
        aim2probedat = read.table(file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/bulk_network_data/Aim2-4/AIM2_AIM4_BS_CARC_HFC_NOTED_ENTREZ_03_16_15_PROBE_DATA.txt",header=T)
        
        #1b. read in Genome Biol 2011 bulk data
        genome_biol_dat = read.table(file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/bulk_network_data/Combined\ data\ set\ from\ Genome\ Biol\ 2011/expr_all_probes.txt",header=T)
        genome_biol_phenodat = read.table(file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/bulk_network_data/Combined\ data\ set\ from\ Genome\ Biol\ 2011/sample_attributes.txt",header=T)
        genome_biol_probedat = read.table(file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/bulk_network_data/Combined\ data\ set\ from\ Genome\ Biol\ 2011/gene_attributes_na31_all_probes.txt",header=T)
        
        #1c. read in Melissa data
        melissa_dat = read.table(file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/bulk_network_data/Melissa\ data\ set\ including\ metastases\ from\ Nat\ Med\ paper/expr_all_probes_20140311.txt",header=T)
        melissa_phenodat = read.delim(file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/bulk_network_data/Melissa\ data\ set\ including\ metastases\ from\ Nat\ Med\ paper/sample_attributes_updated_20140912.txt",header=T)
        melissa_probedat = read.delim(file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/bulk_network_data/Melissa\ data\ set\ including\ metastases\ from\ Nat\ Med\ paper/gene_attributes_MM10_all_probes.txt",header=T)
        melissa_dat[1:10,1:10]
        
        #---------------------------------------------------------------------------------------------#
        #---------------------------------------------------------------------------------------------#
        #Optimize Probe Sets
        #---------------------------------------------------------------------------------------------#
        #Optimize probes for Melissa data
        #Calculate Median Aboslute Deviation for each probe of a gene and choose probe with highest MAD
        #https://support.bioconductor.org/p/70133/
        
        head(melissa_probedat)
        unique_genes = unique(melissa_probedat$symbol)
        
        #loop to ID probes with the highest MAD for a gene
        meliss_optimal_probes = do.call(rbind, lapply(1:length(unique_genes), function(i){
          #i=3
          probes_for_a_gene = subset(melissa_probedat, symbol==unique_genes[i])$IDENTIFIER
          if(length(probes_for_a_gene)>1){
            subdat = melissa_dat[match(probes_for_a_gene,melissa_dat$IDENTIFIER),] #convert to df
            subdat = subdat[,-1] #drop first columns which contains the identifier
            mads = apply(subdat, 1, mad) #calculate mad for each column (each probe)
            optimal_probe = probes_for_a_gene[which(mads==max(mads))] #get the index of the one with the max mad
          } else {optimal_probe = melissa_dat[match(probes_for_a_gene,melissa_dat$IDENTIFIER),]$IDENTIFIER}
          optimal_probe
        }))
        head(meliss_optimal_probes)
        dim(meliss_optimal_probes)
        dim(melissa_probedat)
        length(unique_genes)
        
        melissa_probedat_optimized = melissa_probedat[match(meliss_optimal_probes[,1],melissa_probedat$IDENTIFIER),]
        
        #---------------------------------------------------------------------------------------------#
        #Get the best Probes for Aim2 data
        
        head(aim2probedat)
        aim2dat[1:10,1:10]
        unique_genes = unique(aim2probedat$SYMBOL)
        
        #loop to ID probes with the highest MAD for a gene
        aim2_optimal_probes = do.call(rbind, lapply(1:length(unique_genes), function(i){
          #i=3
          print(i)
          probes_for_a_gene = subset(aim2probedat, SYMBOL==unique_genes[i])$IDENTIFIER
          if(length(probes_for_a_gene)>1){
            subdat = aim2dat[match(probes_for_a_gene,aim2dat$IDENTIFIER),] #convert to df
            subdat = subdat[,-1] #drop first columns which contains the identifier
            mads = apply(subdat, 1, mad) #calculate mad for each column (each probe)
            optimal_probe = probes_for_a_gene[which(mads==max(mads))] #get the index of the one with the max mad
          } else {optimal_probe = aim2dat[match(probes_for_a_gene,aim2dat$IDENTIFIER),]$IDENTIFIER}
          optimal_probe
        }))
        head(aim2_optimal_probes)
        dim(aim2_optimal_probes)
        dim(aim2probedat)
        length(unique_genes)
        
        aim2probedat_optimized = aim2probedat[match(aim2_optimal_probes[,1],aim2probedat$IDENTIFIER),]
        
        #---------------------------------------------------------------------------------------------#
        #---------------------------------------------------------------------------------------------#
        
        #2a. AIM2 subset to specific samples and get sample IDs
        normal_wt_ids = subset(aim2phenodat, TYPE=="NORMAL" & GENOTYPE=="WT")
        normal_wt_ids = normal_wt_ids$IDENTIFIER
        normal_wt_ids = gsub("-",".",normal_wt_ids)  #gsub hyphens to decimals  bc R doesn't allow hyphens in column names and changes hyphens to decimals in the aim2dat dataframe
        tumor_wt_ids = subset(aim2phenodat, TYPE=="TUMOR" & GENOTYPE=="WT")
        tumor_wt_ids = tumor_wt_ids$IDENTIFIER
        tumor_wt_ids = gsub("-",".",tumor_wt_ids)  #gsub hyphens to decimals  bc R doesn't allow hyphens in column names and changes hyphens to decimals in the aim2dat dataframe
        length(tumor_wt_ids)
        
        #2b. Melissa subset to specific samples and get sample IDs
        #add primary tumor + metastasis
        unique(paste(melissa_phenodat$TUMOR_TYPE,melissa_phenodat$HISTOLOGY))
        table(paste(melissa_phenodat$TUMOR_TYPE,melissa_phenodat$HISTOLOGY)) #there are only 3 carcinomas that are both and 1 met that is both
        
        melissa_phenodat$IDENTIFIER = gsub("-",".",melissa_phenodat$IDENTIFIER)  #gsub hyphens to decimals  bc R doesn't allow hyphens in column names and changes hyphens to decimals in the aim2dat dataframe
        melissa_wt_ids = subset(melissa_phenodat, HISTOLOGY=="normal" & GENOTYPE=="WT")$IDENTIFIER
        melissa_pap_ids = subset(melissa_phenodat, HISTOLOGY=="papilloma" & GENOTYPE=="WT")$IDENTIFIER
        melissa_ca_squamous_ids = subset(melissa_phenodat, TUMOR_TYPE=="CA" & HISTOLOGY=="SCC" & GENOTYPE=="WT")$IDENTIFIER
        melissa_ca_spindle_ids = subset(melissa_phenodat,  TUMOR_TYPE=="CA" & HISTOLOGY=="spindle" & GENOTYPE=="WT")$IDENTIFIER
        melissa_met_squamous_ids = subset(melissa_phenodat, TUMOR_TYPE=="MET" & HISTOLOGY=="SCC" & GENOTYPE=="WT")$IDENTIFIER
        melissa_met_spindle_ids = subset(melissa_phenodat,  TUMOR_TYPE=="MET" & HISTOLOGY=="spindle" & GENOTYPE=="WT")$IDENTIFIER
        
        #3a. AIM2 subset expression dat from sample IDs
        normal_wt_dat = subset(aim2dat, select=normal_wt_ids)
        tumor_wt_dat = subset(aim2dat, select=tumor_wt_ids)
        
        #3b. Melissa dat
        melissa_wt_dat = subset(melissa_dat, select=melissa_wt_ids)
        melissa_pap_dat = subset(melissa_dat, select=melissa_pap_ids)
        melissa_squamous_dat = subset(melissa_dat, select=melissa_squamous_ids)
        melissa_spindle_dat = subset(melissa_dat, select=melissa_spindle_ids)
        
        #COME HERE loop to plot each gene one at a time
        
        unique(colData(cds_comb)$integrated_type_bridge)
        
        #make column of normal skin parenchyma
        
        for(i in 1:nrow(seed_gene)) {
          tryCatch({ #suppress error
            
            #i=2
            print(i)
            
            #-----------------------------------------------------------------------------------------------#
            #single-cell plots
            
            #Parenchyma
            cds_ind_gene <- cds_comb[rowData(cds_comb)$gene_short_name %in% as.character(seed_gene[i]),] #extract individual gene to be plotted #PARENCHYMA
            ind_gene_extract = data.frame( (t(cds_ind_gene@assays@data$counts)) , colData(cds_ind_gene) ) #extract the data and make a dataframe
            names(ind_gene_extract)[1] = c("expression")
            ind_gene_extract$gene = seed_gene[i]
            ind_gene_extract.no0 = subset(ind_gene_extract, expression>0) #filter out 0's
            
            #Calculate weighting factors multiply by weighting factor = (1-percentage of cells with 0 expression)
            normal_skin_weight_factor = 1- (nrow(subset(ind_gene_extract,  tissue_updated=="Normal skin" & expression==0)) / nrow(subset(ind_gene_extract,  tissue_updated=="Normal skin")))
            normal_skin_df = subset(ind_gene_extract.no0,  tissue_updated=="Normal skin") #get only cells with expression
            normal_skin_df$expression = normal_skin_df$expression * normal_skin_weight_factor
            
            pap_weight_factor = 1- (nrow(subset(ind_gene_extract,  tissue_updated=="Pap" & expression==0)) / nrow(subset(ind_gene_extract,  tissue_updated=="Pap")))
            pap_df = subset(ind_gene_extract.no0,  tissue_updated=="Pap")
            pap_df$expression = pap_df$expression * pap_weight_factor 
            
            squamous_weight_factor = 1- (nrow(subset(ind_gene_extract,  integrated_type_bridge=="squamous" & expression==0)) / nrow(subset(ind_gene_extract,  integrated_type_bridge=="squamous")))
            squamous_df = subset(ind_gene_extract.no0,  integrated_type_bridge=="squamous")
            squamous_df$expression = squamous_df$expression * squamous_weight_factor 
            
            spindle_weight_factor = 1- (nrow(subset(ind_gene_extract,  integrated_type_bridge=="spindle" & expression==0)) / nrow(subset(ind_gene_extract,  integrated_type_bridge=="spindle")))
            spindle_df = subset(ind_gene_extract.no0,  integrated_type_bridge=="spindle")
            spindle_df$expression = spindle_df$expression * spindle_weight_factor 
            
            #combined weighted expression
            normal_skin_df = normal_skin_df[,c("expression","tissue_updated")]
            names(normal_skin_df)[2] = "tissue"
            pap_df = pap_df[,c("expression","tissue_updated")]
            names(pap_df)[2] = "tissue"
            squamous_df = squamous_df[,c("expression","integrated_type_bridge")]
            names(squamous_df)[2] = "tissue"
            spindle_df = spindle_df[,c("expression","integrated_type_bridge")]
            names(spindle_df)[2] = "tissue"
            
            weighted_df = rbind(normal_skin_df, pap_df, squamous_df, spindle_df)
            unique(weighted_df$tissue)
            
            parenchyma_ind_gene_plot_weighted = ggplot(weighted_df, aes(x=tissue,y=expression)) +
              #facet_grid(. ~ gene) +
              geom_violin(aes(fill=tissue), width=1) +
              geom_boxplot(outlier.size=0, width=0.2, color="black") +
              #geom_jitter(width = 0.1) +
              ggtitle("Seed gene scRNAseq: Parenchyma") + xlab("") + ylab("") + 
              theme(axis.text.x=element_text(size=10,color='black'),
                    axis.text.y=element_text(size=10,color='black'),
                    axis.title=element_text(size=10),
                    axis.ticks.x = element_line(size=0),
                    strip.text.x = element_text(size = 15),
                    legend.position="none") +
              scale_fill_manual(values = c("Normal skin" = "lightblue", "Pap" = "plum","squamous" = "indianred1","spindle" = "red"),
                                name="Tissue") +
              scale_color_manual(values = c("Normal skin" = "lightblue", "Pap" = "plum","squamous" = "indianred1","spindle" = "red"),
                                 name="Tissue") +
              scale_x_discrete(limits=c("Normal skin","Pap","squamous","spindle")) +
              theme(
                # Hide panel borders and remove grid lines
                panel.border = element_blank(),
                panel.grid.major = element_blank(),
                panel.grid.minor = element_blank(),
                panel.background = element_blank(),
                element_line(colour = 'black', size = 0.3),
                # Change axis line
                axis.line = element_line(colour = "black"),
                strip.background = element_rect(colour="white",
                                                fill="white")
              )
            
            #calculate single cell eigengene exprssion value
            #aggregate gene values to get a single expression value
            marker_df = data.frame(balmain_genes[,i],names(balmain_genes)[i])
            marker_df = na.omit(marker_df)
            names(marker_df) = c("gene","cell_type")
            agg_mat_cds_comb = aggregate_gene_expression(
              cds_comb,
              gene_group_df = marker_df,
              cell_group_df = NULL,
              norm_method = c("log"),
              pseudocount = 1,
              scale_agg_values = TRUE,
              #max_agg_value = 5,
              #min_agg_value = -5,
              exclude.na = TRUE
            )
            agg_mat_cds_comb = t(data.frame(agg_mat_cds_comb)) #make eigengene expression into dataframe
            agg_mat_df = data.frame(colData(cds_comb)$parenchyma_tissue ,agg_mat_cds_comb[,1] ) #combine with tissue specificity column
            names(agg_mat_df) = c("tissue","expression")
            agg_mat_df$gene = names(balmain_genes)[i] #add eigengene name from balmain_genes list
            
            parenchyma_eigengene_plot = ggplot(agg_mat_df, aes(x=tissue,y=expression)) +
              facet_grid(. ~ gene) +
              geom_violin(aes(fill=tissue), width=1) +
              geom_boxplot(outlier.size=0, width=0.2, color="black") +
              #geom_jitter(width = 0.1) +
              ggtitle("Eigengene scRNAseq: Parenchyma") + xlab("") + ylab("") + 
              theme(axis.text.x=element_text(size=10,color='black'),
                    axis.text.y=element_text(size=10,color='black'),
                    axis.title=element_text(size=10),
                    axis.ticks.x = element_line(size=0),
                    strip.text.x = element_text(size = 15),
                    legend.position="none") +
              scale_fill_manual(values = c("Normal skin" = "lightblue", "Pap" = "plum","Squamous" = "indianred1","Spindle" = "red"),
                                name="Tissue") +
              scale_color_manual(values = c("Normal skin" = "lightblue", "Pap" = "plum","Squamous" = "indianred1","Spindle" = "red"),
                                 name="Tissue") +
              scale_x_discrete(limits=c("Normal skin","Pap","Squamous","Spindle")) +
              theme(
                # Hide panel borders and remove grid lines
                panel.border = element_blank(),
                panel.grid.major = element_blank(),
                panel.grid.minor = element_blank(),
                panel.background = element_blank(),
                element_line(colour = 'black', size = 0.3),
                # Change axis line
                axis.line = element_line(colour = "black"),
                strip.background = element_rect(colour="white",
                                                fill="white")
              )
            
            
            #All cells
            cds_ind_gene <- big_cds[rowData(big_cds)$gene_short_name %in% seed_gene[i],] #extract individual gene to be plotted #PARENCHYMA
            ind_gene_extract = data.frame( colData(cds_ind_gene)$tissue_updated,(t(cds_ind_gene@assays@data$counts)) ) #extract the data and make a dataframe
            names(ind_gene_extract) = c("tissue","expression")
            ind_gene_extract$gene = seed_gene[i]
            ind_gene_extract.no0 = subset(ind_gene_extract, expression>0) #filter out 0's
            
            #Calculate weighting factors multiply by weighting factor = (1-percentage of cells with 0 expression)
            normal_skin_weight_factor = 1- (nrow(subset(ind_gene_extract,  tissue=="Normal skin" & expression==0)) / nrow(subset(ind_gene_extract,  tissue=="Normal skin")))
            normal_skin_df = subset(ind_gene_extract.no0,  tissue=="Normal skin") #get only cells with expression
            normal_skin_df$expression = normal_skin_df$expression * normal_skin_weight_factor
            
            pap_weight_factor = 1- (nrow(subset(ind_gene_extract,  tissue=="Pap" & expression==0)) / nrow(subset(ind_gene_extract,  tissue=="Pap")))
            pap_df = subset(ind_gene_extract.no0,  tissue=="Pap")
            pap_df$expression = pap_df$expression * pap_weight_factor 
            
            ca_weight_factor = 1- (nrow(subset(ind_gene_extract,  tissue=="CA" & expression==0)) / nrow(subset(ind_gene_extract,  tissue=="CA")))
            ca_df = subset(ind_gene_extract.no0,  tissue=="CA")
            ca_df$expression = ca_df$expression * ca_weight_factor 
            
            #combined weighted expression
            weighted_df = rbind(normal_skin_df, pap_df, ca_df)
            
            allcells_ind_gene_plot_weighted = ggplot(weighted_df, aes(x=tissue,y=expression)) +
              facet_grid(. ~ gene) +
              geom_violin(aes(fill=tissue), width=1) +
              geom_boxplot(outlier.size=0, width=0.2, color="black") +
              #geom_jitter(width = 0.1) +
              ggtitle("Seed gene scRNAseq: All cells") + xlab("") + ylab("") + 
              theme(axis.text.x=element_text(size=10,color='black'),
                    axis.text.y=element_text(size=10,color='black'),
                    axis.title=element_text(size=10),
                    axis.ticks.x = element_line(size=0),
                    strip.text.x = element_text(size = 15),
                    legend.position="none") +
              scale_fill_manual(values = c("Normal skin" = "lightblue", "Pap" = "plum","CA" = "indianred1"),
                                name="Tissue") +
              scale_x_discrete(limits=c("Normal skin","Pap","CA")) +
              theme(
                # Hide panel borders and remove grid lines
                panel.border = element_blank(),
                panel.grid.major = element_blank(),
                panel.grid.minor = element_blank(),
                panel.background = element_blank(),
                element_line(colour = 'black', size = 0.3),
                # Change axis line
                axis.line = element_line(colour = "black"),
                strip.background = element_rect(colour="white",
                                                fill="white")
              )
            
            #calculate single cell eigengene exprssion value
            #aggregate gene values to get a single expression value
            marker_df = data.frame(balmain_genes[,i],names(balmain_genes)[i])
            marker_df = na.omit(marker_df)
            names(marker_df) = c("gene","cell_type")
            agg_mat_cds_comb = aggregate_gene_expression(
              big_cds,
              gene_group_df = marker_df,
              cell_group_df = NULL,
              norm_method = c("log"),
              pseudocount = 1,
              scale_agg_values = TRUE,
              #max_agg_value = 5,
              #min_agg_value = -5,
              exclude.na = TRUE
            )
            agg_mat_cds_comb = t(data.frame(agg_mat_cds_comb)) #make eigengene expression into dataframe
            agg_mat_df = data.frame(colData(big_cds)$tissue_updated ,agg_mat_cds_comb[,1] ) #combine with tissue specificity column
            names(agg_mat_df) = c("tissue","expression")
            agg_mat_df$gene = names(balmain_genes)[i] #add eigengene name from balmain_genes list
            
            allcells_eigengene_plot = ggplot(agg_mat_df, aes(x=tissue,y=expression)) +
              facet_grid(. ~ gene) +
              geom_violin(aes(fill=tissue), width=1) +
              geom_boxplot(outlier.size=0, width=0.2, color="black") +
              #geom_jitter(width = 0.1) +
              ggtitle("Eigengene scRNAseq: All cells") + xlab("") + ylab("") + 
              theme(axis.text.x=element_text(size=10,color='black'),
                    axis.text.y=element_text(size=10,color='black'),
                    axis.title=element_text(size=10),
                    axis.ticks.x = element_line(size=0),
                    strip.text.x = element_text(size = 15),
                    legend.position="none") +
              scale_fill_manual(values = c("Normal skin" = "lightblue", "Pap" = "plum","CA" = "indianred1"),
                                name="Tissue") +
              scale_x_discrete(limits=c("Normal skin","Pap","CA")) +
              theme(
                # Hide panel borders and remove grid lines
                panel.border = element_blank(),
                panel.grid.major = element_blank(),
                panel.grid.minor = element_blank(),
                panel.background = element_blank(),
                element_line(colour = 'black', size = 0.3),
                # Change axis line
                axis.line = element_line(colour = "black"),
                strip.background = element_rect(colour="white",
                                                fill="white")
              )
            
            #-----------------------------------------------------------------------------------------------#
            #-----------------------------------------------------------------------------------------------#
            #-----------------------------------------------------------------------------------------------#
            #Bulk expression plots
            
            #Aim2
            #Individual gene bulk sample
            optimal_probe = aim2probedat_optimized[match(seed_gene[i], aim2probedat_optimized$SYMBOL),] #match seed gene to the optimal probe
            match_index = match(optimal_probe$IDENTIFIER, aim2probedat$IDENTIFIER) #now match optimal probe to its place in the matrix
            
            ns_bulk = data.frame("Normal skin",t(normal_wt_dat[match_index,])) #extract normal skin
            names(ns_bulk) = c("tissue","expression")
            ca_bulk = data.frame("CA",t(tumor_wt_dat[match_index,])) #extract tumor expression
            names(ca_bulk) = c("tissue","expression")
            bulk_comb = rbind(ns_bulk,ca_bulk) #combine normal skin and tumor
            bulk_comb$gene = seed_gene[i] #add column of gene name in order to facet grid
            
            bulkexpression_ind_gene_plot = ggplot(bulk_comb, aes(x=tissue,y=log2(expression))) +
              facet_grid(. ~ gene) +
              geom_violin(aes(fill=tissue), width=1) +
              geom_boxplot(outlier.size=0, width=0.2, color="black") +
              #geom_jitter(width = 0.1) +
              ggtitle("Seed gene: Bulk tissue microarray") + xlab("") + ylab("") + 
              theme(axis.text.x=element_text(size=10,color='black'),
                    axis.text.y=element_text(size=10,color='black'),
                    axis.title=element_text(size=10),
                    axis.ticks.x = element_line(size=0),
                    strip.text.x = element_text(size = 15),
                    legend.position="none") +
              scale_fill_manual(values = c("Normal skin" = "lightblue","CA" = "indianred1"),
                                name="Tissue") +
              scale_x_discrete(limits=c("Normal skin","CA")) +
              theme(
                # Hide panel borders and remove grid lines
                panel.border = element_blank(),
                panel.grid.major = element_blank(),
                panel.grid.minor = element_blank(),
                panel.background = element_blank(),
                element_line(colour = 'black', size = 0.3),
                # Change axis line
                axis.line = element_line(colour = "black"),
                strip.background = element_rect(colour="white",
                                                fill="white")
              )
            
            #Aim2
            #Eigengene bulk sample
            optimal_probe = aim2probedat_optimized[match(balmain_genes[,i], aim2probedat_optimized$SYMBOL),] #match seed gene to the optimal probe
            match_index = match(optimal_probe$IDENTIFIER, aim2probedat$IDENTIFIER) #now match optimal probe to its place in the matrix
            
            ns_bulk = data.frame("Normal skin",t(normal_wt_dat[match_index,])) #extract normal skin
            ns_bulk_eigengene = rowSums(ns_bulk[,c(2:ncol(ns_bulk))],na.rm = T)
            ns_bulk_eigengene = data.frame("Normal skin", ns_bulk_eigengene)
            names(ns_bulk_eigengene) = c("tissue","expression")
            
            ca_bulk = data.frame("CA",t(tumor_wt_dat[match_index,])) #extract tumor expression
            ca_bulk_eigengene = rowSums(ca_bulk[,c(2:ncol(ca_bulk))],na.rm = T)
            ca_bulk_eigengene = data.frame("CA", ca_bulk_eigengene)
            names(ca_bulk_eigengene) = c("tissue","expression")
            
            bulk_comb_eigengene = rbind(ns_bulk_eigengene,ca_bulk_eigengene) #combine normal skin and tumor
            bulk_comb_eigengene$gene = names(balmain_genes)[i] #add column of gene name in order to facet grid
            
            bulkexpression_eigengene_plot = ggplot(bulk_comb_eigengene, aes(x=tissue,y=expression)) +
              facet_grid(. ~ gene) +
              geom_violin(aes(fill=tissue), width=1) +
              geom_boxplot(outlier.size=0, width=0.2, color="black") +
              #geom_jitter(width = 0.1) +
              ggtitle("Eigengene Bulk tissue microarray") + xlab("") + ylab("") + 
              theme(axis.text.x=element_text(size=10,color='black'),
                    axis.text.y=element_text(size=10,color='black'),
                    axis.title=element_text(size=10),
                    axis.ticks.x = element_line(size=0),
                    strip.text.x = element_text(size = 15),
                    legend.position="none") +
              scale_fill_manual(values = c("Normal skin" = "lightblue","CA" = "indianred1"),
                                name="Tissue") +
              scale_x_discrete(limits=c("Normal skin","CA")) +
              theme(
                # Hide panel borders and remove grid lines
                panel.border = element_blank(),
                panel.grid.major = element_blank(),
                panel.grid.minor = element_blank(),
                panel.background = element_blank(),
                element_line(colour = 'black', size = 0.3),
                # Change axis line
                axis.line = element_line(colour = "black"),
                strip.background = element_rect(colour="white",
                                                fill="white")
              )
            
            #MELISSA
            #Individual gene bulk plot
            #i=2
            optimal_probe = melissa_probedat_optimized[match(seed_gene[i], melissa_probedat_optimized$symbol),] #match seed gene to the optimal probe
            match_index = match(optimal_probe$IDENTIFIER, melissa_probedat$IDENTIFIER) #now match optimal probe to its place in the matrix
            
            #now add tissue specificity to melissa genedat
            melissa_wt = data.frame("Normal skin",t(melissa_dat[match_index,match(melissa_wt_ids,colnames(melissa_dat))]))
            names(melissa_wt) = c("tissue","expression")
            melissa_pap = data.frame("Pap",t(melissa_dat[match_index,match(melissa_pap_ids,colnames(melissa_dat))]))
            names(melissa_pap) = c("tissue","expression")
            melissa_ca_squamous = data.frame("Pr-Sq",t(melissa_dat[match_index,match(melissa_ca_squamous_ids,colnames(melissa_dat))]))
            names(melissa_ca_squamous) = c("tissue","expression")
            melissa_ca_spindle = data.frame("Pr-Sp",t(melissa_dat[match_index,match(melissa_ca_spindle_ids,colnames(melissa_dat))]))
            names(melissa_ca_spindle) = c("tissue","expression")
            melissa_met_squamous = data.frame("Met-Sq",t(melissa_dat[match_index,match(melissa_met_squamous_ids,colnames(melissa_dat))]))
            names(melissa_met_squamous) = c("tissue","expression")
            melissa_met_spindle = data.frame("Met-Sp",t(melissa_dat[match_index,match(melissa_met_spindle_ids,colnames(melissa_dat))]))
            names(melissa_met_spindle) = c("tissue","expression")
            
            melissa_combo = rbind(melissa_wt,melissa_pap,
                                  melissa_ca_squamous,melissa_ca_spindle,melissa_met_squamous, melissa_met_spindle) #combine these 
            melissa_combo$gene  = seed_gene[i]
            
            melissa_bulkexpression_ind_gene_plot = ggplot(melissa_combo, aes(x=tissue,y=log2(expression))) +
              facet_grid(. ~ gene) +
              geom_violin(aes(fill=tissue), width=1) +
              geom_boxplot(outlier.size=0, width=0.2, color="black") +
              #geom_jitter(width = 0.1) +
              ggtitle("Seed gene Bulk tissue microarray") + xlab("") + ylab("") + 
              theme(axis.text.x=element_text(size=20,color='black'),
                    axis.text.y=element_text(size=20,color='black'),
                    axis.title=element_text(size=15),
                    axis.ticks.x = element_line(size=0),
                    strip.text.x = element_text(size = 25),
                    legend.position="none") +
              scale_fill_manual(values = c("Normal skin" = "lightblue", "Pap" = "plum",
                                           "Pr-Sq" = "indianred1","Pr-Sp" = "red",
                                           "Met-Sq" = "orange", "Met-Sp" = "darkorange2"),
                                name="Tissue") +
              scale_color_manual(values = c("Normal skin" = "lightblue", "Pap" = "plum",
                                            "Pr-Sq" = "indianred1","Pr-Sp" = "red",
                                            "Met-Sq" = "orange", "Met-Sp" = "darkorange2"),
                                 name="Tissue") +
              scale_x_discrete(limits=c("Normal skin","Pap","Pr-Sq","Pr-Sp",
                                        "Met-Sq","Met-Sp")) +
              theme(
                # Hide panel borders and remove grid lines
                panel.border = element_blank(),
                panel.grid.major = element_blank(),
                panel.grid.minor = element_blank(),
                panel.background = element_blank(),
                element_line(colour = 'black', size = 0.3),
                # Change axis line
                axis.line = element_line(colour = "black"),
                strip.background = element_rect(colour="white",
                                                fill="white")
              )
            
            #MELISSA
            #Eigengene bulk plot
            optimal_probe = melissa_probedat_optimized[match(balmain_genes[,i], melissa_probedat_optimized$symbol),] #match seed gene to the optimal probe
            melissa_match_index = match(optimal_probe$IDENTIFIER, melissa_probedat$IDENTIFIER) #now match optimal probe to its place in the matrix
            
            melissa_bulk_sub = data.frame(t(melissa_dat[melissa_match_index,])) #extract normal skin
            melissa_bulk_sub_eigengene = data.frame(rowSums(melissa_bulk_sub[,c(1:ncol(melissa_bulk_sub))],na.rm = T))
            names(melissa_bulk_sub_eigengene) = c("expression")
            
            #now add tissue specificity to melissa eigengene
            melissa_wt = data.frame("Normal skin",melissa_bulk_sub_eigengene[match(melissa_wt_ids,row.names(melissa_bulk_sub_eigengene)),])
            names(melissa_wt) = c("tissue","expression")
            melissa_pap = data.frame("Pap",melissa_bulk_sub_eigengene[match(melissa_pap_ids,row.names(melissa_bulk_sub_eigengene)),])
            names(melissa_pap) = c("tissue","expression")
            melissa_ca_squamous = data.frame("Pr-Sq",melissa_bulk_sub_eigengene[match(melissa_ca_squamous_ids,row.names(melissa_bulk_sub_eigengene)),])
            names(melissa_ca_squamous) = c("tissue","expression")
            melissa_ca_spindle = data.frame("Pr-Sp",melissa_bulk_sub_eigengene[match(melissa_ca_spindle_ids,row.names(melissa_bulk_sub_eigengene)),])
            names(melissa_ca_spindle) = c("tissue","expression")
            melissa_met_squamous = data.frame("Met-Sq",melissa_bulk_sub_eigengene[match(melissa_met_squamous_ids,row.names(melissa_bulk_sub_eigengene)),])
            names(melissa_met_squamous) = c("tissue","expression")
            melissa_met_spindle = data.frame("Met-Sp",melissa_bulk_sub_eigengene[match(melissa_met_spindle_ids,row.names(melissa_bulk_sub_eigengene)),])
            names(melissa_met_spindle) = c("tissue","expression")
            
            melissa_eigengene_combo = rbind(melissa_wt,melissa_pap,melissa_ca_squamous,melissa_ca_spindle,
                                            melissa_met_squamous,melissa_met_spindle) #combine these 
            melissa_eigengene_combo$gene  = names(balmain_genes)[i]
            
            melissa_bulkexpression_eigengene_plot = ggplot(melissa_eigengene_combo, aes(x=tissue,y=expression)) +
              facet_grid(. ~ gene) +
              geom_violin(aes(fill=tissue), width=1) +
              geom_boxplot(outlier.size=0, width=0.2, color="black") +
              #geom_jitter(width = 0.1) +
              ggtitle("Eigengene Bulk tissue microarray") + xlab("") + ylab("") + 
              theme(axis.text.x=element_text(size=10,color='black'),
                    axis.text.y=element_text(size=10,color='black'),
                    axis.title=element_text(size=10),
                    axis.ticks.x = element_line(size=0),
                    strip.text.x = element_text(size = 15),
                    legend.position="none") +
              scale_fill_manual(values = c("Normal skin" = "lightblue", "Pap" = "plum",
                                           "Pr-Sq" = "indianred1","Pr-Sp" = "red",
                                           "Met-Sq" = "orange", "Met-Sp" = "darkorange2"),
                                name="Tissue") +
              scale_color_manual(values = c("Normal skin" = "lightblue", "Pap" = "plum",
                                            "Pr-Sq" = "indianred1","Pr-Sp" = "red",
                                            "Met-Sq" = "orange", "Met-Sp" = "darkorange2"),
                                 name="Tissue") +
              scale_x_discrete(limits=c("Normal skin","Pap","Pr-Sq","Pr-Sp",
                                        "Met-Sq","Met-Sp")) +
              theme(
                # Hide panel borders and remove grid lines
                panel.border = element_blank(),
                panel.grid.major = element_blank(),
                panel.grid.minor = element_blank(),
                panel.background = element_blank(),
                element_line(colour = 'black', size = 0.3),
                # Change axis line
                axis.line = element_line(colour = "black"),
                strip.background = element_rect(colour="white",
                                                fill="white")
              )
            
            #now make the eigengene and post that shit up in here
            #need to make eigengene for 
            
            png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/stem_cell_mansucript/Figs/Fig3/Fig3C_ind_gene_eigengene_plots/",names(balmain_genes)[i],"_sc_bulk_single_eigen_gene_expression_violinplot.png",sep=""),
                width=16,height=8,units="in",res=300)
            
            print(
              multiplot(bulkexpression_ind_gene_plot, bulkexpression_eigengene_plot,
                        melissa_bulkexpression_ind_gene_plot, melissa_bulkexpression_eigengene_plot,
                        allcells_ind_gene_plot_weighted, allcells_eigengene_plot,
                        parenchyma_ind_gene_plot_weighted, parenchyma_eigengene_plot,
                        cols=4)
            )
            dev.off()  
            
          },  error=function(e){})
        }
        
        ##################################################################################################################################
        ##################################################################################################################################
        ##################################################################################################################################
        ##################################################################################################################################
        ##################################################################################################################################
        ##################################################################################################################################
        ##################################################################################################################################
        
        #Network recruitment across celltypes during progression
        #how peak eigengene expression changes in different cell types during progression.  In other words, how are networks co-opted by cancer at the level of cell-types
        
        #try plotting eigencell valuesand arranging them in different ways
        stages = unique(colData(big_cds)$tissue_updated)
        
        cell_types = c("Epithelial","Outer bulge","Sebaceous Gland","T-cell","Basal cells",
                       "Dendritic cell","Melanocyte","Proliferating Epithelial",
                       "Inner bulge","Neutrophil")
        
        #get specific eigengenes to focus on
        eigencell_genes = read.xlsx(
          xlsxFile="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/stem_cell_mansucript/Figs/Fig3/Fig3B_heatmap_genenames.xlsx",
          colNames=T
        )
        sub_eigengenes = eigencell_genes$eigengene_name
        
        #start loop through eigengenes -> stages -> cell types
        eigengene_network_recruit_sum = do.call(rbind, lapply(1:length(sub_eigengenes), function(i){ #eigengene loop
          
          #i=1
          tryCatch({ #suppress error
            marker_df = data.frame( na.omit(subset(balmain_genes, select = sub_eigengenes[i])), sub_eigengenes[i])
            #only if more than 1 constituent gene
            names(marker_df) = c("gene","cell_type")
            
            #UPDATED NCBI MATCHING!! --> this will product lots of non-matching names that i must resolve with NCBI aliases
            allnetnames = do.call(rbind, lapply(1:nrow(marker_df), function(l){
              #l=1
              netnames = data.frame(marker_df$gene[l],rowData(big_cds)$gene_short_name[match(marker_df$gene[l],rowData(big_cds)$gene_short_name)])
              names(netnames) = c("balmain_network_gene_name","cds_gene_name")
              netnames
            }))
            cds_gene_names = rowData(big_cds)$gene_short_name #get vector of all gene names in cds object
            #get consistent names between cds and network gene names
            match_noms = do.call(rbind, lapply(1:nrow(allnetnames), function(k){
              #k=1 ; 22
              tryCatch({ #suppress error
                if(is.na(allnetnames$cds_gene_name[k])==F) {cds_alias_match = allnetnames$cds_gene_name[k]} else
                { 
                  #search through Gene aliases for a match
                  ncbi_symbol = ncbi_names[grep(allnetnames$balmain_network_gene_name[k],ncbi_names$Aliases),] #search through aliases for a match
                  if (nrow(ncbi_symbol)>0) {
                    ncbi_matches_vector = c(as.character(ncbi_symbol$Symbol), as.character(data.frame(strsplit(as.character(ncbi_symbol$Aliases),", "))[,1]) ) #vector of all aliases that marker_gene matches
                    cds_alias_match = na.omit(cds_gene_names[match(ncbi_matches_vector,cds_gene_names)]) #match this symbol to the cds gene names
                  } else {
                    #search through Gene Symbols
                    ncbi_symbol.1 = ncbi_names[grep(allnetnames$balmain_network_gene_name[k],ncbi_names$Symbol),] 
                    ncbi_matches_vector = c(as.character(ncbi_symbol.1$Symbol), as.character(data.frame(strsplit(as.character(ncbi_symbol.1$Aliases),", "))[,1]) ) #vector of all aliases that marker_gene matches
                    cds_alias_match = na.omit(cds_gene_names[match(ncbi_matches_vector,cds_gene_names)]) #match this symbol to the cds gene names
                  }
                }
                data.frame(allnetnames[k,],cds_alias_match)
              },  error=function(e){})
            }))
            
            #remake the marker df dataframe with the matched gene names
            marker_df = data.frame(na.omit(unique(match_noms$cds_alias_match)),sub_eigengenes[i])
            names(marker_df) = c("gene","cell_type")
            
            tissue_stage_sum = do.call(cbind, lapply(1:length(stages), function(k){ #stage loop
              #k=1
              #subset to a single tissue stage
              sub_cds = big_cds[, colData(big_cds)$tissue_updated %in% stages[k]]
              
              cell_type_sum = do.call(rbind, lapply(1:length(cell_types), function(j){ #cell type loop
                #j=10
                tryCatch({ #suppress error
                  print(paste("i=",i," k=",k," j=",j, sep=""))
                  #subset to a specific cell type
                  sub_sub_cds = sub_cds[,colData(sub_cds)$cell_type %in% cell_types[j]]
                  
                  #condition on if a cell_type exists...if not then assign a value of 0
                  if( dim(sub_sub_cds)[2] > 0) {
                    eigen_cell_def = data.frame(row.names(colData(sub_sub_cds)),colData(sub_sub_cds)$cell_type)
                    names(eigen_cell_def) = c("cell_id","cell_group")
                    
                    #aggregate across cells within all individual genes
                    agg_mat = aggregate_gene_expression(
                      sub_sub_cds, #balanced parenchyma cell tissue
                      gene_group_df = NULL,
                      cell_group_df = eigen_cell_def,
                      norm_method = c("log"),
                      pseudocount = 1,
                      scale_agg_values = TRUE,
                      #max_agg_value = 10,
                      #min_agg_value = -5,
                      exclude.na = TRUE
                    )
                    
                    #extract genes from eigencell
                    agg_matsub = agg_mat[row.names(agg_mat) %in% marker_df$gene,] #get cds names that match the balmain sub network names
                    eigensum = data.frame(sum(agg_matsub)) #just add up the genes now
                    row.names(eigensum) = cell_types[j]
                    names(eigensum) = stages[k]
                  } else {
                    eigensum = data.frame(0)
                    row.names(eigensum) = cell_types[j]
                    names(eigensum) = stages[k]
                  }
                  eigensum
                  
                },  error=function(e){})
                
              })) #end cell type loop
              
              cell_type_sum
              
            })) #end tissue stage loop
            
            tissue_stage_sum$celltype = row.names(tissue_stage_sum) #add column of cell types
            
            #plot Network recruitment by cell type
            tissue_stage_sum.dfm = melt(tissue_stage_sum, id=c("celltype"))
            netrecruit_plot = ggplot() +
              geom_point(data=tissue_stage_sum.dfm, aes(x=celltype, y=variable, 
                                                        size=value, fill=value, color=value), alpha=0.8, shape=21) +
              labs(title=paste(sub_eigengenes[i]))+
              theme(axis.text.y=element_text(size=17,color='black'),
                    axis.text.x=element_text(size=20,color='black', angle=90, hjust=1,vjust=0.4),
                    axis.title.x=element_text(size=0,color='black'),
                    axis.title.y=element_text(size=0,color='black')) +
              theme(plot.title = element_text(size = 20, vjust=1)) +
              theme(panel.grid.major=element_blank(), 
                    panel.grid.minor=element_blank(), 
                    panel.background=element_blank()) +
              theme(legend.position="none") +
              scale_size_continuous(range = c(0.5,20)) +
              scale_colour_gradient(low = "grey", high = "darkblue") +
              scale_fill_gradient(low = "grey", high = "darkblue") +
              theme(axis.ticks = element_blank())
            
            png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/figures/network_recruitment_by_celltype/",sub_eigengenes[i],"_network_recruitment_v1.png",sep=""),
                width=8,height=6,units="in",res=200)
            print(netrecruit_plot)
            dev.off()
            
            tissue_stage_sum$eigengene = sub_eigengenes[i]
            return(tissue_stage_sum)
          },  error=function(e){})
          
        })) #end eigengene loop
        
        write.table(eigengene_network_recruit_sum,
                    file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/figures/network_recruitment_by_celltype/network_celltype_eigencell_value.txt",
                    sep="\t",row.names=F,quote=F)
        
        
        ##################################################################################################################################
        ##################################################################################################################################
        ##################################################################################################################################
        ##################################################################################################################################
        ##################################################################################################################################
        
        #calculate percentages of cells with and without Olfr expression
        
        olfr_names = row.names(cds_comb)[grep("Olfr",row.names(cds_comb))]
        Scgb_names = row.names(cds_comb)[grep("Scgb",row.names(cds_comb))]
        
        all_expressing_genes = do.call(rbind, lapply(1:length(olfr_names), function(i){
          #i=2
          tryCatch({ #suppress error
            print(i)
            sub_cds_counts = data.frame(t(data.frame(cds_comb[row.names(cds_comb) %in% olfr_names[i],]@assays@data$counts))) #extract gene counts
            sub_cds = data.frame( colData(cds_comb)$parenchyma_tissue , sub_cds_counts) #add tissue metadata
            names(sub_cds) = c("tissue","gene")
            
            expression_cells_df = do.call(rbind, lapply(1:length(los_parenchymas), function(j){ 
              #j=1
              not_expressing_n = nrow(subset(sub_cds, tissue==los_parenchymas[j] & gene==0) )
              expressing_n = nrow(subset(sub_cds, tissue==los_parenchymas[j] & gene!=0) )
              expressing_percent = round((expressing_n/not_expressing_n)*100,3)
              data.frame(olfr_names[i], los_parenchymas[j], not_expressing_n, expressing_n, expressing_percent)
            }))
            names(expression_cells_df)[1:2] = c("gene","parenchyma")
            carcinoma_normal_difference = (mean(expression_cells_df$expressing_percent[3:4]) - expression_cells_df$expressing_percent[1]) #difference b/w average squamous and spindle and normal skin
            expression_cells_df$carcinoma_normal_difference = carcinoma_normal_difference
            expression_cells_df
            
          },  error=function(e){})
        }))
        
        all_expressing_genes.result = all_expressing_genes[order(-all_expressing_genes$carcinoma_normal_difference),]
        
        write.table(all_expressing_genes.result, file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/Olfrs/Olfr_binary_expression_in_parenchyma.txt",row.names=F,quote=F,sep="\t")
        
        
        ##################################################################################################################################
        ##################################################################################################################################
        ##################################################################################################################################
        ##################################################################################################################################
        ##################################################################################################################################
        ##################################################################################################################################
        ##################################################################################################################################
        ##################################################################################################################################
        ##################################################################################################################################
        ##################################################################################################################################
        
        net_results = read.xlsx(
          xlsxFile="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/bulk_network_data/eigengene_generation_results/Tigit_normalskin_WT.xlsx",
          colNames=T
        )
        
        x = list("Allan" = na.omit(net_results$`Allan_tigit_skinwt_aim2-4`),
                 "Mark v1" = na.omit(net_results$Mark_v1),
                 "Saumya" = na.omit(net_results$Saumya_v1),
                 "Mark v2" = na.omit(net_results$Mark_v2_top200))
        
        ggVennDiagram(x)
        
        x = list("Allan" = na.omit(net_results$`Allan_tigit_skinwt_aim2-4`),
                 "Saumya" = na.omit(net_results$Saumya_v1),
                 "Mark v2 top 300" = na.omit(net_results$Mark_v2_top300))
        
        ggVennDiagram(x)
        
        
        ##################################################################################################################################
        ##################################################################################################################################
        ##################################################################################################################################
        ##################################################################################################################################
        ##################################################################################################################################
        ##################################################################################################################################
        ##################################################################################################################################
        ##################################################################################################################################
        ##################################################################################################################################
        ##################################################################################################################################
        
        ##################################################################################################################################
        ##################################################################################################################################
        ##################################################################################################################################
        ##################################################################################################################################
        ##################################################################################################################################
        ##################################################################################################################################
        ##################################################################################################################################
        ##################################################################################################################################
        ##################################################################################################################################
        ##################################################################################################################################
        #Network architecture
        
        library(bnlearn)
        
        #first_compare = c("UnTxSkin","TxSkin","TransitionalPap","Pap","TransitionalCA","CA")
        first_compare = c("UnTxSkin","TxSkin","Pap","CA")
        #colpal_transitionnetwork.1 = c("lightblue","blue","cyan","green","purple","red")
        colpal_transitionnetwork.1 = c("lightblue","blue","green","red")
        
        
        pdf(file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/figures/network_analysis/networks_v2.pdf",height=16,width=22)
        #for (p in 1:length(first_compare)) {
        all_neighbors =  do.call(rbind, lapply(1:length(first_compare), function(p){
          #p=3
          tryCatch({ #suppress error
            
            cds_comb <- readRDS(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/matvei/cds_objects/",first_compare[p],".RDS",sep=""))
            
            #now start inner loop on networks
            #for(i in 1:length(sheet_names)) {
            #sheet_graph = do.call(rbind, lapply(1:length(sheet_names), function(i){
            sheet_graph = do.call(rbind, lapply(3:3, function(i){
              
              
              tryCatch({ #suppress error
                #i=3
                print(p)
                print(i)
                
                #Go through the horror of getting the names to match
                network_sub = subset(kras_nets, network_name==sheet_names[i])
                subgenes = c("Kras",network_sub$name_2) #add Kras to the list of Kras network names
                
                #match Balmain network names to cds gene names --> this will product lots of non-matching names that i must resolve with NCBI aliases
                unique_net_names = unique(network_sub$name_2) #get all unique gene names in network_sub
                allnetnames = do.call(rbind, lapply(1:length(unique_net_names), function(l){
                  #l=1
                  netnames = data.frame(unique_net_names[l],rowData(cds_comb)$gene_short_name[match(unique_net_names[l],rowData(cds_comb)$gene_short_name)])
                  names(netnames) = c("balmain_network_gene_name","cds_gene_name")
                  netnames
                }))
                
                cds_gene_names = rowData(cds_comb)$gene_short_name #get vector of all gene names in cds object
                #get consistent names between cds and network gene names
                match_noms = do.call(rbind, lapply(1:nrow(allnetnames), function(k){
                  #k=22 ; 22
                  tryCatch({ #suppress error
                    if(is.na(allnetnames$cds_gene_name[k])==F) {cds_alias_match = allnetnames$cds_gene_name[k]} else
                    { ncbi_symbol = ncbi_names[grep(allnetnames$balmain_network_gene_name[k],ncbi_names$Aliases),]$Symbol #get the NCBI gene symbol for a balmain netowkr name
                    cds_alias_match = cds_gene_names[match(ncbi_symbol,cds_gene_names)] #match this symbol to the cds gene names
                    }
                    data.frame(allnetnames[k,],cds_alias_match)
                  },  error=function(e){})
                }))
                
                cds_matchednames = match_noms[match_noms$balmain_network_gene_name %in% subgenes,]$cds_alias_match #get cds names that match the balmain sub network names
                cds_subset <- cds_comb[rowData(cds_comb)$gene_short_name %in% cds_matchednames,] #subset CDS object to most informative name
                
                #-----------------------------#
                #make before transition network
                deg_matx = cds_subset@assays@data@listData$counts
                deg_matx = as.matrix(deg_matx) #convert to normal matrix
                deg_matx = t(deg_matx) #transpose so genes are being correlated
                cormat = rcorr(deg_matx, type = c("pearson")) #calc correlation matrix
                
                #replace non-sig correlation matrix values with 0
                sigmat = do.call(rbind, lapply(1:nrow(cormat$r), function(rr) {
                  #rr=1
                  loscols = do.call(cbind, lapply(1:ncol(cormat$r), function(cc) {
                    #cc=3
                    if((is.na(cormat$P[rr,cc]) ==T)) {cval = 0} else if((cormat$P[rr,cc] < (0.05/60))) {cval = (cormat$r[rr,cc])} else {cval = 0}
                    data.frame(cval)
                  }))
                  loscols
                }))
                row.names(sigmat) = colnames(deg_matx)
                
                #find where rows/cols are all 0 which represent genes with no significant correlation
                rowsumz = data.frame(rowSums(sigmat))
                rowsumz$index = c(1:nrow(rowsumz))
                index_to_drop = subset(rowsumz, rowSums.sigmat.==0)$index
                if(length(index_to_drop)==0) {sigmat = sigmat} else {sigmat = sigmat[-index_to_drop,-index_to_drop]}
                
                #my data
                net = network(sigmat, directed = FALSE)
                net = network(net, directed = FALSE)
                network.vertex.names(net) = row.names(sigmat) #add gene names to nodes
                net %v% "direction" = as.vector(data.frame(strsplit(as.character(names(sigmat)),"[.]"))[1,])
                
                set.seed(1) #this puts the nodes in the same position
                netgraph = ggnet2(net, mode = "kamadakawai",color=colpal_transitionnetwork.1[p],
                                  label = T,label.color = "black", label.size=2,
                                  size="degree",max_size = 21) +
                  theme(legend.position="none") +
                  theme(panel.background = element_rect(fill = "white")) +
                  theme(legend.position="none") +
                  ggtitle(paste(first_compare[p], sheet_names[i],"Network",sep=" ")) +
                  theme(plot.title = element_text(size = 18, vjust=1))
                print(netgraph)
                
                ig2 <- asIgraph(net) #make network object into igraph object
                firstdegneighb = neighborhood.size(ig2, 1)-1 #find number of first degree neighbors
                seconddegneighb = neighborhood.size(ig2, 2)-d1-1 #find number of second degreeneighbors
                
                neighbors = data.frame(netgraph$data[,1],firstdegneighb,seconddegneighb)
                neighbors$p = p
                neighbors$compare_var = first_compare[p]
                neighbors$network = sheet_names[i]
                return(neighbors)
                
                #learn network things
                
              },  error=function(e){})
            }))
            sheet_graph
          },  error=function(e){})
        }))
        dev.off()    
        
        head(all_neighbors)
        names(all_neighbors)[1] = "gene"
        
        colpal_transitionnetwork.dark = c("lightblue","darkblue","darkgreen","darkred")
        
        first_compare = c("UnTxSkin","TxSkin","Pap","CA")
        colpal_transitionnetwork.dark = c("UnTxSkin"="lightblue","TxSkin"="darkblue","Pap"="darkgreen","CA"="darkred")
        
        ggplot(all_neighbors, aes(x=p, y=seconddegneighb)) +
          geom_point(aes(group=gene,color=compare_var),alpha=0.1) +
          geom_line(aes(group=gene),alpha=0.1) +
          #geom_point(aes(group=gene,color=compare_var),alpha=1) +
          #geom_line(aes(group=gene),alpha=0.6) +
          geom_smooth(linetype ="solid", method="lm",alpha=1,se=F,color="black",size=1.3) +
          ggtitle("Second degree neighbors") +
          xlab("tissue") +
          ylab("second deg connectivity") +
          theme(text = element_text(size=18)) +
          theme(axis.text=element_text(size=18,color='black'),
                axis.title=element_text(size=20,color='black')) +
          theme(plot.title = element_text(size = 30, vjust=1)) +
          theme(axis.text=element_text(size=15,color='black'),
                axis.title=element_text(size=15,color='black')) +
          theme(panel.grid.major=element_line(colour = "darkgrey"), 
                panel.grid.minor=element_blank(), 
                panel.background=element_blank()) +
          theme(legend.key = element_rect(fill = "white")) +
          scale_x_continuous(breaks = c(1,2,3,4),
                             labels=c("UnTxSkin", "TxSkin","Pap","CA")) +
          theme(strip.text.x = element_text(size = 20, colour = "black", angle = 0)) +
          theme(strip.text.y = element_text(size = 20, colour = "black", angle = 0)) +
          theme(legend.text=element_text(size=20), legend.title=element_text(size=20)) +
          scale_colour_manual(values=colpal_transitionnetwork.dark, name='tissue') +
          theme(legend.position="none")
        
        #look at individual transitions
        head(all_neighbors)
        sub.plot = subset(all_neighbors, (compare_var == "UnTxSkin" | compare_var =="TxSkin"))
        sub.plot = subset(all_neighbors, (compare_var == "Pap" | compare_var =="TxSkin"))
        sub.plot = subset(all_neighbors, (compare_var == "Pap" | compare_var =="CA"))
        
        ggplot(sub.plot, aes(x=p, y=firstdegneighb)) +
          geom_point(aes(group=gene,color=compare_var),alpha=0.1) +
          geom_line(aes(group=gene),alpha=0.1) +
          #geom_point(aes(group=gene,color=compare_var),alpha=1) +
          #geom_line(aes(group=gene),alpha=0.6) +
          geom_smooth(linetype ="solid", method="lm",alpha=1,se=F,color="black",size=1.3) +
          ggtitle("") +
          xlab("tissue") +
          ylab("first deg connectivity") +
          theme(text = element_text(size=12)) +
          theme(axis.text=element_text(size=12,color='black'),
                axis.title=element_text(size=12,color='black')) +
          theme(plot.title = element_text(size = 30, vjust=1)) +
          theme(axis.text=element_text(size=12,color='black'),
                axis.title=element_text(size=12,color='black')) +
          theme(panel.grid.major=element_line(colour = "darkgrey"), 
                panel.grid.minor=element_blank(), 
                panel.background=element_blank()) +
          theme(legend.key = element_rect(fill = "white")) +
          scale_x_continuous(breaks = c(1,2,3,4),
                             labels=c("UnTxSkin", "TxSkin","Pap","CA")) +
          theme(strip.text.x = element_text(size = 20, colour = "black", angle = 0)) +
          theme(strip.text.y = element_text(size = 20, colour = "black", angle = 0)) +
          theme(legend.text=element_text(size=20), legend.title=element_text(size=20)) +
          scale_colour_manual(values=colpal_transitionnetwork.dark, name='tissue') +
          theme(legend.position="none")
        
        #first deg neighbors vs second degree
        ggplot(all_neighbors, aes(x=firstdegneighb, y=seconddegneighb,color=compare_var)) +
          geom_point(aes(group=gene),alpha=0.3) +
          #geom_point(aes(group=gene,color=compare_var),alpha=1) +
          #geom_line(aes(group=gene),alpha=0.6) +
          geom_smooth(aes(group=compare_var),linetype ="solid",alpha=1,se=F,size=1.1) +
          ggtitle("1 vs 2 degree neighbors") +
          xlab("first deg connectivity") +
          ylab("second deg connectivity") +
          theme(text = element_text(size=18)) +
          theme(axis.text=element_text(size=18,color='black'),
                axis.title=element_text(size=20,color='black')) +
          theme(plot.title = element_text(size = 30, vjust=1)) +
          theme(axis.text=element_text(size=15,color='black'),
                axis.title=element_text(size=15,color='black')) +
          theme(panel.grid.major=element_line(colour = "darkgrey"), 
                panel.grid.minor=element_blank(), 
                panel.background=element_blank()) +
          theme(legend.key = element_rect(fill = "white")) +
          theme(strip.text.x = element_text(size = 20, colour = "black", angle = 0)) +
          theme(strip.text.y = element_text(size = 20, colour = "black", angle = 0)) +
          theme(legend.text=element_text(size=20), legend.title=element_text(size=20)) +
          scale_colour_manual(values=colpal_transitionnetwork.dark, name='tissue') +
          theme(legend.position="bottom")
        
        hub_genes = subset(all_neighbors, firstdegneighb >= 300 & seconddegneighb >= 400)
        hub_genes = hub_genes[order(-hub_genes$firstdegneighb),]
        data.frame(unique(hub_genes$gene))
        
        drug_genes = subset(all_neighbors, firstdegneighb >= 300 & seconddegneighb <= 300)
        drug_genes = drug_genes[order(-drug_genes$firstdegneighb),]
        data.frame(unique(drug_genes$gene))
        
        ######
        
        #pap_ca
        row.names(pap_ca)[grep("C330",row.names(pap_ca))]
        
        ################################################################################################################################    
        ################################################################################################################################    
        ################################################################################################################################
        ################################################################################################################################    
        ################################################################################################################################    
        ################################################################################################################################
        ################################################################################################################################    
        ################################################################################################################################    
        ################################################################################################################################
        ################################################################################################################################      
        #infer copy number variants CNV
        library("infercnv")
        
        #Get matrix of all cell counts
        direcs = c("second_run_CA_Lgr6","second_run_CA_Progeny","second_run_CA_US",
                   "second_run_PAP_Lgr6","second_run_PAP_Tom","second_run_PAP_US",
                   "second_run_Skin_Lgr6","second_run_Skin_Tom",
                   "first_run_LGR6","first_run_NO_TAM_BS","first_run_PROGENY","first_run_TAM_BS")
        samp_names = c("CA Lgr6","CA Progeny","CA Unsorted",
                       "Pap Lgr6","Pap Progeny","Pap Unsorted",
                       "TxSkin Lgr6","TxSkin Progeny",
                       "UnTxSkin Lgr6","UnTxSkin Unsorted","UnTxSkin Progeny","TxSkin Unsorted")
        tx_names = c("Treated","Treated","Treated",
                     "Treated","Treated","Treated",
                     "Treated","Treated",
                     "Untreated","Untreated","Untreated","Treated")
        batch_names = c("Second run","Second run","Second run",
                        "Second run","Second run","Second run",
                        "Second run","Second run",
                        "First run","First run","First run","First run")
        mingene = c(500,250,500,
                    500,500,750,
                    500,500,
                    500,250,350,500 ) #vector of minimum genes per sample to shwo low-quality/empty droplets
        maxgene = c(6000,5500,6000,
                    5500,5500,5000,
                    3000,3500,
                    4000,2500,3500,3750) #max genes per cell (doublet)
        mito_cutoff = c(10,10,10,
                        10,10,10,
                        15,15,
                        10,10,15,17) #max % mitochondrial genes per cell
        for(i in 1:length(direcs)) {
          
          #i=1
          print(i)
          #1. load data from 10X output: 
          matrix_dir = paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/matvei/cell_ranger_output/avicenna/khorms/projects/collaborations/Allan_Bailman/cellranger_outputs/",direcs[i],"/outs/filtered_feature_bc_matrix/",sep="")
          barcode.path <- paste0(matrix_dir, "barcodes.tsv")
          features.path <- paste0(matrix_dir, "features.tsv")
          matrix.path <- paste0(matrix_dir, "matrix.mtx")
          
          mat <- readMM(file = matrix.path) #read in the expression matrix
          
          feature.names = read.delim(features.path,  #read in gene names
                                     header = FALSE,
                                     stringsAsFactors = FALSE)
          names(feature.names)[1:2] = c("ensemble_id","gene_short_name")
          
          barcode.names = read.delim(barcode.path, #read in barcode names
                                     header = FALSE,
                                     stringsAsFactors = FALSE)
          
          #create cell_metadata dataframe
          cell_metadat = data.frame(matrix(samp_names[i],ncol=1,nrow=nrow(barcode.names)))
          names(cell_metadat) = "sample"
          row.names(cell_metadat) = barcode.names[,1]
          
          #create gene_annotation_file
          gene_metadat = feature.names
          row.names(gene_metadat) = gene_metadat$ensemble_id
          
          #create expression_data matrix
          colnames(mat) = row.names(cell_metadat)
          row.names(mat) = row.names(gene_metadat)
          
          cds.balmain1 <- new_cell_data_set(expression_data = mat,
                                            cell_metadata = cell_metadat,
                                            gene_metadata = gene_metadat)
          
          #----------------------------------------------------------------------------------------------#
          #Start Seurat QC --> cds.balmain2
          #----------------------------------------------------------------------------------------------#
          #Quality Control after: https://broadinstitute.github.io/2019_scWorkshop/data-wrangling-scrnaseq.html#filtering-low-quality-cells
          
          counts_per_cell <- Matrix::colSums(mat)
          counts_per_gene <- Matrix::rowSums(mat)
          genes_per_cell <- Matrix::colSums(mat>0) # count gene only if it has non-zero reads mapped.
          cells_per_gene <- Matrix::rowSums(mat>0) # only count cells where the gene is expressed
          
          #plot counts
          par(mfrow=c(2,2))
          hist((counts_per_cell+1),xlab='counts per cell',main=paste(samp_names[i],'counts per cell'),col='wheat')
          tryCatch({  hist((genes_per_cell+1), xlab='genes per cell', main=paste(samp_names[i],'genes per cell'), col='wheat') + abline(v=c(mingene[i],maxgene[i]), col="red",lwd=3)},  error=function(e){})
          plot(counts_per_cell, xlab='counts per cell',ylab='genes per cell',genes_per_cell, log='xy', main=paste(samp_names[i],'counts vs genees per cell'),col='wheat')
          tryCatch({ plot(sort(genes_per_cell), xlab='cell', ylab='genes per cell', main=paste(samp_names[i],'genes per cell (ordered)'))  + abline(h=c(mingene[i],maxgene[i]), col="red",lwd=3) },  error=function(e){})#distribution of library complexity --> libraries with hardly any genes should probably be dropped
          
          #plots for powerpoint
          #tryCatch({  hist((genes_per_cell+1), xlab='genes per cell', main=paste(samp_names[i],'genes per cell'), col='wheat')},  error=function(e){})
          #tryCatch({ plot(sort(genes_per_cell), xlab='cell', ylab='genes per cell', main=paste(samp_names[i],'genes per cell (ordered)'))},  error=function(e){})#distribution of library complexity --> libraries with hardly any genes should probably be dropped
          #tryCatch({  hist((genes_per_cell+1), xlab='genes per cell', main=paste(samp_names[i],'genes per cell'), col='wheat') + abline(v=c(mingene[i],maxgene[i]), col="red",lwd=3)},  error=function(e){})
          #tryCatch({ plot(sort(genes_per_cell), xlab='cell', ylab='genes per cell', main=paste(samp_names[i],'genes per cell (ordered)'))  + abline(h=c(mingene[i],maxgene[i]), col="red",lwd=3) },  error=function(e){})#distribution of library complexity --> libraries with hardly any genes should probably be dropped
          
          
          par(mfrow=c(1,1))
          tryCatch({ plot(sort(genes_per_cell), xlab='cell', ylab='genes per cell', main=paste(samp_names[i],'genes per cell (ordered)')) + abline(h=c(mingene[i],maxgene[i]), col="red",lwd=3) },  error=function(e){}) #distribution of library complexity --> libraries with hardly any genes should probably be dropped
          
          #create Seurat object
          #make row names into gene symbol, not ENSEMBL
          row.names(mat) = feature.names$gene_short_name
          seurat<-CreateSeuratObject(counts = mat, min.cells = 3, min.features = 200, project = "CA_US")
          
          #QC: add mitochondrial percent
          seurat[["percent.mt"]] <- PercentageFeatureSet(seurat, pattern = "mt-")
          violplot = VlnPlot(seurat, features = c("percent.mt"), ncol = 1) + geom_violin(fill="wheat") + theme(legend.position = "none") + ggtitle(paste(samp_names[i],"% mitochondrial")) + xlab("") + geom_hline(yintercept=mito_cutoff[i], color="red",size=1.25) + theme(axis.text.x = element_text(size=0))
          
          #Getting rid of low-quality cells/empty droplets (<200 genes) ; doublets (>6k genes) ; stresst/dying cells (<10% mito)
          seurat <- subset(seurat, subset = nFeature_RNA > mingene[i] & nFeature_RNA < maxgene[i] & percent.mt < 10)
          
          #Seurat object --> CDS
          exp_mat <- seurat@assays[["RNA"]]@counts #pull out RAW COUNTS counts from Seurat object ######THIS IS FOR INFER CNV
          cell_metadat <- seurat@meta.data #pull out cell meta-data from Seurat object
          cell_metadat$tissue = strsplit(samp_names[i]," ")[[1]][1] #put column of tissue in the cell metadata
          cell_metadat$sorting = strsplit(samp_names[i]," ")[[1]][2] #put column of tissue in the cell metadata
          cell_metadat$sample_nom = samp_names[i]
          cell_metadat$tx_name = tx_names[i]
          cell_metadat$batch_name = as.factor(batch_names[i])
          gene_annot = data.frame(seurat@assays[["RNA"]]@data@Dimnames[[1]])#pull out gene names from Seurat object
          names(gene_annot) = "gene_short_name"
          row.names(gene_annot) = gene_annot$gene_short_name #row.names of gene_metadata must be equal to row.names of expression_data
          
          #Make the CDS object (cds.balmain2) with only Seurat-processed var features
          #exp_mat.varfeat = exp_mat[match(VariableFeatures(seruat),rownames(exp_mat)),] #get only variable features
          #gene_annot.varfeat = data.frame(gene_annot[match(VariableFeatures(seruat),gene_annot$gene_short_name),])
          #names(gene_annot.varfeat) = "gene_short_name"
          #row.names(gene_annot.varfeat) = gene_annot.varfeat$gene_short_name #row.names of gene_metadata must be equal to row.names of expression_data
          
          #assign each cds object to a new name
          name <- paste("cds.", i, sep = "")
          assign(name, new_cell_data_set(exp_mat,
                                         cell_metadata = cell_metadat,
                                         gene_metadata = gene_annot)    #create a bunch of cds objects that we are going to combine
          )
          
        } 
        
        ###################################################################################################################################################
        ###################################################################################################################################################
        ###################################################################################################################################################
        #Infer CNV
        
        #create 1 giant ass cds object from all the cds's created above
        cnv_cds <- combine_cds(list(cds.10,cds.11,cds.9,cds.12,cds.8,cds.7,cds.6,cds.5,cds.4,cds.3,cds.2,cds.1))
        rm(cds.10,cds.11,cds.9,cds.12,cds.8,cds.7,cds.6,cds.5,cds.4,cds.3,cds.2,cds.1,cds.balmain1)
        gc()
        
        #infercnv in each sample individually: only Carcinoma and Papilloma
        samples_to_infercnv = c("Pap Unsorted","CA Unsorted", "Pap Progeny","Pap Lgr6","CA Progeny","CA Lgr6")
        baseline_samples = c("UnTxSkin Unsorted","UnTxSkin Unsorted","UnTxSkin Progeny","UnTxSkin Lgr6","UnTxSkin Progeny","UnTxSkin Lgr6")
        
        mgi_names = read.delim(file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/MGI_genome_names/MRK_List1.txt",header=T) #read in mouse references genes
        
        
        for(i in 1:length(samples_to_infercnv)) {
          
          #i=1
          sub_cnv_cds = cnv_cds[,colData(cnv_cds)$sample_nom %in% c(samples_to_infercnv[i],baseline_samples[i])]
          
          #make Input files directory
          dir.create(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/inferCNV/inferCNV_input_files/", samples_to_infercnv[i],sep=""))
          
          #make Sample annotation file
          sample_annotation = data.frame(row.names(colData(sub_cnv_cds)),colData(sub_cnv_cds)$sample_nom)
          sample_annotation_file_name = paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/inferCNV/inferCNV_input_files/", samples_to_infercnv[i],"/sample_annotation.txt",sep="")
          write.table(sample_annotation,file=sample_annotation_file_name,quote=F,row.names=F,col.names=F,sep="\t")
          
          #make Gene ordering file: https://github.com/broadinstitute/inferCNV/wiki/File-Definitions
          gene_matches = match(row.names(sub_cnv_cds),mgi_names$Marker.Symbol) #get indices where gene names in cds match reference names
          mgi_names$Chr_updated = paste("chr",mgi_names$Chr,sep="") #make a new column of chr
          mgi_names_sub = mgi_names[gene_matches,] #subset to the genes in my data
          gene_order_file = data.frame(mgi_names_sub$Marker.Symbol, mgi_names_sub$Chr_updated, mgi_names_sub$genome.coordinate.start , mgi_names_sub$genome.coordinate.end) #subset the columns you want
          gene_order_file = na.omit(gene_order_file) #get rid of NA's
          gene_order_file_name = paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/inferCNV/inferCNV_input_files/", samples_to_infercnv[i],"/gene_order.txt",sep="")
          write.table(gene_order_file,file=gene_order_file_name,quote=F,row.names=F,col.names=F,sep="\t")
          
          #make sparse raw count matrix
          count_matrix =sub_cnv_cds@assays@data$counts
          
          # create the infercnv object
          infercnv_obj = CreateInfercnvObject(raw_counts_matrix=count_matrix,
                                              annotations_file=sample_annotation_file_name,
                                              delim="\t",
                                              gene_order_file=gene_order_file_name,
                                              ref_group_names=c(baseline_samples[i]))
          
          #Make Results Directory
          dir.create(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/inferCNV/", samples_to_infercnv[i],sep=""))
          outputdir_name = paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/inferCNV/", samples_to_infercnv[i],sep="")
          
          # perform infercnv operations to reveal cnv signal
          infercnv_obj = infercnv::run(infercnv_obj,
                                       cutoff=0.1,  # use 1 for smart-seq, 0.1 for 10x-genomics
                                       out_dir=outputdir_name,  # dir is auto-created for storing outputs
                                       cluster_by_groups=T,   # cluster
                                       denoise=T,
                                       HMM=T
          )
          
        }
        
        #-----------------------------------------------------------------------------------------------------#
        #-----------------------------------------------------------------------------------------------------#
        #-----------------------------------------------------------------------------------------------------#
        #-----------------------------------------------------------------------------------------------------#
        #BINARY METHOD Integrate inferCNVR Results into manifold: Loop across all samples THRESHOLD CUTOFF METHOD binary calls duplicated vs non-duplicated
        #-----------------------------------------------------------------------------------------------------#
        #-----------------------------------------------------------------------------------------------------#
        #-----------------------------------------------------------------------------------------------------#
        #-----------------------------------------------------------------------------------------------------#
        
        all_sample_cnvs = do.call(rbind, lapply(1:length(samples_to_infercnv), function(j) {
          print(j)
          #j=1
          denoisedata = read.table(file= paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/inferCNV/", samples_to_infercnv[j],"/infercnv.21_denoised.observations.txt",sep=""),header=T)
          mgi_names.sub = mgi_names[match(row.names(denoisedata),mgi_names$Marker.Symbol),] #get gene's chr position
          denoisedata$chr = mgi_names.sub$Chr #add column of chr
          
          cutoff_file = read.table(file= paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/inferCNV/", samples_to_infercnv[j],"/infercnv.21_denoised.heatmap_thresholds.txt",sep=""),header=F)#read in data of cut off values from DENOISED heat map
          threshold = cutoff_file[,1][cutoff_file[,1]>1][1] #get the first cutoff >1. This is the threshold on the heatmap to denote duplication (higher than expected expression)
          percent_threshold_for_duplication = 0.33 #this means that if more than this percent of genes on a chr for a cell are above threshold, then it is called a duplication
          
          allchrs = unique(denoisedata$chr)
          
          #---------------------------------------------------------#
          #Aneuploid DUPLICATIONS scan: look within each chromsome
          chromsome_wise_means = do.call(cbind, lapply(1:length(allchrs), function(i){
            #i=7
            print(i)
            denoise_sub = subset(denoisedata, chr==allchrs[i])
            denoise_sub = denoise_sub[,-ncol(denoise_sub)]#drop last column which is the chromosome factor column
            # denoise_sub[1:10,1:10]
            
            dup_status_df = do.call(rbind, lapply(1:ncol(denoise_sub), function(k){ #loop to ask if each cell has more than percent_threshold_for_duplication of genes above the threshold
              #k=1
              percent_exceeding = length((denoise_sub[,k][denoise_sub[,k] > threshold])) / length(denoise_sub[,k]) #calculate percent of genes that exceed expression threshold
              if(percent_exceeding > percent_threshold_for_duplication) {duplication_status = "Duplicated"} else {duplication_status = "Non-Duplicated"}
              status_res_df = data.frame(duplication_status)
              row.names(status_res_df) = names(denoise_sub)[k]
              status_res_df
            }))
            names(dup_status_df) = paste("chr",allchrs[i],sep="")
            dup_status_df
          }))
          
          chromsome_wise_means = data.frame(samples_to_infercnv[j], gsub("\\.","-",row.names(chromsome_wise_means) ),chromsome_wise_means)
          names(chromsome_wise_means)[1:2] = c("sample","cell_id")
          row.names(chromsome_wise_means) = gsub("\\.","-",row.names(chromsome_wise_means) ) #rename rows corrects
          
          #---------------------------------------------------------#
          #WGD: Whole genome duplciation: look across all chromsomes
          upper_threshold = cutoff_file[,1][cutoff_file[,1]>1][1] #get threshold of higher than expected exp
          lessthanone = cutoff_file[,1][cutoff_file[,1]<1]
          lower_threshold = lessthanone[length(lessthanone)] #get threshold of lower than expected exp
          percent_threshold_for_duplication = 0.16 #this means that if more than this percent of genes on a chr for a cell are above threshold, then it is called a duplication
          
          denoise_sub = denoisedata[,-ncol(denoisedata)]#drop last column which is the chromosome factor column
          dup_status_df = do.call(rbind, lapply(1:ncol(denoise_sub), function(k){ #loop to ask if each cell has more than percent_threshold_for_duplication of genes above the threshold
            #k=1
            percent_exceeding = length(denoise_sub[,k][((denoise_sub[,k] > upper_threshold) | (denoise_sub[,k] < lower_threshold))]) / length(denoise_sub[,k]) #calculate percent of genes that exceed expression threshold
            if(percent_exceeding > percent_threshold_for_duplication) {duplication_status = "Duplicated"} else {duplication_status = "Non-Duplicated"}
            status_res_df = data.frame(duplication_status)
            row.names(status_res_df) = names(denoise_sub)[k]
            status_res_df
          }))
          names(dup_status_df) = "WGD_status"
          
          #---------------------------------------------------------#
          #Add WGD column to chr-wise aneuploid scan
          chromsome_wise_means$WGD = dup_status_df$WGD_status
          chromsome_wise_means
        }))
        
        
        table(all_sample_cnvs$sample)
        table(colData(big_cds)$tissue_updated_sorting)
        table(subset(all_sample_cnvs, sample=="Pap Unsorted")$WGD)
        
        #add this to the manifold and plot out
        #add chr and WGD this to the big_cds object
        big_cds = readRDS("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/matvei/cds_objects/integrated_corrected_cds_v2.RDS")
        
        metadat_df = data.frame(colData(big_cds))
        metadat_df$cell_id = row.names(metadat_df) #make cell id column to merge by
        all_sample_cnvs$cell_id = row.names(all_sample_cnvs) #put cell id column to merge by
        metadat_df_updated = full_join(metadat_df, all_sample_cnvs, by="cell_id") #THIS MERGING WORKS BC IT PRESERVES THE CORRECT ORDER
        metadat_df_updated[is.na(metadat_df_updated)] <- "Non-Duplicated" #replae NAs with Non-Duplicated for the Normal skin samples
        
        head(metadat_df)
        head(colData(big_cds))
        head(metadat_df_updated)
        
        colData(big_cds)$chr1 = metadat_df_updated$chr1
        colData(big_cds)$chr2 = metadat_df_updated$chr2
        colData(big_cds)$chr3 = metadat_df_updated$chr3
        colData(big_cds)$chr4 = metadat_df_updated$chr4
        colData(big_cds)$chr5 = metadat_df_updated$chr5
        colData(big_cds)$chr6 = metadat_df_updated$chr6
        colData(big_cds)$chr7 = metadat_df_updated$chr7
        colData(big_cds)$chr8 = metadat_df_updated$chr8
        colData(big_cds)$chr9 = metadat_df_updated$chr9
        colData(big_cds)$chr10 = metadat_df_updated$chr10
        colData(big_cds)$chr11 = metadat_df_updated$chr11
        colData(big_cds)$chr12 = metadat_df_updated$chr12
        colData(big_cds)$chr13 = metadat_df_updated$chr13
        colData(big_cds)$chr14 = metadat_df_updated$chr14
        colData(big_cds)$chr15 = metadat_df_updated$chr15
        colData(big_cds)$chr16 = metadat_df_updated$chr16
        colData(big_cds)$chr17 = metadat_df_updated$chr17
        colData(big_cds)$chr18 = metadat_df_updated$chr18
        colData(big_cds)$chr19 = metadat_df_updated$chr19
        colData(big_cds)$chrMT = metadat_df_updated$chrMT
        colData(big_cds)$WGD = metadat_df_updated$WGD
        
        #Plot WGD
        plot_cells(big_cds,
                   color_cells_by="WGD",
                   label_cell_groups=FALSE,
                   label_leaves=F,
                   label_branch_points=F,
                   graph_label_size=1.5,
                   show_trajectory_graph=F,
                   cell_size=0.6,
                   alpha=0.2) +
          scale_color_manual(values = c("Duplicated" = "orangered4", "Non-Duplicated" = "lightgrey"), name="Aneuploid status") +
          ggtitle("WGD")+ #continuous
          xlab("") + ylab("") + 
          theme(plot.title = element_text(size=20),
                axis.text=element_text(size=0,color='black'),
                axis.title=element_text(size=0),
                axis.ticks = element_line(size=0),
                legend.position="none")
        
        #NOW loop through all of these beauties and print out manifold
        #Fig. 4
        
        chrs = seq(1:19)
        for(i in 1:length(chrs)){
          #i=7
          
          print(paste("chr",chrs[i],sep=""))
          chr_replication_big_cds = plot_cells(big_cds,
                                               color_cells_by=paste("chr",chrs[i],sep=""),
                                               label_cell_groups=FALSE,
                                               label_leaves=F,
                                               label_branch_points=F,
                                               graph_label_size=1.5,
                                               show_trajectory_graph=F,
                                               cell_size=0.6,
                                               alpha=0.2) +
            scale_color_manual(values = c("Duplicated" = "olivedrab4", "Non-Duplicated" = "lightgrey"), name="Aneuploid status") +
            ggtitle(paste("Chr ",chrs[i],sep=""))+ #continuous
            xlab("") + ylab("") + 
            theme(plot.title = element_text(size=12),
                  axis.text=element_text(size=0,color='black'),
                  axis.title=element_text(size=0),
                  axis.ticks = element_line(size=0),
                  legend.position="none",
                  axis.line = element_line(colour = "white"),
                  strip.background = element_blank())
          
          chr_replication_pap_ca = plot_cells(pap_ca,
                                              color_cells_by=paste("chr",chrs[i],sep=""),
                                              label_cell_groups=FALSE,
                                              label_leaves=F,
                                              label_branch_points=F,
                                              graph_label_size=1.5,
                                              show_trajectory_graph=F,
                                              cell_size=1,
                                              alpha=0.6) +
            scale_color_manual(values = c("Duplicated" = "olivedrab4", "Non-Duplicated" = "lightgrey"), name="Aneuploid status") +
            ggtitle(paste("Chr ",chrs[i],sep=""))+ #continuous
            xlab("") + ylab("") + 
            theme(plot.title = element_text(size=12),
                  axis.text=element_text(size=0,color='black'),
                  axis.title=element_text(size=0),
                  axis.ticks = element_line(size=0),
                  legend.position="none")
          
          png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Figs/Fig4/CNV_manifolds/chr",chrs[i],"_full.png",sep=""),
              width=8,height=8,units="in",res=500)
          print(chr_replication_big_cds)
          dev.off()
          
          png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Figs/Fig4/CNV_manifolds/chr",chrs[i],"_papca.png",sep=""),
              width=8,height=8,units="in",res=300)
          print(chr_replication_pap_ca)
          dev.off()
          
        } 
        
        #------------------------------------------------------------------------------------------------------#
        #------------------------------------------------------------------------------------------------------#
        #------------------------------------------------------------------------------------------------------#
        #------------------------------------------------------------------------------------------------------#
        #Fig. 4
        #make venn diagram of multiploids
        #papilloma
        
        meta = data.frame(colData(big_cds))
        
        x = list("Chr1" = row.names(meta[((meta$tissue %in% "Pap") & (meta$chr1 %in% "Duplicated")),]),
                 "Chr6" = row.names(meta[((meta$tissue %in% "Pap") & (meta$chr6 %in% "Duplicated")),]),
                 "Chr7" = row.names(meta[((meta$tissue %in% "Pap") & (meta$chr7 %in% "Duplicated")),]),
                 "Chr10" = row.names(meta[((meta$tissue %in% "Pap") & (meta$chr10 %in% "Duplicated")),])
        )
        
        pap_aneuploid_venn_fig = ggVennDiagram(x, color="black",label="count") +
          scale_fill_gradient(
            low = "white",
            high = "purple",
            guide = "colourbar",
            aesthetics = "fill") +
          theme(legend.position = "none")
        
        #CARCINOMA
        x = list("Chr1" = row.names(meta[((meta$tissue %in% "CA") & (meta$chr1 %in% "Duplicated")),]),
                 "Chr6" = row.names(meta[((meta$tissue %in% "CA") & (meta$chr6 %in% "Duplicated")),]),
                 "Chr7" = row.names(meta[((meta$tissue %in% "CA") & (meta$chr7 %in% "Duplicated")),]),
                 "Chr10" = row.names(meta[((meta$tissue %in% "CA") & (meta$chr10 %in% "Duplicated")),])
        )
        
        ca_aneuploid_venn_fig = ggVennDiagram(x, color="black",label="count") +
          scale_fill_gradient(
            low = "white",
            high = "darkred",
            guide = "colourbar",
            aesthetics = "fill") +
          theme(legend.position = "none")
        
        png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Figs/Fig4/duplication_venn_diagram.png",sep=""),
            width=10,height=5,units="in",res=300)
        print(multiplot( pap_aneuploid_venn_fig, ca_aneuploid_venn_fig, cols=2))
        dev.off()
        
        #------------------------------------------------------------------------------------------------------#
        #------------------------------------------------------------------------------------------------------#
        #------------------------------------------------------------------------------------------------------#
        #------------------------------------------------------------------------------------------------------#
        
        
        
        all_dups = do.call(rbind, lapply(1:length(samples_to_infercnv), function(i){
          #i=1
          sub_cnv = subset(all_sample_cnvs, sample==samples_to_infercnv[i])
          duplication_summary = do.call(rbind, lapply(c(3,8,9,10), function(z){
            #z=3
            dup_summary = data.frame(samples_to_infercnv[i],names(sub_cnv)[z],table(sub_cnv[,z]),(table(sub_cnv[,z])/nrow(sub_cnv))*100)
            dup_summary
          }))
          duplication_summary = duplication_summary[,-4] #drop redundant variable field
          names(duplication_summary) = c("sample","chr","duplication_status","N_cells","Percent_cells")
          duplication_summary
        }))
        
        duplicated_only = subset(all_dups, duplication_status=="Duplicated")
        
        #Plot the duplicated chromsomes
        ggplot(duplicated_only, aes(x=sample, y=chr)) +
          geom_point(aes(size=Percent_cells)) +
          ggtitle("Single-cell chromosomal duplications") +
          xlab("sample") +
          ylab("chromosome") +
          scale_size(range = c(1,35)) +
          theme(axis.text=element_text(size=15,color='black'),
                axis.title=element_text(size=20,color='black')) +
          theme(plot.title = element_text(size = 25, vjust=1)) +
          theme(panel.grid.major=element_line(colour = "darkgrey"), 
                panel.grid.minor=element_blank(), 
                panel.background=element_blank()) +
          theme(legend.key = element_rect(fill = "white")) +
          theme(legend.text=element_text(size=12), legend.title=element_text(size=12)) +
          theme(legend.position="bottom") +
          scale_x_discrete( 
            limits=c("Pap Unsorted","Pap Lgr6","Pap Progeny",
                     "CA Unsorted","CA Lgr6","CA Progeny"))
        
        
        #-----------------------------------------------------------------------------------------------------#
        #-----------------------------------------------------------------------------------------------------#
        #-----------------------------------------------------------------------------------------------------#
        #Ask does CNV correlate to eigengene expression differences?
        
        #Loop through eigengenes and test CNVs: chr 1,6,7,10 (all possible combinations)
        
        eigengene_agg_cnv = do.call(rbind, lapply(1:ncol(balmain_genes), function(i){
          
          #subset eigengene_agg = do.call(cbind, lapply(172:175, function(i){
          
          tryCatch({ #suppress error
            
            #i=268
            print(i)
            #only if more than 1 constituent gene
            marker_df = data.frame(na.omit(balmain_genes[,i]),names(balmain_genes)[i])
            names(marker_df) = c("gene","cell_type")
            
            #remake the marker df dataframe with the matched gene names
            marker_df = data.frame(unique(alias2SymbolTable(unique(marker_df$gene), species = "Mm")),names(balmain_genes)[i],names(balmain_genes)[i])
            names(marker_df) = c("gene","cell_type","module")
            marker_df = na.omit(marker_df) #remove non-matching genes
            
            #aggregate gene values to get a single expression value
            agg_mat = aggregate_gene_expression(
              pap_ca,
              gene_group_df = marker_df,
              cell_group_df = NULL,
              norm_method = c("log"),
              pseudocount = 1,
              scale_agg_values = F,
              max_agg_value = NA,
              min_agg_value = NA,
              exclude.na = TRUE
            )
            
            #combine aggregated values with CNV calls
            agg_gene_CNV = data.frame(c(agg_mat),c(colData(pap_ca)$chr1), c(colData(pap_ca)$chr6),
                                      c(colData(pap_ca)$chr7), c(colData(pap_ca)$chr10))
            names(agg_gene_CNV) = c("eigengene_expression", "chr1","chr6","chr7","chr10")
            glmnb_res <- glm.nb(eigengene_expression ~ chr1*chr6*chr7*chr10, data = agg_gene_CNV)
            
            #get values of duplicated vs non-duplicated changes
            chr1_non_duplicated = median(subset(agg_gene_CNV, chr1=="Non-Duplicated")$eigengene_expression)
            chr1_duplicated = median(subset(agg_gene_CNV, chr1=="Duplicated")$eigengene_expression)
            chr1_change = chr1_non_duplicated - chr1_duplicated
            chr1_percent_change =  round(((chr1_duplicated - chr1_non_duplicated) / chr1_non_duplicated)*100,2)
            
            chr6_non_duplicated = median(subset(agg_gene_CNV, chr6=="Non-Duplicated")$eigengene_expression)
            chr6_duplicated = median(subset(agg_gene_CNV, chr6=="Duplicated")$eigengene_expression)
            chr6_change = chr6_non_duplicated - chr6_duplicated
            chr6_percent_change =  round(((chr6_duplicated - chr6_non_duplicated) / chr6_non_duplicated)*100,2)
            
            chr7_non_duplicated = median(subset(agg_gene_CNV, chr7=="Non-Duplicated")$eigengene_expression)
            chr7_duplicated = median(subset(agg_gene_CNV, chr7=="Duplicated")$eigengene_expression)
            chr7_change = chr7_non_duplicated - chr7_duplicated
            chr7_percent_change =  round(((chr7_duplicated - chr7_non_duplicated) / chr7_non_duplicated)*100,2)
            
            chr10_non_duplicated = median(subset(agg_gene_CNV, chr10=="Non-Duplicated")$eigengene_expression)
            chr10_duplicated = median(subset(agg_gene_CNV, chr10=="Duplicated")$eigengene_expression)
            chr10_change = chr10_non_duplicated - chr10_duplicated
            chr10_percent_change =  round(((chr10_duplicated - chr10_non_duplicated) / chr10_non_duplicated)*100,2)
            
            combined_results = data.frame(names(balmain_genes)[i],
                                          chr1_non_duplicated,chr1_duplicated,chr1_change,chr1_percent_change,
                                          chr6_non_duplicated,chr6_duplicated,chr6_change,chr6_percent_change,
                                          chr7_non_duplicated,chr7_duplicated,chr7_change,chr7_percent_change,
                                          chr10_non_duplicated,chr10_duplicated,chr10_change,chr10_percent_change,
                                          row.names(summary(glmnb_res)$coefficients[-1,]),
                                          summary(glmnb_res)$coefficients[-1,])
            data.frame(names(combined_results))
            names(combined_results)[c(1,18:22)] = c("eigengene","parameter","Est","SE","z","p")
            
            return(combined_results)
            
          },  error=function(e){})
          
        }))
        
        #Write out all CNV~eigengene exprssion results to disk
        write.table(eigengene_agg_cnv, 
                    file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/DEG/eigengene_CNV/eigengene_agg_cnv.txt",
                    quote=F,row.names=F,sep="\t")
        
        #significnat only
        bonferonni_threshold = 0.05 / 310
        eigengene_agg_cnv_sig = subset(eigengene_agg_cnv, p<bonferonni_threshold)
        nrow(eigengene_agg_cnv)
        nrow(eigengene_agg_cnv_sig)
        
        #Write out SIGNIFICANT-ONLY CNV~eigengene exprssion results to disk
        write.table(eigengene_agg_cnv_sig, 
                    file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/DEG/eigengene_CNV/eigengene_agg_cnv_SIGNIFICANT_ONLY.txt",
                    quote=F,row.names=F,sep="\t")
        
        
        #-----------------------------------------------------------------------------------------------------#
        #Process these results in a more intelligeable way
        eigengene_cnv_res = read.delim(file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/DEG/eigengene_CNV/eigengene_agg_cnv.txt", header=T)
        head(eigengene_cnv_res)
        eigengenes_included = unique(eigengene_cnv_res$eigengene)
        
        single_term_results = do.call(rbind, lapply(1:length(eigengenes_included), function(i){
          #i=1
          sub_res = subset(eigengene_cnv_res, eigengene==eigengenes_included[i])
          data.frame(
            eigengenes_included[i],
            sub_res[c(1:4),c(18:22)],
            rbind(sub_res$chr1_percent_change[1],sub_res$chr6_percent_change[1],
                  sub_res$chr7_percent_change[1],sub_res$chr10_percent_change[1])
          )
        }))
        names(single_term_results)[c(1,7)] = c("eigengene","percent_change")
        single_term_results$parameter = substr(single_term_results$parameter,1,5) #extract the chromsome
        single_term_results$parameter = gsub("N","",single_term_results$parameter) #get rid of extra N
        
        #Write out single-term results
        write.table(single_term_results, 
                    file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/DEG/eigengene_CNV/eigengene_agg_cnv_SINGLE_TERM.txt",
                    quote=F,row.names=F,sep="\t")
        
        
        #-----------------------------------------------------------------------------------------------------#
        #-----------------------------------------------------------------------------------------------------#
        #-----------------------------------------------------------------------------------------------------#
        #-----------------------------------------------------------------------------------------------------#
        #-----------------------------------------------------------------------------------------------------#
        #-----------------------------------------------------------------------------------------------------#
        #-----------------------------------------------------------------------------------------------------#
        #-----------------------------------------------------------------------------------------------------#
        
        #Get multi-aneuploids: 6+7
        colData(pap_unsorted_cds)$chr6_7 = paste(colData(pap_unsorted_cds)$chr6,colData(pap_unsorted_cds)$chr7,sep="_") #combine the chr duplication calls
        colData(pap_unsorted_cds)$chr6_7 = gsub("\\bDuplicated_Duplicated\\b","Chr6-7",colData(pap_unsorted_cds)$chr6_7,perl=TRUE) #combine the chr duplication calls
        colData(pap_unsorted_cds)$chr6_7 = gsub("\\bNon-Chr6-7\\b","Chr7",colData(pap_unsorted_cds)$chr6_7,perl=TRUE) #combine the chr duplication calls
        colData(pap_unsorted_cds)$chr6_7 = gsub("\\bDuplicated_Non-Duplicated\\b","Chr6",colData(pap_unsorted_cds)$chr6_7,perl=TRUE) #combine the chr duplication calls
        colData(pap_unsorted_cds)$chr6_7 = gsub("\\bNon-Chr6\\b","None",colData(pap_unsorted_cds)$chr6_7,perl=TRUE) #combine the chr duplication calls
        
        unique(colData(pap_unsorted_cds)$chr6_7)
        
        pap_uns_67 = plot_cells(pap_unsorted_cds,
                                color_cells_by="chr6_7",
                                label_cell_groups=FALSE,
                                label_leaves=F,
                                label_branch_points=F,
                                graph_label_size=1.5,
                                show_trajectory_graph=F,
                                cell_size=0.6,
                                alpha=0.4) +
          xlab("Dim. 1") + ylab("Dim. 2") + 
          theme(legend.position="bottom") +
          scale_color_manual(values = c("None" = "lightgrey", "Chr6" = "darkblue","Chr7"="darkolivegreen4","Chr6-7"="darkred"), name="Aneuploid status") +
          ggtitle(paste("Aneuploid 6+7: Papilloma"))+ #continuous
          theme(axis.text=element_text(size=8,color='black'),
                axis.title=element_text(size=10,color='black')) + theme(plot.title = element_text(size = 14, vjust=1)) +
          theme(strip.text.x = element_text(size = 0)) #remove facet label
        
        #-----------------------------------------------------------------------------------------------------#
        #Get multi-aneuploids: 7+10
        colData(pap_unsorted_cds)$chr7_10 = paste(colData(pap_unsorted_cds)$chr7,colData(pap_unsorted_cds)$chr10,sep="_") #combine the chr duplication calls
        colData(pap_unsorted_cds)$chr7_10 = gsub("\\bDuplicated_Duplicated\\b","Chr7-10",colData(pap_unsorted_cds)$chr7_10,perl=TRUE) #combine the chr duplication calls
        colData(pap_unsorted_cds)$chr7_10 = gsub("\\bNon-Chr7-10\\b","Chr10",colData(pap_unsorted_cds)$chr7_10,perl=TRUE) #combine the chr duplication calls
        colData(pap_unsorted_cds)$chr7_10 = gsub("\\bDuplicated_Non-Duplicated\\b","Chr7",colData(pap_unsorted_cds)$chr7_10,perl=TRUE) #combine the chr duplication calls
        colData(pap_unsorted_cds)$chr7_10 = gsub("\\bNon-Chr7\\b","None",colData(pap_unsorted_cds)$chr7_10,perl=TRUE) #combine the chr duplication calls
        
        unique(colData(pap_unsorted_cds)$chr7_10)
        
        pap_uns_710 = plot_cells(pap_unsorted_cds,
                                 color_cells_by="chr7_10",
                                 label_cell_groups=FALSE,
                                 label_leaves=F,
                                 label_branch_points=F,
                                 graph_label_size=1.5,
                                 show_trajectory_graph=F,
                                 cell_size=0.6,
                                 alpha=0.4) +
          xlab("Dim. 1") + ylab("Dim. 2") + 
          theme(legend.position="bottom") +
          scale_color_manual(values = c("None" = "lightgrey", "Chr10" = "darkblue","Chr7"="darkolivegreen4","Chr7-10"="darkred"), name="Aneuploid status") +
          ggtitle(paste("Aneuploid 7+10: Papilloma"))+ #continuous
          theme(axis.text=element_text(size=8,color='black'),
                axis.title=element_text(size=10,color='black')) + theme(plot.title = element_text(size = 14, vjust=1)) +
          theme(strip.text.x = element_text(size = 0)) #remove facet label
        
        #-----------------------------------------------------------------------------------------------------#
        #Get multi-aneuploids: 7+13
        colData(pap_unsorted_cds)$chr7_13 = paste(colData(pap_unsorted_cds)$chr7,colData(pap_unsorted_cds)$chr13,sep="_") #combine the chr duplication calls
        colData(pap_unsorted_cds)$chr7_13 = gsub("\\bDuplicated_Duplicated\\b","Chr7-13",colData(pap_unsorted_cds)$chr7_13,perl=TRUE) #combine the chr duplication calls
        colData(pap_unsorted_cds)$chr7_13 = gsub("\\bNon-Chr7-13\\b","Chr13",colData(pap_unsorted_cds)$chr7_13,perl=TRUE) #combine the chr duplication calls
        colData(pap_unsorted_cds)$chr7_13 = gsub("\\bDuplicated_Non-Duplicated\\b","Chr7",colData(pap_unsorted_cds)$chr7_13,perl=TRUE) #combine the chr duplication calls
        colData(pap_unsorted_cds)$chr7_13 = gsub("\\bNon-Chr7\\b","None",colData(pap_unsorted_cds)$chr7_13,perl=TRUE) #combine the chr duplication calls
        
        unique(colData(pap_unsorted_cds)$chr7_13)
        
        pap_uns_713 = plot_cells(pap_unsorted_cds,
                                 color_cells_by="chr7_13",
                                 label_cell_groups=FALSE,
                                 label_leaves=F,
                                 label_branch_points=F,
                                 graph_label_size=1.5,
                                 show_trajectory_graph=F,
                                 cell_size=0.6,
                                 alpha=0.4) +
          xlab("Dim. 1") + ylab("Dim. 2") + 
          theme(legend.position="bottom") +
          scale_color_manual(values = c("None" = "lightgrey", "Chr13" = "darkblue","Chr7"="darkolivegreen4","Chr7-13"="darkred"), name="Aneuploid status") +
          ggtitle(paste("Aneuploid 7+13: Papilloma"))+ #continuous
          theme(axis.text=element_text(size=8,color='black'),
                axis.title=element_text(size=10,color='black')) + theme(plot.title = element_text(size = 14, vjust=1)) +
          theme(strip.text.x = element_text(size = 0)) #remove facet label
        
        #-----------------------------------------------------------------------------------------------------#
        #Get multi-aneuploids: 6+7+10 = only need to look at 6+10 overlaps because all of these cells also include 7
        colData(pap_unsorted_cds)$chr6_7_10 = paste(colData(pap_unsorted_cds)$chr6,colData(pap_unsorted_cds)$chr10,sep="_") #combine the chr duplication calls
        colData(pap_unsorted_cds)$chr6_7_10 = gsub("\\bDuplicated_Duplicated\\b","Chr6-7-13",colData(pap_unsorted_cds)$chr6_7_10,perl=TRUE) #combine the chr duplication calls
        colData(pap_unsorted_cds)$chr6_7_10 = gsub("\\bNon-Chr6-7-13\\b","Non-trisomy",colData(pap_unsorted_cds)$chr6_7_10,perl=TRUE) #combine the chr duplication calls
        colData(pap_unsorted_cds)$chr6_7_10 = gsub("\\bDuplicated_Non-Duplicated\\b","Non-trisomy",colData(pap_unsorted_cds)$chr6_7_10,perl=TRUE) #combine the chr duplication calls
        colData(pap_unsorted_cds)$chr6_7_10 = gsub("\\bNon-Non-trisomy\\b","Non-trisomy",colData(pap_unsorted_cds)$chr6_7_10,perl=TRUE) #combine the chr duplication calls
        
        unique(colData(pap_unsorted_cds)$chr6_7_10)
        
        pap_uns_6713 = plot_cells(pap_unsorted_cds,
                                  color_cells_by="chr6_7_10",
                                  label_cell_groups=FALSE,
                                  label_leaves=F,
                                  label_branch_points=F,
                                  graph_label_size=1.5,
                                  show_trajectory_graph=F,
                                  cell_size=0.6,
                                  alpha=0.4) +
          xlab("Dim. 1") + ylab("Dim. 2") + 
          theme(legend.position="bottom") +
          scale_color_manual(values = c("Non-trisomy" = "lightgrey", "Chr6-7-13"="darkred"), name="Aneuploid status") +
          ggtitle(paste("Aneuploid 6+7+13: Papilloma"))+ #continuous
          theme(axis.text=element_text(size=8,color='black'),
                axis.title=element_text(size=10,color='black')) + theme(plot.title = element_text(size = 14, vjust=1)) +
          theme(strip.text.x = element_text(size = 0)) #remove facet label
        
        #print out 4 multiploid plots
        pdf(file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/inferCNV/Manifold_CNV/Unsorted_pap_MULTIPLOID_binary_calls.pdf",height=12,width=12)
        
        print(multiplot(pap_uns_67,pap_uns_710,
                        pap_uns_713, pap_uns_6713, cols=2))
        
        dev.off()
        
        #-----------------------------------------------------------------------------------------------------#
        
        
        #-----------------------------------------------------------------------------------------------------#
        #-----------------------------------------------------------------------------------------------------#
        #-----------------------------------------------------------------------------------------------------#
        #CA Unsorted
        #i=2
        denoisedata = read.table(file= paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/inferCNV/", samples_to_infercnv[i],"/infercnv.21_denoised.observations.txt",sep=""),header=T)
        mgi_names.sub = mgi_names[match(row.names(denoisedata),mgi_names$Marker.Symbol),]
        denoisedata$chr = mgi_names.sub$Chr #add column of chr
        
        cutoff_file = read.table(file= paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/inferCNV/", samples_to_infercnv[i],"/infercnv.heatmap_thresholds.txt",sep=""),header=F)#read in data of cut off values from heat map
        threshold = cutoff_file[11,]#the top 5 cutoffs define where lots of unuusally high expression is; define by looking at heatmap
        percent_threshold_for_duplication = 0.33 #this means that if more than this percent of genes on a chr for a cell are above threshold, then it is called a duplication
        
        allchrs = unique(denoisedata$chr) 
        #loop thru each chromsome and see if it is duplicated in a cell or not based on percentage of genes above expression threshold in a chr
        chromsome_wise_means = do.call(cbind, lapply(1:length(allchrs), function(i){
          #i=7
          print(i)
          denoise_sub = subset(denoisedata, chr==allchrs[i])
          denoise_sub = denoise_sub[,-ncol(denoise_sub)]#drop last column which is the chromosome factor column
          dup_status_df = do.call(rbind, lapply(1:ncol(denoise_sub), function(k){ #loop to ask if each cell has more than half of genes above the threshold
            #k=1
            percent_exceeding = length((denoise_sub[,k][denoise_sub[,k] > threshold])) / length(denoise_sub[,k]) #calculate percent of genes that exceed expression threshold
            if(percent_exceeding > percent_threshold_for_duplication) {duplication_status = "Duplicated"} else {duplication_status = "Non-Duplicated"}
            status_res_df = data.frame(duplication_status)
            row.names(status_res_df) = names(denoise_sub)[k]
            status_res_df
          }))
          names(dup_status_df) = paste("chr",allchrs[i],"_Unsorted_Papilloma_Denoised_binary",sep="")
          dup_status_df
        }))
        
        head(chromsome_wise_means)
        duplication_summary = do.call(rbind, lapply(1:ncol(chromsome_wise_means), function(z){
          #z=1
          dup_summary = data.frame(names(chromsome_wise_means)[z],table(chromsome_wise_means[,z]),(table(chromsome_wise_means[,z])/nrow(chromsome_wise_means))*100)
          dup_summary
        }))
        duplication_summary
        
        #add this to the manifold and plot out
        #add this to the big_cds object
        metadat_df = data.frame(colData(big_cds))
        row.names(chromsome_wise_means) = gsub("\\.","-",row.names(chromsome_wise_means) ) #rename rows corrects
        metadat_df_updated = merge(metadat_df, chromsome_wise_means, by=0, all=TRUE)  #is this true
        
        metadat_df_updated.1 = subset(metadat_df_updated, sample_nom == "CA Unsorted")
        pap_unsorted_cds = big_cds[,colData(big_cds)$sample_nom %in% c("CA Unsorted")]
        
        #chr7
        colData(pap_unsorted_cds)$chr7 = metadat_df_updated.1$chr7_Unsorted_Papilloma_Denoised_binary
        ca_uns_chr7 = plot_cells(pap_unsorted_cds,
                                 color_cells_by="chr7",
                                 label_cell_groups=FALSE,
                                 label_leaves=F,
                                 label_branch_points=F,
                                 graph_label_size=1.5,
                                 show_trajectory_graph=F,
                                 cell_size=0.6,
                                 alpha=0.4) +
          xlab("Dim. 1") + ylab("Dim. 2") + 
          theme(legend.position="bottom") +
          scale_color_manual(values = c("Duplicated" = "darkred", "Non-Duplicated" = "lightgrey"), name="Aneuploid status") +
          ggtitle(paste("Chr 7: Carcinoma"))+ #continuous
          theme(axis.text=element_text(size=8,color='black'),
                axis.title=element_text(size=10,color='black')) + theme(plot.title = element_text(size = 14, vjust=1)) +
          theme(strip.text.x = element_text(size = 0)) #remove facet label
        
        #chr6
        colData(pap_unsorted_cds)$chr6 = metadat_df_updated.1$chr6_Unsorted_Papilloma_Denoised_binary
        ca_uns_chr6 = plot_cells(pap_unsorted_cds,
                                 color_cells_by="chr6",
                                 label_cell_groups=FALSE,
                                 label_leaves=F,
                                 label_branch_points=F,
                                 graph_label_size=1.5,
                                 show_trajectory_graph=F,
                                 cell_size=0.6,
                                 alpha=0.4) +
          xlab("Dim. 1") + ylab("Dim. 2") + 
          theme(legend.position="bottom") +
          scale_color_manual(values = c("Duplicated" = "darkred", "Non-Duplicated" = "lightgrey"), name="Aneuploid status") +
          ggtitle(paste("Chr 6: Carcinoma"))+ #continuous
          theme(axis.text=element_text(size=8,color='black'),
                axis.title=element_text(size=10,color='black')) + theme(plot.title = element_text(size = 14, vjust=1)) +
          theme(strip.text.x = element_text(size = 0)) #remove facet label
        
        #chr10
        colData(pap_unsorted_cds)$chr10 = metadat_df_updated.1$chr10_Unsorted_Papilloma_Denoised_binary
        ca_uns_chr10 = plot_cells(pap_unsorted_cds,
                                  color_cells_by="chr10",
                                  label_cell_groups=FALSE,
                                  label_leaves=F,
                                  label_branch_points=F,
                                  graph_label_size=1.5,
                                  show_trajectory_graph=F,
                                  cell_size=0.6,
                                  alpha=0.4) +
          xlab("Dim. 1") + ylab("Dim. 2") + 
          theme(legend.position="bottom") +
          scale_color_manual(values = c("Duplicated" = "darkred", "Non-Duplicated" = "lightgrey"), name="Aneuploid status") +
          ggtitle(paste("Chr 10: Carcinoma"))+ #continuous
          theme(axis.text=element_text(size=8,color='black'),
                axis.title=element_text(size=10,color='black')) + theme(plot.title = element_text(size = 14, vjust=1)) +
          theme(strip.text.x = element_text(size = 0)) #remove facet label
        
        #chr1
        colData(pap_unsorted_cds)$chr1 = metadat_df_updated.1$chr1_Unsorted_Papilloma_Denoised_binary
        ca_uns_chr1 = plot_cells(pap_unsorted_cds,
                                 color_cells_by="chr1",
                                 label_cell_groups=FALSE,
                                 label_leaves=F,
                                 label_branch_points=F,
                                 graph_label_size=1.5,
                                 show_trajectory_graph=F,
                                 cell_size=0.6,
                                 alpha=0.4) +
          xlab("Dim. 1") + ylab("Dim. 2") + 
          theme(legend.position="bottom") +
          scale_color_manual(values = c("Duplicated" = "darkred", "Non-Duplicated" = "lightgrey"), name="Aneuploid status") +
          ggtitle(paste("Chr 1: Carcinoma"))+ #continuous
          theme(axis.text=element_text(size=8,color='black'),
                axis.title=element_text(size=10,color='black')) + theme(plot.title = element_text(size = 14, vjust=1)) +
          theme(strip.text.x = element_text(size = 0)) #remove facet label
        #-----------------------------------------------------------------------------------------------------#
        
        pdf(file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/inferCNV/Manifold_CNV/Unsorted_pap_ca_aneuploidy_binary_calls.pdf",height=8,width=16)
        
        print(multiplot(pap_uns_7,ca_uns_chr7, cols=2))
        print(multiplot(pap_uns_6,ca_uns_chr6, cols=2))
        print(multiplot(pap_uns_10,ca_uns_chr10, cols=2))
        print(multiplot(pap_uns_13,ca_uns_chr1, cols=2))
        
        
        dev.off()
        
        #-----------------------------------------------------------------------------------------------------#
        #-----------------------------------------------------------------------------------------------------#
        #-----------------------------------------------------------------------------------------------------#
        #Bridge cells for pap_ca
        
        #CA Unsorted
        #i=2
        denoisedata = read.table(file= paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/inferCNV/", samples_to_infercnv[i],"/infercnv.21_denoised.observations.txt",sep=""),header=T)
        mgi_names.sub = mgi_names[match(row.names(denoisedata),mgi_names$Marker.Symbol),]
        denoisedata$chr = mgi_names.sub$Chr #add column of chr
        
        cutoff_file = read.table(file= paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/inferCNV/", samples_to_infercnv[i],"/infercnv.heatmap_thresholds.txt",sep=""),header=F)#read in data of cut off values from heat map
        threshold = cutoff_file[11,]#the top 5 cutoffs define where lots of unuusally high expression is; define by looking at heatmap
        percent_threshold_for_duplication = 0.33 #this means that if more than this percent of genes on a chr for a cell are above threshold, then it is called a duplication
        
        allchrs = unique(denoisedata$chr) 
        #loop thru each chromsome and see if it is duplicated in a cell or not based on percentage of genes above expression threshold in a chr
        ca_unsorted_binary_calls = do.call(cbind, lapply(1:length(allchrs), function(i){
          #i=7
          print(i)
          denoise_sub = subset(denoisedata, chr==allchrs[i])
          denoise_sub = denoise_sub[,-ncol(denoise_sub)]#drop last column which is the chromosome factor column
          dup_status_df = do.call(rbind, lapply(1:ncol(denoise_sub), function(k){ #loop to ask if each cell has more than half of genes above the threshold
            #k=1
            percent_exceeding = length((denoise_sub[,k][denoise_sub[,k] > threshold])) / length(denoise_sub[,k]) #calculate percent of genes that exceed expression threshold
            if(percent_exceeding > percent_threshold_for_duplication) {duplication_status = "Duplicated"} else {duplication_status = "Non-Duplicated"}
            status_res_df = data.frame(duplication_status)
            row.names(status_res_df) = names(denoise_sub)[k]
            status_res_df
          }))
          names(dup_status_df) = paste("chr",allchrs[i],"_Unsorted_Papilloma_Denoised_binary",sep="")
          dup_status_df
        }))
        
        #-----------------------------------------------------------------------------------------------------#
        #Pap Unsorted
        #i=1
        denoisedata = read.table(file= paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/inferCNV/", samples_to_infercnv[i],"/infercnv.21_denoised.observations.txt",sep=""),header=T)
        mgi_names.sub = mgi_names[match(row.names(denoisedata),mgi_names$Marker.Symbol),] #get gene's chr position
        denoisedata$chr = mgi_names.sub$Chr #add column of chr
        
        cutoff_file = read.table(file= paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/inferCNV/", samples_to_infercnv[i],"/infercnv.heatmap_thresholds.txt",sep=""),header=F)#read in data of cut off values from heat map
        threshold = cutoff_file[12,]#the top 5 cutoffs define where lots of unuusally high expression is; define by looking at heatmap
        percent_threshold_for_duplication = 0.33 #this means that if more than this percent of genes on a chr for a cell are above threshold, then it is called a duplication
        
        allchrs = unique(denoisedata$chr) 
        #loop thru each chromsome and see if it is duplicated in a cell or not based on percentage of genes above expression threshold in a chr
        pap_unsorted_binary_calls = do.call(cbind, lapply(1:length(allchrs), function(i){
          #i=7
          print(i)
          denoise_sub = subset(denoisedata, chr==allchrs[i])
          denoise_sub = denoise_sub[,-ncol(denoise_sub)]#drop last column which is the chromosome factor column
          dup_status_df = do.call(rbind, lapply(1:ncol(denoise_sub), function(k){ #loop to ask if each cell has more than half of genes above the threshold
            #k=1
            percent_exceeding = length((denoise_sub[,k][denoise_sub[,k] > threshold])) / length(denoise_sub[,k]) #calculate percent of genes that exceed expression threshold
            if(percent_exceeding > percent_threshold_for_duplication) {duplication_status = "Duplicated"} else {duplication_status = "Non-Duplicated"}
            status_res_df = data.frame(duplication_status)
            row.names(status_res_df) = names(denoise_sub)[k]
            status_res_df
          }))
          names(dup_status_df) = paste("chr",allchrs[i],"_Unsorted_Papilloma_Denoised_binary",sep="")
          dup_status_df
        }))
        
        #-----------------------------------------------------------------------------------------------------#
        #combine pap and carcinoma binary calls
        pap_ca_binary_calls = rbind(pap_unsorted_binary_calls, ca_unsorted_binary_calls)
        
        #plot out pap_ca things
        #something is wrong with the rownames
        pap_ca_unsorted = pap_ca[,colData(pap_ca)$sorting %in% c("Unsorted")] #get only unsorted samples
        
        #somehoe renameing the columns is a bad idea must think around this!
        colnames(pap_ca_unsorted) = substr(colnames(pap_ca_unsorted),1,20) #subtract the extra cell name appendages caused by the merge cds function that craetd the pap_ca cds object
        
        metadat_df = data.frame(colData(pap_ca_unsorted)) #make df
        row.names(pap_ca_binary_calls) = gsub("\\.","-",row.names(pap_ca_binary_calls) ) #rename rows corrects
        metadat_df_updated = merge(metadat_df,pap_ca_binary_calls, by=0, all.x=F) #dont include extra x's for non-matching from parenchyma filter in pap_ca
        
        pap_ca_unsorted_sub = pap_ca_unsorted[,match(metadat_df_updated$Row.names,colnames(pap_ca_unsorted))]
        
        #chr7
        colData(pap_ca_unsorted_sub)$chr7 = metadat_df_updated$chr7_Unsorted_Papilloma_Denoised_binary
        ca_uns_chr7 = plot_cells(pap_ca_unsorted,
                                 color_cells_by="cluster",
                                 #color_cells_by="chr7",
                                 label_cell_groups=FALSE,
                                 label_leaves=F,
                                 label_branch_points=F,
                                 graph_label_size=1.5,
                                 show_trajectory_graph=F,
                                 cell_size=0.6,
                                 alpha=0.4) +
          xlab("Dim. 1") + ylab("Dim. 2") + 
          theme(legend.position="bottom") +
          scale_color_manual(values = c("Duplicated" = "darkred", "Non-Duplicated" = "lightgrey"), name="Aneuploid status") +
          ggtitle(paste("Chr 7: Carcinoma"))+ #continuous
          theme(axis.text=element_text(size=8,color='black'),
                axis.title=element_text(size=10,color='black')) + theme(plot.title = element_text(size = 14, vjust=1)) +
          theme(strip.text.x = element_text(size = 0)) #remove facet label
        
        
        #chr6
        colData(pap_unsorted_cds)$chr6 = metadat_df_updated.1$chr6_Unsorted_Papilloma_Denoised_binary
        ca_uns_chr6 = plot_cells(pap_unsorted_cds,
                                 color_cells_by="chr6",
                                 label_cell_groups=FALSE,
                                 label_leaves=F,
                                 label_branch_points=F,
                                 graph_label_size=1.5,
                                 show_trajectory_graph=F,
                                 cell_size=0.6,
                                 alpha=0.4) +
          xlab("Dim. 1") + ylab("Dim. 2") + 
          theme(legend.position="bottom") +
          scale_color_manual(values = c("Duplicated" = "darkred", "Non-Duplicated" = "lightgrey"), name="Aneuploid status") +
          ggtitle(paste("Chr 6: Carcinoma"))+ #continuous
          theme(axis.text=element_text(size=8,color='black'),
                axis.title=element_text(size=10,color='black')) + theme(plot.title = element_text(size = 14, vjust=1)) +
          theme(strip.text.x = element_text(size = 0)) #remove facet label
        
        #chr10
        colData(pap_unsorted_cds)$chr10 = metadat_df_updated.1$chr10_Unsorted_Papilloma_Denoised_binary
        ca_uns_chr10 = plot_cells(pap_unsorted_cds,
                                  color_cells_by="chr10",
                                  label_cell_groups=FALSE,
                                  label_leaves=F,
                                  label_branch_points=F,
                                  graph_label_size=1.5,
                                  show_trajectory_graph=F,
                                  cell_size=0.6,
                                  alpha=0.4) +
          xlab("Dim. 1") + ylab("Dim. 2") + 
          theme(legend.position="bottom") +
          scale_color_manual(values = c("Duplicated" = "darkred", "Non-Duplicated" = "lightgrey"), name="Aneuploid status") +
          ggtitle(paste("Chr 10: Carcinoma"))+ #continuous
          theme(axis.text=element_text(size=8,color='black'),
                axis.title=element_text(size=10,color='black')) + theme(plot.title = element_text(size = 14, vjust=1)) +
          theme(strip.text.x = element_text(size = 0)) #remove facet label
        
        #chr1
        colData(pap_unsorted_cds)$chr1 = metadat_df_updated.1$chr1_Unsorted_Papilloma_Denoised_binary
        ca_uns_chr1 = plot_cells(pap_unsorted_cds,
                                 color_cells_by="chr1",
                                 label_cell_groups=FALSE,
                                 label_leaves=F,
                                 label_branch_points=F,
                                 graph_label_size=1.5,
                                 show_trajectory_graph=F,
                                 cell_size=0.6,
                                 alpha=0.4) +
          xlab("Dim. 1") + ylab("Dim. 2") + 
          theme(legend.position="bottom") +
          scale_color_manual(values = c("Duplicated" = "darkred", "Non-Duplicated" = "lightgrey"), name="Aneuploid status") +
          ggtitle(paste("Chr 1: Carcinoma"))+ #continuous
          theme(axis.text=element_text(size=8,color='black'),
                axis.title=element_text(size=10,color='black')) + theme(plot.title = element_text(size = 14, vjust=1)) +
          theme(strip.text.x = element_text(size = 0)) #remove facet label
        
        
        ################################################################################################################################    
        ################################################################################################################################    
        ################################################################################################################################
        ################################################################################################################################    
        ################################################################################################################################    
        ################################################################################################################################
        ################################################################################################################################    
        ################################################################################################################################    
        ################################################################################################################################
        ################################################################################################################################
        #Stem cell outline for Allan
        
        
        #Plot the supervised eigengene
        i=8 #column in Balmain eigengenes
        print(i)
        #only if more than 1 constituent gene
        marker_df = data.frame(na.omit(balmain_genes[,i]),names(balmain_genes)[i])
        names(marker_df) = c("gene","cell_type")
        
        #UPDATED NCBI MATCHING!! --> this will product lots of non-matching names that i must resolve with NCBI aliases
        allnetnames = do.call(rbind, lapply(1:nrow(marker_df), function(l){
          #l=1
          netnames = data.frame(marker_df$gene[l],rowData(big_cds)$gene_short_name[match(marker_df$gene[l],rowData(big_cds)$gene_short_name)])
          names(netnames) = c("balmain_network_gene_name","cds_gene_name")
          netnames
        }))
        cds_gene_names = rowData(big_cds)$gene_short_name #get vector of all gene names in cds object
        #get consistent names between cds and network gene names
        match_noms = do.call(rbind, lapply(1:nrow(allnetnames), function(k){
          #k=1 ; 22
          tryCatch({ #suppress error
            if(is.na(allnetnames$cds_gene_name[k])==F) {cds_alias_match = allnetnames$cds_gene_name[k]} else
            { 
              #search through Gene aliases for a match
              ncbi_symbol = ncbi_names[grep(allnetnames$balmain_network_gene_name[k],ncbi_names$Aliases),] #search through aliases for a match
              if (nrow(ncbi_symbol)>0) {
                ncbi_matches_vector = c(as.character(ncbi_symbol$Symbol), as.character(data.frame(strsplit(as.character(ncbi_symbol$Aliases),", "))[,1]) ) #vector of all aliases that marker_gene matches
                cds_alias_match = na.omit(cds_gene_names[match(ncbi_matches_vector,cds_gene_names)]) #match this symbol to the cds gene names
              } else {
                #search through Gene Symbols
                ncbi_symbol.1 = ncbi_names[grep(allnetnames$balmain_network_gene_name[k],ncbi_names$Symbol),] 
                ncbi_matches_vector = c(as.character(ncbi_symbol.1$Symbol), as.character(data.frame(strsplit(as.character(ncbi_symbol.1$Aliases),", "))[,1]) ) #vector of all aliases that marker_gene matches
                cds_alias_match = na.omit(cds_gene_names[match(ncbi_matches_vector,cds_gene_names)]) #match this symbol to the cds gene names
              }
            }
            data.frame(allnetnames[k,],cds_alias_match)
          },  error=function(e){})
        }))
        
        #remake the marker df dataframe with the matched gene names
        marker_df = data.frame(na.omit(unique(match_noms$cds_alias_match)),names(balmain_genes)[i])
        names(marker_df) = c("gene","cell_type")
        
        #Monocle eigene expression
        marker_df$module <- factor(marker_df$cell_type)
        
        #all cells: Seed gene
        individual_gene_manifolds = plot_cells(big_cds,
                                               genes=marker_df$gene, #seed gene is the first row of the markers file
                                               label_cell_groups=FALSE,
                                               label_leaves=F,
                                               label_branch_points=F,
                                               graph_label_size=1.5,
                                               show_trajectory_graph=F) +
          xlab("Dim. 1") + ylab("Dim. 2") + 
          theme(legend.position="bottom") +
          scale_color_viridis_c(option = "inferno",name="log expression") +
          ggtitle(paste("All cells:", names(balmain_genes)[i],"Seed Gene:", marker_df$gene[1],sep=" "))+ #continuous
          theme(axis.text=element_text(size=8,color='black'),
                axis.title=element_text(size=10,color='black')) + theme(plot.title = element_text(size = 0, vjust=1)) +
          theme(strip.text.x = element_text(size = 20)) #remove facet label
        
        png("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/stem_cell_mansucript/Figs/Fig1/Fig1_supervised_stem.png",
            width=15,height=15,units="in",res=400)
        individual_gene_manifolds
        dev.off()
        
        #Plot individual 
        individual_gene_manifolds = plot_cells(big_cds,
                                               genes=marker_df$gene, #seed gene is the first row of the markers file
                                               label_cell_groups=FALSE,
                                               label_leaves=F,
                                               label_branch_points=F,
                                               graph_label_size=1.5,
                                               show_trajectory_graph=F) +
          xlab("Dim. 1") + ylab("Dim. 2") + 
          theme(legend.position="bottom") +
          scale_color_viridis_c(option = "inferno",name="log expression") +
          ggtitle(paste("All cells:", names(balmain_genes)[i],"Seed Gene:", marker_df$gene[1],sep=" "))+ #continuous
          theme(axis.text=element_text(size=8,color='black'),
                axis.title=element_text(size=10,color='black')) + theme(plot.title = element_text(size = 0, vjust=1)) +
          theme(strip.text.x = element_text(size = 20)) #remove facet label
        
        png("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/stem_cell_mansucript/Figs/Fig1/Fig1_supervised_stem.png",
            width=15,height=15,units="in",res=400)
        individual_gene_manifolds
        dev.off()
        
        
        #plot the stem genes
        #Plot the supervised eigengene
        i=334 #column in Balmain eigengenes
        print(i)
        #only if more than 1 constituent gene
        marker_df = data.frame(na.omit(balmain_genes[,i]),names(balmain_genes)[i])
        names(marker_df) = c("gene","cell_type")
        
        #UPDATED NCBI MATCHING!! --> this will product lots of non-matching names that i must resolve with NCBI aliases
        allnetnames = do.call(rbind, lapply(1:nrow(marker_df), function(l){
          #l=1
          netnames = data.frame(marker_df$gene[l],rowData(big_cds)$gene_short_name[match(marker_df$gene[l],rowData(big_cds)$gene_short_name)])
          names(netnames) = c("balmain_network_gene_name","cds_gene_name")
          netnames
        }))
        cds_gene_names = rowData(big_cds)$gene_short_name #get vector of all gene names in cds object
        #get consistent names between cds and network gene names
        match_noms = do.call(rbind, lapply(1:nrow(allnetnames), function(k){
          #k=1 ; 22
          tryCatch({ #suppress error
            if(is.na(allnetnames$cds_gene_name[k])==F) {cds_alias_match = allnetnames$cds_gene_name[k]} else
            { 
              #search through Gene aliases for a match
              ncbi_symbol = ncbi_names[grep(allnetnames$balmain_network_gene_name[k],ncbi_names$Aliases),] #search through aliases for a match
              if (nrow(ncbi_symbol)>0) {
                ncbi_matches_vector = c(as.character(ncbi_symbol$Symbol), as.character(data.frame(strsplit(as.character(ncbi_symbol$Aliases),", "))[,1]) ) #vector of all aliases that marker_gene matches
                cds_alias_match = na.omit(cds_gene_names[match(ncbi_matches_vector,cds_gene_names)]) #match this symbol to the cds gene names
              } else {
                #search through Gene Symbols
                ncbi_symbol.1 = ncbi_names[grep(allnetnames$balmain_network_gene_name[k],ncbi_names$Symbol),] 
                ncbi_matches_vector = c(as.character(ncbi_symbol.1$Symbol), as.character(data.frame(strsplit(as.character(ncbi_symbol.1$Aliases),", "))[,1]) ) #vector of all aliases that marker_gene matches
                cds_alias_match = na.omit(cds_gene_names[match(ncbi_matches_vector,cds_gene_names)]) #match this symbol to the cds gene names
              }
            }
            data.frame(allnetnames[k,],cds_alias_match)
          },  error=function(e){})
        }))
        
        #remake the marker df dataframe with the matched gene names
        marker_df = data.frame(na.omit(unique(match_noms$cds_alias_match)),names(balmain_genes)[i])
        names(marker_df) = c("gene","cell_type")
        
        #Monocle eigene expression
        marker_df$module <- factor(marker_df$cell_type)
        
        #all cells: Seed gene
        individual_gene_manifolds = plot_cells(big_cds,
                                               genes=marker_df$gene, #seed gene is the first row of the markers file
                                               label_cell_groups=FALSE,
                                               label_leaves=F,
                                               label_branch_points=F,
                                               graph_label_size=1.5,
                                               show_trajectory_graph=F) +
          xlab("Dim. 1") + ylab("Dim. 2") + 
          theme(legend.position="bottom") +
          scale_color_viridis_c(option = "inferno",name="log expression") +
          ggtitle(paste("All cells:", names(balmain_genes)[i],"Seed Gene:", marker_df$gene[1],sep=" "))+ #continuous
          theme(axis.text=element_text(size=8,color='black'),
                axis.title=element_text(size=10,color='black')) + theme(plot.title = element_text(size = 0, vjust=1)) +
          theme(strip.text.x = element_text(size = 20)) #remove facet label
        
        png("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/stem_cell_mansucript/tSC_fuch_markers_stem.png",
            width=24,height=16,units="in",res=400)
        individual_gene_manifolds
        dev.off()
        
        #Plot individual 
        individual_gene_manifolds = plot_cells(big_cds,
                                               genes=marker_df$gene, #seed gene is the first row of the markers file
                                               label_cell_groups=FALSE,
                                               label_leaves=F,
                                               label_branch_points=F,
                                               graph_label_size=1.5,
                                               show_trajectory_graph=F) +
          xlab("Dim. 1") + ylab("Dim. 2") + 
          theme(legend.position="bottom") +
          scale_color_viridis_c(option = "inferno",name="log expression") +
          ggtitle(paste("All cells:", names(balmain_genes)[i],"Seed Gene:", marker_df$gene[1],sep=" "))+ #continuous
          theme(axis.text=element_text(size=8,color='black'),
                axis.title=element_text(size=10,color='black')) + theme(plot.title = element_text(size = 0, vjust=1)) +
          theme(strip.text.x = element_text(size = 20)) #remove facet label
        
        png("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/stem_cell_mansucript/Figs/Fig1/Fig1_supervised_stem.png",
            width=15,height=15,units="in",res=400)
        individual_gene_manifolds
        dev.off()
        
        
        
        #all cells eigengene
        agexp_plot = plot_cells(big_cds,
                                genes=marker_df,
                                label_cell_groups=FALSE,
                                label_leaves=F,
                                label_branch_points=F,
                                graph_label_size=1.5,
                                show_trajectory_graph=F,
                                cell_size = 0.5) +
          xlab("Dim. 1") + ylab("Dim. 2") + 
          theme(legend.position="bottom") +
          scale_color_viridis_c(option = "inferno",name="integrated expression") +
          ggtitle(paste("Stem cell eigengene"))+ #continuous
          theme(axis.text=element_text(size=8,color='black'),
                axis.title=element_text(size=10,color='black')) + theme(plot.title = element_text(size = 14, vjust=1)) +
          theme(strip.text.x = element_text(size = 0)) #remove facet label
        
        png("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/stem_cell_mansucript/Figs/Fig1/Fig1_supervised_stem_eigengene.png",
            width=8,height=8,units="in",res=400)
        agexp_plot
        dev.off()
        
        #-----------------------------------------------------------#
        #now plot lower spike up close
        #-----------------------------------------------------------#
        #Lower spike cluster 33
        cds_subset.v3 = big_cds[,colData(big_cds)$cluster %in% c("33")] #filter for cluster 33 because most of the lower spike cells (all of them?) are in that cluster
        write.table(data.frame(colnames(cds_subset.v3)),file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/cell_subsets_by_cell_id/lower_spike_cluster33.txt", sep="\t",row.names=F, quote=F) #write out the cell names of the cells I'm calling the upper spike
        
        #add this to the big_cds object
        matching_cells = match(colnames(cds_subset.v3),colnames(big_cds)) #match extracted cells with all cell names
        colData(big_cds)$lower_spike = "non-lower-spike" #add column of of apparent no matches to ALL the cells; later you will replace with matches
        colData(big_cds)$lower_spike[matching_cells] <- "lower-spike" #now replace matching cells with new name
        
        #now do trajecotry analysis
        subset_pr_test_res.v2 <- graph_test(cds_subset.v3, neighbor_graph="knn", cores=4) #perform trajectory analysis
        pr_deg_ids.v2 <- row.names(subset(subset_pr_test_res.v2, q_value < 0.05)) #extract only sig genes by Moran I test
        gene_module_df.v2 <- data.frame(find_gene_modules(cds_subset.v2[pr_deg_ids.v2,], resolution=0.001)) #find unsupervised modules in genes that are significant in the trajectory analysis
        re_order_gene_changes.v2 = data.frame(subset_pr_test_res.v2[order(subset_pr_test_res.v2$q_value),]) #re-order single gene results from high to low
        graph_test_df_comb.v2 = data.frame(re_order_gene_changes.v2, gene_module_df.v2[match(re_order_gene_changes.v2$gene_short_name, gene_module_df$id),])    #combine with module
        graph_test_df_comb.reord.v2 = graph_test_df_comb.v2[,c(5,8,9,10,11,4,3,2,6)] #re-order columns
        write.table(graph_test_df_comb.reord.v2, file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/DEG/unsupervised_trajectories/lower_spike.txt",sep="\t",row.names=F,quote=F)
        
        #Now find markers: one region vs all others
        marker_test_res <- top_markers(big_cds, group_cells_by="lower_spike",
                                       genes_to_test_per_group = 1000,
                                       cores=8)
        marker_test_res = marker_test_res[order(-marker_test_res$specificity),] #re-order high to low specificity
        marker_test_res.lower_spike = subset(marker_test_res, cell_group == "lower-spike")
        write.table(marker_test_res.lower_spike, file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/top_markers/lower_spike_vs_all_other_cells_top_markers.txt", quote=F,row.names = F,sep="\t")
        
        #plot lower spike
        lower_spike_manifold = plot_cells(big_cds, color_cells_by="lower_spike", group_cells_by="cluster",
                                          group_label_size = 0,
                                          x=1,
                                          y=2,
                                          alpha=0.8,
                                          cell_size = 0.5) +
          ggtitle("Lower spike") + xlab("") + ylab("") + 
          theme(plot.title = element_text(size=20),
                axis.text=element_text(size=0,color='black'),
                axis.title=element_text(size=0),
                axis.ticks = element_line(size=0),
                legend.position="none") +
          scale_color_manual(values = c("non-lower-spike" = "lightgrey", "lower-spike" = "darkred"))
        
        png("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/stem_cell_mansucript/Figs/Fig1/Fig1_lowerspike.png",
            width=8,height=8,units="in",res=400)
        lower_spike_manifold
        dev.off()
        
        #Find overlaps between lower-spike top markers and stem eigengene
        nrow(marker_test_res.lower_spike)
        top_marker_overlap = na.omit(marker_test_res.lower_spike[match(marker_df$gene, marker_test_res.lower_spike$gene_id),])
        
        marker_test_res.lower_spike_sig = subset(marker_test_res.lower_spike, marker_test_q_value<0.05)
        nrow(marker_test_res.lower_spike_sig)
        
        #So are the top markers enriched for stem cell genes: hypergeometric tests
        assayed = as.character(marker_df$gene)
        significant = top_marker_overlap$gene_id
        universe = row.names(big_cds)
        hypergeometric_results = data.frame(hyperg(assayed, significant, universe,
                                                   representation = c("over")))
        
        ##################################################################################################################################
        ##################################################################################################################################
        #Look for stem cell enrichment by lineage tracing
        head(colData(cds_comb))
        parenchymas = unique(colData(cds_comb)$parenchyma_tissue)
        parenchyma_proportion = data.frame(table(colData(cds_comb)$parenchyma_tissue) / dim(cds_comb)[2])
        
        stem_cell_parenchymas = do.call(rbind, lapply(1:length(parenchymas), function(i){
          print(i)
          
          #i=1
          #1. Stage & Lgr6+
          sub_lgr6 = cds_comb[,(colData(cds_comb)$parenchyma_tissue %in% parenchymas[i] & colData(cds_comb)$sorting %in% "Lgr6")] #subset to a specific stage AND  Lgr6+
          agg_mat_sub_lgr6 = aggregate_gene_expression(
            sub_lgr6,
            gene_group_df = marker_df,
            cell_group_df = NULL,
            norm_method = c("log"),
            pseudocount = 1,
            scale_agg_values = TRUE,
            #max_agg_value = 5,
            min_agg_value = 0,
            exclude.na = TRUE
          )
          agg_mat_sub_lgr6 = data.frame(t(agg_mat_sub_lgr6))
          weighted_eigencell = colSums(agg_mat_sub_lgr6) * subset(parenchyma_proportion, Var1==parenchymas[i])$Freq #get sum and multiply by proportion of that stage's cells in the total stage
          agg_mat_sub_lgr6 = data.frame(parenchymas[i],"Lgr6",weighted_eigencell)
          names(agg_mat_sub_lgr6) = c("parenchyma","sorting","weighted_eigencell")
          
          #2. Stage & Progeny
          sub_progeny = cds_comb[,(colData(cds_comb)$parenchyma_tissue %in% parenchymas[i] & colData(cds_comb)$sorting %in% "Progeny")] #subset to a specific stage AND  Lgr6+
          agg_mat_sub_progeny = aggregate_gene_expression(
            sub_progeny,
            gene_group_df = marker_df,
            cell_group_df = NULL,
            norm_method = c("log"),
            pseudocount = 1,
            scale_agg_values = TRUE,
            #max_agg_value = 5,
            min_agg_value = 0,
            exclude.na = TRUE
          )
          agg_mat_sub_progeny = data.frame(t(agg_mat_sub_progeny))
          weighted_eigencell = colSums(agg_mat_sub_progeny) * subset(parenchyma_proportion, Var1==parenchymas[i])$Freq #get sum and multiply by proportion of that stage's cells in the total stage
          agg_mat_sub_progeny = data.frame(parenchymas[i],"Progeny",weighted_eigencell)
          names(agg_mat_sub_progeny) = c("parenchyma","sorting","weighted_eigencell")
          
          #3. combine and return
          rbind(agg_mat_sub_lgr6,agg_mat_sub_progeny)
          
        }))
        stem_cell_parenchymas$scaled_weighted_eigencell = scale(stem_cell_parenchymas$weighted_eigencell, center=F)
        
        ggplot(stem_cell_parenchymas, aes(x=sorting,y=(scaled_weighted_eigencell), fill=sorting)) +
          geom_bar(stat = "identity") +
          facet_wrap(~parenchyma, scales="free", nrow=1) + 
          #geom_jitter(width = 0.1) +
          ggtitle(paste("Stem cell eigengene")) + 
          xlab("") + ylab("weighted scaled stem eigencell expression") + 
          theme(axis.text.x=element_text(size=15,color='black'),
                axis.text.y=element_text(size=10,color='black'),
                axis.title=element_text(size=15),
                axis.ticks.x = element_line(size=0),
                strip.text.x = element_text(size = 15),
                legend.position="none",
                title = element_text(size = 15, face="bold")) +
          scale_fill_manual(values = c("Lgr6" = "lightsteelblue", "Progeny" = "tomato3"),
                            name="Lineage\nsorting") +
          theme(
            # Hide panel borders and remove grid lines
            panel.border = element_blank(),
            panel.grid.major = element_blank(),
            panel.grid.minor = element_blank(),
            panel.background = element_blank(),
            element_line(colour = 'black', size = 0.3),
            # Change axis line
            axis.line = element_line(colour = "black"),
            strip.background = element_rect(colour="white",
                                            fill="white")
          )
        
        
        ##################################################################################################################################
        ##################################################################################################################################
        ##################################################################################################################################  
        ####################################################################################################################
        ####################################################################################################################
        ####################################################################################################################
        #DEG analysis on Lgr6 lineage: Lgr6+ cells vs progeny accroding to FACS
        
        #Subset to stage-specific parenchyma
        cds_comb = big_cds[,colData(big_cds)$partition %in% c(1,2,3,4)]
        tissues = unique(colData(cds_comb)$tissue_updated)
        
        #Start loop to: 1) subset to carcinogenesis stage 2) find top 4k variable genes presentin at least 5% of the data 3) test DEG
        for (i in 2:length(tissues)) {
          
          tryCatch({ #suppress error
            #i=1
            pap_cds = cds_comb[,colData(cds_comb)$tissue_updated %in% tissues[i]] #
            
            #Get minimum number of cells a gene has to be expressed by in order to be considereed
            minimum_cells = round(dim(pap_cds)[2]*0.05)
            
            #now create Seurat object, filtering geenes
            seurat_obj = CreateSeuratObject(
              pap_cds@assays@data$counts,
              project = "CreateSeuratObject",
              assay = "RNA",
              names.field = 1,
              names.delim = "_",
              meta.data = NULL,
              min.cells = minimum_cells)
            
            #FindVariableFeatures
            seurat_obj <- FindVariableFeatures(seurat_obj, selection.method = "vst", nfeatures = 4000)
            variable_genes <- VariableFeatures(seurat_obj) #get variable gene names
            
            to_test_cds = pap_cds[rowData(pap_cds)$gene_short_name %in% variable_genes,] #subset cds object to variable genes
            
            meandiffs = do.call(rbind, lapply(1:length(variable_genes), function(e){
              # meandiffs = do.call(rbind, lapply(1:10, function(e){
              
              tryCatch({ #suppress error
                #e=1
                
                print(paste("i=",i," e=",e,sep=""))
                
                #RFP+ Progeny cells
                ca_data = to_test_cds[rowData(to_test_cds)$gene_short_name %in% variable_genes[e],colData(to_test_cds)$sorting %in% "Progeny"]@assays@data$counts
                ca_mean = mean(ca_data)
                ca_sd = sd(ca_data)
                ca_percent_expressed = (length(ca_data[ca_data>0]) /  length(ca_data))*100
                
                #Lgr6+ cells
                normaskin_data = to_test_cds[rowData(to_test_cds)$gene_short_name %in% variable_genes[e],colData(to_test_cds)$sorting %in% "Lgr6"]@assays@data$counts
                normakin_mean = mean(normaskin_data)
                normaskin_sd = sd(normaskin_data)
                normaskin_percent_expressed = (length(normaskin_data[normaskin_data>0]) /  length(normaskin_data))*100
                
                #differences
                mean_abs_diff = ca_mean - normakin_mean
                mean_per_diff = ((ca_mean - normakin_mean)/normakin_mean)*100
                fc = ca_mean/normakin_mean
                
                #test expression diffs: make dataframe with tissue and p16 status
                ca_dat = data.frame(as.numeric(ca_data), "Progeny")
                normaskin_dat = data.frame(as.numeric(normaskin_data), "Lgr6")
                names(ca_dat) = c("expression","tissue")
                names(normaskin_dat) = c("expression","tissue")
                agg_matdf = rbind(normaskin_dat,ca_dat)
                names(agg_matdf) = c("expression","tissue")
                
                #scale everything up to 0 for negative binormail test
                rescale_value = 0-min(agg_matdf$expression)
                agg_matdf$expression = agg_matdf$expression+rescale_value
                
                #fit negative binomial GLM to eigengene testing for amount-tissue differences
                glmnb_res <- glm.nb(expression~tissue, data = agg_matdf)
                
                #return df
                standard_result_df = data.frame(variable_genes[e],
                                                ca_mean, ca_sd, ca_percent_expressed, 
                                                normakin_mean, normaskin_sd, normaskin_percent_expressed,
                                                mean_abs_diff, mean_per_diff, fc,
                                                t(summary(glmnb_res)$coefficients[2,])
                )
                
                names(standard_result_df)[1:7] = c("gene",
                                                   paste("Progeny","_mean",sep=""),  paste("Progeny","_sd",sep=""),  paste("Progeny","_percent_expressed",sep=""),
                                                   paste("Lgr6","_mean",sep=""),  paste("Lgr6","_sd",sep=""),  paste("Lgr6","_percent_expressed",sep="")
                )
                
                return(standard_result_df) #if there is p16 variation return both results
                
              },  error=function(e){})
            }))
            
            #re-order from high to low percent change
            meandiffs = meandiffs[order(-meandiffs$fc),]
            
            #the infinity values mean that normal_skin is at 0; re-order within the infitnite group based on absolute mean diffchanges
            infdiffs = meandiffs[is.infinite(meandiffs$mean_per_diff),] #extract inifity values
            infdiffs = infdiffs[order(-infdiffs$mean_abs_diff),] #re-order absolute differences high to low
            
            #remove infinite values and then add the re-ordered inf_diffs
            no_inf_diffs <- meandiffs[!is.infinite(meandiffs$mean_per_diff),]
            meandiffs = rbind(infdiffs,no_inf_diffs)
            
            #write out results
            write.table(meandiffs, 
                        file=paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/DEG/lineage_sorting_deg/",
                                   tissues[i],
                                   "lineage_tracing_degs.txt",sep=""),
                        quote=F,sep="\t",row.names=F)
            
            #--------------------------------------------#
            #GO analysis on top DEGs
            
            #create topGO object: top 100 POSITIVE DEGs
            geneUniverse <- row.names(cds_comb)
            genesOfInterest = meandiffs$gene[1:100]
            geneList <- factor(as.integer(geneUniverse %in% genesOfInterest))
            names(geneList) <- geneUniverse
            GOdata <- new("topGOdata",
                          ontology = "BP",
                          allGenes=geneList,
                          #geneSel = selection,
                          annot = annFUN.org, 
                          mapping = "org.Mm.eg.db",
                          ID = "symbol")
            
            
            citation("org.Mm.eg.db")
            
            resultKS <- runTest(GOdata, algorithm = "weight01", statistic = "fisher") #Fisher exact test
            tab <- GenTable(GOdata, raw.p.value = resultKS, topNodes = length(resultKS@score), numChar = 120)
            tab = subset(tab, raw.p.value <=0.01) #keep only sig results 
            
            write.table(tab, 
                        file=paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/DEG/lineage_sorting_deg/",
                                   tissues[i],
                                   "_top100deg_GOterms.txt",sep=""),
                        quote=F,sep="\t",row.names=F)
            
          },  error=function(e){})
          
        } #End outer loop teesting for main DEGs between groups
        
        ##############################################################################################
        ##############################################################################################
        ##############################################################################################
        
        ####################################################################################################################
        ####################################################################################################################
        ####################################################################################################################
        #Plot eigengene expression values analysis from Lgr6 to progeny
        
        eigengene_agg = do.call(rbind, lapply(350:ncol(balmain_genes), function(i){
          #subset eigengene_agg = do.call(cbind, lapply(172:175, function(i){
          #tryCatch({ #suppress error
          
          #i=350
          print(i)
          #only if more than 1 constituent gene
          marker_df = data.frame(na.omit(balmain_genes[,i]),names(balmain_genes)[i])
          names(marker_df) = c("gene","cell_type")
          
          #remake the marker df dataframe with the matched gene names
          marker_df = data.frame(unique(alias2SymbolTable(unique(marker_df$gene), species = "Mm")),names(balmain_genes)[i],names(balmain_genes)[i])
          names(marker_df) = c("gene","cell_type","module")
          marker_df = na.omit(marker_df) #remove non-matching genes
          
          #aggregate gene values to get a single expression value
          agg_mat = aggregate_gene_expression(
            big_cds,
            gene_group_df = marker_df,
            cell_group_df = NULL,
            norm_method = c("log"),
            pseudocount = 1,
            scale_agg_values = F,
            max_agg_value = NA,
            min_agg_value = NA,
            exclude.na = TRUE
          )
          
          #Cluster-specific BEFORE GENE DROPPED
          agg_gene_before = data.frame(c(agg_mat),names(balmain_genes)[i],colData(big_cds))
          names(agg_gene_before)[1:2] = c("eigengene_expression","eigengene")
          agg_gene_before = subset(agg_gene_before, sorting!="Unsorted") #subset to only Lgr6 and progeny
          return(agg_gene_before)
          
        }))
        
        head(eigengene_agg)
        stem_cell_parenchymas$scaled_weighted_eigencell = scale(stem_cell_parenchymas$weighted_eigencell, center=F)
        
        my_comparisons <- list( c("Normal skin", "Pap"), c("Pap", "CA"), c("Normal skin", "CA"), c("Control", "Cisplatin") )
        my_comparisons <- list( c("Progeny", "Lgr6")) 
        
        eigengene_agg$eigengene_sorting = paste(eigengene_agg$eigengene,eigengene_agg$sorting,sep="_")
        eigengene_agg = na.omit(eigengene_agg)
        
        lgr6_progeny_eigengenes = ggplot(eigengene_agg, aes(x=eigengene,y=eigengene_expression, fill=sorting, eigengene=sorting)) +
          stat_compare_means(comparisons = my_comparisons) +
          #geom_violin(aes(fill=sorting), width=1) +
          geom_boxplot(outlier.size=0, width=0.6, color="black") +
          #geom_bar(stat = "identity", fill="darkorange") +
          #facet_wrap(~parenchyma, scales="free", nrow=1) + 
          #geom_jitter(width = 0.1) +
          ggtitle(paste("Super-enahncer-associated eigengenes")) + 
          xlab("") + ylab("eigengene expression") + 
          theme(axis.text.x=element_text(size=15,color='black',angle=90,hjust = 0.98),
                axis.text.y=element_text(size=10,color='black'),
                axis.title=element_text(size=15),
                axis.title.x=element_text(size=15),
                axis.ticks.x = element_line(size=0),
                strip.text.x = element_text(size = 15),
                legend.position="none",
                title = element_text(size = 15, face="bold")) +
          theme(
            # Hide panel borders and remove grid lines
            panel.border = element_blank(),
            panel.grid.major = element_blank(),
            panel.grid.minor = element_blank(),
            panel.background = element_blank(),
            element_line(colour = 'black', size = 0.3),
            # Change axis line
            axis.line = element_line(colour = "black"),
            strip.background = element_rect(colour="white",
                                            fill="white")
          ) +
          scale_fill_manual(values = c("Progeny" = "goldenrod","Lgr6" = "burlywood4")) + theme(legend.position="right")
        
        png("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/sorting_analysis/superenhancer_eigengenes.png",
            width=14,height=14,units="in",res=150)
        print(lgr6_progeny_eigengenes)
        dev.off()
        
        #factor across skin types
        unique(eigengene_agg$cell_type)
        
        eigengene_agg$cell_type_updated = eigengene_agg$cell_type
        eigengene_agg$cell_type_updated = gsub("Bas","Follicular", eigengene_agg$cell_type_updated)
        eigengene_agg$cell_type_updated = gsub("OuB","Follicular", eigengene_agg$cell_type_updated)
        eigengene_agg$cell_type_updated = gsub("Inf","Follicular", eigengene_agg$cell_type_updated)
        eigengene_agg$cell_type_updated = gsub("Isth","Follicular", eigengene_agg$cell_type_updated)
        eigengene_agg$cell_type_updated = gsub("UHF","Follicular", eigengene_agg$cell_type_updated)
        eigengene_agg$cell_type_updated = gsub("unF","Follicular", eigengene_agg$cell_type_updated)
        eigengene_agg$cell_type_updated = gsub("InB","Follicular", eigengene_agg$cell_type_updated)
        eigengene_agg$cell_type_updated = gsub("IfE","Inter-follicular", eigengene_agg$cell_type_updated)
        eigengene_agg$cell_type_updated = gsub("PrE","Inter-follicular", eigengene_agg$cell_type_updated)
        
        eigengene_agg_parenchyma = eigengene_agg[eigengene_agg$cell_type %in% c("Bas","OuB","Inf","Isth","IfE",
                                                                                "UHF","PrE","unF","SG","InB",
                                                                                "nEp"),]
        eigengene_agg_parenchyma$tisue_celltype = paste(eigengene_agg_parenchyma$tissue_updated, eigengene_agg_parenchyma$cell_type_updated)
        eigengene_agg_parenchyma = eigengene_agg_parenchyma[eigengene_agg_parenchyma$tisue_celltype %in% 
                                                              c("Normal skin Follicular",
                                                                "Normal skin Inter-follicular",
                                                                "Pap nEp",
                                                                "CA nEp"),]
        
        lgr6_progeny_eigengene_parenchyma = ggplot(eigengene_agg_parenchyma, aes(x=eigengene,y=eigengene_expression, fill=sorting, eigengene=sorting)) +
          geom_boxplot(outlier.size=0, width=0.6, color="black") +
          facet_wrap(~tisue_celltype, scales="free_y", ncol=1) + 
          ggtitle(paste("Super-enahncer-associated eigengenes")) + 
          xlab("") + ylab("eigengene expression") + 
          theme(axis.text.x=element_text(size=15,color='black',angle=90,hjust = 0.98),
                axis.text.y=element_text(size=10,color='black'),
                axis.title=element_text(size=15),
                axis.title.x=element_text(size=15),
                axis.ticks.x = element_line(size=0),
                strip.text.x = element_text(size = 15),
                legend.position="none",
                title = element_text(size = 15, face="bold")) +
          theme(
            # Hide panel borders and remove grid lines
            panel.border = element_blank(),
            panel.grid.major = element_blank(),
            panel.grid.minor = element_blank(),
            panel.background = element_blank(),
            element_line(colour = 'black', size = 0.3),
            # Change axis line
            axis.line = element_line(colour = "black"),
            strip.background = element_rect(colour="white",
                                            fill="white")
          ) +
          scale_fill_manual(values = c("Progeny" = "goldenrod","Lgr6" = "burlywood4")) + theme(legend.position="right")
        
        
        png("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/sorting_analysis/superenhancer_eigengenes_parenchyma.png",
            width=14,height=30,units="in",res=150)
        print(lgr6_progeny_eigengene_parenchyma)
        dev.off()
        
        ####################################################################################################################
        ####################################################################################################################
        ####################################################################################################################
        ####################################################################################################################
        #find and count overlapping genes within eigengenes and then plot the most overlaping genes into an "eigengene"
        
        #read in eigengenes
        eigengenes_toanalyze = read.xlsx(
          xlsxFile="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/apriori_eigengenes/eigengene_overlaps/eigengenes_to_search_for_overlapping_genes.xlsx",
          colNames=T
        ) 
        
        #start loop to choose a column in the eigengenes_toanalyse file and find overlaps and plot eigengenes within it
        #f=1
        sub_eigen = subset(balmain_genes, select=na.omit(eigengenes_toanalyze[,f])) #select columns of eigengenes from the balmain_genes file
        sub_eigen_vec = stack(sub_eigen) #stack all columns into 1 column
        sub_eigen_vec = na.omit(sub_eigen_vec) #remove NAs
        sub_eigen_vec = subset(sub_eigen_vec, values!="NO ORTHOLOG DETECTED") #remove NO ORTHOLOG DETECTED
        
        length(unique(sub_eigen_vec$values)) #the number of unique genes encompassed by an eigengene
        length(unique(sub_eigen_vec$ind)) #the number of unique eigengenes
        freq_table = data.frame(table(sub_eigen_vec$values)) #calculate frequency table of genes
        names(freq_table)[1] = "gene"
        freq_table = freq_table[order(-freq_table$Freq),] #re-order from most to least frequent
        freq_table$percent_eigengenes = round((freq_table$Freq / length(unique(sub_eigen_vec$ind)))*100,2)
        
        write.table(freq_table, 
                    file=paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/apriori_eigengenes/eigengene_overlaps/", names(eigengenes_toanalyze)[f],"_overlap_frequencies.txt",sep=""),
                    sep="\t",row.names=F,quote=F)
        
        ##################################################################################################################################
        #make directory and plot manifolds into it
        dir.create(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/apriori_eigengenes/eigengene_overlaps/", names(eigengenes_toanalyze)[f],sep=""))
        ##################################################################################################################################
        
        #make eigengenes of top 5% most frequent
        top5_index = round(nrow(freq_table)*0.01,0)
        
        eigengene_agg = do.call(cbind, lapply(0:(top5_index-2), function(i){
          #subset eigengene_agg = do.call(cbind, lapply(106:118, function(i){
          tryCatch({ #suppress error
            
            #i=89
            print(i)
            #only if more than 1 constituent gene
            marker_df = data.frame(na.omit(freq_table[1:(top5_index-i),1]),names(eigengenes_toanalyze)[f])
            names(marker_df) = c("gene","cell_type")
            
            #UPDATED NCBI MATCHING!! --> this will product lots of non-matching names that i must resolve with NCBI aliases
            allnetnames = do.call(rbind, lapply(1:nrow(marker_df), function(l){
              #l=1
              netnames = data.frame(marker_df$gene[l],rowData(big_cds)$gene_short_name[match(marker_df$gene[l],rowData(big_cds)$gene_short_name)])
              names(netnames) = c("balmain_network_gene_name","cds_gene_name")
              netnames
            }))
            cds_gene_names = rowData(big_cds)$gene_short_name #get vector of all gene names in cds object
            #get consistent names between cds and network gene names
            match_noms = do.call(rbind, lapply(1:nrow(allnetnames), function(k){
              #k=1 ; 22
              tryCatch({ #suppress error
                if(is.na(allnetnames$cds_gene_name[k])==F) {cds_alias_match = allnetnames$cds_gene_name[k]} else
                { 
                  #search through Gene aliases for a match
                  ncbi_symbol = ncbi_names[grep(allnetnames$balmain_network_gene_name[k],ncbi_names$Aliases),] #search through aliases for a match
                  if (nrow(ncbi_symbol)>0) {
                    ncbi_matches_vector = c(as.character(ncbi_symbol$Symbol), as.character(data.frame(strsplit(as.character(ncbi_symbol$Aliases),", "))[,1]) ) #vector of all aliases that marker_gene matches
                    cds_alias_match = na.omit(cds_gene_names[match(ncbi_matches_vector,cds_gene_names)]) #match this symbol to the cds gene names
                  } else {
                    #search through Gene Symbols
                    ncbi_symbol.1 = ncbi_names[grep(allnetnames$balmain_network_gene_name[k],ncbi_names$Symbol),] 
                    ncbi_matches_vector = c(as.character(ncbi_symbol.1$Symbol), as.character(data.frame(strsplit(as.character(ncbi_symbol.1$Aliases),", "))[,1]) ) #vector of all aliases that marker_gene matches
                    cds_alias_match = na.omit(cds_gene_names[match(ncbi_matches_vector,cds_gene_names)]) #match this symbol to the cds gene names
                  }
                }
                data.frame(allnetnames[k,],cds_alias_match)
              },  error=function(e){})
            }))
            
            #remake the marker df dataframe with the matched gene names
            marker_df = data.frame(na.omit(unique(match_noms$cds_alias_match)),names(eigengenes_toanalyze)[f])
            names(marker_df) = c("gene","cell_type")
            #Monocle eigene expression
            marker_df$module <- factor(marker_df$cell_type)
            
            #All tissues, all cells eigengene
            agexp_plot = plot_cells(big_cds,
                                    genes=marker_df,
                                    label_cell_groups=FALSE,
                                    label_leaves=F,
                                    label_branch_points=F,
                                    graph_label_size=1.5,
                                    show_trajectory_graph=F) +
              xlab("Dim. 1") + ylab("Dim. 2") + 
              theme(legend.position="bottom") +
              scale_color_viridis_c(option = "inferno",name="integrated expression") +
              ggtitle(paste("All cells: Upper Spike Eigengene N=",nrow(marker_df),sep=""))+ #continuous
              theme(axis.text=element_text(size=8,color='black'),
                    axis.title=element_text(size=10,color='black')) + theme(plot.title = element_text(size = 14, vjust=1)) +
              theme(strip.text.x = element_text(size = 0)) #remove facet label
            
            #Pap_ca eigengene
            agexp_plot_pap_ca = plot_cells(pap_ca.test,
                                           genes=marker_df,
                                           label_cell_groups=FALSE,
                                           label_leaves=F,
                                           label_branch_points=F,
                                           graph_label_size=1.5,
                                           show_trajectory_graph=F,
                                           cell_size=1,
                                           alpha=0.8) +
              xlab("Dim. 1") + ylab("Dim. 2") + 
              theme(legend.position="bottom") +
              scale_color_viridis_c(option = "inferno",name="integrated expression") +
              ggtitle(paste("Pap&CA: Upper Spike Eigengene N=",nrow(marker_df),sep=""))+ #continuous
              theme(axis.text=element_text(size=8,color='black'),
                    axis.title=element_text(size=10,color='black')) + theme(plot.title = element_text(size = 14, vjust=1)) +
              theme(strip.text.x = element_text(size = 0)) #remove facet label
            
            #-----------#
            #print out eigengene plots
            png(
              paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/apriori_eigengenes/eigengene_overlaps/", names(eigengenes_toanalyze)[f],"/Eigengene_Top_Overlapping_Genes_N",nrow(marker_df) ,".png",sep=""),
              width=14,height=7,units="in",res=300)
            
            print(multiplot(agexp_plot, agexp_plot_pap_ca,
                            cols=2))
            dev.off()
            
          },  error=function(e){})
          
        }))
        ################################################################################################################################    
        ################################################################################################################################    
        ################################################################################################################################
        ################################################################################################################################    
        ################################################################################################################################    
        ################################################################################################################################
        ################################################################################################################################    
        ################################################################################################################################    
        ################################################################################################################################
        ################################################################################################################################      
        ################################################################################################################################    
        ################################################################################################################################    
        ################################################################################################################################
        ################################################################################################################################    
        ################################################################################################################################    
        ################################################################################################################################
        ################################################################################################################################    
        ################################################################################################################################    
        ################################################################################################################################
        ################################################################################################################################      
        ################################################################################################################################    
        ################################################################################################################################    
        ################################################################################################################################
        ################################################################################################################################    
        ################################################################################################################################    
        ################################################################################################################################
        ################################################################################################################################    
        ################################################################################################################################    
        ################################################################################################################################
        ################################################################################################################################      
        ################################################################################################################################    
        ################################################################################################################################    
        ################################################################################################################################
        ################################################################################################################################    
        ################################################################################################################################    
        ################################################################################################################################
        ################################################################################################################################    
        ################################################################################################################################    
        ################################################################################################################################
        ################################################################################################################################      
        
        #Combine the new tumor samples data with this.
        
        tumor_cds = readRDS("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/chemotherapy_tumor_project/data/cds/chemotherapy_tumor.RDS")
        tumor_cds = tumor_cds[,colData(tumor_cds)$chemo_tx %in% c("Control","Cisplatin")] #subset to control and cisplatin samples only
        #Replace Sample_Number with a correct sample identifier
        
        #add columns to tumor_cds metadata so they line up with big_cds metadata
        colData(tumor_cds)$tissue_updated = colData(tumor_cds)$chemo_tx
        colData(tumor_cds)$project = "chemo_project"
        colData(big_cds)$project = "original_project"
        
        #combine big_cds with new chemotherapy tumor experiment samples
        tumor_big_cds <- combine_cds(list(big_cds,tumor_cds))
        
        #3b. Pre-process and control for project variation
        tumor_big_cds <- preprocess_cds(tumor_big_cds, 
                                        method="PCA",
                                        num_dim = 100,
                                        norm_method = "none")
        
        #4. A further round of (nonlinear) dimensionality reduction using UMAP ?reduce_dimension
        #4b.
        tumor_big_cds <- reduce_dimension(
          tumor_big_cds,
          max_components = 2,
          reduction_method = c("UMAP"),
          preprocess_method = "PCA",
          umap.metric = "cosine",
          umap.min_dist = 0.1,
          umap.n_neighbors = 15L,
          umap.fast_sgd = FALSE,
          umap.nn_method = "annoy",
          cores = 1,
          verbose = FALSE
        )
        
        #5. Cluster and partition ?cluster_cells cells into supergroups community detection (Phenograph algorithm) Community detection of cells as part of Phenograph algorithm https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4508757/
        #5b. 
        tumor_big_cds <- cluster_cells(tumor_big_cds,
                                       reduction_method = c("UMAP"),
                                       k = 10, #a bigger k will result in lower resolution 
                                       cluster_method = c("leiden"),
                                       num_iter = 2,
                                       partition_qval = 0.05, #Numeric, the q-value cutoff to determine when to partition
                                       weight = FALSE,
                                       resolution = NULL,
                                       random_seed = NULL,
                                       verbose = F)
        
        
        #-------------------------------------------------------------------------------------------#
        #-------------------------------------------------------------------------------------------#
        #Plot all cells and samples
        #plot tissue identity manifold
        unique(colData(tumor_big_cds)$tissue_updated)
        plot_cells(tumor_big_cds, color_cells_by="tissue_updated", group_cells_by="cluster",
                   group_label_size = 0,
                   x=1,
                   y=2) +
          ggtitle("Tissue: Dim. 1 and 2") + xlab("Dim. 1") + ylab("Dim. 2") +
          theme(text = element_text(size=18)) + theme(axis.text=element_text(size=18,color='black'),
                                                      axis.title=element_text(size=20,color='black')) + theme(plot.title = element_text(size = 22, vjust=1)) + theme(axis.text=element_text(size=15,color='black'),
                                                                                                                                                                     axis.title=element_text(size=15,color='black'))+
          scale_color_manual(values = c("Normal skin" = "darkblue","Pap" = "purple", 
                                        "CA" = "darkred",
                                        "Control" = "orange",
                                        "Cisplatin" = "goldenrod1")) + theme(legend.position="bottom")
        
        #all cells eigengene
        agexp_plot_tumor_big = plot_cells(tumor_big_cds,
                                          genes=marker_df,
                                          label_cell_groups=FALSE,
                                          label_leaves=F,
                                          label_branch_points=F,
                                          graph_label_size=1.5,
                                          show_trajectory_graph=F,
                                          cell_size = 0.5) +
          xlab("Dim. 1") + ylab("Dim. 2") + 
          theme(legend.position="bottom") +
          scale_color_viridis_c(option = "inferno",name="integrated expression") +
          ggtitle(paste("Stem cell eigengene"))+ #continuous
          theme(axis.text=element_text(size=8,color='black'),
                axis.title=element_text(size=10,color='black')) + theme(plot.title = element_text(size = 14, vjust=1)) +
          theme(strip.text.x = element_text(size = 0)) #remove facet label
        
        png("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/stem_cell_mansucript/Figs/Fig1/Fig1_supervised_stem_eigengene_tumors.png",
            width=8,height=8,units="in",res=400)
        agexp_plot_tumor_big
        dev.off()
        
        #-------------------------------------------------------------------------------------------#
        #-------------------------------------------------------------------------------------------#
        #Plot carcinomas only
        unique(colData(tumor_big_cds)$tissue_updated)
        carcinoma_samples = tumor_big_cds[,(colData(tumor_big_cds)$tissue_updated %in% c("CA","Control","Cisplatin"))] #subset to all the carcinoma samples
        
        #3b. Pre-process and control for project variation
        ?preprocess_cds
        carcinoma_samples <- preprocess_cds(carcinoma_samples, 
                                            method="PCA",
                                            num_dim = 100,
                                            norm_method = "none")
        
        #4. A further round of (nonlinear) dimensionality reduction using UMAP ?reduce_dimension
        #4b.
        carcinoma_samples <- reduce_dimension(
          carcinoma_samples,
          max_components = 2,
          reduction_method = c("UMAP"),
          preprocess_method = "PCA",
          umap.metric = "cosine",
          umap.min_dist = 0.6,
          umap.n_neighbors = 15L,
          umap.fast_sgd = FALSE,
          umap.nn_method = "annoy",
          cores = 1,
          verbose = FALSE
        )
        
        #5. Cluster and partition ?cluster_cells cells into supergroups community detection (Phenograph algorithm) Community detection of cells as part of Phenograph algorithm https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4508757/
        #5b. 
        carcinoma_samples <- cluster_cells(carcinoma_samples,
                                           reduction_method = c("UMAP"),
                                           k = 10, #a bigger k will result in lower resolution 
                                           cluster_method = c("leiden"),
                                           num_iter = 2,
                                           partition_qval = 0.05, #Numeric, the q-value cutoff to determine when to partition
                                           weight = FALSE,
                                           resolution = NULL,
                                           random_seed = NULL,
                                           verbose = F)
        
        #plot treatment group
        plot_cells(carcinoma_samples, color_cells_by="tissue_updated", group_cells_by="cluster",
                   group_label_size = 0,
                   x=1,
                   y=2,
                   cell_size=0.5) +
          ggtitle("Treatment group") + xlab("Dim. 1") + ylab("Dim. 2") +
          theme(text = element_text(size=18)) + theme(axis.text=element_text(size=18,color='black'),
                                                      axis.title=element_text(size=20,color='black')) + theme(plot.title = element_text(size = 22, vjust=1)) + theme(axis.text=element_text(size=15,color='black'),
                                                                                                                                                                     axis.title=element_text(size=15,color='black'))+
          scale_color_manual(values = c("Normal skin" = "darkblue","Pap" = "purple", 
                                        "CA" = "darkred",
                                        "Control" = "orange",
                                        "Cisplatin" = "navajowhite3")) + theme(legend.position="right")
        
        
        #add the original CA sample name to he Sample number column
        colData(carcinoma_samples)[, 17][is.na(colData(carcinoma_samples)[, 17])] <- "Original CA"
        names(colData(carcinoma_samples))
        
        plot_cells(carcinoma_samples, color_cells_by="sample_name", group_cells_by="cluster",
                   group_label_size = 0,
                   x=1,
                   y=2,
                   cell_size=0.5) +
          ggtitle("Samples") + xlab("Dim. 1") + ylab("Dim. 2") +
          theme(text = element_text(size=18)) + theme(axis.text=element_text(size=18,color='black'),
                                                      axis.title=element_text(size=20,color='black')) + theme(plot.title = element_text(size = 22, vjust=1)) + theme(axis.text=element_text(size=15,color='black'),
                                                                                                                                                                     axis.title=element_text(size=15,color='black'))+
          theme(legend.position="right")
        
        #plot p16 status
        plot_cells(carcinoma_samples,
                   genes="Cdkn2a",
                   label_cell_groups=FALSE,
                   label_leaves=F,
                   label_branch_points=F,
                   graph_label_size=1.5,
                   show_trajectory_graph=F,
                   cell_size = 0.5) +
          xlab("Dim. 1") + ylab("Dim. 2") + 
          theme(legend.position="bottom") +
          scale_color_viridis_c(option = "inferno",name="integrated\nexpression") +
          ggtitle(paste("p16 expression"))+ #continuous
          theme(axis.text=element_text(size=8,color='black'),
                axis.title=element_text(size=10,color='black')) + theme(plot.title = element_text(size = 14, vjust=1)) +
          theme(strip.text.x = element_text(size = 0)) + #remove facet label
          theme(legend.position="right")
        
        #carcinoma samples eigengene
        agexp_plot_tumor_big = plot_cells(carcinoma_samples,
                                          genes=marker_df,
                                          label_cell_groups=FALSE,
                                          label_leaves=F,
                                          label_branch_points=F,
                                          graph_label_size=1.5,
                                          show_trajectory_graph=F,
                                          cell_size = 0.5) +
          xlab("Dim. 1") + ylab("Dim. 2") + 
          theme(legend.position="bottom") +
          scale_color_viridis_c(option = "inferno",name="integrated\nexpression") +
          ggtitle(paste("Stem cell eigengene"))+ #continuous
          theme(axis.text=element_text(size=8,color='black'),
                axis.title=element_text(size=10,color='black')) + theme(plot.title = element_text(size = 14, vjust=1)) +
          theme(strip.text.x = element_text(size = 0)) + #remove facet label
          theme(legend.position="right")
        
        
        ################################################################################################################################
        ################################################################################################################################      
        #compare eigengene expression as eigencell bar plots
        
        head(colData(tumor_big_cds))
        parenchymas = unique(colData(tumor_big_cds)$tissue_updated)
        parenchyma_proportion = data.frame(table(colData(tumor_big_cds)$tissue_updated) / dim(tumor_big_cds)[2])
        
        stem_cell_parenchymas = do.call(rbind, lapply(1:length(parenchymas), function(i){
          print(i)
          
          #i=1
          #1. Stage
          sub_cds_1 = tumor_big_cds[,(colData(tumor_big_cds)$tissue_updated %in% parenchymas[i])] #subset to a specific stage AND  Lgr6+
          agg_mat_sub_lgr6 = aggregate_gene_expression(
            sub_cds_1,
            gene_group_df = marker_df,
            cell_group_df = NULL,
            norm_method = c("log"),
            pseudocount = 1,
            scale_agg_values = TRUE,
            #max_agg_value = 5,
            min_agg_value = 0,
            exclude.na = TRUE
          )
          agg_mat_sub_lgr6 = data.frame(t(agg_mat_sub_lgr6))
          weighted_eigencell = colSums(agg_mat_sub_lgr6) * subset(parenchyma_proportion, Var1==parenchymas[i])$Freq #get sum and multiply by proportion of that stage's cells in the total stage
          
          agg_mat_sub_lgr6 = agg_mat_sub_lgr6[agg_mat_sub_lgr6[,1]!=0,] #remove 0's
          agg_mat_sub_lgr6 = agg_mat_sub_lgr6[agg_mat_sub_lgr6>quantile(agg_mat_sub_lgr6,0.5)] #remove all cells underneath the 50th percentile
          mean_non_zero_expression = mean(agg_mat_sub_lgr6) #get mean for non-zeros above a quantile threshold
          
          agg_mat_sub_lgr6 = data.frame(parenchymas[i],weighted_eigencell,mean_non_zero_expression)
          names(agg_mat_sub_lgr6)[1:2] = c("parenchyma","weighted_eigencell")
          agg_mat_sub_lgr6
        }))
        
        stem_cell_parenchymas$scaled_weighted_eigencell = scale(stem_cell_parenchymas$weighted_eigencell, center=F)
        
        ggplot(stem_cell_parenchymas, aes(x=parenchyma,y=(mean_non_zero_expression))) +
          geom_bar(stat = "identity", fill="darkorange") +
          #facet_wrap(~parenchyma, scales="free", nrow=1) + 
          #geom_jitter(width = 0.1) +
          ggtitle(paste("Stem cell eigengene")) + 
          xlab("") + ylab("mean non-zero stem cell eigengene expression") + 
          theme(axis.text.x=element_text(size=15,color='black'),
                axis.text.y=element_text(size=10,color='black'),
                axis.title=element_text(size=15),
                axis.ticks.x = element_line(size=0),
                strip.text.x = element_text(size = 15),
                legend.position="none",
                title = element_text(size = 15, face="bold")) +
          theme(
            # Hide panel borders and remove grid lines
            panel.border = element_blank(),
            panel.grid.major = element_blank(),
            panel.grid.minor = element_blank(),
            panel.background = element_blank(),
            element_line(colour = 'black', size = 0.3),
            # Change axis line
            axis.line = element_line(colour = "black"),
            strip.background = element_rect(colour="white",
                                            fill="white")
          )
        
        ################################################################################################################################
        ################################################################################################################################      
        #compare eigengene expression as violin plots
        
        head(colData(tumor_big_cds))
        parenchymas = unique(colData(tumor_big_cds)$tissue_updated)
        parenchyma_proportion = data.frame(table(colData(tumor_big_cds)$tissue_updated) / dim(tumor_big_cds)[2])
        
        stem_cell_parenchymas = do.call(rbind, lapply(1:length(parenchymas), function(i){
          print(i)
          
          #i=1
          #1. Stage
          sub_cds_1 = tumor_big_cds[,(colData(tumor_big_cds)$tissue_updated %in% parenchymas[i])] #subset to a specific stage AND  Lgr6+
          agg_mat_sub_lgr6 = aggregate_gene_expression(
            sub_cds_1,
            gene_group_df = marker_df,
            cell_group_df = NULL,
            norm_method = c("log"),
            pseudocount = 1,
            scale_agg_values = TRUE,
            #max_agg_value = 5,
            min_agg_value = 0,
            exclude.na = TRUE
          )
          agg_mat_sub_lgr6 = data.frame(t(agg_mat_sub_lgr6))
          agg_mat_sub_lgr6 = agg_mat_sub_lgr6[agg_mat_sub_lgr6[,1]!=0,] #remove 0's
          agg_mat_sub_lgr6 = data.frame(parenchymas[i],agg_mat_sub_lgr6)
          names(agg_mat_sub_lgr6)[1:2] = c("parenchyma","eigengene_expression")
          agg_mat_sub_lgr6
        }))
        
        stem_cell_parenchymas$scaled_weighted_eigencell = scale(stem_cell_parenchymas$weighted_eigencell, center=F)
        
        my_comparisons <- list( c("Normal skin", "Pap"), c("Pap", "CA"), c("Normal skin", "CA"), c("Control", "Cisplatin") )
        head(stem_cell_parenchymas)
        ggplot(stem_cell_parenchymas, aes(x=parenchyma,y=eigengene_expression, fill=parenchyma)) +
          stat_compare_means(comparisons = my_comparisons) +
          geom_violin(aes(fill=parenchyma), width=1) +
          geom_boxplot(outlier.size=0, width=0.2, color="white") +
          #geom_bar(stat = "identity", fill="darkorange") +
          #facet_wrap(~parenchyma, scales="free", nrow=1) + 
          #geom_jitter(width = 0.1) +
          ggtitle(paste("Stem cell eigengene")) + 
          xlab("") + ylab("mean non-zero stem cell eigengene expression") + 
          theme(axis.text.x=element_text(size=15,color='black'),
                axis.text.y=element_text(size=10,color='black'),
                axis.title=element_text(size=15),
                axis.ticks.x = element_line(size=0),
                strip.text.x = element_text(size = 15),
                legend.position="none",
                title = element_text(size = 15, face="bold")) +
          theme(
            # Hide panel borders and remove grid lines
            panel.border = element_blank(),
            panel.grid.major = element_blank(),
            panel.grid.minor = element_blank(),
            panel.background = element_blank(),
            element_line(colour = 'black', size = 0.3),
            # Change axis line
            axis.line = element_line(colour = "black"),
            strip.background = element_rect(colour="white",
                                            fill="white")
          ) +
          scale_fill_manual(values = c("Normal skin" = "darkblue","Pap" = "purple", 
                                       "CA" = "darkred",
                                       "Control" = "orange",
                                       "Cisplatin" = "goldenrod1")) + theme(legend.position="none")
        
        ################################################################################################################################
        ################################################################################################################################      
        ################################################################################################################################
        ################################################################################################################################      
        ################################################################################################################################
        ################################################################################################################################      
        #Cell Cycle calling with cyclone
        
        cell_cycle_comb = read.table(file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/cell_cycle/cell_cycle_results.txt",header=T)
        
        #subset to cells in my CDS object
        big_cds_cell_cycle = cell_cycle_comb[na.omit(match(colnames(big_cds), cell_cycle_comb$cell_id)),]
        nrow(big_cds_cell_cycle) == length(colnames(big_cds))
        
        #Add cell cycle data to metadata of CDS object
        head(big_cds_cell_cycle)
        colData(big_cds)$cell_cycle = big_cds_cell_cycle$cellcycle_phase
        colData(big_cds)$G1 = big_cds_cell_cycle$G1
        colData(big_cds)$S = big_cds_cell_cycle$S
        colData(big_cds)$G2M = big_cds_cell_cycle$G2M
        
        
        #get cell cycle  percentages of each tissue in each stage
        cycle_counts = data.frame(table(colData(big_cds)$tissue_updated, colData(big_cds)$cell_cycle))
        cycle_counts$freq_percent = round((cycle_counts$Freq / c(table(colData(big_cds)$tissue_updated),table(colData(big_cds)$tissue_updated),table(colData(big_cds)$tissue_updated)))*100, 2)
        names(cycle_counts)[1:2] = c("tissue","cell cycle")
        cycle_counts
        write.table(cycle_counts, file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/cell_cycle/cell_cyle_phase.txt", row.names=F,quote=F, sep="\t")
        
        #get cell cycle  percentages of cell types
        cycle_counts = data.frame(table(colData(big_cds)$cell_type, colData(big_cds)$cell_cycle))
        cycle_counts$freq_percent = round((cycle_counts$Freq / rep(c(table(colData(big_cds)$cell_type)),3) )*100, 2)
        names(cycle_counts)[1:2] = c("cell type","cell cycle")
        write.table(cycle_counts, file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/cell_cycle/cell_cycle_celltypes.txt", row.names=F,quote=F, sep="\t")
        
        #get cell cycle percentages of parenchyma tissues
        #remove the extra "_NUMBER" at the end of cell id's that were added during the cds combo function
        split_names = data.frame(strsplit(colnames(cds_comb), split="_")) #split on Monocle's sample delimiter for cell id's "_"
        restored_cell_ids = do.call(rbind, lapply(1:ncol(split_names), function(i) {
          #i=1
          restored_cell_id = paste(split_names[1,i],"_",split_names[2,i],sep="")
          restored_cell_id
        }))
        colnames(cds_comb) = as.vector(restored_cell_ids) #add restored cell ids to the cds_comb object
        big_cds_cell_cycle = cell_cycle_comb[match(colnames(cds_comb), cell_cycle_comb$cell_id),]
        colData(cds_comb)$cell_cycle = big_cds_cell_cycle$cellcycle_phase
        cycle_counts = data.frame(table(colData(cds_comb)$parenchyma_tissue, colData(cds_comb)$cell_cycle))
        cycle_counts$freq_percent = round((cycle_counts$Freq / rep(c(table(colData(cds_comb)$parenchyma_tissue)),3) )*100, 2)
        names(cycle_counts)[1:2] = c("parenchyma","cell cycle")
        write.table(cycle_counts, file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/cell_cycle/cell_cycle_parenchyma.txt", row.names=F,quote=F, sep="\t")
        
        
        #Manifold cell cycle
        plot_cells(big_cds, color_cells_by="cell_cycle", group_cells_by="cluster",
                   group_label_size = 0,
                   x=1,
                   y=2,
                   cell_size=0.5,
                   alpha=0.5) +
          ggtitle("Cell cycle") + xlab("Dim. 1") + ylab("Dim. 2") +
          theme(text = element_text(size=18)) + theme(axis.text=element_text(size=18,color='black'),
                                                      axis.title=element_text(size=20,color='black')) + theme(plot.title = element_text(size = 22, vjust=1)) + theme(axis.text=element_text(size=15,color='black'),
                                                                                                                                                                     axis.title=element_text(size=15,color='black'))+
          theme(legend.position="right") +
          scale_color_manual(values = c("G1" = "tan",
                                        "G2M" = "darkred", 
                                        "S" = "darkblue",
                                        "NA" = "white"))
        
        ################################################################################################################################
        ################################################################################################################################      
        ################################################################################################################################
        ################################################################################################################################      
        ################################################################################################################################
        ################################################################################################################################      
        ################################################################################################################################
        ################################################################################################################################      
        ###############################################################################################################################      
        ################################################################################################################################
        ################################################################################################################################      
        ################################################################################################################################
        ################################################################################################################################  
        #CellChat : https://github.com/sqjin/CellChat
        #------------------------------------------------------------------------------------------------#
        #Install 
        # install.packages('NMF')
        # Sys.setenv(R_REMOTES_NO_ERRORS_FROM_WARNINGS=TRUE)
        # devtools::install_github("jokergoo/circlize")
        # BiocManager::install("ComplexHeatmap")
        # devtools::install_github("jokergoo/ComplexHeatmap") #this is not working
        # devtools::install_github("sqjin/CellChat")
        
        #Start CellChat analysis
        library(CellChat)
        library(ggplot2)                  
        library(patchwork)
        library(ggalluvial)
        library(igraph)
        library(dplyr)
        options(stringsAsFactors = FALSE)
        
        #------------------------------------------------------------------------------------------------#
        #make master output dir
        outdir_1 = "/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/CellChat/results_v2/"
        
        #------------------------------------------------------------------------------------------------#
        #make factors within the metadata to loop through (called tissue_sources) -> factor cells, will infer communication only within cells
        tissue_sources = unique(colData(big_cds)$tissue_updated)
        
        
        #------------------------------------------------------------------------------------------------#
        #Start CellChat loop through classification schemes of cell populations that you will infer communication among cell populations for
        variable_factors = c("integrated_type_bridge","integrated_celltype_spike")
        
        for(j in 1:length(variable_factors)) {
          
          tryCatch({ #suppress error
            #j=3
            
            #------------------------------------------------------------------------------------------------#
            #Start CellChat loop thru tissue sources
            for(i in 2:length(tissue_sources)) {
              tryCatch({ #suppress error
                
                #i=3
                print(paste("j=",j," i=",i,sep=""))
                
                # get scRNAseq data from CDS object
                data.input = big_cds@assays@data$counts # normalized data matrix
                meta = colData(big_cds) # a dataframe with rownames containing cell mata data
                
                #subset to specific cells
                if(tissue_sources[i] == "Normal skin") {  cell.use = rownames(meta)[meta$tissue_updated == tissue_sources[i] & meta$cell_type!="nEp" & meta$cell_type!=""  & meta$cell_type!="MyFi"] } else # extract the cell names from disease data & Exclude cells that I cannot annotate as a particular cell type
                {cell.use = rownames(meta)[meta$tissue_updated == tissue_sources[i] & meta$cell_type!=""  & meta$cell_type!="MyFi"] }# extract the cell names from disease data & Exclude cells that I cannot annotate as a particular cell type
                print(paste("Number of cells = ",length(cell.use),sep=""))
                
                # Prepare input data for CelChat analysis
                data.input = data.input[, cell.use]
                meta = data.frame(meta[cell.use, ])
                unique(meta$cell_type)
                
                # Create CellChat Object
                cellchat <- createCellChat(object = data.input, meta = meta, group.by = variable_factors[j]) #choose which cell factors to group by
                
                # Set which interactome database to mine
                CellChatDB <- CellChatDB.mouse # use CellChatDB.human if running on mouse data
                CellChatDB.use <- CellChatDB # simply use the default CellChatDB
                # set the used database in the object
                cellchat@DB <- CellChatDB.use  
                
                cellchat <- subsetData(cellchat) # subset the expression data of signaling genes for saving computation cost
                #future::plan("multiprocess", workers = 4) # do parallel
                
                #Preprocessing the expression data for cell-cell communication analysis
                cellchat <- identifyOverExpressedGenes(cellchat)
                cellchat <- identifyOverExpressedInteractions(cellchat)
                cellchat <- projectData(cellchat, PPI.mouse)
                
                #Compute the communication probability and infer cellular communication network
                cellchat <- computeCommunProb(cellchat, raw.use = TRUE, population.size = TRUE)
                
                # Filter out the cell-cell communication if there are only few number of cells in certain cell groups
                cellchat <- filterCommunication(cellchat, min.cells = 10)
                
                # Create output directory
                outdir = paste(outdir_1,tissue_sources[i],"_",variable_factors[j],sep="")
                dir.create(outdir)
                
                #write out interaction results
                df.net <- subsetCommunication(cellchat)
                write.table(df.net, file=paste(outdir,"/interaction_results_cell_type.txt",sep=""), quote=F,row.names=F,sep="\t")
                
                #Infer the cell-cell communication at a signaling pathway level
                cellchat <- computeCommunProbPathway(cellchat)
                
                #Calculate the aggregated cell-cell communication network
                cellchat <- aggregateNet(cellchat)
                
                #Visualize these baddies
                groupSize <- as.numeric(table(cellchat@idents))
                
                png(paste(outdir,"/aggregated_cell_cell_communication_network_NumInt_cell_type.png",sep=""),width=7,height=7,units="in",res=200)
                print( netVisual_circle(cellchat@net$count, vertex.weight = groupSize, weight.scale = T, label.edge= F, title.name = paste(tissue_sources[i], "Number of interactions")) )
                dev.off()
                
                png(paste(outdir,"/aggregated_cell_cell_communication_network_IntWeight_cell_type.png",sep=""),width=7,height=7,units="in",res=200)
                print( netVisual_circle(cellchat@net$weight, vertex.weight = groupSize, weight.scale = T, label.edge= F, title.name = paste(tissue_sources[i], "Interaction weights/strengths")) )
                dev.off()
                
                # Print out individual graphs
                mat <- cellchat@net$weight
                par(mfrow = c(3,4), xpd=TRUE)
                pdf(file=paste(outdir,"/individual_cell_type_interactions.pdf",sep=""),height=7,width=7)
                for (i in 1:nrow(mat)) {
                  #i=1
                  mat2 <- matrix(0, nrow = nrow(mat), ncol = ncol(mat), dimnames = dimnames(mat))
                  mat2[i, ] <- mat[i, ]
                  print(netVisual_circle(mat2, vertex.weight = groupSize, weight.scale = T, edge.weight.max = max(mat), title.name = rownames(mat)[i]) )
                }
                dev.off()
                
                # show all the significant interactions (L-R pairs) from some cell groups (defined by 'sources.use') to other cell groups (defined by 'targets.use')
                # netVisual_bubble(cellchat, sources.use = 4, targets.use = c(5:11), remove.isolate = FALSE)
                png(paste(outdir,"/bubble_plot_communication_prob.png",sep=""),width=15,height=30,units="in",res=200)
                netVisual_bubble(cellchat) + theme(axis.text.x = element_text(size=12))
                dev.off()
                
                #plotGeneExpression(cellchat, signaling = "CXCL")
                
                # Compute the network centrality scores
                cellchat <- netAnalysis_computeCentrality(cellchat, slot.name = "netP") # the slot 'netP' means the inferred intercellular communication network of signaling pathways
                
                #Visualize the dominant senders (sources) and receivers (targets) in a 2D space
                # Signaling role analysis on the aggregated cell-cell communication network from all signaling pathways
                png(paste(outdir,"/interaction_summary.png",sep=""),width=6,height=6,units="in",res=200)
                print( gg1 <- netAnalysis_signalingRole_scatter(cellchat) + ggtitle("Dominant senders and sources") +
                         theme(axis.text= element_text(size=14),
                               axis.title = element_text(size=14),
                               plot.title= element_text(size=16))
                       
                ) 
                dev.off()
                
                #Identify signals contributing most to outgoing or incoming signaling of certain cell groups  
                # Signaling role analysis on the aggregated cell-cell communication network from all signaling pathways
                
                png(paste(outdir,"/outgoing_signals.png",sep=""),width=6,height=10,units="in",res=200)
                tryCatch({ #suppress error
                  print( netAnalysis_signalingRole_heatmap(cellchat, pattern = "outgoing") )
                },  error=function(e){})
                dev.off()
                
                png(paste(outdir,"/incoming_signals.png",sep=""),width=10,height=10,units="in",res=200)
                tryCatch({ #suppress error
                  print( netAnalysis_signalingRole_heatmap(cellchat, pattern = "incoming") )
                },  error=function(e){})
                dev.off()
                
                
                #Identify and visualize outgoing communication pattern of secreting cells
                
                #Incoming communication patterns
                #selectK(cellchat, pattern = "outgoing")
                nPatterns = 5
                cellchat <- identifyCommunicationPatterns(cellchat, pattern = "incoming", k = nPatterns)
                png(paste(outdir,"/incoming_patterns.png",sep=""),width=10,height=10,units="in",res=200)
                netAnalysis_river(cellchat, pattern = "incoming")
                dev.off()
                
                png(paste(outdir,"/incoming_contribution_patterns.png",sep=""),width=20,height=5,units="in",res=200)
                netAnalysis_dot(cellchat, pattern = "incoming") +
                  theme(axis.text= element_text(size=14),
                        axis.title = element_text(size=14),
                        plot.title= element_text(size=16))
                dev.off()
                
                #Outgoing communication patterns
                #selectK(cellchat, pattern = "outgoing")
                nPatterns = 5
                cellchat <- identifyCommunicationPatterns(cellchat, pattern = "outgoing", k = nPatterns)
                png(paste(outdir,"/outgoing_patterns.png",sep=""),width=10,height=10,units="in",res=200)
                netAnalysis_river(cellchat, pattern = "outgoing")
                dev.off()
                
                png(paste(outdir,"/outgoing_contribution_patterns.png",sep=""),width=20,height=5,units="in",res=200)
                netAnalysis_dot(cellchat, pattern = "outgoing") +
                  theme(axis.text= element_text(size=14),
                        axis.title = element_text(size=14),
                        plot.title= element_text(size=16))
                dev.off()
                
              },  error=function(e){})
              
            } #end tissue loop
            
          },  error=function(e){})
          
        } #end variable factor loop
        
        #-----------------------------------------------------------------------------------------------------------------------------------#
        #-----------------------------------------------------------------------------------------------------------------------------------#
        #Interaction in the pap_ca parenchyma
        #Start CellChat analysis
        library(CellChat)
        library(ggplot2)                  
        library(patchwork)
        library(ggalluvial)
        library(igraph)
        library(dplyr)
        options(stringsAsFactors = FALSE)
        
        unique(colData(big_cds)$cell_type)
        pap_ca_immune = big_cds[,(colData(big_cds)$tissue_updated %in% c("Pap","CA") & colData(big_cds)$cell_type %in% c("nEp","T","LC","Mac","Neut"))]
        unique(colData(pap_ca_immune)$integrated_type_bridge)
        
        #------------------------------------------------------------------------------------------------#
        #make master output dir
        outdir_1 = "/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/CellChat/results_v3/"
        
        #------------------------------------------------------------------------------------------------#
        #Start CellChat loop through classification schemes of cell populations that you will infer communication among cell populations for
        
        # get scRNAseq data from CDS object
        data.input = pap_ca_immune@assays@data$counts # normalized data matrix
        meta = colData(pap_ca_immune) # a dataframe with rownames containing cell mata data
        
        #subset to specific cells
        cell.use = rownames(meta)
        print(paste("Number of cells = ",length(cell.use),sep=""))
        
        # Prepare input data for CelChat analysis
        data.input = data.input[, cell.use]
        meta = data.frame(meta[cell.use, ])
        unique(meta$cell_type)
        
        # Create CellChat Object
        cellchat <- createCellChat(object = data.input, meta = meta, group.by = "integrated_type_bridge") #choose which cell factors to group by
        
        # Set which interactome database to mine
        CellChatDB <- CellChatDB.mouse # use CellChatDB.human if running on mouse data
        CellChatDB.use <- CellChatDB # simply use the default CellChatDB
        # set the used database in the object
        cellchat@DB <- CellChatDB.use  
        
        cellchat <- subsetData(cellchat) # subset the expression data of signaling genes for saving computation cost
        #future::plan("multiprocess", workers = 4) # do parallel
        
        #Preprocessing the expression data for cell-cell communication analysis
        cellchat <- identifyOverExpressedGenes(cellchat)
        cellchat <- identifyOverExpressedInteractions(cellchat)
        cellchat <- projectData(cellchat, PPI.mouse)
        
        #Compute the communication probability and infer cellular communication network
        cellchat <- computeCommunProb(cellchat, raw.use = TRUE, population.size = TRUE)
        
        # Filter out the cell-cell communication if there are only few number of cells in certain cell groups
        cellchat <- filterCommunication(cellchat, min.cells = 10)
        
        # Create output directory
        outdir = paste(outdir_1,"pap_ca_immune",sep="")
        dir.create(outdir)
        
        #write out interaction results
        df.net <- subsetCommunication(cellchat)
        write.table(df.net, file=paste(outdir,"/interaction_results_cell_type.txt",sep=""), quote=F,row.names=F,sep="\t")
        
        #Infer the cell-cell communication at a signaling pathway level
        cellchat <- computeCommunProbPathway(cellchat)
        
        #Calculate the aggregated cell-cell communication network
        cellchat <- aggregateNet(cellchat)
        
        #Visualize these baddies
        groupSize <- as.numeric(table(cellchat@idents))
        
        png(paste(outdir,"/aggregated_cell_cell_communication_network_NumInt_cell_type.png",sep=""),width=7,height=7,units="in",res=200)
        print( netVisual_circle(cellchat@net$count, vertex.weight = groupSize, weight.scale = T, label.edge= F, title.name = paste(tissue_sources[i], "Number of interactions")) )
        dev.off()
        
        png(paste(outdir,"/aggregated_cell_cell_communication_network_IntWeight_cell_type.png",sep=""),width=7,height=7,units="in",res=200)
        print( netVisual_circle(cellchat@net$weight, vertex.weight = groupSize, weight.scale = T, label.edge= F, title.name = paste(tissue_sources[i], "Interaction weights/strengths")) )
        dev.off()
        
        # Print out individual graphs
        mat <- cellchat@net$weight
        par(mfrow = c(3,4), xpd=TRUE)
        pdf(file=paste(outdir,"/individual_cell_type_interactions.pdf",sep=""),height=7,width=7)
        for (i in 1:nrow(mat)) {
          #i=1
          mat2 <- matrix(0, nrow = nrow(mat), ncol = ncol(mat), dimnames = dimnames(mat))
          mat2[i, ] <- mat[i, ]
          print(netVisual_circle(mat2, vertex.weight = groupSize, weight.scale = T, edge.weight.max = max(mat), title.name = rownames(mat)[i]) )
        }
        dev.off()
        
        # show all the significant interactions (L-R pairs) from some cell groups (defined by 'sources.use') to other cell groups (defined by 'targets.use')
        # netVisual_bubble(cellchat, sources.use = 4, targets.use = c(5:11), remove.isolate = FALSE)
        png(paste(outdir,"/bubble_plot_communication_prob.png",sep=""),width=20,height=30,units="in",res=200)
        netVisual_bubble(cellchat) + theme(axis.text.x = element_text(size=12))
        dev.off()
        
        #plotGeneExpression(cellchat, signaling = "CXCL")
        
        # Compute the network centrality scores
        cellchat <- netAnalysis_computeCentrality(cellchat, slot.name = "netP") # the slot 'netP' means the inferred intercellular communication network of signaling pathways
        
        #Visualize the dominant senders (sources) and receivers (targets) in a 2D space
        # Signaling role analysis on the aggregated cell-cell communication network from all signaling pathways
        png(paste(outdir,"/interaction_summary.png",sep=""),width=6,height=6,units="in",res=200)
        print( gg1 <- netAnalysis_signalingRole_scatter(cellchat) + ggtitle("Dominant senders and sources") +
                 theme(axis.text= element_text(size=14),
                       axis.title = element_text(size=14),
                       plot.title= element_text(size=16))
               
        ) 
        dev.off()
        
        #Identify signals contributing most to outgoing or incoming signaling of certain cell groups  
        # Signaling role analysis on the aggregated cell-cell communication network from all signaling pathways
        
        png(paste(outdir,"/outgoing_signals.png",sep=""),width=6,height=20,units="in",res=200)
        tryCatch({ #suppress error
          print( netAnalysis_signalingRole_heatmap(cellchat, pattern = "outgoing") )
        },  error=function(e){})
        dev.off()
        
        png(paste(outdir,"/incoming_signals.png",sep=""),width=5,height=7.5,units="in",res=200)
        tryCatch({ #suppress error
          print( netAnalysis_signalingRole_heatmap(cellchat, pattern = "all", height=15,width=7,
                                                   cluster.rows=F,
                                                   cluster.cols=F) )
        },  error=function(e){})
        dev.off()
        
        #Identify and visualize outgoing communication pattern of secreting cells
        
        #Incoming communication patterns
        #selectK(cellchat, pattern = "outgoing")
        nPatterns = 5
        cellchat <- identifyCommunicationPatterns(cellchat, pattern = "incoming", k = nPatterns)
        png(paste(outdir,"/incoming_patterns.png",sep=""),width=10,height=10,units="in",res=200)
        netAnalysis_river(cellchat, pattern = "incoming")
        dev.off()
        
        png(paste(outdir,"/incoming_contribution_patterns.png",sep=""),width=20,height=5,units="in",res=200)
        netAnalysis_dot(cellchat, pattern = "incoming") +
          theme(axis.text= element_text(size=14),
                axis.title = element_text(size=14),
                plot.title= element_text(size=16))
        dev.off()
        
        #Outgoing communication patterns
        #selectK(cellchat, pattern = "outgoing")
        nPatterns = 5
        cellchat <- identifyCommunicationPatterns(cellchat, pattern = "outgoing", k = nPatterns)
        png(paste(outdir,"/outgoing_patterns.png",sep=""),width=10,height=10,units="in",res=200)
        netAnalysis_river(cellchat, pattern = "outgoing")
        dev.off()
        
        png(paste(outdir,"/outgoing_contribution_patterns.png",sep=""),width=20,height=5,units="in",res=200)
        netAnalysis_dot(cellchat, pattern = "outgoing") +
          theme(axis.text= element_text(size=14),
                axis.title = element_text(size=14),
                plot.title= element_text(size=16))
        dev.off()
        
      },  error=function(e){})
      
    } #end tissue loop
    
  },  error=function(e){})
  
} #end variable factor loop



#Get the most common interactions and names

itneraction_sum = read.delim(file = "/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/CellChat/results_v3/pap_ca_immune/interaction_results_cell_type.txt", header=T)
head(itneraction_sum)
unique(itneraction_sum$source)
sources = c("Upper Bridge","Middle Bridge","Lower Bridge")
pathway_sum = do.call(rbind, lapply(1:length(sources), function(i){
  #i=1
  interaction_sub = itneraction_sum[itneraction_sum$source %in% sources[i], ]
  pathway_freq = data.frame(table(interaction_sub$pathway_name))
  pathway_freq = data.frame(sources[i], pathway_freq[order(-pathway_freq$Freq),])
  names(pathway_freq) = c("source","pathway","frequency_significant_interaction")
  pathway_freq
}))


#Now find how often each pathway occurs which tells you how often it exists
pathwayset_sum = data.frame(table(pathway_sum$pathway))

write.table(pathway_sum,
            file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/CellChat/results_v3/pap_ca_immune/pathway_summaries.txt",
            quote=F,row.names=F,sep="\t")

#Now find how often each pathway occurs which tells you how often it exists
pathwayset_sum = data.frame(table(pathway_sum$pathway))
pathwayset_sum = pathwayset_sum[order(-pathwayset_sum$Freq),]
write.table(pathwayset_sum,
            file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/CellChat/results_v3/pap_ca_immune/pathwayset_sum.txt",
            quote=F,row.names=F,sep="\t")

#find which sources have these two signaling modes
pathway_sum[pathway_sum$pathway %in% subset(pathwayset_sum, Freq==2)$Var1,]
pathway_sum[pathway_sum$pathway %in% subset(pathwayset_sum, Freq==1)$Var1,]

#-----------------------------------------------------------------------------------------------------------------------------------#
#-----------------------------------------------------------------------------------------------------------------------------------#
#-----------------------------------------------------------------------------------------------------------------------------------#
#-----------------------------------------------------------------------------------------------------------------------------------#
#Get interactions unique to the carcinoma

#Read in interation results
ca_interactions = read.delim(file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/CellChat/results/CA/interaction_results_cell_type.txt",header=T)
nsk_interactions = read.delim(file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/CellChat/results/Normal\ skin/interaction_results_cell_type.txt",header=T)

length(unique(ca_interactions$interaction_name))
length(unique(nsk_interactions$interaction_name))

ca_specific_interactions = setdiff(ca_interactions$interaction_name,nsk_interactions$interaction_name) #get non-overlapping interactions
ca_interactions_only = ca_interactions[ca_interactions$interaction_name %in% ca_specific_interactions,] #now get the non-overlappign interactions present in the carcinoma
write.table(ca_interactions_only, 
            file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/CellChat/results/carcinoma_specific_interactions.txt",
            row.names=F,quote=F,sep="\t")

#-----------------------------------------------------------------------------------------------------------------------------------#

#Try SCENT package: https://github.com/aet21/SCENT
#https://rdrr.io/github/aet21/SCENT/f/vignettes/SCENT.Rmd

#I could try this with validated networks and my own networks generated from correlation matrices
#This could really unite everything: my networks, empirical networks, and finally stemness or differentiation potential!

library(devtools)
devtools::install_github("aet21/SCENT")

library(SCENT);
data(net13Jun12);
print(dim(net13Jun12.m));

#-----------------------------------------------------------------------------------------------------------------------------------#
#-----------------------------------------------------------------------------------------------------------------------------------#
#Run MAGIC
# https://htmlpreview.github.io/?https://github.com/KrishnaswamyLab/MAGIC/blob/master/Rmagic/inst/examples/emt_tutorial.html#using-magic-for-downstream-analysis
if (!require(viridis)) install.packages("viridis")
if (!require(ggplot2)) install.packages("ggplot2")
if (!require(readr)) install.packages("readr")
if (!require(phateR)) install.packages("phateR")

library(Rmagic)
library(readr)
library(ggplot2)
library(viridis)
library(phateR)

#Now you can impute the data
?magic

#Now you can impute the data
#Get upper and lower spike cells -> transform to Seurat object -> impute their values -> find biggest DEGs between them
data_MAGIC <- magic(data, genes="all_genes",
                    knn=15, init=data_MAGIC)

#-----------------------------------------------------------------------------------------------------------------------------------#
#-----------------------------------------------------------------------------------------------------------------------------------#

#Genie3 on particular cell clusters
library(GENIE3)
library(ggrepel)
library(tidyverse)

#Mouse TF database: https://www.nature.com/articles/ncomms15089#Sec29
mouse_tfs = read.xlsx(
  xlsxFile="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/transcription_factor_database/41467_2017_BFncomms15089_MOESM3449_ESM.xlsx",
  colNames=T,
  sheet=2
)
mouse_tfs = na.omit(unique(alias2SymbolTable(mouse_tfs$GeneSymbol, species = "Mm")))   #Get official gene symbol

#Get cell clusters I wanna infer a GRN on
cell_clusts = c(19,33,5,14,9,12,
                44, 15, 28, 25, 7, 10)
tissues = c("Pap","Pap","Pap","Pap","Pap","Pap",
            "CA","CA","CA","CA","CA","CA")

colData(big_cds)$combined_spike = "non-spike" #add column of of apparent no matches to ALL the cells; later you will replace with matches


#---------------------#
#Now move through this 
all_cell_clust_var_genes = do.call(rbind, lapply(3:length(cell_clusts), function(v){ 
  
  tryCatch({ #suppress error
    
    #v=1
    print(v)
    #subset to a specific clustser
    celltype_cds = big_cds[,colData(big_cds)$cluster %in% cell_clusts[v] & colData(big_cds)$tissue_updated %in% tissues[v]]
    
    #Min cells a gene must be expressed in
    no_cells = dim(celltype_cds)[2] #get number of cells in this subset
    min.cells_sub = round(no_cells*cell_threshold) #get min number
    
    #now create Seurat object from the patient-condition specific object
    seurat_obj = CreateSeuratObject(
      celltype_cds@assays@data$counts,
      project = "CreateSeuratObject",
      assay = "RNA",
      names.field = 1,
      names.delim = "_",
      meta.data = NULL,
      min.cells = min.cells_sub)
    
    passing_genes = dim(seurat_obj)[1] #number of genes passing the min.cells filter
    
    #FindVariableFeatures ONLY if there are more than 2000 genes
    seurat_obj <- FindVariableFeatures(seurat_obj, selection.method = "vst", nfeatures = 1000)
    variable_genes <- VariableFeatures(seurat_obj) #get variable gene names
    
    #Subset to variable genes
    to_test_cds = celltype_cds[rowData(celltype_cds)$gene_short_name %in% variable_genes,] #subset cds object to variable genes
    
    #Find overlap between transcription factors and high-variability genes
    target_tfs = na.omit(data.frame(mouse_tfs[match(variable_genes, mouse_tfs)]))[,1]
    
    #Perform genie inference
    exprMatr = as.matrix(to_test_cds@assays@data$counts)
    set.seed(123) # For reproducibility of results
    weightMat <- GENIE3(exprMatr,
                        nCores=4,
                        regulators = target_tfs,
                        verbose=T)
    
    write.table(weightMat,
                file=paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/GENIE3/cell_cluster_target_tf/","cell_cluster_",cell_clusts[v],".txt",sep=""),
                row.names=T,sep="\t",quote=F)
    
  },  error=function(e){})
  
}))
#--------------------------------------------------------------------------------------------------------------------#
#--------------------------------------------------------------------------------------------------------------------#
#--------------------------------------------------------------------------------------------------------------------#

#now get summaries of all of these lovely inferred GRNs

#Now move through this 
all_cell_clust_var_genes = do.call(rbind, lapply(1:length(cell_clusts), function(v){ 
  
  #v=1
  print(v)
  
  grn_mat = read.delim(file=paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/GENIE3/cell_cluster_target_tf/","cell_cluster_",cell_clusts[v],".txt",sep=""),
                       header=F)
  
  names(grn_mat) = as.vector(c("gene",grn_mat[1,1:(ncol(grn_mat)-1)])) #the column names shifted over 1 during printing out so I need to reshift them back
  grn_mat = grn_mat[-1,]
  
  #get distribution of values in this matrix
  #convert to a matrix
  row.names(grn_mat) = grn_mat$gene
  grn_mat = grn_mat[,-1] #drop gene column
  
  row_names_vec = row.names(grn_mat)
  col_names_vec = colnames(grn_mat)
  
  #convert character dataframe to numeric matrix
  grn_mat = as.matrix(grn_mat)
  mat_num <- matrix(as.numeric(grn_mat),    # Convert to numeric matrix
                    ncol = ncol(grn_mat))
  
  row.names(mat_num) = row_names_vec
  colnames(mat_num) = col_names_vec
  
  #plot distribution of all edge probabilities
  plot_vector = as.numeric(mat_num)
  vector_length = length(plot_vector)
  
  all_edge_weights = ggplot() +
    geom_histogram(aes(x=plot_vector),fill="tan",color="black", bins=75) +
    ggtitle(paste("Cluster ",cell_clusts[v]," complete pairwise regulation inference, N=",vector_length,sep="")) + 
    xlab("GENIE3 edge weight") + ylab("count") + 
    theme(axis.text.x=element_text(size=15,color='black'),
          axis.text.y=element_text(size=10,color='black'),
          axis.title=element_text(size=15),
          axis.ticks.x = element_line(size=0),
          strip.text.x = element_text(size = 15),
          legend.position="none",
          title = element_text(size = 15, face="bold")) +
    theme(
      # Hide panel borders and remove grid lines
      panel.border = element_blank(),
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      panel.background = element_blank(),
      element_line(colour = 'black', size = 0.3),
      # Change axis line
      axis.line = element_line(colour = "black"),
      strip.background = element_rect(colour="white",
                                      fill="white")
    )
  
  #Examine only top quantile of interactions
  quantile_cutoff_edgeprob = quantile(as.numeric(mat_num), 0.99)
  plot_vector.2 = (as.numeric(mat_num)[as.numeric(mat_num)>quantile_cutoff_edgeprob])
  vector_length = length(plot_vector.2)
  
  top_edge_weights = ggplot() +
    geom_histogram(aes(x=plot_vector.2),fill="tan",color="black", bins=75) +
    ggtitle(paste("Cluster ",cell_clusts[v]," top 1% pairwise regulation inference, N=",vector_length,sep="")) + 
    xlab("GENIE3 edge weight") + ylab("count") + 
    theme(axis.text.x=element_text(size=15,color='black'),
          axis.text.y=element_text(size=10,color='black'),
          axis.title=element_text(size=15),
          axis.ticks.x = element_line(size=0),
          strip.text.x = element_text(size = 15),
          legend.position="none",
          title = element_text(size = 15, face="bold")) +
    theme(
      # Hide panel borders and remove grid lines
      panel.border = element_blank(),
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      panel.background = element_blank(),
      element_line(colour = 'black', size = 0.3),
      # Change axis line
      axis.line = element_line(colour = "black"),
      strip.background = element_rect(colour="white",
                                      fill="white")
    )
  
  
  #-------------------------------------#
  #Get top 100 interactions
  linkList <- getLinkList(mat_num)
  top_edge = linkList[1:100,]
  table_layer = data.frame(table(top_edge$regulatoryGene))
  table_layer = subset(table_layer, Freq>0)
  names(table_layer) = c("group","value")
  df = table_layer
  df = droplevels(df) #drop unused empty factor levels
  top100_summary = df[order(-df$value),]
  top_edge100 = top_edge
  
  #mutate dataframe to get percentage i think
  df2 <- df %>% 
    mutate(csum = rev(cumsum(rev(value))), 
           pos = value/2 + lead(csum, 1),
           pos = if_else(is.na(pos), value/2, pos))
  
  #plot out pie chart
  top100_pie = ggplot(df, aes(x = "" , y = value, fill = fct_inorder(group))) +
    geom_col(width = 1, color = 1) +
    coord_polar(theta = "y") +
    #scale_fill_brewer(palette = "Pastel1") +
    geom_label_repel(data = df2,
                     aes(y = pos, label = group),
                     size = 4.5, nudge_x = 1, show.legend = FALSE) +
    guides(fill = guide_legend(title = "Group")) +
    theme_void() +
    ggtitle(paste("Cluster ",cell_clusts[v]," top 100 GRN TFs",sep="")) + 
    theme(legend.position="none",
          title = element_text(size = 15, face="bold"))
  
  #-------------------------------------#
  #Get top 1000 interactions
  linkList <- getLinkList(mat_num)
  top_edge = linkList[1:vector_length,] #get top 1% edges
  table_layer = data.frame(table(top_edge$regulatoryGene))
  table_layer = subset(table_layer, Freq>0)
  names(table_layer) = c("group","value")
  df = table_layer
  df = droplevels(df) #drop unused empty factor levels
  top1000_summary = df[order(-df$value),]
  top_edge1000 = top_edge
  
  #mutate dataframe to get percentage i think
  df2 <- df %>% 
    mutate(csum = rev(cumsum(rev(value))), 
           pos = value/2 + lead(csum, 1),
           pos = if_else(is.na(pos), value/2, pos))
  
  #plot out pie chart
  top1000_pie = ggplot(df, aes(x = "" , y = value, fill = fct_inorder(group))) +
    geom_col(width = 1, color = 1) +
    coord_polar(theta = "y") +
    #scale_fill_brewer(palette = "Pastel1") +
    geom_label_repel(data = df2,
                     aes(y = pos, label = group),
                     size = 4.5, nudge_x = 1, show.legend = FALSE) +
    guides(fill = guide_legend(title = "Group")) +
    theme_void() +
    ggtitle(paste("Cluster ",cell_clusts[v]," top 1% GRN TFs",sep="")) + 
    theme(legend.position="none",
          title = element_text(size = 15, face="bold"))
  
  #-----------------------------------------------------#
  #Graph cell cluster I'm looking at
  colData(big_cds)$combined_spike = "non-spike" #add column of of apparent no matches to ALL the cells; later you will replace with matches
  cds_subset.v3 = big_cds[,colData(big_cds)$cluster %in% c(cell_clusts[v])] #filter for cluster 19 because most of the bridge cells are only in cluster 19
  matching_cells = match(colnames(cds_subset.v3),colnames(big_cds)) #match extracted cells with all cell names
  colData(big_cds)$combined_spike[matching_cells] <- "lower spike" #now replace matching cells with new name
  
  combined_spike_fig = plot_cells(big_cds, color_cells_by="combined_spike", group_cells_by="cluster",
                                  group_label_size = 0,
                                  x=1,
                                  y=2,
                                  alpha=0.4,
                                  cell_size=0.5) +
    ggtitle(paste("Cell cluster ",cell_clusts[v],sep="")) + xlab("") + ylab("") + 
    theme(title = element_text(size = 15, face="bold"),
          axis.text=element_text(size=0,color='black'),
          axis.title=element_text(size=0),
          axis.ticks = element_line(size=0),
          legend.position="none") +
    scale_color_manual(values = c("non-spike" = "lightgrey", "lower spike" = "darkgreen"))
  
  #-----------------------------------------------------#
  #Graph top 5 TFs
  marker_df = data.frame(top1000_summary$group[1:5],"eigengene","eigengene")
  names(marker_df) = c("gene","cell_type","module")
  
  agexp_plot.12 = plot_cells(big_cds,
                             genes=marker_df,
                             label_cell_groups=FALSE,
                             label_leaves=F,
                             label_branch_points=F,
                             graph_label_size=1.5,
                             show_trajectory_graph=F) +
    xlab("") + ylab("") + 
    theme(axis.ticks = element_blank()) +
    theme(legend.position="none") +
    ggtitle(paste("Cell cluster ",cell_clusts[v]," top 5 TFs",sep="")) + 
    scale_color_viridis_c(option = "inferno",name="integrated\nexpression") +
    theme(axis.text=element_text(size=0,color='black'),
          axis.title=element_text(size=0,color='black'),
          title = element_text(size = 15, face="bold")) +
    theme(strip.text.x = element_text(size = 0))  +
    theme(legend.position=c(0.8, 0.2),
          legend.key.size = unit(1, 'cm'),
          legend.text = element_text(size=12.5)) #remove facet label
  
  
  #-----------------------------------------------------#
  #Plot everything out
  png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/GENIE3/cell_cluster_target_tf/summaries/figs/cell_cluster_",cell_clusts[v],".png",sep=""),
      width=24,height=16,units="in",res=200)
  print(
    multiplot(all_edge_weights, top_edge_weights,
              top100_pie, top1000_pie,
              combined_spike_fig, agexp_plot.12,
              cols=3)
  )
  dev.off()
  
  #-----------------------------------------------------#
  #Print summaries out into excel sheet
  OUT <- createWorkbook() #create excel workbook for a sample https://stackoverflow.com/questions/27713310/easy-way-to-export-multiple-data-frame-to-multiple-excel-worksheets
  
  addWorksheet(OUT, "Top100_summary") # Add some sheets to the workbook
  writeData(OUT, sheet = "Top100_summary", x = top100_summary) #Write data to a sheet
  addWorksheet(OUT, "Top100_edge") # Add some sheets to the workbook
  writeData(OUT, sheet = "Top100_edge", x = top_edge100) #Write data to a sheet
  
  addWorksheet(OUT, "Top1000_summary") # Add some sheets to the workbook
  writeData(OUT, sheet = "Top1000_summary", x = top1000_summary) #Write data to a sheet
  addWorksheet(OUT, "Top1000_edge") # Add some sheets to the workbook
  writeData(OUT, sheet = "Top1000_edge", x = top_edge1000) #Write data to a sheet
  
  saveWorkbook(OUT, 
               paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/GENIE3/cell_cluster_target_tf/summaries/excel/cell_cluster_",
                     cell_clusts[v],".xlsx",sep=""),
               overwrite=T)
  
  
}))





#----------------------------------------------------------------------------------------------------------------------------------------#
#----------------------------------------------------------------------------------------------------------------------------------------#
#----------------------------------------------------------------------------------------------------------------------------------------#
#Cell composition of bulk samplees
devtools::install_github('shenorrlab/bseq-sc')

#----------------------------------------------------------------------------------------------------------------------------------------#
#----------------------------------------------------------------------------------------------------------------------------------------#
#----------------------------------------------------------------------------------------------------------------------------------------#
#----------------------------------------------------------------------------------------------------------------------------------------#
#----------------------------------------------------------------------------------------------------------------------------------------#
#Slingshot trajectory analysis

BiocManager::install("slingshot", version="3.10")
library(slingshot)
R.version


#----------------------------------------------------------------------------------------------------------------------------------#
#----------------------------------------------------------------------------------------------------------------------------------#
#----------------------------------------------------------------------------------------------------------------------------------#
#----------------------------------------------------------------------------------------------------------------------------------#
#V1 VERY UNIQUE SPIKE ANALYSIS
#Make lower and upper spike spike manifold
lower_spike <- choose_cells(big_cds)
dim(lower_spike)[2]

#Write out the lower spike cell IDs to disk
write.table(data.frame(colnames(lower_spike)),
            file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/matvei/cds_objects/lower_spike_cellnames.txt",
            quote=F, row.names=F, sep="\t")
table(colData(big_cds)$cluster)

#----------------------------------------------------#
#Read in the lower spike cells
lower_spike_cell_ids = read.table(file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/matvei/cds_objects/lower_spike_cellnames.txt",
                                  header=T)
lower_spike_cell_ids = lower_spike_cell_ids[,1]

#----------------------------------------------------#
#Get the spike cells
upperspike_cds = big_cds[,colData(big_cds)$cluster %in% c("19")] #filter for cluster 19 because most of the bridge cells are only in cluster 19
colData(upperspike_cds)$spike = "Upper spike"
lowerspike_cds = big_cds[,lower_spike_cell_ids]
colData(lowerspike_cds)$spike = "Lower spike"

spikes_cds = combine_cds(list(upperspike_cds,lowerspike_cds))

colData(spikes_cds)

#----------------------------------------------------#
#Optimize manifold of spikes
min_distances = seq(0.05,1,by=0.05) #This is showing a lot of interesting bridgeness with min distanes bewteen 0.4 and 0.6

spikes_cds <- preprocess_cds(spikes_cds, 
                             method="PCA",
                             num_dim = 100,
                             norm_method = "none", #these data have already been normalized by Seurat
                             alignment_group = NULL)

pdf(file=paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/spike_analysis/combined_spike_analysis/manifold_optimization/manifold_min_distances_v1.pdf",sep=""),height=10,width=10)
for (i in 1:length(min_distances)) {
  #i=20
  print(paste("i=",i,sep=""))
  
  #4b. ?reduce_dimension 
  spikes_cds.test <- reduce_dimension(
    spikes_cds,
    max_components = 2,
    reduction_method = c("UMAP"),
    preprocess_method = "PCA", #must now used ALIGNEMNT PRE-PROCESSING METHOD bc it replaces the other pre-process shit
    umap.metric = "cosine",
    umap.min_dist = min_distances[i],
    umap.n_neighbors = 3L,
    umap.fast_sgd = FALSE,
    umap.nn_method = "annoy",
    cores = 1,
    verbose = FALSE
  )
  
  #5. Cluster and partition ?cluster_cells cells into supergroups community detection (Phenograph algorithm) Community detection of cells as part of Phenograph algorithm https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4508757/
  #5b.  ?cluster_cells
  spikes_cds.test <- cluster_cells(spikes_cds.test,
                                   reduction_method = c("UMAP"),
                                   k = 10, #a bigger k will result in lower resolution 
                                   cluster_method = c("leiden"),
                                   num_iter = 2,
                                   partition_qval = 0.05, #Numeric, the q-value cutoff to determine when to partition
                                   weight = FALSE,
                                   resolution = NULL,
                                   random_seed = NULL,
                                   verbose = F)
  print( 
    plot_cells(spikes_cds.test, color_cells_by="spike", group_cells_by="cluster",
               group_label_size = 6,
               show_trajectory_graph=F,
               x=1,
               y=2,
               alpha=0.8,
               cell_size=1) +
      ggtitle(paste(" Minimal distance = ",min_distances[i],sep="")) + xlab("Dim. 1") + ylab("Dim. 2") +
      theme(plot.title = element_text(size = 22, vjust=1)) +
      theme(axis.text=element_text(size=0,color='black'),
            axis.title=element_text(size=0),
            axis.ticks = element_line(size=0)) +
      scale_color_manual(values = c("Upper spike" = "darkgoldenrod4","Lower spike" = "firebrick"))
  )
}
dev.off()

} #End sortings loop

i=20
print(paste("i=",i,sep=""))

?reduce_dimension

#4b. ?reduce_dimension 
spikes_cds.test <- reduce_dimension(
  spikes_cds,
  max_components = 2,
  reduction_method = c("UMAP"),
  preprocess_method = "PCA", #must now used ALIGNEMNT PRE-PROCESSING METHOD bc it replaces the other pre-process shit
  umap.metric = "cosine",
  umap.min_dist = min_distances[i],
  umap.n_neighbors = 3L,
  umap.fast_sgd = FALSE,
  umap.nn_method = "annoy",
  cores = 1,
  verbose = FALSE
)

#5. Cluster and partition ?cluster_cells cells into supergroups community detection (Phenograph algorithm) Community detection of cells as part of Phenograph algorithm https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4508757/
#5b.  ?cluster_cells
spikes_cds.test <- cluster_cells(spikes_cds.test,
                                 reduction_method = c("UMAP"),
                                 k = 10, #a bigger k will result in lower resolution 
                                 cluster_method = c("leiden"),
                                 num_iter = 2,
                                 partition_qval = 0.05, #Numeric, the q-value cutoff to determine when to partition
                                 weight = FALSE,
                                 resolution = NULL,
                                 random_seed = NULL,
                                 verbose = F)
print( 
  plot_cells(spikes_cds.test, color_cells_by="spike", group_cells_by="cluster",
             group_label_size = 6,
             show_trajectory_graph=F,
             x=1,
             y=2,
             alpha=0.8,
             cell_size=1) +
    ggtitle("Optimized spike manifold") + xlab("Dim. 1") + ylab("Dim. 2") +
    theme(plot.title = element_text(size = 22, vjust=1)) +
    theme(axis.text=element_text(size=0,color='black'),
          axis.title=element_text(size=0),
          axis.ticks = element_line(size=0)) +
    scale_color_manual(values = c("Upper spike" = "darkgoldenrod4","Lower spike" = "firebrick"))
)

#Infer trajectory
?learn_graph
spikes_cds.test <- learn_graph(spikes_cds.test, use_partition=F)

print( 
  plot_cells(spikes_cds.test, color_cells_by="spike", group_cells_by="cluster",
             group_label_size = 6,
             show_trajectory_graph=T,
             graph_label_size = 6,
             x=1,
             y=2,
             alpha=0.3,
             cell_size=1) +
    ggtitle("Optimized spike manifold") + xlab("Dim. 1") + ylab("Dim. 2") +
    theme(plot.title = element_text(size = 22, vjust=1)) +
    theme(axis.text=element_text(size=0,color='black'),
          axis.title=element_text(size=0),
          axis.ticks = element_line(size=0)) +
    scale_color_manual(values = c("Upper spike" = "darkgoldenrod4","Lower spike" = "firebrick"))
)

#Get upper spike markers
marker_df = data.frame(unique(alias2SymbolTable(unique( na.omit(balmain_genes[,match("BANF1_CAR_WT",names(balmain_genes))]) ), species = "Mm")),"eigengene","eigengene")
names(marker_df) = c("gene","cell_type","module")
upper_spike_markers = na.omit(marker_df) #remove non-matching genes

#Get upper spike markers
marker_df = data.frame(unique(alias2SymbolTable(unique( na.omit(balmain_genes[,match("Krt15",names(balmain_genes))]) ), species = "Mm")),"eigengene","eigengene")
names(marker_df) = c("gene","cell_type","module")
lower_spike_markers = na.omit(marker_df) #remove non-matching genes


big_upperspike = plot_cells(big_cds,
                            genes=upper_spike_markers, #seed gene is the first row of the markers file
                            label_cell_groups=FALSE,
                            label_leaves=F,
                            label_branch_points=F,
                            graph_label_size=1.5,
                            show_trajectory_graph=F) +
  xlab("") + ylab("") + 
  theme(axis.ticks = element_blank()) +
  theme(legend.position="none") +
  scale_color_viridis_c(option = "inferno",name="expression") +
  ggtitle("BANF1_CAR_WT")+ #continuous
  theme(axis.text=element_text(size=0,color='black'),
        axis.title=element_text(size=0,color='black')) + theme(plot.title = element_text(size = 15, vjust=0)) +
  theme(strip.text.x = element_text(size = 0)) +
  theme(legend.position=c(0.8, 0.2),
        legend.key.size = unit(0.7, 'cm'),
        legend.text = element_text(size=12.5)) #remove facet label

spike_upperspike = plot_cells(spikes_cds.test,
                              genes=upper_spike_markers, #seed gene is the first row of the markers file
                              label_cell_groups=FALSE,
                              label_leaves=F,
                              label_branch_points=F,
                              graph_label_size=1.5,
                              show_trajectory_graph=F,
                              cell_size=2) +
  xlab("") + ylab("") + 
  theme(axis.ticks = element_blank()) +
  theme(legend.position="none") +
  scale_color_viridis_c(option = "inferno",name="expression") +
  ggtitle("BANF1_CAR_WT")+ #continuous
  theme(axis.text=element_text(size=0,color='black'),
        axis.title=element_text(size=0,color='black')) + theme(plot.title = element_text(size = 15, vjust=0)) +
  theme(strip.text.x = element_text(size = 0)) +
  theme(legend.position="none") #remove facet label


big_lowerspike = plot_cells(big_cds,
                            genes=lower_spike_markers, #seed gene is the first row of the markers file
                            label_cell_groups=FALSE,
                            label_leaves=F,
                            label_branch_points=F,
                            graph_label_size=1.5,
                            show_trajectory_graph=F) +
  xlab("") + ylab("") + 
  theme(axis.ticks = element_blank()) +
  theme(legend.position="none") +
  scale_color_viridis_c(option = "inferno",name="expression") +
  ggtitle("SCC Krt15")+ #continuous
  theme(axis.text=element_text(size=0,color='black'),
        axis.title=element_text(size=0,color='black')) + theme(plot.title = element_text(size = 15, vjust=0)) +
  theme(strip.text.x = element_text(size = 0)) +
  theme(legend.position=c(0.8, 0.2),
        legend.key.size = unit(0.7, 'cm'),
        legend.text = element_text(size=12.5)) #remove facet label

spike_lowerspike = plot_cells(spikes_cds.test,
                              genes=lower_spike_markers, #seed gene is the first row of the markers file
                              label_cell_groups=FALSE,
                              label_leaves=F,
                              label_branch_points=F,
                              graph_label_size=1.5,
                              show_trajectory_graph=F,
                              cell_size=2) +
  xlab("") + ylab("") + 
  theme(axis.ticks = element_blank()) +
  theme(legend.position="none") +
  scale_color_viridis_c(option = "inferno",name="expression") +
  ggtitle("SCC Krt15")+ #continuous
  theme(axis.text=element_text(size=0,color='black'),
        axis.title=element_text(size=0,color='black')) + theme(plot.title = element_text(size = 15, vjust=0)) +
  theme(strip.text.x = element_text(size = 0)) +
  theme(legend.position="none") #remove facet label


png("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/spike_analysis/combined_spike_analysis/spikes_alone/markers.png",
    width=14,height=14,units="in",res=400)
print(
  multiplot(big_upperspike, spike_upperspike,
            big_lowerspike, spike_lowerspike,
            cols=2)
)
dev.off()

#----------------------------------------------------------------------------------------------------------------------------------#
#----------------------------------------------------------------------------------------------------------------------------------#
#----------------------------------------------------------------------------------------------------------------------------------#
#----------------------------------------------------------------------------------------------------------------------------------#
#V2 EXPANDING 
#Make lower and upper spike spike manifold
#Get the spike cells
spikes_cds = big_cds[,colData(big_cds)$cluster %in% c("19","5","33")] #filter for cluster 19 because most of the bridge cells are only in cluster 19
colData(spikes_cds)$spike = colData(spikes_cds)$cluster
colData(spikes_cds)$spike = gsub("19","Upper spike",colData(spikes_cds)$spike)
colData(spikes_cds)$spike = gsub("5","Intermediate",colData(spikes_cds)$spike)
colData(spikes_cds)$spike = gsub("33","Lower spike",colData(spikes_cds)$spike)

unique(colData(spikes_cds)$spike)

#----------------------------------------------------#
#Optimize manifold of spikes
min_distances = seq(0.05,1,by=0.05) #This is showing a lot of interesting bridgeness with min distanes bewteen 0.4 and 0.6

spikes_cds <- preprocess_cds(spikes_cds, 
                             method="PCA",
                             num_dim = 100,
                             norm_method = "none", #these data have already been normalized by Seurat
                             alignment_group = NULL)

pdf(file=paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/spike_analysis/combined_spike_analysis/manifold_optimization/manifold_min_distances_v2.pdf",sep=""),height=10,width=10)
for (i in 1:length(min_distances)) {
  #i=11
  print(paste("i=",i,sep=""))
  
  #4b. ?reduce_dimension 
  spikes_cds.test <- reduce_dimension(
    spikes_cds,
    max_components = 2,
    reduction_method = c("UMAP"),
    preprocess_method = "PCA", #must now used ALIGNEMNT PRE-PROCESSING METHOD bc it replaces the other pre-process shit
    umap.metric = "cosine",
    umap.min_dist = min_distances[i],
    umap.n_neighbors = 10L,
    umap.fast_sgd = FALSE,
    umap.nn_method = "annoy",
    cores = 1,
    verbose = FALSE
  )
  
  #5. Cluster and partition ?cluster_cells cells into supergroups community detection (Phenograph algorithm) Community detection of cells as part of Phenograph algorithm https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4508757/
  #5b.  ?cluster_cells
  spikes_cds.test <- cluster_cells(spikes_cds.test,
                                   reduction_method = c("UMAP"),
                                   k = 10, #a bigger k will result in lower resolution 
                                   cluster_method = c("leiden"),
                                   num_iter = 2,
                                   partition_qval = 0.05, #Numeric, the q-value cutoff to determine when to partition
                                   weight = FALSE,
                                   resolution = NULL,
                                   random_seed = NULL,
                                   verbose = F)
  print( 
    plot_cells(spikes_cds.test, color_cells_by="spike", group_cells_by="cluster",
               group_label_size = 6,
               show_trajectory_graph=F,
               x=1,
               y=2,
               alpha=0.8,
               cell_size=1) +
      ggtitle(paste(" Minimal distance = ",min_distances[i],sep="")) + xlab("Dim. 1") + ylab("Dim. 2") +
      theme(plot.title = element_text(size = 22, vjust=1)) +
      theme(axis.text=element_text(size=0,color='black'),
            axis.title=element_text(size=0),
            axis.ticks = element_line(size=0)) +
      scale_color_manual(values = c("Upper spike" = "darkgoldenrod4","Lower spike" = "firebrick",
                                    "Intermediate" = "darkolivegreen4"))
  )
}
dev.off()


i=12
print(paste("i=",i,sep=""))

#4b. ?reduce_dimension 
spikes_cds.test <- reduce_dimension(
  spikes_cds,
  max_components = 2,
  reduction_method = c("UMAP"),
  preprocess_method = "PCA", #must now used ALIGNEMNT PRE-PROCESSING METHOD bc it replaces the other pre-process shit
  umap.metric = "cosine",
  umap.min_dist = min_distances[i],
  umap.n_neighbors = 10L,
  umap.fast_sgd = FALSE,
  umap.nn_method = "annoy",
  cores = 1,
  verbose = FALSE
)

#5. Cluster and partition ?cluster_cells cells into supergroups community detection (Phenograph algorithm) Community detection of cells as part of Phenograph algorithm https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4508757/
#5b.  ?cluster_cells
spikes_cds.test <- cluster_cells(spikes_cds.test,
                                 reduction_method = c("UMAP"),
                                 k = 10, #a bigger k will result in lower resolution 
                                 cluster_method = c("leiden"),
                                 num_iter = 2,
                                 partition_qval = 0.05, #Numeric, the q-value cutoff to determine when to partition
                                 weight = FALSE,
                                 resolution = NULL,
                                 random_seed = NULL,
                                 verbose = F)

#Infer trajectory
spikes_cds.test <- learn_graph(spikes_cds.test, use_partition=F)

trajecotry = plot_cells(spikes_cds.test, color_cells_by="spike", group_cells_by="cluster",
                        group_label_size = 0,
                        show_trajectory_graph=T,
                        graph_label_size = 6,
                        x=1,
                        y=2,
                        alpha=0.3,
                        cell_size=1) +
  ggtitle("Optimized spike manifold") + xlab("Dim. 1") + ylab("Dim. 2") +
  theme(plot.title = element_text(size = 22, vjust=1)) +
  theme(axis.text=element_text(size=0,color='black'),
        axis.title=element_text(size=0),
        axis.ticks = element_line(size=0)) +
  scale_color_manual(values = c("Upper spike" = "darkgoldenrod4","Lower spike" = "firebrick",
                                "Intermediate" = "darkolivegreen4"))

png("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/spike_analysis/combined_spike_analysis/manifold_optimization/optimized_manifold_v2.png",
    width=8,height=8,units="in",res=250)
print(
  trajecotry
)
dev.off()

#Plot clusters
?cluster_cells
set.seed(6)
spikes_cds.test = cluster_cells(spikes_cds.test, resolution=0.005) #re-cluster cells into high-res clusters #high number is MORE resolution
#must optimize the unsupervised clustering until the lower spike falls out
colData(spikes_cds.test)$cluster = spikes_cds.test@clusters@listData$UMAP$clusters
cluster_plot = plot_cells(spikes_cds.test, color_cells_by="cluster", group_cells_by="cluster",
                          group_label_size = 12,
                          label_groups_by_cluster=T,
                          x=1,
                          y=2,
                          show_trajectory_graph=F,
                          cell_size=1,
                          alpha=0.5) +
  ggtitle("Unsupervised high-res cell cluster") + xlab("") + ylab("") +
  theme(plot.title = element_text(size = 22, vjust=1)) +
  theme(axis.text=element_text(size=0,color='black'),
        axis.title=element_text(size=0),
        axis.ticks = element_line(size=0)) +
  scale_color_manual(values=getPalette(length(unique(colData(spikes_cds.test)$cluster))))
png("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/spike_analysis/combined_spike_analysis/manifold_optimization/optimized_manifold_clusterv2.png",
    width=8,height=8,units="in",res=250)
print( cluster_plot )
dev.off()

#--------------------------------------------------------------------------------------------------------------------------------#
#Upper spike graph test
upper_spike_trajectory <- choose_cells(spikes_cds.test)
subset_pr_test_res <- graph_test(upper_spike_trajectory, neighbor_graph="principal_graph", cores=4) #perform trajectory analysis
subset_pr_test_res = data.frame(subset_pr_test_res[order(-subset_pr_test_res$morans_I),]) #re-order single gene results from high to low
re_order_gene_changes <- subset(subset_pr_test_res, q_value < 0.05)
graph_test_df_comb.reord = re_order_gene_changes[,c("gene_short_name","morans_I","morans_test_statistic","p_value","q_value")] #re-order columns
gene_module_df <- data.frame(find_gene_modules(spikes_cds.test[graph_test_df_comb.reord$gene_short_name,], resolution=1e-2)) #infer gene module of thesee genees
graph_test_df_comb.reord$gene_module = gene_module_df$module  #add gene module information to the graph test results
write.table(graph_test_df_comb.reord, file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/spike_analysis/combined_spike_analysis/graph_test/upper_spike_cells_trajecotry.txt",sep="\t",row.names=F,quote=F)


#Calculate aggregate gene expression for unsupervised cell clusters and inffered gene modules
agg_mat = aggregate_gene_expression(
  spikes_cds.test,
  gene_group_df = gene_module_df,
  cell_group_df = data.frame(colnames(spikes_cds.test) ,colData(spikes_cds.test)$cluster ),
  scale_agg_values = TRUE,
  max_agg_value = NA,
  min_agg_value = NA,
  exclude.na = TRUE
)
row.names(agg_mat) <- stringr::str_c("Module ", row.names(agg_mat))
colnames(agg_mat) <- stringr::str_c("Cluster ", colnames(agg_mat))

png("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/spike_analysis/combined_spike_analysis/graph_test/graph_test_upperspike_heatmap.png",
    width=8,height=8,units="in",res=250)
print(
  pheatmap::pheatmap(agg_mat, cluster_rows=TRUE, cluster_cols=TRUE,
                     scale="column", clustering_method="ward.D2",
                     fontsize=12,
                     main="Upper spike gene module trajectory")
)
dev.off()
dev.off()
dev.set(dev.next())
dev.set(dev.next())

#plot expression of various genes
upper_spike_res = read.table(file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/spike_analysis/combined_spike_analysis/graph_test/upper_spike_cells_trajecotry.txt",
                             header=T)
#i=1
plot_cells(spikes_cds.test,
           genes=upper_spike_res$gene_short_name[i], #seed gene is the first row of the markers file
           label_cell_groups=FALSE,
           label_leaves=F,
           label_branch_points=F,
           graph_label_size=1.5,
           show_trajectory_graph=F,
           cell_size=2) +
  xlab("") + ylab("") + 
  theme(axis.ticks = element_blank()) +
  theme(legend.position="bottom") +
  scale_color_viridis_c(option = "inferno",name="expression") +
  ggtitle(paste(upper_spike_res$gene_short_name[i])) + #continuous
  theme(axis.text=element_text(size=0,color='black'),
        axis.title=element_text(size=0,color='black')) + theme(plot.title = element_text(size = 15, vjust=0)) +
  theme(strip.text.x = element_text(size = 0))

plot_cells(spikes_cds.test,
           genes=c(upper_spike_res$gene_short_name[i],"Cdkn2a","Cdkn2b","Trp53"), #seed gene is the first row of the markers file
           label_cell_groups=FALSE,
           label_leaves=F,
           label_branch_points=F,
           graph_label_size=1.5,
           show_trajectory_graph=F,
           cell_size=1.3) +
  xlab("") + ylab("") + 
  theme(axis.ticks = element_blank()) +
  theme(legend.position="bottom") +
  scale_color_viridis_c(option = "inferno",name="expression") +
  #ggtitle(paste(upper_spike_res$gene_short_name[i])) + #continuous
  theme(axis.text=element_text(size=0,color='black'),
        axis.title=element_text(size=0,color='black')) + theme(plot.title = element_text(size = 15, vjust=0)) +
  theme(strip.text.x = element_text(size = 15))

#--------------------------------------------------------------------------------------------------------------------------------#
#Lower spike graph test
lower_spike_trajectory <- choose_cells(spikes_cds.test)
subset_pr_test_res <- graph_test(lower_spike_trajectory, neighbor_graph="principal_graph", cores=4) #perform trajectory analysis
subset_pr_test_res = data.frame(subset_pr_test_res[order(-subset_pr_test_res$morans_I),]) #re-order single gene results from high to low
re_order_gene_changes <- subset(subset_pr_test_res, q_value < 0.05)
graph_test_df_comb.reord = re_order_gene_changes[,c("gene_short_name","morans_I","morans_test_statistic","p_value","q_value")] #re-order columns
gene_module_df <- data.frame(find_gene_modules(spikes_cds.test[graph_test_df_comb.reord$gene_short_name,], resolution=1e-2)) #infer gene module of thesee genees
graph_test_df_comb.reord$gene_module = gene_module_df$module  #add gene module information to the graph test results
write.table(graph_test_df_comb.reord, file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/spike_analysis/combined_spike_analysis/graph_test/lower_spike_cells_trajecotry.txt",sep="\t",row.names=F,quote=F)

#Calculate aggregate gene expression for unsupervised cell clusters and inffered gene modules
agg_mat = aggregate_gene_expression(
  spikes_cds.test,
  gene_group_df = gene_module_df,
  cell_group_df = data.frame(colnames(spikes_cds.test) ,colData(spikes_cds.test)$cluster ),
  scale_agg_values = TRUE,
  max_agg_value = NA,
  min_agg_value = NA,
  exclude.na = TRUE
)
row.names(agg_mat) <- stringr::str_c("Module ", row.names(agg_mat))
colnames(agg_mat) <- stringr::str_c("Cluster ", colnames(agg_mat))

png("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/spike_analysis/combined_spike_analysis/graph_test/graph_test_lowerspike_heatmap.png",
    width=8,height=8,units="in",res=250)
print(
  pheatmap::pheatmap(agg_mat, cluster_rows=TRUE, cluster_cols=TRUE,
                     scale="column", clustering_method="ward.D2",
                     fontsize=12,
                     main="Lower spike gene module trajectory")
)
dev.off()
dev.off()
dev.set(dev.next())
dev.set(dev.next())


#--------------------------------------------------------------------------------------------------------------------------------#
#Pseudotime inference
spikes_cds.test <- order_cells(spikes_cds.test)
pseudotime_plot = plot_cells(spikes_cds.test,
                             color_cells_by = "pseudotime",
                             group_label_size = 0,
                             show_trajectory_graph=T,
                             graph_label_size = 6,
                             x=1,
                             y=2,
                             alpha=0.5,
                             cell_size=1) +
  xlab("") + ylab("") + 
  theme(axis.ticks = element_blank()) +
  scale_color_viridis_c(option = "inferno",name="Pseudotime") +
  ggtitle("Pseudotime from rooted topology")+ #continuous
  theme(axis.text=element_text(size=0,color='black'),
        axis.title=element_text(size=0,color='black')) + theme(plot.title = element_text(size = 15, vjust=0)) +
  theme(strip.text.x = element_text(size = 0)) +
  theme(legend.position="bottom") #remove facet label

png("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/spike_analysis/combined_spike_analysis/manifold_optimization/pseudotime_inerence_v2.png",
    width=8,height=8,units="in",res=250)
print( pseudotime_plot )
dev.off()

#--------------------------------------------------------------------------------------------------------------------------------#
#--------------------------------------------------------------------------------------------------------------------------------#
#Do seed genes correlate to spike-specific graph test trajectories
seed_gene = c("Lgr5",
              "Bmi1","Sox4","Lrig1",
              "Lgr6","Sox9",
              "Sox2","Prom1","Foxa1","Krt15","Krt19","Pitx1","Psca","Prom2","Klf5",
              "Cd44")
upper_spike = read.table(file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/spike_analysis/combined_spike_analysis/graph_test/upper_spike_cells_trajecotry.txt",header=T)
lower_spike = read.table(file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/spike_analysis/combined_spike_analysis/graph_test/lower_spike_cells_trajecotry.txt",header=T)
upper_spike$spike = "upper"
lower_spike$spike = "lower"

all_spike = rbind(upper_spike, lower_spike)
head(all_spike)
all_spike = subset(all_spike, q_value<0.05)
spike_stem_genes = na.omit(all_spike[ match(seed_gene, all_spike$gene_short_name ) , ])
spike_stem_genes = spike_stem_genes[order(-spike_stem_genes$morans_I),] #re-order from high-to-low Moran statistic

us_stem = subset(spike_stem_genes, spike=="upper")
ls_stem = subset(spike_stem_genes, spike=="lower")

#--------------------------------------------------------------------------------------------------------------------------------#
#LS stem genes
for(i in 1:nrow(ls_stem)) {
  #i=1
  ls_stem_ind_genes = plot_cells(big_cds,
                                 genes = ls_stem$gene_short_name[i],
                                 group_label_size = 0,
                                 show_trajectory_graph=F,
                                 graph_label_size =0,
                                 x=1,
                                 y=2,
                                 alpha=0.8,
                                 cell_size=0.5) +
    xlab("") + ylab("") + 
    theme(axis.ticks = element_blank()) +
    scale_color_viridis_c(option = "inferno",name="expression") +
    ggtitle(paste("Moran I = ",round(ls_stem$morans_I[i],2),sep=""))+ #continuous
    theme(axis.text=element_text(size=0,color='black'),
          axis.title=element_text(size=0,color='black')) + theme(plot.title = element_text(size = 15, vjust=0)) +
    theme(strip.text.x = element_text(size = 15)) +
    theme(legend.position="bottom") #remove facet label
  
  png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/spike_analysis/combined_spike_analysis/graph_test_stem_genes/lower_spike/",
            ls_stem$gene_short_name[i],".png",sep=""),
      width=8,height=8,units="in",res=250)
  print( ls_stem_ind_genes )
  dev.off()
  
}

#--------------------------------------------------------------------------------------------------------------------------------#
#US stem genes
for(i in 1:nrow(us_stem)) {
  #i=1
  us_stem_ind_genes = plot_cells(big_cds,
                                 genes = us_stem$gene_short_name[i],
                                 group_label_size = 0,
                                 show_trajectory_graph=F,
                                 graph_label_size =0,
                                 x=1,
                                 y=2,
                                 alpha=0.8,
                                 cell_size=0.5) +
    xlab("") + ylab("") + 
    theme(axis.ticks = element_blank()) +
    scale_color_viridis_c(option = "inferno",name="expression") +
    ggtitle(paste("Moran I = ",round(us_stem$morans_I[i],2),sep=""))+ #continuous
    theme(axis.text=element_text(size=0,color='black'),
          axis.title=element_text(size=0,color='black')) + theme(plot.title = element_text(size = 15, vjust=0)) +
    theme(strip.text.x = element_text(size = 15)) +
    theme(legend.position="bottom") #remove facet label
  
  png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/spike_analysis/combined_spike_analysis/graph_test_stem_genes/upper_spike/",
            us_stem$gene_short_name[i],".png",sep=""),
      width=8,height=8,units="in",res=250)
  print( us_stem_ind_genes )
  dev.off()
  
}

#--------------------------------------------------------------------------------------------------------------------------------#
#--------------------------------------------------------------------------------------------------------------------------------#


#--------------------------------------------------------------------------------------------------------------------------------#
#--------------------------------------------------------------------------------------------------------------------------------#
#--------------------------------------------------------------------------------------------------------------------------------#
#--------------------------------------------------------------------------------------------------------------------------------#
#--------------------------------------------------------------------------------------------------------------------------------#
#--------------------------------------------------------------------------------------------------------------------------------#

#Get upper spike markers
marker_df = data.frame(unique(alias2SymbolTable(unique( na.omit(balmain_genes[,match("BANF1_CAR_WT",names(balmain_genes))]) ), species = "Mm")),"eigengene","eigengene")
names(marker_df) = c("gene","cell_type","module")
upper_spike_markers = na.omit(marker_df) #remove non-matching genes

#Get upper spike markers
marker_df = data.frame(unique(alias2SymbolTable(unique( na.omit(balmain_genes[,match("Krt15",names(balmain_genes))]) ), species = "Mm")),"eigengene","eigengene")
names(marker_df) = c("gene","cell_type","module")
lower_spike_markers = na.omit(marker_df) #remove non-matching genes


big_upperspike = plot_cells(big_cds,
                            genes=upper_spike_markers, #seed gene is the first row of the markers file
                            label_cell_groups=FALSE,
                            label_leaves=F,
                            label_branch_points=F,
                            graph_label_size=1.5,
                            show_trajectory_graph=F) +
  xlab("") + ylab("") + 
  theme(axis.ticks = element_blank()) +
  theme(legend.position="none") +
  scale_color_viridis_c(option = "inferno",name="expression") +
  ggtitle("BANF1_CAR_WT")+ #continuous
  theme(axis.text=element_text(size=0,color='black'),
        axis.title=element_text(size=0,color='black')) + theme(plot.title = element_text(size = 15, vjust=0)) +
  theme(strip.text.x = element_text(size = 0)) +
  theme(legend.position=c(0.8, 0.2),
        legend.key.size = unit(0.7, 'cm'),
        legend.text = element_text(size=12.5)) #remove facet label

spike_upperspike = plot_cells(spikes_cds.test,
                              genes=upper_spike_markers, #seed gene is the first row of the markers file
                              label_cell_groups=FALSE,
                              label_leaves=F,
                              label_branch_points=F,
                              graph_label_size=1.5,
                              show_trajectory_graph=F,
                              cell_size=2) +
  xlab("") + ylab("") + 
  theme(axis.ticks = element_blank()) +
  theme(legend.position="none") +
  scale_color_viridis_c(option = "inferno",name="expression") +
  ggtitle("BANF1_CAR_WT")+ #continuous
  theme(axis.text=element_text(size=0,color='black'),
        axis.title=element_text(size=0,color='black')) + theme(plot.title = element_text(size = 15, vjust=0)) +
  theme(strip.text.x = element_text(size = 0)) +
  theme(legend.position="none") #remove facet label


big_lowerspike = plot_cells(big_cds,
                            genes=lower_spike_markers, #seed gene is the first row of the markers file
                            label_cell_groups=FALSE,
                            label_leaves=F,
                            label_branch_points=F,
                            graph_label_size=1.5,
                            show_trajectory_graph=F) +
  xlab("") + ylab("") + 
  theme(axis.ticks = element_blank()) +
  theme(legend.position="none") +
  scale_color_viridis_c(option = "inferno",name="expression") +
  ggtitle("SCC Krt15")+ #continuous
  theme(axis.text=element_text(size=0,color='black'),
        axis.title=element_text(size=0,color='black')) + theme(plot.title = element_text(size = 15, vjust=0)) +
  theme(strip.text.x = element_text(size = 0)) +
  theme(legend.position=c(0.8, 0.2),
        legend.key.size = unit(0.7, 'cm'),
        legend.text = element_text(size=12.5)) #remove facet label

spike_lowerspike = plot_cells(spikes_cds.test,
                              genes=lower_spike_markers, #seed gene is the first row of the markers file
                              label_cell_groups=FALSE,
                              label_leaves=F,
                              label_branch_points=F,
                              graph_label_size=1.5,
                              show_trajectory_graph=F,
                              cell_size=2) +
  xlab("") + ylab("") + 
  theme(axis.ticks = element_blank()) +
  theme(legend.position="none") +
  scale_color_viridis_c(option = "inferno",name="expression") +
  ggtitle("SCC Krt15")+ #continuous
  theme(axis.text=element_text(size=0,color='black'),
        axis.title=element_text(size=0,color='black')) + theme(plot.title = element_text(size = 15, vjust=0)) +
  theme(strip.text.x = element_text(size = 0)) +
  theme(legend.position="none") #remove facet label


png("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/spike_analysis/combined_spike_analysis/spikes_alone/markers_v2.png",
    width=14,height=14,units="in",res=400)
print(
  multiplot(big_upperspike, spike_upperspike,
            big_lowerspike, spike_lowerspike,
            cols=2)
)
dev.off()

#############################################################################################
#############################################################################################
#DGE Analysis: https://medium.com/biosyntax/using-limma-to-find-differentially-expressed-genes-1c12c905e3b1

#Combine NSK and SCC microarray data together
bulk_dat = as.matrix(cbind(normal_wt_dat,tumor_wt_dat))

#Make gene names
row.names(bulk_dat) = make.names(aim2probedat$SYMBOL, unique=T)
#bulk_dat = log(bulk_dat) #log transform bulk data
bulk_dat[1:10,1:10]

# Create a design matrix of NSk and SCC
design <- model.matrix(~ 0+factor(c(rep(1,ncol(normal_wt_dat)),rep(2,ncol(tumor_wt_dat)))))
# assign the column names
colnames(design) <- c("NSk", "SCC")

# Make contrast matrix
cont_matrix <- makeContrasts(NSkvsSCC = NSk-SCC, levels=design)

# Fit the expression matrix to a linear model
fit <- lmFit(bulk_dat, design)
# Compute contrast
fit_contrast <- contrasts.fit(fit, cont_matrix)
# Bayes statistics of differential expression
# *There are several options to tweak!*
fit_contrast <- eBayes(fit_contrast)
# Generate a vocalno plot to visualize differential expression
volcanoplot(fit_contrast)
# Generate a list of top 100 differentially expressed genes
top_genes <- topTable(fit_contrast, number = 100, adjust = "BH")
# Summary of results (number of differentially expressed genes)
result <- decideTests(fit_contrast)
summary(result)

#Extract results
result_df = data.frame( fit_contrast$coefficients, fit_contrast$t, fit_contrast$p.value)
names(result_df) = c("coef","t","p")
result_df = result_df[order(result_df$p),]
result_df$p_adj = p.adjust(result_df$p, method="BH") #adjust p-value
result_df$gene = row.names(result_df)
result_df$coef = result_df$coef*-1 #coefficients are in the wrong direction so multiply by -1

#Plot out
topfc = subset(result_df, coef < -2.5 | coef > 2) #get high FC
top_p = result_df[c(1:150),] #get high significance
top_p = top_p[-c(na.omit(match(topfc$gene,top_p$gene))),] #get rid of top_p that were already grabbed in the topfc
nrow(topfc)
nrow(top_p)

volcano_plot = ggplot() +
  geom_point(data=result_df, aes(coef,(-log10(p_adj))), alpha=0.2, color="black") +
  geom_point(data=topfc, aes(coef,(-log10(p_adj))), alpha=0.9, color="darkred") +
  geom_point(data=top_p, aes(coef,(-log10(p_adj))), alpha=0.9, color="darkred") +
  geom_text_repel(data=topfc, aes(coef,(-log10(p_adj)),label=gene),box.padding = 0.2, max.overlaps = 25) + #add labels
  geom_text_repel(data=top_p, aes(coef,(-log10(p_adj)),label=gene),box.padding = 0.2, max.overlaps = 25) + #add labels
  
  geom_vline(xintercept=0, color="black") +
  xlab("log2(FC)") +
  ylab("-log10(p)") +
  ggtitle("LIMMA NSk vs SCC") +
  theme(axis.text=element_text(size=15,color='black'),
        axis.title=element_text(size=15,color='black')) +
  theme(plot.title = element_text(size = 20, vjust=2)) +
  theme(strip.background = element_rect(fill = 'white', color='black', size=1)) +
  theme(panel.background=element_blank(),
        panel.grid.major=element_blank(),
        panel.grid.minor=element_blank()) +
  #scale_y_continuous(breaks=number_ticks(6)) +
  #scale_x_continuous(breaks=number_ticks(6)) +
  theme(axis.line = element_line(colour = 'black'))


ggsave("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/DEG_bulk/WT_nsk_scc/wt_nsk_vs_scc.png",
       plot=volcano_plot,
       width = 14, height = 14, units = "in")


#--------------------------------------------------------------------------------------#
#--------------------------------------------------------------------------------------#
#make eigengeenes of top 200 up DEG and top 200 down DEG by FC
aim2probedat$gene_name = make.names(aim2probedat$SYMBOL, unique=T)
degs_to_plot = 100
eigengene_size = 100

plot_degs_in_scmanifold = function(seed_gene, output_direct) {
  
  for(i in 1:length(seed_gene)) {
    
    tryCatch({ #suppress error
      #i=1
      print(i)
      
      #Get the correct match_index
      if( seed_gene[i]=="Prom1" ) {match_index = 12747} else {
        match_index = match(seed_gene[i], aim2probedat$gene_name) #match seed gene to its place in the probe data
      }
      
      #---------------------------------------#
      #Seed gene
      #make marker df for grahing eigengene
      seed_gene_to_plot = alias2SymbolTable(aim2probedat[match_index,]$SYMBOL, species = "Mm")
      
      seed_gene_plot = 
        plot_cells(big_cds,
                   genes=seed_gene_to_plot,
                   label_cell_groups=FALSE,
                   label_leaves=F,
                   label_branch_points=F,
                   graph_label_size=1.5,
                   show_trajectory_graph=F) +
        theme(axis.text=element_text(size=0,color='black'),
              axis.title=element_text(size=0,color='black')) + 
        theme(axis.ticks = element_blank()) +
        theme(plot.title = element_text(size = 15, vjust=0)) +
        scale_color_viridis_c(option = "inferno",name="log\nexpression") +
        ggtitle(paste("Seed gene", seed_gene[i],sep=" "))+ #continuous
        theme(strip.text.x = element_text(size = 0))  +
        theme(legend.position=c(0.8, 0.2),
              legend.key.size = unit(1, 'cm'),
              legend.text = element_text(size=12.5)) #remove facet label #remove facet label
      
      
      #---------------------------------------#
      #POSITIVE Nsk
      symbol_match = data.frame(aim2probedat$SYMBOL, rhodf[,match_index], p_df[,match_index]) #put together gene names, expression data, and p-vals
      symbol_match = symbol_match[order(-symbol_match$rhodf...match_index.),] #order from high to low gene values
      names(symbol_match) = c("gene","rho","p")
      
      #make marker df for grahing eigengene
      marker_df = data.frame(unique(alias2SymbolTable(symbol_match$gene[1:(eigengene_size)], species = "Mm")),"eigengene","eigengene")
      names(marker_df) = c("gene","cell_type","module")
      marker_df = na.omit(marker_df)
      
      classic_eigengene_expression_plot_pos_nsk = 
        plot_cells(big_cds,
                   genes=marker_df,
                   label_cell_groups=FALSE,
                   label_leaves=F,
                   label_branch_points=F,
                   graph_label_size=1.5,
                   show_trajectory_graph=F) +
        theme(axis.text=element_text(size=0,color='black'),
              axis.title=element_text(size=0,color='black')) + 
        theme(axis.ticks = element_blank()) +
        theme(plot.title = element_text(size = 15, vjust=0)) +
        scale_color_viridis_c(option = "inferno",name="integrated\nexpression") +
        ggtitle(paste("POS NSk", seed_gene[i],sep=" "))+ #continuous
        theme(strip.text.x = element_text(size = 0))  +
        theme(legend.position=c(0.8, 0.2),
              legend.key.size = unit(1, 'cm'),
              legend.text = element_text(size=12.5)) #remove facet label #remove facet label
      
      
      #---------------------------------------#
      #CARCINOMA EIGENGENES
      symbol_match = data.frame(aim2probedat$SYMBOL, tumor_rhodf[,match_index], tumor_p_df[,match_index]) #put together gene names, expression data, and p-vals
      symbol_match = symbol_match[order(-symbol_match$tumor_rhodf...match_index.),] #order from high to low gene values
      names(symbol_match) = c("gene","rho","p")
      
      #---------------------------------------#
      #POSITIVE SCC
      #make marker df for grahing eigengene
      marker_df = data.frame(unique(alias2SymbolTable(symbol_match$gene[1:(eigengene_size)], species = "Mm")),paste(seed_gene[i]," Normal skin",sep=""),paste(seed_gene[i]," Normal skin",sep=""))
      names(marker_df) = c("gene","cell_type","module")
      marker_df = na.omit(marker_df)
      
      classic_eigengene_expression_plot_pos_scc = 
        plot_cells(big_cds,
                   genes=marker_df,
                   label_cell_groups=FALSE,
                   label_leaves=F,
                   label_branch_points=F,
                   graph_label_size=1.5,
                   show_trajectory_graph=F) +
        theme(axis.text=element_text(size=0,color='black'),
              axis.title=element_text(size=0,color='black')) + 
        theme(axis.ticks = element_blank()) +
        theme(plot.title = element_text(size = 15, vjust=0)) +
        scale_color_viridis_c(option = "inferno",name="integrated\nexpression") +
        ggtitle(paste("POS SCC", seed_gene[i],sep=" "))+ #continuous
        theme(strip.text.x = element_text(size = 0))  +
        theme(legend.position=c(0.8, 0.2),
              legend.key.size = unit(1, 'cm'),
              legend.text = element_text(size=12.5)) #remove facet label #remove facet label
      
      #-----------------------------------------------------#
      #NEGATIVE SCC
      marker_df = data.frame(unique(alias2SymbolTable(symbol_match$gene[(nrow(symbol_match)-eigengene_size):nrow(symbol_match)], species = "Mm")),paste(seed_gene[i]," Normal skin",sep=""),paste(seed_gene[i]," Normal skin",sep=""))
      names(marker_df) = c("gene","cell_type","module")
      marker_df = na.omit(marker_df)
      
      classic_eigengene_expression_plot_neg_scc = 
        plot_cells(big_cds,
                   genes=marker_df,
                   label_cell_groups=FALSE,
                   label_leaves=F,
                   label_branch_points=F,
                   graph_label_size=1.5,
                   show_trajectory_graph=F) +
        theme(axis.text=element_text(size=0,color='black'),
              axis.title=element_text(size=0,color='black')) + 
        theme(axis.ticks = element_blank()) +
        theme(plot.title = element_text(size = 15, vjust=0)) +
        scale_color_viridis_c(option = "inferno",name="integrated\nexpression") +
        ggtitle(paste("NEG SCC", seed_gene[i],sep=" "))+ #continuous
        theme(strip.text.x = element_text(size = 0))  +
        theme(legend.position=c(0.8, 0.2),
              legend.key.size = unit(1, 'cm'),
              legend.text = element_text(size=12.5)) #remove facet label #remove facet label
      
      
      #-----------#
      #print out binary and classic plots
      png(paste(output_direct,i,"_",seed_gene[i],".png",sep=""),
          width=15,height=15,units="in",res=200)
      
      print(multiplot(seed_gene_plot,
                      classic_eigengene_expression_plot_pos_nsk,
                      classic_eigengene_expression_plot_pos_scc,
                      classic_eigengene_expression_plot_neg_scc,
                      cols=2))
      
      dev.off()
      
    },  error=function(e){})
    
  }#end positive and negative eigengene loop
} #end function to do all this shit
#--------------------------------------------------------------------------------------#
#--------------------------------------------------------------------------------------#

#Plot UP-REGULATED DEGS
seed_gene = result_df[order(-result_df$coef),]$gene[1:degs_to_plot]
output_direct = "/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/DEG_bulk/WT_nsk_scc/DEG_up_sc_manifold/"
plot_degs_in_scmanifold(seed_gene, output_direct)


#Plot DOWN-REGULATED DEGS
seed_gene = result_df[order(result_df$coef),]$gene[1:degs_to_plot]
output_direct = "/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/DEG_bulk/WT_nsk_scc/DEG_down_sc_manifold/"
plot_degs_in_scmanifold(seed_gene, output_direct)

#--------------------------------------------------------------------------------------#
#--------------------------------------------------------------------------------------#
#--------------------------------------------------------------------------------------#
#--------------------------------------------------------------------------------------#
#--------------------------------------------------------------------------------------#
#STRING analysis: 
# Each group of genes (MIF vs. NLF, MAF vs. NLF, and MAF vs. MIF) were subjected to PPIs analysis using the STRING platform (Szklarczyk et al., 2017). The minimum confidence of interaction was defined as confidence > 0.3 and connections based on text-mining were excluded. Groups of under four genes were excluded, narrowing the size of each group by ~50%.

#Install
BiocManager::install("STRINGdb", version="3.10", force=T)
install.packages("igraph")

#Load library
library(STRINGdb)

string_db <- STRINGdb$new( version="10", species=10090,
                           score_threshold=200, input_directory="")

#Query string for a specific protein
lgr6 = string_db$mp( "trp53" )

string_db$get_neighbors( c(lgr6) )
string_db$get_neighbors(lgr6)


library(igraph)
#--------------------------------------------------------------------------------------#
#--------------------------------------------------------------------------------------#
#--------------------------------------------------------------------------------------#

#Machine learning: https://www.r-bloggers.com/2021/08/machine-learning-strategy-part-1/

library(WGCNA)

#Check fold change with averages
#Check fold change with averages

row.names(normal_wt_dat) = make.names(aim2probedat$SYMBOL, unique=T)
row.names(tumor_wt_dat) = make.names(aim2probedat$SYMBOL, unique=T)


do.call(rbind, lapply(1:nrow(result_df), function(i){
  
  #i=1
  result_df[i,]$gene
  nsk_mean = mean(as.numeric(normal_wt_dat[ match(result_df[i,]$gene, row.names(normal_wt_dat)), ]))
  tumor_mean = mean(as.numeric(tumor_wt_dat[ match(result_df[i,]$gene, row.names(tumor_wt_dat)), ]))
  fc = tumor_mean / nsk_mean
  result_df[i,]
}))



#--------------------------------------------------------------------------------------#
#--------------------------------------------------------------------------------------#

#Make chromomap input files from eigengene correlations

#Use NCBI to get positions
ncbi_names = read.delim(file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Balmains_networks/mouse_genome_gene_result_reduced.txt",header=T)

seed_gene = c("Lgr5",
              "Bmi1","Sox4","Lrig1",
              "Lgr6","Sox9",
              "Sox2","Prom1","Foxa1","Krt15","Krt19","Pitx1","Psca","Prom2","Klf5",
              "Cd44")
eigengene_size = 100

#Loop through seed genes and get those sweet sweet tumor eigengenes
tumor_rhos = do.call(rbind, lapply(1:length(seed_gene), function(i){
  #i=1
  print(i)
  if( seed_gene[i]=="Prom1" ) {match_index = 12747} else {
    match_index = match(seed_gene[i], aim2probedat$SYMBOL) #match seed gene to its place in the matrix
  }
  symbol_match = data.frame(aim2probedat$SYMBOL, tumor_rhodf[,match_index]) #put together gene names, expression data, and p-vals
  symbol_match = symbol_match[order(-symbol_match$tumor_rhodf...match_index.),] #order from high to low gene values
  names(symbol_match) = c("gene","rho")
  symbol_match$seed_gene = seed_gene[i]
  symbol_match = symbol_match[-1,] #drop the top-correlated gene
  
  symbol_match$official_symbol = alias2SymbolTable(symbol_match$gene, species = "Mm") #get official symbol
  chromo_map_dat = data.frame( ncbi_names[match(symbol_match$official_symbol,ncbi_names$Symbol),c(2,5,7,8)], symbol_match$rho) #get matches and re-order
  chromo_map_dat = na.omit(chromo_map_dat)
  
  
  write.table(chromo_map_dat,
              file=paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/chromoMap/input_files/",seed_gene[i],".txt",sep=""),
              col.names = F, row.names = F, quote=F, sep="\t"
  )
  
}))

#--------------------------------------------------------------------------------------#
#--------------------------------------------------------------------------------------#
#--------------------------------------------------------------------------------------#
#--------------------------------------------------------------------------------------#
#WGCNA on our data

#Tutorial: https://bioinformaticsworkbook.org/tutorials/wgcna.html#gsc.tab=0


# download from ze internet
# cd /Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/code/WGCNA\ tutorial/tutorial_data
# wget ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE61nnn/GSE61333/suppl/GSE61333_ligule_count.txt.gz
# curl -o GSE61333_ligule_count.txt.gz ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE61nnn/GSE61333/suppl/GSE61333_ligule_count.txt.gz
# gunzip GSE61333_ligule_count.txt.gz

library(tidyverse)     # tidyverse will pull in ggplot2, readr, other useful libraries
library(magrittr)      # provides the %>% operator
library(WGCNA)   

#1. read in the data
aim2dat = read.table(file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/bulk_network_data/Aim2-4/AIM2_AIM4_BS_CARC_HFC_NOTED_ENTREZ_03_16_15_EXPRESSION.txt",header=T)
aim2dat[1:10,1:10]
dim(aim2dat)
aim2phenodat = read.table(file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/bulk_network_data/Aim2-4/AIM2_AIM4_BS_CARC_HFC_NOTED_ENTREZ_03_16_15_PHENOTYPE_DATA.txt",header=T)
aim2phenodat[1:10,1:10]
dim(aim2phenodat)
aim2probedat = read.table(file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/bulk_network_data/Aim2-4/AIM2_AIM4_BS_CARC_HFC_NOTED_ENTREZ_03_16_15_PROBE_DATA.txt",header=T)
aim2probedat[1:10,]


#2a. subset to specific samples and get sample IDs
# normal_wt_ids = subset(aim2phenodat, TYPE=="NORMAL" & GENOTYPE=="WT") my original settings cf Allan's settings
normal_wt_ids = subset(aim2phenodat, TISSUE=="BACK" & GENOTYPE_INFERRED=="WT") #use Allan's filter settings
normal_wt_ids = normal_wt_ids$IDENTIFIER
normal_wt_ids = gsub("-",".",normal_wt_ids)  #gsub hyphens to decimals  bc R doesn't allow hyphens in column names and changes hyphens to decimals in the aim2dat dataframe
length(normal_wt_ids)

#2b. subset to specific samples and get sample IDs
tumor_wt_ids = subset(aim2phenodat, TISSUE=="CARC" & GENOTYPE_INFERRED=="WT")
tumor_wt_ids = tumor_wt_ids$IDENTIFIER
tumor_wt_ids = gsub("-",".",tumor_wt_ids)  #gsub hyphens to decimals  bc R doesn't allow hyphens in column names and changes hyphens to decimals in the aim2dat dataframe
length(tumor_wt_ids) -71

#3a. subset expression dat from sample IDs
normal_wt_dat = subset(aim2dat, select=normal_wt_ids)
#normal_wt_dat.v2 = aim2dat[,match(normal_wt_ids, colnames(aim2dat))] #alternate way of selecting correct columns

#3b. subset expression dat from sample IDs
tumor_wt_dat = subset(aim2dat, select=tumor_wt_ids)
#tumor_wt_dat.v2 = aim2dat[,match(tumor_wt_ids, colnames(aim2dat))] #alternate way of selecting correct columns

#Add row unique gene nemes to rows
row.names(tumor_wt_dat) = make.unique(aim2probedat$SYMBOL, sep="_")
tumor_wt_dat_t = t(tumor_wt_dat) #Transpose for WGCNA
tumor_wt_dat_t[1:10,1:10]

# Choose a set of soft-thresholding powers
powers = c(c(1:10), seq(from = 12, to = 20, by = 2))

# Call the network topology analysis function
sft = pickSoftThreshold(
  tumor_wt_dat_t,             # <= Input data
  #blockSize = 30,
  powerVector = powers,
  verbose = 5
)

#-----------------------------------------#
#Explore scale parameters 

par(mfrow = c(1,2));
cex1 = 0.9;

plot(sft$fitIndices[, 1],
     -sign(sft$fitIndices[, 3]) * sft$fitIndices[, 2],
     xlab = "Soft Threshold (power)",
     ylab = "Scale Free Topology Model Fit, signed R^2",
     main = paste("Scale independence")
)
text(sft$fitIndices[, 1],
     -sign(sft$fitIndices[, 3]) * sft$fitIndices[, 2],
     labels = powers, cex = cex1, col = "red"
)
abline(h = 0.90, col = "red")
plot(sft$fitIndices[, 1],
     sft$fitIndices[, 5],
     xlab = "Soft Threshold (power)",
     ylab = "Mean Connectivity",
     type = "n",
     main = paste("Mean connectivity")
)
text(sft$fitIndices[, 1],
     sft$fitIndices[, 5],
     labels = powers,
     cex = cex1, col = "red")

input_mat = tumor_wt_dat_t
picked_power = 6
temp_cor <- cor       
cor <- WGCNA::cor         # Force it to use WGCNA cor function (fix a namespace conflict issue)
netwk <- blockwiseModules(input_mat,                # <= input here
                          
                          # == Adjacency Function ==
                          power = picked_power,                # <= power here
                          networkType = "signed",
                          
                          # == Tree and Block Options ==
                          deepSplit = 2,
                          pamRespectsDendro = F,
                          # detectCutHeight = 0.75,
                          minModuleSize = 30,
                          maxBlockSize = 4000,
                          
                          # == Module Adjustments ==
                          reassignThreshold = 0,
                          mergeCutHeight = 0.25,
                          
                          # == TOM == Archive the run results in TOM file (saves time)
                          saveTOMs = T,
                          saveTOMFileBase = "ER",
                          
                          # == Output Options
                          numericLabels = T,
                          verbose = 3)

# Convert labels to colors for plotting
mergedColors = labels2colors(netwk$colors)
# Plot the dendrogram and the module colors underneath
plotDendroAndColors(
  netwk$dendrograms[[1]],
  mergedColors[netwk$blockGenes[[1]]],
  "Module colors",
  dendroLabels = FALSE,
  hang = 0.03,
  addGuide = TRUE,
  guideHang = 0.05 
)

module_df <- data.frame(
  gene_id = names(netwk$colors),
  colors = labels2colors(netwk$colors),
  module_id = netwk$colors
)

module_df = module_df[order(module_df$module_id),] #Re-order module assignments from most populated to least
gene_module_summary = data.frame(table(module_df$module_id)) #get the number of genes in each module
names(gene_module_summary) = c("module_id","gene_no")

#-----------------------------------------------------------------#
#Get the module assignemtns of my beloved genes MAH BELOVERDZ genes
seed_gene = c("Lgr5",
              "Bmi1","Sox4","Lrig1",
              "Lgr6","Sox9",
              "Sox2","Prom1","Foxa1","Krt15","Krt19","Pitx1","Psca","Prom2","Klf5",
              "Cd44")
seed_gene_modules = do.call(rbind, lapply(1:length(seed_gene), function(k){
  #k=1
  print(k)
  module_df[grep(seed_gene[k],module_df$gene_id),]
}))

#-----------------------------------------------------------------#
#Plot specific pairwise eigengene values
eigengene_df = netwk$MEs

str(eigengene_df)
#Plot eigengene 3 vs all the others

for(i in 1:ncol(eigengene_df)) {
  #i=1
  print(i)
  eigen_plot = ggplot(data=eigengene_df, aes(x=eigengene_df$ME3, y=eigengene_df[,i])) +
    geom_point(alpha=0.4,color="black") +
    geom_smooth(method='lm', color="black", se=F, size=0.7) +
    stat_ellipse(color="orangered3") +
    theme_classic() +
    xlab("ME3") +
    ylab(names(eigengene_df)[i]) +
    ggtitle(paste("Module 3 Eigengene vs ",names(eigengene_df)[i],
                  ", r = ",
                  round(cor(eigengene_df$ME3,eigengene_df[,i],method="pearson"),3),
                  sep=""))
  
  png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/WGCNA/aim2_tumor_wt/module3_eigengene_pairwise/module3_vs_",
            names(eigengene_df)[i],"_eeigengene.png",sep=""),
      width=6,height=6,units="in",res=300)
  print(eigen_plot)
  dev.off()
  
}

#Pairsplot of all eigengenes
library(GGally)
#Fig. 5
png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/WGCNA/aim2_tumor_wt/module3_eigengene_pairwise/0_pairwise_eigengene.png",
          sep=""),
    width=15,height=15,units="in",res=300)

cor = ggcorr(eigengene_df, label = TRUE, label_round=2,
             low="darkred",high="steelblue",hjust=0.94,geom="circle",min_size=16,max_size=16,
             size = 7,layout.exp=5,
             legend.position = "right",
             legend.size = 20)
print(cor)
dev.off()


#-----------------------------------------------------------------#
#Write out results
write.table(seed_gene_modules, 
            file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/WGCNA/aim2_tumor_wt/seed_gene_modules.txt",
            quote=F, row.names=F,sep="\t")
write.table(gene_module_summary, 
            file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/WGCNA/aim2_tumor_wt/all_gene_modules_summary.txt",
            quote=F, row.names=F,sep="\t")
write.table(module_df, 
            file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/WGCNA/aim2_tumor_wt/all_gene_module_assignments.txt",
            quote=F, row.names=F,sep="\t")

#-----------------------------------------------------------------#
#Do GO Enrichment on these modules
los_modules_ids = unique(module_df$module_id)
los_modules_colors = unique(module_df$colors)

combined_go_res = do.call(rbind, lapply(1:length(los_modules_ids), function(i) {
  tryCatch({ #suppress error
    
    #i=20
    print(i)
    
    #create topGO object
    geneUniverse <- unique(aim2probedat$SYMBOL)
    genesOfInterest <- subset(module_df, module_df$module_id==los_modules_ids[i])$gene_id #get genes within a module
    geneList <- factor(as.integer(geneUniverse %in% genesOfInterest))
    names(geneList) <- geneUniverse
    GOdata <- new("topGOdata",
                  ontology = "BP",
                  allGenes=geneList,
                  #geneSel = selection,
                  annot = annFUN.org, 
                  mapping = "org.Mm.eg.db",
                  ID = "symbol")
    
    
    resultKS <- runTest(GOdata, algorithm = "weight01", statistic = "fisher") #Fisher exact test
    tab <- GenTable(GOdata, raw.p.value = resultKS, topNodes = length(resultKS@score), numChar = 120)
    tab$module_id = as.character(los_modules_ids[i]) #add module id column
    tab$module_color = subset(module_df, module_id==los_modules_ids[i])$colors[1] #add module color coulmn
    tab = subset(tab, raw.p.value <=0.01) #keep only sig results 
    
    #Get the genes that are contributing to a term
    go_candidates = genesOfInterest
    selcGenes <- genesInTerm(GOdata, whichGO=tab$GO.ID)
    genes_in_go = do.call(rbind, lapply(1:length(selcGenes), function(i){
      #i=1
      data.frame(names(selcGenes)[i],  paste( c(na.omit(selcGenes[[i]][match(go_candidates,  selcGenes[[i]])])) , collapse=", ") )
    }))
    names(genes_in_go) = c("GO_ID","sig_genes_in_term")
    tab = cbind(tab, genes_in_go)
    
    #Figures to show enrichment path
    pdf(file=paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/WGCNA/aim2_tumor_wt/module_GO/figs/module_",
                   los_modules_ids[i],"_GO.pdf",sep=""),width=12,height=8)
    print(showSigOfNodes(GOdata, score(resultKS), firstSigNodes = 10, useInfo = "def"))
    print(showSigOfNodes(GOdata, score(resultKS), firstSigNodes = 5, useInfo = "def"))
    dev.off()
    
    write.table(tab, file=paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/WGCNA/aim2_tumor_wt/module_GO/module_",
                                los_modules_ids[i],"_GO_enrichment.txt",sep=""),
                quote=F, row.names=F,sep="\t")
    
    return(tab)
  },  error=function(e){})
  
})) #end topGO loop plz

write.table(combined_go_res, 
            file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/WGCNA/aim2_tumor_wt/module_GO/0_combined_GO_analysis.txt",
            quote=F, row.names=F,sep="\t")

#-----------------------------------------------------------------#
#Plot the modules' eigengene expression in the scRNAseq map
for(i in 1:length(los_modules_ids)) {
  tryCatch({ #suppress error
    
    #i=16
    print(i)
    
    #create topGO object
    marker_df <- data.frame(subset(module_df, module_df$module_id==los_modules_ids[i])$gene_id, "module","module") #get genes within a module
    names(marker_df) = c("gene","cell_type","module")
    
    agexp_plot = plot_cells(big_cds,
                            genes=marker_df,
                            label_cell_groups=FALSE,
                            label_leaves=F,
                            label_branch_points=F,
                            graph_label_size=1.5,
                            show_trajectory_graph=F) +
      xlab("Dim. 1") + ylab("Dim. 2") + 
      theme(legend.position="bottom") +
      scale_color_viridis_c(option = "inferno",name="") +
      ggtitle(paste("Module ", los_modules_ids[i]," ",subset(module_df, module_id==los_modules_ids[i])$colors[1],
                    " (N=",nrow(marker_df),")",
                    sep=""))+ #continuous
      theme(axis.text=element_text(size=0,color='black'),
            axis.title=element_text(size=0,color='black'),
            axis.ticks = element_line(size=0)) + 
      theme(plot.title = element_text(size = 15, vjust=1)) +
      theme(strip.text = element_text(size = 0))
    
    png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/WGCNA/aim2_tumor_wt/module_eigengene_manifolds/module_",
              los_modules_ids[i],"_eigengene_expression.png",sep=""),
        width=6,height=6,units="in",res=300)
    print(agexp_plot)
    dev.off()
    
  },  error=function(e){})
  
} #End module eigengene loop

#-----------------------------------------------------------------#
#-----------------------------------------------------------------#
#Okay now find the seed gene networks within these broader networks

modules_of_interest = unique(seed_gene_modules$colors)

for(i in 1:length(modules_of_interest)) {
  
  #i=4
  genes_of_interest = module_df %>%
    subset(colors %in% modules_of_interest[i]) #extract gene ID's annotated within a module
  
  expr_of_interest = tumor_wt_dat_t[,genes_of_interest$gene_id] #now extract the gene expression data
  expr_of_interest[1:5,1:5]
  
  # Only recalculate Topological Overemap Matrix (TOM) for modules of interest (faster, altho there's some online discussion if this will be slightly off)
  TOM = TOMsimilarityFromExpr(expr_of_interest,
                              power = picked_power)
  
  # Add gene names to row and columns
  row.names(TOM) = colnames(expr_of_interest)
  colnames(TOM) = colnames(expr_of_interest)
  
  edge_list = data.frame(TOM) %>%
    mutate(
      gene1 = row.names(.)
    ) %>%
    pivot_longer(-gene1) %>%
    dplyr::rename(gene2 = name, correlation = value) %>%
    unique() %>%
    subset(!(gene1==gene2)) %>%
    mutate(
      module1 = module_df[gene1,]$colors,
      module2 = module_df[gene2,]$colors
    )
  
  #Keep only the top 1% of edges
  # edge_list = subset(edge_list, correlation>=(quantile(edge_list$correlation, 0.99)))
  # tail(edge_list)
  
  write_delim(edge_list,
              file = paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/WGCNA/aim2_tumor_wt/cytoscape_input_files/module_",
                           subset(module_df, colors==modules_of_interest[i])$module_id[1],
                           "_",
                           modules_of_interest[i],
                           "_",
                           "edgelist.tsv",sep=""),
              delim = "\t")
  
}

seed_gene_edgelist = subset(edge_list, gene1==seed_gene)
table(seed_gene_edgelist$gene1)
table(edge_list$gene1)

summary(seed_gene_edgelist$correlation)

#Compare overlaps seed gene networks from WGCNA compared to basic correlations
cor_eigengenes = read.delim(file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/WGCNA/seed_gene_rank_cor_mats.txt",
                            header=T)

seed_eigen_colnames = data.frame(t(data.frame(strsplit(colnames(cor_eigengenes),"_"))))
names(seed_eigen_colnames) = c("gene_id","eigengene_type")

seed_gene_modules_sub = subset(seed_gene_modules, module_id==3) #pick a particular module to find overlapping genes in
module3_overlaps = do.call(rbind, lapply(1:nrow(seed_gene_modules_sub), function(i){
  #i=5
  print(i)
  tryCatch({ #suppress error
    
    wgcna_module_genes = subset(seed_gene_edgelist, gene1==seed_gene_modules_sub$gene_id[i])$gene2
    rank_genes = cor_eigengenes[,match (paste(seed_gene_modules_sub$gene_id[i],"_Car",sep=""),  names(cor_eigengenes) )] #get the correct eigengene column
    interesecting_genes = intersect(wgcna_module_genes ,data.frame(t(data.frame(strsplit(rank_genes," "))))[,1] ) #find interesection between WGCNA and our eigengenes
    wgcna_interesection = data.frame(seed_gene_modules_sub[i,],length(wgcna_module_genes),length(interesecting_genes),paste(interesecting_genes,collapse=", "))
    
    names(wgcna_interesection) = c("seed_gene","colors","module_id","wgcna_edge","N_intersection_rankcor_eigengene","interesecting_genes")
    wgcna_interesection
  },  error=function(e){})
  
}))


#-----------------------------------------------------------------#
#-----------------------------------------------------------------#
#-----------------------------------------------------------------#
#-----------------------------------------------------------------#
#-----------------------------------------------------------------#
#-----------------------------------------------------------------#
#-----------------------------------------------------------------#

#WGCNA on normal skin healthy tissue
#Add row unique gene nemes to rows
row.names(normal_wt_dat) = make.unique(aim2probedat$SYMBOL, sep="_")
normal_wt_dat = t(normal_wt_dat) #Transpose for WGCNA
normal_wt_dat[1:10,1:10]

# Choose a set of soft-thresholding powers
powers = c(c(1:10), seq(from = 12, to = 20, by = 2))

# Call the network topology analysis function
sft = pickSoftThreshold(
  normal_wt_dat,             # <= Input data
  #blockSize = 30,
  powerVector = powers,
  verbose = 5
)

#-----------------------------------------#
#Explore scale parameters 

par(mfrow = c(1,2));
cex1 = 0.9;

plot(sft$fitIndices[, 1],
     -sign(sft$fitIndices[, 3]) * sft$fitIndices[, 2],
     xlab = "Soft Threshold (power)",
     ylab = "Scale Free Topology Model Fit, signed R^2",
     main = paste("Scale independence")
)
text(sft$fitIndices[, 1],
     -sign(sft$fitIndices[, 3]) * sft$fitIndices[, 2],
     labels = powers, cex = cex1, col = "red"
)
abline(h = 0.90, col = "red")
plot(sft$fitIndices[, 1],
     sft$fitIndices[, 5],
     xlab = "Soft Threshold (power)",
     ylab = "Mean Connectivity",
     type = "n",
     main = paste("Mean connectivity")
)
text(sft$fitIndices[, 1],
     sft$fitIndices[, 5],
     labels = powers,
     cex = cex1, col = "red")

input_mat = normal_wt_dat
picked_power = 6
temp_cor <- cor       
cor <- WGCNA::cor         # Force it to use WGCNA cor function (fix a namespace conflict issue)
netwk <- blockwiseModules(input_mat,                # <= input here
                          
                          # == Adjacency Function ==
                          power = picked_power,                # <= power here
                          networkType = "signed",
                          
                          # == Tree and Block Options ==
                          deepSplit = 2,
                          pamRespectsDendro = F,
                          # detectCutHeight = 0.75,
                          minModuleSize = 30,
                          maxBlockSize = 4000,
                          
                          # == Module Adjustments ==
                          reassignThreshold = 0,
                          mergeCutHeight = 0.25,
                          
                          # == TOM == Archive the run results in TOM file (saves time)
                          saveTOMs = T,
                          saveTOMFileBase = "ER",
                          
                          # == Output Options
                          numericLabels = T,
                          verbose = 3)

# Convert labels to colors for plotting
mergedColors = labels2colors(netwk$colors)
# Plot the dendrogram and the module colors underneath
plotDendroAndColors(
  netwk$dendrograms[[1]],
  mergedColors[netwk$blockGenes[[1]]],
  "Module colors",
  dendroLabels = FALSE,
  hang = 0.03,
  addGuide = TRUE,
  guideHang = 0.05 
)

module_df <- data.frame(
  gene_id = names(netwk$colors),
  colors = labels2colors(netwk$colors),
  module_id = netwk$colors
)

module_df = module_df[order(module_df$module_id),] #Re-order module assignments from most populated to least
gene_module_summary = data.frame(table(module_df$module_id)) #get the number of genes in each module
names(gene_module_summary) = c("module_id","gene_no")

#-----------------------------------------------------------------#
#Get the module assignemtns of my beloved genes MAH BELOVERDZ genes
seed_gene = c("Lgr5",
              "Bmi1","Sox4","Lrig1",
              "Lgr6","Sox9",
              "Sox2","Prom1","Foxa1","Krt15","Krt19","Pitx1","Psca","Prom2","Klf5",
              "Cd44")
seed_gene_modules = do.call(rbind, lapply(1:length(seed_gene), function(k){
  #k=1
  print(k)
  module_df[grep(seed_gene[k],module_df$gene_id),]
}))

#-----------------------------------------------------------------#
#Plot specific pairwise eigengene values
eigengene_df = netwk$MEs

str(eigengene_df)
#Plot eigengene 3 vs all the others

for(i in 1:ncol(eigengene_df)) {
  #i=1
  print(i)
  eigen_plot = ggplot(data=eigengene_df, aes(x=eigengene_df$ME3, y=eigengene_df[,i])) +
    geom_point(alpha=0.4,color="black") +
    geom_smooth(method='lm', color="black", se=F, size=0.7) +
    stat_ellipse(color="orangered3") +
    theme_classic() +
    xlab("ME3") +
    ylab(names(eigengene_df)[i]) +
    ggtitle(paste("Module 3 Eigengene vs ",names(eigengene_df)[i],
                  ", r = ",
                  round(cor(eigengene_df$ME3,eigengene_df[,i],method="pearson"),3),
                  sep=""))
  
  png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/WGCNA/aim2_tumor_wt/aim2_normalskin_wt/module3_vs_",
            names(eigengene_df)[i],"_eeigengene.png",sep=""),
      width=6,height=6,units="in",res=300)
  print(eigen_plot)
  dev.off()
  
}

#Pairsplot of all eigengenes
library(GGally)
#Fig. 5
png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/WGCNA/aim2_normalskin_wt/0_pairwise_eigengene.png",
          sep=""),
    width=15,height=15,units="in",res=300)

cor = ggcorr(eigengene_df, label = TRUE, label_round=2,
             low="darkred",high="steelblue",hjust=0.94,geom="circle",min_size=16,max_size=16,
             size = 7,layout.exp=5,
             legend.position = "right",
             legend.size = 20)
print(cor)
dev.off()


#-----------------------------------------------------------------#
#Write out results
write.table(seed_gene_modules, 
            file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/WGCNA/aim2_normalskin_wt/seed_gene_modules.txt",
            quote=F, row.names=F,sep="\t")
write.table(gene_module_summary, 
            file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/WGCNA/aim2_normalskin_wt/all_gene_modules_summary.txt",
            quote=F, row.names=F,sep="\t")
write.table(module_df, 
            file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/WGCNA/aim2_normalskin_wt/all_gene_module_assignments.txt",
            quote=F, row.names=F,sep="\t")

#-----------------------------------------------------------------#
#Do GO Enrichment on these modules
los_modules_ids = unique(module_df$module_id)
los_modules_colors = unique(module_df$colors)

combined_go_res = do.call(rbind, lapply(1:length(los_modules_ids), function(i) {
  tryCatch({ #suppress error
    
    #i=1
    print(i)
    
    #create topGO object
    geneUniverse <- unique(aim2probedat$SYMBOL)
    genesOfInterest <- subset(module_df, module_df$module_id==los_modules_ids[i])$gene_id #get genes within a module
    geneList <- factor(as.integer(geneUniverse %in% genesOfInterest))
    names(geneList) <- geneUniverse
    GOdata <- new("topGOdata",
                  ontology = "BP",
                  allGenes=geneList,
                  #geneSel = selection,
                  annot = annFUN.org, 
                  mapping = "org.Mm.eg.db",
                  ID = "symbol")
    
    
    resultKS <- runTest(GOdata, algorithm = "weight01", statistic = "fisher") #Fisher exact test
    tab <- GenTable(GOdata, raw.p.value = resultKS, topNodes = length(resultKS@score), numChar = 120)
    tab$module_id = as.character(los_modules_ids[i]) #add module id column
    tab$module_color = subset(module_df, module_id==los_modules_ids[i])$colors[1] #add module color coulmn
    tab = subset(tab, raw.p.value <=0.01) #keep only sig results 
    
    #Get the genes that are contributing to a term
    go_candidates = genesOfInterest
    selcGenes <- genesInTerm(GOdata, whichGO=tab$GO.ID)
    genes_in_go = do.call(rbind, lapply(1:length(selcGenes), function(i){
      #i=1
      data.frame(names(selcGenes)[i],  paste( c(na.omit(selcGenes[[i]][match(go_candidates,  selcGenes[[i]])])) , collapse=", ") )
    }))
    names(genes_in_go) = c("GO_ID","sig_genes_in_term")
    tab = cbind(tab, genes_in_go)
    
    #Figures to show enrichment path
    pdf(file=paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/WGCNA/aim2_normalskin_wt/module_GO/figs/module_",
                   los_modules_ids[i],"_GO.pdf",sep=""),width=12,height=8)
    print(showSigOfNodes(GOdata, score(resultKS), firstSigNodes = 10, useInfo = "def"))
    print(showSigOfNodes(GOdata, score(resultKS), firstSigNodes = 5, useInfo = "def"))
    dev.off()
    
    write.table(tab, file=paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/WGCNA/aim2_normalskin_wt/module_GO/module_",
                                los_modules_ids[i],"_GO_enrichment.txt",sep=""),
                quote=F, row.names=F,sep="\t")
    
    return(tab)
  },  error=function(e){})
  
})) #end topGO loop plz

write.table(combined_go_res, 
            file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/WGCNA/aim2_normalskin_wt/module_GO/0_combined_GO_analysis.txt",
            quote=F, row.names=F,sep="\t")

#-----------------------------------------------------------------#
#Plot the modules' eigengene expression in the scRNAseq map
for(i in 1:length(los_modules_ids)) {
  tryCatch({ #suppress error
    
    #i=16
    print(i)
    
    #create topGO object
    marker_df <- data.frame(subset(module_df, module_df$module_id==los_modules_ids[i])$gene_id, "module","module") #get genes within a module
    names(marker_df) = c("gene","cell_type","module")
    
    agexp_plot = plot_cells(big_cds,
                            genes=marker_df,
                            label_cell_groups=FALSE,
                            label_leaves=F,
                            label_branch_points=F,
                            graph_label_size=1.5,
                            show_trajectory_graph=F) +
      xlab("Dim. 1") + ylab("Dim. 2") + 
      theme(legend.position="bottom") +
      scale_color_viridis_c(option = "inferno",name="") +
      ggtitle(paste("Module ", los_modules_ids[i]," ",subset(module_df, module_id==los_modules_ids[i])$colors[1],
                    " (N=",nrow(marker_df),")",
                    sep=""))+ #continuous
      theme(axis.text=element_text(size=0,color='black'),
            axis.title=element_text(size=0,color='black'),
            axis.ticks = element_line(size=0)) + 
      theme(plot.title = element_text(size = 15, vjust=1)) +
      theme(strip.text = element_text(size = 0))
    
    png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/WGCNA/aim2_normalskin_wt/module_eigengene_manifolds/module_",
              los_modules_ids[i],"_eigengene_expression.png",sep=""),
        width=6,height=6,units="in",res=300)
    print(agexp_plot)
    dev.off()
    
  },  error=function(e){})
  
} #End module eigengene loop

#-----------------------------------------------------------------#
#-----------------------------------------------------------------#
#Okay now find the seed gene networks within these broader networks

modules_of_interest = unique(seed_gene_modules$colors)

for(i in 1:length(modules_of_interest)) {
  
  #i=4
  genes_of_interest = module_df %>%
    subset(colors %in% modules_of_interest[i]) #extract gene ID's annotated within a module
  
  expr_of_interest = tumor_wt_dat_t[,genes_of_interest$gene_id] #now extract the gene expression data
  expr_of_interest[1:5,1:5]
  
  # Only recalculate Topological Overemap Matrix (TOM) for modules of interest (faster, altho there's some online discussion if this will be slightly off)
  TOM = TOMsimilarityFromExpr(expr_of_interest,
                              power = picked_power)
  
  # Add gene names to row and columns
  row.names(TOM) = colnames(expr_of_interest)
  colnames(TOM) = colnames(expr_of_interest)
  
  edge_list = data.frame(TOM) %>%
    mutate(
      gene1 = row.names(.)
    ) %>%
    pivot_longer(-gene1) %>%
    dplyr::rename(gene2 = name, correlation = value) %>%
    unique() %>%
    subset(!(gene1==gene2)) %>%
    mutate(
      module1 = module_df[gene1,]$colors,
      module2 = module_df[gene2,]$colors
    )
  
  #Keep only the top 1% of edges
  # edge_list = subset(edge_list, correlation>=(quantile(edge_list$correlation, 0.99)))
  # tail(edge_list)
  
  write_delim(edge_list,
              file = paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/WGCNA/aim2_tumor_wt/cytoscape_input_files/module_",
                           subset(module_df, colors==modules_of_interest[i])$module_id[1],
                           "_",
                           modules_of_interest[i],
                           "_",
                           "edgelist.tsv",sep=""),
              delim = "\t")
  
}

seed_gene_edgelist = subset(edge_list, gene1==seed_gene)
table(seed_gene_edgelist$gene1)
table(edge_list$gene1)

summary(seed_gene_edgelist$correlation)

#Compare overlaps seed gene networks from WGCNA compared to basic correlations
cor_eigengenes = read.delim(file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/WGCNA/seed_gene_rank_cor_mats.txt",
                            header=T)

seed_eigen_colnames = data.frame(t(data.frame(strsplit(colnames(cor_eigengenes),"_"))))
names(seed_eigen_colnames) = c("gene_id","eigengene_type")

seed_gene_modules_sub = subset(seed_gene_modules, module_id==3) #pick a particular module to find overlapping genes in
module3_overlaps = do.call(rbind, lapply(1:nrow(seed_gene_modules_sub), function(i){
  #i=5
  print(i)
  tryCatch({ #suppress error
    
    wgcna_module_genes = subset(seed_gene_edgelist, gene1==seed_gene_modules_sub$gene_id[i])$gene2
    rank_genes = cor_eigengenes[,match (paste(seed_gene_modules_sub$gene_id[i],"_Car",sep=""),  names(cor_eigengenes) )] #get the correct eigengene column
    interesecting_genes = intersect(wgcna_module_genes ,data.frame(t(data.frame(strsplit(rank_genes," "))))[,1] ) #find interesection between WGCNA and our eigengenes
    wgcna_interesection = data.frame(seed_gene_modules_sub[i,],length(wgcna_module_genes),length(interesecting_genes),paste(interesecting_genes,collapse=", "))
    
    names(wgcna_interesection) = c("seed_gene","colors","module_id","wgcna_edge","N_intersection_rankcor_eigengene","interesecting_genes")
    wgcna_interesection
  },  error=function(e){})
  
}))

#-----------------------------------------------------------------#
#-----------------------------------------------------------------#
#-----------------------------------------------------------------#
#-----------------------------------------------------------------#
#-----------------------------------------------------------------#

#-----------------------------------------------------------------#
#-----------------------------------------------------------------#
#-----------------------------------------------------------------#
#-----------------------------------------------------------------#
#-----------------------------------------------------------------#

#-----------------------------------------------------------------#
#-----------------------------------------------------------------#
#-----------------------------------------------------------------#
#-----------------------------------------------------------------#
#-----------------------------------------------------------------#

#End WGCNA analysis
#=====================================================================================#
#Revision Fig : WGCNA module sizes for top 5 modulese in Normal skin vs Carcinoma
#=====================================================================================#

module_size = read.xlsx("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/WGCNA/normal_tumor_top5_modules.xlsx", 
                        sheet =  1)
names(module_size)

module_size_fig = ggplot(module_size, aes(x=Module.ID, y=Tissue.Num, fill=Tissue)) + 
  geom_point(aes(size=Ngene, color=Tissue),alpha=0.5)+
  ggtitle("Top 5 WGCNA Gene Modules") + xlab("Module") + ylab("Tissue") +
  scale_color_manual(values = c("Normal skin" = "darkblue","Carcinoma" = "darkred"), name="Tissue") + 
  theme(legend.position="bottom") + theme_classic() +
  theme(axis.text.y=element_text(size=0,color='black'),
        axis.text.x=element_text(size=20,color='black'),
        axis.title=element_text(size=0,color='black'),
        legend.title=element_text(size=20), 
        legend.text=element_text(size=20),
        axis.ticks = element_line(size=0),
        title = element_text(size=0)) +
  #scale_y_continuous(breaks = c(1,2), limits=c(-8,8)) #get rid of space at bottom and top
  scale_size(range=c(5,70)) +
  theme(legend.position = "none")

png("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Nature/Round\ 2/Figs/Revised\ Fig/ModuleSize_fig.png",
    width=5,height=2.5,units="in",res=250)
print(module_size_fig)
dev.off()

module_size_fig = ggplot(module_size, aes(x=Module.ID, y=Tissue.Num, fill=Tissue)) + 
  geom_point(aes(size=Ngene, color=Tissue),alpha=0.5)+
  ggtitle("Top 5 WGCNA Gene Modules") + xlab("Module") + ylab("Tissue") +
  scale_color_manual(values = c("Normal skin" = "darkblue","Carcinoma" = "darkred"), name="Tissue") + 
  theme(legend.position="bottom") + theme_classic() +
  theme(axis.text.y=element_text(size=0,color='black'),
        axis.text.x=element_text(size=20,color='black'),
        axis.title=element_text(size=0,color='black'),
        legend.title=element_text(size=20), 
        legend.text=element_text(size=20),
        axis.ticks = element_line(size=0),
        title = element_text(size=0)) +
  #scale_y_continuous(breaks = c(1,2), limits=c(-8,8)) #get rid of space at bottom and top
  scale_size(range=c(5,70)) +
  theme(legend.position = "bottom")

png("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Nature/Round\ 2/Figs/Revised\ Fig/ModuleSize_fig_LEGEND.png",
    width=30,height=10,units="in",res=250)
print(module_size_fig)
dev.off()


#-----------------------------------------------------------------#
#-----------------------------------------------------------------#
#-----------------------------------------------------------------#
#-----------------------------------------------------------------#
#-----------------------------------------------------------------#

#Plot DEGs for cisplatin vs control into the original manifold

cisplatin_degs = read.xlsx("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/combined_projects/results/DEG/Chemo\ project\ -\ Chemo\ -\ Control\ VS\ Chemo\ project\ -\ Chemo\ -\ Cisplatin/Chemo\ project\ -\ Chemo\ -\ Control\ VS\ Chemo\ project\ -\ Chemo\ -\ Cisplatincollated_results.xlsx", 
                           sheet = "All DEGs")

pos_degs = subset(cisplatin_degs, fc>1 & leclass=="significant")
neg_degs = subset(cisplatin_degs, fc<1 & leclass=="significant")

#--------------------------------------------------------------------------#
#Top 100 Pos and Neg DEGs
#--------------------------------------------------------------------------#

pos_marker_df = data.frame(pos_degs$gene[1:100],"eigengene","eigengene")
neg_marker_df = data.frame(tail(neg_degs$gene,100),"eigengene","eigengene")

plot_cells(big_cds,
           genes=pos_marker_df,
           label_cell_groups=FALSE,
           label_leaves=F,
           label_branch_points=F,
           graph_label_size=1.5,
           show_trajectory_graph=F) +
  ggtitle("Top 100 Pos DEGs: control->cisplatin") + xlab("") + ylab("") + 
  theme(axis.text=element_text(size=0,color='black'),
        title=element_text(size=18),
        axis.title=element_text(size=0),
        axis.ticks = element_line(size=0),
        legend.position="none")+ 
  theme(strip.text.x = element_text(size = 0)) + #remove facet label
  scale_color_viridis_c(option = "inferno",name="integrated expression")

plot_cells(big_cds,
           genes=neg_marker_df,
           label_cell_groups=FALSE,
           label_leaves=F,
           label_branch_points=F,
           graph_label_size=1.5,
           show_trajectory_graph=F) +
  ggtitle("Top 100 Neg DEGs: control->cisplatin") + xlab("") + ylab("") + 
  theme(axis.text=element_text(size=0,color='black'),
        title=element_text(size=18),
        axis.title=element_text(size=0),
        axis.ticks = element_line(size=0),
        legend.position="none")+ 
  theme(strip.text.x = element_text(size = 0)) + #remove facet label
  scale_color_viridis_c(option = "inferno",name="integrated expression")

#--------------------------------------------------------------------------#
#All DEGs
#--------------------------------------------------------------------------#

pos_marker_df = data.frame(pos_degs$gene,"eigengene","eigengene")
neg_marker_df = data.frame(neg_degs$gene,"eigengene","eigengene")

plot_cells(big_cds,
           genes=pos_marker_df,
           label_cell_groups=FALSE,
           label_leaves=F,
           label_branch_points=F,
           graph_label_size=1.5,
           show_trajectory_graph=F) +
  ggtitle(paste("All Pos DEGs (N=" ,nrow(pos_marker_df),"):control->cisplatin",sep="") ) + xlab("") + ylab("") + 
  theme(axis.text=element_text(size=0,color='black'),
        title=element_text(size=18),
        axis.title=element_text(size=0),
        axis.ticks = element_line(size=0),
        legend.position="none")+ 
  theme(strip.text.x = element_text(size = 0)) + #remove facet label
  scale_color_viridis_c(option = "inferno",name="integrated expression")

plot_cells(big_cds,
           genes=neg_marker_df,
           label_cell_groups=FALSE,
           label_leaves=F,
           label_branch_points=F,
           graph_label_size=1.5,
           show_trajectory_graph=F) +
  ggtitle(paste("All Neg DEGs (N=" ,nrow(neg_marker_df),"):control->cisplatin",sep="")) + xlab("") + ylab("") + 
  theme(axis.text=element_text(size=0,color='black'),
        title=element_text(size=18),
        axis.title=element_text(size=0),
        axis.ticks = element_line(size=0),
        legend.position="none")+ 
  theme(strip.text.x = element_text(size = 0)) + #remove facet label
  scale_color_viridis_c(option = "inferno",name="integrated expression")

#------------------------------------------------------------------------------------------#
#------------------------------------------------------------------------------------------#
#Add chemo samples and re-align across batch
#------------------------------------------------------------------------------------------#
#Now add chemo_cds to the original data and infer a lineage trajectory
rm(parenchyma_cds)
big_cds = readRDS("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/matvei/cds_objects/integrated_corrected_cds_v3.RDS")
chemo_cds = readRDS(file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/chemotherapy_tumor_project/data/cds/chemotherapy_tumor.RDS")

#Get only control and cisplatin samples
unique(colData(chemo_cds)$Chemo)
chemo_cds = chemo_cds[,colData(chemo_cds)$Chemo %in% c("Control","Cisplatin")] #filter for cluster 19 because most of the bridge cells are only in cluster 19

#Combine original data + control/cisplatin samples
all_cds <- combine_cds(list(big_cds,chemo_cds))

#Metadata corrections
colData(all_cds)$Genotype = gsub("Kras 4A KO","Kras4A KO",colData(all_cds)$Genotype)
colData(all_cds)$Sample.Info = gsub("Lineage traced carcinoma tissue ","Lineage traced carcinoma tissue",colData(all_cds)$Sample.Info)
colData(all_cds)$tissue.name = gsub("CA","Car",colData(all_cds)$tissue.name)
colData(all_cds)$cell_type_hires = gsub("delta-gamma T","dgT",colData(all_cds)$cell_type_hires)
colData(all_cds)$cell_type_hires = gsub("unF","UnF",colData(all_cds)$cell_type_hires)

#Process data and print out
all_cds <- preprocess_cds(all_cds, 
                          method="PCA",
                          num_dim = 50,
                          norm_method = "none", #these data have already been normalized by Seurat
                          alignment_group = NULL)


#3c. plot PC variation explained ?plot_pc_variance_explained
plot_pc_variance_explained(all_cds) + ggtitle("All-sample Scree Plot") +
  theme(text = element_text(size=18)) +
  theme(axis.text=element_text(size=18,color='black'),
        axis.title=element_text(size=20,color='black')) +
  theme(plot.title = element_text(size = 22, vjust=1)) +
  theme(axis.text=element_text(size=15,color='black'),
        axis.title=element_text(size=15,color='black')) +
  theme(panel.grid.major=element_line(colour = "darkgrey"), 
        panel.grid.minor=element_blank(), 
        panel.background=element_blank()) +
  theme(legend.key = element_rect(fill = "white")) +
  theme(strip.text.x = element_text(size = 14, colour = "black", angle = 0)) +
  theme(strip.text.y = element_text(size = 14, colour = "black", angle = 0)) +
  theme(legend.text=element_text(size=14), legend.title=element_text(size=14))

#4. A further round of (nonlinear) dimensionality reduction using UMAP ?reduce_dimension
#4b.
all_cds <- reduce_dimension(
  all_cds,
  max_components = 2,
  reduction_method = c("UMAP"),
  preprocess_method = "PCA",
  umap.metric = "cosine",
  umap.min_dist = 0.1,
  umap.n_neighbors = 10L,
  umap.fast_sgd = FALSE,
  umap.nn_method = "annoy",
  cores = 1,
  verbose = FALSE
)

#5. Cluster and partition ?cluster_cells cells into supergroups community detection (Phenograph algorithm) Community detection of cells as part of Phenograph algorithm https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4508757/

#5b. 
all_cds <- cluster_cells(all_cds,
                         reduction_method = c("UMAP"),
                         k = 10, #a bigger k will result in lower resolution 
                         cluster_method = c("leiden"),
                         num_iter = 2,
                         partition_qval = 0.05, #Numeric, the q-value cutoff to determine when to partition
                         weight = FALSE,
                         resolution = NULL,
                         random_seed = NULL,
                         verbose = F)

#Loop through selected metadata metrics
colData(all_cds)$cluster = all_cds@clusters@listData$UMAP$clusters
colData(all_cds)$partition = all_cds@clusters@listData$UMAP$partitions

data.frame(names(colData(all_cds)))

#Find metadata column index of metrics that I want to color cells by 
for(i in c(16,39:59,76)) {
  tryCatch({ #suppress error
    
    #i=76
    colourCount = length(unique(colData(all_cds)[,i]))
    getPalette = colorRampPalette(brewer.pal(9, "Set1"))
    
    png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/combined_projects/chemo_original_data/unbatch_corrected/manifolds/",names(colData(all_cds))[i] ,".png",sep=""),
        width=9,height=7,units="in",res=300)
    print(
      plot_cells(all_cds, color_cells_by=names(colData(all_cds))[i], group_cells_by="cluster",
                 group_label_size = 3,
                 x=1,
                 y=2,
                 alpha=0.4,
                 cell_size=0.4) +
        ggtitle(paste(names(colData(all_cds))[i])) + xlab("") + ylab("") + 
        theme(axis.text=element_text(size=0,color='black'),
              axis.title=element_text(size=0),
              title=element_text(size=15),
              axis.ticks = element_line(size=0),
              legend.position="right") +
        scale_color_manual(values=getPalette(colourCount))
    )
    dev.off()
  },  error=function(e){})
  
}

##################################################################################################################################
#Batch correct on Project name instead of batch bc batch was already corrected for
unique(colData(all_cds)$project.name)
all_cds = align_cds(cds = all_cds,
                    num_dim = 50,
                    preprocess_method = "PCA",
                    alignment_group = "project.name",
                    verbose=T)


#3c. plot PC variation explained ?plot_pc_variance_explained
plot_pc_variance_explained(all_cds) + ggtitle("All-sample Scree Plot: Batch-corrected") +
  theme(text = element_text(size=18)) +
  theme(axis.text=element_text(size=18,color='black'),
        axis.title=element_text(size=20,color='black')) +
  theme(plot.title = element_text(size = 22, vjust=1)) +
  theme(axis.text=element_text(size=15,color='black'),
        axis.title=element_text(size=15,color='black')) +
  theme(panel.grid.major=element_line(colour = "darkgrey"), 
        panel.grid.minor=element_blank(), 
        panel.background=element_blank()) +
  theme(legend.key = element_rect(fill = "white")) +
  theme(strip.text.x = element_text(size = 14, colour = "black", angle = 0)) +
  theme(strip.text.y = element_text(size = 14, colour = "black", angle = 0)) +
  theme(legend.text=element_text(size=14), legend.title=element_text(size=14))

#4. A further round of (nonlinear) dimensionality reduction using UMAP ?reduce_dimension
#4b.
all_cds <- reduce_dimension(
  all_cds,
  max_components = 2,
  reduction_method = c("UMAP"),
  preprocess_method = "Aligned",
  umap.metric = "cosine",
  umap.min_dist = 0.1,
  umap.n_neighbors = 15L,
  umap.fast_sgd = FALSE,
  umap.nn_method = "annoy",
  cores = 1,
  verbose = FALSE
)

#5. Cluster and partition ?cluster_cells cells into supergroups community detection (Phenograph algorithm) Community detection of cells as part of Phenograph algorithm https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4508757/

#5b. 
all_cds <- cluster_cells(all_cds,
                         reduction_method = c("UMAP"),
                         k = 10, #a bigger k will result in lower resolution 
                         cluster_method = c("leiden"),
                         num_iter = 2,
                         partition_qval = 0.05, #Numeric, the q-value cutoff to determine when to partition
                         weight = FALSE,
                         resolution = NULL,
                         random_seed = NULL,
                         verbose = F)

#Find metadata column index of metrics that I want to color cells by 
for(i in c(16,39:59,76)) {
  tryCatch({ #suppress error
    
    #i=16
    colourCount = length(unique(colData(all_cds)[,i]))
    getPalette = colorRampPalette(brewer.pal(9, "Set1"))
    
    png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/combined_projects/chemo_original_data/batch_corrected/manifolds/",i,"_",names(colData(all_cds))[i] ,".png",sep=""),
        width=9,height=7,units="in",res=300)
    print(
      plot_cells(all_cds, color_cells_by=names(colData(all_cds))[i], group_cells_by="cluster",
                 group_label_size = 3,
                 x=1,
                 y=2,
                 alpha=0.4,
                 cell_size=0.4) +
        ggtitle(paste(names(colData(all_cds))[i])) + xlab("") + ylab("") + 
        theme(axis.text=element_text(size=0,color='black'),
              axis.title=element_text(size=0),
              title=element_text(size=15),
              axis.ticks = element_line(size=0),
              legend.position="right") +
        scale_color_manual(values=getPalette(colourCount))
    )
    dev.off()
  },  error=function(e){})
  
}

#------------------------------------------------------------------------------------------------------------------------------------------------------------------#
#START COMBINATORIC BRIDGE ANALYSIS on original data+chemo data parenchyma to test how robust bridges are to sample addition/deletion
#------------------------------------------------------------------------------------------------------------------------------------------------------------------#
#Start outer loop that is adding/dropping all combos of chemo parenchyma

original_sample = c("first_run_TAM_BS","first_run_PROGENY","first_run_NO_TAM_BS","first_run_LGR6","second_run_Skin_Tom","second_run_Skin_Lgr6","second_run_PAP_US","second_run_PAP_Tom","second_run_PAP_Lgr6","second_run_CA_US","second_run_CA_Progeny","second_run_CA_Lgr6")
chemo_samples = c("EK-3205_01","EK-3205_02","EK-3205_03","EK-3205_04")
combn(chemo_samples,3)

#Start outer loop that goes through the number of elements to choose from chemo_samples
for(k in 0:length(chemo_samples)) {
  #k=4
  combos = combn(chemo_samples,k)
  
  for(p in 1:ncol(combos)) {
    #p=1
    output_dir = paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/combined_projects/chemo_original_data/combinatoric_bridge_analysis/batch_corrected/",
                       k," ",paste(combos[,p],collapse=" "),sep="")
    dir.create(output_dir)
    
    pap_ca = all_cds[,colData(all_cds)$tissue.name %in% c("Pap","Car") & 
                       colData(all_cds)$cell_type %in% c("nEp","nEp NeuroEndDiff") &
                       colData(all_cds)$sample_name %in% c(original_sample,
                                                           combos[,p])] #Get all original samples + a specific combo of chemo samples
    
    pap_ca = align_cds(cds = pap_ca,
                       num_dim = 50,
                       preprocess_method = "PCA",
                       alignment_group = "project.name",
                       verbose=T)
    
    #-----------------------------------------------#
    #Min_distances loop: [0.05, 0.10, 0.15, ... , 1]
    #-----------------------------------------------#
    min_distances = seq(0.05,1,by=0.05) #This is showing a lot of interesting bridgeness with min distanes bewteen 0.4 and 0.6
    
    for (i in 1:length(min_distances)) {
      #i=11
      print(paste("k=",k," p=",p," i=",i,sep=""))
      #loop to plot in Pap and CA in different ways
      #4b. ?reduce_dimension 
      pap_ca.test <- reduce_dimension(
        pap_ca,
        max_components = 2,
        reduction_method = c("UMAP"),
        preprocess_method = "PCA", #must now used ALIGNEMNT PRE-PROCESSING METHOD bc it replaces the other pre-process shit
        umap.metric = "cosine",
        umap.min_dist = min_distances[i],
        umap.n_neighbors = 15L,
        umap.fast_sgd = FALSE,
        umap.nn_method = "annoy",
        cores = 1,
        verbose = FALSE
      )
      
      #5. Cluster and partition ?cluster_cells cells into supergroups community detection (Phenograph algorithm) Community detection of cells as part of Phenograph algorithm https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4508757/
      #5b.  ?cluster_cells
      pap_ca.test <- cluster_cells(pap_ca.test,
                                   reduction_method = c("UMAP"),
                                   k = 10, #a bigger k will result in lower resolution 
                                   cluster_method = c("leiden"),
                                   num_iter = 2,
                                   partition_qval = 0.05, #Numeric, the q-value cutoff to determine when to partition
                                   weight = FALSE,
                                   resolution = NULL,
                                   random_seed = NULL,
                                   verbose = F)
      
      #Plot cell type with original data's bridges!
      colData(pap_ca.test)$integrated_type_bridge[is.na(colData(pap_ca.test)$integrated_type_bridge)]="nEp" #replace chemo NAs with nEp bc chemo's cell type was not annotated in this way
      colourCount = length(unique(colData(pap_ca.test)$integrated_type_bridge))
      getPalette = colorRampPalette(brewer.pal(9, "Set1"))
      
      bridge_cell_manifold = plot_cells(pap_ca.test, color_cells_by="integrated_type_bridge", group_cells_by="cluster",
                                        group_label_size = 0,
                                        x=1,
                                        y=2,
                                        alpha=0.6,
                                        cell_size=0.9) +
        ggtitle(paste("Min dist = ",min_distances[i],sep="")) + xlab("") + ylab("") + 
        theme(axis.text=element_text(size=0,color='black'),
              axis.title=element_text(size=0),
              title=element_text(size=15),
              axis.ticks = element_line(size=0),
              legend.position="right") +
        scale_color_manual(values=getPalette(colourCount))
      
      #Plot sample name
      colourCount = length(unique(colData(pap_ca.test)$sample_name))
      getPalette = colorRampPalette(brewer.pal(9, "Set1"))
      
      sample_manifold = plot_cells(pap_ca.test, color_cells_by="sample_name", group_cells_by="cluster",
                                   group_label_size = 0,
                                   x=1,
                                   y=2,
                                   alpha=0.6,
                                   cell_size=0.9) +
        ggtitle(paste("Min dist = ",min_distances[i],sep="")) + xlab("") + ylab("") + 
        theme(axis.text=element_text(size=0,color='black'),
              axis.title=element_text(size=0),
              title=element_text(size=15),
              axis.ticks = element_line(size=0),
              legend.position="right") +
        scale_color_manual(values=getPalette(colourCount))
      
      
      #Print out these manifolds
      png(paste(output_dir,"/i_",i,"_mindistance_", min_distances[i],".png",sep=""),
          width=18,height=7,units="in",res=300)
      print(
        multiplot(bridge_cell_manifold,sample_manifold,cols=2)
      )
      dev.off()
      
    } #End min distance loop (incrementing on i)
    
    
    #-----------------------------------------------#
    #Nearest Neighbor loop: [0.05, 0.10, 0.15, ... , 1]
    #-----------------------------------------------#
    nearest_neighbors = seq(5,200,by=15) #This is showing a lot of interesting bridgeness with min distanes bewteen 0.4 and 0.6
    
    for (q in 1:length(nearest_neighbors)) {
      #q=14
      print(paste("k=",k," p=",p," q=",q,sep=""))
      #loop to plot in Pap and CA in different ways
      #4b. ?reduce_dimension 
      pap_ca.test <- reduce_dimension(
        pap_ca,
        max_components = 2,
        reduction_method = c("UMAP"),
        preprocess_method = "PCA", #must now used ALIGNEMNT PRE-PROCESSING METHOD bc it replaces the other pre-process shit
        umap.metric = "cosine",
        umap.min_dist = 0.5,
        umap.n_neighbors = nearest_neighbors[q],
        umap.fast_sgd = FALSE,
        umap.nn_method = "annoy",
        cores = 1,
        verbose = FALSE
      )
      
      #5. Cluster and partition ?cluster_cells cells into supergroups community detection (Phenograph algorithm) Community detection of cells as part of Phenograph algorithm https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4508757/
      #5b.  ?cluster_cells
      pap_ca.test <- cluster_cells(pap_ca.test,
                                   reduction_method = c("UMAP"),
                                   k = 10, #a bigger k will result in lower resolution 
                                   cluster_method = c("leiden"),
                                   num_iter = 2,
                                   partition_qval = 0.05, #Numeric, the q-value cutoff to determine when to partition
                                   weight = FALSE,
                                   resolution = NULL,
                                   random_seed = NULL,
                                   verbose = F)
      
      #Plot cell type with original data's bridges!
      colData(pap_ca.test)$integrated_type_bridge[is.na(colData(pap_ca.test)$integrated_type_bridge)]="nEp" #replace chemo NAs with nEp bc chemo's cell type was not annotated in this way
      colourCount = length(unique(colData(pap_ca.test)$integrated_type_bridge))
      getPalette = colorRampPalette(brewer.pal(9, "Set1"))
      
      bridge_cell_manifold = plot_cells(pap_ca.test, color_cells_by="integrated_type_bridge", group_cells_by="cluster",
                                        group_label_size = 0,
                                        x=1,
                                        y=2,
                                        alpha=0.6,
                                        cell_size=0.9) +
        ggtitle(paste("Nn = ",nearest_neighbors[q],sep="")) + xlab("") + ylab("") + 
        theme(axis.text=element_text(size=0,color='black'),
              axis.title=element_text(size=0),
              title=element_text(size=15),
              axis.ticks = element_line(size=0),
              legend.position="right") +
        scale_color_manual(values=getPalette(colourCount))
      
      #Plot sample name
      colourCount = length(unique(colData(pap_ca.test)$sample_name))
      getPalette = colorRampPalette(brewer.pal(9, "Set1"))
      
      sample_manifold = plot_cells(pap_ca.test, color_cells_by="sample_name", group_cells_by="cluster",
                                   group_label_size = 0,
                                   x=1,
                                   y=2,
                                   alpha=0.6,
                                   cell_size=0.9) +
        ggtitle(paste("Nn = ",nearest_neighbors[q],sep="")) + xlab("") + ylab("") + 
        theme(axis.text=element_text(size=0,color='black'),
              axis.title=element_text(size=0),
              title=element_text(size=15),
              axis.ticks = element_line(size=0),
              legend.position="right") +
        scale_color_manual(values=getPalette(colourCount))
      
      
      #Print out these manifolds
      png(paste(output_dir,"/q_",q,"_Nn_", nearest_neighbors[q],".png",sep=""),
          width=18,height=7,units="in",res=300)
      print(
        multiplot(bridge_cell_manifold,sample_manifold,cols=2)
      )
      dev.off()
      
    } #End Nearest neighbor (Nn) loop (incrementing on q)
    
  } #End combos loop (incrementing on p)
  
} #End number of elements to choosee from (incrementing on l)

#---------------------------------------------------------------------------------------------#
#---------------------------------------------------------------------------------------------#
#---------------------------------------------------------------------------------------------#
#--------------------------------------------------------------------------------------#
#Now print out metadata characteristics in optimized parenchyma bridge
#--------------------------------------------------------------------------------------#
optimal_min_distance = 0.55
pap_ca = all_cds[,colData(all_cds)$tissue.name %in% c("Pap","Car") & colData(all_cds)$cell_type %in% c("nEp","nEp NeuroEndDiff")]

pap_ca = align_cds(cds = pap_ca,
                   num_dim = 50,
                   preprocess_method = "PCA",
                   alignment_group = "project.name",
                   verbose=T)

#4b. ?reduce_dimension 
pap_ca <- reduce_dimension(
  pap_ca,
  max_components = 2,
  reduction_method = c("UMAP"),
  preprocess_method = "PCA", #must now used ALIGNEMNT PRE-PROCESSING METHOD bc it replaces the other pre-process shit
  umap.metric = "cosine",
  umap.min_dist = optimal_min_distance, #this is based on optimizing the manifold to visualize bridge cells (See pap_ca_manifold_min_distances.pdf)
  umap.n_neighbors = 15L,
  umap.fast_sgd = FALSE,
  umap.nn_method = "annoy",
  cores = 1,
  verbose = FALSE
)

#5. Cluster and partition ?cluster_cells cells into supergroups community detection (Phenograph algorithm) Community detection of cells as part of Phenograph algorithm https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4508757/
#5b.  ?cluster_cells
pap_ca <- cluster_cells(pap_ca,
                        reduction_method = c("UMAP"),
                        k = 10, #a bigger k will result in lower resolution 
                        cluster_method = c("leiden"),
                        num_iter = 2,
                        partition_qval = 0.05, #Numeric, the q-value cutoff to determine when to partition
                        weight = FALSE,
                        resolution = NULL,
                        random_seed = NULL,
                        verbose = F)


#Find metadata column index of metrics that I want to color cells by 
colnames(colData(pap_ca))
for(i in c(16:18,39:59,76)) {
  tryCatch({ #suppress error
    
    #i=16
    colourCount = length(unique(colData(pap_ca)[,i]))
    getPalette = colorRampPalette(brewer.pal(9, "Set1"))
    
    png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/combined_projects/chemo_original_data/batch_corrected/nEp\ parenchyma/metadata_manifold/",i,"_",names(colData(all_cds))[i] ,".png",sep=""),
        width=9,height=7,units="in",res=300)
    print(
      plot_cells(pap_ca, color_cells_by=names(colData(pap_ca))[i], group_cells_by="cluster",
                 group_label_size = 0,
                 x=1,
                 y=2,
                 alpha=0.4,
                 cell_size=1) +
        ggtitle(paste(names(colData(pap_ca))[i])) + xlab("") + ylab("") + 
        theme(axis.text=element_text(size=0,color='black'),
              axis.title=element_text(size=0),
              title=element_text(size=15),
              axis.ticks = element_line(size=0),
              legend.position="right") +
        scale_color_manual(values=getPalette(colourCount))
    )
    dev.off()
  },  error=function(e){})
  
}

pap_ca_chemo = pap_ca

colData(chemo_cds)

colnames(colData(pap_ca))
for(i in c(1,16:18,39:59,76)) {
  tryCatch({ #suppress error
    
    #i=16
    colourCount = length(unique(colData(chemo_cds)[,i]))
    getPalette = colorRampPalette(brewer.pal(9, "Set1"))
    
    png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/chemotherapy_tumor_project/data/figures/manifolds/metadata_manifolds_v2/",i,"_",names(colData(all_cds))[i] ,".png",sep=""),
        width=9,height=7,units="in",res=300)
    print(
      plot_cells(chemo_cds, color_cells_by=names(colData(chemo_cds))[i], group_cells_by="cluster",
                 group_label_size = 0,
                 x=1,
                 y=2,
                 alpha=0.4,
                 cell_size=0.6) +
        ggtitle(paste(names(colData(chemo_cds))[i])) + xlab("") + ylab("") + 
        theme(axis.text=element_text(size=0,color='black'),
              axis.title=element_text(size=0),
              title=element_text(size=15),
              axis.ticks = element_line(size=0),
              legend.position="right") +
        scale_color_manual(values=getPalette(colourCount))
    )
    dev.off()
  },  error=function(e){})
  
}

plot_cells(pap_ca, color_cells_by="Sample.Info", group_cells_by="cluster",
           group_label_size = 0,
           x=1,
           y=2,
           alpha=0.4,
           cell_size=0.6) +
  ggtitle(paste("Sample.Info")) + xlab("") + ylab("") + 
  theme(axis.text=element_text(size=0,color='black'),
        axis.title=element_text(size=0),
        title=element_text(size=15),
        axis.ticks = element_line(size=0),
        legend.position="right") +
  scale_color_manual(values=getPalette(colourCount))


plot_cells(pap_ca,
           genes="Fah",
           label_cell_groups=FALSE,
           label_leaves=F,
           label_branch_points=F,
           graph_label_size=1.5,
           show_trajectory_graph=F,
           cell_size=1.2) +
  xlab("") + ylab("") + 
  ggtitle(paste("Hras",sep="")) +
  theme(axis.ticks = element_blank()) +
  scale_color_viridis_c(option = "inferno",name="") +
  theme(axis.text=element_text(size=0,color='black'),
        axis.title=element_text(size=0,color='black')) + theme(plot.title = element_text(size = 15, vjust=0)) +
  theme(strip.text.x = element_text(size = 0)) +
  theme(legend.position="bottom")

#---------------------------------------------------------------------------------------------#
#---------------------------------------------------------------------------------------------#
#---------------------------------------------------------------------------------------------#
#---------------------------------------------------------------------------------------------#
#---------------------------------------------------------------------------------------------#
#Print various bridge-specific a priori eigengenes
balmain_genes = read.xlsx(
  xlsxFile="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/apriori_eigengenes/Mark_Eigengenes.xlsx",
  colNames=T
)

bridge_gene_basename = c("Tmem","Pdcd1","Cd47","MB21D1","Tgfb","274","Cd80") #these are the gene name stems that occur in the column names of genes that allan made
bridge_eigengenes = do.call(c, lapply(1:length(bridge_gene_basename), function(i){
  #i=1
  colnames(balmain_genes)[grep(bridge_gene_basename[i],colnames(balmain_genes))]
}))

eigengene_agg = do.call(cbind, lapply(1:length(bridge_eigengenes), function(i){
  
  tryCatch({ #suppress error
    
    #i=1
    print(i)
    col_index = match(bridge_eigengenes[i],colnames(balmain_genes)) #get column with genes you want homie
    #only if more than 1 constituent gene
    marker_df = data.frame(na.omit(balmain_genes[,col_index]),names(balmain_genes)[col_index])
    names(marker_df) = c("gene","cell_type")
    
    #remake the marker df dataframe with the matched gene names
    marker_df = data.frame(unique(alias2SymbolTable(unique(marker_df$gene), species = "Mm")),"eigengene","eigengene")
    names(marker_df) = c("gene","cell_type","module")
    marker_df = na.omit(marker_df) #remove non-matching genes
    
    #big_cds
    big_cds_manifold = plot_cells(big_cds,
                                  genes=marker_df,
                                  label_cell_groups=FALSE,
                                  label_leaves=F,
                                  label_branch_points=F,
                                  graph_label_size=1.5,
                                  show_trajectory_graph=F) +
      xlab("") + ylab("") + 
      ggtitle(paste(names(balmain_genes)[col_index]," eigengene",sep="")) +
      theme(axis.ticks = element_blank()) +
      scale_color_viridis_c(option = "inferno",name="") +
      theme(axis.text=element_text(size=0,color='black'),
            axis.title=element_text(size=0,color='black')) + theme(plot.title = element_text(size = 15, vjust=0)) +
      theme(strip.text.x = element_text(size = 0)) +
      theme(legend.position="bottom")
    
    #original pap_ca
    original_pap_ca_manifold = plot_cells(pap_ca,
                                          genes=marker_df,
                                          label_cell_groups=FALSE,
                                          label_leaves=F,
                                          label_branch_points=F,
                                          graph_label_size=1.5,
                                          show_trajectory_graph=F,
                                          cell_size=1) +
      xlab("") + ylab("") + 
      ggtitle(paste(names(balmain_genes)[col_index]," eigengene",sep="")) +
      theme(axis.ticks = element_blank()) +
      scale_color_viridis_c(option = "inferno",name="") +
      theme(axis.text=element_text(size=0,color='black'),
            axis.title=element_text(size=0,color='black')) + theme(plot.title = element_text(size = 15, vjust=0)) +
      theme(strip.text.x = element_text(size = 0)) +
      theme(legend.position="bottom")
    
    #pap_ca_chemo chemo neoplastic parenchyma 
    chemo_pap_ca_manifold = plot_cells(pap_ca_chemo,
                                       genes=marker_df,
                                       label_cell_groups=FALSE,
                                       label_leaves=F,
                                       label_branch_points=F,
                                       graph_label_size=1.5,
                                       show_trajectory_graph=F,
                                       cell_size=1) +
      xlab("") + ylab("") + 
      ggtitle(paste(names(balmain_genes)[col_index]," eigengene",sep="")) +
      theme(axis.ticks = element_blank()) +
      scale_color_viridis_c(option = "inferno",name="") +
      theme(axis.text=element_text(size=0,color='black'),
            axis.title=element_text(size=0,color='black')) + theme(plot.title = element_text(size = 15, vjust=0)) +
      theme(strip.text.x = element_text(size = 0)) +
      theme(legend.position="bottom")
    
    #Now print to png object
    png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/combined_projects/chemo_original_data/batch_corrected/nEp\ parenchyma/bridge_eigengenes/",i,"_",names(balmain_genes)[col_index],".png",sep=""),
        width=24,height=8,units="in",res=300)
    print(
      multiplot(big_cds_manifold,original_pap_ca_manifold,chemo_pap_ca_manifold,
                cols=3)
    )
    dev.off()
    
  },  error=function(e){})
}))

#---------------------------------------------------------------------------------------------#
#---------------------------------------------------------------------------------------------#
#Classon LINE paper DTP gene signatures from Table S5
#Make eigengenes from her RNAseq results and print them out

n_gene_vector = c(100,250,500,1000,1500,2000)

#Start out loop through various sizes of DEGs
for(j in 1:length(n_gene_vector)){
  
  #j=1
  n_gene = n_gene_vector[j] #get number of top DEGs from Classon's tables
  
  #Start inner loop through the different sheets
  for(i in 1:3) {
    
    print(paste("j=",j," i=",i,sep=""))
    #i=2
    #Read in each sheet from Classon's Table S5 excel workbook
    marker_file = read.xlsx(
      xlsxFile="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/apriori_eigengenes/Classon_LINE_paper/1-s2.0-S1535610817302945-mmc5.xlsx",
      sheet=i,
      colNames=T
    )
    
    #get sheet names
    sheet_names <- getSheetNames("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/apriori_eigengenes/Classon_LINE_paper/1-s2.0-S1535610817302945-mmc5.xlsx")
    
    #remake the marker df dataframe with the matched gene names
    marker_df = data.frame(unique(convertHumanGeneList(marker_file$X2[2:(n_gene+1)])),"eigengene","eigengene")
    names(marker_df) = c("gene","cell_type","module")
    marker_df = na.omit(marker_df) #remove non-matching genes
    
    #big_cds
    big_cds_manifold = plot_cells(big_cds,
                                  genes=marker_df,
                                  label_cell_groups=FALSE,
                                  label_leaves=F,
                                  label_branch_points=F,
                                  graph_label_size=1.5,
                                  show_trajectory_graph=F,
                                  cell_size=0.8) +
      xlab("") + ylab("") + 
      ggtitle(paste("Classon: ",sheet_names[i],"\ntop ",n_gene,"-gene eigengene",sep="")) +
      theme(axis.ticks = element_blank()) +
      scale_color_viridis_c(option = "inferno",name="") +
      theme(axis.text=element_text(size=0,color='black'),
            axis.title=element_text(size=0,color='black')) + theme(plot.title = element_text(size = 15, vjust=0)) +
      theme(strip.text.x = element_text(size = 0)) +
      theme(legend.position="bottom")
    
    #original pap_ca
    original_pap_ca_manifold = plot_cells(pap_ca,
                                          genes=marker_df,
                                          label_cell_groups=FALSE,
                                          label_leaves=F,
                                          label_branch_points=F,
                                          graph_label_size=1.5,
                                          show_trajectory_graph=F,
                                          cell_size=1) +
      xlab("") + ylab("") + 
      ggtitle(paste("Classon: ",sheet_names[i],"\ntop ",n_gene,"-gene eigengene",sep="")) +
      theme(axis.ticks = element_blank()) +
      scale_color_viridis_c(option = "inferno",name="") +
      theme(axis.text=element_text(size=0,color='black'),
            axis.title=element_text(size=0,color='black')) + theme(plot.title = element_text(size = 15, vjust=0)) +
      theme(strip.text.x = element_text(size = 0)) +
      theme(legend.position="bottom")
    
    #pap_ca_chemo chemo neoplastic parenchyma 
    chemo_pap_ca_manifold = plot_cells(pap_ca_chemo,
                                       genes=marker_df,
                                       label_cell_groups=FALSE,
                                       label_leaves=F,
                                       label_branch_points=F,
                                       graph_label_size=1.5,
                                       show_trajectory_graph=F,
                                       cell_size=1) +
      xlab("") + ylab("") + 
      ggtitle(paste("Classon: ",sheet_names[i],"\ntop ",n_gene,"-gene eigengene",sep="")) +
      theme(axis.ticks = element_blank()) +
      scale_color_viridis_c(option = "inferno",name="") +
      theme(axis.text=element_text(size=0,color='black'),
            axis.title=element_text(size=0,color='black')) + theme(plot.title = element_text(size = 15, vjust=0)) +
      theme(strip.text.x = element_text(size = 0)) +
      theme(legend.position="bottom")
    
    #Now print to png object
    png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/balmain_eigene/Classon_eigengenes/",i,"_n",n_gene,"_",sheet_names[i],".png",sep=""),
        width=24,height=8,units="in",res=300)
    print(
      multiplot(big_cds_manifold,original_pap_ca_manifold,chemo_pap_ca_manifold,
                cols=3)
    )
    dev.off()
    
  }
}


#---------------------------------------------------------------------------------------------#
#-----------------------------------------------------------#
#Plot p16 expression in optimized chemo+original data manifold

cds_subset <- pap_ca[row.names(subset(rowData(pap_ca),
                                      gene_short_name %in% c("Cdkn2a"))),]


plot_genes_violin(cds_subset, group_cells_by="sample_name", ncol=2,
) +
  theme(axis.text.x=element_text(angle=45, hjust=1))

plot_cells(pap_ca,
           genes="Cdkn2a",
           label_cell_groups=FALSE,
           label_leaves=F,
           label_branch_points=F,
           graph_label_size=1.5,
           show_trajectory_graph=F) +
  theme(legend.position="bottom") +
  scale_color_viridis_c(option = "inferno",name="") +
  ggtitle("p16 (Cdkn2a)")+ #continuous
  theme(axis.text=element_text(size=0,color='black'),
        axis.title=element_text(size=0,color='black'),
        axis.ticks = element_line(size=0)) + 
  theme(plot.title = element_text(size = 15, vjust=1)) +
  theme(strip.text = element_text(size = 0))


#-----------------------------------------------------------#
#Test DEG dependence on p16

parenchyma_cds = readRDS("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/combined_projects/results/all_cds/data/cds/parenchyma_cds.RDS")

#This is the chemo DEG
group_a = list("Chemo project","Chemo","Control")
group_b = list("Chemo project","Chemo","Cisplatin")
input_cds = parenchyma_cds
min_frac = 0.1 #input the number of cells to be expressed

#--------------------------------------------#
#1. Isolate the cell populations to be compared
#Get all cell IDs

#Get field index in metadata that I want to isolate a factor level on
match_index_a = match(unlist(group_a[2]), names(colData(input_cds))) #get column index of metadata field I wanna look at
match_index_b = match(unlist(group_b[2]), names(colData(input_cds))) #get column index of metadata field I wanna look at

#Get metadata dataframe
metadata_df = colData(input_cds)

#--------------------------------------------------------------------------------------------------#
#Isolate the project, metadata field factor level from column index
selected_cds_a =  input_cds [ , (metadata_df[,"project.name"] %in% unlist(group_a[1]) & (metadata_df[,match_index_a] %in% unlist(group_a[3])))  ]
colData(selected_cds_a)$comparison_column = paste(unlist(group_a[1]), unlist(group_a[2]), unlist(group_a[3]),sep=" - ")
group_a_name = paste(unlist(group_a[1]), unlist(group_a[2]), unlist(group_a[3]),sep=" - ")

selected_cds_b =  input_cds [ , (metadata_df[,"project.name"] %in% unlist(group_b[1]) & (metadata_df[,match_index_b] %in% unlist(group_b[3]))) ]
colData(selected_cds_b)$comparison_column = paste(unlist(group_b[1]), unlist(group_b[2]), unlist(group_b[3]),sep=" - ")
group_b_name = paste(unlist(group_b[1]), unlist(group_b[2]), unlist(group_b[3]),sep=" - ")

selected_cds = combine_cds(list(selected_cds_a,selected_cds_b))

#--------------------------------------------------------------------------------------------------#
#Add metadata column of p16 status
p16_status_col = do.call(rbind, lapply(1:nrow(colData(selected_cds)), function(i){
  #i=1
  print(i)
  if( colData(selected_cds)$sample_name[i] == as.character("EK-3205_01") | colData(selected_cds)$sample_name[i] == "EK-3205_04" ) {p16_status = "p16_neg"} else {p16_status = "p16_pos"}
  p16_status
}))

colData(selected_cds)$p16_status = p16_status_col #Add p16 column
table(colData(selected_cds)$p16_status,colData(selected_cds)$Chemo) #Get contingency table

#--------------------------------------------------------------------------------------------------#
#--------------------------------------------------------------------------------------------------#
#Do DEGs independently for p16 together (which was the same), p16 pos, p16 neg, and GLM testing for interaction
p16_model_names = c("p16 pos+neg", "p16 pos", "p16 neg")
p16_models = list(together = c("p16_pos","p16_neg"),
                  p16_pos = "p16_pos",
                  p16_neg = "p16_neg")


for(l in 1:length(p16_models)) {
  
  #l=1
  #Select the correct p16 status
  selected_cds_p16 =  selected_cds [ , colData(selected_cds)$p16_status %in% c(unlist(p16_models[l])) ]
  
  
  #--------------------------------------------------------------------------------------------------#
  #Filter candidate DEGs
  
  #Get genes expressed only in minimum_cells
  minimum_cells = round(dim(selected_cds_p16)[2]*min_frac)
  #now create Seurat object from the patient-condition specific object
  seurat_obj = CreateSeuratObject(
    selected_cds_p16@assays@data$counts,
    project = "CreateSeuratObject",
    assay = "RNA",
    names.field = 1,
    names.delim = "_",
    meta.data = NULL,
    min.cells = minimum_cells)
  
  #FindVariableFeatures
  seurat_obj <- FindVariableFeatures(seurat_obj, selection.method = "vst", nfeatures = 5000) #testing top 5k most variable genes this is a lot of genes this is fine
  variable_genes <- VariableFeatures(seurat_obj) #get variable gene names
  
  #Select
  to_test_cds = selected_cds_p16[rowData(selected_cds_p16)$gene_short_name %in% variable_genes,] #subset cds object to variable genes
  
  #--------------------------------------------------------------------------------------------------#
  #Start loop 
  meandiffs = do.call(rbind, lapply(1:nrow(to_test_cds@assays@data$counts), function(e){
    # meandiffs = do.call(rbind, lapply(1:10, function(e){
    
    tryCatch({ #suppress error
      #e=1
      print(paste(" e=",e,sep=""))
      
      to_test_cds@assays@data$counts[1:10,1:10]
      
      #Add metadata to imputed data
      evaluate_df = data.frame(to_test_cds@assays@data$counts[e,],colData(to_test_cds)$comparison_column) 
      names(evaluate_df) = c("test_gene", "dis_category")
      
      #Mean and std of dis factor level
      df <- as.data.frame(aggregate(test_gene ~ dis_category, evaluate_df, function(x) c(M = mean(x), SE = sd(x)/sqrt(length(x)))))
      group_a_res = subset(df, dis_category==group_a_name)
      group_b_res = subset(df, dis_category==group_b_name)
      
      #Mann-Whitney
      wilcox_res = wilcox.test(evaluate_df$test_gene ~ evaluate_df$dis_category)
      
      #Gather percentage of positive expression among cells
      group_a_expressed = nrow(subset(evaluate_df, dis_category==group_a_name & test_gene>0)) / nrow(subset(evaluate_df, dis_category==group_a_name))
      group_b_expressed = nrow(subset(evaluate_df, dis_category==group_b_name & test_gene>0)) / nrow(subset(evaluate_df, dis_category==group_b_name))
      
      #Gather results
      str(group_a_res)
      res_df = data.frame(rownames(to_test_cds@assays@data$counts)[e], 
                          group_a_res$test_gene[1,1], group_a_res$test_gene[1,2], group_a_expressed,
                          group_b_res$test_gene[1,1], group_b_res$test_gene[1,2], group_b_expressed,
                          as.numeric(group_b_res$test_gene[1,1])/as.numeric(group_a_res$test_gene[1,1]), log2(as.numeric(group_b_res$test_gene[1,1])/as.numeric(group_a_res$test_gene[1,1])),
                          wilcox_res$statistic, wilcox_res$p.value ) 
      
      names(res_df) = c("gene",
                        "group_a_mean","group_a_sd","group_a_fraction_posexp",
                        "group_b_mean","group_b_sd","group_b_fraction_posexp",
                        "fc","log2fc",
                        "W","p")
      #Return results file
      res_df
      
    },  error=function(e){})
  }))
  
  #order from high to low FC
  meandiffs = meandiffs[order(-meandiffs$fc),]
  meandiffs$p_adjusted = p.adjust(meandiffs$p, method="fdr")
  meandiffs.1 <- meandiffs[!is.infinite(meandiffs$fc),] #Remove infinite values
  meandiffs.1 = subset(meandiffs.1, fc>0)
  meandiffs.1 = meandiffs.1[order(-meandiffs.1$fc),]
  
  #Now Volcano plots
  topfc = head(meandiffs.1,20)
  topfc.2 = tail(meandiffs.1,20)
  
  transcript_class = do.call(rbind, lapply(1:nrow(meandiffs.1), function(j){
    #i=1
    subdf = meandiffs.1[j,]
    if( subdf$p_adjusted<0.05) {leclass = "significant"} else {leclass = "non-significant"}
    data.frame(subdf, leclass)
  }))
  
  volcano_plot = ggplot() +
    geom_point(data=transcript_class, aes(log2fc,(-log10(p_adjusted)), color=leclass), alpha=1) +
    geom_text_repel(data=topfc, aes(log2fc,(-log10(p_adjusted)),label=gene),box.padding = 0.3) + #add labels
    geom_text_repel(data=topfc.2, aes(log2fc,(-log10(p_adjusted)),label=gene),box.padding = 0.3) + #add labels
    geom_point(data=topfc, aes(log2fc,(-log10(p_adjusted))), alpha=1,color="firebrick3") +
    geom_point(data=topfc.2, aes(log2fc,(-log10(p_adjusted))), alpha=1,color="steelblue4") +
    geom_vline(xintercept=0, color="black") +
    xlab("log2(FC)") +
    ylab("-log10(FDR)") +
    ggtitle(paste("Positive: ",paste(unlist(group_b[1]), unlist(group_b[2]), unlist(group_b[3]),sep=" - "),sep="")) + #list the factor level corresponding to an increase in expression
    theme(axis.text=element_text(size=15,color='black'),
          axis.title=element_text(size=15,color='black')) +
    theme(plot.title = element_text(size = 20, vjust=2)) +
    theme(strip.background = element_rect(fill = 'white', color='black', size=1)) +
    theme(panel.background=element_blank(),
          panel.grid.major=element_blank(),
          panel.grid.minor=element_blank()) +
    #scale_y_continuous(breaks=number_ticks(6)) +
    #scale_x_continuous(breaks=number_ticks(6)) +
    theme(axis.line = element_line(colour = 'black')) +
    scale_colour_manual(values = c("significant" = "wheat3", "non-significant" = "lightgrey"), name="significance") +
    theme(legend.position = "none")
  
  #--------------------------------------------------------------------------#
  #Plot Expression of Pos and Neg DEGs as eigengenes in the original manifold
  #--------------------------------------------------------------------------#
  pos_degs = subset(transcript_class, fc>1 & leclass=="significant")
  neg_degs = subset(transcript_class, fc<1 & leclass=="significant")
  
  #Top 100 Pos and Neg DEGs
  pos_marker_df = data.frame(pos_degs$gene[1:100],"eigengene","eigengene")
  neg_marker_df = data.frame(tail(neg_degs$gene,100),"eigengene","eigengene")
  
  pos_top100_manifold = plot_cells(big_cds,
                                   genes=pos_marker_df,
                                   label_cell_groups=FALSE,
                                   label_leaves=F,
                                   label_branch_points=F,
                                   graph_label_size=1.5,
                                   show_trajectory_graph=F) +
    ggtitle("Top 100 Pos DEGs") + xlab("") + ylab("") + 
    theme(axis.text=element_text(size=0,color='black'),
          title=element_text(size=18),
          axis.title=element_text(size=0),
          axis.ticks = element_line(size=0),
          legend.position="none")+ 
    theme(strip.text.x = element_text(size = 0)) + #remove facet label
    scale_color_viridis_c(option = "inferno",name="integrated expression")
  
  neg_top100_manifold = plot_cells(big_cds,
                                   genes=neg_marker_df,
                                   label_cell_groups=FALSE,
                                   label_leaves=F,
                                   label_branch_points=F,
                                   graph_label_size=1.5,
                                   show_trajectory_graph=F) +
    ggtitle("Top 100 Neg DEGs") + xlab("") + ylab("") + 
    theme(axis.text=element_text(size=0,color='black'),
          title=element_text(size=18),
          axis.title=element_text(size=0),
          axis.ticks = element_line(size=0),
          legend.position="none")+ 
    theme(strip.text.x = element_text(size = 0)) + #remove facet label
    scale_color_viridis_c(option = "inferno",name="integrated expression")
  
  #All DEGs
  pos_marker_df = data.frame(pos_degs$gene,"eigengene","eigengene")
  neg_marker_df = data.frame(neg_degs$gene,"eigengene","eigengene")
  
  pos_all_manifold = plot_cells(big_cds,
                                genes=pos_marker_df,
                                label_cell_groups=FALSE,
                                label_leaves=F,
                                label_branch_points=F,
                                graph_label_size=1.5,
                                show_trajectory_graph=F) +
    ggtitle(paste("All Pos DEGs (N=" ,nrow(pos_marker_df),")",sep="") ) + xlab("") + ylab("") + 
    theme(axis.text=element_text(size=0,color='black'),
          title=element_text(size=18),
          axis.title=element_text(size=0),
          axis.ticks = element_line(size=0),
          legend.position="none")+ 
    theme(strip.text.x = element_text(size = 0)) + #remove facet label
    scale_color_viridis_c(option = "inferno",name="integrated expression")
  
  neg_all_manifold = plot_cells(big_cds,
                                genes=neg_marker_df,
                                label_cell_groups=FALSE,
                                label_leaves=F,
                                label_branch_points=F,
                                graph_label_size=1.5,
                                show_trajectory_graph=F) +
    ggtitle(paste("All Neg DEGs (N=" ,nrow(neg_marker_df),")",sep="")) + xlab("") + ylab("") + 
    theme(axis.text=element_text(size=0,color='black'),
          title=element_text(size=18),
          axis.title=element_text(size=0),
          axis.ticks = element_line(size=0),
          legend.position="none")+ 
    theme(strip.text.x = element_text(size = 0)) + #remove facet label
    scale_color_viridis_c(option = "inferno",name="integrated expression")
  
  
  
  #--------------------------------------------#
  #GO analysis on top 100 pos and neg DEGs
  #--------------------------------------------#
  #Top 100 POSITIVE DEGs
  geneUniverse <- row.names(input_cds)
  genesOfInterest = transcript_class$gene[1:100]
  geneList <- factor(as.integer(geneUniverse %in% genesOfInterest))
  names(geneList) <- geneUniverse
  GOdata <- new("topGOdata",
                ontology = "BP",
                allGenes=geneList,
                #geneSel = selection,
                annot = annFUN.org, 
                mapping = "org.Mm.eg.db",
                ID = "symbol")
  
  resultKS <- runTest(GOdata, algorithm = "weight01", statistic = "fisher") #Fisher exact test
  tab <- GenTable(GOdata, raw.p.value = resultKS, topNodes = length(resultKS@score), numChar = 120)
  tab = subset(tab, raw.p.value <=0.01) #keep only sig results
  go_candidates = genesOfInterest
  selcGenes <- genesInTerm(GOdata, whichGO=tab$GO.ID)
  genes_in_go = do.call(rbind, lapply(1:length(selcGenes), function(i){
    #i=1
    data.frame(names(selcGenes)[i],  paste( c(na.omit(selcGenes[[i]][match(go_candidates,  selcGenes[[i]])])) , collapse=", ") )
  }))
  names(genes_in_go) = c("GO_ID","sig_genes_in_term")
  tab = cbind(tab, genes_in_go)
  Pos_TopDEG_res = tab
  
  #All POSITIVE DEGs
  geneUniverse <- row.names(input_cds)
  genesOfInterest = subset(transcript_class, fc>1 & leclass=="significant")$gene
  geneList <- factor(as.integer(geneUniverse %in% genesOfInterest))
  names(geneList) <- geneUniverse
  GOdata <- new("topGOdata",
                ontology = "BP",
                allGenes=geneList,
                #geneSel = selection,
                annot = annFUN.org, 
                mapping = "org.Mm.eg.db",
                ID = "symbol")
  
  resultKS <- runTest(GOdata, algorithm = "weight01", statistic = "fisher") #Fisher exact test
  tab <- GenTable(GOdata, raw.p.value = resultKS, topNodes = length(resultKS@score), numChar = 120)
  tab = subset(tab, raw.p.value <=0.01) #keep only sig results
  go_candidates = genesOfInterest
  selcGenes <- genesInTerm(GOdata, whichGO=tab$GO.ID)
  genes_in_go = do.call(rbind, lapply(1:length(selcGenes), function(i){
    #i=1
    data.frame(names(selcGenes)[i],  paste( c(na.omit(selcGenes[[i]][match(go_candidates,  selcGenes[[i]])])) , collapse=", ") )
  }))
  names(genes_in_go) = c("GO_ID","sig_genes_in_term")
  tab = cbind(tab, genes_in_go)
  Pos_AllDEG_res = tab
  
  
  #Top 100 NEGATIVE DEGs
  geneUniverse <- row.names(input_cds)
  genesOfInterest = tail(transcript_class,100)$gene[1:100]
  geneList <- factor(as.integer(geneUniverse %in% genesOfInterest))
  names(geneList) <- geneUniverse
  GOdata <- new("topGOdata",
                ontology = "BP",
                allGenes=geneList,
                #geneSel = selection,
                annot = annFUN.org, 
                mapping = "org.Mm.eg.db",
                ID = "symbol")
  
  resultKS <- runTest(GOdata, algorithm = "weight01", statistic = "fisher") #Fisher exact test
  tab <- GenTable(GOdata, raw.p.value = resultKS, topNodes = length(resultKS@score), numChar = 120)
  tab = subset(tab, raw.p.value <=0.01) #keep only sig results 
  go_candidates = genesOfInterest
  selcGenes <- genesInTerm(GOdata, whichGO=tab$GO.ID)
  genes_in_go = do.call(rbind, lapply(1:length(selcGenes), function(i){
    #i=1
    data.frame(names(selcGenes)[i],  paste( c(na.omit(selcGenes[[i]][match(go_candidates,  selcGenes[[i]])])) , collapse=", ") )
  }))
  names(genes_in_go) = c("GO_ID","sig_genes_in_term")
  tab = cbind(tab, genes_in_go)
  Neg_TopDEG_res = tab
  
  #All NEGATIVE DEGs
  geneUniverse <- row.names(input_cds)
  genesOfInterest = subset(transcript_class, fc<1 & leclass=="significant")$gene
  geneList <- factor(as.integer(geneUniverse %in% genesOfInterest))
  names(geneList) <- geneUniverse
  GOdata <- new("topGOdata",
                ontology = "BP",
                allGenes=geneList,
                #geneSel = selection,
                annot = annFUN.org, 
                mapping = "org.Mm.eg.db",
                ID = "symbol")
  
  resultKS <- runTest(GOdata, algorithm = "weight01", statistic = "fisher") #Fisher exact test
  tab <- GenTable(GOdata, raw.p.value = resultKS, topNodes = length(resultKS@score), numChar = 120)
  tab = subset(tab, raw.p.value <=0.01) #keep only sig results
  go_candidates = genesOfInterest
  selcGenes <- genesInTerm(GOdata, whichGO=tab$GO.ID)
  genes_in_go = do.call(rbind, lapply(1:length(selcGenes), function(i){
    #i=1
    data.frame(names(selcGenes)[i],  paste( c(na.omit(selcGenes[[i]][match(go_candidates,  selcGenes[[i]])])) , collapse=", ") )
  }))
  names(genes_in_go) = c("GO_ID","sig_genes_in_term")
  tab = cbind(tab, genes_in_go)
  Neg_AllDEG_res = tab
  
  #Write out volcano  plot and results
  group_a_faclevels = paste(unlist(group_a[1]), unlist(group_a[2]), unlist(group_a[3]),sep=" - ")
  group_b_faclevels = paste(unlist(group_b[1]), unlist(group_b[2]), unlist(group_b[3]),sep=" - ")
  output_dir = paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/combined_projects/results/DEG/p16\ factors/",p16_model_names[l]," ",group_a_faclevels," VS ",group_b_faclevels,sep="")
  dir.create(output_dir)
  
  
  #-----------------------------------------------------------------------------------#
  #Write out to excel workbook
  
  OUT <- createWorkbook() #create excel workbook for a sample https://stackoverflow.com/questions/27713310/easy-way-to-export-multiple-data-frame-to-multiple-excel-worksheets
  
  addWorksheet(OUT, "All DEGs") # Add some sheets to the workbook
  writeData(OUT, sheet = "All DEGs", x = transcript_class) #Write data to a shee
  
  addWorksheet(OUT, "Pos_TopDEG_res") # Add some sheets to the workbook
  writeData(OUT, sheet = "Pos_TopDEG_res", x = Pos_TopDEG_res) #Write data to a shee
  
  addWorksheet(OUT, "Pos_AllDEG_res") # Add some sheets to the workbook
  writeData(OUT, sheet = "Pos_AllDEG_res", x = Pos_AllDEG_res) #Write data to a shee
  
  addWorksheet(OUT, "Neg_TopDEG_res") # Add some sheets to the workbook
  writeData(OUT, sheet = "Neg_TopDEG_res", x = Neg_TopDEG_res) #Write data to a shee
  
  addWorksheet(OUT, "Neg_AllDEG_res") # Add some sheets to the workbook
  writeData(OUT, sheet = "Neg_AllDEG_res", x = Neg_AllDEG_res) #Write data to a shee
  
  saveWorkbook(OUT, 
               paste(output_dir,"/",paste(group_a_faclevels," VS ",group_b_faclevels,"collated_results.xlsx",sep=""),sep=""),
               overwrite=T)
  
  #-----------------------------------------------------------------------------------#
  #Print out volcano plot
  png(paste(output_dir,"/volcano_plot.png",sep=""),
      width=7,height=7,units="in",res=350)
  print(volcano_plot)
  dev.off()
  
  #-----------------------------------------------------------------------------------#
  #Print out manifold plots
  png(paste(output_dir,"/manifold_eigengene_plot.png",sep=""),
      width=12,height=12,units="in",res=500)
  multiplot(pos_top100_manifold,pos_all_manifold,
            neg_top100_manifold,neg_all_manifold,
            cols=2)
  dev.off()
  
} #End p16 models loop (incremented on l)
#--------------------------------------------------------------------------------------------------#
#--------------------------------------------------------------------------------------------------#
#--------------------------------------------------------------------------------------------------#
#--------------------------------------------------------------------------------------------------#
#--------------------------------------------------------------------------------------------------#
#--------------------------------------------------------------------------------------------------#

#DEG Lgr6 to Progeny

tissues = c("Normal skin","Pap","CA" )

min_frac=0.1
group_a_name="Lgr6"
group_b_name="Progeny"

for(l in 1:length(tissues)) {
  
  #l=1
  #Select the right tissue, cell types, and FACS sorting (Lgr6, progeny)
  sub_cds =  big_cds [ , colData(big_cds)$tissue_updated %in% c(tissues[l]) & 
                         colData(big_cds)$cell_type %in% c("IfE","Inf","OuB","UHF","SG","Bas",
                                                           "Isth","unF","PrE","InB","nEp") &
                         colData(big_cds)$sorting %in% c("Lgr6","Progeny")]
  
  
  #--------------------------------------------------------------------------------------------------#
  #Filter candidate DEGs
  
  #Get genes expressed only in minimum_cells
  minimum_cells = round(dim(sub_cds)[2]*min_frac)
  #now create Seurat object from the patient-condition specific object
  seurat_obj = CreateSeuratObject(
    sub_cds@assays@data$counts,
    project = "CreateSeuratObject",
    assay = "RNA",
    names.field = 1,
    names.delim = "_",
    meta.data = NULL,
    min.cells = minimum_cells)
  
  #FindVariableFeatures
  seurat_obj <- FindVariableFeatures(seurat_obj, selection.method = "vst", nfeatures = 5000) #testing top 5k most variable genes this is a lot of genes this is fine
  variable_genes <- VariableFeatures(seurat_obj) #get variable gene names
  
  #Select high-var genes
  to_test_cds = sub_cds[rowData(sub_cds)$gene_short_name %in% variable_genes,] #subset cds object to variable genes
  
  #--------------------------------------------------------------------------------------------------#
  #Start loop 
  meandiffs = do.call(rbind, lapply(1:nrow(to_test_cds@assays@data$counts), function(e){
    # meandiffs = do.call(rbind, lapply(1:10, function(e){
    
    tryCatch({ #suppress error
      #e=1
      print(paste("l=",l," e=",e,sep=""))
      #Add metadata to imputed data
      evaluate_df = data.frame(to_test_cds@assays@data$counts[e,],colData(to_test_cds)$sorting) 
      names(evaluate_df) = c("test_gene", "dis_category")
      
      #Mean and std of dis factor level
      df <- as.data.frame(aggregate(test_gene ~ dis_category, evaluate_df, function(x) c(M = mean(x), SE = sd(x)/sqrt(length(x)))))
      group_a_res = subset(df, dis_category==group_a_name)
      group_b_res = subset(df, dis_category==group_b_name)
      
      #Mann-Whitney
      wilcox_res = wilcox.test(evaluate_df$test_gene ~ evaluate_df$dis_category)
      
      #Gather percentage of positive expression among cells
      group_a_expressed = nrow(subset(evaluate_df, dis_category==group_a_name & test_gene>0)) / nrow(subset(evaluate_df, dis_category==group_a_name))
      group_b_expressed = nrow(subset(evaluate_df, dis_category==group_b_name & test_gene>0)) / nrow(subset(evaluate_df, dis_category==group_b_name))
      
      #Gather results
      res_df = data.frame(rownames(to_test_cds@assays@data$counts)[e], 
                          group_a_res$test_gene[1,1], group_a_res$test_gene[1,2], group_a_expressed,
                          group_b_res$test_gene[1,1], group_b_res$test_gene[1,2], group_b_expressed,
                          as.numeric(group_b_res$test_gene[1,1])/as.numeric(group_a_res$test_gene[1,1]), log2(as.numeric(group_b_res$test_gene[1,1])/as.numeric(group_a_res$test_gene[1,1])),
                          wilcox_res$statistic, wilcox_res$p.value ) 
      
      names(res_df) = c("gene",
                        "group_a_mean","group_a_sd","group_a_fraction_posexp",
                        "group_b_mean","group_b_sd","group_b_fraction_posexp",
                        "fc","log2fc",
                        "W","p")
      #Return results file
      res_df
      
    },  error=function(e){})
  }))
  
  #order from high to low FC
  meandiffs = meandiffs[order(-meandiffs$fc),]
  meandiffs$p_adjusted = p.adjust(meandiffs$p, method="fdr")
  meandiffs.1 <- meandiffs[!is.infinite(meandiffs$fc),] #Remove infinite values
  meandiffs.1 = subset(meandiffs.1, fc>0)
  meandiffs.1 = meandiffs.1[order(-meandiffs.1$fc),]
  
  #Now Volcano plots
  topfc = head(meandiffs.1,20)
  topfc.2 = tail(meandiffs.1,20)
  
  transcript_class = do.call(rbind, lapply(1:nrow(meandiffs.1), function(j){
    #i=1
    subdf = meandiffs.1[j,]
    if( subdf$p_adjusted<0.05) {leclass = "significant"} else {leclass = "non-significant"}
    data.frame(subdf, leclass)
  }))
  
  volcano_plot = ggplot() +
    geom_point(data=transcript_class, aes(log2fc,(-log10(p_adjusted)), color=leclass), alpha=1) +
    geom_text_repel(data=topfc, aes(log2fc,(-log10(p_adjusted)),label=gene),box.padding = 0.3) + #add labels
    geom_text_repel(data=topfc.2, aes(log2fc,(-log10(p_adjusted)),label=gene),box.padding = 0.3) + #add labels
    geom_point(data=topfc, aes(log2fc,(-log10(p_adjusted))), alpha=1,color="firebrick3") +
    geom_point(data=topfc.2, aes(log2fc,(-log10(p_adjusted))), alpha=1,color="steelblue4") +
    geom_vline(xintercept=0, color="black") +
    xlab("log2(FC)") +
    ylab("-log10(FDR)") +
    ggtitle(paste("Lgr6->Progeny",sep="")) + #list the factor level corresponding to an increase in expression
    theme(axis.text=element_text(size=15,color='black'),
          axis.title=element_text(size=15,color='black')) +
    theme(plot.title = element_text(size = 20, vjust=2)) +
    theme(strip.background = element_rect(fill = 'white', color='black', size=1)) +
    theme(panel.background=element_blank(),
          panel.grid.major=element_blank(),
          panel.grid.minor=element_blank()) +
    #scale_y_continuous(breaks=number_ticks(6)) +
    #scale_x_continuous(breaks=number_ticks(6)) +
    theme(axis.line = element_line(colour = 'black')) +
    scale_colour_manual(values = c("significant" = "wheat3", "non-significant" = "lightgrey"), name="significance") +
    theme(legend.position = "none")
  
  #--------------------------------------------------------------------------#
  #Plot Expression of Pos and Neg DEGs as eigengenes in the original manifold
  #--------------------------------------------------------------------------#
  pos_degs = subset(transcript_class, fc>1 & leclass=="significant")
  neg_degs = subset(transcript_class, fc<1 & leclass=="significant")
  
  #Top 100 Pos and Neg DEGs
  pos_marker_df = data.frame(pos_degs$gene[1:100],"eigengene","eigengene")
  neg_marker_df = data.frame(tail(neg_degs$gene,100),"eigengene","eigengene")
  
  pos_top100_manifold = plot_cells(big_cds,
                                   genes=pos_marker_df,
                                   label_cell_groups=FALSE,
                                   label_leaves=F,
                                   label_branch_points=F,
                                   graph_label_size=1.5,
                                   show_trajectory_graph=F) +
    ggtitle("Top 100 Pos DEGs") + xlab("") + ylab("") + 
    theme(axis.text=element_text(size=0,color='black'),
          title=element_text(size=18),
          axis.title=element_text(size=0),
          axis.ticks = element_line(size=0),
          legend.position="none")+ 
    theme(strip.text.x = element_text(size = 0)) + #remove facet label
    scale_color_viridis_c(option = "inferno",name="integrated expression")
  
  neg_top100_manifold = plot_cells(big_cds,
                                   genes=neg_marker_df,
                                   label_cell_groups=FALSE,
                                   label_leaves=F,
                                   label_branch_points=F,
                                   graph_label_size=1.5,
                                   show_trajectory_graph=F) +
    ggtitle("Top 100 Neg DEGs") + xlab("") + ylab("") + 
    theme(axis.text=element_text(size=0,color='black'),
          title=element_text(size=18),
          axis.title=element_text(size=0),
          axis.ticks = element_line(size=0),
          legend.position="none")+ 
    theme(strip.text.x = element_text(size = 0)) + #remove facet label
    scale_color_viridis_c(option = "inferno",name="integrated expression")
  
  #All DEGs
  pos_marker_df = data.frame(pos_degs$gene,"eigengene","eigengene")
  neg_marker_df = data.frame(neg_degs$gene,"eigengene","eigengene")
  
  pos_all_manifold = plot_cells(big_cds,
                                genes=pos_marker_df,
                                label_cell_groups=FALSE,
                                label_leaves=F,
                                label_branch_points=F,
                                graph_label_size=1.5,
                                show_trajectory_graph=F) +
    ggtitle(paste("All Pos DEGs (N=" ,nrow(pos_marker_df),")",sep="") ) + xlab("") + ylab("") + 
    theme(axis.text=element_text(size=0,color='black'),
          title=element_text(size=18),
          axis.title=element_text(size=0),
          axis.ticks = element_line(size=0),
          legend.position="none")+ 
    theme(strip.text.x = element_text(size = 0)) + #remove facet label
    scale_color_viridis_c(option = "inferno",name="integrated expression")
  
  neg_all_manifold = plot_cells(big_cds,
                                genes=neg_marker_df,
                                label_cell_groups=FALSE,
                                label_leaves=F,
                                label_branch_points=F,
                                graph_label_size=1.5,
                                show_trajectory_graph=F) +
    ggtitle(paste("All Neg DEGs (N=" ,nrow(neg_marker_df),")",sep="")) + xlab("") + ylab("") + 
    theme(axis.text=element_text(size=0,color='black'),
          title=element_text(size=18),
          axis.title=element_text(size=0),
          axis.ticks = element_line(size=0),
          legend.position="none")+ 
    theme(strip.text.x = element_text(size = 0)) + #remove facet label
    scale_color_viridis_c(option = "inferno",name="integrated expression")
  
  #--------------------------------------------#
  #GO analysis on top 100 pos and neg DEGs
  #--------------------------------------------#
  #Top 100 POSITIVE DEGs
  geneUniverse <- row.names(to_test_cds)
  genesOfInterest = transcript_class$gene[1:100]
  geneList <- factor(as.integer(geneUniverse %in% genesOfInterest))
  names(geneList) <- geneUniverse
  GOdata <- new("topGOdata",
                ontology = "BP",
                allGenes=geneList,
                #geneSel = selection,
                annot = annFUN.org, 
                mapping = "org.Mm.eg.db",
                ID = "symbol")
  
  resultKS <- runTest(GOdata, algorithm = "weight01", statistic = "fisher") #Fisher exact test
  tab <- GenTable(GOdata, raw.p.value = resultKS, topNodes = length(resultKS@score), numChar = 120)
  tab = subset(tab, raw.p.value <=0.01) #keep only sig results
  go_candidates = genesOfInterest
  selcGenes <- genesInTerm(GOdata, whichGO=tab$GO.ID)
  genes_in_go = do.call(rbind, lapply(1:length(selcGenes), function(i){
    #i=1
    data.frame(names(selcGenes)[i],  paste( c(na.omit(selcGenes[[i]][match(go_candidates,  selcGenes[[i]])])) , collapse=", ") )
  }))
  names(genes_in_go) = c("GO_ID","sig_genes_in_term")
  tab = cbind(tab, genes_in_go)
  Pos_TopDEG_res = tab
  
  #All POSITIVE DEGs
  geneUniverse <- row.names(to_test_cds)
  genesOfInterest = subset(transcript_class, fc>1 & leclass=="significant")$gene
  geneList <- factor(as.integer(geneUniverse %in% genesOfInterest))
  names(geneList) <- geneUniverse
  GOdata <- new("topGOdata",
                ontology = "BP",
                allGenes=geneList,
                #geneSel = selection,
                annot = annFUN.org, 
                mapping = "org.Mm.eg.db",
                ID = "symbol")
  
  resultKS <- runTest(GOdata, algorithm = "weight01", statistic = "fisher") #Fisher exact test
  tab <- GenTable(GOdata, raw.p.value = resultKS, topNodes = length(resultKS@score), numChar = 120)
  tab = subset(tab, raw.p.value <=0.01) #keep only sig results
  go_candidates = genesOfInterest
  selcGenes <- genesInTerm(GOdata, whichGO=tab$GO.ID)
  genes_in_go = do.call(rbind, lapply(1:length(selcGenes), function(i){
    #i=1
    data.frame(names(selcGenes)[i],  paste( c(na.omit(selcGenes[[i]][match(go_candidates,  selcGenes[[i]])])) , collapse=", ") )
  }))
  names(genes_in_go) = c("GO_ID","sig_genes_in_term")
  tab = cbind(tab, genes_in_go)
  Pos_AllDEG_res = tab
  
  
  #Top 100 NEGATIVE DEGs
  geneUniverse <- row.names(to_test_cds)
  genesOfInterest = tail(transcript_class,100)$gene[1:100]
  geneList <- factor(as.integer(geneUniverse %in% genesOfInterest))
  names(geneList) <- geneUniverse
  GOdata <- new("topGOdata",
                ontology = "BP",
                allGenes=geneList,
                #geneSel = selection,
                annot = annFUN.org, 
                mapping = "org.Mm.eg.db",
                ID = "symbol")
  
  resultKS <- runTest(GOdata, algorithm = "weight01", statistic = "fisher") #Fisher exact test
  tab <- GenTable(GOdata, raw.p.value = resultKS, topNodes = length(resultKS@score), numChar = 120)
  tab = subset(tab, raw.p.value <=0.01) #keep only sig results 
  go_candidates = genesOfInterest
  selcGenes <- genesInTerm(GOdata, whichGO=tab$GO.ID)
  genes_in_go = do.call(rbind, lapply(1:length(selcGenes), function(i){
    #i=1
    data.frame(names(selcGenes)[i],  paste( c(na.omit(selcGenes[[i]][match(go_candidates,  selcGenes[[i]])])) , collapse=", ") )
  }))
  names(genes_in_go) = c("GO_ID","sig_genes_in_term")
  tab = cbind(tab, genes_in_go)
  Neg_TopDEG_res = tab
  
  #All NEGATIVE DEGs
  geneUniverse <- row.names(to_test_cds)
  genesOfInterest = subset(transcript_class, fc<1 & leclass=="significant")$gene
  geneList <- factor(as.integer(geneUniverse %in% genesOfInterest))
  names(geneList) <- geneUniverse
  GOdata <- new("topGOdata",
                ontology = "BP",
                allGenes=geneList,
                #geneSel = selection,
                annot = annFUN.org, 
                mapping = "org.Mm.eg.db",
                ID = "symbol")
  
  resultKS <- runTest(GOdata, algorithm = "weight01", statistic = "fisher") #Fisher exact test
  tab <- GenTable(GOdata, raw.p.value = resultKS, topNodes = length(resultKS@score), numChar = 120)
  tab = subset(tab, raw.p.value <=0.01) #keep only sig results
  go_candidates = genesOfInterest
  selcGenes <- genesInTerm(GOdata, whichGO=tab$GO.ID)
  genes_in_go = do.call(rbind, lapply(1:length(selcGenes), function(i){
    #i=1
    data.frame(names(selcGenes)[i],  paste( c(na.omit(selcGenes[[i]][match(go_candidates,  selcGenes[[i]])])) , collapse=", ") )
  }))
  names(genes_in_go) = c("GO_ID","sig_genes_in_term")
  tab = cbind(tab, genes_in_go)
  Neg_AllDEG_res = tab
  
  #-----------------------------------------------------------------------------------#
  #Make output dir
  output_dir = paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/DEG/lineage_sorting_deg/",tissues[l],sep="")
  dir.create(output_dir)
  
  #-----------------------------------------------------------------------------------#
  #Write out to excel workbook
  
  OUT <- createWorkbook() #create excel workbook for a sample https://stackoverflow.com/questions/27713310/easy-way-to-export-multiple-data-frame-to-multiple-excel-worksheets
  
  addWorksheet(OUT, "All DEGs") # Add some sheets to the workbook
  writeData(OUT, sheet = "All DEGs", x = transcript_class) #Write data to a shee
  
  addWorksheet(OUT, "Pos_TopDEG_res") # Add some sheets to the workbook
  writeData(OUT, sheet = "Pos_TopDEG_res", x = Pos_TopDEG_res) #Write data to a shee
  
  addWorksheet(OUT, "Pos_AllDEG_res") # Add some sheets to the workbook
  writeData(OUT, sheet = "Pos_AllDEG_res", x = Pos_AllDEG_res) #Write data to a shee
  
  addWorksheet(OUT, "Neg_TopDEG_res") # Add some sheets to the workbook
  writeData(OUT, sheet = "Neg_TopDEG_res", x = Neg_TopDEG_res) #Write data to a shee
  
  addWorksheet(OUT, "Neg_AllDEG_res") # Add some sheets to the workbook
  writeData(OUT, sheet = "Neg_AllDEG_res", x = Neg_AllDEG_res) #Write data to a shee
  
  saveWorkbook(OUT, 
               paste(output_dir,"/", tissues[l],"_Lgr6_vs_progeny_collated_results.xlsx",sep=""),
               overwrite=T)
  
  #-----------------------------------------------------------------------------------#
  #Print out volcano plot
  png(paste(output_dir,"/", tissues[l],"_Lgr6_vs_progeny_volcanoplot.png",sep=""),
      width=7,height=7,units="in",res=350)
  print(volcano_plot)
  dev.off()
  
  #-----------------------------------------------------------------------------------#
  #Print out manifold plots
  png(paste(output_dir,"/", tissues[l],"_Lgr6_vs_progeny_manifolds.png",sep=""),
      width=12,height=12,units="in",res=500)
  multiplot(pos_top100_manifold,pos_all_manifold,
            neg_top100_manifold,neg_all_manifold,
            cols=2)
  dev.off()
  
} #End tissue Lgr6 vs Progeny loop (incremented on l)

#----------------------------------------------------------------------------------------------------------------#
#----------------------------------------------------------------------------------------------------------------#
#----------------------------------------------------------------------------------------------------------------#
#----------------------------------------------------------------------------------------------------------------#
#GO anaylsis on metagenes
#----------------------------------------------------------------------------------------------------------------#

#Read in metagenes reported in the paper
eigengenes = read.xlsx(
  xlsxFile="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Nature/Round\ 2/SI\ Tables.xlsx",
  colNames=T,
  startRow = 3
)
eigengenes[1:10,1:10]

aim2probedat = read.table(file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/bulk_network_data/Aim2-4/AIM2_AIM4_BS_CARC_HFC_NOTED_ENTREZ_03_16_15_PROBE_DATA.txt",header=T)
aim2probedat[1:10,]


metagene_GOs = do.call(rbind, lapply(1:ncol(eigengenes), function(j){
  tryCatch({ #suppress error
    
    #j=1
    #Start genes of interest GO scan
    geneUniverse <- unique(aim2probedat$SYMBOL) #get all the genes on the microarray
    genesOfInterest = t(as.vector(data.frame(strsplit(eigengenes[,j]," "))[1,]) )[,1] #extract genes from metagenes sheet
    geneList <- factor(as.integer(geneUniverse %in% genesOfInterest))
    names(geneList) <- geneUniverse
    GOdata <- new("topGOdata",
                  ontology = "BP",
                  allGenes=geneList,
                  #geneSel = selection,
                  annot = annFUN.org, 
                  mapping = "org.Mm.eg.db",
                  ID = "symbol")
    
    resultKS <- runTest(GOdata, algorithm = "weight01", statistic = "fisher") #Fisher exact test
    tab <- GenTable(GOdata, raw.p.value = resultKS, topNodes = length(resultKS@score), numChar = 120)
    tab = subset(tab, raw.p.value <=0.01) #keep only sig results
    go_candidates = genesOfInterest
    selcGenes <- genesInTerm(GOdata, whichGO=tab$GO.ID)
    genes_in_go = do.call(rbind, lapply(1:length(selcGenes), function(i){
      #i=1
      data.frame(names(selcGenes)[i],  paste( c(na.omit(selcGenes[[i]][match(go_candidates,  selcGenes[[i]])])) , collapse=", ") )
    }))
    names(genes_in_go) = c("GO_ID","sig_genes_in_term")
    tab = cbind(tab, genes_in_go)
    tab = cbind(names(eigengenes)[j] , tab) #Add an initial col telling me what this gene is, then add a row of that separates 
    names(tab)[1] = "metagene"
    return(tab)
    
  },  error=function(e){})
  
}))

#------------#
#Write out to excel sheet
OUT <- createWorkbook() #create excel workbook for a sample https://stackoverflow.com/questions/27713310/easy-way-to-export-multiple-data-frame-to-multiple-excel-worksheets

addWorksheet(OUT, "Metagene GO") # Add some sheets to the workbook
writeData(OUT, sheet = "Metagene GO", x = metagene_GOs) #Write data to a shee

saveWorkbook(OUT, 
             paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Nature/Round\ 2/SI\ Tables/Metagene_GOs.xlsx",sep=""),
             overwrite=T)

#----------------------------------------------------------------------------------------------------------------#
#----------------------------------------------------------------------------------------------------------------#
#----------------------------------------------------------------------------------------------------------------#
#Impute data

#Can't impute everything at once like some goddam BOESHIT
#Try imputing within sample and saving only my genes of interest?

seed_gene = c("Lgr5",
              "Bmi1","Sox4","Lrig1",
              "Lgr6","Sox9",
              "Sox2","Prom1","Foxa1","Krt15","Krt19","Pitx1","Psca","Prom2","Klf5",
              "Cd44")

clusts = seq(1:49)

samples = unique(colData(big_cds)$sample_nom)

#imputed_genes = do.call(rbind, lapply(1:length(clusts), function(i){

imputed_genes = do.call(rbind, lapply(1:length(samples), function(i){
  
  #Try imputing on sample
  #i=1
  print(paste("i=",i,sep=""))
  #subset to a cluster
  # cds_subset = big_cds[,colData(big_cds)$cluster %in% clusts[i]] 
  
  #subset to a sample
  cds_subset = big_cds[,colData(big_cds)$sample_nom %in% samples[i]] 
  
  #Make iCellR object
  my.data <- as.data.frame(as.matrix(cds_subset@assays@data$counts))
  myUMAP <- as.data.frame(cds_subset@int_colData@listData$reducedDims@listData$UMAP) #Get UMAP embeddings
  myCond <- as.data.frame(colData(cds_subset)) #get metadata
  MyCols <- row.names(myUMAP) #get cell names
  my.obj <- make.obj(my.data) #make iCellR object
  my.obj@main.data <- my.obj@raw.data #add the datums
  my.obj@umap.data <- myUMAP #Add the UMAP coords
  my.obj@metadata <- myCond #Add the metadata
  
  #imputation
  my.obj <- run.pca(my.obj, top.rank = 500) #Run PCA betch
  my.obj <- find.dim.genes(my.obj, dims = 1:30,top.pos = 20, top.neg = 20) #HAD TO CHANGE dims to max number of cells included in the object!  dims cannot exceed nubmer of cells!
  my.obj <- run.impute(my.obj, dims = 1:10, nn = 50, data.type = "pca") #Run imputation betch, nn is improtant
  
  ?run.impute
  
  return(t(my.obj@imputed.data[seed_gene,])) #extract seed gene values and return them
  
}))

#Now graph these
head(imputed_genes)
colData(big_cds)$cell_id = row.names(colData(big_cds)) #make column of cell IDs
colData(big_cds)$cell_id <- gsub('-', '.', colData(big_cds)$cell_id) #replace hypthen with decimal in cell id's to match imputed data cell IDs (rownames in imputed data)

for(i in 1:ncol(imputed_genes)){
  tryCatch({ #suppress error
    
    #i=8
    print(i)
    #Empiric expression
    expresion_empiric = plot_cells(big_cds,
                                   genes=colnames(imputed_genes)[i],
                                   label_cell_groups=FALSE,
                                   label_leaves=F,
                                   label_branch_points=F,
                                   graph_label_size=1.5,
                                   show_trajectory_graph=F) +
      xlab("") + ylab("") + 
      ggtitle(paste(colnames(imputed_genes)[i]," empiric",sep="")) +
      theme(axis.ticks = element_blank()) +
      scale_color_viridis_c(option = "inferno",name="") +
      theme(axis.text=element_text(size=0,color='black'),
            axis.title=element_text(size=0,color='black')) + theme(plot.title = element_text(size = 15, vjust=0)) +
      theme(strip.text.x = element_text(size = 0)) +
      theme(legend.position="bottom")
    
    #Imputed expression
    #Try adding a gene to be plotted
    colData(big_cds)$imputed_gene = imputed_genes[colData(big_cds)$cell_id,i]  #add imputed data
    expresion_imputed = plot_cells(big_cds,
                                   color_cells_by="imputed_gene",
                                   label_cell_groups=FALSE,
                                   label_leaves=F,
                                   label_branch_points=F,
                                   graph_label_size=1.5,
                                   show_trajectory_graph=F,
                                   alpha=0.5) +
      xlab("") + ylab("") + 
      ggtitle(paste(colnames(imputed_genes)[i]," sample-wise imputation",sep="")) +
      theme(axis.ticks = element_blank()) +
      scale_color_viridis(option = "inferno",name="",direction=-1) +
      theme(axis.text=element_text(size=0,color='black'),
            axis.title=element_text(size=0,color='black')) + theme(plot.title = element_text(size = 15, vjust=0)) +
      theme(strip.text.x = element_text(size = 0)) +
      theme(legend.position="bottom")
    
    
    png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/imputation/seed_gene_figs/",
              colnames(imputed_genes)[i],".png",sep=""),     
        width=14,height=7,units="in",res=400)
    print(
      multiplot(expresion_empiric, expresion_imputed,
                cols=2)
    )
    dev.off()
  },  error=function(e){})
  
  
}

#-------------------------------------------------------------------------------------------------------------------------#
#-------------------------------------------------------------------------------------------------------------------------#

#Violin plot versions of Metagene expression

#Seed genes
seed_gene = c("Lgr5",
              "Bmi1","Sox4","Lrig1",
              "Lgr6","Sox9",
              "Sox2","Prom1","Foxa1","Krt15","Krt19","Pitx1","Psca","Prom2","Klf5",
              "Cd44",
              "E2f1","Tmem173", "Atm","Atr",
              "Cdkn2a","Cdkn2b","Cdkn1a",
              "Tacstd2","Ly6d","Lypd3",
              "Slc4a11","Tigit") #Now some upper spike metagenes

eigengene_size=99

for(i in 27:length(seed_gene)) {
  
  #i=1
  print(i)
  
  #Get the correct match_index
  if( seed_gene[i]=="Prom1" ) {match_index = 12747} else {
    match_index = match(seed_gene[i], aim2probedat$SYMBOL) #match seed gene to its place in the matrix
  }
  
  #---------------------------------------#
  #Tumor
  #---------------------------------------#
  if( seed_gene[i]=="Prom1" ) {match_index = 12747} else {
    match_index = match(seed_gene[i], aim2probedat$SYMBOL) #match seed gene to its place in the matrix
  }
  symbol_match = data.frame(aim2probedat$SYMBOL, tumor_rhodf[,match_index], tumor_p_df[,match_index]) #put together gene names, expression data, and p-vals
  symbol_match = symbol_match[order(-symbol_match$tumor_rhodf...match_index.),] #order from high to low gene values
  names(symbol_match) = c("gene","rho","p")
  
  #make marker df for grahing eigengene
  marker_df = data.frame(unique(alias2SymbolTable(symbol_match$gene[1:(eigengene_size+1)], species = "Mm")),paste(seed_gene[i],"Carcinoma",sep=""),paste(seed_gene[i],"Carcinoma",sep=""))
  names(marker_df) = c("gene","cell_type","module")
  marker_df = na.omit(marker_df)
  
  #-------------------------------------#
  #2. Tumor WT Violin plot
  #-------------------------------------#
  #2a. aggregate gene values to get a single expression value
  agg_mat = aggregate_gene_expression(
    big_cds,
    gene_group_df = marker_df,
    cell_group_df = NULL,
    norm_method = c("log"),
    pseudocount = 1,
    scale_agg_values = F,
    max_agg_value = NA,
    min_agg_value = NA,
    exclude.na = TRUE
  )
  
  #2b. combine aggregated values with CNV calls
  agg_gene_clindat = data.frame(c(agg_mat),colData(big_cds))
  names(agg_gene_clindat)[1] = c("eigengene_expression")
  
  #2c. compute lower and upper whiskers
  ylim1 = boxplot.stats(agg_gene_clindat$eigengene_expression)$stats[c(1, 5)]
  
  #2d. Rename cell clusteres into corrct comparisons
  agg_gene_clindat$parenchyma_cluster = agg_gene_clindat$cluster
  agg_gene_clindat$parenchyma_cluster = gsub( "\\b5\\b", "nEp", agg_gene_clindat$parenchyma_cluster,perl=TRUE)
  agg_gene_clindat$parenchyma_cluster = gsub( "\\b9\\b", "nEp", agg_gene_clindat$parenchyma_cluster,perl=TRUE)
  agg_gene_clindat$parenchyma_cluster = gsub( "\\b12\\b", "nEp", agg_gene_clindat$parenchyma_cluster,perl=TRUE)
  agg_gene_clindat$parenchyma_cluster = gsub( "\\b14\\b", "nEp", agg_gene_clindat$parenchyma_cluster,perl=TRUE)
  agg_gene_clindat$parenchyma_cluster = gsub( "\\b9\\b", "nEp", agg_gene_clindat$parenchyma_cluster,perl=TRUE)
  agg_gene_clindat$parenchyma_cluster = gsub( "\\b7\\b", "Sp", agg_gene_clindat$parenchyma_cluster,perl=TRUE)
  agg_gene_clindat$parenchyma_cluster = gsub( "\\b10\\b", "Sp", agg_gene_clindat$parenchyma_cluster,perl=TRUE)
  agg_gene_clindat$parenchyma_cluster = gsub( "\\b15\\b", "Sq", agg_gene_clindat$parenchyma_cluster,perl=TRUE)
  agg_gene_clindat$parenchyma_cluster = gsub( "\\b19\\b", "US", agg_gene_clindat$parenchyma_cluster,perl=TRUE)
  agg_gene_clindat$parenchyma_cluster = gsub( "\\b33\\b", "LS", agg_gene_clindat$parenchyma_cluster,perl=TRUE)
  
  #2e. Subset and re-ordere
  agg_gene_clindat.1 = agg_gene_clindat[agg_gene_clindat$parenchyma_cluster %in% c("nEp","Sp","Sq","US","LS"),]
  agg_gene_clindat.1$parenchyma_cluster <- factor(agg_gene_clindat.1$parenchyma_cluster,levels = c("nEp","US","LS","Sq","Sp"))
  
  #2f. 
  car_violinplot = ggplot(agg_gene_clindat.1, aes(x=parenchyma_cluster,y=eigengene_expression)) +
    stat_compare_means(label = "p.signif", method = "wilcox.test",
                       ref.group = ".all.")  +
    geom_violin(width=1, alpha=0.4) +
    geom_boxplot(outlier.size=0, width=0.3, color="black", size=0.7, alpha=0.4) +
    theme_classic() +
    #geom_bar(stat = "identity", fill="darkorange") +
    #facet_wrap(~parenchyma, scales="free", nrow=1) + 
    #geom_jitter(width = 0.1, size=0.3, alpha=0.4) +
    ggtitle(paste(seed_gene[i]," NSk metagene",sep="")) + 
    xlab("") + ylab("") + 
    theme(axis.text.x=element_text(size=15,color='black'),
          axis.text.y=element_text(size=15,color='black'),
          title=element_text(size=0),
          axis.ticks.x = element_line(size=0),
          strip.text.x = element_text(size = 0),
          legend.position="none") +
    scale_x_discrete(breaks=c("nEp","US","LS","Sq","Sp"))
  
  #Print out figs
  png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Nature/Round\ 2/Figs/Revised\ Fig/violin_plots/",
            seed_gene[i],".png",sep=""),width=3,height=6,units="in",res=300)
  print(
    car_violinplot
  )
  dev.off()
}

colData(big_cds)$cell_type

#-------------------------------------------------------------------------------------------------------------------------#
#-------------------------------------------------------------------------------------------------------------------------#

#Zoom in on the papilloma parenchyma

#Seed genes
seed_gene = c("Lgr5",
              "Bmi1","Sox4","Lrig1",
              "Lgr6","Sox9",
              "Sox2","Prom1","Foxa1","Krt15","Krt19","Pitx1","Psca","Prom2","Klf5",
              "Cd44",
              "E2f1","Tmem173", "Atm","Atr",
              "Cdkn2a","Cdkn2b","Cdkn1a",
              "Tacstd2","Ly6d","Lypd3",
              "Slc4a101","Tigit") #Now some upper spike metagenes

eigengene_size=99

for(i in 24:length(seed_gene)) {
  
  #i=17
  print(i)
  
  #Get the correct match_index
  if( seed_gene[i]=="Prom1" ) {match_index = 12747} else {
    match_index = match(seed_gene[i], aim2probedat$SYMBOL) #match seed gene to its place in the matrix
  }
  
  #---------------------------------------#
  #Normal skin
  symbol_match = data.frame(aim2probedat$SYMBOL, rhodf[,match_index], p_df[,match_index]) #put together gene names, expression data, and p-vals
  symbol_match = symbol_match[order(-symbol_match$rhodf...match_index.),] #order from high to low gene values
  names(symbol_match) = c("gene","rho","p")
  
  #make marker df for grahing eigengene
  marker_df = data.frame(unique(alias2SymbolTable(symbol_match$gene[1:(eigengene_size+1)], species = "Mm")),paste(seed_gene[i]," Normal skin",sep=""),paste(seed_gene[i]," Normal skin",sep=""))
  names(marker_df) = c("gene","cell_type","module")
  marker_df = na.omit(marker_df)
  
  #-------------------------------------#
  #1a. Normal skin WT metagene UMAP
  #-------------------------------------#
  agexp_plot.12 = plot_cells(big_cds,
                             genes=marker_df,
                             label_cell_groups=FALSE,
                             label_leaves=F,
                             label_branch_points=F,
                             graph_label_size=1.5,
                             show_trajectory_graph=F,
                             cell_size=1.2) +
    xlab("") + ylab("") + 
    theme(axis.ticks = element_blank()) +
    scale_color_viridis_c(option = "inferno",name="",) +
    theme(axis.text=element_text(size=0,color='black'),
          axis.title=element_text(size=0,color='black')) + theme(plot.title = element_text(size = 0, vjust=0)) +
    theme(strip.text.x = element_text(size = 0))  +
    annotate("text", label = paste(seed_gene[i],"\nNSk metagene",sep=""), 
             x = -5, y = 7, vjust = 0,size=10) +
    theme(legend.position=c(-0.8, -0.23),
          legend.key.size = unit(1, 'cm'),
          legend.text = element_text(size=20)) +
    xlim(-4.1,0.9) +
    ylim(-7.5, -3.8)
  
  #-------------------------------------#
  #2. Normal skin WT BarGraph
  #-------------------------------------#
  #2a. aggregate gene values to get a single expression value
  agg_mat = aggregate_gene_expression(
    big_cds,
    gene_group_df = marker_df,
    cell_group_df = NULL,
    norm_method = c("log"),
    pseudocount = 1,
    scale_agg_values = F,
    max_agg_value = NA,
    min_agg_value = NA,
    exclude.na = TRUE
  )
  
  #2b. combine aggregated values with CNV calls
  agg_gene_clindat = data.frame(c(agg_mat),colData(big_cds))
  names(agg_gene_clindat)[1] = c("eigengene_expression")
  
  #2c. compute lower and upper whiskers
  ylim1 = boxplot.stats(agg_gene_clindat$eigengene_expression)$stats[c(1, 5)]
  
  #2d. Rename cell clusteres into corrct comparisons
  agg_gene_clindat$parenchyma_cluster = agg_gene_clindat$cluster
  agg_gene_clindat$parenchyma_cluster = gsub( "\\b5\\b", "nEp", agg_gene_clindat$parenchyma_cluster,perl=TRUE)
  agg_gene_clindat$parenchyma_cluster = gsub( "\\b9\\b", "nEp", agg_gene_clindat$parenchyma_cluster,perl=TRUE)
  agg_gene_clindat$parenchyma_cluster = gsub( "\\b12\\b", "nEp", agg_gene_clindat$parenchyma_cluster,perl=TRUE)
  agg_gene_clindat$parenchyma_cluster = gsub( "\\b14\\b", "nEp", agg_gene_clindat$parenchyma_cluster,perl=TRUE)
  agg_gene_clindat$parenchyma_cluster = gsub( "\\b9\\b", "nEp", agg_gene_clindat$parenchyma_cluster,perl=TRUE)
  agg_gene_clindat$parenchyma_cluster = gsub( "\\b7\\b", "Sp", agg_gene_clindat$parenchyma_cluster,perl=TRUE)
  agg_gene_clindat$parenchyma_cluster = gsub( "\\b10\\b", "Sp", agg_gene_clindat$parenchyma_cluster,perl=TRUE)
  agg_gene_clindat$parenchyma_cluster = gsub( "\\b15\\b", "Sq", agg_gene_clindat$parenchyma_cluster,perl=TRUE)
  agg_gene_clindat$parenchyma_cluster = gsub( "\\b19\\b", "US", agg_gene_clindat$parenchyma_cluster,perl=TRUE)
  agg_gene_clindat$parenchyma_cluster = gsub( "\\b33\\b", "LS", agg_gene_clindat$parenchyma_cluster,perl=TRUE)
  
  #2e. Subset and re-ordere
  agg_gene_clindat.1 = agg_gene_clindat[agg_gene_clindat$parenchyma_cluster %in% c("nEp","Sp","Sq","US","LS"),]
  agg_gene_clindat.1$parenchyma_cluster <- factor(agg_gene_clindat.1$parenchyma_cluster,levels = c("nEp","US","LS","Sq","Sp"))
  
  my_comparisons <- list( c("US", "LS") )
  
  #2f. 
  agg_gene_clindat.1 = agg_gene_clindat.1[agg_gene_clindat.1$parenchyma_cluster %in% c("US","LS"), ] #subset to onlh LW and US
  nsk_violinplot = ggplot(agg_gene_clindat.1, aes(x=parenchyma_cluster,y=eigengene_expression)) +
    stat_compare_means(comparisons = my_comparisons)  +
    geom_violin(width=1, alpha=0.4) +
    geom_boxplot(outlier.size=0, width=0.3, color="black", size=0.7, alpha=0.4) +
    theme_classic() +
    #geom_bar(stat = "identity", fill="darkorange") +
    #facet_wrap(~parenchyma, scales="free", nrow=1) + 
    #geom_jitter(width = 0.1, size=0.3, alpha=0.4) +
    ggtitle(paste(seed_gene[i]," NSk metagene",sep="")) + 
    xlab("") + ylab("") + 
    theme(axis.text.x=element_text(size=25,color='black'),
          axis.text.y=element_text(size=25,color='black'),
          title=element_text(size=0),
          axis.ticks.x = element_line(size=0),
          strip.text.x = element_text(size = 0),
          legend.position="none") +
    scale_x_discrete(breaks=c("US","LS"))
  
  
  #---------------------------------------#
  #Tumor
  #---------------------------------------#
  if( seed_gene[i]=="Prom1" ) {match_index = 12747} else {
    match_index = match(seed_gene[i], aim2probedat$SYMBOL) #match seed gene to its place in the matrix
  }
  symbol_match = data.frame(aim2probedat$SYMBOL, tumor_rhodf[,match_index], tumor_p_df[,match_index]) #put together gene names, expression data, and p-vals
  symbol_match = symbol_match[order(-symbol_match$tumor_rhodf...match_index.),] #order from high to low gene values
  names(symbol_match) = c("gene","rho","p")
  
  #make marker df for grahing eigengene
  marker_df = data.frame(unique(alias2SymbolTable(symbol_match$gene[1:(eigengene_size+1)], species = "Mm")),paste(seed_gene[i],"Carcinoma",sep=""),paste(seed_gene[i],"Carcinoma",sep=""))
  names(marker_df) = c("gene","cell_type","module")
  marker_df = na.omit(marker_df)
  
  #all cells eigengene
  agexp_plot_tumor.12 = plot_cells(big_cds,
                                   genes=marker_df,
                                   label_cell_groups=FALSE,
                                   label_leaves=F,
                                   label_branch_points=F,
                                   graph_label_size=1.5,
                                   show_trajectory_graph=F,
                                   cell_size=1.2) +
    xlab("") + ylab("") + 
    theme(axis.ticks = element_blank()) +
    scale_color_viridis_c(option = "inferno",name="") +
    theme(axis.text=element_text(size=0,color='black'),
          axis.title=element_text(size=0,color='black')) + theme(plot.title = element_text(size = 0, vjust=0)) +
    theme(strip.text.x = element_text(size = 0))  +
    annotate("text", label = paste(seed_gene[i],"\nCar metagene",sep=""), 
             x = -5, y = 7, vjust = 0,size=10) +
    theme(legend.position=c(5, 5),
          legend.key.size = unit(1, 'cm'),
          legend.text = element_text(size=20)) +
    xlim(-4.1,0.9) +
    ylim(-7.5, -3.8)
  
  #-------------------------------------#
  #2. Tumor WT BarGraph
  #-------------------------------------#
  #2a. aggregate gene values to get a single expression value
  agg_mat = aggregate_gene_expression(
    big_cds,
    gene_group_df = marker_df,
    cell_group_df = NULL,
    norm_method = c("log"),
    pseudocount = 1,
    scale_agg_values = F,
    max_agg_value = NA,
    min_agg_value = NA,
    exclude.na = TRUE
  )
  
  #2b. combine aggregated values with CNV calls
  agg_gene_clindat = data.frame(c(agg_mat),colData(big_cds))
  names(agg_gene_clindat)[1] = c("eigengene_expression")
  
  #2c. compute lower and upper whiskers
  ylim1 = boxplot.stats(agg_gene_clindat$eigengene_expression)$stats[c(1, 5)]
  
  #2d. Rename cell clusteres into corrct comparisons
  agg_gene_clindat$parenchyma_cluster = agg_gene_clindat$cluster
  agg_gene_clindat$parenchyma_cluster = gsub( "\\b5\\b", "nEp", agg_gene_clindat$parenchyma_cluster,perl=TRUE)
  agg_gene_clindat$parenchyma_cluster = gsub( "\\b9\\b", "nEp", agg_gene_clindat$parenchyma_cluster,perl=TRUE)
  agg_gene_clindat$parenchyma_cluster = gsub( "\\b12\\b", "nEp", agg_gene_clindat$parenchyma_cluster,perl=TRUE)
  agg_gene_clindat$parenchyma_cluster = gsub( "\\b14\\b", "nEp", agg_gene_clindat$parenchyma_cluster,perl=TRUE)
  agg_gene_clindat$parenchyma_cluster = gsub( "\\b9\\b", "nEp", agg_gene_clindat$parenchyma_cluster,perl=TRUE)
  agg_gene_clindat$parenchyma_cluster = gsub( "\\b7\\b", "Sp", agg_gene_clindat$parenchyma_cluster,perl=TRUE)
  agg_gene_clindat$parenchyma_cluster = gsub( "\\b10\\b", "Sp", agg_gene_clindat$parenchyma_cluster,perl=TRUE)
  agg_gene_clindat$parenchyma_cluster = gsub( "\\b15\\b", "Sq", agg_gene_clindat$parenchyma_cluster,perl=TRUE)
  agg_gene_clindat$parenchyma_cluster = gsub( "\\b19\\b", "US", agg_gene_clindat$parenchyma_cluster,perl=TRUE)
  agg_gene_clindat$parenchyma_cluster = gsub( "\\b33\\b", "LS", agg_gene_clindat$parenchyma_cluster,perl=TRUE)
  
  #2e. Subset and re-ordere
  agg_gene_clindat.1 = agg_gene_clindat[agg_gene_clindat$parenchyma_cluster %in% c("nEp","Sp","Sq","US","LS"),]
  agg_gene_clindat.1$parenchyma_cluster <- factor(agg_gene_clindat.1$parenchyma_cluster,levels = c("nEp","US","LS","Sq","Sp"))
  
  #2f.
  agg_gene_clindat.1 = agg_gene_clindat.1[agg_gene_clindat.1$parenchyma_cluster %in% c("US","LS"), ] #subset to onlh LW and US
  car_violinplot = ggplot(agg_gene_clindat.1, aes(x=parenchyma_cluster,y=eigengene_expression)) +
    stat_compare_means(comparisons = my_comparisons)  +
    geom_violin(width=1, alpha=0.4) +
    geom_boxplot(outlier.size=0, width=0.3, color="black", size=0.7, alpha=0.4) +
    theme_classic() +
    #geom_bar(stat = "identity", fill="darkorange") +
    #facet_wrap(~parenchyma, scales="free", nrow=1) + 
    #geom_jitter(width = 0.1, size=0.3, alpha=0.4) +
    ggtitle(paste(seed_gene[i]," NSk metagene",sep="")) + 
    xlab("") + ylab("") + 
    theme(axis.text.x=element_text(size=25,color='black'),
          axis.text.y=element_text(size=25,color='black'),
          title=element_text(size=0),
          axis.ticks.x = element_line(size=0),
          strip.text.x = element_text(size = 0),
          legend.position="none") +
    scale_x_discrete(breaks=c("US","LS")) 
  
  #Print out figs
  png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Nature/Round\ 2/Figs/Revised\ Fig/violiln_plots_pap_only/",
            seed_gene[i],".png",sep=""),width=10,height=9.5,units="in",res=400)
  print(
    multiplot(agexp_plot.12, agexp_plot_tumor.12,
              nsk_violinplot,car_violinplot,
              cols=2)
  )
  dev.off()
}





#-------------------------------------------------------------------------------------------------------------------------#
#-------------------------------------------------------------------------------------------------------------------------#
#Papilloma side-by-side
#Extended data Fig. 4

#Seed genes
seed_gene = c("Lgr5","Bmi1","Sox4","Lrig1",
              "Lgr6","Sox9",
              "Sox2","Prom1","Foxa1","Krt15","Krt19","Pitx1","Psca","Prom2","Klf5","Cd44",
              "Cdkn2a", "Cdkn2b", "E2f1", "Foxm1","Atm", "Atr",
              "Tacstd2","Ly6d","Lypd3","Slc4a11","Tigit","Foxa1") #Now some upper spike metagenes

eigengene_size=99

for(i in 1:length(seed_gene)) {
  
  #i=18
  print(i)
  
  seed_plot = plot_cells(big_cds,
             genes=seed_gene[i],
             label_cell_groups=FALSE,
             label_leaves=F,
             label_branch_points=F,
             graph_label_size=1.5,
             show_trajectory_graph=F) +
    xlab("") + ylab("") + 
    theme(axis.ticks = element_blank()) +
    scale_color_viridis_c(option = "inferno",name="") +
    theme(axis.text=element_text(size=0,color='black'),
          axis.title=element_text(size=0,color='black')) + theme(plot.title = element_text(size = 0, vjust=0)) +
    theme(strip.text.x = element_text(size = 0))  +
    annotate("text", label = paste(seed_gene[i],"\nseed gene",sep=""), 
             x = -5, y = 6.3, vjust = 0,size=8) +
    theme(legend.position=c(5, 5),
          legend.key.size = unit(1, 'cm'),
          legend.text = element_text(size=20))
  
  #---------------------------------------#
  #Tumor
  #---------------------------------------#
  if( seed_gene[i]=="Prom1" ) {match_index = 12747} else {
    match_index = match(seed_gene[i], aim2probedat$SYMBOL) #match seed gene to its place in the matrix
  }
  symbol_match = data.frame(aim2probedat$SYMBOL, tumor_rhodf[,match_index], tumor_p_df[,match_index]) #put together gene names, expression data, and p-vals
  symbol_match = symbol_match[order(-symbol_match$tumor_rhodf...match_index.),] #order from high to low gene values
  names(symbol_match) = c("gene","rho","p")
  
  #make marker df for grahing eigengene
  marker_df = data.frame(unique(alias2SymbolTable(symbol_match$gene[1:(eigengene_size+1)], species = "Mm")),paste(seed_gene[i],"Carcinoma",sep=""),paste(seed_gene[i],"Carcinoma",sep=""))
  names(marker_df) = c("gene","cell_type","module")
  marker_df = na.omit(marker_df)
  
  #all cells eigengene
  car_metagene_pap_only = plot_cells(big_cds,
                                   genes=marker_df,
                                   label_cell_groups=FALSE,
                                   label_leaves=F,
                                   label_branch_points=F,
                                   graph_label_size=1.5,
                                   show_trajectory_graph=F,
                                   cell_size=1.2) +
    xlab("") + ylab("") + 
    theme(axis.ticks = element_blank()) +
    scale_color_viridis_c(option = "inferno",name="") +
    theme(axis.text=element_text(size=0,color='black'),
          axis.title=element_text(size=0,color='black')) + theme(plot.title = element_text(size = 0, vjust=0)) +
    theme(strip.text.x = element_text(size = 0))  +
    annotate("text", label = paste(seed_gene[i],"\nCar metagene",sep=""), 
             x = -2.5, y = -4.4, vjust = 0,size=8) +
    theme(legend.position=c(5, 5),
          legend.key.size = unit(1, 'cm'),
          legend.text = element_text(size=20)) +
    xlim(-4.1,0.9) +
    ylim(-7.5, -3.8)
  
  #Print out figs
  png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Nature/Round\ 2/Extended\ data/Figs/seed_gene_car_pap/",
            i, "_",seed_gene[i],".png",sep=""),width=10,height=5,units="in",res=400)
  print(
    multiplot(seed_plot,car_metagene_pap_only,
              cols=2)
  )
  dev.off()
  
  
  #Print out figs
  png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Nature/Round\ 2/Extended\ data/Figs/seed_gene_car_pap/",
            i, "_",seed_gene[i],"_seedgene.png",sep=""),width=5,height=5,units="in",res=400)
  print(
    seed_plot
  )
  dev.off()
  
  
  #Print out figs
  png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Nature/Round\ 2/Extended\ data/Figs/seed_gene_car_pap/",
            i, "_",seed_gene[i],"_carmetagene.png",sep=""),width=5,height=5,units="in",res=400)
  print(
    car_metagene_pap_only
  )
  dev.off()
  
}



#-------------------------------------------------------------------------------------------------------------------------#
#-------------------------------------------------------------------------------------------------------------------------#
#-------------------------------------------------------------------------------------------------------------------------#
#-------------------------------------------------------------------------------------------------------------------------#
#-------------------------------------------------------------------------------------------------------------------------#
#Radar plots

#Loope through and calculate meaen+var expression for metagenees in various cell groups
#Seed genes
seed_gene = c("Lgr5",
              "Bmi1","Sox4","Lrig1",
              "Lgr6","Sox9",
              "Sox2","Prom1","Foxa1","Krt15","Krt19","Pitx1","Psca","Prom2","Klf5",
              "Cd44")
eigengene_size=99

combined_sc_sums = do.call(rbind, lapply(1:length(seed_gene), function(i){
  #i=1
  print(i)
  
  #Get the correct match_index
  if( seed_gene[i]=="Prom1" ) {match_index = 12747} else {
    match_index = match(seed_gene[i], aim2probedat$SYMBOL) #match seed gene to its place in the matrix
  }
  
  #---------------------------------------#
  #Normal skin
  symbol_match = data.frame(aim2probedat$SYMBOL, rhodf[,match_index], p_df[,match_index]) #put together gene names, expression data, and p-vals
  symbol_match = symbol_match[order(-symbol_match$rhodf...match_index.),] #order from high to low gene values
  names(symbol_match) = c("gene","rho","p")
  
  #make marker df for grahing eigengene
  marker_df = data.frame(unique(alias2SymbolTable(symbol_match$gene[1:(eigengene_size+1)], species = "Mm")),paste(seed_gene[i]," Normal skin",sep=""),paste(seed_gene[i]," Normal skin",sep=""))
  names(marker_df) = c("gene","cell_type","module")
  marker_df = na.omit(marker_df)
  
  
  #-------------------------------------#
  #2. Normal skin WT BarGraph
  #-------------------------------------#
  #2a. aggregate gene values to get a single expression value
  agg_mat = aggregate_gene_expression(
    big_cds,
    gene_group_df = marker_df,
    cell_group_df = NULL,
    norm_method = c("log"),
    pseudocount = 1,
    scale_agg_values = F,
    max_agg_value = NA,
    min_agg_value = NA,
    exclude.na = TRUE
  )
  
  #2b. combine aggregated values with CNV calls
  agg_gene_clindat = data.frame(c(agg_mat),colData(big_cds))
  names(agg_gene_clindat)[1] = c("eigengene_expression")
  
  #2c. compute lower and upper whiskers
  ylim1 = boxplot.stats(agg_gene_clindat$eigengene_expression)$stats[c(1, 5)]
  
  #2d. Rename cell clusteres into corrct comparisons
  agg_gene_clindat$parenchyma_cluster = agg_gene_clindat$cluster
  agg_gene_clindat$parenchyma_cluster = gsub( "\\b5\\b", "nEp", agg_gene_clindat$parenchyma_cluster,perl=TRUE)
  agg_gene_clindat$parenchyma_cluster = gsub( "\\b9\\b", "nEp", agg_gene_clindat$parenchyma_cluster,perl=TRUE)
  agg_gene_clindat$parenchyma_cluster = gsub( "\\b12\\b", "nEp", agg_gene_clindat$parenchyma_cluster,perl=TRUE)
  agg_gene_clindat$parenchyma_cluster = gsub( "\\b14\\b", "nEp", agg_gene_clindat$parenchyma_cluster,perl=TRUE)
  agg_gene_clindat$parenchyma_cluster = gsub( "\\b9\\b", "nEp", agg_gene_clindat$parenchyma_cluster,perl=TRUE)
  agg_gene_clindat$parenchyma_cluster = gsub( "\\b7\\b", "Sp", agg_gene_clindat$parenchyma_cluster,perl=TRUE)
  agg_gene_clindat$parenchyma_cluster = gsub( "\\b10\\b", "Sp", agg_gene_clindat$parenchyma_cluster,perl=TRUE)
  agg_gene_clindat$parenchyma_cluster = gsub( "\\b15\\b", "Sq", agg_gene_clindat$parenchyma_cluster,perl=TRUE)
  agg_gene_clindat$parenchyma_cluster = gsub( "\\b19\\b", "US", agg_gene_clindat$parenchyma_cluster,perl=TRUE)
  agg_gene_clindat$parenchyma_cluster = gsub( "\\b33\\b", "LS", agg_gene_clindat$parenchyma_cluster,perl=TRUE)
  
  #2e. Subset and re-ordere
  agg_gene_clindat.1 = agg_gene_clindat[agg_gene_clindat$parenchyma_cluster %in% c("nEp","Sp","Sq","US","LS"),]
  agg_gene_clindat.1$parenchyma_cluster <- factor(agg_gene_clindat.1$parenchyma_cluster,levels = c("nEp","US","LS","Sq","Sp"))
  
  #Calculate means and var metrics for these expression metrics
  dfc_nsk = summarySE(agg_gene_clindat.1, measurevar='eigengene_expression', groupvars=c('parenchyma_cluster'))
  dfc_nsk$tissue = "NSk"
  
  #---------------------------------------#
  #Tumor
  #---------------------------------------#
  if( seed_gene[i]=="Prom1" ) {match_index = 12747} else {
    match_index = match(seed_gene[i], aim2probedat$SYMBOL) #match seed gene to its place in the matrix
  }
  symbol_match = data.frame(aim2probedat$SYMBOL, tumor_rhodf[,match_index], tumor_p_df[,match_index]) #put together gene names, expression data, and p-vals
  symbol_match = symbol_match[order(-symbol_match$tumor_rhodf...match_index.),] #order from high to low gene values
  names(symbol_match) = c("gene","rho","p")
  
  #make marker df for grahing eigengene
  marker_df = data.frame(unique(alias2SymbolTable(symbol_match$gene[1:(eigengene_size+1)], species = "Mm")),paste(seed_gene[i],"Carcinoma",sep=""),paste(seed_gene[i],"Carcinoma",sep=""))
  names(marker_df) = c("gene","cell_type","module")
  marker_df = na.omit(marker_df)
  
  #-------------------------------------#
  #2. Tumor WT BarGraph
  #-------------------------------------#
  #2a. aggregate gene values to get a single expression value
  agg_mat = aggregate_gene_expression(
    big_cds,
    gene_group_df = marker_df,
    cell_group_df = NULL,
    norm_method = c("log"),
    pseudocount = 1,
    scale_agg_values = F,
    max_agg_value = NA,
    min_agg_value = NA,
    exclude.na = TRUE
  )
  
  #2b. combine aggregated values with CNV calls
  agg_gene_clindat = data.frame(c(agg_mat),colData(big_cds))
  names(agg_gene_clindat)[1] = c("eigengene_expression")
  
  #2c. compute lower and upper whiskers
  ylim1 = boxplot.stats(agg_gene_clindat$eigengene_expression)$stats[c(1, 5)]
  
  #2d. Rename cell clusteres into corrct comparisons
  agg_gene_clindat$parenchyma_cluster = agg_gene_clindat$cluster
  agg_gene_clindat$parenchyma_cluster = gsub( "\\b5\\b", "nEp", agg_gene_clindat$parenchyma_cluster,perl=TRUE)
  agg_gene_clindat$parenchyma_cluster = gsub( "\\b9\\b", "nEp", agg_gene_clindat$parenchyma_cluster,perl=TRUE)
  agg_gene_clindat$parenchyma_cluster = gsub( "\\b12\\b", "nEp", agg_gene_clindat$parenchyma_cluster,perl=TRUE)
  agg_gene_clindat$parenchyma_cluster = gsub( "\\b14\\b", "nEp", agg_gene_clindat$parenchyma_cluster,perl=TRUE)
  agg_gene_clindat$parenchyma_cluster = gsub( "\\b9\\b", "nEp", agg_gene_clindat$parenchyma_cluster,perl=TRUE)
  agg_gene_clindat$parenchyma_cluster = gsub( "\\b7\\b", "Sp", agg_gene_clindat$parenchyma_cluster,perl=TRUE)
  agg_gene_clindat$parenchyma_cluster = gsub( "\\b10\\b", "Sp", agg_gene_clindat$parenchyma_cluster,perl=TRUE)
  agg_gene_clindat$parenchyma_cluster = gsub( "\\b15\\b", "Sq", agg_gene_clindat$parenchyma_cluster,perl=TRUE)
  agg_gene_clindat$parenchyma_cluster = gsub( "\\b19\\b", "US", agg_gene_clindat$parenchyma_cluster,perl=TRUE)
  agg_gene_clindat$parenchyma_cluster = gsub( "\\b33\\b", "LS", agg_gene_clindat$parenchyma_cluster,perl=TRUE)
  
  #2e. Subset and re-ordere
  agg_gene_clindat.1 = agg_gene_clindat[agg_gene_clindat$parenchyma_cluster %in% c("nEp","Sp","Sq","US","LS"),]
  agg_gene_clindat.1$parenchyma_cluster <- factor(agg_gene_clindat.1$parenchyma_cluster,levels = c("nEp","US","LS","Sq","Sp"))
  
  #Calculate means and var metrics for these expression metrics
  dfc_car = summarySE(agg_gene_clindat.1, measurevar='eigengene_expression', groupvars=c('parenchyma_cluster'))
  dfc_car$tissue = "Car"
  
  #---------------------------------------#
  #Combine expression summaries together
  dfc_comb = rbind(dfc_nsk, dfc_car)
  dfc_comb$gene = seed_gene[i]
  return(dfc_comb)
  
}))


us_ls_barplot = combined_sc_sums[combined_sc_sums$tissue %in% "Car" & (combined_sc_sums$parenchyma_cluster %in% c("US","LS")),]

#Calculate US - LS

diff_df = do.call(rbind, lapply(1:length(seed_gene), function(i){
  #i=1
  us_ls_dif = us_ls_barplot[us_ls_barplot$gene %in% seed_gene[i] & us_ls_barplot$parenchyma_cluster %in% "US", ]$eigengene_expression -
    us_ls_barplot[us_ls_barplot$gene %in% seed_gene[i] & us_ls_barplot$parenchyma_cluster %in% "LS", ]$eigengene_expression
  data.frame(seed_gene[i], us_ls_dif)
}))
names(diff_df)[1] = "gene" 

diff_df$gene <- factor(diff_df$gene, levels=c("Lgr5",
                                              "Bmi1","Sox4","Lrig1",
                                              "Lgr6","Sox9",
                                              "Sox2","Prom1","Foxa1","Krt15","Krt19","Pitx1","Psca","Prom2","Klf5",
                                              "Cd44"))

rel_dif_fig = ggplot(diff_df, aes(x=gene, y=us_ls_dif)) +       # Note that id is a factor. If x is numeric, there is some space between the first bar
  geom_bar(stat="identity",position="dodge", fill="wheat4") +
  #ylim(-7000,13000) +
  theme(
    panel.grid.major.y = element_blank(),
    panel.grid.minor.y = element_blank()
  ) +
  theme(panel.background = element_rect(fill = 'white', color = 'white')) +
  geom_hline(yintercept=0, linetype="solid", 
             color = "black", size=1) +
  theme(axis.ticks = element_blank()) +
  theme(axis.text=element_text(size=0,color='black'),
        axis.title=element_text(size=0,color='black')) + theme(plot.title = element_text(size = 0, vjust=0))

png("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Nature/Round\ 2/Figs/Revised\ Fig/Fig\ 3\ revised/relative_car_metagene_diffs.png",width=7,height=5,units="in",res=400)
print(
  rel_dif_fig
)
dev.off()


#Circular bar plot: https://www.data-to-viz.com/graph/circularbarplot.html 
ggplot(combined_sc_sums, aes(x=gene, y=eigengene_expression)) +       # Note that id is a factor. If x is numeric, there is some space between the first bar
  geom_bar(stat="identity", aes(fill="tissue")) +
  #ylim(-7000,13000) +
  theme_minimal() +
  theme(
    axis.text = element_blank(),
    axis.title = element_blank(),
    panel.grid = element_blank(),
    plot.margin = unit(rep(-1,4), "cm") 
  ) +
  coord_polar(start = 0) + 
  geom_text(data=label_tmp, aes(x=id, y=Value+2, label=Country ), color="black", fontface="bold",alpha=0.6, size=2.5, angle= label_tmp$angle, hjust=label_tmp$hjust, inherit.aes = FALSE ) +
  geom_text( aes(x=24, y=8000, label="Who sells more weapons?"), color="black", inherit.aes = FALSE)

#-------------------------------------------------------------------------------------------------------------------------#
#-------------------------------------------------------------------------------------------------------------------------#
#-------------------------------------------------------------------------------------------------------------------------#
#-------------------------------------------------------------------------------------------------------------------------#
#-------------------------------------------------------------------------------------------------------------------------#
#Test for significance of overlap between upregulated DEGs from cisplatin experiment and lower-spike genes

lower_spike_genes = read.delim(file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/top_markers/lower_spike_vs_all_other_cells_top_markers.txt", header=T)
cisplatin_degs.1 = read.xlsx("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/combined_projects/results/DEG/Chemo\ project\ -\ Chemo\ -\ Control\ VS\ Chemo\ project\ -\ Chemo\ -\ Cisplatin/Chemo\ project\ -\ Chemo\ -\ Control\ VS\ Chemo\ project\ -\ Chemo\ -\ Cisplatincollated_results.xlsx", 
                             1)

#Get significant upregulated DEGs expressed in at least 10% cisplatin-treated cells
cisplatin_degs = subset(cisplatin_degs.1, fc>1 & leclass=="significant" & group_b_fraction_posexp>=0.1
                        & group_a_fraction_posexp>=0.1) #ordered from high to low fold change

#Get significant upregulated DEGs expressed in at least 10% cisplatin-treated cells
lower_spike_genes = subset(lower_spike_genes, marker_test_q_value<0.05)
lower_spike_genes = lower_spike_genes[order(-lower_spike_genes$marker_score),] #Order from high to low marker score
head(lower_spike_genes)

nrow(cisplatin_degs)
nrow(lower_spike_genes)

#find all possible genes in common between DEG list and genes in big_cds
possible_genes = row.names(big_cds)

#--------------------------------------------------------------------------------------------------#
#For each gene, calculate the percentage of cells with positive expression
big_cds@assays@data$counts[1:10,1:10]

all_n = length(big_cds@assays@data$counts[i,]) #number of all cells in the cds object
gene_percent_cell_expressed = do.call(rbind, lapply(1:nrow(big_cds@assays@data$counts), function(i){
  tryCatch({ #suppress error
    #i=1
    print(i)
    df_res = as.data.frame(table(big_cds@assays@data$counts[i,]>0))
    pos_exp_n = df_res[df_res$Var1=="TRUE",]$Freq
    percent_exp_df = data.frame(row.names(big_cds)[i], round((pos_exp_n / all_n)*100,3))
    names(percent_exp_df) = c("gene","percent_expressed")
    return(percent_exp_df)
  },  error=function(e){})
}))

write.table(gene_percent_cell_expressed,
            file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/matvei/cds_objects/gene_exp_percentage_big_cds.txt",
            sep="\t", quote=F, row.names=F)

gene_percent_cell_expressed = read.table(file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/data/matvei/cds_objects/gene_exp_percentage_big_cds.txt",
                                         header=T)

head(gene_percent_cell_expressed)
#get nubmer of genes in big_cds that are expressed in at least 10% of cells
gene_percent_cell_expressed.sub = subset(gene_percent_cell_expressed, percent_expressed>=10)
nrow(gene_percent_cell_expressed)
n_gene_cell_frac = nrow(gene_percent_cell_expressed.sub)

cell_frac = nrow(gene_percent_cell_expressed.sub) / nrow(gene_percent_cell_expressed) #fraction of cells in genome that are expressed in at least 10% of cells in this expt
#36.0% of genes are expressed in at least 10% of cells
#--------------------------------------------------------------------------------------------------#

#Now loop through overlapping sizes and test for significance
all_intersection_res = do.call(rbind, lapply(2:nrow(lower_spike_genes), function(i){
  
  #i=10
  print(i)
  lower_spike_genes_sub = lower_spike_genes$gene_id[1:i]
  cisplatin_degs_sub = cisplatin_degs$gene[1:i]
  intersect_n = length(intersect(lower_spike_genes_sub, cisplatin_degs_sub)) #size of intersedtions between these 2 vectors
  
  n_A = length(lower_spike_genes_sub)
  n_B = length(cisplatin_degs_sub)
  n_C = n_gene_cell_frac #length(possible_genes)
  n_A_B = intersect_n
  
  p_value = phyper(n_A_B - 1, n_A, n_C-n_A, n_B, lower.tail = FALSE)
  
  result_df = data.frame(i, n_A, n_B, n_C, n_A_B, p_value)
  names(result_df) = c("genes_tested","n_LS_markers","n_DEGs","n_gene_pool",
                       "n_intersection","p")
  return(result_df)
  
}))

all_intersection_res$fdr =  p.adjust(all_intersection_res$p, method="BH") #adjust p-value 
write.table(all_intersection_res,
            file="/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/combined_projects/results/DEG/Chemo\ project\ -\ Chemo\ -\ Control\ VS\ Chemo\ project\ -\ Chemo\ -\ Cisplatin/LS_spike_combination/overlap_significance_lower_spike_DEGs.txt",
            sep="\t", quote=F, row.names=F)



tested_intersect_plot = ggplot(data=all_intersection_res,aes(x=genes_tested, y=n_intersection)) +
  geom_point(size=0.5) +
  geom_line() +
  theme_classic() +
  xlab("genes tested") +
  ylab("intersection size") +
  theme(axis.text=element_text(size=20,color='black'),
        axis.title=element_text(size=20,color='black'),
        legend.title=element_text(size=20), 
        legend.text=element_text(size=20))

tested_fdr_plot = ggplot(data=all_intersection_res, aes(x=n_intersection, y=-log10(fdr))) +
  geom_point(size=0.5) +
  geom_line() +
  geom_hline(yintercept=-log10(0.05), color="red") +
  geom_vline(xintercept=57, color="blue") + #this comes from looking at an infelction point
  theme_classic() +
  xlab("intersection size") +
  ylab("-log10(FDR)") +
  theme(axis.text=element_text(size=20,color='black'),
        axis.title=element_text(size=20,color='black'),
        legend.title=element_text(size=20), 
        legend.text=element_text(size=20))

multiplot(tested_intersect_plot,tested_fdr_plot,cols=1)


#--------------------------------------------------------------------------------------------------#
#--------------------------------------------------------------------------------------------------#
#--------------------------------------------------------------------------------------------------#
#--------------------------------------------------------------------------------------------------#
#--------------------------------------------------------------------------------------------------#

#Now make a big ass loop 

dummy_list = do.call(rbind, lapply(1:1000, function(j){
  
  #j=1
  random_list = sample(gene_percent_cell_expressed.sub$gene, nrow(lower_spike_genes)) #get random list
  
  all_intersection_res = do.call(rbind, lapply(2:nrow(lower_spike_genes), function(i){
    
    #i=10
    print(paste("j=",j," i=",i,sep=""))
    lower_spike_genes_sub = random_list[1:i] #REPLACE LOWER SPIKE MARKERS WITH RANDOM DUMMY LIST
    cisplatin_degs_sub = cisplatin_degs$gene[1:i]
    intersect_n = length(intersect(lower_spike_genes_sub, cisplatin_degs_sub)) #size of intersedtions between these 2 vectors
    
    n_A = length(lower_spike_genes_sub)
    n_B = length(cisplatin_degs_sub)
    n_C = n_gene_cell_frac #length(possible_genes)
    n_A_B = intersect_n
    
    p_value = phyper(n_A_B - 1, n_A, n_C-n_A, n_B, lower.tail = FALSE)
    
    result_df = data.frame(j,i, n_A, n_B, n_C, n_A_B, p_value)
    names(result_df) = c("iteration","genes_tested","n_LS_markers","n_DEGs","n_gene_pool",
                         "n_intersection","p")
    return(result_df)
    
  }))
  
  all_intersection_res
  
}))



head(dummy_list)
head(all_intersection_res) #rerun the LOWER SPIKE ONE ABOVE
ggplot() +
  geom_line(data=dummy_list,aes(x=genes_tested, y=n_intersection, group=iteration), size=0.3,alpha=0.1) +
  geom_line(data=all_intersection_res,aes(x=genes_tested, y=n_intersection), size=1, color="red") +
  theme_classic() +
  xlab("genes tested") +
  ylab("intersection size") +
  theme(axis.text=element_text(size=20,color='black'),
        axis.title=element_text(size=20,color='black'),
        legend.title=element_text(size=20), 
        legend.text=element_text(size=20))


#--------------------------------------------------------------------------------------------------#
#--------------------------------------------------------------------------------------------------#
#--------------------------------------------------------------------------------------------------#
#--------------------------------------------------------------------------------------------------#
#--------------------------------------------------------------------------------------------------#

#Now find top_markers from all cell clusters and test

match_index = 14 #This is the cluster id column index
OUT <- createWorkbook() #create excel workbook for a sample https://stackoverflow.com/questions/27713310/easy-way-to-export-multiple-data-frame-to-multiple-excel-worksheets

factor_levels = as.factor(1:49) #Put the cell clusters in order
for(j in 1:length(factor_levels)){
  tryCatch({ #suppress error
    #j=43
    print(paste("j=",j,sep=""))
    
    matching_cell_ids = row.names(colData(big_cds)[ colData(big_cds)[,match_index] %in% factor_levels[j] , ]) #get cell id's that match a factor level
    
    #Add dummy column
    colData(big_cds)$isolated_cell_population = "Non-match"
    
    #Now match isolated cell IDs with all cell IDs and replace non-match with match in dummy column
    sub_cell_matches = match(matching_cell_ids,colnames(big_cds)) #match extracted cells with all cell names
    colData(big_cds)$isolated_cell_population[sub_cell_matches] <- "Match" #now replace matching cells with new name
    
    #Now find markers of a particular cluster
    marker_test_res <- top_markers(big_cds, group_cells_by="isolated_cell_population",
                                   genes_to_test_per_group = 1000,
                                   cores=8)
    marker_test_res = marker_test_res[order(-marker_test_res$specificity),] #re-order high to low specificity
    marker_test_res.sub = subset(marker_test_res, cell_group == "Match")
    
    #Get significant upregulated DEGs expressed in at least 10% cisplatin-treated cells
    marker_test_genes = subset(marker_test_res.sub, marker_test_q_value<0.05)
    marker_test_genes = marker_test_genes[order(-marker_test_genes$marker_score),]
    
    #Now loop through overlapping sizes and test for significance
    subcluster_intersection_res = do.call(rbind, lapply(2:nrow(marker_test_genes), function(i){
      tryCatch({ #suppress error
        #i=10
        print(paste("j=",j," i=",i,sep=""))
        lower_spike_genes_sub = marker_test_genes$gene_id[1:i] #THESE ARE NO LONGER LOWER-SPIKE GENES. THEY ARE FROM THE MARKER TEST DONE RIGHT ABOVE
        cisplatin_degs_sub = cisplatin_degs$gene[1:i]
        intersect_n = length(intersect(lower_spike_genes_sub, cisplatin_degs_sub)) #size of intersedtions between these 2 vectors
        
        n_A = length(lower_spike_genes_sub)
        n_B = length(cisplatin_degs_sub)
        n_C = n_gene_cell_frac #length(possible_genes)
        n_A_B = intersect_n
        
        p_value = phyper(n_A_B - 1, n_A, n_C-n_A, n_B, lower.tail = FALSE)
        
        result_df = data.frame(i, n_A, n_B, n_C, n_A_B, p_value)
        names(result_df) = c("genes_tested","n_LS_markers","n_DEGs","n_gene_pool",
                             "n_intersection","p")
        return(result_df)
      },  error=function(e){})
    })) #end loop moving through specific markers
    
    subcluster_intersection_res$fdr =  p.adjust(subcluster_intersection_res$p, method="BH") #adjust p-value 
    
    intersection_plot = ggplot() +
      geom_line(data=dummy_list,aes(x=genes_tested, y=n_intersection, group=iteration), size=0.3,alpha=0.1) +
      geom_line(data=all_intersection_res,aes(x=genes_tested, y=n_intersection), size=1, color="red") + #lower-spike marker set
      geom_line(data=subcluster_intersection_res,aes(x=genes_tested, y=n_intersection), size=1, color="blue") + #specific cluster set
      theme_classic() +
      xlab("genes tested") +
      ylab("intersection size") +
      ggtitle(paste("Cisplatin DEG intersection with lower-spike markers (red) & \ncell cluster ",factor_levels[j]," markers (blue)",sep="")) +
      theme(axis.text=element_text(size=20,color='black'),
            axis.title=element_text(size=20,color='black'),
            legend.title=element_text(size=20), 
            legend.text=element_text(size=20),
            title=element_text(size=17,color='black'))
    
    #Add marker sheet to Excel Workbook
    addWorksheet(OUT, paste("Clust ",factor_levels[j]," markers",sep="") ) # Add some sheets to the workbook
    writeData(OUT, sheet = paste("Clust ",factor_levels[j]," markers",sep=""), x = marker_test_genes) #Write data to a shee
    
    #Add Cisplatin DEG overlap w cluster marker results to Excel Workbook
    addWorksheet(OUT, paste("Clust ",factor_levels[j]," intersxn",sep="") ) # Add some sheets to the workbook
    writeData(OUT, sheet = paste("Clust ",factor_levels[j]," intersxn",sep=""), x = subcluster_intersection_res) #Write data to a shee
    
    #Plot intersersection figure
    png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/DEG_cisplatin_marker_overlap/figs/clust_",
              factor_levels[j],"_cisplatinDEG_intersxn.png",sep=""),width=9,height=8,units="in",res=400)
    print(intersection_plot)
    dev.off()
    
  },  error=function(e){})
  
} #End factor level loop

# Export the Excel file
saveWorkbook(OUT, 
             "/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/results/DEG_cisplatin_marker_overlap/DEG_cisplatin_overlap_cellclust_markers.xlsx",
             overwrite=T)

#-------------------------------------------------------------------------------------------------------------------#
#Volcano plot: Cisplatin DEGs
#-------------------------------------------------------------------------------------------------------------------#

#Find specific genes 
highlight_genes = cisplatin_degs.1[cisplatin_degs.1$gene %in% c("Lgr5",
                                                                "Bmi1","Sox4","Lrig1",
                                                                "Lgr6","Sox9",
                                                                "Sox2","Prom1","Foxa1","Krt15","Krt19","Pitx1","Psca","Prom2","Klf5",
                                                                "Cd44",
                                                                "Tacstd1", "Tacstd2","Foxa1", 
                                                                "Lypd3", "Slc4a11", "Tigit"),]


#Plot out
volcano_plot = ggplot() +
  geom_point(data=cisplatin_degs.1, aes(log2(fc),(-log10(p_adjusted))), alpha=0.1, color="burlywood4") +
  geom_point(data=highlight_genes, aes(log2(fc),(-log10(p_adjusted))), alpha=0.9, color="darkred", size=2) +
  geom_text_repel(data=highlight_genes, aes(log2(fc),(-log10(p_adjusted)),label=gene),
                  box.padding = 0.5, max.overlaps = 25, size=6) + #add labels
  geom_hline(yintercept = (-log10(0.05)), linetype="dotted", color="black") +
  geom_vline(xintercept=0, color="black") +
  xlab("log2(FC)") +
  ylab("-log10(p)") +
  xlim(-1,1.2) +
  ylim(0,100) +
  ggtitle("Cisplatin DEGs") +
  theme(axis.text=element_text(size=15,color='black'),
        axis.title=element_text(size=15,color='black')) +
  theme(plot.title = element_text(size = 20, vjust=2)) +
  theme(strip.background = element_rect(fill = 'white', color='black', size=1)) +
  theme(panel.background=element_blank(),
        panel.grid.major=element_blank(),
        panel.grid.minor=element_blank()) +
  #scale_y_continuous(breaks=number_ticks(6)) +
  #scale_x_continuous(breaks=number_ticks(6)) +
  theme(axis.line = element_line(colour = 'black'))
volcano_plot

#----------------------------------------------------------------------------------------------#
#Violin plots of top and bottom cisplatin DEGs
#----------------------------------------------------------------------------------------------#

#Zoom in on the papilloma parenchyma

head(cisplatin_degs.1)

top_bttm_degs = data.frame(pos_degs = cisplatin_degs.1$gene[1:100], 
                           neg_degs = tail(cisplatin_degs.1$gene, 100)
)

for(i in 1:ncol(top_bttm_degs)) {
  
  print(i)
  #i=1
  #make marker df for grahing eigengene
  marker_df = data.frame(top_bttm_degs[,i],"eigengene","eigengene")
  names(marker_df) = c("gene","cell_type","module")
  marker_df = na.omit(marker_df)
  
  #-------------------------------------#
  #2. Normal skin WT BarGraph
  #-------------------------------------#
  #2a. aggregate gene values to get a single expression value
  agg_mat = aggregate_gene_expression(
    big_cds,
    gene_group_df = marker_df,
    cell_group_df = NULL,
    norm_method = c("log"),
    pseudocount = 1,
    scale_agg_values = F,
    max_agg_value = NA,
    min_agg_value = NA,
    exclude.na = TRUE
  )
  
  #2b. combine aggregated values with CNV calls
  agg_gene_clindat = data.frame(c(agg_mat),colData(big_cds))
  names(agg_gene_clindat)[1] = c("eigengene_expression")
  
  #2c. compute lower and upper whiskers
  ylim1 = boxplot.stats(agg_gene_clindat$eigengene_expression)$stats[c(1, 5)]
  
  #2d. Rename cell clusteres into corrct comparisons
  agg_gene_clindat$parenchyma_cluster = agg_gene_clindat$cluster
  agg_gene_clindat$parenchyma_cluster = gsub( "\\b5\\b", "nEp", agg_gene_clindat$parenchyma_cluster,perl=TRUE)
  agg_gene_clindat$parenchyma_cluster = gsub( "\\b9\\b", "nEp", agg_gene_clindat$parenchyma_cluster,perl=TRUE)
  agg_gene_clindat$parenchyma_cluster = gsub( "\\b12\\b", "nEp", agg_gene_clindat$parenchyma_cluster,perl=TRUE)
  agg_gene_clindat$parenchyma_cluster = gsub( "\\b14\\b", "nEp", agg_gene_clindat$parenchyma_cluster,perl=TRUE)
  agg_gene_clindat$parenchyma_cluster = gsub( "\\b9\\b", "nEp", agg_gene_clindat$parenchyma_cluster,perl=TRUE)
  agg_gene_clindat$parenchyma_cluster = gsub( "\\b19\\b", "US", agg_gene_clindat$parenchyma_cluster,perl=TRUE)
  agg_gene_clindat$parenchyma_cluster = gsub( "\\b33\\b", "LS", agg_gene_clindat$parenchyma_cluster,perl=TRUE)
  
  #2e. Subset and re-ordere
  agg_gene_clindat.1 = agg_gene_clindat[agg_gene_clindat$parenchyma_cluster %in% c("nEp","US","LS"),]
  agg_gene_clindat.1$parenchyma_cluster <- factor(agg_gene_clindat.1$parenchyma_cluster,levels = c("nEp","US","LS"))
  
  my_comparisons <- list( c("nEp", "US"),c("US", "LS"),c("nEp", "LS") )
  
  #2f. 
  agg_gene_clindat.1 = agg_gene_clindat.1[agg_gene_clindat.1$parenchyma_cluster %in% c("US","LS","nEp"), ] #subset to onlh LW and US
  deg_violinplot = ggplot(agg_gene_clindat.1, aes(x=parenchyma_cluster,y=eigengene_expression)) +
    #stat_compare_means(comparisons = my_comparisons)  + #evereything is significant
    geom_violin(width=1, alpha=0.4) +
    geom_boxplot(outlier.size=0, width=0.3, color="black", size=0.7, alpha=0.4) +
    theme_classic() +
    #geom_bar(stat = "identity", fill="darkorange") +
    #facet_wrap(~parenchyma, scales="free", nrow=1) + 
    #geom_jitter(width = 0.1, size=0.3, alpha=0.4) +
    ggtitle(paste(names(top_bttm_degs)[i]," agg exp",sep="")) + 
    xlab("") + ylab("") + 
    theme(axis.text.x=element_text(size=25,color='black'),
          axis.text.y=element_text(size=25,color='black'),
          title=element_text(size=25),
          axis.ticks.x = element_line(size=0),
          strip.text.x = element_text(size = 0),
          legend.position="none") +
    scale_x_discrete(breaks=c("nEp","US","LS"))
  
  #Print out figs
  png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Nature/Round\ 2/Figs/Revised\ Fig/violin_plots_cisplatin_DEGs/",
            names(top_bttm_degs)[i],".png",sep=""),width=3,height=5.5,units="in",res=400)
  print(
    deg_violinplot
  )
  dev.off()
}


#Papilloma UMAP plots
for(i in 1:ncol(top_bttm_degs)) {
  
  print(i)
  #i=1
  #make marker df for grahing eigengene
  marker_df = data.frame(top_bttm_degs[,i],"eigengene","eigengene")
  names(marker_df) = c("gene","cell_type","module")
  marker_df = na.omit(marker_df)

#all cells eigengene
deg_umap = plot_cells(big_cds,
                                   genes=marker_df,
                                   label_cell_groups=FALSE,
                                   label_leaves=F,
                                   label_branch_points=F,
                                   graph_label_size=1.5,
                                   show_trajectory_graph=F,
                                   cell_size=1.2) +
  xlab("") + ylab("") + 
  theme(axis.ticks = element_blank()) +
  scale_color_viridis_c(option = "inferno",name="") +
  theme(axis.text=element_text(size=0,color='black'),
        axis.title=element_text(size=0,color='black')) + theme(plot.title = element_text(size = 0, vjust=0)) +
  theme(strip.text.x = element_text(size = 0))  +
  theme(legend.position=c(5, 5),
        legend.key.size = unit(1, 'cm'),
        legend.text = element_text(size=20)) +
  xlim(-4.1,0.9) +
  ylim(-7.5, -3.8)

#Print out figs
  png(paste("/Users/marktaylor/Desktop/Projects/Balmain\ lab/mouse_scRNAseq/Network\ eigengenes\ manuscript/Nature/Round\ 2/Figs/Revised\ Fig/violin_plots_cisplatin_DEGs/",
          names(top_bttm_degs)[i],"_umap.png",sep=""),width=5,height=5,units="in",res=400)
  print(deg_umap)
  dev.off()
}





